{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eafe6834",
   "metadata": {},
   "source": [
    "# Hash table\n",
    "\n",
    "**A hash table** is a data structure that implements an associative array, also called a dictionary, which is an abstract data type that maps keys to values.\n",
    "\n",
    "We want to\n",
    "\n",
    "- add an element\n",
    "\n",
    "- delete an element\n",
    "\n",
    "- check availability\n",
    "\n",
    "faster than $O(logn)$\n",
    "\n",
    "A hash table allows you to access an element based on its corresponding key (an alternative to an index in an array). Unlike binay search trees, there is no partial order between keys.\n",
    "\n",
    "**A hash function** converts a **key** (array of $n$ elements) to an index of $m$ bits. Two or more different keys can map to the same index, so collisions must be resolved.\n",
    "\n",
    "$0 \\leq H < 2^{m}$\n",
    "\n",
    "A good hash function should be computed quickly and have a minimal number of collisions. Result indices should be evenly distributed between 0 and $2^{m}$.\n",
    "\n",
    "## Simple hash function examples:\n",
    "\n",
    "- $H(k) = k mod M$,  where M is prime\n",
    "\n",
    "   Bad case is  $M = 2^k$, the hash function will not depend on the high rank bytes of $k$.\n",
    "   \n",
    "   Example for decimal : $k mod 1000$\n",
    "\n",
    "- $H(k)= [M \\cdot \\{A \\cdot k\\}]$, where {} - fractional part, [] - rounding down\n",
    "\n",
    "    $0 < A < 1$ - real\n",
    "\n",
    "    For example: A = 0.6180339887... for good results, $\\{A \\cdot k\\}$ should be highly random\n",
    "\n",
    "    This hash function should not be calculated using float variables because of the error they introduce during multiplication:\n",
    "\n",
    "    if $M = 2^p$, where $p \\leq 32$\n",
    "\n",
    "    than \n",
    "\n",
    "    $A = \\frac{s}{2^{32}} = \\frac{2654435769}{2^{32}}$\n",
    "\n",
    "    It can be shown that \n",
    "\n",
    "     $H(k)= ((k \\cdot s) mod 2^{32}) >> (32-p)$\n",
    "     \n",
    "- For string $s = \\{s_0, s_1..., s_{n-1}\\}$:\n",
    "\n",
    "    $H(s) = (s_0 + s_1a + s_2 a^2 ... s_{n-1}a^{n-1})mod M$\n",
    "\n",
    "    or\n",
    "\n",
    "    $H(s) = (s_{n-1} + s_{n-2}a + s_{n-3} a^2 ... s_{0}a^{n-1})mod M$\n",
    "\n",
    "    All terms $(s_ia^{i}) modM$ should be different\n",
    "\n",
    "    It can be shown that a and M should be coprime.\n",
    "\n",
    "    Horner's method can be used for efficient calculation:\n",
    "\n",
    "    $H(s) = (((s_0a + s_1)a + s_2)a ... s_{n-2}) + s_{n-1})$\n",
    "    \n",
    "## Popular hash functions\n",
    "\n",
    "Cyclec redundancy check (CRC)\n",
    "\n",
    "MD1,..., MD5, MD6\n",
    "\n",
    "SHA1, SHA2\n",
    "\n",
    "\n",
    "### Load factor\n",
    "\n",
    "For a hash table of size $m$ that contains $n$ elements\n",
    "\n",
    "$\\alpha = \\frac{n}{m}$ - load factor\n",
    "\n",
    "## Separate chaining\n",
    "\n",
    "![Title](img/Separate_chaining.png)\n",
    "\n",
    "Add element -  $O(1)$, $O(n)$ if we are checking for duplicates\n",
    "\n",
    "Delete element -  $O(n)$\n",
    "\n",
    "Get element -  $O(n)$\n",
    "\n",
    "Average time (expected value) for a good hash function:\n",
    "\n",
    "$T =  O(1 + \\frac{n}{m}) = O(1 + \\alpha)$\n",
    "\n",
    "## Open addressing\n",
    "\n",
    "If cell $A[H(k)]$ is not empty, check the following cells until we find an empty one. It's much more efficient to check them out in order, depending on the $k$:\n",
    "\n",
    "$p(k, 1)$, $p(k, 2)$, ... $p(k, m-1)$ - cell indeces to check\n",
    "\n",
    "It's important that the set of cells to check is equal to the set of all cells in hash table $m$.\n",
    "\n",
    "![Title](img/open_addressing_simple.png)\n",
    "\n",
    "How to compute next $p(k, i)$?\n",
    "\n",
    "- Linear probing - clulsters  of values are generated\n",
    "\n",
    "$p(k, i) = (H(k) + i) modm$\n",
    "\n",
    "\n",
    "- Quadratic probing\n",
    "\n",
    "$p(k, i) = (H(k) +c_1 \\cdot i + c_2\\cdot i^2) modm$\n",
    "\n",
    "for example $c_1 = c_2 = 0.5$\n",
    "\n",
    "Still we are following the same chain of cells while searching for an empty cell\n",
    "\n",
    "- Double hashing\n",
    "\n",
    "$p(k, i) = (H_1(k) +i \\cdot H_2(k)) modm$\n",
    "\n",
    "Generally, $H_{1}$ and $H_{2}$ are selected from a set of universal hash functions; $H_{1}$ is selected to have a range of $\\{0,|m|-1\\}$ and $H_{2}$ to have a range of $\\{1,|m|-1\\}$.\n",
    "\n",
    "$H_2(k)$ and m should be coprime: if $m = 2^t$ than $H_2$ - odd, if m - prime: $H_2 < m$\n",
    "\n",
    "Example:\n",
    "\n",
    "$H_1(k) = k mod m_1$ and $H_2(k) = 1 + k mod m_2$\n",
    "\n",
    "## Interface\n",
    "\n",
    "Add element - $O(n)$ in worst case\n",
    "\n",
    "Delete element - we can mark an element as deleted (not empty!)\n",
    "\n",
    "Get element -  $O(n)$ in worst case\n",
    "\n",
    "$O(\\frac{1}{1-\\alpha})$ - on average\n",
    "\n",
    "## Example: k-mer index\n",
    "\n",
    "![Title](img/kmer_index.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0cf3a4a",
   "metadata": {},
   "source": [
    "# Bloom filter\n",
    "\n",
    "Probobalistic data structure to use less memory than a hash table. Using bloom filters we can check if element $a$ is in set $S$.\n",
    "\n",
    "Results: \n",
    "- No\n",
    "- Probably yes.\n",
    "\n",
    "False positives exist.\n",
    "\n",
    "Bloom filters are primarily used in bioinformatics to test the existence of a k-mer in a sequence or set of sequences.\n",
    "\n",
    "![Title](img/bloom.png)\n",
    "\n",
    "We can combine all k filters with sizes $n_1, n_2,..., n_k$ to one array of size $n = n_1+n_2+...+n_k$ bits.\n",
    "\n",
    "\n",
    "https://www.youtube.com/watch?v=9J8pRQ_EInA\n",
    "\n",
    "Probability of a false positive for k Bloom filters of length n with m items added= $p^k$, where p is a fraction of bits in k Bloom filters that are set to 1, $p = 1-(1-\\frac{1}{n})^{mk}$\n",
    "\n",
    "$(1-\\frac{1}{n})^{mk} = e^{ln((1-\\frac{1}{n})^{mk})} = e^{mkln((1-\\frac{1}{n}))}$\n",
    "\n",
    "Transform using Taylor expension $ln(1+x) = x - \\frac{x^2}{2} + \\frac{x^3}{3} - \\frac{x^4}{4} + ... $\n",
    "\n",
    "$ln(1-\\frac{1}{n}) \\approx -\\frac{1}{n}$ for large n\n",
    "\n",
    "$(1-\\frac{1}{n})^{mk} =e^{-\\frac{mk}{n}}$\n",
    "\n",
    "if we take the minimum of the function $f(k) = (1 - e^{-\\frac{mk}{n}})^k$:\n",
    "\n",
    "An optimal number of hash functions will be: $ k = \\frac{m}{n}ln2$\n",
    "\n",
    "When the number of functions is chosen optimaly ~ 50% of the bits are set, 50% of the bits are unset (p = 0.5).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bfcc74f",
   "metadata": {},
   "source": [
    "\n",
    "# Sequence Bloom Trees\n",
    "\n",
    "Task: given n databases $d_1, d_2...d_n$ - find a set of $d_i$ where sequence q is present.\n",
    "\n",
    "0) Build tree: for each dataset $d_i$ with Bloom filter $B(d_i)$\n",
    "\n",
    "Walk from the root to the leaves:\n",
    "\n",
    "- if current node u has a single child - insert $d_i$ as u’s second child\n",
    "\n",
    "- If u has two children, $B(d_i)$ is compared against the bloom filters B(left(u)) and B(right(u)) of the left left(u) and right right(u) children of u. The child with the more similar filter under the Hamming distance between the filters becomes the current node\n",
    "\n",
    "- if u has no children, a new union node v is created as a child of u’s parent. This new node has two children: u and a new node representing $d_i$.\n",
    "\n",
    "1) Query: Split q to k-mers and do search for each k-mer\n",
    "\n",
    "2) Flowing each kmer over SBT starting from the root. At each node u, the bloom filter B(u) at that node is queried for each of the kmers in Kq. If more than θ|Kq| kmers are reported to be present in B(u), the search proceeds to all of the children of u, where θ is a cutoff between 0 and 1 governing the stringency required of the match. \n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4804353/\n",
    "\n",
    "![Title](img/sbt.png)\n",
    "\n",
    "\n",
    "# Bit-sliced Genomic Signature index\n",
    "\n",
    "http://dx.doi.org/10.1038/s41587-018-0010-1\n",
    "\n",
    "![Title](https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41587-018-0010-1/MediaObjects/41587_2018_10_Fig2_HTML.png?as=webp)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6a3392f",
   "metadata": {},
   "source": [
    "## REINDEER \n",
    "\n",
    "REINDEER performs indexing of sequences and records their abundances across a collection of datasets\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7355249/\n",
    "    \n",
    "http://dx.doi.org/10.1038/s41587-018-0010-1\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
