{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n"
     ]
    }
   ],
   "source": [
    "import PyPDF2\n",
    " \n",
    "#create file object variable\n",
    "#opening method will be rb\n",
    "pdffileobj=open('PMC6746701.pdf','rb')\n",
    " \n",
    "#create reader variable that will read the pdffileobj\n",
    "pdfreader=PyPDF2.PdfReader(pdffileobj)\n",
    " \n",
    "#This will store the number of pages of this pdf file\n",
    "x=len(pdfreader.pages)\n",
    " \n",
    "print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = \"\"\n",
    "for i in range(len(pdfreader.pages)):\n",
    "    pageObj = pdfreader.pages[i]\n",
    "    output += pageObj.extract_text()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARTICLE\n",
      "A phyB-PIF1-SPA1 kinase regulatory complex\n",
      "promotes photomorphogenesis in Arabidopsis\n",
      "Inyup Paik1, Fulu Chen1,3, Vinh Ngoc Pham1, Ling Zhu1,4, Jeong-Il Kim1,2& Enamul Huq1\n",
      "CONSTITUTIVELY PHOTOMORPHOGENIC1 (COP1) is a highly conserved E3 ubiquitin ligase\n",
      "from plants to animals and acts as a central repressor of photomorphogenesis in plants.SUPPRESSOR OF PHYA-105 1 family members (SPA1-SPA4) directly interact with COP1 andenhance COP1 activity. Despite the presence of a kinase domain at the N-terminus, no COP1-independent role of SPA proteins has been reported. Here we show that SPA1 acts as aserine/threonine kinase and directly phosphorylates PIF1 in vitro and in vivo. SPAs arenecessary for the light-induced phosphorylation, ubiquitination and subsequent degradationof PIF1. Moreover, the red/far-red light photoreceptor phyB interacts with SPA1 through itsC-terminus and enhances the recruitment of PIF1 for phosphorylation. These data provide amechanistic view on how the COP1-SPA complexes serve as an example of a cognate kinase-E3 ligase complex that selectively triggers rapid phosphorylation and removal of its sub-strates, and how phyB modulates this process to promote photomorphogenesis.https://doi.org/10.1038/s41467-019-12110-y OPEN\n",
      "1Department of Molecular Biosciences and The Institute for Cellular and Molecular Biology, The University of Texas at Austin, Austin, TX 78712, USA.\n",
      "2Department of Biotechnology and Kumho Life Science Laboratory, Chonnam National University, Gwangju 61186, Korea.3Present address: MOA Key\n",
      "Laboratory of Soybean Biology (Beijing), National Key Facility of Crop Gene Resource and Genetic Improvement, Institute of Crop Sciences, Chinese\n",
      "Academy of Agricultural Sciences, 100081 Beijing, China.4Present address: Syngenta Crop Protection, LLC., Research Triangle Park, NC 27709, USA.\n",
      "Correspondence and requests for materials should be addressed to E.H. (email: huq@austin.utexas.edu )\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 11234567890():,;The ubiquitin-26S proteasome system (UPS) plays critical\n",
      "roles in controlling growth and development of all eukar-\n",
      "yotes1,2and is involved in multiple developmental and\n",
      "environmental responses, including photomorphogenesis inArabidopsis\n",
      "2,3. This system targets speci ﬁc proteins by selectively\n",
      "attaching a polyubiquitin chain by the action of three enzymes:\n",
      "E1, E2, and E3, and the polyubiquitin chain serves as a tag for\n",
      "speciﬁc recognition and degradation of the substrate by the UPS\n",
      "pathway1,2.\n",
      "CONTITUTIVE PHOTOMORPHOGENIC 1 (COP1) is a\n",
      "RING ﬁnger protein containing WD40 repeats and a coiled-coil\n",
      "domain and is highly conserved from plants to mammals4.I ti s\n",
      "one of the best characterized E3 ubiquitin ligases with broad roles\n",
      "as a central repressor of light signaling in plants to cancer biology\n",
      "in mammals5–7. COP1 interacts with multiple substrates either\n",
      "directly or indirectly using accessory proteins and targets them\n",
      "for UPS-mediated degradation. For example, in mammals, COP1\n",
      "interacts with Tribbles pseudokinases (Trib1-3), and the COP1-\n",
      "Trib complex then recruits multiple substrates for UPS-mediated\n",
      "degradation8,9. In plants, COP1 interacts with SUPPRESSOR OF\n",
      "PHYA-105 1 family members (SPA1-SPA4), and the COP1-SPA\n",
      "complexes target multiple transcription factors for rapid degra-\n",
      "dation10. Mutations in COP1 orSPA genes result in either leth-\n",
      "ality at an early seedling development stage (for cop1 null alleles)\n",
      "or severe developmental defects including dwar ﬁsm and early\n",
      "ﬂowering11–13. One characteristic phenotype of the COP1-SPA-\n",
      "related mutants is premature light-induced development even in\n",
      "the absence of light called constitutive photomorphogenic (cop)\n",
      "development11. This is because COP1-SPA complex targets\n",
      "positively acting transcription factors (e.g., HY5/LAF1/HFR1 and\n",
      "others) for degradation in darkness. In cop1 and spaQ (spa1s-\n",
      "pa2spa3spa4 quadruple) mutants, the stabilized positively acting\n",
      "transcription factors promote photomorphogenic development\n",
      "even in the dark10.\n",
      "SPA1 was originally identi ﬁed from a genetic screen to isolate a\n",
      "suppressor of a weak phytochrome A mutant phyA-10514,15.\n",
      "Since then, four SPA (SPA1 ,SPA2 ,SPA3 and SPA4 ) genes have\n",
      "been characterized with differential roles in plant develop-\n",
      "ment13,16,17. SPAs contain a ser/thr kinase domain at the N-\n",
      "terminus, a coiled-coil domain in the middle, and four WD-40\n",
      "repeats in the C-terminus that serves as an interaction domain\n",
      "with speci ﬁc substrates14. SPAs directly interact with COP1\n",
      "through their coiled-coil domain and form multiple COP1-SPA\n",
      "complexes18. Genetic evidence shows that SPAs can enhance the\n",
      "COP1 activity in vivo. For example, weak alleles of the cop1 spa1\n",
      "double mutant display strongly enhanced photomorphogenic\n",
      "phenotypes in the dark, suggesting SPA1 genetically interacts\n",
      "with COP119,20. In addition, spaQ displays a very similar phe-\n",
      "notype as a strong allele of cop1 mutant, suggesting the require-\n",
      "ment of SPAs in COP1 activity in vivo21. In vitro biochemical\n",
      "assays showed that the SPA1 coiled-coil domain strongly\n",
      "enhances the COP1 E3 ligase activity19. In addition, the COP1-\n",
      "SPA complex associates with CULLIN4, and the CUL4COP1-SPA\n",
      "promotes degradation of the positively acting transcription fac-tors in the dark to repress photomorphogenesis\n",
      "22.\n",
      "In response to environmental light signals, the red/far-red\n",
      "photoreceptors called phytochromes (phy) undergo allosteric\n",
      "changes in conformation (an inactive Pr to a biologically active\n",
      "Pfr form) and migrate into the nucleus23,24. The activated phy-\n",
      "tochromes then interact with the COP1-SPA complex and reor-\n",
      "ganizes the complex to inhibit the E3 ligase activity25,26. COP1 is\n",
      "also excluded from the nucleus in response to light27–29, and the\n",
      "reduction in COP1 in the nucleus as well as the light-induced\n",
      "inhibition of COP1 activity contribute to the accumulation of the\n",
      "positively acting transcription factors that promote photo-\n",
      "morphogenesis in the light30.In addition to the inhibition of the COP1-SPA complex, light-\n",
      "activated phytochromes also directly interact with a group of\n",
      "bHLH transcription factors called PHYTOCHROME INTER-\n",
      "ACTING FACTORs (PIFs)31,32. PIFs mainly function negatively\n",
      "in phytochrome signaling pathways. Moreover, PIFs also regulate\n",
      "a wide range of plant responses to light32,33. The red-light-\n",
      "activated phytochromes directly interact with PIFs and trigger\n",
      "rapid phosphorylation, ubiquitination and degradation of PIFs to\n",
      "release photomorphogenic development. In this process, multiple\n",
      "kinases and E3 ubiquitin ligases participate in selective removal of\n",
      "PIFs to promote photomorphogenesis32,34.\n",
      "Among all PIFs, PIF1 is unique as it is the only PIF that\n",
      "exclusively represses seed germination in the dark35,36. PIF1\n",
      "interacts with the Pfr forms of both phyA and phyB37, and\n",
      "undergoes the fastest degradation in response to red light with a\n",
      "half-life of <2 min among all PIFs38. PIF1 also directly interacts\n",
      "with the COP1-SPA complex, and both PIF1 and COP1-SPA\n",
      "complex repress photomorphogenesis in the dark in a synergistic\n",
      "manner30,39,40. However, upon red-light exposure, PIF1 is ubi-\n",
      "quitinated by the COP1-SPA complex and is rapidly degraded to\n",
      "allow seed germination to proceed41. This selective substrate\n",
      "recognition and ubiquitination by the COP1-SPA complex is\n",
      "promoted by light-induced PIF1 phosphorylation38. Casein\n",
      "Kinase 2 (CK2) was previously shown to phosphorylate PIF1\n",
      "in vitro; however, was not involved in the light-induced phos-\n",
      "phorylation of PIF142. Therefore, the protein kinase necessary for\n",
      "the rapid light-induced PIF1 phosphorylation has not been\n",
      "identi ﬁed.\n",
      "Similar to the mammalian COP1-associated pseudo-kinase,\n",
      "Trib, plant COP1-associated SPA kinases have no known sub-\n",
      "strate, although the kinase domain of both Trib and SPAs are\n",
      "necessary for their biological functions9,43. Here we provide evi-\n",
      "dence that SPA1, a component of the COP1-SPA E3 ubiquitin\n",
      "ligase complex itself, acts as a protein kinase. By performing\n",
      "extensive biochemical, genetic and genomic analyses, we provide\n",
      "strong evidence that SPA1 is a bona ﬁde serine/threonine kinase\n",
      "that is necessary for the light-induced phosphorylation of PIF1.We also provide mechanistic details on how the red/far-red light\n",
      "photoreceptor phyB directly recruits SPA1 kinase and this\n",
      "recruitment plays a pivotal role in inducing PIF1 phosphorylation\n",
      "and degradation to promote seed germination and seedling\n",
      "development. These data highlight the importance of COP1-\n",
      "associated kinases not only in enhancing COP1 activity, but also\n",
      "speciﬁc phosphorylation of their substrates for rapid ubiquitina-\n",
      "tion and subsequent degradation.\n",
      "Results\n",
      "SPA1 shows a protein kinase activity in vitro . Despite the\n",
      "presence of the majority of the signature sequences of a ser/thr\n",
      "kinase-like domain at its N-terminus\n",
      "14, no kinase activity has\n",
      "been reported for SPA proteins. To examine whether SPA1 can\n",
      "act as a kinase, we puri ﬁed the GST and strep-tagged N-terminal\n",
      "half of SPA1 (SPA1-Kin) containing the ser/thr kinase domain.\n",
      "The puri ﬁed SPA1-Kin domain displayed autophosphorylation\n",
      "activity in a protein concentration-dependent manner (Fig. 1a).\n",
      "To test whether the kinase domain is necessary for the kinase\n",
      "activity of SPA1, we created a point mutation in an amino acid\n",
      "(R517E) on the SPA1 kinase domain, which is conserved in\n",
      "SPA1 sequences from multiple plants (Supplementary Fig. 1).\n",
      "This residue is part of a conserved Glu-Arg salt bridge that\n",
      "deﬁnes eukaryotic protein kinases44, and has recently been shown\n",
      "to be critical for its biological function in the dark43. We puri ﬁed\n",
      "both the wild type and the mutant GST-SPA1-Kin proteins\n",
      "and examined their autophosphorylation activities. Results show\n",
      "a signi ﬁcantly reduced auto-phosphorylation activity for theARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "2 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsSPA1-KinR517E(mSPA1) mutant compared to the wild type\n",
      "SPA1-Kin domain (Fig. 1b), suggesting that the mutation affects\n",
      "the kinase activity of SPA1 in vitro.\n",
      "To examine whether the SPA1-Kin domain can phosphorylate\n",
      "a substrate, we performed a kinase assay using myelin basic\n",
      "protein (MBP), which has been widely used as a general substrate\n",
      "for many kinases45. In vitro kinase assays showed that the SPA1-\n",
      "Kin domain phosphorylates MBP in vitro in a concentration-\n",
      "dependent manner (Fig. 1c). Notably, the SPA1-Kin auto-\n",
      "phosphorylation activity reduces as the MBP phosphorylation\n",
      "increases, showing a general feature of a kinase-substrate\n",
      "relationship in that auto-phosphorylation of a kinase reduces as\n",
      "the substrate phosphorylation increases.\n",
      "SPA1 can directly phosphorylate PIF1 in vitro . Previously, we\n",
      "reported that the COP1-SPA1 E3 ligase can interact with PIF1 in\n",
      "a red light-induced manner and promotes its rapid degradation41.\n",
      "To test whether PIF1 could be a native substrate of the SPA1\n",
      "kinase, a strep-tagged full-length SPA1 protein was puri ﬁed from\n",
      "a eukaryotic expression host, Pichia pastoris (Supplementary\n",
      "Fig. 2a, b), and the kinase assay was performed with PIF1 as a\n",
      "substrate. PIF1 was strongly phosphorylated by SPA1 in vitro\n",
      "(Fig. 1d), suggesting that PIF1 might be a bona ﬁde native sub-\n",
      "strate of SPA1 kinase. SPA1 also phosphorylated PIF1 in an ATP\n",
      "concentration- and time-dependent manner (Fig. 1e, f). SinceSPA1 harbors a predicted ser/thr kinase domain, we further\n",
      "investigated the nature of PIF1 phosphorylation by acid-base\n",
      "sensitivity assays (Supplementary Fig. 2c). In the presence of\n",
      "KOH, PIF1 was dephosphorylated, while PIF1 phosphorylation\n",
      "was stable in HCl solution, suggesting that SPA1 is a ser/thr\n",
      "kinase for PIF1.\n",
      "To examine whether the conserved amino acid mutation\n",
      "(R517E) on the full-length SPA1 kinase domain affects PIF1\n",
      "phosphorylation, we puri ﬁed the full-length mSPA1 from Pichia\n",
      "pastoris and performed the kinase assay using PIF1 as a substrate.\n",
      "The mSPA1 showed signi ﬁcantly reduced PIF1 phosphorylation\n",
      "compared to the wild type SPA1 (Fig. 1g). In addition, the SPA1-\n",
      "Kin domain alone was able to phosphorylate PIF1, while the\n",
      "mutant mSPA1-Kin exhibited reduced PIF1 phosphorylation\n",
      "(Supplementary Fig. 2d). To examine if SPA1 can bind to ATP,\n",
      "we performed a concentration-dependent ATP binding assay\n",
      "using TNP-ATP. Results show that both wild type and mutant\n",
      "form of SPA1-Kin were able to bind to TNP-ATP in a\n",
      "concentration-dependent manner (Supplementary Fig. 2e). Both\n",
      "wild type and mutant SPA1 can directly interact with PIF1\n",
      "in vitro (Supplementary Fig. 2f, g), suggesting that the reduction\n",
      "in PIF1 phosphorylation by the mSPA1 mutant is not due to a\n",
      "reduction in physical interaction between mSPA1 and PIF1.\n",
      "Taken together, these data strongly suggest that SPA1 is a bona\n",
      "ﬁde ser/thr kinase that phosphorylates PIF1 in vitro.\n",
      "a\n",
      "bc\n",
      "SPA1-Kin\n",
      "SPA1-KinWT SPA1SPA1-Kin\n",
      "MBP\n",
      "g\n",
      "PIF1SPA1WT mSPA1PIF1SPA1\n",
      "f e\n",
      "5 30 60 90 120 (min)d\n",
      "PIF1SPA1SPA1 amount\n",
      "M\n",
      "MM++\n",
      "––+\n",
      "+\n",
      "TimeMBP amount\n",
      "SPA1-Kin\n",
      "SPA1-Kin\n",
      "PIF1SPA1\n",
      "PIF1SPA1PIF1SPA1PIF1SPA1\n",
      "100 -\n",
      "100 -150 -\n",
      "100 -\n",
      "100 -100 - 100 -kDkDkD\n",
      "kD\n",
      "kD\n",
      "kDkD\n",
      "PIF1SPA1\n",
      "PIF1SPA1mMATP 2.5 0.5 0.1 0.02 0.004mSPA1\n",
      "Fig. 1 SPA1 acts as a ser/thr protein kinase. aSPA1 kinase domain (SPA1-Kin) puri ﬁed from E. coli showed an auto-phosphorylation activity in a\n",
      "concentration-dependent manner (autoradiogram on top panel). Bottom panel shows the protein level in a Coomassie-stained gel. M, indicates protei n\n",
      "marker. bA conserved amino acid mutation on the SPA1 kinase domain reduces the auto-phosphorylation activity of SPA1 (autoradiogram on top panel).\n",
      "Bottom panel shows the protein level in a Coomassie-stained gel. cThe N-terminal kinase domain of SPA1 exhibits kinase activity in the presence of myelin\n",
      "basic protein (MBP); a general kinase substrate (autoradiogram on top panel). Bottom panel shows the protein levels in a Coomassie-stained gel. dFull-\n",
      "length SPA1 puri ﬁed from Pichia pastoris phosphorylates PIF1 in vitro (autoradiogram on top panel). Bottom panel shows the protein levels in a Coomassie-\n",
      "stained gel. eATP-dependent kinase assays of full-length SPA1 on PIF1 (autoradiogram on top panel). ATP concentrations used were 0.004, 0.02, 0.1, 0.5,\n",
      "and 2.5 mM. Bottom panel shows the protein levels in a Coomassie-stained gel. fKinetic analysis of the kinase activity of the full-length SPA1 on PIF1\n",
      "(autoradiogram on top panel). Bottom panel shows the protein levels in a Coomassie-stained gel. gA conserved amino acid mutation on the full-length\n",
      "SPA1 kinase domain reduces the SPA1 kinase activity toward PIF1 (autoradiogram on top panel). Bottom panel shows the protein levels in a Coomassie-\n",
      "stained gelNATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 3SPAs are necessary for light-induced PIF1 phosphorylation .\n",
      "Since PIF1 is rapidly phosphorylated in response to light and\n",
      "displays a characteristic mobility shift by immunoblot analy-\n",
      "sis38,41, we used spaQ plants expressing pPIF1::TAP-PIF1 to\n",
      "examine the in vivo effect of SPAs on the PIF1 phosphorylation.\n",
      "Strikingly, the rapid red light-induced phosphorylation and\n",
      "consequent band shift of TAP-PIF1 observed in wild type plants\n",
      "was completely abolished in the spaQ mutant (Fig. 2a). The PIF1\n",
      "phosphorylation status was further examined by utilizing Phos-\n",
      "tag containing SDS-PAGE gels (Fig. 2b). In wild type plants,\n",
      "TAP-PIF1 showed clear mobility shift in the presence of 15 and\n",
      "25 µM Phos-tag. However, in spaQ mutant, no mobility shift was\n",
      "observed under these conditions (Fig. 2b, top and middle panel),\n",
      "suggesting a complete absence of the light-induced phosphor-\n",
      "ylation of PIF1 in vivo. As previously shown38, the TAP-PIF1mobility shift was due to the light-induced phosphorylation of\n",
      "PIF1, since treatment with the native calf intestinal phosphatase\n",
      "(CIP) removed the shifted bands from TAP-PIF1 (Fig. 2b, bottom\n",
      "panel). Overexpression of SPA1 in wild type plants resulted in a\n",
      "faster and more robust phosphorylation of PIF1-HA under weak\n",
      "red-light conditions (Fig. 2c), suggesting an in vivo role of SPA1\n",
      "as a PIF1 kinase. Taken together, these loss-of-function and gain-\n",
      "of-function analyses suggest that SPAs are essential for the light-\n",
      "induced PIF1 phosphorylation in vivo.\n",
      "Since the light-induced PIF1 phosphorylation was largely\n",
      "absent in the spaQ mutant, we tested if the subsequent PIF1\n",
      "degradation is also defective. As expected, the PIF1 degradation\n",
      "was almost completely absent in the spaQ mutant compared to\n",
      "wild type as previously shown (Fig. 2d). In addition, the light-\n",
      "induced degradation of PIF1 was signi ﬁcantly delayed in the\n",
      "b\n",
      "TAP-PIF1\n",
      "RpTAP-PIF1/spaQ\n",
      "D 10 20 D 10 20 minRpa\n",
      "TAP-PIF1\n",
      "RpTAP-PIF1/spaQ\n",
      "D 10 20 D 10 20 minRpRegular gel 15 μM Phos-tag gel\n",
      "c\n",
      "d eTAP-PIF1TAP-PIF1\n",
      "RPT5TAP-PIF1\n",
      "RpTAP-PIF1/spaQ\n",
      "Rp\n",
      "TAP-PIF125 μM Phos-tag gel\n",
      "TAP-PIF1D – +B +CIPRpPIF1-HA\n",
      "RpPIF1-HA/TAP-SPA1\n",
      "D25 D 2 5 minRp\n",
      "PIF1-HA\n",
      "RPT5\n",
      "Col-0 spaQ\n",
      "0 5 10 0 5 10 min\n",
      "TAP-SPA1\n",
      "PIF1-HAIPed with PIF1 HA\n",
      "PelletTAP-SPA1TAP-SPA1/\n",
      "PIF1-HA\n",
      "Red D Red\n",
      "Input TAP-SPA1RPT5PIF1Rp Rp100 -\n",
      "50 -100 -kD\n",
      "50 -\n",
      "50 -75 -kD\n",
      "kD100 -\n",
      "100 -\n",
      "100 -\n",
      "50 -\n",
      "150 -kDkDD 10 20 D 10 20 min kD\n",
      "Fig. 2 SPAs are necessary for the light-induced phosphorylation and degradation of PIF1 in vivo. aImmunoblots showing the light-induced phosphorylation\n",
      "of TAP-PIF1 is defective in the spaQ mutant compared to wild type. Four-day-old dark-grown seedlings were either kept in darkness or exposed to a pulse of\n",
      "red light (300 μmolm−2) and then incubated in the dark for the duration indicated before being sampled for protein extraction. All dash lines show the dark\n",
      "position of the PIF1 band. bImmunoblots show a defect in the light-induced phosphorylation of TAP-PIF1 in spaQ background compared to wild type in gels\n",
      "containing 15 (top) and 25 (middle) µM phostag. (Bottom panel) The red light-induced slow-migrating band is a phosphorylated form of TAP-PIF1 as\n",
      "indicated by the phosphatase treatment. +CIP, native Calf Intestinal Phosphatase; +B, heat-inactivated boiled CIP. cOverexpression of SPA1 induces faster\n",
      "phosphorylation of PIF1-HA than wild type under weak red-light pulse (1 μmolm−2). Seedlings were pretreated with the proteasome inhibitor (40 μM\n",
      "bortezomib) for 5 h before being exposed to red light. Blots were probed with anti-HA and anti-RPT5 antibodies. dImmunoblots showing native PIF1 level in\n",
      "spaQ mutant compared to wild type. eImmunoblots show that TAP-SPA1 can interact with PIF1-HA in a red light-inducible manner in vivoARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "4 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsspa123 and spa124 triple mutants compared to wild type\n",
      "(Supplementary Fig. 3a), indicating that SPAs are redundantly\n",
      "functioning in light-mediated PIF1 degradation. In accordance\n",
      "with a previous report41, the poly-ubiquitination of PIF1 in\n",
      "response to light was largely reduced in the spaQ compared to\n",
      "wild type (Supplementary Fig. 3b), con ﬁrming a role of the\n",
      "COP1-SPA complex in the light-induced ubiquitination of PIF1.\n",
      "In addition, as previously shown, the in vivo interaction between\n",
      "SPA1 and PIF1 is strongly induced under red light (Fig. 2e).\n",
      "Because PIF1 is phosphorylated in response to red light, the red\n",
      "light-induced PIF1-SPA1 interaction further supports that SPA1\n",
      "might be recruited as a kinase for the light-induced PIF1\n",
      "phosphorylation.\n",
      "Since SPAs are usually present in complexes with COP1 in\n",
      "planta18, and function together with COP121, we investigated\n",
      "whether COP1 has any role in the light-induced PIF1\n",
      "phosphorylation. However, the light-induced PIF1 phosphoryla-\n",
      "tion is not defective in the cop1-4 mutant (Supplementary Fig. 3c).\n",
      "Since cop1-4 is a weak allele and still maintains some COP1\n",
      "activity12, further experiments are necessary to exclude the role of\n",
      "COP1 in PIF1 phosphorylation.\n",
      "SPA1 kinase activity is crucial for seed germination . PIF1 has\n",
      "been shown to be a master negative regulator of red light-\n",
      "mediated seed germination35. As reported previously41, the spaQ\n",
      "mutant exhibited signi ﬁcantly delayed germination in response to\n",
      "light possibly due to the increased stability of PIF1 (Fig. 3a). To\n",
      "examine the biological signi ﬁcance of the SPA1-mediated phos-\n",
      "phorylation of PIF1, we expressed both the wild type and themutant form of SPA1 as a fusion protein with Luciferase (LUC) in\n",
      "thespaQ background and selected homozygous transgenic plants\n",
      "expressing either the wild type LUC-SPA1 or the LUC-mSPA1 in\n",
      "thespaQ quadruple background (Supplementary Fig. 10c). Phe-\n",
      "notypic analyses showed that the wild type LUC-SPA1 rescued the\n",
      "seed germination phenotype to levels comparable to the wild type\n",
      "(Fig. 3a). However, the kinase mutant LUC-mSPA1 failed to\n",
      "complement the light-induced seed germination phenotype of the\n",
      "spaQ , supporting a role of the SPA1 kinase activity in regulating\n",
      "the light-induced seed germination through PIF1. At a molecular\n",
      "level, LUC-mSPA1 plants showed a delayed phosphorylation and\n",
      "subsequent degradation of PIF1 compared to the wild type LUC-\n",
      "SPA1 in the spa1spa2spa3 triple background in vivo (Fig. 3b).\n",
      "This is consistent with our biochemical data that mSPA1 has\n",
      "reduced kinase activity; hence, reducing the phosphorylation of\n",
      "PIF1 in vitro and in vivo. As previously reported43, the LUC-\n",
      "mSPA1 also largely failed to rescue the constitutive photo-\n",
      "morphogenic phenotypes of the spaQ in the dark compared to the\n",
      "wild type LUC-SPA1 (Fig. 3c). These data suggest that the SPA1\n",
      "kinase function is not only important for regulating seed germi-\n",
      "nation, but also the skotomorphogenic development of Arabi-\n",
      "dopsis seedlings.\n",
      "SPA1 interacts with the C-terminus of phyB under red light .\n",
      "The red/far-red photoreceptor phyB plays a major role in the red\n",
      "light-mediated PIF1 phosphorylation and degradation in the\n",
      "phyA background38. To investigate the functional relationship\n",
      "between phyB and SPA1, we ﬁrst performed yeast two hybrid\n",
      "assays to identify the domains necessary for their interaction. We\n",
      "a\n",
      "bcGermination frequency (×100%)\n",
      "Red light fluence ( μmolm–2)\n",
      "Hypocotyl length relative to spaQSPA1 #1/spaQ\n",
      "1.2\n",
      "1\n",
      "0.80.60.4\n",
      "0.2\n",
      "0\n",
      "1 10 30 60 120 180 1000SPA1 #2/spaQ\n",
      "mSPA1 #1/spaQ mSPA1 #2/spaQ\n",
      "spaQ Col-0\n",
      "TAP-PIF1\n",
      "RPT5spa123 /LUC-SPA1/ spa123 /LUC-mSPA1/\n",
      "TAP-PIF1\n",
      "0 5 10 0 5 10 minTAP-PIF1\n",
      "100 -\n",
      "50 -kD10.0\n",
      "7.5\n",
      "5.0\n",
      "2.5spaQ\n",
      "spaQ/LUC-\n",
      "mSPA1 #1\n",
      "spaQ/LUC-\n",
      "mSPA1 #2\n",
      "spaQ/LUC-\n",
      "SPA1 #1\n",
      "spaQ/LUC-\n",
      "SPA1 #2 Col-0\n",
      "spaQ\n",
      "spaQ/LUC-\n",
      "mSPA1 #1\n",
      "spaQ/LUC-\n",
      "mSPA1 #2\n",
      "spaQ/LUC-\n",
      "SPA1 #1\n",
      "spaQ/LUC-\n",
      "SPA1 #2 Col-0\n",
      "Fig. 3 The kinase activity of SPA1 is necessary for its biological function. aSeed germination phenotypes of wild type and mutant SPA1 expressed in the\n",
      "spaQ background in response to an increasing ﬂuence of red light. Col-0 and spaQ were used as controls. Homozygous transgenic plants expressing similar\n",
      "levels of wild type and mutant SPA1 were selected. Error bars indicate s.e.m. ( n=3).bImmunoblots showing the phosphorylation and degradation of TAP-\n",
      "PIF1 in spa123 seedlings expressing either the wild type or the mutant form of LUC-SPA1. c(top) Photograph showing the seedling phenotypes of two\n",
      "independent lines of wild type LUC-SPA1 and LUC-mSPA1 expressed in spaQ background grown in darkness for 4 days. The mutant LUC-mSPA1 failed to\n",
      "complement the short hypocotyl phenotype whereas the wild type LUC-SPA1 largely complements the phenotype. (Bottom) Dot plot shows the hypocotyl\n",
      "lengths of seedlings shown in the top panel. Scale bar =5 mm. Line =median, ( n=11)NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 5mapped the interaction domain between the coiled-coil (CC)\n",
      "domain and WD-40 repeat (WD40) in the C-terminal half of\n",
      "SPA1, and the C-terminal half of phyB (PHYB CT) using yeast-\n",
      "two-hybrid assays (Fig. 4a). Further deletion analysis showed that\n",
      "the last 80 amino acids of the C-terminal end of phyB is required\n",
      "for the interaction with full-length SPA1 (Fig. 4a). Furthermore,\n",
      "yeast two-hybrid growth assays using SPA1 and the two halves ofphyB showed that that the C-terminal half of phyB, but not the\n",
      "N-terminal half of phyB (NG) interacts with SPA1 in a light-\n",
      "independent manner, while NG interacted with PIF3 under these\n",
      "conditions (Fig. 4b). Conversely, the extreme C-terminal (C)\n",
      "domain of phyB was able to interact with SPA1 and the inter-\n",
      "action level was comparable to the full-length C-terminal PHYB\n",
      "(Fig. 4c). These data are consistent with a previous report46, and\n",
      "a\n",
      "d\n",
      "e\n",
      "f\n",
      "MBP-SPA1phyB-GFP FL\n",
      "IP: αGFPIP: αGFP\n",
      "Pellet\n",
      "α-GFPphyB-GFP /afii9797C\n",
      "D Red D Red1% Input\n",
      "TAP-SPA1\n",
      "InputphyB-GFP FL phyB-GFP /afii9797C\n",
      "α-GFPD Red D Red\n",
      "TAP-SPA1phyB FL\n",
      "phyB ΔC (1–1092)GFP\n",
      "GFPDKinase CC WD40\n",
      "521 545 696 1029Kinase CCKinase CCWD40CC WD40\n",
      "PHYB CT ΔCAD\n",
      "AD-C509\n",
      "AD-WD40\n",
      "AD-CC\n",
      "AD-N545AD-N696\n",
      "AD-SPA1\n",
      "0 5 10 15PHYB CT\n",
      "D\n",
      "g\n",
      "phyB-FL phyB- /afii9797C\n",
      "Dark Red Dark Red\n",
      "phyB-FL-GFP\n",
      "phyB- /afii9797C-GFPMBP-SPA1\n",
      "IP: αGFPh\n",
      "phyB-FL-GFP\n",
      "phyB- /afii9797C-GFPGST-PIF1phyB-FL\n",
      "Dark Red Dark RedphyB- /afii9797C\n",
      "IP: αGFPbc\n",
      "PIF3-AD\n",
      "Dark Red Dark RedSPA1-AD AD\n",
      "D153-phyB NT-\n",
      "BD\n",
      "PHYB CT-BD– LTH +\n",
      "5 mM 3-AT – LT\n",
      "Relative β-galactosidase activityCADC A\n",
      "1172 1092 988 625PAS-A PAS-B HKRD\n",
      "863 1044β-galactosidase activity (M.U)\n",
      "100 -\n",
      "100 -\n",
      "100 -150 -\n",
      "150 -\n",
      "150 -\n",
      "150 -150 -\n",
      "150 -kD\n",
      "kDkDkDB-CT : GAD625 994 1092 1172B-CB-CAPHYB CT (B-CT)\n",
      "PHYB CT ΔC (B- ΔC)\n",
      "B-CT : SPA1\n",
      "B-CA : SPA1\n",
      "B-C : SPA1\n",
      "0.0 0.1 0.2 0.3B-CA : GAD\n",
      "B-C : GADPIF3 SPA1 AD\n",
      "D153-PHYB NT-\n",
      "BD\n",
      "PHYB CT-BD\n",
      "CAARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "6 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationssuggest that the putative SPA1 interaction domain at the very C-\n",
      "terminal end of phyB is both necessary and suf ﬁcient for SPA1\n",
      "interaction in yeast. To con ﬁrm this hypothesis by an indepen-\n",
      "dent method, we cloned the full length PHYB (FL) and phyB ΔC\n",
      "with a GFP tag in a binary vector (Fig. 4d). Stable transgenic\n",
      "plants expressing similar levels of the full-length phyB-GFP FL or\n",
      "phyB-GFP ΔC were generated in the phyA-211 phyB-9 back-\n",
      "ground (Supplementary Fig. 10a). These lines were crossed with\n",
      "the tandem af ﬁnity puri ﬁcation (TAP) tagged SPA1 in phyAB\n",
      "background ( phyAB /TAP-SPA1) , and the phyB-SPA1 interaction\n",
      "was tested by a light-dependent in vivo co-immunoprecipitation\n",
      "assay as previously described26. Interestingly, the phyB-GFP FL\n",
      "robustly interacted with TAP-SPA1 in a red light-dependent\n",
      "manner in vivo, whereas the phyB-GFP ΔC displayed much\n",
      "reduced interaction with TAP-SPA1 (Fig. 4e). In addition, in a\n",
      "semi-in vitro pull-down assay, we also observed strong interac-\n",
      "tion between the phyB-GFP FL and MBP-SPA1. However, phyB-\n",
      "GFP ΔC showed much reduced interaction with MBP-SPA1\n",
      "(Fig. 4f). Because phyB can homo- and hetero-dimerize among\n",
      "other type II phytochromes47, and these in vivo phyB-GFP FL\n",
      "and phyB-GFP ΔC preparations might have both homo- and\n",
      "hetero-dimers, we performed an in vitro pull-down assay using\n",
      "phyB-GFP FL and phyB-GFP ΔC puri ﬁed from yeast cells and\n",
      "MBP-SPA1 from E. coli . The in vitro pull-down assay showed\n",
      "that phyB-GFP ΔC failed to bind to MBP-SPA1, while phyB-GFP\n",
      "FL strongly interacted with MBP-SPA1 in a light-dependent\n",
      "manner (Fig. 4g). The integrity of the phyB-GFP FL and phyB-\n",
      "GFP ΔC has been examined by the in vitro pull-down assay\n",
      "where both proteins interacted with GST-PIF1 in a light-\n",
      "dependent manner (Fig. 4h). These data collectively suggest\n",
      "that the C-terminal 80 amino acids of phyB are indispensable for\n",
      "the light-induced SPA1 interaction in Arabidopsis.\n",
      "To rule out the possibility that SPA1 might still interact with\n",
      "the N-terminal domain of phyB, we performed semi-in vivo co-\n",
      "immunoprecipitation assays using the previously described PBG\n",
      "(phyB-GFP FL ), and NG (phyB N terminal-GFP-GUS-NLS )\n",
      "transgenic plants48. Results show that PBG can interact with\n",
      "MBP-SPA1 in a light-stimulated manner (Supplementary Fig. 4),\n",
      "whereas NG failed to interact with MBP-SPA1 under these\n",
      "conditions. Overall, these data suggest that SPA1 interacts with\n",
      "phyB through the C-terminal domain.\n",
      "phyB enhances the phyB-SPA1-PIF1 complex formation\n",
      "in vitro . To address the biochemical relevance of the phyB-SPA1\n",
      "interaction, we performed an in vitro pull-down assay to test the\n",
      "phyB-SPA1-PIF1 trimolecular complex formation. We expressedGST-PIF1-strep protein in E. coli and examined its interaction\n",
      "with SPA1 in the absence or presence of an increasing amount of\n",
      "phyB. Pull-down using glutathione beads showed an interaction\n",
      "between PIF1 and SPA1 in vitro even in the absence of phyB.\n",
      "However, the PIF1-SPA1 interaction was strongly enhanced in\n",
      "the presence of the Pfr form of phyB (Fig. 5a, b). Moreover, the\n",
      "PIF1-SPA1 interaction was further enhanced when the phyB\n",
      "amount was increased to two-fold. Thus, phyB-SPA1-PIF1 can\n",
      "form a trimolecular complex in vitro.\n",
      "Both SPA1 and phyB phosphorylate PIF1 in vitro . Phyto-\n",
      "chromes displayed ser/thr kinase activity in vitro49,50. Although,\n",
      "the C-terminal half of phytochromes have sequence similarity to\n",
      "histidine kinase-related domains (HKRD), a recent study mapped\n",
      "the kinase activity to the N-terminus of Oat phyA50. SPA1 shows\n",
      "kinase activity in vitro, and both phyB and SPA1 are necessary for\n",
      "PIF1 phosphorylation and degradation in vivo. Hence, we\n",
      "examined the effect of phyB on the SPA1-mediated phosphor-ylation of PIF1 in vitro. We performed a light-dependent in vitro\n",
      "kinase assay to elucidate the molecular effect of phyB on the SPA1\n",
      "kinase activity and vice versa. Figure 5c shows that both SPA1\n",
      "and phyB independently phosphorylates PIF1 in vitro. Moreover,\n",
      "PIF1 phosphorylation was strongly enhanced in the presence of\n",
      "both SPA1 and phyB, possibly due to mutual enhancement of\n",
      "PIF1 phosphorylation by each other. However, unlike the in vivo\n",
      "data, no light-induced phosphorylation of PIF1 was observed\n",
      "in vitro. Since the intrinsic kinase activity of Arabidopsis phyB is\n",
      "still debatable, we established a method to purify functional\n",
      "phytochromes from Saccharomyces cerevisiae . By using holo-\n",
      "phytochromes, we repeated the in vitro kinase assay for both\n",
      "phyA and phyB. In this assay, both phyA and phyB phos-\n",
      "phorylated PIF1 in vitro (Supplementary Fig. 5a). The quality of\n",
      "these phytochrome preparations has been examined by zinc blot\n",
      "(Supplementary Fig. 5a), and in vitro interaction with GST-PIF1\n",
      "in a light-dependent manner (Supplementary Fig. 5b). These data\n",
      "support the hypothesis that phytochromes might function as\n",
      "protein kinases for PIFs.\n",
      "phyB-SPA1 interaction is necessary for PIF1 phosphorylation .\n",
      "To elucidate the biological relevance underlying the phyB-SPA1\n",
      "interaction, we examined PIF1 levels in both wild type PHYB-\n",
      "GFP FL andPHYB-GFP ΔCtransgenic plants. The light-induced\n",
      "degradation of PIF1 is strongly reduced in the PHYB-GFP ΔC\n",
      "lines compared to PHYB-GFP FL lines (Fig. 6a). In the PHYB-\n",
      "GFPΔClines, even after 2 hours of red light exposure, the PIF1Fig. 4 PhyB interacts with SPA1 in a light-dependent manner through its C-terminal domain. aMapping of the interaction domains for phyB and SPA1 using\n",
      "the Yeast-two-hybrid assays. Right panel shows the full-length and various domains of SPA1. Left, β-galactosidase assays show that the Coiled-Coil (CC)\n",
      "domain and WD40 repeat at the C-terminus of SPA1 is necessary for interaction with the C-terminal domain of phyB (PHYB CT, aa 625-1172). On the\n",
      "other hand, the extreme C-terminal 80 amino acids of PHYB-CT is necessary for SPA1 interaction. β-galactosidase activity was normalized by activation\n",
      "domain (AD) empty vector control. LexA-PHYB CT (aa 625-1172), LexA-PHYB CT ΔC (aa 625-1092). Error bar =SD, ( n=2).b(Top panel) Yeast-two-\n",
      "hybrid assays in the presence of chromophore phycocyanobillin (PCB). Yeast colonies were grown under red light for two days on –LTH (5 mM 3-AT)\n",
      "drop-out media containing 20 µM of PCB. PIF3 was used as a positive control. The full-length phyB (D153-PHYB) can interact with PIF3 in both dark and\n",
      "red light, while the N-terminal half of phyB (D153-PHYB NT) can only interact with PIF3 under red light. The C-terminal half of phyB also showed\n",
      "interaction with PIF3 independent of light. However, under the same condition both full-length and N-terminal half of phyB failed to show interactio n with\n",
      "SPA1, whereas the C-terminal half of phyB showed interaction with SPA1. (Bottom panel) Growth on control plate is normal for all combinations of\n",
      "constructs. cThe C-terminal 80 amino acids of phyB are suf ﬁcient for interaction with SPA1. (Top panel) Schematic diagram of various truncated\n",
      "constructs of phyB fused to BD vector. Full-length SPA1 was fused to AD vector. Error bar =SD, ( n=2).dSchematic diagram showing the structures of\n",
      "phyB-GFP FL and phyB-GFP ΔC as a fusion protein with green ﬂuorescent protein (GFP). eIn vivo co immunoprecipitation assay shows that phyB interacts\n",
      "with SPA1 in response to red light using the last 80 amino acids. phyB-SPA1 interaction is drastically reduced when the putative SPA1-interaction dom ain in\n",
      "phyB C-terminus was deleted. fSemi-in vitro pull-down assay shows that the C-terminal domain of phyB is necessary for SPA1 interaction. gIn vitro pull-\n",
      "down assay with phyB-GFP puri ﬁed from yeast and MBP-SPA1 expressed from E. coli shows light- and C-terminal-dependent interaction between the two\n",
      "proteins. hIntegrity of the puri ﬁed phyB-GFP and phyB ΔC-GFP has been shown by in vitro light-dependent interaction with GST-PIF1NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 7level remained at ~60% of the dark level (Fig. 6a, b). However, the\n",
      "PHYB-GFP FL lines showed complete PIF1 degradation in <30\n",
      "min of red light, similar to wild type (Fig. 6a, b). We also\n",
      "examined the phosphorylation status of PIF1 using Phos-tag and\n",
      "regular SDS-PAGE gels. In these assays, TAP-PIF1 displayed no\n",
      "band-shift in response to red light in the PHYB-GFP ΔCplants,\n",
      "while the wild type PHYB-GFP FL induced a band-shift of TAP-\n",
      "PIF1 under these conditions (Fig. 6c, d), indicating an absence of\n",
      "phosphorylation in the PHYB-GFP ΔCtransgenic plants. These\n",
      "data strongly suggest that the phyB-SPA1 interaction is essential\n",
      "for phyB to recruit SPA1 into a phyB-SPA1-PIF1 trimolecular\n",
      "complex and facilitate PIF1 phosphorylation and degradation in\n",
      "response to light.\n",
      "phyB C-terminal domain is essential for photobody formation .\n",
      "phyB has been shown to form nuclear photobodies in response to\n",
      "red light, and the size and number of photobodies have been\n",
      "implicated to phyB function24,51,52. To examine whether the C-\n",
      "terminal domain of phyB is necessary for the photobody for-\n",
      "mation, we grew PHYB-GFP FL and PHYB-GFP ΔCtransgenic\n",
      "seedlings in the dark and then exposed them to red light for\n",
      "various time periods. phyB-GFP FL seedlings displayed normal\n",
      "nuclear photobodies as previously described51. However, phyB-\n",
      "GFPΔCtransgenic seedlings failed to form nuclear photobodies\n",
      "even after prolonged light exposure (Fig. 6e, Supplementary\n",
      "Fig. 6). To examine if the C-terminal domain is necessary for\n",
      "phyB nuclear transport, we measured the cytosolic GFP signal in\n",
      "the hypocotyl cells of both PHYB-GFP FL and PHYB-GFP ΔC\n",
      "transgenic seedlings. Results show that the GFP signal is much\n",
      "more abundant in the cytoplasm of PHYB-GFP ΔCcompared to\n",
      "PHYB-GFP FL seedlings (Fig. 6f). These data are consistent with a\n",
      "previous report that phyB-NG-NLS which lacks the C-terminal\n",
      "half of phyB failed to form nuclear photobodies under red light.\n",
      "In contrast, the C-terminal domain of phyB is suf ﬁcient to form\n",
      "nuclear bodies even in darkness48. Thus, these results suggest that\n",
      "the interaction between phyB and SPA1 and possibly otherproteins including other SPAs through the C-terminal domain\n",
      "might be necessary for robust phyB nuclear transport and pho-\n",
      "tobody formation in response to light.\n",
      "phyB C-terminus is necessary for multiple phytochrome\n",
      "responses . To examine the biological signi ﬁcance of phyB-SPA1\n",
      "interaction, we examined a variety of phenotypes of PHYB-GFP\n",
      "FLand PHY-GFP ΔCtransgenic plants compared to wild type\n",
      "(Fig. 7a). Because of PIF1 ’s exclusive role in light-regulated ger-\n",
      "mination, we ﬁrst examined the seed germination phenotypes of\n",
      "these genotypes. As shown in Fig. 7b,PHYB-GFP ΔCseeds\n",
      "showed signi ﬁcantly delayed germination compared to PHYB-\n",
      "GFP FL , which is similar to spaQ seeds (Fig. 3a), again supporting\n",
      "the role of the SPA1-interaction domain of phyB in regulating\n",
      "seed germination in response to red light. Previous genetic\n",
      "screens have described the isolation of multiple phyB hypo-\n",
      "sensitive mutants53. Among them, phyB-28 mutant has a pre-\n",
      "mature stop codon resulting in the production of phyB with\n",
      "similar deletion (1-991) to phyB-GFP ΔC (1-1092) (Fig. 7a)54.T o\n",
      "examine whether phyB-28 shows similar germination defects as\n",
      "observed in PHYB-GFP ΔC, we generated transgenic plants\n",
      "expressing PHYB-28 in the phyA-211phyB-9 background similar\n",
      "to a wild type phyA-211 level (Supplementary Fig. 10b). Com-\n",
      "pared to wild type ( phyA-211 ),phyB-28 showed a signi ﬁcant delay\n",
      "in germination rate in response to red light (Fig. 7c), suggesting\n",
      "that the hyposensitive phenotypes of the phyB-28 allele might be\n",
      "due to a defect in the interaction with SPA proteins. Consistent\n",
      "with the above genetic analyses, the light-induced phosphoryla-\n",
      "tion and degradation of PIF1 is also defective in phyB-28 allele\n",
      "compared to phyA211 control (Fig. 7d).\n",
      "In addition to seed germination, other phytochrome-mediated\n",
      "responses are also signi ﬁcantly reduced in the phyB-GFP ΔCand\n",
      "phyB-28 mutants compared to wild type phyB-GFP FL . Hypocotyl\n",
      "lengths under different ﬂuences of red light were measured for\n",
      "phyB-GFP FL and phyB-GFP ΔCas well as phyB-28 expressing\n",
      "plants. Results show that the hypocotyl lengths of phyB-GFP ΔC\n",
      "a\n",
      "phyB-FLMBP-SPA1\n",
      "GST-PIF1GST Pull down Input (5%)c\n",
      "PIF1-GSTphyB-strepPIF1-GST\n",
      "SPA1-strep\n",
      "phyB+\n",
      "RedphyB++\n",
      "RedphyB–10\n",
      "86\n",
      "42\n",
      "0Fold enrichmentbphyB-FLMBP-SPA1\n",
      "GST-PIF1+ ++R ––\n",
      "+R++ ++ +\n",
      "++ ++–+ DSPA1\n",
      "phyB\n",
      "PIF1 PIF1 PIF1 PIF1 PIF1 PIF1D Red D RedSPA1\n",
      "phyB\n",
      "150 -\n",
      "150 -\n",
      "100 -150 -\n",
      "100 -kD\n",
      "kD\n",
      "Fig. 5 PhyB interaction with SPA1 promotes phyB-SPA1-PIF1 trimolecular complex formation and enhanced phosphorylation of PIF1 in vitro. aIn vitro pull-\n",
      "down assay shows that phyB enhances PIF1-SPA1 interaction in a light- and concentration-dependent manner. bBar graph shows the enhanced interaction\n",
      "between SPA1 and PIF1 in the presence of phyB. Error bar =SD, ( n=3).c(top panel) Autoradiogram shows that phyB enhances SPA1-mediated\n",
      "phosphorylation of PIF1 in vitro. (Bottom panel) Coomassie-stained gel shows the amount of various proteins used in the assayARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "8 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsandphyB-28 are signi ﬁcantly longer compared to controls under\n",
      "various ﬂuences of red light (Supplementary Fig. 7a, b, c).\n",
      "Measurement of the petiole lengths using 3-week-old plants show\n",
      "that phyB-GFP ΔChas signi ﬁcantly longer petiole lengths\n",
      "compared to those of phyB-GFP FL (Supplementary Fig. 7d),\n",
      "suggesting a role of the phyB C-terminus in multiple phyto-\n",
      "chrome responses. Consistent with our results, the previously\n",
      "reported NGand CGseedlings showed more stable PIF1 under\n",
      "red light than PBG (Supplementary Fig. 7e); although, both NG\n",
      "and CG induced a slight degradation of PIF1 compared to\n",
      "phyAphyB . Recently, CGhas been shown to induce degradation of\n",
      "PIFs55. Because NG undergoes rapid degradation compared to\n",
      "PBG56, and NG and the isolated N-terminal photosensory\n",
      "domain can interact with PIFs57,58, it is possible that NG\n",
      "undergoes co-degradation with PIF1 through PPK kinases or\n",
      "additional unknown kinases59. Taken together, by analyzing\n",
      "multiple phyB mutants, these data strongly support thehypothesis that the extreme C-terminal domain of phyB is\n",
      "necessary for many phyB-mediated responses during\n",
      "photomorphogenesis.\n",
      "Several lines of evidence show that the C-terminal end of phyB\n",
      "is speci ﬁcally necessary for SPA1 interaction and PIF1 degrada-\n",
      "tion, as opposed to a defect in interaction with PIF1 and/or lack\n",
      "of dimerization. As shown in Supplementary Fig. 8a, PIF1\n",
      "interacts with both phyB-GFP FL and phyB-GFP ΔC with a\n",
      "similar strength in vivo. Second, an in vitro native PAGE gel\n",
      "shows that phyB-GFP ΔC can dimerize in a similar manner to the\n",
      "phyB-GFP FL protein (Supplementary Fig. 8b). However, theisolated phyB CT ΔC failed to dimerize in yeast assays, despite\n",
      "similar dimerization activity of the phyB CT ΔC with phyB CT in\n",
      "yeast two-hybrid assays (Supplementary Fig. 8c), suggesting that\n",
      "the dimerization might be defective in the isolated phyB CT ΔC,\n",
      "but not in the phyB-GFP ΔC context. Third, chromatin\n",
      "immunoprecipitation (ChIP) assay showed that phyB-GFP FL,\n",
      "aCol-0 phyAB phyAB /phyB FL\n",
      "#31\n",
      "phyAB /phyB FL\n",
      "#24phyAB /phyB ΔC\n",
      "#44phyAB /phyB ΔC\n",
      "#27PIF1\n",
      "RPT50 0.5 1 2 0 0.5 1 2 0 0.5 1 2 (h)\n",
      "(h) 0 0.5 1 2 0 0.5 1 2 0 0.5 1 2b\n",
      "PIF1 level to dark\n",
      "00.20.40.60.811.2\n",
      "Dark 0.5 1 2Col-0\n",
      "phyAB\n",
      "phyB FL #31\n",
      "phyB FL #24\n",
      "phyB /afii9797C #44\n",
      "phyB /afii9797C #27\n",
      "phyAB /phyB FL phyAB /phyB ΔCc\n",
      "dphyAB /phyB FL phyAB /phyB ΔC\n",
      "0 10 20 0 10 20\n",
      "TAP-PIF1\n",
      "60\n",
      "40\n",
      "20\n",
      "0\n",
      "phyB-GFP FL phyB-GFP ΔCRegular gelTime (h)\n",
      "phyB-GFP /afii9797C phyB-GFP FL e\n",
      "GFP Brightfield GFP Brightfield\n",
      "GFP signal intensityf- 50\n",
      "- 50\n",
      "- 50\n",
      "- 50100 -\n",
      "100 -kD\n",
      "kD\n",
      "kDkD\n",
      "Phostag gel\n",
      "TAP-\n",
      "PIF10 10 20 30 0 10 20 30 (min)(min)\n",
      "PIF1\n",
      "RPT5\n",
      "Fig. 6 The C-terminal SPA1-interaction domain of phyB is essential for photobody formation and PIF1 phosphorylation. aImmunoblots showing the level of\n",
      "native PIF1 in two independent transgenic lines of phyB-GFP FL (#31 and #24) and phyB-GFP ΔC (#44 and #27) along with wild type and phyAB as\n",
      "controls. Total protein was extracted from 4-day-old dark-grown seedlings either kept in darkness or exposed to continuous red light (3.5 μmolm−2s−1)\n",
      "over time. bA graph showing the amount of PIF1 levels in response to red light exposure over time. c,dThe light-induced phosphorylation of PIF1 is\n",
      "defective in phyB-GFP ΔC compared to phyB-GFP FL. eThe phyB-SPA1 interaction is necessary for phyB nuclear transport and photobody formation. The\n",
      "phyB-GFP FL produces nuclear photobodies after 5 h of red light (10 µmolm−2s−1) treatment (left), whereas phyB-GFP ΔC showed dispersed nuclear\n",
      "signals (right). Bar =15μm.fThe cytoplasmic GFP signals were quanti ﬁed from phyB-GFP FL and phyB-GFP ΔC images by imageJ ( n=12), Error bar =SD.\n",
      "***p-value < 0.00001, student ’st-testNATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 9as well as phyB-GFP ΔC equally sequester PIF1 from the G-box\n",
      "on the PIL1 promoter in response to red light (Supplementary\n",
      "Fig. 8d). Fourth, we introduced mutations in the four conserved\n",
      "amino acids in the HKRD domain to test whether the possible\n",
      "HKRD-related activity was affected in phyB-GFP ΔC (Supple-\n",
      "mentary Fig. 8e), and made homozygous transgenic plants\n",
      "expressing the wild type phyB-GFP FL and phyB-GFP 4 M in\n",
      "thephyA211-phyB-9 background. Immunoblots show that the red\n",
      "light-induced PIF1 degradation was not affected in the phyB-GFP\n",
      "4Mcompared to phyB-GFP FL (Supplementary Fig. 8f), suggest-\n",
      "ing that the residues within the HKRD-related domain are not\n",
      "necessary for PIF1 degradation. Finally, we puri ﬁed phyB-GFP\n",
      "FL, phyB-GFP ΔC, and phyB-GFP 4 M proteins from Sacchar-\n",
      "omyces cerevisiae and con ﬁrmed no change in the intrinsic phyB\n",
      "kinase activities in these mutants (Supplementary Fig. 8g).\n",
      "Because phyB-GFP ΔC failed to form nuclear photobodies under\n",
      "red light and also defective in SPA1 interaction, the light-induced\n",
      "degradation of PIF1 might be promoted by either phyB-SPA1\n",
      "interaction and/or phyB photobody formation in Arabidopsis.\n",
      "SPAs are necessary for PIF-target gene regulation . To under-\n",
      "stand the molecular basis of SPA- and phyB-mediated regulation\n",
      "of PIF1 stability, we also performed RNA-seq analysis for spaQ ,\n",
      "cop1-4 and Col-0 using 3-day-old dark-grown seedlings and dark-\n",
      "grown seedlings exposed to red light for 1 h. In wild type Col-0,\n",
      "2702 genes are differentially expressed between dark and 1 h of\n",
      "red light treated samples (Fig. 8a), which are categorized as red\n",
      "light responsive genes in the wild type. Red light responsive genes\n",
      "between 1 h of red light treatment compared to dark in cop1-4andspaQ are also presented in Supplemental Data 1. In spaQ and\n",
      "cop1-4 mutants, however, 1861 genes are no longer responding to\n",
      "the red-light treatment, suggesting the importance of COP1-SPA\n",
      "complex in regulating red light-responsive gene expression\n",
      "(Supplemental Data 2). Remarkably, the spaQ mutant showed\n",
      "854 genes that do not overlap with cop1-4 differentially expressed\n",
      "genes (Fig. 8a, Supplemental Data 3). The existence of these\n",
      "exclusively SPA-dependent genes supports a COP1-independent\n",
      "function of SPA proteins.\n",
      "We also performed RNA-seq analysis for phyB-GFP FL and\n",
      "phyB-GFP ΔCunder the same conditions. A total of 1117 genes\n",
      "were differentially expressed in phyB-GFP FL in response to red\n",
      "light within 1 h (Fig. 8b, Supplemental Data 1, 4). Among them,\n",
      "561 genes were misregulated in phyB-GFP ΔCmutant that did not\n",
      "display light responsiveness (Supplemental Data 5). Thus, we\n",
      "categorized these genes into phyB C-terminal-dependent red-\n",
      "responsive genes (Fig. 8b). By comparing the differential gene\n",
      "expression in Col-0 Red/Dark and spaQ Red/Dark (Fig. 8a), 2199\n",
      "genes no longer responding to red light treatment in spaQ and are\n",
      "deﬁned as SPA-dependent red responsive genes. Because we\n",
      "identi ﬁed the phyB C-terminus as a SPA1 interaction domain\n",
      "(Fig. 4), we compared the list of phyB C-terminal-dependent red-\n",
      "responsive genes with the SPA-dependent red-responsive genes.\n",
      "These analyses showed that ~59% (333/561) of the phyB\n",
      "C-terminal-dependent genes also overlap with SPA-dependent\n",
      "genes (Fig. 8c, Supplemental Data 6), suggesting a common\n",
      "regulatory pathway shared by phyB and SPA proteins. Red light\n",
      "responsive spaQ -regulated genes were identi ﬁed as differentially\n",
      "expressed genes between spaQ in red light compared to Col-0 in\n",
      "b a\n",
      "cd\n",
      "Germination frequency (percent) Red light fluence ( μmolm–2)020406080100120phyB FL #31\n",
      "phyB FL #24\n",
      "phyB /afii9797C #44\n",
      "phyB /afii9797C #27Germination frequency (percent)\n",
      "Red light fluence ( μmolm–2)phyA/TAP-PIF1 phyA/phyB-28/TAP-PIF1\n",
      "TAP-PIF1\n",
      "RPT5020406080100120\n",
      "D 10 100 1000phyA-211\n",
      "phyA-211/phyB-28phyA phyA/phyB-28D 2.4 9.6 24\n",
      "0 0.5 1 2 0 0.5 1 2 (h)0 5 10 0 5 10 (min)\n",
      "TAP-PIF1\n",
      "RPT5phyB FL\n",
      "phyB ΔC (1–1092)GFP\n",
      "GFPD\n",
      "DCA\n",
      "1092 625PAS-A PAS-B HKRD\n",
      "863phyB-28 (1–991) GFP D\n",
      "991 1172\n",
      "100 -\n",
      "50 -\n",
      "100 -\n",
      "50 -kDCA\n",
      "Fig. 7 Direct interaction between phyB and SPA1 is necessary for phyB-mediated seed germination. aSchematic diagram of phyB-28, phyB-GFP ΔC and\n",
      "phyB-GFP FL proteins. b,cThe C-terminal SPA1-interaction domain of phyB is necessary for phyB-mediated seed germination. phyB-GFP ΔCdisplays\n",
      "reduced germination compared to phyB-GFP FL lines, while phyB-28 displays slower germination compared to wild type ( phyA-211 that contains wild type\n",
      "PHYB ). The error bars indicate s.e.m. ( n=3).dThe light-induced phosphorylation and degradation of TAP-PIF1 is defective in phyB-28 compared to wild\n",
      "type ( phyA-211 that contains wild type PHYB ). Total protein was extracted from four-day-old dark-grown seedlings either kept in darkness or exposed to\n",
      "continuous red light (3.5 μmolm−2s−1) over timeARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "10 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsred light ( spaQ Red/ Col-0 Red) (7191 genes). PIF-mis-regulated\n",
      "genes (2025 genes) in darkness were previously described60.W e\n",
      "found ~54% (1086/2025) of the PIF-mis-regulated genes in\n",
      "darkness overlapped with the spaQ -regulated genes in response to\n",
      "red light (Fig. 8d, Supplemental Data 7). We generated a heatmap\n",
      "for these 1086 overlapping genes and compared their gene\n",
      "expression changes in two different genotypes ( pifQ vsspaQ ). A\n",
      "large number of genes showed a similar pattern of up (red)- or\n",
      "down (green)-regulation compared to wild type (Fig. 8e),\n",
      "suggesting that SPA-regulated genes in early red light are largely\n",
      "PIF-regulated as well. Among these genes, Fig. 8f shows an\n",
      "example of the expression level for one light-inducible and one\n",
      "light-repressed gene in spaQ and cop1-4 in comparison to wild\n",
      "type. Moreover, we selected a few of these genes and veri ﬁed their\n",
      "expression levels by an independent method using RT-qPCR\n",
      "assays. Results show that the RNA-seq data are largelyreproducible (Supplementary Fig. 9). Taken together, the RNA-\n",
      "seq data further supports the proposed trimolecular framework\n",
      "among phyB-SPA-PIF1 in regulating gene expression and\n",
      "photomorphogenesis in response to red light.\n",
      "Discussion\n",
      "Substrate recognition by E3 ubiquitin ligases and subsequent\n",
      "degradation is one of the most critical cellular processes that is\n",
      "regulated by multiple signaling cascades in plants and animals2,61.\n",
      "Speci ﬁc recognition of a substrate by an E3 ubiquitin ligase can be\n",
      "enhanced by phosphorylation of a substrate which increases the\n",
      "afﬁnity between the substrate and the E3 ligase1,62. To promote\n",
      "photomorphogenesis, PIFs are phosphorylated, ubiquitinated and\n",
      "then degraded under light31,32. This study provides a mechanistic\n",
      "view on how PIF1 is phosphorylated and ubiquitinated by a\n",
      "aCol-0 R/D\n",
      "2702\n",
      "1861338 397\n",
      "96 556 561228 333 1866339\n",
      "164 258\n",
      "690\n",
      "939 1086 6105\n",
      "pifQ D\n",
      "/Col-0 DspaQ  R/D\n",
      "1451\n",
      "spaQ  R/Col-0 R\n",
      "7191\n",
      "pifQ D/Col-0 D\n",
      "2025cop1  R/D\n",
      "1332phyB FL R/FL D\n",
      "1117\n",
      "phyB C dependent\n",
      "red responsive genes\n",
      "561SPA-dependent\n",
      "red responsive genes\n",
      "2199\n",
      "phyB ΔC R/ ΔC D\n",
      "652c\n",
      "f AT4G29930\n",
      "AT1G80440RPKM RPKM\n",
      "Col-0 D Col-0 R cop1 D\n",
      "cop1 R\n",
      "spaQ D\n",
      "spaQ RLog2FC20\n",
      "15\n",
      "10\n",
      "5\n",
      "0\n",
      "150\n",
      "100\n",
      "50\n",
      "0\n",
      "–6 –2 2 6spaQ  R\n",
      "/Col-0 Re db\n",
      "Fig. 8 RNA-sequencing revealed unique roles of SPAs in PIF-regulated gene expression. aVenn diagram of differentially expressed genes (DEGs) in a\n",
      "three-way comparison as indicated (Col-0 R/D, spaQ R/D and cop1-4 R/D). Genes showing differential expression in red light vs dark in three genotypes\n",
      "are presented. bVenn diagram of DEGs in phyB-GFP FL vs phyB-GFP ΔC in response to red light. A total of 561 genes was identi ﬁed as DEGs in phyB-GFP\n",
      "FL that did not show differential expression in phyB-GFP ΔC.cVenn diagram showing DEGs in a pairwise comparison between SPA-dependent red light\n",
      "responsive genes and phyB C-terminal-dependent red light responsive genes. Sixty percent of phyB C terminal-dependent red light responsive genes a re\n",
      "overlapped with SPA-dependent red responsive genes, supporting a role of the phyB C-terminal domain in SPA-regulated gene expression in response to\n",
      "red light. dVenn diagram showing DEGs in a pairwise comparison between pifQ-dependent genes in darkness and spaQ -dependent red light responsive\n",
      "genes. Approximately, 54% of the PIF target genes are mis-regulated in spaQ mutant. eHierarchical clustering of 1086 DEGs in two different pairwise\n",
      "comparisons as indicated ( pifQD/Col-0 D and spaQ R/Col-0 R). fBar graph showing expression level of one light-inducible and one light-repressive gene\n",
      "in three genotypes as indicated. Reads Per Kilobase of transcript per Million (RPKM) values were obtained from RNAseq data. Error bar =SD, ( n=3)NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 11cognate kinase-E3 ubiquitin ligase (COP1-SPA) for rapid light-\n",
      "induced degradation.\n",
      "Although, initial studies on the SPA proteins were focused on\n",
      "their accessory role in enhancing COP1 E3 ligase function, aCOP1-independent role for SPAs has not been shown yet, despite\n",
      "the presence of a kinase-like domain at the N-terminus. Here we\n",
      "show that the N-terminal domain of SPA1 possesses a ser/thr\n",
      "kinase activity. Recombinant full-length SPA1 and the N-terminal\n",
      "kinase-like catalytic domain (SPA1-Kin) display autopho-\n",
      "sphorylation activity (Fig. 1). Both proteins also phosphorylate a\n",
      "generic substrate MBP and a native plant substrate PIF1 in a\n",
      "concentration-dependent manner in vitro. The SPA1-mediated\n",
      "phosphorylation is acid stable and base labile, suggesting that the\n",
      "phosphorylated residues are serine and/or threonine (Supple-\n",
      "mentary Fig. 2c). Mutation in a critical residue of SPA1 (R517E)\n",
      "abrogating a conserved Glu-Arg salt bridge essential for eukar-\n",
      "yotic protein kinases results in reduced kinase activity toward\n",
      "MBP and PIF1 in vitro (Fig. 1). The wild type and mutant SPA1-\n",
      "Kin domains equally bound to an ATP analog (TNP-ATP) in a\n",
      "concentration-dependent manner in vitro (Supplementary\n",
      "Fig. 2e), suggesting that the mutant SPA1-Kin is not defective in\n",
      "ATP binding. Consistently, the light-induced phosphorylation of\n",
      "PIF1 is reduced in the mutant SPA1 compared to wild type SPA1\n",
      "transgenic plants (Fig. 3c). Moreover, the light-induced phos-\n",
      "phorylation of PIF1 is drastically reduced in spaQ compared to\n",
      "wild type (Fig. 2). In addition, the mutant SPA1 failed to rescue\n",
      "the seedling de-etiolation phenotype in the dark and the light-\n",
      "induced seed germination phenotype compared to wild type\n",
      "SPA1 transgenic plants (Fig. 3a, b). Taken together, these data\n",
      "strongly support the conclusion that SPA1 acts as a bona ﬁde ser/\n",
      "thr kinase for PIF1.\n",
      "Recently, it was shown that SPA1\n",
      "R517E mutant failed to rescue\n",
      "the seedling deetiolation phenotype of spaQ43, and these authors\n",
      "hypothesized that the kinase domain of SPA1 may provide\n",
      "structural information critical for SPA1 function. We have\n",
      "reproduced these data independently and also show additional\n",
      "phenotypes defective in this mutant. Our data show that thekinase domain displays a protein kinase activity in phosphor-\n",
      "ylating PIF1 in vitro and in vivo. Moreover, the kinase like\n",
      "domain of SPA1 may act as a molecular scaffold for potential\n",
      "protein-protein interaction. The arginine 517 in SPA1 is very well\n",
      "conserved among many plant species, suggesting an importance\n",
      "of the structural integrity of the kinase domain (Supplementary\n",
      "Fig. 1). It is possible that the SPA1\n",
      "R517Emutant compromised not\n",
      "only the kinase activity, but also the structural integrity of the\n",
      "SPA1 kinase domain, resulting in a defect in PIF1 degradation\n",
      "and seed germination in response to light. The biochemical basis\n",
      "for the failure of the SPA1R517Emutant to rescue the spaQ\n",
      "seedling de-etiolation phenotype in the dark is still unknown.\n",
      "However, our hypothesis is that the kinase activity by itself and/or\n",
      "the structural information included within the kinase domain\n",
      "contribute to regulating seedling de-etiolation phenotypes.\n",
      "Direct interaction of PIF1 with the Pfr form of phyB is\n",
      "necessary for the rapid light-induced phosphorylation and\n",
      "degradation of PIF138. The in vitro and in vivo data presented\n",
      "here show that phyB is promoting the PIF1-SPA1-phyB complex\n",
      "formation in a red light-inducible manner. Domain mapping\n",
      "analyses shows the last 80 amino acids at the C-terminal end of\n",
      "phyB is necessary and suf ﬁcient for SPA1 interaction (Fig. 4a–e).\n",
      "Deletion of the SPA1 binding domain from phyB (phyB ΔC)\n",
      "results in reduced PIF1 phosphorylation and degradation and\n",
      "hyposensitive phenotypes of phyB-GFP ΔCcompared to wild type\n",
      "phyB-GFP FL plants (Figs. 6,7, Supplementary Fig. 7). In addi-\n",
      "tion, three other phyB C-terminal deletion mutants, phyB-28 ,CG\n",
      "and NG, that were previously reported48,54, also showed\n",
      "reduced phosphorylation and degradation of PIF1 similar to thatofphyB-GFP ΔC(Fig. 7, Supplementary Fig. 7), further sup-\n",
      "porting that the C-terminal end of phyB is required for phyB-\n",
      "SPA1 interaction. Overall, these data strongly suggest that the\n",
      "recruitment of SPA1 kinase in a PIF1-SPA1-phyB trimolecularcomplex is a critical step for the rapid light-induced phosphor-\n",
      "ylation of PIF1.\n",
      "The importance of the C-terminal domain of phytochromes in\n",
      "mediating light signaling has been in focus for decades\n",
      "63.\n",
      "Recently, Qiu et al showed that overexpression of the C-terminal\n",
      "domain of phyB (625 -1172 aa) induces degradation of PIF3 and\n",
      "activates light-inducible gene expression in the dark55. Consistent\n",
      "with this study, Park et al (2018) showed that phyB regulates PIF\n",
      "activity by two independent methods: PIF sequestration, which is\n",
      "induced by the N-terminal half of phyB57,64, and PIF degradation,\n",
      "which is induced by the C-terminal half of phyB64. These two\n",
      "separable activities of phyB are located in two halves of phyB\n",
      "protein. Although, these data highlight the importance of the C-\n",
      "terminal domain of phyB in regulation of PIF abundance, our\n",
      "data show that the last 80 amino acids of the C-terminal domain\n",
      "of phyB is necessary for the light-dependent SPA1 interaction\n",
      "both in vitro and in vivo. Moreover, this domain is necessary for\n",
      "phyB nuclear translocation and photobody formation in response\n",
      "to light (Fig. 6e). SPA proteins co-translocate with phyB into the\n",
      "nucleus46. However, it is highly unlikely that SPA proteins are the\n",
      "only factors facilitating transport of phyB into nucleus, as PIFs\n",
      "also promote nuclear translocation of phyB65. It is possible that\n",
      "this domain interacts with SPA proteins as well as other factors,\n",
      "and these combined interactions are required for phyB nuclear\n",
      "transport and photobody formation. Because phyB-GFP ΔC\n",
      "behaves normally in terms of chromophore binding, dimeriza-\n",
      "tion, PIF1 interaction and sequestration, and complements the\n",
      "phyB phenotypes to a large extent, the defects observed in phyB-\n",
      "GFPΔCseedlings might not be due to a non-functional protein,\n",
      "rather due in part to a lack of phyB-SPA1 binding in addition to\n",
      "other potential unknown interactions necessary for robust phyB\n",
      "nuclear transport and photobody formation. However, it is still\n",
      "possible that phyB- ΔC might have additional intrinsic defects that\n",
      "were not fully addressed in our study. For example, the phyB-\n",
      "GFP ΔC exhibits defect in the photobody formation similar to the\n",
      "PBY18-YFP (D1040V), which cannot dimerize55. But unlike\n",
      "PBY18-YFP (D1040V) which lacks PIF3 interaction, phyB-GFP\n",
      "ΔC still interacts with PIF1 in vivo. Because the isolated phyB CT\n",
      "ΔC failed to dimerize in yeast, it is still possible that the last 80\n",
      "amino acids of phyB contributes to the dimerization and pho-\n",
      "tobody formation of phyB-GFP ΔC.\n",
      "Multiple kinases phosphorylate PIFs in vitro including, phy-\n",
      "tochrome itself32,42,49,50,59,66. However, Casein Kinase II (CK2)\n",
      "and BRASSINOSTEROID- INSENSITIVE2 (BIN2) were shown\n",
      "to phosphorylate PIF1 and PIF4, respectively, only in vitro in a\n",
      "light-independent manner. Photoregulatory Protein Kinases\n",
      "(PPK1 –PPK4) phosphorylate PIF3 both in vitro and in vivo, and\n",
      "the light-induced phosphorylation of PIF3 is defective in ppk\n",
      "mutants compared to wild type59. However, PPKs do not interact\n",
      "with phyB in a light-dependent manner in vitro. Moreover, PPK-\n",
      "mediated phosphorylation of PIF3 might be involved in signal\n",
      "attenuation as opposed to signal transduction, as the ppkmutants\n",
      "display hypersensitive phenotypes in response to red light. A\n",
      "recent study also showed that phytochrome itself acts as a kinase\n",
      "for PIFs50. However, the kinase activity of oat phyA has been\n",
      "mapped to the N-terminus as opposed to a conserved HKRD\n",
      "located at the C-terminus. Although the location of the kinase\n",
      "domain in Arabidopsis phys is still being investigated, puri ﬁed\n",
      "Arabidopsis phyA and phyB from two different expression hosts\n",
      "induce phosphorylation of PIF1 in vitro (Supplementary Fig. 5).\n",
      "Moreover, phyB and SPA1 mutually enhanced phosphorylation\n",
      "of PIF1 in vitro (Fig. 5). Thus, it appears that multiple kinases areARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "12 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsnecessary for sequential and possibly cooperative phosphoryla-\n",
      "tion of PIF1. Phenotypic analyses showed that only spaQ exhibits\n",
      "robustly delayed germination phenotype and an absence of the\n",
      "light-induced PIF1 phosphorylation in vivo. Therefore, SPAs are\n",
      "necessary for phyB-mediated signal transduction to regulate PIF1\n",
      "abundance (Fig. 9).\n",
      "Because SPAs are crucial for the COP1-SPA E3 ubiquitin ligase\n",
      "complex, this study identi ﬁed a novel mechanism by which a\n",
      "cognate kinase-E3 ligase complex selectively interacts with its\n",
      "substrate and triggers rapid degradation in a light-inducible\n",
      "manner. This unique kinase-E3 ligase complex may account for\n",
      "the rapid degradation speed observed under red light for PIF138.\n",
      "Finally, our ﬁndings suggest that other known COP1-SPA targets\n",
      "might also be phosphorylated by SPAs, given that the WD40\n",
      "repeat domains present in COP1 and SPA proteins prefer phos-\n",
      "phorylated substrates for binding67,68. Future experiments are\n",
      "necessary to address whether other COP1-SPA substrates are also\n",
      "phosphorylated by SPA kinase, and if the cognate kinase-E3\n",
      "ubiquitin ligase model presented here may also be applicable to\n",
      "other E3 ligase complexes in plants and animals.\n",
      "Methods\n",
      "Plant materials and growth conditions . Wild type Col-0, various mutants, and\n",
      "transgenic plants in the Col-0 background were used in this study unless otherwise\n",
      "indicated. Plants were grown in soil under 24-h light at 22 ± 0.5 °C.\n",
      "Construction of vectors and generation of transgenic plants .TAP-PIF1 ,TAP-\n",
      "SPA1 ,PIF1-HA ,TAP-SPA1/PIF1-HA , and TAP-PIF1/spaQ transgenic plants were\n",
      "reported previously13,22,38,41,42. To generate phyB-GFP FL and phyB-GFP ΔC,\n",
      "PHYB F KpnI-PHYB R SmaI, PHYB-GFP ΔC R SmaI primers were used to\n",
      "amplify PHYB coding sequence. These were ﬁrst subcloned into pEZS-NL vector.\n",
      "The expression cassettes from pEZS-NL-PHYB( ΔC) were cut out with NotI and\n",
      "cloned into pART27 binary vector with NotI restriction enzyme. Then pART27-\n",
      "PHYB( ΔC)-GFP were transformed into phyA-211phyB-9 double mutant and\n",
      "transformants were selected in the presence of Kanamycin antibiotics. For SPA1,full-length SPA1-6his coding sequence was ampli ﬁed by PCR using the primers\n",
      "SPA1 F BamHI -His R XhoI and cloned into pCAMBIA C-luc vector. The mutant\n",
      "version of SPA1, mSPA1 was generated using Quickchage II site-directed muta-\n",
      "genesis kit (Agilent, Catalog #200523) in pBSK-SPA1. Both wild type and the\n",
      "mutant SPA1 were transformed into GV3101 and then transformed into spa123\n",
      "mutant. Obtained SPA1 transgenic plants in spa123 were further crossed with spaQ\n",
      "mutant to generate SPA1/ spaQ and mSPA1/ spaQ transgenic plants.\n",
      "Protein puri ﬁcation from E. coli . SPA1 kinase domain was subcloned into\n",
      "pASK75 vector with SPA1 F BamHI- SPA1 kin R SalI primers. SPA1 kinase\n",
      "domain was then PCR ampli ﬁed with SPA1 F BamHI-Strep R NotI primers andcloned into pGEX4T-1. For generation of mSPA1, pASK75-SPA1 kinase domain\n",
      "clone was mutagenized using Quickchange II site-directed mutagenesis kit (Agi-\n",
      "lent, Catalog #200523). Each plasmid was transformed into BL21(DE3) cells.\n",
      "Protein expression was induced under 16 °C for 12 h with 0.1 mM IPTG. Collectedcells were sonicated in binding buffer (100 mM Tris, pH 7.5, 150 mM NaCl, 0.2%\n",
      "Tergitol NP-40, 0.1 mM EDTA, 1· Protease inhibitor cocktail, and 1 mM PMSF)\n",
      "and puri ﬁed using Strep-Tactin Sepharose column (IBA, Cat# 2-1209-001). Col-\n",
      "umns were washed three times thoroughly by applying 5 mL of the binding buffer.\n",
      "Proteins were eluted with the elution buffer (Tris 7.5 100 mM, EDTA 1 mM, 10%glycerol, 10 mM d-desthiobiotin) into separate fractions. The eluted proteins were\n",
      "analyzed on a SDS-PAGE gel and used for kinase assays. pGEX PIF1-strep clone\n",
      "was previously reported\n",
      "50. GST-PIF1-strep was puri ﬁed using the same method as\n",
      "described in this section. For MBP-SPA1, pMAL p2x SPA1 construct was used as\n",
      "described previously41.\n",
      "SPA1 and phytochrome puri ﬁcation from Pichia pastoris . The full-length SPA1\n",
      "coding sequence was PCR-ampli ﬁed with SPA1 F BamHI-SPA1 R SalI primers and\n",
      "subcloned into pASK75 vector. Then, SPA1-strep coding sequence was PCR-\n",
      "ampli ﬁed with SPA1 F SnaBI- strep R NotI primers and cloned into pPIC3.5 K\n",
      "vector. Flag sequences (3·) were further cloned into the SnaBI site of the pPIC3.5 K\n",
      "SPA1-strep clone using Flag F-Flag R primers. Approximately 30 µg of SPA1-strep\n",
      "pPIC3.5 K plasmid DNA was linearized using PmeI restriction enzyme. Pichia\n",
      "pastoris strain GS115 was grown in YPD media and the electro-competent cell was\n",
      "prepared according to the manufacturer ’s instructions (Life Technology, Pichia\n",
      "Expression Kit, Cat.# K1710-01). Electroporation was performed using a generalsetup provided by electroporator (Bio-rad) for Pichia pastoris . Transformed colo-\n",
      "nies were selected on the MD media plates (1· YNB, 1· biotin, 2% Dextrose, 1.8%\n",
      "agar). Plates were incubated at 30 °C for two days to obtain ~5000 colonies.\n",
      "Multiple genomic integrations of the plasmids were further con ﬁrmed by antibiotic\n",
      "resistance at 0.5, 1, 1.5, 2, 2.5, 3, 3.5, and 4 µM of G418 Geneticin (Sigma, Cat.#11811023) containing YPD media for ﬁve days of incubation. The selected colonies\n",
      "were tested their Mut\n",
      "sphenotype on solid MM media (1· YNB, 1· biotin, 5%\n",
      "methanol, 1.8% agar) and used for protein puri ﬁcation. For induction of the\n",
      "protein expression, the colony was initially inoculated in 10 mL of liquid MM\n",
      "media for 24 h, and then cultured in 200 mL of liquid MGY (1x YNB, 1x biotin, 2%\n",
      "glycerol) media for an additional 24 h. The cells were collected and MGY mediawas completely removed. Then, the collected cells were resuspended and diluted\n",
      "into 1 L of liquid MM media (1x YNB, 1x biotin, 5% methanol) to induce protein\n",
      "expression for 24 h at 30 °C.\n",
      "The collected induced cells were resuspended in the lysis buffer (100 mM Tris\n",
      "pH 7.5, 150 mM NaCl, 0.2% Tergitol NP-40, 0.1 mM EDTA, 1x Protease inhibitorcocktail, 1 mM PMSF) and ground in liquid nitrogen using physical force with a\n",
      "homogenizer (Nohonseiki Kaisha LTD Japan, model AM-3). The whole cell\n",
      "extracts were ﬁltered through a 0.45 µm ﬁlter and ran through the Strep-Tactin\n",
      "Sepharose column (IBA Lifesciences, Cat.# 2-1202-001) for puri ﬁcation. The\n",
      "column was washed 5 times thoroughly before eluting with the elution buffer\n",
      "(100 mM Tris pH 7.5, 1 mM EDTA, 10% glycerol, 10 mM d-desthiobiotin).\n",
      "Puriﬁed protein was analyzed on a SDS-PAGE gel.\n",
      "Arabidopsis phytochrome B was puri ﬁed from Pichia pastoris as described\n",
      "previously\n",
      "69. In brief, PHYB expressing Pichia pastoris GS115 strain was grown in\n",
      "200 mL MGY media for 24 h before transferring into 1 L MM media to induce\n",
      "protein expression. The cells were harvested and homogenized. Total proteins wereCOP1\n",
      "COP1SPAs\n",
      "SPAsPIF1PIF1\n",
      "HY5\n",
      "COP1COP1SPAsSPAsPIF1PIF1\n",
      "HY5PIF1\n",
      "ppUbi\n",
      "Ubi\n",
      "Ubi\n",
      "Ubi\n",
      "Ubi\n",
      "26 S\n",
      "Protea-\n",
      "someppppphyB in Pr\n",
      "phyB in Pfr\n",
      "Synergistic\n",
      "PhotomorphogenesisPhotomorphogenesis\n",
      "Fig. 9 Model showing phyB-SPA1-COP1-PIF1 relationships in dark and light conditions. In darkness, inactive Pr form of phyB is present in the cytoplasm,\n",
      "while COP1-SPA complex together with PIF1 as a co-factor induces degradation of positively acting transcription factors (HY5/HFR1 and possibly oth ers).\n",
      "However, upon light exposure, activated Pfr form of phyB translocates into the nucleus and interacts with PIF1 as well as SPA1 to trigger rapid\n",
      "phosphorylation of PIF1. phyB interacts with PIFs mostly through its N-terminal domain, while its extreme C-terminal 80 amino acids (shown as a red\n",
      "patch) are necessary for SPA1 interaction. By stabilizing the phyB-SPA1-PIF1 tripartite complex, phyB can initiate the light-induced phosphoryla tion of PIF1\n",
      "by SPA1 kinase. Phosphorylated PIF1 is then recognized by the CUL4COP1-SPAE3 ubiquitin ligase for rapid poly-ubiquitination and subsequent degradation\n",
      "through the 26S proteasome. Degradation of PIFs and stabilization of HY5 result in promotion of photomorphogenesisNATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 13precipitated by ammonium sulfate salting-out (0.23 g/mL) methods. The pellet was\n",
      "resuspended and assembled with a chromophore (10 µM), phycocyanobilin\n",
      "(Frontier Scienti ﬁc, Cat.# P14137). Strep-Tactin Sepharose column was used to\n",
      "purify phyB holoprotein. The puri ﬁed protein was analyzed on a SDS-PAGE gel.\n",
      "Phytochrome puri ﬁcation from Saccharomyces cerevisiae . We obtained Sac-\n",
      "charomyces strain RKY1293 from Dr. Arlen Johnson at UT. The full-length PHYB\n",
      "coding sequence was ﬁrst subcloned into pEZS-NL using phyB F KpnI- phyB R\n",
      "SmaI primers. By using phyB F KpnI- GFP R XbaI primers, PHYB-GFP was PCR-ampli ﬁed and cloned into pYES2 (Invitrogen, Carlsbad, USA) vector. The pYES2-\n",
      "PHYB-GFP was transformed into yeast strain RKY1293 using EZ transformation\n",
      "kit and transformants were selected on –URA dropout media. The selected colonies\n",
      "were grown in the 50 mL of liquid –URA dropout media with 2% raf ﬁnose as a\n",
      "carbon source. By adding 2% of galactose, phyB expression was induced for 12 h.\n",
      "The induced cells were collected and resuspended in lysis buffer (100 mM Tris pH\n",
      "7.5, 150 mM NaCl, 0.1% Tergitol NP-40, 1 mM EDTA, 1x Protease inhibitor\n",
      "cocktail, 1 mM PMSF), and then frozen in liquid nitrogen to be ground thoroughlyusing a pestle and mortar. The ﬁne powder was thawed on the mortar and collected\n",
      "in the test tubes. Tubes were centrifuged at a maximum speed 20200 × g for 20 min\n",
      "at 4 °C. For puri ﬁcation of the apoprotein, one microgram of anti-GFP (Abcam #\n",
      "ab6556) antibody was preincubated with 20 µL of protein A Dynabeads for 15 min.\n",
      "The antibody-bound Dynabeads were washed with the lysis buffer and incubated\n",
      "with the supernatant. After two hours of incubation, the Dynabeads were collectedand washed three times with the lysis buffer using a magnet rack. The immuno-\n",
      "precipitated PHYB containing Dynabeads were resuspended in 100 µL of the lysis\n",
      "buffer supplemented with 10 ⌠M phycocyanobilin (PCB) and incubated in the dark\n",
      "for 1 h. The Dynabeads were collected and washed four times thoroughly with the\n",
      "lysis buffer and used for other assays.\n",
      "In vitro kinase assay . For SPA1 kinase assay, about 500 ng of SPA1 and 1 µg of\n",
      "GST-PIF1 were used. For Pichia puriﬁed phyB, about 1 µg of holoprotein was used\n",
      "for the kinase assay. For Saccharomyces puriﬁed phyA and phyB, about 500 ng of\n",
      "holoproteins were used. All kinase assays were performed in kinase buffer (50 mM\n",
      "Tris, pH 7.5, 4 mM β-mercaptoethanol, 1 mM EDTA, 10 mM MgCl\n",
      "2) except the\n",
      "phyB+SPA1 light dependent kinase assay, which contains 100 mM NaCl.32P\n",
      "radio-labeled gamma-ATP (Perkin Elmer Cat# BLU502A) was added to the\n",
      "reaction and incubated at 28 °C for 1 h unless otherwise indicated. SDS sample\n",
      "buffer (6x) was added to stop the reaction and the boiled proteins were separated\n",
      "on a SDS-PAGE gel. The gels were dried and exposed to a phosphor screen and\n",
      "then scanned with Typhoon FLA 9500 (GE healthcare, Chicago, USA).\n",
      "Acid and base sensitivity assay . The kinase assay was performed as described in\n",
      "the previous section. The total reaction containing SPA1 and PIF1 was boiled and\n",
      "divided into three lanes in a SDS-PAGE gel. The proteins were blotted onto PVDFmembrane and cut into three pieces. Each piece was treated with neutral (Tris, pH\n",
      "7.5), basic (3 M KOH) and acid (1 M HCl) solution as described previously\n",
      "49. The\n",
      "membrane was then exposed to a phosphor screen and scanned by Typhoon FLA9500 (GE healthcare, Chicago, USA).\n",
      "PIF1 mobility shift assay . To observe TAP-PIF1 mobility shift in Col-0 and spaQ ,\n",
      "total protein was separated in a 10·10.5 cm 6.5% SDS-PAGE. For phostag gelanalysis, each sample was ground in the buffer (50 mM Tris, pH 7.5, 150 mM NaCl,\n",
      "1 mM EDTA, pH 8.0, 0.1% Tween20, 1 × protease inhibitor cocktail [Sigma-\n",
      "Aldrich, Cat# P9599], 1 mM PMSF, 20 μM bortezomib, 25 mM β-glyceropho-\n",
      "sphate, 10 mM NaF, and 2 mM Na orthovanadate) and immunoprecipitated using\n",
      "anti-myc (Sigma, Cat.# C3956) antibody. The immunoprecipitated TAP-PIF1 wasthen separated on a 6.5% SDS-PAGE gel containing 15-30 µM phostag acrylamide\n",
      "(Wako Japan, Cat# AAL-107). The phostag gel was prepared according to the\n",
      "manufacturer ’s instructions.\n",
      "Yeast two-hybrid Yeast two-hybrid assay . LexA-phyB CT indicates the C-\n",
      "terminal domain of PHYB (aa 625-1172) and LexA-phyB CT ΔC indicates the C-\n",
      "terminal half of PHYB without the last 80 amino acids (aa 625-1092). For the baitproteins, phyB CT and phyB CT ΔC PCR products were ampli ﬁed using PHYB-F-\n",
      "BamHI/PHYB-R-SalI and PHYB-F-BamHI /PHYB CT ΔC-R-SalI primers,\n",
      "respectively. PCR products were then cloned into pEG202 vector with LexA fusion.\n",
      "For the prey proteins, the full-length and various deletion fragments of AD-SPA1\n",
      "fusion constructs were described previously\n",
      "41. All the constructed were con ﬁrmed\n",
      "by enzyme digestion and sequencing.\n",
      "Plasmids containing bait and prey constructs were transformed into EGY48\n",
      "yeast strain and selected on -His, -Ura, -Trp dropout medium at 30 °C for 3 days.Yeast colonies grown on this minimal medium were cultured overnight in liquid\n",
      "media supplemented with 2% (w/v) glucose. Overnight cultures (1 mL) were then\n",
      "transferred to 4 mL liquid medium supplemented with 2% (w/v) galactose and 1%\n",
      "(w/v) raf ﬁnose to induce the expression of proteins. β-galactosidase assays were\n",
      "performed according to the Matchmaker Two-Hybrid System, ( Takara\n",
      "Biosciences ). Three independent replicates were performed and the bar graphs\n",
      "represent the β-galactosidase activity normalized by activation domain (AD) vector\n",
      "control.The light-dependent Yeast two-hybrid assays were performed as described\n",
      "26.\n",
      "The AD-PIF3, AD-SPA1, and D153-phyB NT-BD constructs have been described\n",
      "previously41,70. Brie ﬂy, AD and BD fusion combination plasmids were transformed\n",
      "into AH109 yeast cells and selected on drop out media without Leu and Trp.\n",
      "Individual colonies were cultured in the same liquid media and 10 μL of each\n",
      "culture were spotted on a plate containing 5 mM 3-AT and 20 μM phycocyanobilin\n",
      "(PCB), but lacking Leu, Trp and His. Colonies were grown at 30 °C for 2 days and\n",
      "then photographed. A control plate without Leu and Trp was grown to shownormal growth.\n",
      "In vivo co-immunoprecipitation assay . For in vivo co-immunoprecipitation\n",
      "assays, ~400 4-day-old dark-grown seedlings were collected and ground in liquid\n",
      "nitrogen. Total proteins were solubilized in the IP buffer (50 mM Tris, pH 7.5,\n",
      "150 mM NaCl, 1 mM EDTA, pH 8.0, 0.1% Tween-20, 0.25 mM DTT, 1 × proteaseinhibitor cocktail [Sigma-Aldrich Co., St. Louis, MO, Cat# P9599], 1 mM PMSF,\n",
      "20μM bortezomib, 25 mM β-glycerophosphate, 10 mM NaF, and 2 mM Na\n",
      "orthovanadate), cleared by centrifugation at 20200 xg for 10 mins. Anti-HA(Abcam, Cat.# ab9110), and anti-GFP (Abcam, Cat.# ab6556) were bound to\n",
      "dynabeads before adding into total cell extractions. After 2 h of incubation at 4 °C\n",
      "in the dark, the dynabeads (20 μl/μg antibody; Life Technologies, Carlsbad, USA)\n",
      "were collected and washed using IP buffer on magnet rack. The beads were then\n",
      "boiled with 6X SDS sample buffer, ran on a 6.5% SDS-PAGE gel, and blotted withcorresponding antibodies.\n",
      "In vitro/semi-in vitro pull-down assay . For in vitro pull-down assay, GST-PIF1,\n",
      "GST, SPA1-strep, and mSPA1-strep proteins were puri ﬁed as described above.\n",
      "Total 2 µg of GST and GST PIF1 proteins were bound to 25 µl of GST agarose\n",
      "beads (Pierce, Waltham, USA. Cat.# 20211). One microgram of SPA1 and mSPA1\n",
      "was added to each sample with 500 µL binding buffer (50 mM Tris, pH 7.5,\n",
      "150 mM NaCl, 1 mM EDTA, 0.1% Tergitol NP-40, 0.5 mM DTT, 1 × protease\n",
      "inhibitor cocktail, 1 mM PMSF). After 2 h of incubation with gentle rotation at 4 °C, beads were collected with centrifugation at 1800 ×g. The collected beads were\n",
      "thoroughly washed ﬁve (x 5 mins) with 1 ml of the binding buffer by gentle\n",
      "rotation. Beads were then boiled with 2x SDS sample buffer and separated on anSDS-PAGE gel to visualize protein with Coomassie staining.\n",
      "For semi-in vitro immunoprecipitation assay, phyB-GFP FL and phyB-GFP ΔC\n",
      "seeds were grown in the dark for 4 days before grinding in liquid nitrogen. Totalprotein was solubilized in IP buffer (50 mM Tris, pH 7.5, 150 mM NaCl, 1 mM\n",
      "EDTA, 0.1% Tween20, 0.25 mM DTT, 1 × protease inhibitor cocktail [Sigma-\n",
      "Aldrich Co., St. Louis, MO, Cat# P9599], 1 mM PMSF). The extracts were cleared\n",
      "by centrifugation at 20200 × g for 10 min. The supernatant was incubated with\n",
      "anti-GFP (Abcam, Cat.# ab6556) antibody-bound dynabeads (20 μL/μg antibody)\n",
      "for one hour in the dark. Immunoprecipitated phyB-GFP proteins were washed\n",
      "three times with IP buffer. Whole cell extract from IPTG induced E. coli harboring\n",
      "pMal p2x SPA1 plasmids were added into immunoprecipitated beads and eachtube was treated either red (10 µmolm\n",
      "−2s−1) or far-red light (15 µmole/s/m2) for\n",
      "5 mins. After an additional two hours incubation at 4 °C, beads were washed four\n",
      "times, ﬁve mins each with extraction buffer (100 mM phosphate buffer pH 7.8, 150\n",
      "mM NaCl, 0.5 mM DTT, 0.2% Nondidet P-40, 1x protease inhibitor cocktail, 1 mM\n",
      "PMSF). Beads were then boiled with 2X SDS sample buffer and separated on aSDS-PAGE gel. Anti-MBP (NEB, Cat.# E8032S) antibody was used to detect co-\n",
      "immunoprecipitated SPA1-MBP protein.\n",
      "For the tripartite complex pull-down assay, 2 µg of puri ﬁed GST-PIF1-strep\n",
      "were bound to GST agarose beads. Twenty μL of the GST-bound beads were\n",
      "resuspended in the binding buffer (50 mM Tris, pH 7.5, 150 mM NaCl, 1 mM\n",
      "EDTA, 0.1% Tween 20, 0.25 mM DTT, 1 × protease inhibitor cocktail,1 mM\n",
      "PMSF). SPA1-MBP protein expression was induced in the BL21 cell as described in\n",
      "the previous section. The induced cell pellets were resuspended and sonicated inthe lysis buffer (50 mM Tris, pH 7.5, 150 mM NaCl, 1 mM EDTA, 0.2% Tween 20,\n",
      "0.25 mM DTT, 1 × protease inhibitor cocktail,1 mM PMSF), cleared by\n",
      "centrifugation at 20200 × g for 30 mins. The supernatant was then added toprepared GST-PIF1 beads and incubated with or without increasing amount of\n",
      "puriﬁed phyB (1 –2 µg) from Pichia pastoris . After the pull-down, the beads were\n",
      "washed three times, ﬁve min each with the binding buffer. Co-precipitated MBP-\n",
      "SPA1, phyB and GST-PIF1 was immunoblotted with anti-MBP (NEB, Cat.#\n",
      "E8032S), anti-strep (Sigma, Cat.# GT661), and anti-GST HRP (GE healthcare, Cat.\n",
      "# RPN1236) antibodies.\n",
      "In vitro phytochrome dimerization assay . Total 5 µg of Pichia pastoris puriﬁed\n",
      "phyB-GFP FL and ΔC holoproteins were mixed with 2X loading buffer (62.5 mM\n",
      "Tris-HCl, pH 6.8, 25% Glycerol, 1% Bromophenol Blue). Native 6% PAGE gel was\n",
      "prepared and ran in 4 °C, 100 v for 3 hours. The native PAGE gel was stained with\n",
      "Coomassie brilliant blue staining solution and the proteins were visualized.\n",
      "PIF1 sequestration assay using chromatin immunoprecipitation . Chromatin\n",
      "immunoprecipitation (ChIP) was performed as previously described\n",
      "71. In brief,\n",
      "phyB-GFP FL/TAP-PIF1 andphyB-GFP ΔC/TAP-PIF1 were plated in MS media\n",
      "and grown for 4 days in the dark. Final 10 µM bortezomib was pretreated in the\n",
      "dark for 4 h before the red light irradiation to block 26 S proteasome-mediatedARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "14 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunicationsdegradation of the TAP-PIF1. A pulse of red light (300 µmol s−2) treated for the\n",
      "red samples and whole seedlings were subjected to cross-linking in 1% for-\n",
      "maldehyde buffer for 15 min. The crosslinking was quenched by adding 2 M gly-\n",
      "cine and incubated additional 5 min. The seedlings were collected and ground in\n",
      "liquid nitrogen, resuspended in the nuclear isolation buffer (0.25 M sucrose, 15 mMPIPES, 5 mM MgCl2, 60 mM KCl, 15 mM NaCl, 1% Triton X-100, 2 mM PMSF,\n",
      "1X protease inhibitor cocktail). Cleared by 15 min centrifugation at 20200 × g at\n",
      "4 °C, the pellets were resuspended in 1 mL lysis buffer (50 mM HEPES pH7.5,150 mM NaCl, 1 mM EDTA, 1% Triton X-100, 0.1% SDS, 2 mM PMSF, 1X pro-\n",
      "tease inhibitor cocktail). Each sample was sonicated at 10% amplitude with\n",
      "Branson Digital Soni ﬁer (Model# PC BD 400 W 20 kHz) sonication machine (10\n",
      "seconds each for ten times), total 100 seconds sonication. Samples were centrifuged\n",
      "at 20,200 ×gfor 20 min. Ten percent of the supernatant was saved for input DNA\n",
      "isolation. Anti-myc (Sigma, Cat.# C3956) antibody-bound Dyanbeads were then\n",
      "added to the supernatant for PIF1 chromatin immunoprecipitation overnight.\n",
      "Bound ChIP samples were washed sequentially with low salt, high salt, LiCl, andTE wash buffer for 5 mins each. Eluted chromatin was then subjected to reverse\n",
      "crosslinking by overnight incubation in the presence of NaCl. After proteinase K\n",
      "treatment for an hour, DNA was isolated by DNA puri ﬁcation kit (Qiagen, Cat.#\n",
      "28106). Enrichment was measured by quantitative PCR using primers corre-\n",
      "sponding to the G-box region on PIL1 promoter and PIL1 coding region were\n",
      "compared as a negative control.\n",
      "Protein extraction and western blot analyses . Seeds were surface sterilized and\n",
      "grown in the dark for four-days otherwise indicated. For total protein extraction,\n",
      "about 100 seeds were plated. The tissue was collected and ground in 100 μL\n",
      "denaturing extraction buffer [100 mM Tris pH 7.5, 1 mM EDTA, 8 M Urea, 1x\n",
      "protease inhibitor cocktail (Sigma-Aldrich Co., St. Louis, MO, cat# 59), 2 mM\n",
      "PMSF,+optional 25 mM β-GP, 10 mM NaF and 2 mM Na-orthovanadate for\n",
      "PIF1 included samples] and cleared by centrifugation at 20,200 ×gfor 10 min at\n",
      "4 °C. Samples were boiled for 5 min with 1X SDS sample buffer and separated on a\n",
      "6.5% SDS-PAGE gel, blotted onto PVDF membranes, and probed with corre-\n",
      "sponding antibodies. Antibodies used in these studies are anti-PIF1\n",
      "38, anti-myc\n",
      "(Sigma, Cat.# C3956 for immunoprecipitation, EMD Millipore, Cat#OP10-200UGfor Western blot), anti-GFP (Abcam, Cat.# ab6556 for immunoprecipitation), anti-\n",
      "GFP (Santa Cruz. Cat.# sc-390394 for Western blot) anti-HA (Covance Cat.#\n",
      "MMS-101P for Western blot), anti-MBP (NEB, Cat.# E8032S), and anti-RPT5antibodies (Enzo Life Sciences, Cat.# BML-PW8770-0025). Secondary HRP bound\n",
      "antibodies were visualized with Super Signal West Femto Chemiluminescent\n",
      "substrate (Pierce Biotechnology Inc., Waltham, USA.), and developed with an X-rayﬁlm. The PIF1 amount was quanti ﬁed with image J from three independent\n",
      "blots and normalized by its corresponding RPT5 levels.\n",
      "Light-induced seed germination assay and measurement of hypocotyl lengths .\n",
      "For measurement of hypocotyl lengths, the same batch of seeds were harvested and\n",
      "dried in the presence of desiccation rocks at room temperature for 1 month.\n",
      "Photographs of 4-day-old dark- and light-grown seedlings were taken and at least20 seedlings were measured using the ImageJ software ( http://rsb.info.nih.gov/ij/ ).\n",
      "Seeds were harvested and dried at room temperature in the presence of\n",
      "desiccation rocks for one month. Light-induced germination assay was performedas described previously\n",
      "35. In brief, surface sterilized seeds were imbibed on the MS\n",
      "media for one hour. Plates were initially treated with far-red light (10 µmol/s/m2)\n",
      "for 5 mins and then irradiated with red light with the amount of light as indicated\n",
      "in the ﬁgure. The plates were wrapped with aluminum foil twice and incubated at\n",
      "22 °C for four days before counting germination frequency.\n",
      "RNAseq analysis . Total six genotypes; Col-0, cop1-4 ,spaQ ,phyAphyB /phyB-GFP\n",
      "FL, and phyAphyB /phyB-GFP ΔCplants were used for mRNA sequence analysis.\n",
      "The seeds were surface sterilized and plated on the MS media without sucrose, cold\n",
      "strati ﬁed for three days before treated for 3 hours of white light and 21 h of dark\n",
      "incubation to induce germination. After the initial 24 h, seeds were further treated\n",
      "with far-red light (10 µmolm−2s−1) to inactivate phytochrome activity. After an\n",
      "additional two days in the dark, seedlings were either collected (dark sample) ortreated with 1 h of continuous red light (red sample) (3.5 µmolm\n",
      "−2s−1). Total\n",
      "RNA was isolated using the plant total RNA kit (Sigma, Cat.# STRN250). NEB-\n",
      "Next ®Ultra ™DNA Library Prep Kit for Illumina ®(NEB Cat.# E7370S) was used to\n",
      "generate mRNA sequencing library according to the manufacturer ’s instructions.\n",
      "The constructed library was quality checked by Agilent 2100 Bioanalyzer (Wald-broon, Germany) and sequenced on Illumina NextSeq 500 platform. Raw data and\n",
      "processed data were deposited into the Gene Expression Omnibus database\n",
      "(accession number GSE117114).\n",
      "For the RNA-seq data analysis, the quality of raw reads (FASTQ) was initially\n",
      "checked using FastQC before aligned to the Arabidopsis genome using TopHat\n",
      "72\n",
      "calling Bowtie273. Arabidopsis genome was acquired from TAIR10 ( https://www.\n",
      "arabidopsis.org/ ).\n",
      "HTSeq-count v0.8.0 was used to generate counts of reads mapped to annotated\n",
      "Arabidopsis genes74(http://htseq.readthedocs.io/en/master/index.html ). The data\n",
      "preprocessing, mapping and counting reads were performed using TACC\n",
      "Stampede2 supercomputer.Differentially expressed genes in each mutant compared to wild-type (WT)\n",
      "were analyzed using DESeq2 package in R programming based on the negative\n",
      "binomial distribution75. The differential gene expression from DESeq2 was selected\n",
      "for signi ﬁcant changes with the |log2FC| > 2 for Col-0, spaQ ,cop1-4 and |log2FC| >\n",
      "1.5 for phyB-GFP FL ,phyB-GFP ΔCwith adjusted p-value < 0.05, Wald test as part\n",
      "of the DESeq2 package.\n",
      "Venny v2.1.0 and ComplexHeatmap76package in R were used to generate Venn\n",
      "diagrams and heat maps. Enrichment analysis (Gene Ontology) was implementedusing the Database for Annotation, Visualization and Integrated Discovery\n",
      "(DAVID) v6.8 . In addition, to obtain the number of read per kilo bases per million\n",
      "reads (RPKM), the aligned reads from TopHat were assembled and quanti ﬁed\n",
      "using Cuf ﬂinks\n",
      "72. Differentially expressed genes between two conditions were\n",
      "analyzed by using Cuffdiff from Cuf ﬂinks. Signi ﬁcantly difference was selected with\n",
      "adjusted p-value < 0.05, Cuffdiff analysis.\n",
      "Reporting summary . Further information on research design is available in\n",
      "the Nature Research Reporting Summary linked to this article.\n",
      "Data availability\n",
      "RNA sequencing data were deposited into the Gene Expression Omnibus database\n",
      "(accession numbers GSE112662 and GSE117114). Arabidopsis mutants and transgeniclines, as well as plasmids and antibodies generated during the current study are availablefrom the corresponding author upon reasonable request. The authors declare that the\n",
      "data supporting the ﬁndings of this study including the gels and immunoblots in Figs. 1–7\n",
      "and Supplementary Figs. 1 –10 are available within the paper and its supplementary\n",
      "information ﬁles.\n",
      "Received: 24 October 2018 Accepted: 22 August 2019\n",
      "References\n",
      "1. Deshaies, R. J. SCF and Cullin/RING H2-based ubiquitin ligases. Annu. Rev.\n",
      "Cell Dev. Biol. 15, 435 –467 (1999).\n",
      "2. Vierstra, R. D. The ubiquitin –26S proteasome system at the nexus of plant\n",
      "biology. Nat. Rev. Mol. Cell Biol. 10, 385 –397 (2009).\n",
      "3. Castillon, A., Shen, H. & Huq, E. Phytochrome Interacting Factors: central\n",
      "players in phytochrome-mediated light signaling networks. Trends Plant Sci.\n",
      "12, 514 –521 (2007).\n",
      "4. Yi, C. & Deng, X. W. COP1 - from plant photomorphogenesis to mammalian\n",
      "tumorigenesis. Trends Cell Biol. 15, 618 –625 (2005).\n",
      "5. Marine, J.-C. Spotlight on the role of COP1 in tumorigenesis. Nat. Rev. Cancer\n",
      "12, 455 –464 (2012).\n",
      "6. Podolec, R. & Ulm, R. Photoreceptor-mediated regulation of the COP1/SPA\n",
      "E3 ubiquitin ligase. Curr. Opin. Plant Biol. 45,1 8–25 (2018).\n",
      "7. Lau, O. S. & Deng, X. W. The photomorphogenic repressors COP1 and DET1:\n",
      "20 years later. Trends Plant Sci. 17, 584 –593 (2012).\n",
      "8. Uljon, S. et al. Structural basis for substrate selectivity of the E3 ligase COP1.\n",
      "Structure 24, 687 –696 (2016).\n",
      "9. Keeshan, K. et al. Transformation by Tribbles homolog 2 (Trib2) requires both\n",
      "the Trib2 kinase domain and COP1 binding. Blood 116, 4948 –4957 (2010).\n",
      "10. Hoecker, U. The activities of the E3 ubiquitin ligase COP1/SPA, a key\n",
      "repressor in light signaling. Curr. Opin. Plant Biol. 37,6 3–69 (2017).\n",
      "11. Deng, X.-W. et al. COP1 , an arabidopsis regulatory gene, encodes a protein\n",
      "with both a zinc-binding motif and a G βhomologous domain. Cell71,\n",
      "791–801 (1992).\n",
      "12. McNellis, T. W. et al. Genetic and molecular analysis of an allelic series of\n",
      "cop1 mutants suggests functional roles for the multiple protein domains. The\n",
      "Plant Cell 6, 487 –500 (1994).\n",
      "13. Laubinger, S., Fittinghoff, K. & Hoecker, U. The SPA quartet: a family of WD-\n",
      "repeat proteins with a central role in suppression of photomorphogenesis inarabidopsis. Plant Cell 16, 2293 –2306 (2004).\n",
      "14. Hoecker, U., Tepperman, J. M. & Quail, P. H. SPA1, a WD-repeat protein\n",
      "speciﬁc to phytochrome A signal transduction.\n",
      "Science 284, 496 –499 (1999).\n",
      "15. Hoecker, U., Xu, Y. & Quail, P. H. SPA1 : a new genetic locus involved in\n",
      "phytochrome A-speci ﬁc signal transduction. Plant Cell 10,1 9–33 (1998).\n",
      "16. Laubinger, S. & Hoecker, U. The SPA1-like proteins SPA3 and SPA4 repress\n",
      "photomorphogenesis in the light. Plant J. 35, 373 –385 (2003).\n",
      "17. Fittinghoff, K. et al. Functional and expression analysis of Arabidopsis SPA\n",
      "genes during seedling photomorphogenesis and adult growth. Plant J. 47,\n",
      "577–590 (2006).\n",
      "18. Zhu, D. M. et al. Biochemical characterization of arabidopsis complexes\n",
      "containing CONSTITUTIVELY PHOTOMORPHOGENIC1 andSUPPRESSOR OF PHYA proteins in light control of plant development. Plant\n",
      "Cell20, 2307 –2323 (2008).NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 1519. Seo, H. S. et al. LAF1 ubiquitination by COP1 controls photomorphogenesis\n",
      "and is stimulated by SPA1. Nature 423, 995 –999 (2003).\n",
      "20. Saijo, Y. et al. The COP1-SPA1 interaction de ﬁnes a critical step in\n",
      "phytochrome A-mediated regulation of HY5 activity. Genes Dev. 17,\n",
      "2642 –2647 (2003).\n",
      "21. Ordoñez-Herrera, N. et al. A cop1 spa mutant de ﬁcient in COP1 and SPA\n",
      "proteins reveals partial Co-Action of COP1 and SPA during Arabidopsis post-embryonic development and photomorphogenesis. Mol. Plant 8, 479–481 (2015).\n",
      "22. Chen, H. D. et al. Arabidopsis CULLIN4-damaged DNA binding protein 1\n",
      "interacts with CONSTITUTIVELY PHOTOMORPHOGENIC1-\n",
      "SUPPRESSOR OF PHYA complexes to regulate photomorphogenesis andﬂowering time. Plant Cell 22, 108 –123 (2010).\n",
      "23. Bae, G. & Choi, G. Decoding of light signals by plant phytochromes and their\n",
      "interacting proteins. Annu Rev. Plant Biol. 59, 281 –311 (2008).\n",
      "24. Klose, C., Viczián, A., Kircher, S., Schäfer, E. & Nagy, F. Molecular\n",
      "mechanisms for mediating light ‐dependent nucleo/cytoplasmic partitioning of\n",
      "phytochrome photoreceptors. New Phytol. 206, 965 –971 (2015).\n",
      "25. Sheerin, D. J. et al. Light-activated phytochrome A and B interact with\n",
      "members of the SPA family to promote photomorphogenesis in Arabidopsisby reorganizing the COP1/SPA complex. Plant Cell 27, 189 –201 (2015).\n",
      "26. Lu, X.-D. et al. Red-light-dependent interaction of phyB with SPA1 promotes\n",
      "COP1 –SPA1 dissociation and photomorphogenic development in\n",
      "Arabidopsis. Mol. Plant 8, 467 –478 (2015).\n",
      "27. Subramanian, C. et al. The Arabidopsis repressor of light signaling, COP1, is\n",
      "regulated by nuclear exclusion: mutational analysis by bioluminescenceresonance energy transfer. Proc. Natl Acad. Sci. Usa. 101, 6798 –6802 (2004).\n",
      "28. Pacín, M., Legris, M. & Casal, J. J. Rapid decline in nuclear COSTITUTIVE\n",
      "PHOTOMORPHOGENESIS1 abundance anticipates the stabilization of itstarget ELONGATED HYPOCOTYL5 in the Light. Plant Physiol. 164,\n",
      "1134 –1138 (2014).\n",
      "29. Balcerowicz, M., Kerner, K., Schenkel, C. & Hoecker, U. SPA Proteins affect\n",
      "the subcellular localization of COP1 in the COP1/SPA ubiquitin ligasecomplex during photomorphogenesis. Plant Physiol. 174, 1314 –1321 (2017).\n",
      "30. Xu, X., Paik, I., Zhu, L. & Huq, E. Illuminating progress in phytochrome-\n",
      "mediated light signaling pathways. Trends plant Sci. 20, 641 –650 (2015).\n",
      "31. Leivar, P. & Quail, P. H. PIFs: pivotal components in a cellular signaling hub.\n",
      "Trends Plant Sci. 16\n",
      ",1 9–28 (2011).\n",
      "32. Pham, V. N., Kathare, P. K. & Huq, E. Phytochromes and phytochrome\n",
      "interacting factors. Plant Physiol. 176, 1025 –1038 (2018).\n",
      "33. Paik, I., Kathare, P. K., Kim, J.-I. & Huq, E. Expanding roles of PIFs in signal\n",
      "integration from multiple processes. Mol. Plant 10, 1035 –1046 (2017).\n",
      "34. Pham, V. N., Kathare, P. K. & Huq, E. Dynamic regulation of PIF5 by COP1-\n",
      "SPA complex to optimize photomorphogenesis in Arabidopsis. Plant J. 96,\n",
      "260–273 (2018).\n",
      "35. Oh, E. et al. PIL5, a phytochrome-interacting basic helix-loop-helix protein, is\n",
      "a key negative regulator of seed germination in Arabidopsis thaliana. Plant\n",
      "Cell16, 3045 –3058 (2004).\n",
      "36. Shen, H., Moon, J. & Huq, E. PIF1 is regulated by light-mediated degradation\n",
      "through the ubiquitin-26S proteasome pathway to optimize seedlingphotomorphogenesis in Arabidopsis. Plant J. 44, 1023 –1035 (2005).\n",
      "37. Huq, E. et al. Phytochrome-interacting factor 1 is a critical bHLH regulator of\n",
      "chlorophyll biosynthesis. Science 305, 1937 –1941 (2004).\n",
      "38. Shen, H. et al. Light-induced phosphorylation and degradation of the negative\n",
      "regulator PHYTOCHROME INTERACTING FACTOR 1 depends upon itsdirect physical interactions with photoactivated phytochromes. Plant Cell 20,\n",
      "1586 –1602 (2008).\n",
      "39. Xu, X. et al. PHYTOCHROME INTERACTING FACTOR1 enhances the E3\n",
      "ligase activity of CONSTITUTIVE PHOTOMORPHOGENIC1 tosynergistically repress photomorphogenesis in Arabidopsis. Plant Cell 26,\n",
      "1992 –2006 (2014).\n",
      "40. Xu, X. et al. Reciprocal proteasome-mediated degradation of PIFs and HFR1\n",
      "underlies photomorphogenic development in Arabidopsis. Development 144,\n",
      "1831 –1840 (2017).\n",
      "41. Zhu, L. et al. CUL4 forms an E3 ligase with COP1 and SPA to promote light-\n",
      "induced degradation of PIF1. Nat. Commun. 6, 7245 (2015).\n",
      "42. Bu, Q. et al. Phosphorylation by CK2 enhances the rapid light-induced\n",
      "degradation of PIF1. J. Biol. Chem. 286, 12066 –12074 (2011).\n",
      "43. Holtkotte, X. et al. Mutations in the N-terminal kinase-like domain of the\n",
      "repressor of photomorphogenesis SPA1 severely impair SPA1 function but\n",
      "not light responsiveness in Arabidopsis. Plant J. 88, 205 –218 (2016).\n",
      "44. Yang, J. et al. A conserved Glu –Arg salt bridge connects coevolved motifs\n",
      "that de ﬁne the eukaryotic protein kinase fold. J. Mol. Biol. 415, 666 –\n",
      "679 (2012).\n",
      "45. Ulmer, J. B., Edwards, A. M., McMorris, F. A. & Braun, P. E. Cyclic AMP\n",
      "decreases the phosphorylation state of myelin basic proteins in rat brain cell\n",
      "cultures. J. Biol. Chem. 262, 1748 –1755 (1987).\n",
      "46. Zheng, X. et al. Arabidopsis phytochrome B promotes SPA1 nuclear\n",
      "accumulation to repress photomorphogenesis under far-red light. Plant Cell\n",
      "25, 115 –133 (2013).47. Sharrock, R. A. & Clack, T. Heterodimerization of type II phytochromes in\n",
      "Arabidopsis. Proc. Natl Acad. Sci. USA 101, 11500 –11505 (2004).\n",
      "48. Matsushita, T., Mochizuki, N. & Nagatani, A. Dimers of the N-terminal\n",
      "domain of phytochrome B are functional in the nucleus. Nature 424, 571 –574\n",
      "(2003).\n",
      "49. Yeh, K. C. & Lagarias, J. C. Eukaryotic phytochromes: light-regulated serine/\n",
      "threonine protein kinases with histidine kinase ancestry. Proc. Natl Acad. Sci.\n",
      "USA 95, 13976 –13981 (1998).\n",
      "50. Shin, A.-Y. et al. Evidence that phytochrome functions as a protein kinase in\n",
      "plant light signalling. Nat. Commun. 7, 11545 (2016).\n",
      "51. Buskirk, E. K. V., Reddy, A. K., Nagatani, A. & Chen, M. Photobody\n",
      "localization of phytochrome B is tightly correlated with prolonged and light-dependent inhibition of hypocotyl elongation in the dark. Plant Physiol. 165,\n",
      "595–617 (2014).\n",
      "52. Van Buskirk, E. K., Decker, P. V. & Chen, M. Photobodies in light signaling.\n",
      "Plant Physiol. 158,5 2–60 (2011).\n",
      "53. Reed, J. W., Nagpal, P., Poole, D. S., Furuya, M. & Chory, J. Mutations in the\n",
      "gene for the red/far-red light receptor phytochrome B alter cell elongation and\n",
      "physiological responses throughout Arabidopsis development. Plant Cell 5,\n",
      "147–157 (1993).\n",
      "54. Krall, L. & Reed, J. W. The histidine kinase-related domain participates in\n",
      "phytochrome B function but is dispensable. Proc. Natl. Acad. Sci. USA 97,\n",
      "8169 –8174 (2000).\n",
      "55. Qiu, Y. et al. Mechanism of early light signaling by the carboxy-terminal\n",
      "output module of Arabidopsis phytochrome B. Nature .Communications 8,\n",
      "1905 (2017).\n",
      "56. Jang, I.-C., Henriques, R., Seo, H. S., Nagatani, A. & Chua, N.-H. Arabidopsis\n",
      "PHYTOCHROME INTERACTING FACTOR Proteins PromotePhytochrome B Polyubiquitination by COP1 E3 Ligase in the Nucleus. Plant\n",
      "Cell22, 2370 –2383 (2010).\n",
      "57. Park, E. et al. Phytochrome B inhibits binding of phytochrome-interacting\n",
      "factors to their target promoters. Plant J. 72, 537 –546 (2012).\n",
      "58. Ni, M., Tepperman, J. M. & Quail, P. H. Binding of phytochrome B to its\n",
      "nuclear signalling partner PIF3 is reversibly induced by light. Nature\n",
      "400,\n",
      "781–784 (1999).\n",
      "59. Ni, W. et al. PPKs mediate direct signal transfer from phytochrome\n",
      "photoreceptors to transcription factor PIF3. Nature .Communications 8, 15236\n",
      "(2017).\n",
      "60. Zhang, Y. et al. A quartet of PIF bHLH factors provides a transcriptionally\n",
      "centered signaling hub that regulates seedling morphogenesis through\n",
      "differential expression-patterning of shared target genes in Arabidopsis. Plos\n",
      "Genet. 9, e1003244 (2013).\n",
      "61. Deshaies, R. J. & Joazeiro, C. A. P. RING Domain E3 Ubiquitin Ligases. Annu.\n",
      "Rev. Biochem. 78, 399 –434 (2009).\n",
      "62. Hunter, T. The age of crosstalk: phosphorylation, ubiquitination, and beyond.\n",
      "Mol. Cell 28, 730 –738 (2007).\n",
      "63. Quail, P. H. et al. Phytochromes: photosensory perception and signal\n",
      "transduction. Science 268, 675 –680 (1995).\n",
      "64. Park, E., Kim, Y. & Choi, G. Phytochrome B requires PIF degradation and\n",
      "sequestration to induce light responses across a wide range of light conditions.Plant Cell 30, 1277 –1292 (2018).\n",
      "65. Pfeiffer, A. et al. Interaction with plant transcription factors can mediate\n",
      "nuclear import of phytochrome B. Proc. Natl Acad. Sci. 109, 5892 –5897\n",
      "(2012).\n",
      "66. Bernardo-García, S. et al. BR-dependent phosphorylation modulates PIF4\n",
      "transcriptional activity and shapes diurnal hypocotyl growth. Genes Dev. 28,\n",
      "1681 –1694 (2014).\n",
      "67. Yaron, A. et al. Identi ﬁcation of the receptor component of the Icriptional\n",
      "activity an. Nature 396, 590 (1998).\n",
      "68. Winston, J. T. et al. The SCF, J.T. –ubiquitin ligase complex associates\n",
      "speciﬁcally with phosphorylated destruction motifs in I κBαandβ-catenin and\n",
      "stimulates I κBαubiquitination in vitro. Genes Dev. 13, 270 –283 (1999).\n",
      "69. Shin, A.-Y., Han, Y.-J., Song, P.-S. & Kim, J.-I. Expression of recombinant full-\n",
      "length plant phytochromes assembled with phytochromobilin in Pichia\n",
      "pastoris. FEBS Lett. 588, 2964 –2970 (2014).\n",
      "70. Sato, S., Huq, E., Tepperman, J. M. & Quail, P. H. A light-switchable gene\n",
      "promoter system.\n",
      "Nat. Biotech. 20, 1041 –1044 (2002).\n",
      "71. Moon, J., Zhu, L., Shen, H. & Huq, E. PIF1 directly and indirectly regulates\n",
      "chlorophyll biosynthesis to optimize the greening process in Arabidopsis.\n",
      "Proc. Natl Acad. Sci. USA 105, 9433 –9438 (2008).\n",
      "72. Trapnell, C. et al. Differential gene and transcript expression analysis of RNA-\n",
      "seq experiments with TopHat and Cuf ﬂinks. Nat. Protoc. 7, 562 –578 (2012).\n",
      "73. Langmead, B. & Salzberg, S. L. Fast gapped-read alignment with Bowtie 2. Nat.\n",
      "Methods 9, 357 (2012).\n",
      "74. Anders, S., Pyl, P. T. & Huber, W. HTSeq —a Python framework to work with\n",
      "high-throughput sequencing data. Bioinformatics 31, 166 –169 (2015).\n",
      "75. Love, M. I., Huber, W. & Anders, S. Moderated estimation of fold change and\n",
      "dispersion for RNA-seq data with DESeq2. Genome Biol. 15, 550 (2014).ARTICLE NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y\n",
      "16 NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications76. Gu, Z., Eils, R. & Schlesner, M. Complex heatmaps reveal patterns and\n",
      "correlations in multidimensional genomic data. Bioinformatics 32, 2847 –2849\n",
      "(2016).\n",
      "Acknowledgements\n",
      "We thank Dr. Karen Browning and the members of the Huq laboratory for critical\n",
      "reading of the paper, Akira Nagatani for sharing seeds, Peter Quail for sharing anti-phyBantibody. This work was supported by grants from the National Institute of Health(NIH) (GM-114297) and National Science Foundation (MCB-1543813) to E.H., and\n",
      "Next-Generation BioGreen 21 Program from Rural Development Administration,\n",
      "Republic of Korea (grant no. PJ01325301) to J.K. The authors acknowledge the TexasAdvanced Computing Center (TACC) at The University of Texas at Austin for providingHigh Performance Computing, visualization, and database resources that have con-\n",
      "tributed to the research results reported in this paper.\n",
      "Author contributions\n",
      "I.P., F.C., V.P., L.Z. and E.H. designed experiments. J.K. contributed essential reagent.I.P., F.C., V.P. and L.Z. carried out experiments. I.P., F.C., V.P., L.Z., J.K. and E.H.\n",
      "analyzed data. I.P. and E.H. wrote the paper. I.P., F.C., V.P., L.Z., J.K. and E.H. editedthe paper.\n",
      "Additional information\n",
      "Supplementary Information accompanies this paper at https://doi.org/10.1038/s41467-\n",
      "019-12110-y .Competing interests: The authors declare no competing interests.\n",
      "Reprints and permission information is available online at http://npg.nature.com/\n",
      "reprintsandpermissions/\n",
      "Peer review information Nature Communications thanks the anonymous reviewers for\n",
      "their contribution to the peer review of this work. Peer reviewer reports are available\n",
      "Publisher ’s note Springer Nature remains neutral with regard to jurisdictional claims in\n",
      "published maps and institutional af ﬁliations.\n",
      "Open Access This article is licensed under a Creative Commons Attri-\n",
      "bution 4.0 International License, which permits use, sharing, adaptation,\n",
      "distribution and reproduction in any medium or format, as long as you give appropriatecredit to the original author(s) and the source, provide a link to the Creative Commons\n",
      "license, and indicate if changes were made. The images or other third party material in this\n",
      "article are included in the article ’s Creative Commons license, unless indicated otherwise in\n",
      "a credit line to the material. If material is not included in the article ’sC r e a t i v eC o m m o n s\n",
      "license and your intended use is not permitted by statutory regulation or exceeds thepermitted use, you will need to obtain permission directly from the copyright holder. Toview a copy of this license, visit http://creativecommons.org/licenses/by/4.0/ .\n",
      "© The Author(s) 2019NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-019-12110-y ARTICLE\n",
      "NATURE COMMUNICATIONS |         (2019) 10:4216 | https://doi.org/10.1038/s41467-019-12110-y | www.nature.com/naturecommunications 17\n"
     ]
    }
   ],
   "source": [
    "print(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a variable that will select the selected number of pages\n",
    "pageobj=pdfreader.pages[0]\n",
    " \n",
    "#(x+1) because python indentation starts with 0.\n",
    "#create text variable which will store all text datafrom pdf file\n",
    "text=pageobj.extract_text()\n",
    " \n",
    "#save the extracted data from pdf to a txt file\n",
    "#we will use file handling here\n",
    "#dont forget to put r before you put the file path\n",
    "#go to the file location copy the path by right clicking on the file\n",
    "#click properties and copy the location path and paste it here.\n",
    "#put \"\\\\your_txtfilename\"\n",
    "file1=open(r\"C:\\Users\\ammar\\OneDrive\\Desktop\\phoenixbio\\Phoenix Bioinformatics\\test6.txt\",\"a\", encoding=\"utf-8\")\n",
    "file1.writelines(text)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
