{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from itertools import combinations\n",
    "import random\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASETS_DIR = \"../data/pickles/\"\n",
    "INPUT_DATASET_PATH = \"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\"\n",
    "OUTPUT_TABLE = defaultdict(dict)\n",
    "OUTPUT_PAIR_TUPLES = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       2868                 2365\n",
      "1   total       2868                 2365\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"  \n",
    "groups = load_from_pickle(lloyd_subsets_filename)\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_phenotypes.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared = lambda x,y: len(set(id_to_group_ids[x]).intersection(set(id_to_group_ids[y])))>0\n",
    "pair_list = [shared(x,y) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"phenotype\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a biochemical pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       1769                 1469\n",
      "1     gmx          1                    1\n",
      "2     mtr          2                    2\n",
      "3     osa          3                    3\n",
      "4     sly         18                   18\n",
      "5     zma        185                  160\n",
      "6   total       1978                 1653\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Using the KEGG pathways from the saved groupings object.\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\"                         \n",
    "kegg_groups = load_from_pickle(kegg_pathways_filename)\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(data)\n",
    "# Using the PlantCyc pathways from the saved groupings object.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"  \n",
    "pmn_groups = load_from_pickle(pmn_pathways_filename)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(data)\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "group_mapped_ids = []\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_kegg_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_pmn_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids = list(set(group_mapped_ids))\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_pathways.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared_kegg = lambda x,y: len(set(id_to_kegg_group_ids[x]).intersection(set(id_to_kegg_group_ids[y])))>0\n",
    "shared_pmn = lambda x,y: len(set(id_to_pmn_group_ids[x]).intersection(set(id_to_pmn_group_ids[y])))>0\n",
    "pair_list = [(shared_kegg(x,y) or shared_pmn(x,y)) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"pathways\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"pathways\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"pathways\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"pathways\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether a protein-protein interaction exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       3674                 2950\n",
      "1     mtr         10                   10\n",
      "2     osa         44                   42\n",
      "3     sly          3                    3\n",
      "4     zma        160                  147\n",
      "5   total       3891                 3152\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "data.filter_with_ids(string_data.ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_interactions.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the ProteinInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#interaction = lambda x,y: ((string_data.df[\"from\"] == x) & (string_data.df[\"to\"] == y)).any()\n",
    "#interaction_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(string_data.df[\"from\"].values, string_data.df[\"to\"].values)])\n",
    "#has_interaction = lambda x,y: \"{}:{}\".format(int(x),int(y)) in interaction_strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = string_data.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"interactions\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"interactions\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes are orthologous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath        350                  303\n",
      "1     osa         86                   80\n",
      "2     sly          7                    7\n",
      "3     zma        443                  337\n",
      "4   total        886                  727\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\"\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "data.filter_with_ids(ortholog_edgelist.ids)\n",
    "print(\"\\n\", data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_orthologs.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the AnyInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#ortholog_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(ortholog_edgelist.df[\"from\"].values, ortholog_edgelist.df[\"to\"].values)])\n",
    "#are_orthologs = lambda x,y: \"{}:{}\".format(int(x),int(y)) in ortholog_strings\n",
    "#pair_list = [are_orthologs(x,y) for x,y in list(combinations(data.get_ids(), 2))]\n",
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"positive\"] = pair_list.count(True)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"negative\"] = pair_list.count(False)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes, without enumerating.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = ortholog_edgelist.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"orthologs\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarzing table of the classes for each prediction problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relevant_genes</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "      <th>class_ratio</th>\n",
       "      <th>pos_fraction</th>\n",
       "      <th>neg_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>phenotype</th>\n",
       "      <td>2868</td>\n",
       "      <td>458461</td>\n",
       "      <td>3652817</td>\n",
       "      <td>0.1255</td>\n",
       "      <td>0.111513</td>\n",
       "      <td>0.888487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathways</th>\n",
       "      <td>1978</td>\n",
       "      <td>52530</td>\n",
       "      <td>1902723</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>0.0268661</td>\n",
       "      <td>0.973134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>interactions</th>\n",
       "      <td>3891</td>\n",
       "      <td>450298</td>\n",
       "      <td>7.11575e+06</td>\n",
       "      <td>0.0633</td>\n",
       "      <td>0.0595004</td>\n",
       "      <td>0.940243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orthologs</th>\n",
       "      <td>886</td>\n",
       "      <td>77</td>\n",
       "      <td>391535</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>0.000196401</td>\n",
       "      <td>0.998674</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             relevant_genes positive     negative class_ratio pos_fraction neg_fraction\n",
       "phenotype              2868   458461      3652817      0.1255     0.111513     0.888487\n",
       "pathways               1978    52530      1902723      0.0276    0.0268661     0.973134\n",
       "interactions           3891   450298  7.11575e+06      0.0633    0.0595004     0.940243\n",
       "orthologs               886       77       391535      0.0002  0.000196401     0.998674"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = pd.DataFrame(OUTPUT_TABLE).transpose()\n",
    "summary_df[\"pos_fraction\"] = summary_df[\"positive\"]/(((summary_df[\"relevant_genes\"]**2)-summary_df[\"relevant_genes\"])/2)\n",
    "summary_df[\"neg_fraction\"] = summary_df[\"negative\"]/(((summary_df[\"relevant_genes\"]**2)-summary_df[\"relevant_genes\"])/2)\n",
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phenotype 1</th>\n",
       "      <th>Phenotype 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dwarf; late flowering</td>\n",
       "      <td>abnormal root hairs; increased trichome density</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>abnormal inflorescence stem gravitropism; few ...</td>\n",
       "      <td>complete lack of recognizable stomata; cluster...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>early flowering; early flowering, normal statu...</td>\n",
       "      <td>abnormal leaf morphology; early flowering</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chlorotic cotyledons; no homozygous mutant pla...</td>\n",
       "      <td>increased shoot branching; increased hypocotyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>early flowering; long hypocotyl under blue light</td>\n",
       "      <td>this mutant has delayed flowering compared to ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>embryo defective-transition; embryo defective;...</td>\n",
       "      <td>embryo defective-transition; pale green cotyle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>germination insensitive to aba</td>\n",
       "      <td>increased seed dormancy; sensitive to aba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>intermediate salt sensitive phenotype; sensiti...</td>\n",
       "      <td>plants grown on agar plates have roots that sk...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>complete female gametophyte defective; embryo ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>short telomeres; no visible phenotype</td>\n",
       "      <td>decrease in the total chlorophyll content and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chlorotic leaves under continuous light; dwarf</td>\n",
       "      <td>embryo defective-transition; pale green cotyle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>long hypocotyl; root growth insensitive to pho...</td>\n",
       "      <td>abnormal growth in the dark; anthocyanin accum...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>abnormal circadian rhythms; early flowering in...</td>\n",
       "      <td>few rosette leaves; late flowering</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo lethal; embryo development was arrested...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>slightly elevated levels of photorespiratory i...</td>\n",
       "      <td>no homozygotes were recovered from the heteroz...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>dwarf; elevated chloroplast starch levels</td>\n",
       "      <td>dwarf; late flowering</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>abnormal lateral root formation in response to...</td>\n",
       "      <td>susceptible to viral infection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>short roots</td>\n",
       "      <td>short roots; the hlr-1 (halted root-1) mutant ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>seedlings appear short and thick, compressed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>amorphous, club-like seedlings; reduced growth...</td>\n",
       "      <td>suppressor of superoot2-1 phenotype; reduced a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>few lateral roots; reduced root gravitropism</td>\n",
       "      <td>leaf chlorosis; resistant to bacterial infection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sensitive to aba; slightly delayed germination</td>\n",
       "      <td>no visible phenotype; t-dna insertion lines sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>abnormal sam organization; disorganized rosett...</td>\n",
       "      <td>reduced aba sensitivity, resulting in the foll...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>curled rosette leaves; delayed germination</td>\n",
       "      <td>abnormal seed shape; low germination rate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>abnormal cotyledon morphology; collapsed root tip</td>\n",
       "      <td>pigment defective embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>abnormal seedling root cell patterning; embryo...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>complete loss of circadian rhythms under red l...</td>\n",
       "      <td>sensitive to oryzalin (microtubule-disrupting ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>delayed growth; embryo defective-globular</td>\n",
       "      <td>surface of seedling appears rough due to lack ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>helical growth; right-handed petiole torsions</td>\n",
       "      <td>resistant to several  aminoglycoside antibioti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective-preglobular / globular; embry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>germination sensitive to aba; germination sens...</td>\n",
       "      <td>resistant to 2,4-d. resistant to pcib (toxic a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>male sterile, failed cytokinesis in microsporo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>most of the male gametophyte degenerates befor...</td>\n",
       "      <td>delay in flowering time; lower growth rates un...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>dwarf; thin inflorescence stems</td>\n",
       "      <td>complete loss of cotyledonary petiole; large sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>germination sensitive to aba; seedling growth ...</td>\n",
       "      <td>delayed germination; germination of the mutant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>abnormal flowers; abnormal seed morphology</td>\n",
       "      <td>developmental arrest of mutant embryos occurs ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>increased lateral root number; sensitive to salt</td>\n",
       "      <td>embryo defective-cotyledon; low vitamin c levels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-transition; embryo defective;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>distorted leaves; few rosette leaves</td>\n",
       "      <td>severe albino phenotype; inhibited in chloropl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>dwarf; embryo defective-preglobular</td>\n",
       "      <td>defective interfascicular fiber cells in inflo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>embryo defective-preglobular; enlarged embryo ...</td>\n",
       "      <td>short telomeres; no visible phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>long roots</td>\n",
       "      <td>abnormal leaf morphology; abnormal phyllotaxy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>abnormal pistil; abnormal sepals</td>\n",
       "      <td>complete male gametophyte defective; female ga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>short lateral roots; reduced growth of lateral...</td>\n",
       "      <td>short roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>dwarf; fragile inflorescence stems</td>\n",
       "      <td>inappropriate cortical microtubule reorientati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>slightly smaller meristems; slow flower develo...</td>\n",
       "      <td>reduced apical dominance, organ clustering, al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-transition; embryo defective;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>abnormal glucosinolate composition; elevated f...</td>\n",
       "      <td>abnormal suberin composition in roots; abnorma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>late flowering; late flowering; strong influen...</td>\n",
       "      <td>slightly smaller than the wild type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>germination sensitive to aba; germination sens...</td>\n",
       "      <td>delayed growth during later stages of developm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>susceptible to pseudomonas syringae; delayed l...</td>\n",
       "      <td>reduced aba sensitivity, resulting in the foll...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>abnormal flowers; abnormal seed morphology</td>\n",
       "      <td>male sterile, defective in homolog separation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>altered response to far-red light; abnormal su...</td>\n",
       "      <td>lesions on leaves, systemic acquired resistance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chlorotic leaves at low temperature; elevated ...</td>\n",
       "      <td>elevated amino acid levels at night; low amino...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>abnormal chloroplasts in leaf mesophyll layer;...</td>\n",
       "      <td>mutant kasi plants have multiple morphological...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>embryo defective; gametophyte defective (infer...</td>\n",
       "      <td>early embryo defective (inferred); female game...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>abnormal pectin levels in cell walls; abnormal...</td>\n",
       "      <td>elevated sinapoylglucose levels; low sinapoylc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>insensitive to auxin; reduced root gravitropism</td>\n",
       "      <td>most of the male gametophyte degenerates befor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chlorotic leaves under continuous light; dwarf</td>\n",
       "      <td>hypocotyl hypersensitivity to red light; hypoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>heterozygotes segregate for embryo lethal phen...</td>\n",
       "      <td>heterozygous plants consistently display an al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>crinkled leaves; embryo lethal</td>\n",
       "      <td>arrested chloroplast division; early flowering</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>bent leaf margins; dark green leaves</td>\n",
       "      <td>yellow first leaves</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>albino inflorescence nodes; albino late flowers</td>\n",
       "      <td>pale green leaves; short roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>embryo defective; female gametophyte defective</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "      <td>embryo defective-transition; embryo defective;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>abnormal pollen exine layer (no reduction in f...</td>\n",
       "      <td>abnormal sepal development; fused floral organs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>completely sterile; pollen development arreste...</td>\n",
       "      <td>complete male gametophyte defective; dwarf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>elevated hydrogen peroxide levels under photoo...</td>\n",
       "      <td>germination resistant to uniconazole (inhibito...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>abnormal floral morphology; curled leaves</td>\n",
       "      <td>abnormal petal development; abnormal phyllotaxy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>completely male sterile; dark green</td>\n",
       "      <td>mutant kasi plants have multiple morphological...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>long roots</td>\n",
       "      <td>sensitive to hyperosmotic stress; small root s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>helical growth; right-handed petiole torsions</td>\n",
       "      <td>chlorotic leaves; slow growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>reduced apical hook straightening under far-re...</td>\n",
       "      <td>dwarfed, necrotic lesions under white light; p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>albino plants where total clorophylls and cart...</td>\n",
       "      <td>pigment defective embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>abnormal cauline leaf phyllotaxy; abnormal flo...</td>\n",
       "      <td>dwarf phenotype of ga5-1 plants was rescued by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>dwarf; early leaf senescence</td>\n",
       "      <td>female gametophyte defective; male gametophyte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>complete loss of hypersensitive response in re...</td>\n",
       "      <td>completely male sterile; dark green</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>abnormal cotyledon shape; abnormal veins in co...</td>\n",
       "      <td>dot2-1 mutants are dwarved plants that display...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>wildtype roots; cytokinin-induced partial inhi...</td>\n",
       "      <td>strong allele, embryo defective, lethal; embry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>abnormal senescence; dark green</td>\n",
       "      <td>seedling lethal; sterile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>sensitive to auxin; sensitive to npa (inhibito...</td>\n",
       "      <td>no visible phenotype; exhibited the highest le...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>completely male sterile; dark green</td>\n",
       "      <td>increased branching; short petioles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>abnormal rosette leaf morphology; decreased le...</td>\n",
       "      <td>albino; seedling lethal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>semi-dwarf; sensitive to high light</td>\n",
       "      <td>short hypocotyl in the dark</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>collapsed xylem; dwarf</td>\n",
       "      <td>male and female sterile; defects in meiosis be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>after-ripening not required for germination; g...</td>\n",
       "      <td>etiolated seedlings of both wild type and mtk ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>embryo defective</td>\n",
       "      <td>abnormal suspensor; embryo defective-globular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>abnormal trichome morphology; increased tricho...</td>\n",
       "      <td>abnormal pavement cell morphology; abnormal tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>resistant to fungal infection; semi-dwarf</td>\n",
       "      <td>susceptible to alternaria brassicicola; highly...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>germination sensitive to hydrogen peroxide; in...</td>\n",
       "      <td>this double mutant has a lower chiasma frequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>abnormal vacuolar trafficking</td>\n",
       "      <td>excess absorbed light energy cannot be dissipa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective-preglobular / globular; embry...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>abnormal cotyledon morphology; collapsed root tip</td>\n",
       "      <td>resistant to das734 (herbicide)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>albino; dwarf</td>\n",
       "      <td>complete loss of  fructose accumulation at low...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>dwarf in the dark; no apical hook in the dark</td>\n",
       "      <td>lesions on leaves, systemic acquired resistance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Phenotype 1                                        Phenotype 2\n",
       "0                               dwarf; late flowering    abnormal root hairs; increased trichome density\n",
       "1   abnormal inflorescence stem gravitropism; few ...  complete lack of recognizable stomata; cluster...\n",
       "2   early flowering; early flowering, normal statu...          abnormal leaf morphology; early flowering\n",
       "3   chlorotic cotyledons; no homozygous mutant pla...  increased shoot branching; increased hypocotyl...\n",
       "4    early flowering; long hypocotyl under blue light  this mutant has delayed flowering compared to ...\n",
       "5   embryo defective-transition; embryo defective;...  embryo defective-transition; pale green cotyle...\n",
       "6                      germination insensitive to aba          increased seed dormancy; sensitive to aba\n",
       "7   intermediate salt sensitive phenotype; sensiti...  plants grown on agar plates have roots that sk...\n",
       "8   embryo defective-preglobular; embryo defective...  embryo defective-preglobular; embryo defective...\n",
       "9   embryo defective-globular; embryo defective; g...  complete female gametophyte defective; embryo ...\n",
       "10              short telomeres; no visible phenotype  decrease in the total chlorophyll content and ...\n",
       "11     chlorotic leaves under continuous light; dwarf  embryo defective-transition; pale green cotyle...\n",
       "12  long hypocotyl; root growth insensitive to pho...  abnormal growth in the dark; anthocyanin accum...\n",
       "13  abnormal circadian rhythms; early flowering in...                 few rosette leaves; late flowering\n",
       "14  embryo defective-globular; embryo defective; g...  embryo lethal; embryo development was arrested...\n",
       "15  slightly elevated levels of photorespiratory i...  no homozygotes were recovered from the heteroz...\n",
       "16          dwarf; elevated chloroplast starch levels                              dwarf; late flowering\n",
       "17  abnormal lateral root formation in response to...                     susceptible to viral infection\n",
       "18                                        short roots  short roots; the hlr-1 (halted root-1) mutant ...\n",
       "19  embryo defective-globular; embryo defective; g...  seedlings appear short and thick, compressed i...\n",
       "20  amorphous, club-like seedlings; reduced growth...  suppressor of superoot2-1 phenotype; reduced a...\n",
       "21       few lateral roots; reduced root gravitropism   leaf chlorosis; resistant to bacterial infection\n",
       "22     sensitive to aba; slightly delayed germination  no visible phenotype; t-dna insertion lines sh...\n",
       "23  abnormal sam organization; disorganized rosett...  reduced aba sensitivity, resulting in the foll...\n",
       "24         curled rosette leaves; delayed germination          abnormal seed shape; low germination rate\n",
       "25  abnormal cotyledon morphology; collapsed root tip                           pigment defective embryo\n",
       "26  embryo defective-globular; embryo defective; g...  embryo defective; developmental arrest of muta...\n",
       "27  abnormal seedling root cell patterning; embryo...  embryo defective; developmental arrest of muta...\n",
       "28  complete loss of circadian rhythms under red l...  sensitive to oryzalin (microtubule-disrupting ...\n",
       "29          delayed growth; embryo defective-globular  surface of seedling appears rough due to lack ...\n",
       "30      helical growth; right-handed petiole torsions  resistant to several  aminoglycoside antibioti...\n",
       "31  embryo defective-cotyledon; embryo defective; ...  embryo defective-globular; embryo defective; g...\n",
       "32  embryo defective-globular; embryo defective; g...  embryo defective-preglobular / globular; embry...\n",
       "33  germination sensitive to aba; germination sens...  resistant to 2,4-d. resistant to pcib (toxic a...\n",
       "34  embryo defective-globular; embryo defective; g...  male sterile, failed cytokinesis in microsporo...\n",
       "35  most of the male gametophyte degenerates befor...  delay in flowering time; lower growth rates un...\n",
       "36                    dwarf; thin inflorescence stems   complete loss of cotyledonary petiole; large sam\n",
       "37  germination sensitive to aba; seedling growth ...  delayed germination; germination of the mutant...\n",
       "38         abnormal flowers; abnormal seed morphology  developmental arrest of mutant embryos occurs ...\n",
       "39   increased lateral root number; sensitive to salt   embryo defective-cotyledon; low vitamin c levels\n",
       "40  embryo defective-cotyledon; embryo defective; ...  embryo defective-preglobular; embryo defective...\n",
       "41  embryo defective-cotyledon; embryo defective; ...  embryo defective-transition; embryo defective;...\n",
       "42               distorted leaves; few rosette leaves  severe albino phenotype; inhibited in chloropl...\n",
       "43                dwarf; embryo defective-preglobular  defective interfascicular fiber cells in inflo...\n",
       "44  embryo defective-preglobular; enlarged embryo ...              short telomeres; no visible phenotype\n",
       "45                                         long roots      abnormal leaf morphology; abnormal phyllotaxy\n",
       "46                   abnormal pistil; abnormal sepals  complete male gametophyte defective; female ga...\n",
       "47  short lateral roots; reduced growth of lateral...                                        short roots\n",
       "48                 dwarf; fragile inflorescence stems  inappropriate cortical microtubule reorientati...\n",
       "49  slightly smaller meristems; slow flower develo...  reduced apical dominance, organ clustering, al...\n",
       "50  embryo defective-cotyledon; embryo defective; ...  embryo defective-transition; embryo defective;...\n",
       "51  abnormal glucosinolate composition; elevated f...  abnormal suberin composition in roots; abnorma...\n",
       "52  late flowering; late flowering; strong influen...                slightly smaller than the wild type\n",
       "53  germination sensitive to aba; germination sens...  delayed growth during later stages of developm...\n",
       "54  susceptible to pseudomonas syringae; delayed l...  reduced aba sensitivity, resulting in the foll...\n",
       "55         abnormal flowers; abnormal seed morphology      male sterile, defective in homolog separation\n",
       "56  altered response to far-red light; abnormal su...    lesions on leaves, systemic acquired resistance\n",
       "57  chlorotic leaves at low temperature; elevated ...  elevated amino acid levels at night; low amino...\n",
       "58  abnormal chloroplasts in leaf mesophyll layer;...  mutant kasi plants have multiple morphological...\n",
       "59  embryo defective; gametophyte defective (infer...  early embryo defective (inferred); female game...\n",
       "60  abnormal pectin levels in cell walls; abnormal...  elevated sinapoylglucose levels; low sinapoylc...\n",
       "61    insensitive to auxin; reduced root gravitropism  most of the male gametophyte degenerates befor...\n",
       "62     chlorotic leaves under continuous light; dwarf  hypocotyl hypersensitivity to red light; hypoc...\n",
       "63  heterozygotes segregate for embryo lethal phen...  heterozygous plants consistently display an al...\n",
       "64                     crinkled leaves; embryo lethal     arrested chloroplast division; early flowering\n",
       "65               bent leaf margins; dark green leaves                                yellow first leaves\n",
       "66    albino inflorescence nodes; albino late flowers                     pale green leaves; short roots\n",
       "67     embryo defective; female gametophyte defective  embryo defective; developmental arrest of muta...\n",
       "68  embryo defective-preglobular; embryo defective...  embryo defective-transition; embryo defective;...\n",
       "69  abnormal pollen exine layer (no reduction in f...    abnormal sepal development; fused floral organs\n",
       "70  completely sterile; pollen development arreste...         complete male gametophyte defective; dwarf\n",
       "71  elevated hydrogen peroxide levels under photoo...  germination resistant to uniconazole (inhibito...\n",
       "72  embryo defective-preglobular; embryo defective...  embryo defective; developmental arrest of muta...\n",
       "73          abnormal floral morphology; curled leaves    abnormal petal development; abnormal phyllotaxy\n",
       "74                completely male sterile; dark green  mutant kasi plants have multiple morphological...\n",
       "75                                         long roots  sensitive to hyperosmotic stress; small root s...\n",
       "76      helical growth; right-handed petiole torsions                      chlorotic leaves; slow growth\n",
       "77  reduced apical hook straightening under far-re...  dwarfed, necrotic lesions under white light; p...\n",
       "78  albino plants where total clorophylls and cart...                           pigment defective embryo\n",
       "79  abnormal cauline leaf phyllotaxy; abnormal flo...  dwarf phenotype of ga5-1 plants was rescued by...\n",
       "80                       dwarf; early leaf senescence  female gametophyte defective; male gametophyte...\n",
       "81  complete loss of hypersensitive response in re...                completely male sterile; dark green\n",
       "82  abnormal cotyledon shape; abnormal veins in co...  dot2-1 mutants are dwarved plants that display...\n",
       "83  wildtype roots; cytokinin-induced partial inhi...  strong allele, embryo defective, lethal; embry...\n",
       "84                    abnormal senescence; dark green                           seedling lethal; sterile\n",
       "85  sensitive to auxin; sensitive to npa (inhibito...  no visible phenotype; exhibited the highest le...\n",
       "86                completely male sterile; dark green                increased branching; short petioles\n",
       "87  abnormal rosette leaf morphology; decreased le...                            albino; seedling lethal\n",
       "88                semi-dwarf; sensitive to high light                        short hypocotyl in the dark\n",
       "89                             collapsed xylem; dwarf  male and female sterile; defects in meiosis be...\n",
       "90  after-ripening not required for germination; g...  etiolated seedlings of both wild type and mtk ...\n",
       "91                                   embryo defective      abnormal suspensor; embryo defective-globular\n",
       "92  abnormal trichome morphology; increased tricho...  abnormal pavement cell morphology; abnormal tr...\n",
       "93          resistant to fungal infection; semi-dwarf  susceptible to alternaria brassicicola; highly...\n",
       "94  germination sensitive to hydrogen peroxide; in...  this double mutant has a lower chiasma frequen...\n",
       "95                      abnormal vacuolar trafficking  excess absorbed light energy cannot be dissipa...\n",
       "96  embryo defective-globular; embryo defective; g...  embryo defective-preglobular / globular; embry...\n",
       "97  abnormal cotyledon morphology; collapsed root tip                    resistant to das734 (herbicide)\n",
       "98                                      albino; dwarf  complete loss of  fructose accumulation at low...\n",
       "99      dwarf in the dark; no apical hook in the dark    lesions on leaves, systemic acquired resistance"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Methodology\n",
    "# These are phenotype pairs drawn from the \n",
    "\n",
    "pairs = pd.DataFrame(OUTPUT_PAIR_TUPLES)\n",
    "pairs.columns = [\"Phenotype 1\",\"Phenotype 2\"]\n",
    "pairs[\"Phenotype 1\"] = pairs[\"Phenotype 1\"].str.casefold()\n",
    "pairs[\"Phenotype 2\"] = pairs[\"Phenotype 2\"].str.casefold()\n",
    "pairs[\"Phenotype 1\"] = pairs[\"Phenotype 1\"].map(lambda x: \"; \".join([s[:-1] for s in sent_tokenize(x)[0:2]]))\n",
    "pairs[\"Phenotype 2\"] = pairs[\"Phenotype 2\"].map(lambda x: \"; \".join([s[:-1] for s in sent_tokenize(x)[0:2]]))\n",
    "pairs.to_csv(\"../data/corpus_related_files/phenotype_pairs/unscored.csv\",index=False)\n",
    "pairs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
