{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SARS-2 and Human ACE2 Variant Analysis\n",
    "\n",
    "## Purpose\n",
    "The intent of this pipeline is to identify SNVs in publically available, GISAID submitted SARS-2 genomes that differ from the original Wuhan strain. Identified amino acid variants in the S protein were then modeled for their 3D structure and interactions were predicted agains the Human ACE2 receptor and its reported variants.\n",
    "\n",
    "### 1. What are the mutations within the SARS-2 genome and how prevalent are they?\n",
    "In order to identify the mutations that are in circulation we can use the SARS-2 genomes from GISAID and compare them to the established Wuhan strain (the only RefSeq in NCBI). \n",
    "\n",
    "A caveat: I do not know how these genomes are generated but they are presumably MinION data so performing traidional QC is not possible.\n",
    "\n",
    "From there I can use the MUMMER4 package to identify mutations and their subsequent effect on translation.\n",
    "\n",
    "### 2. Do any of these mutations have an effect on the binding affinity to ACE2?\n",
    "Hin Hark is using these mutations in his modeling. I will ask him for a summary of his methods.\n",
    "\n",
    "### 3. What are the mutations within the ACE2 receptor and do they have any effect on the binding affinity to the SARS-2 spike (s) protein?\n",
    "Data for the mutations were obtained from dbSNP and is using the data from the sources mentioned below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNP Analysis\n",
    "\n",
    "To start we are looking only at the spike protein. Using the SARS-2 reference sequence from NCBI I aligned all of the GISAID CDS in protein-space to the reference sequence. I then filtered the alignments to only the first reading frame and removed the ambiguous bases in the resulting SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Format sequence identifier to play nice with Promer\n",
    "perl -pe 's/^.+EPI_ISL_(\\d+).+/>$1/g'  gisaid_hcov-19_2020_05_04_17.fasta | perl -pe 's/-/n/g' > gisaid_hcov-19_2020_05_04_17.refactored.fasta\n",
    "\n",
    "promer -p SARS-2.refseq.cds.s_prot.promer SARS-2.refseq.cds.s_prot.fasta gisaid_cov2020_sequences.refactored.fasta\n",
    "show-coords -clT SARS-2.refseq.cds.s_prot.promer.delta > SARS-2.refseq.cds.s_prot.promer.coords\n",
    "show-snps -STH gisaid_hcov-19_2020_05_04_17.s_prot.promer.delta < <(awk '$14 == \"1\" && $15 == \"1\" { print $0 }' gisaid_hcov-19_2020_05_04_17.s_prot.promer.coords) | cut -f1-3 | sort | uniq -c | sort -nrb | perl -pe 's/ +/\\t/g' | perl -pe 's/^\\s+//g' > gisaid_hcov-19_2020_05_04_17.s_prot.promer.snps-with-nonsense-mutations.counts.tsv\n",
    "# Removes nonsense mutations\n",
    "# show-snps -ST SARS-2.refseq.cds.s_prot.promer.delta < <(awk '$14 == \"1\" && $15 == \"1\" { print $0 }' SARS-2.refseq.cds.s_prot.promer.coords) | grep -v 'X' > SARS-2.refseq.cds.s_prot.promer.snps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsed the dbSNP into a TSV file with the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Added this into the show-snps one liner above\n",
    "# perl -pe 's/\\n$/\\t/g' snp_result.txt| perl -pe 's/--\\t/\\n/g' | perl -pe 's/\\d+. rs/\\nrs/g' | perl -pe 's/\\r//g' > snp_result.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From there I formatted the data into an extended TSV by adding the frequencies of the mutation per data source and extracted the other relevant info (discarding much of it data provided by dbSNP that I deeded irrelevent at this stage in the analysis)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = []\n",
    "\n",
    "import re\n",
    "\n",
    "def check_snp_src(l, src):\n",
    "    index = [i for i, s in enumerate(l) if src in s if i is not \"\"]\n",
    "    if index:\n",
    "        m = re.search(r'0\\.\\d+',l[index[0]])\n",
    "#         print(l[index[0]])\n",
    "#         print (m.group(0))\n",
    "#         print(m)\n",
    "        if m:\n",
    "            return m.group(0)\n",
    "        else:\n",
    "            return 'NA'\n",
    "    else:\n",
    "        return 'NA'\n",
    "\n",
    "with open('snp_result.tsv', 'r') as f:\n",
    "    for line in f.readlines():\n",
    "        if line == '\\n':\n",
    "            continue\n",
    "        snp = {\n",
    "                'id':'',\n",
    "                'snv':'',\n",
    "                'position':'',\n",
    "                '1000Genomes':'',\n",
    "                'TWINSUK':'',\n",
    "                'GnomAD':'',\n",
    "                'ALSPAC':'',\n",
    "                'TOPMED':''\n",
    "              }\n",
    "        l = line.split('\\t')\n",
    "        snp['id'] = l[0]\n",
    "        snp['snv'] = l[1]\n",
    "        snp['position'] = l[2]\n",
    "        snp['1000Genomes'] = check_snp_src(l,'1000Genomes')\n",
    "        snp['TWINSUK'] = check_snp_src(l,'TWINSUK')\n",
    "        snp['GnomAD'] = check_snp_src(l,'GnomAD')\n",
    "        snp['ALSPAC'] = check_snp_src(l,'ALSPAC')\n",
    "        snp['TOPMED'] = check_snp_src(l,'TOPMED')\n",
    "        \n",
    "#         print(snp)\n",
    "        \n",
    "        snps.append(snp)\n",
    "    \n",
    "# with open('snp_result.cleaned.tsv', 'wb', )\n",
    "df = pandas.DataFrame(snps)\n",
    "df.to_csv('snp_result.cleaned.tsv', sep='\\t', )\n",
    "\n",
    "# print (snps)\n",
    "    \n",
    "#         for i in l:\n",
    "#             if i == l[0]:\n",
    "#                 snp['id'] = l[0]\n",
    "#                 continue\n",
    "#             elif i == l[1]:\n",
    "#                 snp['snv'] = l[1]\n",
    "#                 continue\n",
    "#             elif i == l[2]:\n",
    "#                 snp['position'] = l[2]\n",
    "#                 continue\n",
    "#             elif\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP time course analysis\n",
    "To look at the occurence of mutations throughout the course of the pandemic we need to filter out the sequences that by the month. To do so, we need to link the metatdata with the sequences and the filter per month and run the SNP analysis pipeline for each subsequent month.\n",
    "\n",
    "Also, to make results more reliable, I will implement some QC filtering methods as mentioned in these two papers:\n",
    "https://virological.org/t/issues-with-sars-cov-2-sequencing-data/473\n",
    "https://www.biorxiv.org/content/10.1101/2020.04.26.062422v2.full.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "\n",
    "metadata = pandas.read_csv(\"metadata_2020-10-01_08-36.tsv\", delimiter='\\t', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:4: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "#Remove non-human isolates and those with incomplete collection date info.\n",
    "# Fix entries with no collection day in datetime field\n",
    "metadata['date'] = metadata['date'].str.replace('XX','01')\n",
    "metadata.filtered = metadata.loc[(metadata['host'] == 'Human') & (metadata['date'] != '2020')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>region_exposure</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>pangolin_lineage</th>\n",
       "      <th>GISAID_clade</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>submitting_lab</th>\n",
       "      <th>authors</th>\n",
       "      <th>url</th>\n",
       "      <th>title</th>\n",
       "      <th>paper_url</th>\n",
       "      <th>date_submitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>130280</th>\n",
       "      <td>Wuhan/IPBCAMS-WH-01/2019</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402123</td>\n",
       "      <td>MT019529</td>\n",
       "      <td>2019-12-24</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>B</td>\n",
       "      <td>L</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Lili Ren et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130290</th>\n",
       "      <td>Wuhan/WH01/2019</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_406798</td>\n",
       "      <td>LR757998</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>B</td>\n",
       "      <td>L</td>\n",
       "      <td>General Hospital of Central Theater Command of...</td>\n",
       "      <td>BGI &amp; Institute of Microbiology, Chinese Acade...</td>\n",
       "      <td>Weijun Chen et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>Genomic characterisation and epidemiology of 2...</td>\n",
       "      <td>https://dx.doi.org/10.1016/S0140-6736(20)30251-8</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130274</th>\n",
       "      <td>Wuhan/Hu-1/2019</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402125</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B</td>\n",
       "      <td>L</td>\n",
       "      <td>National Institute for Communicable Disease Co...</td>\n",
       "      <td>National Institute for Communicable Disease Co...</td>\n",
       "      <td>Zhang et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>A new coronavirus associated with human respir...</td>\n",
       "      <td>https://dx.doi.org/10.1038/s41586-020-2008-3</td>\n",
       "      <td>2020-01-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130311</th>\n",
       "      <td>Wuhan/WIV05/2019</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402128</td>\n",
       "      <td>MN996529</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>...</td>\n",
       "      <td>Female</td>\n",
       "      <td>B</td>\n",
       "      <td>L</td>\n",
       "      <td>Wuhan Jinyintan Hospital</td>\n",
       "      <td>Wuhan Institute of Virology, Chinese Academy o...</td>\n",
       "      <td>Peng Zhou et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130312</th>\n",
       "      <td>Wuhan/WIV06/2019</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402129</td>\n",
       "      <td>MN996530</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>B</td>\n",
       "      <td>L</td>\n",
       "      <td>Wuhan Jinyintan Hospital</td>\n",
       "      <td>Wuhan Institute of Virology, Chinese Academy o...</td>\n",
       "      <td>Peng Zhou et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51194</th>\n",
       "      <td>England/NOTT-113555/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_549512</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-20</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B.1</td>\n",
       "      <td>G</td>\n",
       "      <td>Queens Medical Centre, Clinical Microbiology D...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) Consortium</td>\n",
       "      <td>Gemma Clark et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51195</th>\n",
       "      <td>England/NOTT-113564/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_549513</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-20</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B.1.79</td>\n",
       "      <td>G</td>\n",
       "      <td>Queens Medical Centre, Clinical Microbiology D...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) Consortium</td>\n",
       "      <td>Gemma Clark et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51196</th>\n",
       "      <td>England/NOTT-113573/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_549514</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-20</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B.1</td>\n",
       "      <td>G</td>\n",
       "      <td>Queens Medical Centre, Clinical Microbiology D...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) Consortium</td>\n",
       "      <td>Gemma Clark et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51197</th>\n",
       "      <td>England/NOTT-113582/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_549515</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-20</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B.1</td>\n",
       "      <td>G</td>\n",
       "      <td>Queens Medical Centre, Clinical Microbiology D...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) Consortium</td>\n",
       "      <td>Gemma Clark et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51198</th>\n",
       "      <td>England/NOTT-113591/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_549516</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-20</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>GR</td>\n",
       "      <td>Queens Medical Centre, Clinical Microbiology D...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) Consortium</td>\n",
       "      <td>Gemma Clark et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-09-28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128826 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          strain virus  gisaid_epi_isl genbank_accession  \\\n",
       "130280  Wuhan/IPBCAMS-WH-01/2019  ncov  EPI_ISL_402123          MT019529   \n",
       "130290           Wuhan/WH01/2019  ncov  EPI_ISL_406798          LR757998   \n",
       "130274           Wuhan/Hu-1/2019  ncov  EPI_ISL_402125          MN908947   \n",
       "130311          Wuhan/WIV05/2019  ncov  EPI_ISL_402128          MN996529   \n",
       "130312          Wuhan/WIV06/2019  ncov  EPI_ISL_402129          MN996530   \n",
       "...                          ...   ...             ...               ...   \n",
       "51194   England/NOTT-113555/2020  ncov  EPI_ISL_549512                 ?   \n",
       "51195   England/NOTT-113564/2020  ncov  EPI_ISL_549513                 ?   \n",
       "51196   England/NOTT-113573/2020  ncov  EPI_ISL_549514                 ?   \n",
       "51197   England/NOTT-113582/2020  ncov  EPI_ISL_549515                 ?   \n",
       "51198   England/NOTT-113591/2020  ncov  EPI_ISL_549516                 ?   \n",
       "\n",
       "              date  region         country division location region_exposure  \\\n",
       "130280  2019-12-24    Asia           China    Hubei    Wuhan            Asia   \n",
       "130290  2019-12-26    Asia           China    Hubei    Wuhan            Asia   \n",
       "130274  2019-12-26    Asia           China    Hubei    Wuhan            Asia   \n",
       "130311  2019-12-30    Asia           China    Hubei    Wuhan            Asia   \n",
       "130312  2019-12-30    Asia           China    Hubei    Wuhan            Asia   \n",
       "...            ...     ...             ...      ...      ...             ...   \n",
       "51194   2020-09-20  Europe  United Kingdom  England      NaN          Europe   \n",
       "51195   2020-09-20  Europe  United Kingdom  England      NaN          Europe   \n",
       "51196   2020-09-20  Europe  United Kingdom  England      NaN          Europe   \n",
       "51197   2020-09-20  Europe  United Kingdom  England      NaN          Europe   \n",
       "51198   2020-09-20  Europe  United Kingdom  England      NaN          Europe   \n",
       "\n",
       "        ...     sex pangolin_lineage GISAID_clade  \\\n",
       "130280  ...    Male                B            L   \n",
       "130290  ...    Male                B            L   \n",
       "130274  ...       ?                B            L   \n",
       "130311  ...  Female                B            L   \n",
       "130312  ...    Male                B            L   \n",
       "...     ...     ...              ...          ...   \n",
       "51194   ...       ?              B.1            G   \n",
       "51195   ...       ?           B.1.79            G   \n",
       "51196   ...       ?              B.1            G   \n",
       "51197   ...       ?              B.1            G   \n",
       "51198   ...       ?            B.1.1           GR   \n",
       "\n",
       "                                          originating_lab  \\\n",
       "130280  Institute of Pathogen Biology, Chinese Academy...   \n",
       "130290  General Hospital of Central Theater Command of...   \n",
       "130274  National Institute for Communicable Disease Co...   \n",
       "130311                           Wuhan Jinyintan Hospital   \n",
       "130312                           Wuhan Jinyintan Hospital   \n",
       "...                                                   ...   \n",
       "51194   Queens Medical Centre, Clinical Microbiology D...   \n",
       "51195   Queens Medical Centre, Clinical Microbiology D...   \n",
       "51196   Queens Medical Centre, Clinical Microbiology D...   \n",
       "51197   Queens Medical Centre, Clinical Microbiology D...   \n",
       "51198   Queens Medical Centre, Clinical Microbiology D...   \n",
       "\n",
       "                                           submitting_lab            authors  \\\n",
       "130280  Institute of Pathogen Biology, Chinese Academy...     Lili Ren et al   \n",
       "130290  BGI & Institute of Microbiology, Chinese Acade...  Weijun Chen et al   \n",
       "130274  National Institute for Communicable Disease Co...        Zhang et al   \n",
       "130311  Wuhan Institute of Virology, Chinese Academy o...    Peng Zhou et al   \n",
       "130312  Wuhan Institute of Virology, Chinese Academy o...    Peng Zhou et al   \n",
       "...                                                   ...                ...   \n",
       "51194            COVID-19 Genomics UK (COG-UK) Consortium  Gemma Clark et al   \n",
       "51195            COVID-19 Genomics UK (COG-UK) Consortium  Gemma Clark et al   \n",
       "51196            COVID-19 Genomics UK (COG-UK) Consortium  Gemma Clark et al   \n",
       "51197            COVID-19 Genomics UK (COG-UK) Consortium  Gemma Clark et al   \n",
       "51198            COVID-19 Genomics UK (COG-UK) Consortium  Gemma Clark et al   \n",
       "\n",
       "                           url  \\\n",
       "130280  https://www.gisaid.org   \n",
       "130290  https://www.gisaid.org   \n",
       "130274  https://www.gisaid.org   \n",
       "130311  https://www.gisaid.org   \n",
       "130312  https://www.gisaid.org   \n",
       "...                        ...   \n",
       "51194   https://www.gisaid.org   \n",
       "51195   https://www.gisaid.org   \n",
       "51196   https://www.gisaid.org   \n",
       "51197   https://www.gisaid.org   \n",
       "51198   https://www.gisaid.org   \n",
       "\n",
       "                                                    title  \\\n",
       "130280                                                  ?   \n",
       "130290  Genomic characterisation and epidemiology of 2...   \n",
       "130274  A new coronavirus associated with human respir...   \n",
       "130311                                                  ?   \n",
       "130312                                                  ?   \n",
       "...                                                   ...   \n",
       "51194                                                   ?   \n",
       "51195                                                   ?   \n",
       "51196                                                   ?   \n",
       "51197                                                   ?   \n",
       "51198                                                   ?   \n",
       "\n",
       "                                               paper_url date_submitted  \n",
       "130280                                                 ?     2020-01-11  \n",
       "130290  https://dx.doi.org/10.1016/S0140-6736(20)30251-8     2020-01-30  \n",
       "130274      https://dx.doi.org/10.1038/s41586-020-2008-3     2020-01-12  \n",
       "130311                                                 ?     2020-01-18  \n",
       "130312                                                 ?     2020-01-18  \n",
       "...                                                  ...            ...  \n",
       "51194                                                  ?     2020-09-28  \n",
       "51195                                                  ?     2020-09-28  \n",
       "51196                                                  ?     2020-09-28  \n",
       "51197                                                  ?     2020-09-28  \n",
       "51198                                                  ?     2020-09-28  \n",
       "\n",
       "[128826 rows x 26 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The earliest is 2019-12 and the latest is 2020-09\n",
    "metadata.filtered.sort_values('date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n"
     ]
    }
   ],
   "source": [
    "# Making sure dates are in the proper format\n",
    "\n",
    "\n",
    "# from dateutil.parser import parse\n",
    "# print(metadata.filtered.loc[0,].date)\n",
    "# print(parse(metadata.filtered.loc[0,].date))\n",
    "\n",
    "# print(metadata.filtered.sort_values('date').date.unique())\n",
    "\n",
    "metadata.filtered['date'] = pandas.to_datetime(metadata.filtered['date'])\n",
    "metadata.filtered['month_year'] = metadata.filtered['date'].dt.strftime('%Y-%m')\n",
    "dates = metadata.filtered.sort_values('month_year')['month_year'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wuhan/HBCDC-HB-01/2019\n",
      "Wuhan/HBCDC-HB-02/2019\n",
      "Wuhan/HBCDC-HB-03/2019\n",
      "Wuhan/HBCDC-HB-04/2019\n",
      "Wuhan/Hu-1/2019\n",
      "Wuhan/IME-WH01/2020\n",
      "Wuhan/IME-WH02/2020\n",
      "Wuhan/IME-WH03/2020\n",
      "Wuhan/IME-WH04/2020\n",
      "Wuhan/IME-WH05/2020\n",
      "Wuhan/IPBCAMS-WH-01/2019\n",
      "Wuhan/IPBCAMS-WH-02/2019\n",
      "Wuhan/IPBCAMS-WH-03/2019\n",
      "Wuhan/IPBCAMS-WH-04/2019\n",
      "Wuhan/IVDC-HB-01/2019\n",
      "Wuhan/IVDC-HB-05/2019\n",
      "Wuhan/IVDC-HB-GX02/2019\n",
      "Wuhan/WH01/2019\n",
      "Wuhan/WH02/2019\n",
      "Wuhan/WIV02/2019\n",
      "Wuhan/WIV04/2019\n",
      "Wuhan/WIV05/2019\n",
      "Wuhan/WIV06/2019\n",
      "Wuhan/WIV07/2019\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n\".join(metadata.filtered['strain'].loc[metadata.filtered['month_year'] <= date].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "for date in dates:\n",
    "#     print(date)\n",
    "#     print(len(metadata.filtered['strain'].loc[metadata.filtered['month_year'] <= date]))\n",
    "    with open('covid_'+date+'_seqs.txt', 'w') as f:\n",
    "        f.write('\\n'.join(metadata.filtered['strain'].loc[metadata.filtered['month_year'] <= date].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ran on system\n",
    "# Grabbing all sequences for each timepoint\n",
    "for i in $(ls covid*.txt | perl -pe 's/\\.txt//g'); do \n",
    "    seqtk subseq sequences_2020-10-01_07-14.fasta $i.txt >! $i.fasta ; \n",
    "    \n",
    "    promer -p $i.promer.s_prot SARS-2.refseq.cds.s_prot.fasta $i.fasta\n",
    "    show-coords -clT $i.promer.s_prot.delta > $i.promer.s_prot.coords\n",
    "    show-snps -STH $i.promer.s_prot.delta < <(awk '$14 == \"1\" && $15 == \"1\" { print $0 }' $i.promer.s_prot.coords) | cut -f1-3 | sort | uniq -c | sort -nrb | perl -pe 's/ +/\\t/g' | perl -pe 's/^\\s+//g' > $i.promer.s_prot.snps-with-nonsense-mutations.counts.tsv\n",
    "    # Removes nonsense mutations\n",
    "    # show-snps -ST SARS-2.refseq.cds.promer.delta < <(awk '$14 == \"1\" && $15 == \"1\" { print $0 }' SARS-2.refseq.cds.s_prot.promer.coords) | grep -v 'X' > SARS-2.refseq.cds.s_prot.promer.snps\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
