{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "248a91bb",
   "metadata": {},
   "source": [
    "$$\n",
    "\\def\\CC{\\bf C}\n",
    "\\def\\QQ{\\bf Q}\n",
    "\\def\\RR{\\bf R}\n",
    "\\def\\ZZ{\\bf Z}\n",
    "\\def\\NN{\\bf N}\n",
    "$$\n",
    "<img src=\"https://bitbucket.org/charade/elsa/raw/master/doc/images/elsa_logo.png\" width=\"100\" height=\"50\" alt=\"logo.png\" />\n",
    "\n",
    "# ELSA - Finding Time-Dependent Associations (Now Augmented with Liquid Associaiton)\n",
    "\n",
    "## QUICK LINKS\n",
    "\n",
    "[Extended Local Similarity Analysis (ELSA)\n",
    "Manuals](https://bitbucket.org/charade/elsa/wiki/Manual)\n",
    "\n",
    "[Extended Liquid Associaiton Analysis (ELA)\n",
    "Manuals](https://bitbucket.org/charade/elsa/wiki/Manual_ela)\n",
    "\n",
    "[FAQ](https://bitbucket.org/charade/elsa/wiki/FAQ)\n",
    "\n",
    "## INSTRODUCTION\n",
    "\n",
    "In recent years, advances in molecular technologies have empowered\n",
    "researchers with the ability to spatially and temporally characterize\n",
    "natural microbial communities without lab cultivation. Mining and\n",
    "analyzing co-occurrence patterns in these new datasets are fundamental\n",
    "to revealing existing symbiosis relations and microbe-environment\n",
    "interactions. Time series data, in particular, are receiving more and\n",
    "more attention, since not only undirected but also directed associations\n",
    "can be inferred from these datasets.\n",
    "\n",
    "Researchers typically use techniques like principal component analysis\n",
    "(PCA), multidimensional scaling (MDS), discriminant function analysis\n",
    "(DFA) and canonical correlation analysis (CCA)) to analyze microbial\n",
    "community data under various conditions. Different from these methods,\n",
    "the Local Similarity Analysis (LSA) technique is unique to capture the\n",
    "time-dependent associations (possibly time-shifted) between microbes and\n",
    "between microbe and environmental factors. Significant LSA associations\n",
    "can be interpreted as a partially directed association network for\n",
    "further network-based analysis.\n",
    "\n",
    "ALso, another limitation of previous investigations was their\n",
    "restriction toward pairwise relationships, while there was no lack of\n",
    "abundance of third-party mediated functionalities in microbial\n",
    "community. Liquid Association analysis was originally developed by K.C.\n",
    "Li for characterizing the internal evolution of expression pattern of a\n",
    "pair of genes (ùëã, ùëå) depending on a ‚Äòscouting‚Äô (mediator) gene Z. In the\n",
    "newly implemented ELSA, we extended Liquid Association and combined it\n",
    "with LSA to explore the mediated co-varying microbial dynamics, which\n",
    "provides a novel perspective and a useful toolkit to analyze and\n",
    "interpret microbial community time series data.\n",
    "\n",
    "## METHODS\n",
    "\n",
    "<img src=\"https://bitbucket.org/charade/elsa/raw/master/doc/images/elsa_pipeline.png\" width=\"540\" height=\"450\" alt=\"lsa_pipeline.png\" />\n",
    "\n",
    "Figure 1. The analysis workflow of Local Similarity Analysis (LSA)\n",
    "tools. Users start with raw data (matrices of time series) as input and\n",
    "specify their requirements as parameters. The LSA tools subsequently\n",
    "F-transform and normalize the raw data and then calculate the Local\n",
    "Similarity (LS) Scores and the Pearson‚Äôs Correlation Coefficients. The\n",
    "tools then assess the statistical significance (P-values) of these\n",
    "correlation statistics using permutation test and filter out\n",
    "insignificant results. Finally, the tools construct a partially directed\n",
    "association network from significant associations.\n",
    "\n",
    "<img src=\"https://bitbucket.org/charade/elsa/raw/master/doc/images/ela_pipeline.png\" width=\"540\" height=\"450\" alt=\"ela_pipeline.png\" />\n",
    "\n",
    "Figure 2. Liquid Association / Mediated correlation and example\n",
    "Cytoscape diagrams for all liquid association types of factors X, Y and\n",
    "Z: (A) High Z level enhances the positive correlation between X and Y;\n",
    "(B) Low Z level enhances the negative correlation between X and Y; (C)\n",
    "Low Z level enhances the positive correlation between X and Y; (D) High\n",
    "Z level enhances the negative correlation between X and Y. And (E) A\n",
    "flowchart for incorporating Liquid Association (LA) with Local\n",
    "Similarity (LS) Analysis (LSA). First we use LSA to find candidate local\n",
    "and without time-delayed associations between factors X and Y. The\n",
    "results were filtered based on p-values, q-values and effect (LS score).\n",
    "Then, given the significant LSA factors X and Y, we compute LA score to\n",
    "scout any environmental/OTU factors to discover potential mediating\n",
    "factor Z. Next, a permutation test for liquid association is performed\n",
    "and the results were filtered based on p-values, q-values and effect (LA\n",
    "score) to remove insignificant triplets. Finally, we use the software\n",
    "Cytoscape to visualize the results.\n",
    "\n",
    "## SOFTWARE\n",
    "\n",
    "> Extended Local Similarity Analysis(ELSA)\n",
    ">\n",
    "> Currently the package is maintained only for Linux (Ubuntu). For other\n",
    "> platforms, see the DOCKER section for use.\n",
    ">\n",
    "> More current information of this package is available at:\n",
    "> <http://github.com/labxscut/elsa>\n",
    "\n",
    "## DOCKER\n",
    "\n",
    "> A Dockerfile is provided to build elsa enabled docker image from a\n",
    "> standard Ubuntu docker image. To build a docker image using 'docker\n",
    "> build \\$ELSAPKG', where \\$ELSAPKG is the unzipped path of elsa.\n",
    ">\n",
    "> > ```{.python .input}\n",
    "> > host> docker build --no-cache -t labx-elsa .\n",
    "> > ```\n",
    ">\n",
    "> Name the built container as your:container; Then mount current data\n",
    "> directory to /var/data accessible by docker:\n",
    ">\n",
    "> > ```{.python .input}\n",
    "> > host> docker run --name elsa -it -v `pwd` :/var/data/ labx-elsa \n",
    "> > elsa> lsa_compute --help\n",
    "> > ```\n",
    "\n",
    "## INSTALL\n",
    "\n",
    "> (1). \\# for use \\#\n",
    ">\n",
    "> First, install the prerequisites: C++ (build-essential), Python(dev),\n",
    "> conda, numpy, scipy and matplotlib as specified in setup.py, create\n",
    "> and activate a virtual env named elsa.\n",
    ">\n",
    "> Same for Linux, Mac OS, Windows Subsystem Linux, with a Conda-like\n",
    "> Virtual Env\n",
    ">\n",
    "> Then, download and unzip the latest master branch to the elsa folder.\n",
    ">\n",
    "> ```{.python .input}\n",
    "> elsa>  pip install .                            # or python setup.py install\n",
    "> elsa>  cd test && . test.sh                     # a test script is available\n",
    "> ```\n",
    ">\n",
    "> (2). \\# for development \\#\n",
    ">\n",
    "> eLSA is open source and your contributions are greatly welcome.\n",
    ">\n",
    "> First, use git to fork a copy of elsa on github.com:\n",
    ">\n",
    "> ```{.python .input}\n",
    "> elsa> git clone ssh://git@github.com/your/elsa elsa\n",
    "> ```\n",
    ">\n",
    "> Then, make your edits and create a pull request to merge back.\n",
    "\n",
    "## EXECUTABLES\n",
    "\n",
    "> 1.  The following executables will be available from your python\n",
    ">     scripts directory (typically already in \\$PATH).\n",
    "> 2.  Use '\\$script\\_name -h' for its usage.\n",
    "> 3.  A few examples are available in 'test/test.sh' and explained\n",
    ">     there.\n",
    ">\n",
    "> ```{.python .input}\n",
    "> lsa_compute                       # main script for LSA/LTA computation to generat the .lsa result\n",
    "> lsa_version                       # display the git commit hash, i.e. version, of installed esla\n",
    ">\n",
    "> lsa_chkdat                        # provided as is, check the input file's format compatibility\n",
    "> lsa_query                         # provided as is, query .lsa result and generate network (requires rpy)\n",
    "> lsa_infer                         # provided as is, plot pairwise .lsa result (requires rpy)\n",
    "> lsa_sim                           # provided as is, simulate a pair of time series\n",
    "> lsa_para                          # provided as is, generate parallel jobs for a large input file \n",
    "> lsa_fixqv                         # provided as is, fix the q-values for the merged parallel .lsa results \n",
    "> lsa_totrend                       # provided as is, convert original series to trend series\n",
    "> ```\n",
    "\n",
    "## NOTES\n",
    "\n",
    "> The historical R version of LSA is no longer offered, all its\n",
    "> computation capacity is avaiblable through eLSA.\n",
    "\n",
    "## CONTACT\n",
    "\n",
    "> lcxia at scut dot edu dot cn\n",
    "\n",
    "## CITATIONS\n",
    "\n",
    "Please cite the references 1, 2, 3, 4 and 5 if the LSA or LTA algorithms\n",
    "and statistical theories was used or reviewed in your work. Please cite\n",
    "the references 2 and 3 if any scripts of the ELSA software package was\n",
    "used in your work. Please also cite the reference 4 if local trend\n",
    "analysis (LTA) scripts was used in your work. Please also cite the\n",
    "reference 6 if local liquid association analysis (LLA) scripts was used\n",
    "in your work.\n",
    "\n",
    "> 1.  1.  Dongmei Ai, Lulu Chen, Jiemin Xie, Longwei Cheng, Fang Zhang,\n",
    ">         Yihui Luan, Yang Li, Shengwei Hou, Fengzhu Sun, Li Charlie\n",
    ">         Xia. Identifying local associations in biological time series:\n",
    ">         algorithms, statistical significance, and applications.\n",
    ">         Briefings in Bioinformatics 2023, 24(6):bbad390.\n",
    ">         (<https://doi.org/10.1093/bib/bbad390>)\n",
    ">\n",
    "> 2.  1.  Li C Xia, Dongmei Ai, Jacob Cram, Jed A Fuhrman, Fengzhu Sun.\n",
    ">         Efficient Statistical Significance Approximation for Local\n",
    ">         Association Analysis of High-Throughput Time Series Data.\n",
    ">         Bioinformatics 2013, 29(2):230-237.\n",
    ">         (<https://doi.org/10.1093/bioinformatics/bts668>)\n",
    ">\n",
    "> 3.  1.  Li C Xia, Joshua A Steele, Jacob A Cram, Zoe G Cardon, Sheri L\n",
    ">         Simmons, Joseph J Vallino, Jed A Fuhrman and Fengzhu Sun.\n",
    ">         Extended local similarity analysis (eLSA) of microbial\n",
    ">         community and other time series data with replicates. BMC\n",
    ">         Systems Biology 2011, 5(S2):S15\n",
    ">         (<https://doi.org/10.1186/1752-0509-5-S2-S15>)\n",
    ">\n",
    "> 4.  1.  Li C Xia, Dongmei Ai, Jacob Cram, Xiaoyi Liang, Jed Fuhrman,\n",
    ">         Fengzhu Sun. Statistical significance approximation in local\n",
    ">         trend analysis of high-throughput time-series data using the\n",
    ">         theory of Markov chains. BMC Bioinformatics 2015, 16, 301\n",
    ">         (<https://doi.org/10.1186/s12859-015-0732-8>)\n",
    ">\n",
    "> 5.  1.  Quansong Ruan, Debojyoti Dutta, Michael S. Schwalbach,\n",
    ">         Joshua A. Steele, Jed A. Fuhrman and Fengzhu Sun Local\n",
    ">         similarity analysis reveals unique associations among marine\n",
    ">         bacterioplankton species and environmental factors\n",
    ">         Bioinformatics 2006, 22(20):2532-2538\n",
    ">\n",
    "> 6.  1.  Dongmei Ai, Xiaoxin Li, Hongfei Pan, Li Charlie Xia\\#. Explore\n",
    ">         Mediated Co-varying Dynamics in Microbial Community using\n",
    ">         Integrated Local Similarity and Liquid Association Analysis.\n",
    ">         Accepted bu APBC, to appear in BMC Genomics (2019)."
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
