{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot\n",
    "import seaborn\n",
    "from tqdm import tqdm\n",
    "from scipy.stats import ttest_ind\n",
    "from collections import Counter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_duplicates(input_list):\n",
    "    procd = [''.join(i.split()).lower() for i in input_list]\n",
    "    counts = Counter(input_list)\n",
    "    dups = [k for k,v in counts.items() if v > 1]\n",
    "#     print(dups)\n",
    "    return dups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in raw data\n",
    "enzyme_order_df = pd.read_excel(\"data/initial_enzymes_1.xlsx\")\n",
    "sp_df = pd.read_excel(\"data/sp_prot_translations.xls\")\n",
    "\n",
    "excel_raw = pd.ExcelFile(\"data/191203_mastersheet_BGsubtracted_raw.xls\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheet_names = ['mastersheet',\n",
    "    'amylase_1',\n",
    "    'amylase_2',\n",
    "    'amylase_1_10x',\n",
    "    'amylase_2_10x',\n",
    "    'protease_1',\n",
    "    'protease_2',\n",
    "    'xylanase_1',\n",
    "    'xylanase_1_50x',\n",
    "    'lipase_1',\n",
    "    'lipase_2',\n",
    "    'positive_1',\n",
    "    'positive_2',\n",
    "    'positive_amy_10x',\n",
    "    'positive_xyl_50x',\n",
    "    'negative']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_few = ['Glyc stock plate', 'Well location', 'Planned Protein', 'Planned SP', 'Sequence Name', 'Correct Construct', 'Full Sequencing', 'SP start', 'SP end', 'SP coverage', 'SP actual DNA', 'SP expected DNA', 'SP expected name', 'Protein start', 'Protein end', 'Protein coverage', 'Protein actual DNA', 'Protein expected DNA', 'Protein expected name', 'Unnamed: 19', 'SP expected name.1']\n",
    "next_few = ['Glyc stock plate', 'Well location', 'Planned Protein', 'Planned SP', 'Sequence Name', 'Correct Construct', 'Full Sequencing', 'SP start', 'SP end', 'SP coverage', 'SP actual DNA', 'SP expected DNA', 'SP expected name', 'Protein start', 'Protein end', 'Protein coverage', 'Protein actual DNA', 'Protein expected DNA', 'Protein expected name', 'SP expected name.1', 'Unnamed: 20', 'Assay1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for duplicate columns\n",
    "# actually this is built in to pandas, SP.expected name is doubled\n",
    "# Checking for correct column names\n",
    "\n",
    "# Amylase sheets have an extra column\n",
    "amylase_firsts = ['Glyc stock plate', 'Well location', 'Planned Protein', 'Planned SP', 'Sequence Name', 'Correct Construct', 'Full Sequencing', 'SP start', 'SP end', 'SP coverage', 'SP actual DNA', 'SP expected DNA', 'SP expected name', 'Protein start', 'Protein end', 'Protein coverage', 'Protein actual DNA', 'Protein expected DNA', 'Protein expected name', 'Unnamed: 19', 'SP expected name.1']\n",
    "other_firsts = ['Glyc stock plate', 'Well location', 'Planned Protein', 'Planned SP', 'Sequence Name', 'Correct Construct', 'Full Sequencing', 'SP start', 'SP end', 'SP coverage', 'SP actual DNA', 'SP expected DNA', 'SP expected name', 'Protein start', 'Protein end', 'Protein coverage', 'Protein actual DNA', 'Protein expected DNA', 'Protein expected name', 'SP expected name.1', 'Unnamed: 20', 'Assay1']\n",
    "\n",
    "def get_duplicates(input_list):\n",
    "    procd = [''.join(i.split()).lower() for i in input_list]\n",
    "    counts = Counter(input_list)\n",
    "    dups = [k for k,v in counts.items() if v > 1]\n",
    "    return dups\n",
    "\n",
    "all_cols = []\n",
    "\n",
    "# Iterate through excel sheets, grab columns and check for correct data\n",
    "for i in range(1,len(sheet_names)): # skip master sheet\n",
    "    _df = excel_raw.parse(excel_raw.sheet_names[i])\n",
    "    a = list(_df.columns)\n",
    "    dups = get_duplicates(a)\n",
    "    all_cols = all_cols + a\n",
    "#     print(sheet_names[i])\n",
    "    if i <= 4:\n",
    "        assert a[:len(amylase_firsts)] == amylase_firsts, 'incorrect first columns'\n",
    "#         print(a[len(amylase_firsts):])\n",
    "    else:\n",
    "        assert a[:len(other_firsts)] == other_firsts, 'incorrect first columns'\n",
    "#         print(a[len(other_firsts):])\n",
    "#     if 'Assay5' in a:\n",
    "#         print(sheet_names[i])\n",
    "#         print(a[len(other_firsts):])\n",
    "#     print('----')\n",
    "    assert len(dups) == 0, \"Error: Duplicate column names\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_cols = {'Unnamed: 19': 'blank_col',\n",
    "            'Unnamed: 21': 'id_concatd',\n",
    "            'Unnamed: 23': 'assay_1_value',\n",
    "            'Unnamed: 25': 'assay_2_value',\n",
    "            'Unnamed: 27': 'assay_3_value',\n",
    "            'Unnamed: 29': 'assay_4_value',\n",
    "            'Assay4-BG': 'Assay4- BG',    \n",
    "}\n",
    "\n",
    "# Data after amylase 2 doesn't have blank column 19 \n",
    "new_cols_2 = {\n",
    "            'Unnamed: 20': 'id_concatd',\n",
    "            'Unnamed: 22': 'assay_1_value',\n",
    "            'Unnamed: 24': 'assay_2_value',\n",
    "            'Unnamed: 26': 'assay_3_value',\n",
    "            'Unnamed: 28': 'assay_4_value',\n",
    "            'Assay4-BG': 'Assay4- BG',    \n",
    "}\n",
    "\n",
    "# Because the single blank line in amylase only offsets the columns by 1, \n",
    "# and there is an identifier column spacing out assay values\n",
    "# can merge column renaming dictionaries and process in one go\n",
    "new_cols.update(new_cols_2)\n",
    "\n",
    "# There's also a 5th assay completed for the xylanases we'll deal with later:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "amylase_1 96\n",
      "amylase_2 91\n",
      "amylase_1_10x 96\n",
      "amylase_2_10x 91\n",
      "protease_1 96\n",
      "protease_2 32\n",
      "xylanase_1 96\n",
      "xylanase_1_50x 96\n",
      "lipase_1 96\n",
      "lipase_2 96\n",
      "positive_1 96\n",
      "positive_2 51\n",
      "positive_amy_10x 32\n",
      "positive_xyl_50x 30\n",
      "negative 64\n"
     ]
    }
   ],
   "source": [
    "# Merge all data into tidy spreadsheet\n",
    "\n",
    "df = pd.DataFrame()\n",
    "rowsum = 0\n",
    "for sheet_index in range(1, len(sheet_names)):  # skip master sheet\n",
    "    \n",
    "    col_update = new_cols.copy()\n",
    "    # Deal with Assay 5 in Xylanase\n",
    "    if sheet_names[sheet_index] == 'xylanase_1':\n",
    "        col_update.update({'Unnamed: 34': 'assay_5_value'})\n",
    "#         print(col_update)\n",
    "    \n",
    "    _df = excel_raw.parse(excel_raw.sheet_names[sheet_index]).dropna(thresh=5)  # protease_1 has extra rows\n",
    "    _df.rename(columns=col_update, inplace=True)  \n",
    "    _df['run_label'] = sheet_names[sheet_index]\n",
    "    print(sheet_names[sheet_index], len(_df))\n",
    "    df = df.append(_df, sort=False)\n",
    "    \n",
    "df.to_csv(\"data/191203_combined_data.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Go through and look at signal peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for checking if SP exists\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import pairwise2\n",
    "\n",
    "all_sps = list(set(df['SP expected DNA'].values))\n",
    "\n",
    "def check_sp(row, verbose=True, return_sp=False):\n",
    "    ''' Given row in data, see if there's any SP sequence\n",
    "    '''\n",
    "    seq = row['Full Sequencing']\n",
    "\n",
    "    sp_exists = False\n",
    "    # SP alignment against expected\n",
    "    best_sp = row['SP expected DNA']\n",
    "    best_score = pairwise2.align.localxs(best_sp, seq, -1, -1, one_alignment_only=True)[0][2] \n",
    "    best_score = best_score / len(best_sp)\n",
    "\n",
    "    # Print cases that do not match spreadsheet sent\n",
    "    if best_score > 0.9:\n",
    "        sp_exists = True\n",
    "        if row['Correct Construct'] != True:\n",
    "            aln = pairwise2.align.localxs(best_sp, seq, -1, -1, one_alignment_only=True)\n",
    "            if verbose:\n",
    "                print('SP found when not found previously')\n",
    "                print(format_alignment(*aln[0]))\n",
    "\n",
    "    # Look for non-expected SPs (uses all_sps) above\n",
    "    if sp_exists == False:\n",
    "        # Iterate through other sps\n",
    "        for sp in all_sps:\n",
    "            aln = pairwise2.align.localxs(sp, seq, -1, -1, one_alignment_only=True)\n",
    "            score = aln[0][2] \n",
    "            norm_score = score / len(sp)\n",
    "            if norm_score > 0.9:\n",
    "                # If SP is found -- true\n",
    "                sp_exists = True\n",
    "                best_sp = sp\n",
    "                # Print mistake\n",
    "                if verbose:\n",
    "                    print('other sp found')\n",
    "                    print(format_alignment(*aln[0]))\n",
    "                    print('---\\n')\n",
    "                break\n",
    "        \n",
    "    if return_sp==False:            \n",
    "        return sp_exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SP found when not found previously\n",
      "  1 ATGGGATTCAGACTCAAAGCTCTCCTTGTCGGCTGTTTAATTTTTCTGGCGGTATCTAGCGCTATAGCCGGGGCT\n",
      "    ||||||||||.|||||||||||||||||||||||||||||||||||||||||.|||||.||||||||||||||||\n",
      "357 ATGGGATTCACACTCAAAGCTCTCCTTGTCGGCTGTTTAATTTTTCTGGCGGGATCTANCGCTATAGCCGGGGCT\n",
      "  Score=72\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTGAGTTTTAGCTCTTTACTAGCGGCAGCATCGCTAGCCGTGGTTAATGCGGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||\n",
      "355 ATGGTGAGTTTTAGCTCTTTACTAGCGGCAGCATCGCTAGCTGTGGTTAATGCGGGGGCT\n",
      "  Score=59\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "    ||||||||||||||||||.||.||||||||||||||||||||.||||||..||||||||||||||||||\n",
      "361 ATGAAAAAACGGTTGCACTTANGACTATTGCTATCACTCATCTCCTTCCNNGCCGGATTTGCGGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "    |||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "357 ATGAAGAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAATTCCAAGATCTCACACTTGTTCTGAGTTTGTCGACAGCTTTGGCTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||.||||||||||||||||||\n",
      "357 ATGAAATTCCAAGATCTCACACTTGTTCTGAGTTTGTCCACAGCTTTGGCTGGGGCT\n",
      "  Score=56\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGCCAGCTGCTCGTAGGTTGGCACTGTTTGCTGCAGTTGCTCTCGCTGCCGTGGGTTTGTCCCCTGCAGCGCTTGCAGGGGCT\n",
      "    ||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGCCAGCCGCTCGTAGGTTGGCACTGTTTGCTGCAGTTGCTCTCGCTGCCGTGGGTTTGTCCCCTGCAGCGCTTGCAGGGGCT\n",
      "  Score=83\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTCACTATGAAACTGCGGCTTATCGCGCTGGCTGTATGTCTCTGTACCTTTATCAACGCATCCTTCGCAGGGGCT\n",
      "    ||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGGTCACTATGGAACTGCGGCTTATCGCGCTGGCTGTATGTCTCTGTACCTTTATCAACGCATCCTTCGCAGGGGCT\n",
      "  Score=77\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGGATTCAGACTCAAAGCTCTCCTTGTCGGCTGTTTAATTTTTCTGGCGGTATCTAGCGCTATAGCCGGGGCT\n",
      "    ||||||||||.|||||||||||||||||||||||||||||||||||||||||.|||||.||||||||||||||||\n",
      "357 ATGGGATTCACACTCAAAGCTCTCCTTGTCGGCTGTTTAATTTTTCTGGCGGGATCTANCGCTATAGCCGGGGCT\n",
      "  Score=72\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTGAGTTTTAGCTCTTTACTAGCGGCAGCATCGCTAGCCGTGGTTAATGCGGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||\n",
      "355 ATGGTGAGTTTTAGCTCTTTACTAGCGGCAGCATCGCTAGCTGTGGTTAATGCGGGGGCT\n",
      "  Score=59\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "    ||||||||||||||||||.||.||||||||||||||||||||.||||||..||||||||||||||||||\n",
      "361 ATGAAAAAACGGTTGCACTTANGACTATTGCTATCACTCATCTCCTTCCNNGCCGGATTTGCGGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "    |||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "357 ATGAAGAAACGGTTGCACATAGGACTATTGCTATCACTCATCGCCTTCCAGGCCGGATTTGCGGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAATTCCAAGATCTCACACTTGTTCTGAGTTTGTCGACAGCTTTGGCTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||.||||||||||||||||||\n",
      "357 ATGAAATTCCAAGATCTCACACTTGTTCTGAGTTTGTCCACAGCTTTGGCTGGGGCT\n",
      "  Score=56\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGCCAGCTGCTCGTAGGTTGGCACTGTTTGCTGCAGTTGCTCTCGCTGCCGTGGGTTTGTCCCCTGCAGCGCTTGCAGGGGCT\n",
      "    ||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGCCAGCCGCTCGTAGGTTGGCACTGTTTGCTGCAGTTGCTCTCGCTGCCGTGGGTTTGTCCCCTGCAGCGCTTGCAGGGGCT\n",
      "  Score=83\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTCACTATGAAACTGCGGCTTATCGCGCTGGCTGTATGTCTCTGTACCTTTATCAACGCATCCTTCGCAGGGGCT\n",
      "    ||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGGTCACTATGGAACTGCGGCTTATCGCGCTGGCTGTATGTCTCTGTACCTTTATCAACGCATCCTTCGCAGGGGCT\n",
      "  Score=77\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAATTTTTTAATCCGTTCAAGGTAATTGCGCTTGCGTGCATTAGTGGAGCGCTAGCGACGGCCCAAGCTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||\n",
      "355 ATGAAATTTTTTAATCCGTTCAAGGTAATTGCGCTTGCGTGCATTAGTGGAGCGCTAGCGCCGGCCCAAGCTGGGGCT\n",
      "  Score=77\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGATGGCCGCAGTGGTGCGTGCCGTCGCTGCAACCTTAATTTTGATACTCTGTGGCGCCGAACTGGCAGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "357 ATGATGGCCGCAGTGGTGCGTGCCGTCGCTGCAACCTTAATTTTGATACTCTGTGGCGCCGAACTGGCAGGGGCT\n",
      "  Score=75\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGTTGAAGCGCTTCGTCAAGCTGGCAGTCATAGCATTAGCGTTCGCGTACGTGTCGGCAGGGGCT\n",
      "    ||||||||.||||||||||||||||||||||||||||||.||||||||||||||||||||||||||\n",
      "353 ATGTTGAAACGCTTCGTCAAGCTGGCAGTCATAGCATTANCGTTCGCGTACGTGTCGGCAGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAAACTGGGTTCATTGGCAAGACCCTCGCCCTTGTCATAGCGGCTGGTATGGCGGGGACAGCGGCGTTCGCGGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||\n",
      "352 ATGAAAAAAACTGGGTTCATTGGCAAGACCCTCGCCCTTGTCATAGCGGCTGGCATGGCGGGGACAGCGGCGTTCGCGGGGGCT\n",
      "  Score=83\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTCAAAATAGTGTTCGCGGTGGCCGCTATAGCTCCAGTCCTACACAGTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "351 ATGAAGCTCAAAATAGTGTTCGCGGTGGCCGCTATAGCTCCAGTCCTACACAGTGGGGCT\n",
      "  Score=60\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTGTCATTCAGTTCCCTGAATGCCTTATTTTTAGCCACAGTCCTGGCAGGGGCT\n",
      "    |||||||||||||||||||||||||||.|||||||||||||||||||||||||||||\n",
      "357 ATGGTGTCATTCAGTTCCCTGAATGCCCTATTTTTAGCCACAGTCCTGGCAGGGGCT\n",
      "  Score=56\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTATCGCAATCCCTCACTTATTTGGCCGTATTGGGGCTTGCGGCTGGTGCTAATGCGGGGGCT\n",
      "    |||||.|||.|||||||||||.||||||||||||.|||||||||||||||||||||||||||.||||||\n",
      "356 ATGAACCTAACGCAATCCCTCCCTTATTTGGCCGCATTGGGGCTTGCGGCTGGTGCTAATGCCGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAACATATCGATCTTCGTGGGTAAGCTGGCGCTAGCAGCGCTGGGATCAGCGTTGGTCGCAGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||\n",
      "359 ATGAACATATCGATCTTCGTGGGTAAGCTGGCGCTAGCAGCGCTGGGATCAGCGTCGGTCGCAGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGTTTAAACTCAAGGACATTCTAATAGGCCTTACAGGAATCTTACTATCGAGTCTTTTTGCGGGGGCT\n",
      "    ||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "357 ATGTTTAAACCCAAGGACATTCTAATAGGCCTTACAGGAATCTTACTATCGAGTCTTTTTGCGGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGTGTCATTCAGTTCCCTGAATGCCTTATTTTTAGCCACAGTCCTGGCAGGGGCT\n",
      "    |||||||||||||||||||||||||||.|||||||||||||||||||||||||||||\n",
      "357 ATGGTGTCATTCAGTTCCCTGAATGCCCTATTTTTAGCCACAGTCCTGGCAGGGGCT\n",
      "  Score=56\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTATCGCAATCCCTCACTTATTTGGCCGTATTGGGGCTTGCGGCTGGTGCTAATGCGGGGGCT\n",
      "    |||||.|||.|||||||||||.||||||||||||.|||||||||||||||||||||||||||.||||||\n",
      "356 ATGAACCTAACGCAATCCCTCCCTTATTTGGCCGCATTGGGGCTTGCGGCTGGTGCTAATGCCGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAACATATCGATCTTCGTGGGTAAGCTGGCGCTAGCAGCGCTGGGATCAGCGTTGGTCGCAGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||\n",
      "359 ATGAACATATCGATCTTCGTGGGTAAGCTGGCGCTAGCAGCGCTGGGATCAGCGTCGGTCGCAGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGTTTAAACTCAAGGACATTCTAATAGGCCTTACAGGAATCTTACTATCGAGTCTTTTTGCGGGGGCT\n",
      "    ||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "357 ATGTTTAAACCCAAGGACATTCTAATAGGCCTTACAGGAATCTTACTATCGAGTCTTTTTGCGGGGGCT\n",
      "  Score=68\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAATTTGCTCTTCTCGACGCTTACTGCTGTATTGATTACATCCGTCAGTTTTGCGGGGGCT\n",
      "    |||||||| |||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "360 ATGAAAAA-TTGCTCTTCTCGACGCTTACTGCTGTATTGATTACATCCGTCAGTTTTGCGGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGGTGTTTACGTTAGCATTCGCAATTATATGTCAATTGTTTGCCTCTGCAGGGGCT\n",
      "    ||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||\n",
      "355 ATGAAGGTGTTTACGTTAGCATTCGCAATTGTATGTCAATTGTTTGCCTCTGCAGGGGCT\n",
      "  Score=59\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGCTTTCAATAGATACGTCCTCGACGAGACGCGTCGTCCCGAATACTGCCCTTTTCCCGAATACTCACCGTCGTGACTTTGCTACCGCGGGTCAATTGTTGGCAATGGCCTCAGCAGTACTAACGGGGGCGCCCGCGCACGCGGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||||||||\n",
      "356 ATGCTTTCAATAGATACGTCCTCGACGAGACGCGTCGTCCCGAATACTGCCCTTTTCCCGAATACTCACCGTCGTGACTTTGCTACCGCGGGTCAATTGTTGGCAATGGCCTCAGCAGTACTAACGAGGGCGCCCGCGCACGCGGGGGCT\n",
      "  Score=149\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAGAAATAGATGGTGGATTATCTTGTTACTGTTTCTCGTTTTCCTTCCTAAGACGAGCTTTGCGCATGCTGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGAAAAGAAATAGATGGTGGATTATCTTGTTACTGTTTCTCGTTTTCCTTCCTAAGACGAGCTTTGCGCATGCTGGGGCT\n",
      "  Score=81\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAGATCAAAAAAACTTTGGATCAGCTTGTTATTCGCGCTCACACTGATTTTTACGATGGCTTTTTCCAATATGTCTGTTCAGGCTGGGGCT\n",
      "    |||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGAGATCAAAAAAACTTTGGATCATCTTGTTATTCGCGCTCACACTGATTTTTACGATGGCTTTTTCCAATATGTCTGTTCAGGCTGGGGCT\n",
      "  Score=92\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAGGTTCTGATGGCCTTTATCATTTGCTTGTCTCTTATTCTCTCAGTGTTAGCGGCACCTCCGAGCGGCGCTAAGGCTGGGGCT\n",
      "    |||||||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "355 ATGAAAAAGGTTCTGGTGGCCTTTATCATTTGCTTGTCTCTTATTCTCTCAGTGTTAGCGGCACCTCCGAGCGGCGCTAAGGCTGGGGCT\n",
      "  Score=89\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAGAAATAGATGGTGGATTATCTTGTTACTGTTTCTCGTTTTCCTTCCTAAGACGAGCTTTGCGCATGCTGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGAAAAGAAATAGATGGTGGATTATCTTGTTACTGTTTCTCGTTTTCCTTCCTAAGACGAGCTTTGCGCATGCTGGGGCT\n",
      "  Score=81\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAGATCAAAAAAACTTTGGATCAGCTTGTTATTCGCGCTCACACTGATTTTTACGATGGCTTTTTCCAATATGTCTGTTCAGGCTGGGGCT\n",
      "    |||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGAGATCAAAAAAACTTTGGATCATCTTGTTATTCGCGCTCACACTGATTTTTACGATGGCTTTTTCCAATATGTCTGTTCAGGCTGGGGCT\n",
      "  Score=92\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAAAAGGTTCTGATGGCCTTTATCATTTGCTTGTCTCTTATTCTCTCAGTGTTAGCGGCACCTCCGAGCGGCGCTAAGGCTGGGGCT\n",
      "    |||||||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "355 ATGAAAAAGGTTCTGGTGGCCTTTATCATTTGCTTGTCTCTTATTCTCTCAGTGTTAGCGGCACCTCCGAGCGGCGCTAAGGCTGGGGCT\n",
      "  Score=89\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGCGGGGCTA\n",
      "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|..||.|\n",
      "356 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGGGCTGCCA\n",
      "  Score=63\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAACTGGTTAGTTATAATGTTCGCTATTGGACTGGCGTGTAGTACTTCTTTGGCCTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||||||||||| ||||||||||| ||||||||||\n",
      "355 ATGAACTGGTTAGTTATAATGTTCGCTATTGGACTGGCGTG-AGTACTTCTTT-GCCTGGGGCT\n",
      "  Score=60\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAACTGGTTAGTTATAATGTTCGCTATTGGACTGGCGTGTAGTACTTCTTTGGCCTGGGGCT\n",
      "    ||||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||||\n",
      "357 ATGAACTGGTTAGTTATAATGTTCGCTATTGGCCTGGCGTGTAGTACTTCTTTGGCCTGGGGCT\n",
      "  Score=63\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTCCTTGTGCTCTTAATATTGATACTGATACCTGTCGTAGCCTGGGCCGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||||||||| |||||||||||||||||\n",
      "356 ATGAAGCTCCTTGTGCTCTTAATATTGATACTGATACCTGTC-TAGCCTGGGCCGGGGCT\n",
      "  Score=58\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAACTGGTTAGTTATAATGTTCGCTATTGGACTGGCGTGTAGTACTTCTTTGGCCTGGGGCT\n",
      "    |||||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||\n",
      "359 ATGAACTGGTTAGTTATAATGTTCGCTATT-GACTGGCGTGTAGTACTTCTTTGGCCTGGGGCT\n",
      "  Score=62\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "    |||||||||||||||||||||||||||||||||||| |||||||||||||||||||||||||||||\n",
      "356 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTT-TGTTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "  Score=64\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "    |||||||||||||||..||||||||||||||||||||||.||||||||||||||||||||||||||\n",
      "355 ATGGAACATACACTCCTAATCCTTGTTTGCTCGCTTCTGGTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "  Score=63\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGGAACATACACTCAGAATCCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "    ||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||\n",
      "355 ATGGAACATACACTCAGAATTCTTGTTTGCTCGCTTCTGTTTACAGTTGCCATTTGGGGCGGGGCT\n",
      "  Score=65\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTCCTTGTGCTCTTAATATTGATACTGATACCTGTCGTAGCCTGGGCCGGGGCT\n",
      "    |||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||\n",
      "355 ATGAAGCTCCTTGTGCTCTTAA-ATTGATACTGATACCTGTCGTAGCCTGGGCCGGGGCT\n",
      "  Score=58\n",
      "\n",
      "SP found when not found previously\n",
      "  1 ATGAAGCTCCTTGTGCTCTTAATATTGATACTGATACCTGTCGTAGCCTGGGCCGGGGCT\n",
      "    ||||||||||||||| ||||||||||||||||||||||||||||||||||||||||||||\n",
      "356 ATGAAGCTCCTTGTG-TCTTAATATTGATACTGATACCTGTCGTAGCCTGGGCCGGGGCT\n",
      "  Score=58\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(df)):\n",
    "    row = df.iloc[i]\n",
    "    check_sp(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Glyc stock plate</th>\n",
       "      <th>Well location</th>\n",
       "      <th>Planned Protein</th>\n",
       "      <th>Planned SP</th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>Correct Construct</th>\n",
       "      <th>Full Sequencing</th>\n",
       "      <th>SP start</th>\n",
       "      <th>SP end</th>\n",
       "      <th>SP coverage</th>\n",
       "      <th>...</th>\n",
       "      <th>Average</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "      <th>Assay5</th>\n",
       "      <th>assay_5_value</th>\n",
       "      <th>Unnamed: 35</th>\n",
       "      <th>Unnamed: 39</th>\n",
       "      <th>Unnamed: 40</th>\n",
       "      <th>Unnamed: 34</th>\n",
       "      <th>Average CV</th>\n",
       "      <th>sp_correct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A10</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A10-fwd_A10.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A11</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A11-fwd_A11.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>440.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A12</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A12-fwd_A12.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGGNNGTATTACATATGTAAG...</td>\n",
       "      <td>358.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A1</td>\n",
       "      <td>A001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A1-fwd_A01.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACANTGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>425.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A2</td>\n",
       "      <td>A006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A2-fwd_A02.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNNNNNNNGNANNNCNNNNNNNANNATTACATATGTAA...</td>\n",
       "      <td>359.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Glyc stock plate Well location Planned Protein  Planned SP  \\\n",
       "0   Alina_Amylase1           A10            A003         NaN   \n",
       "1   Alina_Amylase1           A11            A003         NaN   \n",
       "2   Alina_Amylase1           A12            A003         NaN   \n",
       "3   Alina_Amylase1            A1            A001         NaN   \n",
       "4   Alina_Amylase1            A2            A006         NaN   \n",
       "\n",
       "               Sequence Name Correct Construct  \\\n",
       "0  >amylase1-A10-fwd_A10.ab1              True   \n",
       "1  >amylase1-A11-fwd_A11.ab1              True   \n",
       "2  >amylase1-A12-fwd_A12.ab1              True   \n",
       "3   >amylase1-A1-fwd_A01.ab1              True   \n",
       "4   >amylase1-A2-fwd_A02.ab1              True   \n",
       "\n",
       "                                     Full Sequencing  SP start  SP end  \\\n",
       "0  NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   434.0   \n",
       "1  NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   440.0   \n",
       "2  NNNNNNNNNNTTACAATGTAGTCTTTGGNNGTATTACATATGTAAG...     358.0   417.0   \n",
       "3  NNNNNNNNNNTTACANTGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   425.0   \n",
       "4  NNNNNNNNNNNNNNNNGNANNNCNNNNNNNANNATTACATATGTAA...     359.0   427.0   \n",
       "\n",
       "   SP coverage  ... Average Unnamed: 32 Assay5  assay_5_value  Unnamed: 35  \\\n",
       "0         99.0  ...     NaN         NaN    NaT            NaN          NaN   \n",
       "1         99.0  ...     NaN         NaN    NaT            NaN          NaN   \n",
       "2         98.0  ...     NaN         NaN    NaT            NaN          NaN   \n",
       "3         99.0  ...     NaN         NaN    NaT            NaN          NaN   \n",
       "4         99.0  ...     NaN         NaN    NaT            NaN          NaN   \n",
       "\n",
       "   Unnamed: 39 Unnamed: 40 Unnamed: 34 Average CV  sp_correct  \n",
       "0          NaN         NaN         NaN        NaN        True  \n",
       "1          NaN         NaN         NaN        NaN        True  \n",
       "2          NaN         NaN         NaN        NaN        True  \n",
       "3          NaN         NaN         NaN        NaN        True  \n",
       "4          NaN         NaN         NaN        NaN        True  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply to df\n",
    "df['sp_correct'] = df.apply(lambda row: check_sp(row, verbose=False), axis = 1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for proteins: ### Go through and look at signal peptides"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# 92, 279, 424 need some further analysis\n",
    "\n",
    "# Conclusions\n",
    "92 and 279 belong to Arnold_001 construct 39_39-4\n",
    "Both are nonfunctional, but there is another functional construct.\n",
    "They were previously at norm_score thresholds 0.74, the new threshold cutoff of 0.75 removes these nonfunctional sequences.\n",
    "\n",
    "424 also does not have a functional SP sequence, and would already be labeled as a nonfunctional sequence by definition. No further steps are taken to deal with it.\n",
    "\n",
    "These issues arose when previously using a norm_score cutoff of 0.7. However, as can be seen below, no issues arise when using a cutoff of 0.75 for protein sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Similar process for proteins\n",
    "\n",
    "all_prots = list(set(df['Protein expected DNA'].values))\n",
    "\n",
    "def check_prot(row, verbose=True):\n",
    "    \"\"\" Given a row in protein, returns True if prot is found, otherwise False\n",
    "    \"\"\"\n",
    "    seq = row['Full Sequencing']\n",
    "\n",
    "    prot_exists = False\n",
    "    # Prot alignment\n",
    "    best_prot = row['Protein expected DNA'][:300]     # only take first few since sequencing cuts off\n",
    "    best_score = pairwise2.align.localxs(best_prot, seq, -.5, -.5, one_alignment_only=True)[0][2]\n",
    "    best_score = best_score / len(best_prot)\n",
    "\n",
    "    if best_score > 0.8:\n",
    "        prot_exists = True\n",
    "\n",
    "    # Check other proteins if expected is not found\n",
    "    if prot_exists == False:\n",
    "        # Iterate through other prots\n",
    "        for prot in all_prots:\n",
    "            prot = prot[:300]\n",
    "            aln = pairwise2.align.localxs(prot, seq, -.5, -.5, one_alignment_only=True)\n",
    "            score = aln[0][2] \n",
    "            norm_score = score / len(prot)\n",
    "            if norm_score >= 0.75:\n",
    "                # If prot is found -- true\n",
    "                prot_exists = True\n",
    "                best_prot = prot\n",
    "                # Print mistake\n",
    "                if verbose:\n",
    "                    print('other prot found')\n",
    "                    print('original', row['Planned Protein'])\n",
    "                    print('new')\n",
    "                    print('best_prot')\n",
    "                    print(format_alignment(*aln[0]))\n",
    "                    print(norm_score)\n",
    "                    print(best_prot)\n",
    "                    print('---')\n",
    "                    return 'further analysis'\n",
    "                break\n",
    "                \n",
    "#     # If protein is not found, show alignment            \n",
    "#     if verbose:\n",
    "#         if prot_exists == False:\n",
    "#             aln = pairwise2.align.localxs(best_prot, seq, -.5, -.5, one_alignment_only=True)\n",
    "#             print('protein not found')\n",
    "#             print(format_alignment(*aln[0]))\n",
    "#             print(best_score)\n",
    "#             print('---')\n",
    "        \n",
    "    return prot_exists\n",
    "\n",
    "# check_prot(df.iloc[599]    # Test\n",
    "# s = 'TGTCGGCTACGGAGCGTACGACTTGTATGATCTTGGCGAATTTCATCAAAAAGGTACTGTACGTACAA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0  for further analysis\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for i in range(len(df)):\n",
    "    row = df.iloc[i]\n",
    "    a = check_prot(row)\n",
    "    if a == 'further analysis':\n",
    "        count += 1\n",
    "        print(a)\n",
    "print(str(count), ' for further analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Glyc stock plate</th>\n",
       "      <th>Well location</th>\n",
       "      <th>Planned Protein</th>\n",
       "      <th>Planned SP</th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>Correct Construct</th>\n",
       "      <th>Full Sequencing</th>\n",
       "      <th>SP start</th>\n",
       "      <th>SP end</th>\n",
       "      <th>SP coverage</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "      <th>Assay5</th>\n",
       "      <th>assay_5_value</th>\n",
       "      <th>Unnamed: 35</th>\n",
       "      <th>Unnamed: 39</th>\n",
       "      <th>Unnamed: 40</th>\n",
       "      <th>Unnamed: 34</th>\n",
       "      <th>Average CV</th>\n",
       "      <th>sp_correct</th>\n",
       "      <th>prot_correct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A10</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A10-fwd_A10.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A11</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A11-fwd_A11.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>440.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A12</td>\n",
       "      <td>A003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A12-fwd_A12.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACAATGTAGTCTTTGGNNGTATTACATATGTAAG...</td>\n",
       "      <td>358.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A1</td>\n",
       "      <td>A001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A1-fwd_A01.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNTTACANTGTAGTCTTTGAAGTATTACATATGTAAGA...</td>\n",
       "      <td>357.0</td>\n",
       "      <td>425.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Alina_Amylase1</td>\n",
       "      <td>A2</td>\n",
       "      <td>A006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;amylase1-A2-fwd_A02.ab1</td>\n",
       "      <td>True</td>\n",
       "      <td>NNNNNNNNNNNNNNNNGNANNNCNNNNNNNANNATTACATATGTAA...</td>\n",
       "      <td>359.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Glyc stock plate Well location Planned Protein  Planned SP  \\\n",
       "0   Alina_Amylase1           A10            A003         NaN   \n",
       "1   Alina_Amylase1           A11            A003         NaN   \n",
       "2   Alina_Amylase1           A12            A003         NaN   \n",
       "3   Alina_Amylase1            A1            A001         NaN   \n",
       "4   Alina_Amylase1            A2            A006         NaN   \n",
       "\n",
       "               Sequence Name Correct Construct  \\\n",
       "0  >amylase1-A10-fwd_A10.ab1              True   \n",
       "1  >amylase1-A11-fwd_A11.ab1              True   \n",
       "2  >amylase1-A12-fwd_A12.ab1              True   \n",
       "3   >amylase1-A1-fwd_A01.ab1              True   \n",
       "4   >amylase1-A2-fwd_A02.ab1              True   \n",
       "\n",
       "                                     Full Sequencing  SP start  SP end  \\\n",
       "0  NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   434.0   \n",
       "1  NNNNNNNNNNTTACAATGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   440.0   \n",
       "2  NNNNNNNNNNTTACAATGTAGTCTTTGGNNGTATTACATATGTAAG...     358.0   417.0   \n",
       "3  NNNNNNNNNNTTACANTGTAGTCTTTGAAGTATTACATATGTAAGA...     357.0   425.0   \n",
       "4  NNNNNNNNNNNNNNNNGNANNNCNNNNNNNANNATTACATATGTAA...     359.0   427.0   \n",
       "\n",
       "   SP coverage  ... Unnamed: 32 Assay5 assay_5_value  Unnamed: 35  \\\n",
       "0         99.0  ...         NaN    NaT           NaN          NaN   \n",
       "1         99.0  ...         NaN    NaT           NaN          NaN   \n",
       "2         98.0  ...         NaN    NaT           NaN          NaN   \n",
       "3         99.0  ...         NaN    NaT           NaN          NaN   \n",
       "4         99.0  ...         NaN    NaT           NaN          NaN   \n",
       "\n",
       "   Unnamed: 39  Unnamed: 40 Unnamed: 34 Average CV sp_correct  prot_correct  \n",
       "0          NaN          NaN         NaN        NaN       True          True  \n",
       "1          NaN          NaN         NaN        NaN       True          True  \n",
       "2          NaN          NaN         NaN        NaN       True          True  \n",
       "3          NaN          NaN         NaN        NaN       True          True  \n",
       "4          NaN          NaN         NaN        NaN       True          True  \n",
       "\n",
       "[5 rows x 57 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['prot_correct'] = df.apply(lambda row: check_prot(row, verbose=False), axis = 1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take out relevant columns\n",
    "df = df[['assay_1_value', 'assay_2_value', 'assay_3_value', 'assay_4_value', 'assay_5_value', 'Correct Construct', 'Protein expected name', 'SP expected name', 'run_label','sp_correct', 'prot_correct']]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "SP IDs as ordered from Twist match sp_prot_translations.xls.\n",
    "Reminder: sp_prot_translations is from Alycia's word file containing bs_5 outputs\n",
    "\n",
    "However,the enzymes are not correct. Need to take order from sp_df and match with enzyme_order_df\n",
    "\n",
    "ie: order on r1_191003 is label_ID:SP_ID by linking\n",
    "    1. sequence from sp_prot_translations : true_id\n",
    "    2. sequence from initial_enzymes_1: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pickle' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-f5d9e7503107>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"data/enzID_to_trueID.p\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"wb\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mpickle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menzID_to_trueID\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'pickle' is not defined"
     ]
    }
   ],
   "source": [
    "# Map true protein ids\n",
    "enzyme_order_df['prot_seq_trunc'] = enzyme_order_df['Protein-met-sigp'].str[:80]\n",
    "sp_df['prot_seq_trunc'] = sp_df['prot_seq'].str[1:81]\n",
    "\n",
    "merge_df = enzyme_order_df.merge(sp_df, how='inner', on='prot_seq_trunc')\n",
    "\n",
    "# The following dictionary maps the TWIST labeled enzyme ID to the corresponding SP ID\n",
    "enzID_to_trueID = pd.Series(merge_df.seqID.values,index=merge_df.enzyme_id).to_dict()\n",
    "assert len(list(enzID_to_trueID.keys())) == len(list(enzID_to_trueID.values())), \"dics unaligned, check sequence identities\"\n",
    "\n",
    "with open(\"data/enzID_to_trueID.p\",\"wb\") as f:\n",
    "    pickle.dump(enzID_to_trueID, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get nice column names\n",
    "col_name_switch = dict()\n",
    "for i in range(5):\n",
    "    col_name_switch.update({'assay_value_' + str(i) : 'Assay' + str(i)})\n",
    "df.rename(columns=col_name_switch, inplace=True)\n",
    "df.rename(columns={'Protein expected name':'protein_id', 'SP expected name':'sp_id', 'Correct Construct':'correct'}, inplace=True)\n",
    "\n",
    "df['true_prot_id'] = df['protein_id'].map(enzID_to_trueID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update families -- but keep 10x/50x separate for now\n",
    "def get_familyid(row):\n",
    "    run_id = row['run_label']\n",
    "    \n",
    "    if run_id[-1] == 'x':\n",
    "        return run_id\n",
    "    else:\n",
    "        return run_id.split('_')[0]   # just take family name\n",
    "\n",
    "df['family'] = df.apply(lambda row: get_familyid(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add run_id, write to csv\n",
    "\n",
    "def get_rowid(row):\n",
    "    spid = row['sp_id']\n",
    "    \n",
    "    if spid[:3] == 'sps':\n",
    "        return row['true_prot_id'][3:] + '_' + row['sp_id'][3:]\n",
    "    elif spid[:3] == 'Sig':\n",
    "        return row['true_prot_id'][3:] + '_' + row['sp_id'][3:] + \"_pos\"\n",
    "    elif spid[:3] == 'Neg':\n",
    "        return row['true_prot_id'][3:] + '_' + row['sp_id'][3:] + \"_neg\"\n",
    "    else:\n",
    "        raise spid\n",
    "        \n",
    "df['run_id'] = df.apply(lambda row: get_rowid(row), axis=1)\n",
    "df.to_csv(\"data/preprocessed.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_1_value</th>\n",
       "      <th>assay_2_value</th>\n",
       "      <th>assay_3_value</th>\n",
       "      <th>assay_4_value</th>\n",
       "      <th>assay_5_value</th>\n",
       "      <th>correct</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>sp_id</th>\n",
       "      <th>run_label</th>\n",
       "      <th>sp_correct</th>\n",
       "      <th>prot_correct</th>\n",
       "      <th>true_prot_id</th>\n",
       "      <th>family</th>\n",
       "      <th>run_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>0.752</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.736</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Arnold_040</td>\n",
       "      <td>Neg1</td>\n",
       "      <td>negative</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>seq26</td>\n",
       "      <td>negative</td>\n",
       "      <td>26_1_neg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>0.860</td>\n",
       "      <td>0.892</td>\n",
       "      <td>0.672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Arnold_040</td>\n",
       "      <td>Neg1</td>\n",
       "      <td>negative</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>seq26</td>\n",
       "      <td>negative</td>\n",
       "      <td>26_1_neg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>1.452</td>\n",
       "      <td>1.782</td>\n",
       "      <td>1.640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Arnold_040</td>\n",
       "      <td>Neg3</td>\n",
       "      <td>negative</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>seq26</td>\n",
       "      <td>negative</td>\n",
       "      <td>26_3_neg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>0.574</td>\n",
       "      <td>0.644</td>\n",
       "      <td>0.622</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Arnold_040</td>\n",
       "      <td>Neg1</td>\n",
       "      <td>negative</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>seq26</td>\n",
       "      <td>negative</td>\n",
       "      <td>26_1_neg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>0.432</td>\n",
       "      <td>0.490</td>\n",
       "      <td>1.290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Arnold_040</td>\n",
       "      <td>Neg2</td>\n",
       "      <td>negative</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>seq26</td>\n",
       "      <td>negative</td>\n",
       "      <td>26_2_neg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    assay_1_value  assay_2_value  assay_3_value  assay_4_value  assay_5_value  \\\n",
       "62          0.752          0.524          0.736            NaN            NaN   \n",
       "63          0.860          0.892          0.672            NaN            NaN   \n",
       "64          1.452          1.782          1.640            NaN            NaN   \n",
       "65          0.574          0.644          0.622            NaN            NaN   \n",
       "66          0.432          0.490          1.290            NaN            NaN   \n",
       "\n",
       "   correct  protein_id sp_id run_label  sp_correct  prot_correct true_prot_id  \\\n",
       "62    True  Arnold_040  Neg1  negative        True          True        seq26   \n",
       "63    True  Arnold_040  Neg1  negative        True          True        seq26   \n",
       "64    True  Arnold_040  Neg3  negative        True          True        seq26   \n",
       "65    True  Arnold_040  Neg1  negative        True          True        seq26   \n",
       "66    True  Arnold_040  Neg2  negative        True          True        seq26   \n",
       "\n",
       "      family    run_id  \n",
       "62  negative  26_1_neg  \n",
       "63  negative  26_1_neg  \n",
       "64  negative  26_3_neg  \n",
       "65  negative  26_1_neg  \n",
       "66  negative  26_2_neg  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
