{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Healthcare Hack Nights: Part I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## RNA-Seq\n",
    "Questions that can be answered by RNA-seq:\n",
    "    - What genes are differentially expressed between group samples?\n",
    "    - How does gene expression change across time or conditions? (eg, in benign vs malignant tumors)\n",
    "    - What pathways or processes are enriched under a condition?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(423, 20531)"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc = pd.read_csv('../lihc_rnaseq.csv')\n",
    "lihc.set_index('bcr_patient_barcode', inplace=True)\n",
    "lihc.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this count matrix, each column represents an Ensembl gene transcript, each row a patient sequenced RNA library, and the values give the raw numbers of fragments that were uniquely assigned to the respective gene in each library. We also have additional information on each of the patients samples (the rows of the count matrix) and on each of the genes (the columns of the matrix).\n",
    "                    \n",
    "We now have all the ingredients to prepare our data object in a form that is suitable for analysis, namely:\n",
    "\n",
    " - countdata: a table with the fragment counts\n",
    "\n",
    " - rowdata: a table with information about the patient samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>?|100130426</th>\n",
       "      <th>?|100133144</th>\n",
       "      <th>?|100134869</th>\n",
       "      <th>?|10357</th>\n",
       "      <th>?|10431</th>\n",
       "      <th>?|136542</th>\n",
       "      <th>?|155060</th>\n",
       "      <th>?|26823</th>\n",
       "      <th>?|280660</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA|7789</th>\n",
       "      <th>ZXDB|158586</th>\n",
       "      <th>ZXDC|79364</th>\n",
       "      <th>ZYG11A|440590</th>\n",
       "      <th>ZYG11B|79699</th>\n",
       "      <th>ZYX|7791</th>\n",
       "      <th>ZZEF1|23140</th>\n",
       "      <th>ZZZ3|26009</th>\n",
       "      <th>psiTPTE22|387590</th>\n",
       "      <th>tAKR|389932</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5051</td>\n",
       "      <td>3.7074</td>\n",
       "      <td>90.1124</td>\n",
       "      <td>1017.1038</td>\n",
       "      <td>0.0</td>\n",
       "      <td>141.3911</td>\n",
       "      <td>0.6516</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>24.7597</td>\n",
       "      <td>273.6602</td>\n",
       "      <td>794.2662</td>\n",
       "      <td>18.2440</td>\n",
       "      <td>499.1041</td>\n",
       "      <td>3172.5037</td>\n",
       "      <td>890.0472</td>\n",
       "      <td>510.1808</td>\n",
       "      <td>3.9094</td>\n",
       "      <td>6.5157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.4120</td>\n",
       "      <td>2.6663</td>\n",
       "      <td>71.0054</td>\n",
       "      <td>639.2311</td>\n",
       "      <td>0.0</td>\n",
       "      <td>122.7206</td>\n",
       "      <td>1.4786</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>68.5067</td>\n",
       "      <td>632.8241</td>\n",
       "      <td>1153.7703</td>\n",
       "      <td>71.4638</td>\n",
       "      <td>1000.4929</td>\n",
       "      <td>5301.1336</td>\n",
       "      <td>755.5446</td>\n",
       "      <td>860.5224</td>\n",
       "      <td>6.4071</td>\n",
       "      <td>482.9966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-2Y-A9GT-01A-11R-A38B-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.4833</td>\n",
       "      <td>95.5122</td>\n",
       "      <td>742.4344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.0460</td>\n",
       "      <td>1.7933</td>\n",
       "      <td>0.8967</td>\n",
       "      <td>...</td>\n",
       "      <td>46.6263</td>\n",
       "      <td>1219.4575</td>\n",
       "      <td>1133.3782</td>\n",
       "      <td>12.5532</td>\n",
       "      <td>1289.3970</td>\n",
       "      <td>3219.0092</td>\n",
       "      <td>860.7935</td>\n",
       "      <td>523.6494</td>\n",
       "      <td>14.3466</td>\n",
       "      <td>83.3894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-2Y-A9GU-01A-11R-A38B-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.7222</td>\n",
       "      <td>5.1216</td>\n",
       "      <td>61.6679</td>\n",
       "      <td>1186.9807</td>\n",
       "      <td>0.0</td>\n",
       "      <td>280.2709</td>\n",
       "      <td>0.8341</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.3511</td>\n",
       "      <td>285.2758</td>\n",
       "      <td>1150.2786</td>\n",
       "      <td>9.1755</td>\n",
       "      <td>941.7437</td>\n",
       "      <td>3092.9899</td>\n",
       "      <td>1339.6283</td>\n",
       "      <td>343.6655</td>\n",
       "      <td>2.5024</td>\n",
       "      <td>2.5024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-2Y-A9GV-01A-11R-A38B-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.4975</td>\n",
       "      <td>5.4230</td>\n",
       "      <td>104.4670</td>\n",
       "      <td>878.1726</td>\n",
       "      <td>0.0</td>\n",
       "      <td>282.5719</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>41.4552</td>\n",
       "      <td>999.1540</td>\n",
       "      <td>1631.9797</td>\n",
       "      <td>4.2301</td>\n",
       "      <td>1380.7107</td>\n",
       "      <td>2902.7073</td>\n",
       "      <td>575.2961</td>\n",
       "      <td>665.8206</td>\n",
       "      <td>2.5381</td>\n",
       "      <td>119.2893</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20532 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            bcr_patient_barcode  ?|100130426  ?|100133144  ?|100134869  \\\n",
       "0  TCGA-2V-A95S-01A-11R-A37K-07          0.0       1.5051       3.7074   \n",
       "1  TCGA-2Y-A9GS-01A-12R-A38B-07          0.0      26.4120       2.6663   \n",
       "2  TCGA-2Y-A9GT-01A-11R-A38B-07          0.0       0.0000       4.4833   \n",
       "3  TCGA-2Y-A9GU-01A-11R-A38B-07          0.0       5.7222       5.1216   \n",
       "4  TCGA-2Y-A9GV-01A-11R-A38B-07          0.0      11.4975       5.4230   \n",
       "\n",
       "    ?|10357    ?|10431  ?|136542  ?|155060  ?|26823  ?|280660  ...  ZXDA|7789  \\\n",
       "0   90.1124  1017.1038       0.0  141.3911   0.6516    0.0000  ...    24.7597   \n",
       "1   71.0054   639.2311       0.0  122.7206   1.4786    0.0000  ...    68.5067   \n",
       "2   95.5122   742.4344       0.0   95.0460   1.7933    0.8967  ...    46.6263   \n",
       "3   61.6679  1186.9807       0.0  280.2709   0.8341    0.0000  ...    18.3511   \n",
       "4  104.4670   878.1726       0.0  282.5719   0.0000    0.0000  ...    41.4552   \n",
       "\n",
       "   ZXDB|158586  ZXDC|79364  ZYG11A|440590  ZYG11B|79699   ZYX|7791  \\\n",
       "0     273.6602    794.2662        18.2440      499.1041  3172.5037   \n",
       "1     632.8241   1153.7703        71.4638     1000.4929  5301.1336   \n",
       "2    1219.4575   1133.3782        12.5532     1289.3970  3219.0092   \n",
       "3     285.2758   1150.2786         9.1755      941.7437  3092.9899   \n",
       "4     999.1540   1631.9797         4.2301     1380.7107  2902.7073   \n",
       "\n",
       "   ZZEF1|23140  ZZZ3|26009  psiTPTE22|387590  tAKR|389932  \n",
       "0     890.0472    510.1808            3.9094       6.5157  \n",
       "1     755.5446    860.5224            6.4071     482.9966  \n",
       "2     860.7935    523.6494           14.3466      83.3894  \n",
       "3    1339.6283    343.6655            2.5024       2.5024  \n",
       "4     575.2961    665.8206            2.5381     119.2893  \n",
       "\n",
       "[5 rows x 20532 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataset contains the number of counts for ~20k genes defined by their Entrez transcript ID x 423 deidentified patients. \n",
    "\n",
    "Entrez (https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html) is a data retrieval system that provides users access to NCBIâ€™s databases such as PubMed, GenBank, GEO, and many others. You can access Entrez from a web browser to manually enter queries, or you can use Biopythonâ€™s Bio.Entrez module for programmatic access to Entrez. \n",
    "\n",
    "Entrez gene IDs are unique gene identifiers that can be used to trace a particular gene or transcript to the genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['100130426', '100133144', '100134869', '10357', '10431'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Entrez transcript IDs\n",
    "\n",
    "ids = pd.Series(lihc.columns.values[1:]).apply(lambda x: x.split('|')[1]).values\n",
    "ids[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioMart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pybiomart import Server, Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000200036</td>\n",
       "      <td>ENSG00000200036.1</td>\n",
       "      <td>ENST00000363166</td>\n",
       "      <td>ENST00000363166.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000252396</td>\n",
       "      <td>ENSG00000252396.1</td>\n",
       "      <td>ENST00000516587</td>\n",
       "      <td>ENST00000516587.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000252429</td>\n",
       "      <td>ENSG00000252429.2</td>\n",
       "      <td>ENST00000516620</td>\n",
       "      <td>ENST00000516620.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000221643</td>\n",
       "      <td>ENSG00000221643.1</td>\n",
       "      <td>ENST00000408716</td>\n",
       "      <td>ENST00000408716.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000264371</td>\n",
       "      <td>ENSG00000264371.1</td>\n",
       "      <td>ENST00000580572</td>\n",
       "      <td>ENST00000580572.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36496</td>\n",
       "      <td>ENSG00000196290</td>\n",
       "      <td>ENSG00000196290.15</td>\n",
       "      <td>ENST00000409357</td>\n",
       "      <td>ENST00000409357.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36497</td>\n",
       "      <td>ENSG00000196290</td>\n",
       "      <td>ENSG00000196290.15</td>\n",
       "      <td>ENST00000409129</td>\n",
       "      <td>ENST00000409129.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36498</td>\n",
       "      <td>ENSG00000196290</td>\n",
       "      <td>ENSG00000196290.15</td>\n",
       "      <td>ENST00000409588</td>\n",
       "      <td>ENST00000409588.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36499</td>\n",
       "      <td>ENSG00000196290</td>\n",
       "      <td>ENSG00000196290.15</td>\n",
       "      <td>ENST00000436412</td>\n",
       "      <td>ENST00000436412.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36500</td>\n",
       "      <td>ENSG00000196290</td>\n",
       "      <td>ENSG00000196290.15</td>\n",
       "      <td>ENST00000416651</td>\n",
       "      <td>ENST00000416651.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36501 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0      ENSG00000200036      ENSG00000200036.1      ENST00000363166   \n",
       "1      ENSG00000252396      ENSG00000252396.1      ENST00000516587   \n",
       "2      ENSG00000252429      ENSG00000252429.2      ENST00000516620   \n",
       "3      ENSG00000221643      ENSG00000221643.1      ENST00000408716   \n",
       "4      ENSG00000264371      ENSG00000264371.1      ENST00000580572   \n",
       "...                ...                    ...                  ...   \n",
       "36496  ENSG00000196290     ENSG00000196290.15      ENST00000409357   \n",
       "36497  ENSG00000196290     ENSG00000196290.15      ENST00000409129   \n",
       "36498  ENSG00000196290     ENSG00000196290.15      ENST00000409588   \n",
       "36499  ENSG00000196290     ENSG00000196290.15      ENST00000436412   \n",
       "36500  ENSG00000196290     ENSG00000196290.15      ENST00000416651   \n",
       "\n",
       "      Transcript stable ID version  \n",
       "0                ENST00000363166.1  \n",
       "1                ENST00000516587.1  \n",
       "2                ENST00000516620.2  \n",
       "3                ENST00000408716.1  \n",
       "4                ENST00000580572.1  \n",
       "...                            ...  \n",
       "36496            ENST00000409357.5  \n",
       "36497            ENST00000409129.2  \n",
       "36498            ENST00000409588.1  \n",
       "36499            ENST00000436412.1  \n",
       "36500            ENST00000416651.5  \n",
       "\n",
       "[36501 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retrieving a dataset directly with known dataset name\n",
    "\n",
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')\n",
    "\n",
    "dataset.query(\n",
    "              filters={'chromosome_name': ['1','2']})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `attributes` attribute can be used to pull up a list of additional fields available from the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ensembl_gene_id',\n",
       " 'ensembl_gene_id_version',\n",
       " 'ensembl_transcript_id',\n",
       " 'ensembl_transcript_id_version',\n",
       " 'ensembl_peptide_id',\n",
       " 'ensembl_peptide_id_version',\n",
       " 'ensembl_exon_id',\n",
       " 'description',\n",
       " 'chromosome_name',\n",
       " 'start_position',\n",
       " 'end_position',\n",
       " 'strand',\n",
       " 'band',\n",
       " 'transcript_start',\n",
       " 'transcript_end',\n",
       " 'transcription_start_site',\n",
       " 'transcript_length',\n",
       " 'transcript_tsl',\n",
       " 'transcript_gencode_basic',\n",
       " 'transcript_appris',\n",
       " 'transcript_mane_select',\n",
       " 'external_gene_name',\n",
       " 'external_gene_source',\n",
       " 'external_transcript_name',\n",
       " 'external_transcript_source_name',\n",
       " 'transcript_count',\n",
       " 'percentage_gene_gc_content',\n",
       " 'gene_biotype',\n",
       " 'transcript_biotype',\n",
       " 'source',\n",
       " 'transcript_source',\n",
       " 'version',\n",
       " 'transcript_version',\n",
       " 'peptide_version',\n",
       " 'phenotype_description',\n",
       " 'Source_name',\n",
       " 'study_external_id',\n",
       " 'strain_name',\n",
       " 'strain_gender',\n",
       " 'p_value',\n",
       " 'go_id',\n",
       " 'name_1006',\n",
       " 'definition_1006',\n",
       " 'go_linkage_type',\n",
       " 'namespace_1003',\n",
       " 'goslim_goa_accession',\n",
       " 'goslim_goa_description',\n",
       " 'ccds',\n",
       " 'chembl',\n",
       " 'clone_based_ensembl_gene',\n",
       " 'clone_based_ensembl_transcript',\n",
       " 'dbass3_name',\n",
       " 'dbass3_id',\n",
       " 'dbass5_name',\n",
       " 'dbass5_id',\n",
       " 'ens_hs_transcript',\n",
       " 'ens_hs_translation',\n",
       " 'entrezgene_trans_name',\n",
       " 'embl',\n",
       " 'arrayexpress',\n",
       " 'genedb',\n",
       " 'hgnc_id',\n",
       " 'hgnc_symbol',\n",
       " 'hgnc_trans_name',\n",
       " 'hpa_accession',\n",
       " 'hpa_id',\n",
       " 'protein_id',\n",
       " 'kegg_enzyme',\n",
       " 'ens_lrg_gene',\n",
       " 'ens_lrg_transcript',\n",
       " 'merops',\n",
       " 'mim_gene_description',\n",
       " 'mim_gene_accession',\n",
       " 'mim_morbid_description',\n",
       " 'mim_morbid_accession',\n",
       " 'mirbase_accession',\n",
       " 'mirbase_id',\n",
       " 'mirbase_trans_name',\n",
       " 'entrezgene_description',\n",
       " 'entrezgene_accession',\n",
       " 'entrezgene_id',\n",
       " 'pdb',\n",
       " 'reactome',\n",
       " 'reactome_gene',\n",
       " 'reactome_transcript',\n",
       " 'refseq_mrna',\n",
       " 'refseq_mrna_predicted',\n",
       " 'refseq_ncrna',\n",
       " 'refseq_ncrna_predicted',\n",
       " 'refseq_peptide',\n",
       " 'refseq_peptide_predicted',\n",
       " 'rfam',\n",
       " 'rfam_trans_name',\n",
       " 'rnacentral',\n",
       " 'ucsc',\n",
       " 'uniparc',\n",
       " 'uniprot_gn_symbol',\n",
       " 'uniprot_gn_id',\n",
       " 'uniprotswissprot',\n",
       " 'uniprotsptrembl',\n",
       " 'wikigene_description',\n",
       " 'wikigene_name',\n",
       " 'wikigene_id',\n",
       " 'affy_hc_g110',\n",
       " 'affy_hg_focus',\n",
       " 'affy_hg_u133a',\n",
       " 'affy_hg_u133a_2',\n",
       " 'affy_hg_u133b',\n",
       " 'affy_hg_u133_plus_2',\n",
       " 'affy_hg_u95a',\n",
       " 'affy_hg_u95av2',\n",
       " 'affy_hg_u95b',\n",
       " 'affy_hg_u95c',\n",
       " 'affy_hg_u95d',\n",
       " 'affy_hg_u95e',\n",
       " 'affy_hta_2_0',\n",
       " 'affy_huex_1_0_st_v2',\n",
       " 'affy_hugenefl',\n",
       " 'affy_hugene_1_0_st_v1',\n",
       " 'affy_hugene_2_0_st_v1',\n",
       " 'affy_primeview',\n",
       " 'affy_u133_x3p',\n",
       " 'agilent_cgh_44b',\n",
       " 'agilent_gpl6848',\n",
       " 'agilent_sureprint_g3_ge_8x60k',\n",
       " 'agilent_sureprint_g3_ge_8x60k_v2',\n",
       " 'agilent_wholegenome',\n",
       " 'agilent_wholegenome_4x44k_v1',\n",
       " 'agilent_wholegenome_4x44k_v2',\n",
       " 'codelink_codelink',\n",
       " 'illumina_humanht_12_v3',\n",
       " 'illumina_humanht_12_v4',\n",
       " 'illumina_humanref_8_v3',\n",
       " 'illumina_humanwg_6_v1',\n",
       " 'illumina_humanwg_6_v2',\n",
       " 'illumina_humanwg_6_v3',\n",
       " 'phalanx_onearray',\n",
       " 'family',\n",
       " 'family_description',\n",
       " 'cdd',\n",
       " 'cdd_start',\n",
       " 'cdd_end',\n",
       " 'gene3d',\n",
       " 'gene3d_start',\n",
       " 'gene3d_end',\n",
       " 'hamap',\n",
       " 'hamap_start',\n",
       " 'hamap_end',\n",
       " 'hmmpanther',\n",
       " 'hmmpanther_start',\n",
       " 'hmmpanther_end',\n",
       " 'pfam',\n",
       " 'pfam_start',\n",
       " 'pfam_end',\n",
       " 'pirsf',\n",
       " 'pirsf_start',\n",
       " 'pirsf_end',\n",
       " 'prints',\n",
       " 'prints_start',\n",
       " 'prints_end',\n",
       " 'blastprodom',\n",
       " 'blastprodom_start',\n",
       " 'blastprodom_end',\n",
       " 'scanprosite',\n",
       " 'scanprosite_start',\n",
       " 'scanprosite_end',\n",
       " 'pfscan',\n",
       " 'pfscan_start',\n",
       " 'pfscan_end',\n",
       " 'sfld',\n",
       " 'sfld_start',\n",
       " 'sfld_end',\n",
       " 'smart',\n",
       " 'smart_start',\n",
       " 'smart_end',\n",
       " 'superfamily',\n",
       " 'superfamily_start',\n",
       " 'superfamily_end',\n",
       " 'tigrfam',\n",
       " 'tigrfam_start',\n",
       " 'tigrfam_end',\n",
       " 'interpro',\n",
       " 'interpro_short_description',\n",
       " 'interpro_description',\n",
       " 'interpro_start',\n",
       " 'interpro_end',\n",
       " 'mobidblite',\n",
       " 'mobidblite_start',\n",
       " 'mobidblite_end',\n",
       " 'ncoils',\n",
       " 'ncoils_start',\n",
       " 'ncoils_end',\n",
       " 'seg',\n",
       " 'seg_start',\n",
       " 'seg_end',\n",
       " 'sifts_import',\n",
       " 'sifts_import_start',\n",
       " 'sifts_import_end',\n",
       " 'signalp',\n",
       " 'signalp_start',\n",
       " 'signalp_end',\n",
       " 'tmhmm',\n",
       " 'tmhmm_start',\n",
       " 'tmhmm_end',\n",
       " 'structure_gene_stable_id',\n",
       " 'structure_gene_stable_id_version',\n",
       " 'structure_gene_version',\n",
       " 'structure_transcript_stable_id',\n",
       " 'structure_transcript_stable_id_version',\n",
       " 'structure_transcript_version',\n",
       " 'structure_translation_stable_id',\n",
       " 'structure_translation_stable_id_version',\n",
       " 'structure_peptide_version',\n",
       " 'structure_canonical_transcript_id',\n",
       " 'structure_chrom_name',\n",
       " 'structure_gene_chrom_start',\n",
       " 'structure_gene_chrom_end',\n",
       " 'structure_transcript_chrom_start',\n",
       " 'structure_transcript_chrom_end',\n",
       " 'structure_transcription_start_site',\n",
       " 'structure_transcript_length',\n",
       " 'structure_transcript_chrom_strand',\n",
       " 'structure_external_gene_name',\n",
       " 'structure_external_source_name',\n",
       " 'structure_5_utr_start',\n",
       " 'structure_5_utr_end',\n",
       " 'structure_3_utr_start',\n",
       " 'structure_3_utr_end',\n",
       " 'structure_cds_length',\n",
       " 'structure_cdna_length',\n",
       " 'structure_peptide_length',\n",
       " 'struct_transcript_count',\n",
       " 'structure_translation_count',\n",
       " 'structure_description',\n",
       " 'structure_biotype',\n",
       " 'structure_ensembl_exon_id',\n",
       " 'exon_chrom_start',\n",
       " 'exon_chrom_end',\n",
       " 'is_constitutive',\n",
       " 'rank',\n",
       " 'phase',\n",
       " 'end_phase',\n",
       " 'cdna_coding_start',\n",
       " 'cdna_coding_end',\n",
       " 'genomic_coding_start',\n",
       " 'genomic_coding_end',\n",
       " 'structure_cds_start',\n",
       " 'structure_cds_end',\n",
       " 'homologs_ensembl_gene_id',\n",
       " 'homologs_gene_stable_id_version',\n",
       " 'homologs_gene_version',\n",
       " 'homologs_ensembl_transcript_id',\n",
       " 'homologs_transcript_stable_id_version',\n",
       " 'homologs_transcript_version',\n",
       " 'homologs_ensembl_peptide_id',\n",
       " 'homologs_translation_stable_id_version',\n",
       " 'homologs_peptide_version',\n",
       " 'homologs_canonical_transcript_id',\n",
       " 'homologs_chromosome_name',\n",
       " 'homologs_start_position',\n",
       " 'homologs_end_position',\n",
       " 'homologs_strand',\n",
       " 'homologs_band',\n",
       " 'homologs_external_gene_name',\n",
       " 'homologs_external_gene_source',\n",
       " 'homologs_ensembl_CDS_length',\n",
       " 'homologs_ensembl_cDNA_length',\n",
       " 'homologs_ensembl_peptide_length',\n",
       " 'homologs_transcript_count',\n",
       " 'homologs_percentage_gc_content',\n",
       " 'homologs_description',\n",
       " 'cabingdonii_homolog_ensembl_gene',\n",
       " 'cabingdonii_homolog_associated_gene_name',\n",
       " 'cabingdonii_homolog_ensembl_peptide',\n",
       " 'cabingdonii_homolog_chromosome',\n",
       " 'cabingdonii_homolog_chrom_start',\n",
       " 'cabingdonii_homolog_chrom_end',\n",
       " 'cabingdonii_homolog_canonical_transcript_protein',\n",
       " 'cabingdonii_homolog_subtype',\n",
       " 'cabingdonii_homolog_orthology_type',\n",
       " 'cabingdonii_homolog_perc_id',\n",
       " 'cabingdonii_homolog_perc_id_r1',\n",
       " 'cabingdonii_homolog_goc_score',\n",
       " 'cabingdonii_homolog_wga_coverage',\n",
       " 'cabingdonii_homolog_dn',\n",
       " 'cabingdonii_homolog_ds',\n",
       " 'cabingdonii_homolog_orthology_confidence',\n",
       " 'gagassizii_homolog_ensembl_gene',\n",
       " 'gagassizii_homolog_associated_gene_name',\n",
       " 'gagassizii_homolog_ensembl_peptide',\n",
       " 'gagassizii_homolog_chromosome',\n",
       " 'gagassizii_homolog_chrom_start',\n",
       " 'gagassizii_homolog_chrom_end',\n",
       " 'gagassizii_homolog_canonical_transcript_protein',\n",
       " 'gagassizii_homolog_subtype',\n",
       " 'gagassizii_homolog_orthology_type',\n",
       " 'gagassizii_homolog_perc_id',\n",
       " 'gagassizii_homolog_perc_id_r1',\n",
       " 'gagassizii_homolog_goc_score',\n",
       " 'gagassizii_homolog_wga_coverage',\n",
       " 'gagassizii_homolog_dn',\n",
       " 'gagassizii_homolog_ds',\n",
       " 'gagassizii_homolog_orthology_confidence',\n",
       " 'mspretus_homolog_ensembl_gene',\n",
       " 'mspretus_homolog_associated_gene_name',\n",
       " 'mspretus_homolog_ensembl_peptide',\n",
       " 'mspretus_homolog_chromosome',\n",
       " 'mspretus_homolog_chrom_start',\n",
       " 'mspretus_homolog_chrom_end',\n",
       " 'mspretus_homolog_canonical_transcript_protein',\n",
       " 'mspretus_homolog_subtype',\n",
       " 'mspretus_homolog_orthology_type',\n",
       " 'mspretus_homolog_perc_id',\n",
       " 'mspretus_homolog_perc_id_r1',\n",
       " 'mspretus_homolog_goc_score',\n",
       " 'mspretus_homolog_wga_coverage',\n",
       " 'mspretus_homolog_dn',\n",
       " 'mspretus_homolog_ds',\n",
       " 'mspretus_homolog_orthology_confidence',\n",
       " 'vpacos_homolog_ensembl_gene',\n",
       " 'vpacos_homolog_associated_gene_name',\n",
       " 'vpacos_homolog_ensembl_peptide',\n",
       " 'vpacos_homolog_chromosome',\n",
       " 'vpacos_homolog_chrom_start',\n",
       " 'vpacos_homolog_chrom_end',\n",
       " 'vpacos_homolog_canonical_transcript_protein',\n",
       " 'vpacos_homolog_subtype',\n",
       " 'vpacos_homolog_orthology_type',\n",
       " 'vpacos_homolog_perc_id',\n",
       " 'vpacos_homolog_perc_id_r1',\n",
       " 'vpacos_homolog_goc_score',\n",
       " 'vpacos_homolog_wga_coverage',\n",
       " 'vpacos_homolog_dn',\n",
       " 'vpacos_homolog_ds',\n",
       " 'vpacos_homolog_orthology_confidence',\n",
       " 'mmarmota_homolog_ensembl_gene',\n",
       " 'mmarmota_homolog_associated_gene_name',\n",
       " 'mmarmota_homolog_ensembl_peptide',\n",
       " 'mmarmota_homolog_chromosome',\n",
       " 'mmarmota_homolog_chrom_start',\n",
       " 'mmarmota_homolog_chrom_end',\n",
       " 'mmarmota_homolog_canonical_transcript_protein',\n",
       " 'mmarmota_homolog_subtype',\n",
       " 'mmarmota_homolog_orthology_type',\n",
       " 'mmarmota_homolog_perc_id',\n",
       " 'mmarmota_homolog_perc_id_r1',\n",
       " 'mmarmota_homolog_goc_score',\n",
       " 'mmarmota_homolog_wga_coverage',\n",
       " 'mmarmota_homolog_dn',\n",
       " 'mmarmota_homolog_ds',\n",
       " 'mmarmota_homolog_orthology_confidence',\n",
       " 'pformosa_homolog_ensembl_gene',\n",
       " 'pformosa_homolog_associated_gene_name',\n",
       " 'pformosa_homolog_ensembl_peptide',\n",
       " 'pformosa_homolog_chromosome',\n",
       " 'pformosa_homolog_chrom_start',\n",
       " 'pformosa_homolog_chrom_end',\n",
       " 'pformosa_homolog_canonical_transcript_protein',\n",
       " 'pformosa_homolog_subtype',\n",
       " 'pformosa_homolog_orthology_type',\n",
       " 'pformosa_homolog_perc_id',\n",
       " 'pformosa_homolog_perc_id_r1',\n",
       " 'pformosa_homolog_goc_score',\n",
       " 'pformosa_homolog_wga_coverage',\n",
       " 'pformosa_homolog_dn',\n",
       " 'pformosa_homolog_ds',\n",
       " 'pformosa_homolog_orthology_confidence',\n",
       " 'ccanadensis_homolog_ensembl_gene',\n",
       " 'ccanadensis_homolog_associated_gene_name',\n",
       " 'ccanadensis_homolog_ensembl_peptide',\n",
       " 'ccanadensis_homolog_chromosome',\n",
       " 'ccanadensis_homolog_chrom_start',\n",
       " 'ccanadensis_homolog_chrom_end',\n",
       " 'ccanadensis_homolog_canonical_transcript_protein',\n",
       " 'ccanadensis_homolog_subtype',\n",
       " 'ccanadensis_homolog_orthology_type',\n",
       " 'ccanadensis_homolog_perc_id',\n",
       " 'ccanadensis_homolog_perc_id_r1',\n",
       " 'ccanadensis_homolog_goc_score',\n",
       " 'ccanadensis_homolog_wga_coverage',\n",
       " 'ccanadensis_homolog_dn',\n",
       " 'ccanadensis_homolog_ds',\n",
       " 'ccanadensis_homolog_orthology_confidence',\n",
       " 'bbison_homolog_ensembl_gene',\n",
       " 'bbison_homolog_associated_gene_name',\n",
       " 'bbison_homolog_ensembl_peptide',\n",
       " 'bbison_homolog_chromosome',\n",
       " 'bbison_homolog_chrom_start',\n",
       " 'bbison_homolog_chrom_end',\n",
       " 'bbison_homolog_canonical_transcript_protein',\n",
       " 'bbison_homolog_subtype',\n",
       " 'bbison_homolog_orthology_type',\n",
       " 'bbison_homolog_perc_id',\n",
       " 'bbison_homolog_perc_id_r1',\n",
       " 'bbison_homolog_goc_score',\n",
       " 'bbison_homolog_wga_coverage',\n",
       " 'bbison_homolog_dn',\n",
       " 'bbison_homolog_ds',\n",
       " 'bbison_homolog_orthology_confidence',\n",
       " 'uamericanus_homolog_ensembl_gene',\n",
       " 'uamericanus_homolog_associated_gene_name',\n",
       " 'uamericanus_homolog_ensembl_peptide',\n",
       " 'uamericanus_homolog_chromosome',\n",
       " 'uamericanus_homolog_chrom_start',\n",
       " 'uamericanus_homolog_chrom_end',\n",
       " 'uamericanus_homolog_canonical_transcript_protein',\n",
       " 'uamericanus_homolog_subtype',\n",
       " 'uamericanus_homolog_orthology_type',\n",
       " 'uamericanus_homolog_perc_id',\n",
       " 'uamericanus_homolog_perc_id_r1',\n",
       " 'uamericanus_homolog_goc_score',\n",
       " 'uamericanus_homolog_wga_coverage',\n",
       " 'uamericanus_homolog_dn',\n",
       " 'uamericanus_homolog_ds',\n",
       " 'uamericanus_homolog_orthology_confidence',\n",
       " 'nvison_homolog_ensembl_gene',\n",
       " 'nvison_homolog_associated_gene_name',\n",
       " 'nvison_homolog_ensembl_peptide',\n",
       " 'nvison_homolog_chromosome',\n",
       " 'nvison_homolog_chrom_start',\n",
       " 'nvison_homolog_chrom_end',\n",
       " 'nvison_homolog_canonical_transcript_protein',\n",
       " 'nvison_homolog_subtype',\n",
       " 'nvison_homolog_orthology_type',\n",
       " 'nvison_homolog_perc_id',\n",
       " 'nvison_homolog_perc_id_r1',\n",
       " 'nvison_homolog_goc_score',\n",
       " 'nvison_homolog_wga_coverage',\n",
       " 'nvison_homolog_dn',\n",
       " 'nvison_homolog_ds',\n",
       " 'nvison_homolog_orthology_confidence',\n",
       " 'cpalliatus_homolog_ensembl_gene',\n",
       " 'cpalliatus_homolog_associated_gene_name',\n",
       " 'cpalliatus_homolog_ensembl_peptide',\n",
       " 'cpalliatus_homolog_chromosome',\n",
       " 'cpalliatus_homolog_chrom_start',\n",
       " 'cpalliatus_homolog_chrom_end',\n",
       " 'cpalliatus_homolog_canonical_transcript_protein',\n",
       " 'cpalliatus_homolog_subtype',\n",
       " 'cpalliatus_homolog_orthology_type',\n",
       " 'cpalliatus_homolog_perc_id',\n",
       " 'cpalliatus_homolog_perc_id_r1',\n",
       " 'cpalliatus_homolog_goc_score',\n",
       " 'cpalliatus_homolog_wga_coverage',\n",
       " 'cpalliatus_homolog_dn',\n",
       " 'cpalliatus_homolog_ds',\n",
       " 'cpalliatus_homolog_orthology_confidence',\n",
       " 'acarolinensis_homolog_ensembl_gene',\n",
       " 'acarolinensis_homolog_associated_gene_name',\n",
       " 'acarolinensis_homolog_ensembl_peptide',\n",
       " 'acarolinensis_homolog_chromosome',\n",
       " 'acarolinensis_homolog_chrom_start',\n",
       " 'acarolinensis_homolog_chrom_end',\n",
       " 'acarolinensis_homolog_canonical_transcript_protein',\n",
       " 'acarolinensis_homolog_subtype',\n",
       " 'acarolinensis_homolog_orthology_type',\n",
       " 'acarolinensis_homolog_perc_id',\n",
       " 'acarolinensis_homolog_perc_id_r1',\n",
       " 'acarolinensis_homolog_goc_score',\n",
       " 'acarolinensis_homolog_wga_coverage',\n",
       " 'acarolinensis_homolog_dn',\n",
       " 'acarolinensis_homolog_ds',\n",
       " 'acarolinensis_homolog_orthology_confidence',\n",
       " 'uparryii_homolog_ensembl_gene',\n",
       " 'uparryii_homolog_associated_gene_name',\n",
       " 'uparryii_homolog_ensembl_peptide',\n",
       " 'uparryii_homolog_chromosome',\n",
       " 'uparryii_homolog_chrom_start',\n",
       " 'uparryii_homolog_chrom_end',\n",
       " 'uparryii_homolog_canonical_transcript_protein',\n",
       " 'uparryii_homolog_subtype',\n",
       " 'uparryii_homolog_orthology_type',\n",
       " 'uparryii_homolog_perc_id',\n",
       " 'uparryii_homolog_perc_id_r1',\n",
       " 'uparryii_homolog_goc_score',\n",
       " 'uparryii_homolog_wga_coverage',\n",
       " 'uparryii_homolog_dn',\n",
       " 'uparryii_homolog_ds',\n",
       " 'uparryii_homolog_orthology_confidence',\n",
       " 'smerianae_homolog_ensembl_gene',\n",
       " 'smerianae_homolog_associated_gene_name',\n",
       " 'smerianae_homolog_ensembl_peptide',\n",
       " 'smerianae_homolog_chromosome',\n",
       " 'smerianae_homolog_chrom_start',\n",
       " 'smerianae_homolog_chrom_end',\n",
       " 'smerianae_homolog_canonical_transcript_protein',\n",
       " 'smerianae_homolog_subtype',\n",
       " 'smerianae_homolog_orthology_type',\n",
       " 'smerianae_homolog_perc_id',\n",
       " 'smerianae_homolog_perc_id_r1',\n",
       " 'smerianae_homolog_goc_score',\n",
       " 'smerianae_homolog_wga_coverage',\n",
       " 'smerianae_homolog_dn',\n",
       " 'smerianae_homolog_ds',\n",
       " 'smerianae_homolog_orthology_confidence',\n",
       " 'dnovemcinctus_homolog_ensembl_gene',\n",
       " 'dnovemcinctus_homolog_associated_gene_name',\n",
       " 'dnovemcinctus_homolog_ensembl_peptide',\n",
       " 'dnovemcinctus_homolog_chromosome',\n",
       " 'dnovemcinctus_homolog_chrom_start',\n",
       " 'dnovemcinctus_homolog_chrom_end',\n",
       " 'dnovemcinctus_homolog_canonical_transcript_protein',\n",
       " 'dnovemcinctus_homolog_subtype',\n",
       " 'dnovemcinctus_homolog_orthology_type',\n",
       " 'dnovemcinctus_homolog_perc_id',\n",
       " 'dnovemcinctus_homolog_perc_id_r1',\n",
       " 'dnovemcinctus_homolog_goc_score',\n",
       " 'dnovemcinctus_homolog_wga_coverage',\n",
       " 'dnovemcinctus_homolog_dn',\n",
       " 'dnovemcinctus_homolog_ds',\n",
       " 'dnovemcinctus_homolog_orthology_confidence',\n",
       " 'sformosus_homolog_ensembl_gene',\n",
       " 'sformosus_homolog_associated_gene_name',\n",
       " 'sformosus_homolog_ensembl_peptide',\n",
       " 'sformosus_homolog_chromosome',\n",
       " 'sformosus_homolog_chrom_start',\n",
       " 'sformosus_homolog_chrom_end',\n",
       " 'sformosus_homolog_canonical_transcript_protein',\n",
       " 'sformosus_homolog_subtype',\n",
       " 'sformosus_homolog_orthology_type',\n",
       " 'sformosus_homolog_perc_id',\n",
       " 'sformosus_homolog_perc_id_r1',\n",
       " 'sformosus_homolog_goc_score',\n",
       " 'sformosus_homolog_wga_coverage',\n",
       " 'sformosus_homolog_dn',\n",
       " 'sformosus_homolog_ds',\n",
       " 'sformosus_homolog_orthology_confidence',\n",
       " 'cporosus_homolog_ensembl_gene',\n",
       " 'cporosus_homolog_associated_gene_name',\n",
       " 'cporosus_homolog_ensembl_peptide',\n",
       " 'cporosus_homolog_chromosome',\n",
       " 'cporosus_homolog_chrom_start',\n",
       " 'cporosus_homolog_chrom_end',\n",
       " 'cporosus_homolog_canonical_transcript_protein',\n",
       " 'cporosus_homolog_subtype',\n",
       " 'cporosus_homolog_orthology_type',\n",
       " 'cporosus_homolog_perc_id',\n",
       " 'cporosus_homolog_perc_id_r1',\n",
       " 'cporosus_homolog_goc_score',\n",
       " 'cporosus_homolog_wga_coverage',\n",
       " 'cporosus_homolog_dn',\n",
       " 'cporosus_homolog_ds',\n",
       " 'cporosus_homolog_orthology_confidence',\n",
       " 'lbergylta_homolog_ensembl_gene',\n",
       " 'lbergylta_homolog_associated_gene_name',\n",
       " 'lbergylta_homolog_ensembl_peptide',\n",
       " 'lbergylta_homolog_chromosome',\n",
       " 'lbergylta_homolog_chrom_start',\n",
       " 'lbergylta_homolog_chrom_end',\n",
       " 'lbergylta_homolog_canonical_transcript_protein',\n",
       " 'lbergylta_homolog_subtype',\n",
       " 'lbergylta_homolog_orthology_type',\n",
       " 'lbergylta_homolog_perc_id',\n",
       " 'lbergylta_homolog_perc_id_r1',\n",
       " 'lbergylta_homolog_goc_score',\n",
       " 'lbergylta_homolog_wga_coverage',\n",
       " 'lbergylta_homolog_dn',\n",
       " 'lbergylta_homolog_ds',\n",
       " 'lbergylta_homolog_orthology_confidence',\n",
       " 'lcalcarifer_homolog_ensembl_gene',\n",
       " 'lcalcarifer_homolog_associated_gene_name',\n",
       " 'lcalcarifer_homolog_ensembl_peptide',\n",
       " 'lcalcarifer_homolog_chromosome',\n",
       " 'lcalcarifer_homolog_chrom_start',\n",
       " 'lcalcarifer_homolog_chrom_end',\n",
       " 'lcalcarifer_homolog_canonical_transcript_protein',\n",
       " 'lcalcarifer_homolog_subtype',\n",
       " 'lcalcarifer_homolog_orthology_type',\n",
       " 'lcalcarifer_homolog_perc_id',\n",
       " 'lcalcarifer_homolog_perc_id_r1',\n",
       " 'lcalcarifer_homolog_goc_score',\n",
       " 'lcalcarifer_homolog_wga_coverage',\n",
       " 'lcalcarifer_homolog_dn',\n",
       " 'lcalcarifer_homolog_ds',\n",
       " 'lcalcarifer_homolog_orthology_confidence',\n",
       " 'ldomestica_homolog_ensembl_gene',\n",
       " 'ldomestica_homolog_associated_gene_name',\n",
       " 'ldomestica_homolog_ensembl_peptide',\n",
       " 'ldomestica_homolog_chromosome',\n",
       " 'ldomestica_homolog_chrom_start',\n",
       " 'ldomestica_homolog_chrom_end',\n",
       " 'ldomestica_homolog_canonical_transcript_protein',\n",
       " 'ldomestica_homolog_subtype',\n",
       " 'ldomestica_homolog_orthology_type',\n",
       " 'ldomestica_homolog_perc_id',\n",
       " 'ldomestica_homolog_perc_id_r1',\n",
       " 'ldomestica_homolog_goc_score',\n",
       " 'ldomestica_homolog_wga_coverage',\n",
       " 'ldomestica_homolog_dn',\n",
       " 'ldomestica_homolog_ds',\n",
       " 'ldomestica_homolog_orthology_confidence',\n",
       " 'spartitus_homolog_ensembl_gene',\n",
       " 'spartitus_homolog_associated_gene_name',\n",
       " 'spartitus_homolog_ensembl_peptide',\n",
       " 'spartitus_homolog_chromosome',\n",
       " 'spartitus_homolog_chrom_start',\n",
       " 'spartitus_homolog_chrom_end',\n",
       " 'spartitus_homolog_canonical_transcript_protein',\n",
       " 'spartitus_homolog_subtype',\n",
       " 'spartitus_homolog_orthology_type',\n",
       " 'spartitus_homolog_perc_id',\n",
       " 'spartitus_homolog_perc_id_r1',\n",
       " 'spartitus_homolog_goc_score',\n",
       " 'spartitus_homolog_wga_coverage',\n",
       " 'spartitus_homolog_dn',\n",
       " 'spartitus_homolog_ds',\n",
       " 'spartitus_homolog_orthology_confidence',\n",
       " 'rbieti_homolog_ensembl_gene',\n",
       " 'rbieti_homolog_associated_gene_name',\n",
       " 'rbieti_homolog_ensembl_peptide',\n",
       " 'rbieti_homolog_chromosome',\n",
       " 'rbieti_homolog_chrom_start',\n",
       " 'rbieti_homolog_chrom_end',\n",
       " 'rbieti_homolog_canonical_transcript_protein',\n",
       " 'rbieti_homolog_subtype',\n",
       " 'rbieti_homolog_orthology_type',\n",
       " 'rbieti_homolog_perc_id',\n",
       " 'rbieti_homolog_perc_id_r1',\n",
       " 'rbieti_homolog_goc_score',\n",
       " 'rbieti_homolog_wga_coverage',\n",
       " 'rbieti_homolog_dn',\n",
       " 'rbieti_homolog_ds',\n",
       " 'rbieti_homolog_orthology_confidence',\n",
       " 'ccaeruleus_homolog_ensembl_gene',\n",
       " 'ccaeruleus_homolog_associated_gene_name',\n",
       " 'ccaeruleus_homolog_ensembl_peptide',\n",
       " 'ccaeruleus_homolog_chromosome',\n",
       " 'ccaeruleus_homolog_chrom_start',\n",
       " 'ccaeruleus_homolog_chrom_end',\n",
       " 'ccaeruleus_homolog_canonical_transcript_protein',\n",
       " 'ccaeruleus_homolog_subtype',\n",
       " 'ccaeruleus_homolog_orthology_type',\n",
       " 'ccaeruleus_homolog_perc_id',\n",
       " 'ccaeruleus_homolog_perc_id_r1',\n",
       " 'ccaeruleus_homolog_goc_score',\n",
       " 'ccaeruleus_homolog_wga_coverage',\n",
       " 'ccaeruleus_homolog_dn',\n",
       " 'ccaeruleus_homolog_ds',\n",
       " 'ccaeruleus_homolog_orthology_confidence',\n",
       " 'lcoronata_homolog_ensembl_gene',\n",
       " 'lcoronata_homolog_associated_gene_name',\n",
       " 'lcoronata_homolog_ensembl_peptide',\n",
       " 'lcoronata_homolog_chromosome',\n",
       " 'lcoronata_homolog_chrom_start',\n",
       " 'lcoronata_homolog_chrom_end',\n",
       " 'lcoronata_homolog_canonical_transcript_protein',\n",
       " 'lcoronata_homolog_subtype',\n",
       " 'lcoronata_homolog_orthology_type',\n",
       " 'lcoronata_homolog_perc_id',\n",
       " 'lcoronata_homolog_perc_id_r1',\n",
       " 'lcoronata_homolog_goc_score',\n",
       " 'lcoronata_homolog_wga_coverage',\n",
       " 'lcoronata_homolog_dn',\n",
       " 'lcoronata_homolog_ds',\n",
       " 'lcoronata_homolog_orthology_confidence',\n",
       " 'sboliviensis_homolog_ensembl_gene',\n",
       " 'sboliviensis_homolog_associated_gene_name',\n",
       " 'sboliviensis_homolog_ensembl_peptide',\n",
       " 'sboliviensis_homolog_chromosome',\n",
       " 'sboliviensis_homolog_chrom_start',\n",
       " 'sboliviensis_homolog_chrom_end',\n",
       " 'sboliviensis_homolog_canonical_transcript_protein',\n",
       " 'sboliviensis_homolog_subtype',\n",
       " 'sboliviensis_homolog_orthology_type',\n",
       " 'sboliviensis_homolog_perc_id',\n",
       " 'sboliviensis_homolog_perc_id_r1',\n",
       " 'sboliviensis_homolog_goc_score',\n",
       " 'sboliviensis_homolog_wga_coverage',\n",
       " 'sboliviensis_homolog_dn',\n",
       " 'sboliviensis_homolog_ds',\n",
       " 'sboliviensis_homolog_orthology_confidence',\n",
       " 'ppaniscus_homolog_ensembl_gene',\n",
       " 'ppaniscus_homolog_associated_gene_name',\n",
       " 'ppaniscus_homolog_ensembl_peptide',\n",
       " 'ppaniscus_homolog_chromosome',\n",
       " 'ppaniscus_homolog_chrom_start',\n",
       " 'ppaniscus_homolog_chrom_end',\n",
       " 'ppaniscus_homolog_canonical_transcript_protein',\n",
       " 'ppaniscus_homolog_subtype',\n",
       " 'ppaniscus_homolog_orthology_type',\n",
       " 'ppaniscus_homolog_perc_id',\n",
       " 'ppaniscus_homolog_perc_id_r1',\n",
       " 'ppaniscus_homolog_goc_score',\n",
       " 'ppaniscus_homolog_wga_coverage',\n",
       " 'ppaniscus_homolog_dn',\n",
       " 'ppaniscus_homolog_ds',\n",
       " 'ppaniscus_homolog_orthology_confidence',\n",
       " 'caperea_homolog_ensembl_gene',\n",
       " 'caperea_homolog_associated_gene_name',\n",
       " 'caperea_homolog_ensembl_peptide',\n",
       " 'caperea_homolog_chromosome',\n",
       " 'caperea_homolog_chrom_start',\n",
       " 'caperea_homolog_chrom_end',\n",
       " 'caperea_homolog_canonical_transcript_protein',\n",
       " 'caperea_homolog_subtype',\n",
       " 'caperea_homolog_orthology_type',\n",
       " 'caperea_homolog_perc_id',\n",
       " 'caperea_homolog_perc_id_r1',\n",
       " 'caperea_homolog_goc_score',\n",
       " 'caperea_homolog_wga_coverage',\n",
       " 'caperea_homolog_dn',\n",
       " 'caperea_homolog_ds',\n",
       " 'caperea_homolog_orthology_confidence',\n",
       " 'mundulatus_homolog_ensembl_gene',\n",
       " 'mundulatus_homolog_associated_gene_name',\n",
       " 'mundulatus_homolog_ensembl_peptide',\n",
       " 'mundulatus_homolog_chromosome',\n",
       " 'mundulatus_homolog_chrom_start',\n",
       " 'mundulatus_homolog_chrom_end',\n",
       " 'mundulatus_homolog_canonical_transcript_protein',\n",
       " 'mundulatus_homolog_subtype',\n",
       " 'mundulatus_homolog_orthology_type',\n",
       " 'mundulatus_homolog_perc_id',\n",
       " 'mundulatus_homolog_perc_id_r1',\n",
       " 'mundulatus_homolog_goc_score',\n",
       " 'mundulatus_homolog_wga_coverage',\n",
       " 'mundulatus_homolog_dn',\n",
       " 'mundulatus_homolog_ds',\n",
       " 'mundulatus_homolog_orthology_confidence',\n",
       " 'hburtoni_homolog_ensembl_gene',\n",
       " 'hburtoni_homolog_associated_gene_name',\n",
       " 'hburtoni_homolog_ensembl_peptide',\n",
       " 'hburtoni_homolog_chromosome',\n",
       " 'hburtoni_homolog_chrom_start',\n",
       " 'hburtoni_homolog_chrom_end',\n",
       " 'hburtoni_homolog_canonical_transcript_protein',\n",
       " 'hburtoni_homolog_subtype',\n",
       " 'hburtoni_homolog_orthology_type',\n",
       " 'hburtoni_homolog_perc_id',\n",
       " 'hburtoni_homolog_perc_id_r1',\n",
       " 'hburtoni_homolog_goc_score',\n",
       " 'hburtoni_homolog_wga_coverage',\n",
       " 'hburtoni_homolog_dn',\n",
       " 'hburtoni_homolog_ds',\n",
       " 'hburtoni_homolog_orthology_confidence',\n",
       " 'ogarnettii_homolog_ensembl_gene',\n",
       " 'ogarnettii_homolog_associated_gene_name',\n",
       " 'ogarnettii_homolog_ensembl_peptide',\n",
       " 'ogarnettii_homolog_chromosome',\n",
       " 'ogarnettii_homolog_chrom_start',\n",
       " 'ogarnettii_homolog_chrom_end',\n",
       " 'ogarnettii_homolog_canonical_transcript_protein',\n",
       " 'ogarnettii_homolog_subtype',\n",
       " 'ogarnettii_homolog_orthology_type',\n",
       " 'ogarnettii_homolog_perc_id',\n",
       " 'ogarnettii_homolog_perc_id_r1',\n",
       " 'ogarnettii_homolog_goc_score',\n",
       " 'ogarnettii_homolog_wga_coverage',\n",
       " 'ogarnettii_homolog_dn',\n",
       " 'ogarnettii_homolog_ds',\n",
       " 'ogarnettii_homolog_orthology_confidence',\n",
       " 'cintestinalis_homolog_ensembl_gene',\n",
       " 'cintestinalis_homolog_associated_gene_name',\n",
       " 'cintestinalis_homolog_ensembl_peptide',\n",
       " 'cintestinalis_homolog_chromosome',\n",
       " 'cintestinalis_homolog_chrom_start',\n",
       " 'cintestinalis_homolog_chrom_end',\n",
       " 'cintestinalis_homolog_canonical_transcript_protein',\n",
       " 'cintestinalis_homolog_subtype',\n",
       " 'cintestinalis_homolog_orthology_type',\n",
       " 'cintestinalis_homolog_perc_id',\n",
       " 'cintestinalis_homolog_perc_id_r1',\n",
       " 'cintestinalis_homolog_goc_score',\n",
       " 'cintestinalis_homolog_wga_coverage',\n",
       " 'cintestinalis_homolog_dn',\n",
       " 'cintestinalis_homolog_ds',\n",
       " 'cintestinalis_homolog_orthology_confidence',\n",
       " 'csavignyi_homolog_ensembl_gene',\n",
       " 'csavignyi_homolog_associated_gene_name',\n",
       " 'csavignyi_homolog_ensembl_peptide',\n",
       " 'csavignyi_homolog_chromosome',\n",
       " 'csavignyi_homolog_chrom_start',\n",
       " 'csavignyi_homolog_chrom_end',\n",
       " 'csavignyi_homolog_canonical_transcript_protein',\n",
       " 'csavignyi_homolog_subtype',\n",
       " 'csavignyi_homolog_orthology_type',\n",
       " 'csavignyi_homolog_perc_id',\n",
       " 'csavignyi_homolog_perc_id_r1',\n",
       " 'csavignyi_homolog_goc_score',\n",
       " 'csavignyi_homolog_wga_coverage',\n",
       " 'csavignyi_homolog_dn',\n",
       " 'csavignyi_homolog_ds',\n",
       " 'csavignyi_homolog_orthology_confidence',\n",
       " 'celegans_homolog_ensembl_gene',\n",
       " 'celegans_homolog_associated_gene_name',\n",
       " 'celegans_homolog_ensembl_peptide',\n",
       " 'celegans_homolog_chromosome',\n",
       " 'celegans_homolog_chrom_start',\n",
       " 'celegans_homolog_chrom_end',\n",
       " 'celegans_homolog_canonical_transcript_protein',\n",
       " 'celegans_homolog_subtype',\n",
       " 'celegans_homolog_orthology_type',\n",
       " 'celegans_homolog_perc_id',\n",
       " 'celegans_homolog_perc_id_r1',\n",
       " 'celegans_homolog_goc_score',\n",
       " 'celegans_homolog_wga_coverage',\n",
       " 'celegans_homolog_dn',\n",
       " 'celegans_homolog_ds',\n",
       " 'celegans_homolog_orthology_confidence',\n",
       " 'ccapucinus_homolog_ensembl_gene',\n",
       " 'ccapucinus_homolog_associated_gene_name',\n",
       " 'ccapucinus_homolog_ensembl_peptide',\n",
       " 'ccapucinus_homolog_chromosome',\n",
       " 'ccapucinus_homolog_chrom_start',\n",
       " 'ccapucinus_homolog_chrom_end',\n",
       " 'ccapucinus_homolog_canonical_transcript_protein',\n",
       " 'ccapucinus_homolog_subtype',\n",
       " 'ccapucinus_homolog_orthology_type',\n",
       " 'ccapucinus_homolog_perc_id',\n",
       " 'ccapucinus_homolog_perc_id_r1',\n",
       " 'ccapucinus_homolog_goc_score',\n",
       " 'ccapucinus_homolog_wga_coverage',\n",
       " 'ccapucinus_homolog_dn',\n",
       " 'ccapucinus_homolog_ds',\n",
       " 'ccapucinus_homolog_orthology_confidence',\n",
       " 'fcatus_homolog_ensembl_gene',\n",
       " 'fcatus_homolog_associated_gene_name',\n",
       " 'fcatus_homolog_ensembl_peptide',\n",
       " 'fcatus_homolog_chromosome',\n",
       " 'fcatus_homolog_chrom_start',\n",
       " 'fcatus_homolog_chrom_end',\n",
       " 'fcatus_homolog_canonical_transcript_protein',\n",
       " 'fcatus_homolog_subtype',\n",
       " 'fcatus_homolog_orthology_type',\n",
       " 'fcatus_homolog_perc_id',\n",
       " 'fcatus_homolog_perc_id_r1',\n",
       " 'fcatus_homolog_goc_score',\n",
       " 'fcatus_homolog_wga_coverage',\n",
       " 'fcatus_homolog_dn',\n",
       " 'fcatus_homolog_ds',\n",
       " 'fcatus_homolog_orthology_confidence',\n",
       " 'pvitticeps_homolog_ensembl_gene',\n",
       " 'pvitticeps_homolog_associated_gene_name',\n",
       " 'pvitticeps_homolog_ensembl_peptide',\n",
       " 'pvitticeps_homolog_chromosome',\n",
       " 'pvitticeps_homolog_chrom_start',\n",
       " 'pvitticeps_homolog_chrom_end',\n",
       " 'pvitticeps_homolog_canonical_transcript_protein',\n",
       " 'pvitticeps_homolog_subtype',\n",
       " 'pvitticeps_homolog_orthology_type',\n",
       " 'pvitticeps_homolog_perc_id',\n",
       " 'pvitticeps_homolog_perc_id_r1',\n",
       " 'pvitticeps_homolog_goc_score',\n",
       " 'pvitticeps_homolog_wga_coverage',\n",
       " 'pvitticeps_homolog_dn',\n",
       " 'pvitticeps_homolog_ds',\n",
       " 'pvitticeps_homolog_orthology_confidence',\n",
       " 'ipunctatus_homolog_ensembl_gene',\n",
       " 'ipunctatus_homolog_associated_gene_name',\n",
       " 'ipunctatus_homolog_ensembl_peptide',\n",
       " 'ipunctatus_homolog_chromosome',\n",
       " 'ipunctatus_homolog_chrom_start',\n",
       " 'ipunctatus_homolog_chrom_end',\n",
       " 'ipunctatus_homolog_canonical_transcript_protein',\n",
       " 'ipunctatus_homolog_subtype',\n",
       " 'ipunctatus_homolog_orthology_type',\n",
       " 'ipunctatus_homolog_perc_id',\n",
       " 'ipunctatus_homolog_perc_id_r1',\n",
       " 'ipunctatus_homolog_goc_score',\n",
       " 'ipunctatus_homolog_wga_coverage',\n",
       " 'ipunctatus_homolog_dn',\n",
       " 'ipunctatus_homolog_ds',\n",
       " 'ipunctatus_homolog_orthology_confidence',\n",
       " 'ggallus_homolog_ensembl_gene',\n",
       " 'ggallus_homolog_associated_gene_name',\n",
       " 'ggallus_homolog_ensembl_peptide',\n",
       " 'ggallus_homolog_chromosome',\n",
       " 'ggallus_homolog_chrom_start',\n",
       " 'ggallus_homolog_chrom_end',\n",
       " 'ggallus_homolog_canonical_transcript_protein',\n",
       " 'ggallus_homolog_subtype',\n",
       " 'ggallus_homolog_orthology_type',\n",
       " 'ggallus_homolog_perc_id',\n",
       " 'ggallus_homolog_perc_id_r1',\n",
       " 'ggallus_homolog_goc_score',\n",
       " 'ggallus_homolog_wga_coverage',\n",
       " 'ggallus_homolog_dn',\n",
       " 'ggallus_homolog_ds',\n",
       " 'ggallus_homolog_orthology_confidence',\n",
       " 'nperdicaria_homolog_ensembl_gene',\n",
       " 'nperdicaria_homolog_associated_gene_name',\n",
       " 'nperdicaria_homolog_ensembl_peptide',\n",
       " 'nperdicaria_homolog_chromosome',\n",
       " 'nperdicaria_homolog_chrom_start',\n",
       " 'nperdicaria_homolog_chrom_end',\n",
       " 'nperdicaria_homolog_canonical_transcript_protein',\n",
       " 'nperdicaria_homolog_subtype',\n",
       " 'nperdicaria_homolog_orthology_type',\n",
       " 'nperdicaria_homolog_perc_id',\n",
       " 'nperdicaria_homolog_perc_id_r1',\n",
       " 'nperdicaria_homolog_goc_score',\n",
       " 'nperdicaria_homolog_wga_coverage',\n",
       " 'nperdicaria_homolog_dn',\n",
       " 'nperdicaria_homolog_ds',\n",
       " 'nperdicaria_homolog_orthology_confidence',\n",
       " 'ptroglodytes_homolog_ensembl_gene',\n",
       " 'ptroglodytes_homolog_associated_gene_name',\n",
       " 'ptroglodytes_homolog_ensembl_peptide',\n",
       " 'ptroglodytes_homolog_chromosome',\n",
       " 'ptroglodytes_homolog_chrom_start',\n",
       " 'ptroglodytes_homolog_chrom_end',\n",
       " 'ptroglodytes_homolog_canonical_transcript_protein',\n",
       " 'ptroglodytes_homolog_subtype',\n",
       " 'ptroglodytes_homolog_orthology_type',\n",
       " 'ptroglodytes_homolog_perc_id',\n",
       " 'ptroglodytes_homolog_perc_id_r1',\n",
       " 'ptroglodytes_homolog_goc_score',\n",
       " 'ptroglodytes_homolog_wga_coverage',\n",
       " 'ptroglodytes_homolog_dn',\n",
       " 'ptroglodytes_homolog_ds',\n",
       " 'ptroglodytes_homolog_orthology_confidence',\n",
       " 'cchok1gshd_homolog_ensembl_gene',\n",
       " 'cchok1gshd_homolog_associated_gene_name',\n",
       " 'cchok1gshd_homolog_ensembl_peptide',\n",
       " 'cchok1gshd_homolog_chromosome',\n",
       " 'cchok1gshd_homolog_chrom_start',\n",
       " 'cchok1gshd_homolog_chrom_end',\n",
       " 'cchok1gshd_homolog_canonical_transcript_protein',\n",
       " 'cchok1gshd_homolog_subtype',\n",
       " 'cchok1gshd_homolog_orthology_type',\n",
       " 'cchok1gshd_homolog_perc_id',\n",
       " 'cchok1gshd_homolog_perc_id_r1',\n",
       " 'cchok1gshd_homolog_goc_score',\n",
       " 'cchok1gshd_homolog_wga_coverage',\n",
       " 'cchok1gshd_homolog_dn',\n",
       " 'cchok1gshd_homolog_ds',\n",
       " 'cchok1gshd_homolog_orthology_confidence',\n",
       " 'ccrigri_homolog_ensembl_gene',\n",
       " 'ccrigri_homolog_associated_gene_name',\n",
       " 'ccrigri_homolog_ensembl_peptide',\n",
       " 'ccrigri_homolog_chromosome',\n",
       " 'ccrigri_homolog_chrom_start',\n",
       " 'ccrigri_homolog_chrom_end',\n",
       " 'ccrigri_homolog_canonical_transcript_protein',\n",
       " 'ccrigri_homolog_subtype',\n",
       " 'ccrigri_homolog_orthology_type',\n",
       " 'ccrigri_homolog_perc_id',\n",
       " 'ccrigri_homolog_perc_id_r1',\n",
       " 'ccrigri_homolog_goc_score',\n",
       " 'ccrigri_homolog_wga_coverage',\n",
       " 'ccrigri_homolog_dn',\n",
       " 'ccrigri_homolog_ds',\n",
       " 'ccrigri_homolog_orthology_confidence',\n",
       " 'cpicr_homolog_ensembl_gene',\n",
       " 'cpicr_homolog_associated_gene_name',\n",
       " 'cpicr_homolog_ensembl_peptide',\n",
       " 'cpicr_homolog_chromosome',\n",
       " 'cpicr_homolog_chrom_start',\n",
       " 'cpicr_homolog_chrom_end',\n",
       " 'cpicr_homolog_canonical_transcript_protein',\n",
       " 'cpicr_homolog_subtype',\n",
       " 'cpicr_homolog_orthology_type',\n",
       " 'cpicr_homolog_perc_id',\n",
       " 'cpicr_homolog_perc_id_r1',\n",
       " 'cpicr_homolog_goc_score',\n",
       " 'cpicr_homolog_wga_coverage',\n",
       " 'cpicr_homolog_dn',\n",
       " 'cpicr_homolog_ds',\n",
       " 'cpicr_homolog_orthology_confidence',\n",
       " 'psinensis_homolog_ensembl_gene',\n",
       " 'psinensis_homolog_associated_gene_name',\n",
       " 'psinensis_homolog_ensembl_peptide',\n",
       " 'psinensis_homolog_chromosome',\n",
       " 'psinensis_homolog_chrom_start',\n",
       " 'psinensis_homolog_chrom_end',\n",
       " 'psinensis_homolog_canonical_transcript_protein',\n",
       " 'psinensis_homolog_subtype',\n",
       " 'psinensis_homolog_orthology_type',\n",
       " 'psinensis_homolog_perc_id',\n",
       " 'psinensis_homolog_perc_id_r1',\n",
       " 'psinensis_homolog_goc_score',\n",
       " 'psinensis_homolog_wga_coverage',\n",
       " 'psinensis_homolog_dn',\n",
       " 'psinensis_homolog_ds',\n",
       " 'psinensis_homolog_orthology_confidence',\n",
       " 'atestudineus_homolog_ensembl_gene',\n",
       " 'atestudineus_homolog_associated_gene_name',\n",
       " 'atestudineus_homolog_ensembl_peptide',\n",
       " 'atestudineus_homolog_chromosome',\n",
       " 'atestudineus_homolog_chrom_start',\n",
       " 'atestudineus_homolog_chrom_end',\n",
       " 'atestudineus_homolog_canonical_transcript_protein',\n",
       " 'atestudineus_homolog_subtype',\n",
       " 'atestudineus_homolog_orthology_type',\n",
       " 'atestudineus_homolog_perc_id',\n",
       " 'atestudineus_homolog_perc_id_r1',\n",
       " 'atestudineus_homolog_goc_score',\n",
       " 'atestudineus_homolog_wga_coverage',\n",
       " 'atestudineus_homolog_dn',\n",
       " 'atestudineus_homolog_ds',\n",
       " 'atestudineus_homolog_orthology_confidence',\n",
       " 'aocellaris_homolog_ensembl_gene',\n",
       " 'aocellaris_homolog_associated_gene_name',\n",
       " 'aocellaris_homolog_ensembl_peptide',\n",
       " 'aocellaris_homolog_chromosome',\n",
       " 'aocellaris_homolog_chrom_start',\n",
       " 'aocellaris_homolog_chrom_end',\n",
       " 'aocellaris_homolog_canonical_transcript_protein',\n",
       " 'aocellaris_homolog_subtype',\n",
       " 'aocellaris_homolog_orthology_type',\n",
       " ...]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dataset.attributes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can map the gene stable ID with the mappings from the to get a sense of which pathways are linked to a particular gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'link_so_mini_closure': <biomart.Filter name='link_so_mini_closure', type='list'>,\n",
       " 'link_go_closure': <biomart.Filter name='link_go_closure', type='text'>,\n",
       " 'link_ensembl_transcript_stable_id': <biomart.Filter name='link_ensembl_transcript_stable_id', type='text'>,\n",
       " 'gene_id': <biomart.Filter name='gene_id', type='text'>,\n",
       " 'transcript_id': <biomart.Filter name='transcript_id', type='text'>,\n",
       " 'link_ensembl_gene_id': <biomart.Filter name='link_ensembl_gene_id', type='text'>,\n",
       " 'chromosome_name': <biomart.Filter name='chromosome_name', type='text'>,\n",
       " 'start': <biomart.Filter name='start', type='text'>,\n",
       " 'end': <biomart.Filter name='end', type='text'>,\n",
       " 'band_start': <biomart.Filter name='band_start', type='drop_down_basic_filter'>,\n",
       " 'band_end': <biomart.Filter name='band_end', type='drop_down_basic_filter'>,\n",
       " 'marker_start': <biomart.Filter name='marker_start', type='drop_down_basic_filter'>,\n",
       " 'marker_end': <biomart.Filter name='marker_end', type=''>,\n",
       " 'hsapiens_encode.type': <biomart.Filter name='hsapiens_encode.type', type='text'>,\n",
       " 'hsapiens_encode.encode_region': <biomart.Filter name='hsapiens_encode.encode_region', type=''>,\n",
       " 'strand': <biomart.Filter name='strand', type='text'>,\n",
       " 'chromosomal_region': <biomart.Filter name='chromosomal_region', type='text'>,\n",
       " 'id_list_xrefs_filters': <biomart.Filter name='id_list_xrefs_filters', type='boolean_list'>,\n",
       " 'id_list_limit_xrefs_filters': <biomart.Filter name='id_list_limit_xrefs_filters', type='id_list'>,\n",
       " 'id_list_microarray_filters': <biomart.Filter name='id_list_microarray_filters', type='boolean_list'>,\n",
       " 'id_list_limit_microarray_filters': <biomart.Filter name='id_list_limit_microarray_filters', type='id_list'>,\n",
       " 'transcript_count_greater_than': <biomart.Filter name='transcript_count_greater_than', type='text'>,\n",
       " 'transcript_count_less_than': <biomart.Filter name='transcript_count_less_than', type='text'>,\n",
       " 'biotype': <biomart.Filter name='biotype', type='list'>,\n",
       " 'transcript_biotype': <biomart.Filter name='transcript_biotype', type='list'>,\n",
       " 'source': <biomart.Filter name='source', type='list'>,\n",
       " 'transcript_source': <biomart.Filter name='transcript_source', type='list'>,\n",
       " 'transcript_tsl': <biomart.Filter name='transcript_tsl', type='boolean'>,\n",
       " 'transcript_gencode_basic': <biomart.Filter name='transcript_gencode_basic', type='boolean'>,\n",
       " 'transcript_appris': <biomart.Filter name='transcript_appris', type='boolean'>,\n",
       " 'mane_select': <biomart.Filter name='mane_select', type='boolean'>,\n",
       " 'phenotype_description': <biomart.Filter name='phenotype_description', type='list'>,\n",
       " 'phenotype_source': <biomart.Filter name='phenotype_source', type='list'>,\n",
       " 'go_clos': <biomart.Filter name='go_clos', type=''>,\n",
       " 'go_name': <biomart.Filter name='go_name', type=''>,\n",
       " 'go_evidence_code': <biomart.Filter name='go_evidence_code', type='list'>,\n",
       " 'homolog_filters': <biomart.Filter name='homolog_filters', type='boolean_list'>,\n",
       " 'id_list_protein_domain_and_feature_filters': <biomart.Filter name='id_list_protein_domain_and_feature_filters', type='boolean_list'>,\n",
       " 'id_list_limit_protein_domain_filters': <biomart.Filter name='id_list_limit_protein_domain_filters', type='id_list'>,\n",
       " 'germ_line_variation_source': <biomart.Filter name='germ_line_variation_source', type='list'>,\n",
       " 'somatic_variation_source': <biomart.Filter name='somatic_variation_source', type='list'>,\n",
       " 'with_validated_snp': <biomart.Filter name='with_validated_snp', type='boolean'>,\n",
       " 'so_consequence_name': <biomart.Filter name='so_consequence_name', type='list'>}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>NCBI gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>251176</td>\n",
       "      <td>ENST00000644207</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251177</td>\n",
       "      <td>ENST00000647544</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251178</td>\n",
       "      <td>ENST00000642596</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251179</td>\n",
       "      <td>ENST00000643537</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251180</td>\n",
       "      <td>ENST00000644633</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251181</td>\n",
       "      <td>ENST00000642800</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251182</td>\n",
       "      <td>ENST00000645112</td>\n",
       "      <td>56169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251183</td>\n",
       "      <td>ENST00000642712</td>\n",
       "      <td>56169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251184</td>\n",
       "      <td>ENST00000646090</td>\n",
       "      <td>56169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>251185</td>\n",
       "      <td>ENST00000643960</td>\n",
       "      <td>56169.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Transcript stable ID  NCBI gene ID\n",
       "251176      ENST00000644207           NaN\n",
       "251177      ENST00000647544           NaN\n",
       "251178      ENST00000642596           NaN\n",
       "251179      ENST00000643537           NaN\n",
       "251180      ENST00000644633           NaN\n",
       "251181      ENST00000642800           NaN\n",
       "251182      ENST00000645112       56169.0\n",
       "251183      ENST00000642712       56169.0\n",
       "251184      ENST00000646090       56169.0\n",
       "251185      ENST00000643960       56169.0"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl_entrez_gene_ids = dataset.query(attributes=['ensembl_transcript_id', 'entrezgene_id'])\n",
    "ensmbl_entrez_gene_ids.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{4535: 'ENST00000361390',\n",
       " 4536: 'ENST00000361453',\n",
       " 4512: 'ENST00000361624',\n",
       " 113219467: 'ENST00000387416',\n",
       " 4513: 'ENST00000361739',\n",
       " 4509: 'ENST00000361851',\n",
       " 4508: 'ENST00000361899',\n",
       " 4514: 'ENST00000362079',\n",
       " 4537: 'ENST00000361227',\n",
       " 4539: 'ENST00000361335',\n",
       " 4538: 'ENST00000361381',\n",
       " 4540: 'ENST00000361567',\n",
       " 4541: 'ENST00000361681',\n",
       " 4519: 'ENST00000361789',\n",
       " 1028: 'ENST00000471157',\n",
       " 51621: 'ENST00000616962',\n",
       " 255027: 'ENST00000567442',\n",
       " 9665: 'ENST00000549219',\n",
       " 4849: 'ENST00000611667',\n",
       " 102723475: 'ENST00000622690',\n",
       " 9093: 'ENST00000431375',\n",
       " 79165: 'ENST00000619669',\n",
       " 57348: 'ENST00000423529',\n",
       " 1113: 'ENST00000556876',\n",
       " 5165: 'ENST00000493226',\n",
       " 89927: 'ENST00000565913',\n",
       " 94104: 'ENST00000464256',\n",
       " 1007: 'ENST00000511822',\n",
       " 147798: 'ENST00000620520',\n",
       " 23532: 'ENST00000398741',\n",
       " 338755: 'ENST00000338569',\n",
       " 144125: 'ENST00000307401',\n",
       " 54664: 'ENST00000462754',\n",
       " 51277: 'ENST00000534855',\n",
       " 440243: 'ENST00000612056',\n",
       " 161725: 'ENST00000560598',\n",
       " 2558: 'ENST00000400081',\n",
       " 285148: 'ENST00000490621',\n",
       " 83787: 'ENST00000425331',\n",
       " 643401: 'ENST00000671061',\n",
       " 7726: 'ENST00000477569',\n",
       " 2533: 'ENST00000503065',\n",
       " 56245: 'ENST00000479548',\n",
       " 26531: 'ENST00000546665',\n",
       " 442194: 'ENST00000550017',\n",
       " 222236: 'ENST00000425379',\n",
       " 9962: 'ENST00000496331',\n",
       " 102723996: 'ENST00000620481',\n",
       " 81696: 'ENST00000230223',\n",
       " 26530: 'ENST00000509836',\n",
       " 81797: 'ENST00000550681',\n",
       " 2562: 'ENST00000622697',\n",
       " 26529: 'ENST00000377140',\n",
       " 8590: 'ENST00000332601',\n",
       " 6868: 'ENST00000648002',\n",
       " 51314: 'ENST00000476435',\n",
       " 7761: 'ENST00000536068',\n",
       " 102724023: 'ENST00000642815',\n",
       " 120775: 'ENST00000317834',\n",
       " 5002: 'ENST00000495518',\n",
       " 7762: 'ENST00000610573',\n",
       " 90527: 'ENST00000613425',\n",
       " 6424: 'ENST00000447200',\n",
       " 144124: 'ENST00000299454',\n",
       " 120776: 'ENST00000299459',\n",
       " 283297: 'ENST00000379829',\n",
       " 341276: 'ENST00000307322',\n",
       " 54749: 'ENST00000425345',\n",
       " 102724159: 'ENST00000634021',\n",
       " 5003: 'ENST00000455942',\n",
       " 8340: 'ENST00000377401',\n",
       " 51439: 'ENST00000259963',\n",
       " 8329: 'ENST00000358739',\n",
       " 8357: 'ENST00000369163',\n",
       " 353345: 'ENST00000334425',\n",
       " 8227: 'ENST00000381261',\n",
       " 3009: 'ENST00000331442',\n",
       " 55975: 'ENST00000469845',\n",
       " 8332: 'ENST00000613174',\n",
       " 221806: 'ENST00000614403',\n",
       " 9104: 'ENST00000457380',\n",
       " 8362: 'ENST00000611927',\n",
       " 8354: 'ENST00000616365',\n",
       " 3705: 'ENST00000553655',\n",
       " 8330: 'ENST00000618958',\n",
       " 8342: 'ENST00000621112',\n",
       " 56265: 'ENST00000380605',\n",
       " 57718: 'ENST00000556470',\n",
       " 8336: 'ENST00000359611',\n",
       " 8363: 'ENST00000355057',\n",
       " 8356: 'ENST00000479986',\n",
       " 8368: 'ENST00000618305',\n",
       " 8331: 'ENST00000333151',\n",
       " 8341: 'ENST00000396980',\n",
       " 90525: 'ENST00000561278',\n",
       " 223075: 'ENST00000451887',\n",
       " 105375283: 'ENST00000653190',\n",
       " 1022: 'ENST00000629350',\n",
       " 10669: 'ENST00000260595',\n",
       " 405753: 'ENST00000491993',\n",
       " 57407: 'ENST00000572559',\n",
       " 26716: 'ENST00000400689',\n",
       " 10398: 'ENST00000346786',\n",
       " 3795: 'ENST00000464371',\n",
       " 50506: 'ENST00000603300',\n",
       " 26015: 'ENST00000569829',\n",
       " 22839: 'ENST00000339266',\n",
       " 51156: 'ENST00000393096',\n",
       " 79143: 'ENST00000620311',\n",
       " 8100: 'ENST00000470695',\n",
       " 59307: 'ENST00000525070',\n",
       " 53905: 'ENST00000558744',\n",
       " 3163: 'ENST00000613539',\n",
       " 116511: 'ENST00000383549',\n",
       " 8468: 'ENST00000252037',\n",
       " 105372749: 'ENST00000662613',\n",
       " 105375009: 'ENST00000416126',\n",
       " 8348: 'ENST00000616182',\n",
       " 81697: 'ENST00000303324',\n",
       " 10842: 'ENST00000498609',\n",
       " 11215: 'ENST00000025301',\n",
       " 651746: 'ENST00000296657',\n",
       " 64112: 'ENST00000556883',\n",
       " 219771: 'ENST00000339497',\n",
       " 54765: 'ENST00000532066',\n",
       " 80831: 'ENST00000249044',\n",
       " 26212: 'ENST00000244623',\n",
       " 55644: 'ENST00000556252',\n",
       " 202658: 'ENST00000552136',\n",
       " 57477: 'ENST00000376020',\n",
       " 101930100: 'ENST00000624813',\n",
       " 56658: 'ENST00000552337',\n",
       " 438: 'ENST00000432523',\n",
       " 3460: 'ENST00000421802',\n",
       " 160287: 'ENST00000615355',\n",
       " 283080: 'ENST00000614533',\n",
       " 56243: 'ENST00000430453',\n",
       " 7932: 'ENST00000453513',\n",
       " 10537: 'ENST00000414436',\n",
       " 9246: 'ENST00000526659',\n",
       " 100288797: 'ENST00000380585',\n",
       " 390598: 'ENST00000341418',\n",
       " 128653: 'ENST00000380589',\n",
       " 79944: 'ENST00000556393',\n",
       " 2550: 'ENST00000546913',\n",
       " 3092: 'ENST00000616821',\n",
       " 7251: 'ENST00000535077',\n",
       " 100131244: 'ENST00000434396',\n",
       " 5330: 'ENST00000559634',\n",
       " 866: 'ENST00000557225',\n",
       " 757: 'ENST00000459909',\n",
       " 219770: 'ENST00000321660',\n",
       " 79897: 'ENST00000552136',\n",
       " 8325: 'ENST00000374694',\n",
       " 390502: 'ENST00000553483',\n",
       " 8326: 'ENST00000344575',\n",
       " 3812: 'ENST00000639252',\n",
       " 170954: 'ENST00000443517',\n",
       " 107987457: 'ENST00000443517',\n",
       " 7140: 'ENST00000673505',\n",
       " 167153: 'ENST00000453514',\n",
       " 11033: 'ENST00000617043',\n",
       " 200958: 'ENST00000625621',\n",
       " 80741: 'ENST00000494996',\n",
       " 55148: 'ENST00000555329',\n",
       " 3106: 'ENST00000640475',\n",
       " 3120: 'ENST00000417780',\n",
       " 51604: 'ENST00000639058',\n",
       " 29947: 'ENST00000431166',\n",
       " 5269: 'ENST00000644875',\n",
       " 140831: 'ENST00000255152',\n",
       " 84520: 'ENST00000556566',\n",
       " 8600: 'ENST00000239849',\n",
       " 26175: 'ENST00000415050',\n",
       " 9210: 'ENST00000252677',\n",
       " 140873: 'ENST00000444723',\n",
       " 387357: 'ENST00000368248',\n",
       " 2492: 'ENST00000419927',\n",
       " 140890: 'ENST00000612404',\n",
       " 56975: 'ENST00000673583',\n",
       " 90204: 'ENST00000372520',\n",
       " 23564: 'ENST00000483514',\n",
       " 79042: 'ENST00000616209',\n",
       " 255967: 'ENST00000483842',\n",
       " 7407: 'ENST00000455956',\n",
       " 100507170: 'ENST00000419982',\n",
       " 6227: 'ENST00000450116',\n",
       " 6725: 'ENST00000217188',\n",
       " 55388: 'ENST00000485659',\n",
       " 6048: 'ENST00000488204',\n",
       " 54962: 'ENST00000561773',\n",
       " 64795: 'ENST00000471113',\n",
       " 245938: 'ENST00000382410',\n",
       " 55611: 'ENST00000422031',\n",
       " 30818: 'ENST00000360990',\n",
       " 4585: 'ENST00000478685',\n",
       " 255313: 'ENST00000457020',\n",
       " 105375118: 'ENST00000670013',\n",
       " 84569: 'ENST00000375500',\n",
       " 2000: 'ENST00000615377',\n",
       " 7844: 'ENST00000472680',\n",
       " 81623: 'ENST00000542572',\n",
       " 728036: 'ENST00000430448',\n",
       " 6535: 'ENST00000457723',\n",
       " 153478: 'ENST00000283426',\n",
       " 3196: 'ENST00000621092',\n",
       " 2019: 'ENST00000546667',\n",
       " 90639: 'ENST00000466853',\n",
       " 140688: 'ENST00000616976',\n",
       " 9031: 'ENST00000466844',\n",
       " 896: 'ENST00000616010',\n",
       " 81850: 'ENST00000344363',\n",
       " 338323: 'ENST00000299481',\n",
       " 284428: 'ENST00000329753',\n",
       " 79644: 'ENST00000514398',\n",
       " 81851: 'ENST00000306271',\n",
       " 10382: 'ENST00000598635',\n",
       " 3133: 'ENST00000462399',\n",
       " 81872: 'ENST00000542137',\n",
       " 132001: 'ENST00000630288',\n",
       " 728042: 'ENST00000417256',\n",
       " 6538: 'ENST00000464828',\n",
       " 79591: 'ENST00000311122',\n",
       " 1237: 'ENST00000414803',\n",
       " 653656: 'ENST00000381394',\n",
       " 728279: 'ENST00000542910',\n",
       " 100526767: 'ENST00000440757',\n",
       " 143458: 'ENST00000529759',\n",
       " 114823: 'ENST00000616932',\n",
       " 730755: 'ENST00000391418',\n",
       " 125997: 'ENST00000381393',\n",
       " 51652: 'ENST00000485465',\n",
       " 401138: 'ENST00000504451',\n",
       " 122258: 'ENST00000631817',\n",
       " 140850: 'ENST00000382388',\n",
       " 54977: 'ENST00000642442',\n",
       " 442179: 'ENST00000402365',\n",
       " 102724889: 'ENST00000605945',\n",
       " 245939: 'ENST00000334391',\n",
       " 51617: 'ENST00000521146',\n",
       " 85461: 'ENST00000496406',\n",
       " 84993: 'ENST00000673596',\n",
       " 128497: 'ENST00000372519',\n",
       " 80345: 'ENST00000340487',\n",
       " 401251: 'ENST00000489292',\n",
       " 4987: 'ENST00000672146',\n",
       " 23769: 'ENST00000246841',\n",
       " 7718: 'ENST00000377325',\n",
       " 53335: 'ENST00000647038',\n",
       " 29928: 'ENST00000617832',\n",
       " 28992: 'ENST00000538595',\n",
       " 6203: 'ENST00000625297',\n",
       " 23423: 'ENST00000650754',\n",
       " 29: 'ENST00000628631',\n",
       " 54939: 'ENST00000567935',\n",
       " 222: 'ENST00000349015',\n",
       " 85294: 'ENST00000394015',\n",
       " 94059: 'ENST00000611161',\n",
       " 2683: 'ENST00000535206',\n",
       " 55667: 'ENST00000380424',\n",
       " 811: 'ENST00000586967',\n",
       " 10771: 'ENST00000381604',\n",
       " 139728: 'ENST00000484705',\n",
       " 23251: 'ENST00000559272',\n",
       " 10949: 'ENST00000314940',\n",
       " 8685: 'ENST00000494979',\n",
       " 729020: 'ENST00000441178',\n",
       " 57189: 'ENST00000482493',\n",
       " 8814: 'ENST00000534566',\n",
       " 3601: 'ENST00000397251',\n",
       " 5306: 'ENST00000539476',\n",
       " 1351: 'ENST00000314133',\n",
       " 9159: 'ENST00000540028',\n",
       " 9118: 'ENST00000369849',\n",
       " 5407: 'ENST00000534513',\n",
       " 55750: 'ENST00000629555',\n",
       " 10071: 'ENST00000536621',\n",
       " 5624: 'ENST00000402125',\n",
       " 55553: 'ENST00000396356',\n",
       " 2832: 'ENST00000369768',\n",
       " 10371: 'ENST00000490883',\n",
       " 11025: 'ENST00000445347',\n",
       " 102725035: 'ENST00000445347',\n",
       " 107987425: 'ENST00000445347',\n",
       " 107987462: 'ENST00000445347',\n",
       " 92521: 'ENST00000492188',\n",
       " 9499: 'ENST00000508938',\n",
       " 79803: 'ENST00000299238',\n",
       " 168090: 'ENST00000491176',\n",
       " 9542: 'ENST00000541337',\n",
       " 3107: 'ENST00000618255',\n",
       " 100233156: 'ENST00000559466',\n",
       " 64210: 'ENST00000355839',\n",
       " 27109: 'ENST00000672612',\n",
       " 100996750: 'ENST00000571574',\n",
       " 129881: 'ENST00000419478',\n",
       " 10752: 'ENST00000620033',\n",
       " 114799: 'ENST00000622333',\n",
       " 728224: 'ENST00000318329',\n",
       " 123606: 'ENST00000557930',\n",
       " 100132386: 'ENST00000617453',\n",
       " 653240: 'ENST00000391413',\n",
       " 83755: 'ENST00000394014',\n",
       " 3908: 'ENST00000617695',\n",
       " 112577516: 'ENST00000652090',\n",
       " 100533177: 'ENST00000391353',\n",
       " 2334: 'ENST00000286437',\n",
       " 399814: 'ENST00000329446',\n",
       " 1192: 'ENST00000614982',\n",
       " 157506: 'ENST00000521013',\n",
       " 51306: 'ENST00000510804',\n",
       " 27290: 'ENST00000379721',\n",
       " 202333: 'ENST00000505466',\n",
       " 116985: 'ENST00000455638',\n",
       " 105375501: 'ENST00000660838',\n",
       " 6129: 'ENST00000431653',\n",
       " 7402: 'ENST00000460618',\n",
       " 3111: 'ENST00000492557',\n",
       " 573: 'ENST00000673069',\n",
       " 11270: 'ENST00000610775',\n",
       " 79616: 'ENST00000541762',\n",
       " 83732: 'ENST00000493691',\n",
       " 51092: 'ENST00000431081',\n",
       " 51274: 'ENST00000482150',\n",
       " 5265: 'ENST00000355814',\n",
       " 10247: 'ENST00000519155',\n",
       " 8722: 'ENST00000526010',\n",
       " 126917: 'ENST00000416166',\n",
       " 55900: 'ENST00000627982',\n",
       " 57695: 'ENST00000338465',\n",
       " 5476: 'ENST00000606000',\n",
       " 112936: 'ENST00000531741',\n",
       " 80036: 'ENST00000360823',\n",
       " 9991: 'ENST00000334318',\n",
       " 91801: 'ENST00000530933',\n",
       " 26285: 'ENST00000286808',\n",
       " 6347: 'ENST00000582017',\n",
       " 23352: 'ENST00000497018',\n",
       " 7328: 'ENST00000490974',\n",
       " 9073: 'ENST00000399899',\n",
       " 10944: 'ENST00000524508',\n",
       " 23310: 'ENST00000532445',\n",
       " 339665: 'ENST00000451479',\n",
       " 23095: 'ENST00000622724',\n",
       " 643803: 'ENST00000340345',\n",
       " 150290: 'ENST00000442752',\n",
       " 100131902: 'ENST00000416044',\n",
       " 340252: 'ENST00000473601',\n",
       " 6840: 'ENST00000375398',\n",
       " 85477: 'ENST00000476649',\n",
       " 3113: 'ENST00000448154',\n",
       " 6354: 'ENST00000394630',\n",
       " 326: 'ENST00000337909',\n",
       " 25987: 'ENST00000612930',\n",
       " 110599588: 'ENST00000607690',\n",
       " 4439: 'ENST00000548003',\n",
       " 10940: 'ENST00000517435',\n",
       " 54529: 'ENST00000420250',\n",
       " 3115: 'ENST00000488076',\n",
       " 148170: 'ENST00000472703',\n",
       " 151525: 'ENST00000409124',\n",
       " 57583: 'ENST00000367090',\n",
       " 10075: 'ENST00000262854',\n",
       " 207063: 'ENST00000430536',\n",
       " 55735: 'ENST00000465911',\n",
       " 79168: 'ENST00000270464',\n",
       " 1776: 'ENST00000463694',\n",
       " 9437: 'ENST00000638974',\n",
       " 84893: 'ENST00000397269',\n",
       " 6632: 'ENST00000577906',\n",
       " 1643: 'ENST00000614884',\n",
       " 27039: 'ENST00000511176',\n",
       " 9189: 'ENST00000381223',\n",
       " 388818: 'ENST00000360542',\n",
       " 100528021: 'ENST00000494999',\n",
       " 10588: 'ENST00000560919',\n",
       " 2563: 'ENST00000640688',\n",
       " 29058: 'ENST00000379276',\n",
       " 221044: 'ENST00000463405',\n",
       " 1734: 'ENST00000422005',\n",
       " 338440: 'ENST00000525857',\n",
       " 8861: 'ENST00000425280',\n",
       " 64318: 'ENST00000461562',\n",
       " 79025: 'ENST00000370097',\n",
       " 25837: 'ENST00000566724',\n",
       " 64419: 'ENST00000606184',\n",
       " 2204: 'ENST00000640830',\n",
       " 6356: 'ENST00000305869',\n",
       " 30012: 'ENST00000296921',\n",
       " 5368: 'ENST00000522209',\n",
       " 57176: 'ENST00000550773',\n",
       " 79603: 'ENST00000599275',\n",
       " 728049: 'ENST00000457977',\n",
       " 6355: 'ENST00000394620',\n",
       " 100996455: 'ENST00000316124',\n",
       " 81793: 'ENST00000622002',\n",
       " 653282: 'ENST00000434883',\n",
       " 155051: 'ENST00000478106',\n",
       " 3904: 'ENST00000351841',\n",
       " 4798: 'ENST00000531318',\n",
       " 100507608: 'ENST00000391355',\n",
       " 85280: 'ENST00000334109',\n",
       " 81870: 'ENST00000394008',\n",
       " 728062: 'ENST00000673456',\n",
       " 83700: 'ENST00000533711',\n",
       " 151230: 'ENST00000448589',\n",
       " 100526832: 'ENST00000448589',\n",
       " 9081: 'ENST00000472391',\n",
       " 5228: 'ENST00000555234',\n",
       " 6876: 'ENST00000529622',\n",
       " 221527: 'ENST00000375527',\n",
       " 222950: 'ENST00000489641',\n",
       " 90326: 'ENST00000480647',\n",
       " 629: 'ENST00000486534',\n",
       " 255426: 'ENST00000615330',\n",
       " 4650: 'ENST00000397274',\n",
       " 389257: 'ENST00000328278',\n",
       " 7096: 'ENST00000308979',\n",
       " 493911: 'ENST00000617738',\n",
       " 140881: 'ENST00000246105',\n",
       " 7262: 'ENST00000314222',\n",
       " 1021: 'ENST00000491250',\n",
       " 4123: 'ENST00000618257',\n",
       " 123036: 'ENST00000555302',\n",
       " 9782: 'ENST00000502394',\n",
       " 9061: 'ENST00000504987',\n",
       " 219464: 'ENST00000313264',\n",
       " 390154: 'ENST00000303059',\n",
       " 400830: 'ENST00000382376',\n",
       " 83900: 'ENST00000411528',\n",
       " 83899: 'ENST00000377721',\n",
       " 7429: 'ENST00000419986',\n",
       " 10333: 'ENST00000610323',\n",
       " 653657: 'ENST00000333843',\n",
       " 55727: 'ENST00000554968',\n",
       " 5886: 'ENST00000591467',\n",
       " 92092: 'ENST00000275766',\n",
       " 196996: 'ENST00000569450',\n",
       " 6993: 'ENST00000367085',\n",
       " 55296: 'ENST00000513596',\n",
       " 255725: 'ENST00000530810',\n",
       " 4773: 'ENST00000496054',\n",
       " 81847: 'ENST00000616343',\n",
       " 196335: 'ENST00000316529',\n",
       " 4999: 'ENST00000457595',\n",
       " 120787: 'ENST00000311352',\n",
       " 5360: 'ENST00000420868',\n",
       " 81928: 'ENST00000453274',\n",
       " 94120: 'ENST00000611299',\n",
       " 100507632: 'ENST00000653444',\n",
       " 643812: 'ENST00000382835',\n",
       " 165257: 'ENST00000272520',\n",
       " 337963: 'ENST00000334160',\n",
       " 729597: 'ENST00000563237',\n",
       " 6742: 'ENST00000612337',\n",
       " 337959: 'ENST00000399889',\n",
       " 5179: 'ENST00000521153',\n",
       " 140258: 'ENST00000355459',\n",
       " 55312: 'ENST00000479197',\n",
       " 9575: 'ENST00000513033',\n",
       " 337960: 'ENST00000390690',\n",
       " 728072: 'ENST00000439466',\n",
       " 133957: 'ENST00000441693',\n",
       " 284827: 'ENST00000334068',\n",
       " 254950: 'ENST00000334067',\n",
       " 3206: 'ENST00000613671',\n",
       " 165545: 'ENST00000498552',\n",
       " 5154: 'ENST00000426681',\n",
       " 337882: 'ENST00000390689',\n",
       " 79820: 'ENST00000553676',\n",
       " 2556: 'ENST00000417858',\n",
       " 90480: 'ENST00000316939',\n",
       " 337969: 'ENST00000334055',\n",
       " 337970: 'ENST00000334063',\n",
       " 100505724: 'ENST00000391354',\n",
       " 10869: 'ENST00000465902',\n",
       " 54809: 'ENST00000620985',\n",
       " 337971: 'ENST00000334058',\n",
       " 258: 'ENST00000613447',\n",
       " 23244: 'ENST00000506262',\n",
       " 337972: 'ENST00000334151',\n",
       " 23016: 'ENST00000459856',\n",
       " 710: 'ENST00000619430',\n",
       " 1004: 'ENST00000504835',\n",
       " 4709: 'ENST00000454214',\n",
       " 717: 'ENST00000549972',\n",
       " 100289561: 'ENST00000492837',\n",
       " 7145: 'ENST00000615025',\n",
       " 390155: 'ENST00000641368',\n",
       " 65125: 'ENST00000530271',\n",
       " 56729: 'ENST00000629642',\n",
       " 728075: 'ENST00000672380',\n",
       " 219469: 'ENST00000610894',\n",
       " 79706: 'ENST00000468316',\n",
       " 105379487: 'ENST00000623227',\n",
       " 84071: 'ENST00000368972',\n",
       " 10942: 'ENST00000635038',\n",
       " 6357: 'ENST00000577681',\n",
       " 285172: 'ENST00000494881',\n",
       " 7174: 'ENST00000651823',\n",
       " 3205: 'ENST00000489695',\n",
       " 92689: 'ENST00000513966',\n",
       " 337973: 'ENST00000334046',\n",
       " 337974: 'ENST00000334849',\n",
       " 100288287: 'ENST00000382830',\n",
       " 50809: 'ENST00000443615',\n",
       " 3921: 'ENST00000495394',\n",
       " 337968: 'ENST00000391624',\n",
       " 337967: 'ENST00000334897',\n",
       " 337979: 'ENST00000334680',\n",
       " 494551: 'ENST00000493845',\n",
       " 337966: 'ENST00000329122',\n",
       " 337975: 'ENST00000334664',\n",
       " 5111: 'ENST00000379160',\n",
       " 51304: 'ENST00000478669',\n",
       " 5813: 'ENST00000331327',\n",
       " 345757: 'ENST00000505792',\n",
       " 55251: 'ENST00000299468',\n",
       " 56829: 'ENST00000471652',\n",
       " 54539: 'ENST00000276062',\n",
       " 57491: 'ENST00000512529',\n",
       " 80017: 'ENST00000298858',\n",
       " 140825: 'ENST00000372518',\n",
       " 219285: 'ENST00000610760',\n",
       " 177: 'ENST00000551827',\n",
       " 23287: 'ENST00000337006',\n",
       " 144809: 'ENST00000537894',\n",
       " 160298: 'ENST00000316375',\n",
       " 8241: 'ENST00000329236',\n",
       " 5077: 'ENST00000258387',\n",
       " 55531: 'ENST00000534236',\n",
       " 9319: 'ENST00000508430',\n",
       " 101927378: 'ENST00000446000',\n",
       " 9083: 'ENST00000618574',\n",
       " 442867: 'ENST00000618574',\n",
       " 442868: 'ENST00000618574',\n",
       " 10124: 'ENST00000396663',\n",
       " 27288: 'ENST00000306904',\n",
       " 22907: 'ENST00000619982',\n",
       " 10344: 'ENST00000394905',\n",
       " 4661: 'ENST00000360149',\n",
       " 151835: 'ENST00000613455',\n",
       " 55858: 'ENST00000515591',\n",
       " 100151643: 'ENST00000382828',\n",
       " 22827: 'ENST00000533362',\n",
       " 7301: 'ENST00000568490',\n",
       " 337976: 'ENST00000330798',\n",
       " 79631: 'ENST00000558974',\n",
       " 337985: 'ENST00000382826',\n",
       " 100288323: 'ENST00000444335',\n",
       " 337977: 'ENST00000335093',\n",
       " 337978: 'ENST00000333892',\n",
       " 10016: 'ENST00000628729',\n",
       " 116535: 'ENST00000441623',\n",
       " 7123: 'ENST00000428034',\n",
       " 5634: 'ENST00000461630',\n",
       " 6194: 'ENST00000380381',\n",
       " 1139: 'ENST00000645153',\n",
       " 102724631: 'ENST00000612601',\n",
       " 79815: 'ENST00000519324',\n",
       " 199675: 'ENST00000598851',\n",
       " 5926: 'ENST00000348476',\n",
       " 79230: 'ENST00000414706',\n",
       " 619279: 'ENST00000519936',\n",
       " 389610: 'ENST00000647027',\n",
       " 6892: 'ENST00000456592',\n",
       " 401: 'ENST00000544057',\n",
       " 6138: 'ENST00000611050',\n",
       " 348654: 'ENST00000381254',\n",
       " 126003: 'ENST00000595985',\n",
       " 1672: 'ENST00000642865',\n",
       " 3636: 'ENST00000541756',\n",
       " 1823: 'ENST00000257198',\n",
       " 2968: 'ENST00000400450',\n",
       " 55203: 'ENST00000512108',\n",
       " 55824: 'ENST00000519326',\n",
       " 124935: 'ENST00000613070',\n",
       " 4122: 'ENST00000360468',\n",
       " 2173: 'ENST00000356535',\n",
       " 6588: 'ENST00000305991',\n",
       " 22906: 'ENST00000613405',\n",
       " 115701: 'ENST00000590642',\n",
       " 54790: 'ENST00000540549',\n",
       " 54476: 'ENST00000476345',\n",
       " 513: 'ENST00000614837',\n",
       " 149775: 'ENST00000443966',\n",
       " 151278: 'ENST00000295226',\n",
       " 344752: 'ENST00000356517',\n",
       " 54784: 'ENST00000498283',\n",
       " 81490: 'ENST00000632264',\n",
       " 2517: 'ENST00000374479',\n",
       " 6346: 'ENST00000225842',\n",
       " 6389: 'ENST00000617470',\n",
       " 1671: 'ENST00000646187',\n",
       " 55620: 'ENST00000601482',\n",
       " 29062: 'ENST00000485942',\n",
       " 54510: 'ENST00000617302',\n",
       " 1824: 'ENST00000648081',\n",
       " 80228: 'ENST00000356387',\n",
       " 55655: 'ENST00000616903',\n",
       " 2778: 'ENST00000306090',\n",
       " 5834: 'ENST00000471359',\n",
       " 79839: 'ENST00000581103',\n",
       " 6850: 'ENST00000375746',\n",
       " 129049: 'ENST00000610372',\n",
       " 114898: 'ENST00000393975',\n",
       " 728082: 'ENST00000441330',\n",
       " 63920: 'ENST00000523213',\n",
       " 728090: 'ENST00000458218',\n",
       " 285761: 'ENST00000487076',\n",
       " 728096: 'ENST00000672229',\n",
       " 11027: 'ENST00000616629',\n",
       " 6598: 'ENST00000626269',\n",
       " 79677: 'ENST00000489535',\n",
       " 51530: 'ENST00000311873',\n",
       " 140893: 'ENST00000252998',\n",
       " 55879: 'ENST00000598523',\n",
       " 157777: 'ENST00000415737',\n",
       " 6654: 'ENST00000402219',\n",
       " 9975: 'ENST00000472780',\n",
       " 57406: 'ENST00000480457',\n",
       " 1611: 'ENST00000508646',\n",
       " 80008: 'ENST00000495029',\n",
       " 2747: 'ENST00000673493',\n",
       " 11046: 'ENST00000643789',\n",
       " 8892: 'ENST00000556668',\n",
       " 26090: 'ENST00000461204',\n",
       " 4488: 'ENST00000507785',\n",
       " 26154: 'ENST00000412081',\n",
       " 57120: 'ENST00000535237',\n",
       " 7920: 'ENST00000548592',\n",
       " 2208: 'ENST00000360067',\n",
       " 54872: 'ENST00000505800',\n",
       " 23082: 'ENST00000413464',\n",
       " 79290: 'ENST00000612185',\n",
       " 202243: 'ENST00000396499',\n",
       " 7033: 'ENST00000291525',\n",
       " 29099: 'ENST00000526789',\n",
       " 25942: 'ENST00000565264',\n",
       " 115273: 'ENST00000465518',\n",
       " 10924: 'ENST00000539041',\n",
       " 6883: 'ENST00000495422',\n",
       " 3605: 'ENST00000648244',\n",
       " 28512: 'ENST00000614374',\n",
       " 219473: 'ENST00000623845',\n",
       " 8449: 'ENST00000446074',\n",
       " 122553: 'ENST00000556765',\n",
       " 51015: 'ENST00000173527',\n",
       " 3293: 'ENST00000650386',\n",
       " 105375851: 'ENST00000661064',\n",
       " 6369: 'ENST00000416943',\n",
       " 90342: 'ENST00000641731',\n",
       " 3005: 'ENST00000340857',\n",
       " 6098: 'ENST00000403284',\n",
       " 64374: 'ENST00000508744',\n",
       " 1244: 'ENST00000648523',\n",
       " 8225: 'ENST00000326153',\n",
       " 219: 'ENST00000635162',\n",
       " 55607: 'ENST00000456331',\n",
       " 729288: 'ENST00000668878',\n",
       " 51175: 'ENST00000603722',\n",
       " 51284: 'ENST00000484204',\n",
       " 157574: 'ENST00000519471',\n",
       " 254272: 'ENST00000405044',\n",
       " 347252: 'ENST00000377694',\n",
       " 100507303: 'ENST00000563192',\n",
       " 6854: 'ENST00000425297',\n",
       " 240: 'ENST00000622238',\n",
       " 51311: 'ENST00000311912',\n",
       " 3112: 'ENST00000413252',\n",
       " 55344: 'ENST00000381663',\n",
       " 221143: 'ENST00000460374',\n",
       " 158405: 'ENST00000536272',\n",
       " 4869: 'ENST00000524204',\n",
       " 10809: 'ENST00000540587',\n",
       " 81614: 'ENST00000398014',\n",
       " 91319: 'ENST00000626238',\n",
       " 158471: 'ENST00000223609',\n",
       " 2310: 'ENST00000454745',\n",
       " 53342: 'ENST00000468605',\n",
       " 84263: 'ENST00000480881',\n",
       " 101926971: 'ENST00000665808',\n",
       " 54928: 'ENST00000517461',\n",
       " 8938: 'ENST00000628027',\n",
       " 1269: 'ENST00000374472',\n",
       " 199713: 'ENST00000615699',\n",
       " 124944: 'ENST00000547709',\n",
       " 7697: 'ENST00000359735',\n",
       " 55629: 'ENST00000579103',\n",
       " 340602: 'ENST00000342995',\n",
       " 150160: 'ENST00000359963',\n",
       " 170685: 'ENST00000356450',\n",
       " 55893: 'ENST00000519730',\n",
       " 8838: 'ENST00000409166',\n",
       " 79623: 'ENST00000485468',\n",
       " 9938: 'ENST00000479844',\n",
       " 221545: 'ENST00000468808',\n",
       " 84902: 'ENST00000592401',\n",
       " 203611: 'ENST00000544303',\n",
       " 51011: 'ENST00000470100',\n",
       " 23191: 'ENST00000627816',\n",
       " 107983955: 'ENST00000620688',\n",
       " 134829: 'ENST00000368438',\n",
       " 9063: 'ENST00000590944',\n",
       " 51427: 'ENST00000423627',\n",
       " 23657: 'ENST00000509248',\n",
       " 203074: 'ENST00000646980',\n",
       " 7079: 'ENST00000287814',\n",
       " 79661: 'ENST00000569758',\n",
       " 199699: 'ENST00000317060',\n",
       " 390157: 'ENST00000279783',\n",
       " 8744: 'ENST00000245817',\n",
       " 80179: 'ENST00000634006',\n",
       " 55190: 'ENST00000375992',\n",
       " 140680: 'ENST00000400269',\n",
       " 7032: 'ENST00000475297',\n",
       " 100507436: 'ENST00000546529',\n",
       " 1669: 'ENST00000642522',\n",
       " 440400: 'ENST00000552321',\n",
       " 2650: 'ENST00000442371',\n",
       " 55683: 'ENST00000478492',\n",
       " 10117: 'ENST00000472597',\n",
       " 3643: 'ENST00000600492',\n",
       " 94240: 'ENST00000536042',\n",
       " 5411: 'ENST00000557680',\n",
       " 4771: 'ENST00000413209',\n",
       " 11177: 'ENST00000360310',\n",
       " 149951: 'ENST00000645058',\n",
       " 27430: 'ENST00000521838',\n",
       " 140453: 'ENST00000497859',\n",
       " 339318: 'ENST00000492450',\n",
       " 4340: 'ENST00000551993',\n",
       " 7127: 'ENST00000333007',\n",
       " 3512: 'ENST00000543780',\n",
       " 2171: 'ENST00000481695',\n",
       " 8495: 'ENST00000534552',\n",
       " 102724219: 'ENST00000611026',\n",
       " 619208: 'ENST00000604268',\n",
       " 162699: 'ENST00000620881',\n",
       " 51224: 'ENST00000620522',\n",
       " 1828: 'ENST00000462981',\n",
       " 284805: 'ENST00000360785',\n",
       " 91404: 'ENST00000452991',\n",
       " 728929: 'ENST00000611323',\n",
       " 7031: 'ENST00000291527',\n",
       " 100506888: 'ENST00000330682',\n",
       " 10569: 'ENST00000520664',\n",
       " 83473: 'ENST00000591522',\n",
       " 9618: 'ENST00000618771',\n",
       " 54943: 'ENST00000617313',\n",
       " 105375622: 'ENST00000650562',\n",
       " 159963: 'ENST00000528822',\n",
       " 10630: 'ENST00000617617',\n",
       " 81844: 'ENST00000487252',\n",
       " 79148: 'ENST00000612672',\n",
       " 219477: 'ENST00000303039',\n",
       " 345778: 'ENST00000617335',\n",
       " 147409: 'ENST00000359747',\n",
       " 504189: 'ENST00000627176',\n",
       " 85364: 'ENST00000500893',\n",
       " 504190: 'ENST00000626074',\n",
       " 1667: 'ENST00000644528',\n",
       " 440934: 'ENST00000439237',\n",
       " 80023: 'ENST00000621012',\n",
       " 5725: 'ENST00000627714',\n",
       " 970: 'ENST00000597430',\n",
       " 57819: 'ENST00000478692',\n",
       " 23066: 'ENST00000626378',\n",
       " 79722: 'ENST00000519114',\n",
       " 7430: 'ENST00000392177',\n",
       " 387758: 'ENST00000318627',\n",
       " 23514: 'ENST00000541342',\n",
       " 51091: 'ENST00000513285',\n",
       " 92799: 'ENST00000602239',\n",
       " 441308: 'ENST00000320901',\n",
       " 7253: 'ENST00000298171',\n",
       " 55240: 'ENST00000393107',\n",
       " 54905: 'ENST00000415893',\n",
       " 2321: 'ENST00000543394',\n",
       " 8424: 'ENST00000263182',\n",
       " 9468: 'ENST00000496020',\n",
       " 57414: 'ENST00000318622',\n",
       " 239: 'ENST00000406228',\n",
       " 169270: 'ENST00000320552',\n",
       " 3572: 'ENST00000396816',\n",
       " 346673: 'ENST00000275764',\n",
       " 401027: 'ENST00000673386',\n",
       " 27346: 'ENST00000336687',\n",
       " 1832: 'ENST00000506617',\n",
       " 5375: 'ENST00000519260',\n",
       " 196792: 'ENST00000489000',\n",
       " 55328: 'ENST00000437752',\n",
       " 130074: 'ENST00000389915',\n",
       " 101928359: 'ENST00000665977',\n",
       " 6666: 'ENST00000342665',\n",
       " 170949: 'ENST00000453425',\n",
       " 100287083: 'ENST00000453425',\n",
       " 90410: 'ENST00000588477',\n",
       " 54733: 'ENST00000524991',\n",
       " 728358: 'ENST00000644617',\n",
       " 646486: 'ENST00000360464',\n",
       " 5408: 'ENST00000591655',\n",
       " 102724334: 'ENST00000464664',\n",
       " 50624: 'ENST00000338948',\n",
       " 646480: 'ENST00000379071',\n",
       " 64328: 'ENST00000400602',\n",
       " 91584: 'ENST00000321063',\n",
       " 7114: 'ENST00000380633',\n",
       " 818: 'ENST00000423381',\n",
       " 6161: 'ENST00000396957',\n",
       " 123720: 'ENST00000560964',\n",
       " 2167: 'ENST00000521734',\n",
       " 107986954: 'ENST00000656790',\n",
       " 51272: 'ENST00000479463',\n",
       " 9252: 'ENST00000556594',\n",
       " 4967: 'ENST00000631326',\n",
       " 255349: 'ENST00000423535',\n",
       " 64866: 'ENST00000490471',\n",
       " 203076: 'ENST00000643037',\n",
       " 142910: 'ENST00000531458',\n",
       " 6891: 'ENST00000468636',\n",
       " 85481: 'ENST00000523010',\n",
       " 9972: 'ENST00000537253',\n",
       " 55231: 'ENST00000333861',\n",
       " 4336: 'ENST00000479860',\n",
       " 94137: 'ENST00000329335',\n",
       " 245972: 'ENST00000285393',\n",
       " 10892: 'ENST00000587438',\n",
       " 117583: 'ENST00000622699',\n",
       " 1668: 'ENST00000642311',\n",
       " 147166: 'ENST00000395671',\n",
       " 100463482: 'ENST00000633925',\n",
       " 6741: 'ENST00000494051',\n",
       " 113746: 'ENST00000342593',\n",
       " 166929: 'ENST00000504754',\n",
       " 8740: 'ENST00000599359',\n",
       " 441239: 'ENST00000340779',\n",
       " 145508: 'ENST00000281129',\n",
       " 10793: 'ENST00000395375',\n",
       " 1794: 'ENST00000540750',\n",
       " 80742: 'ENST00000475643',\n",
       " 541565: 'ENST00000615223',\n",
       " 64711: 'ENST00000293937',\n",
       " 337879: 'ENST00000329621',\n",
       " 1670: 'ENST00000647159',\n",
       " 337878: 'ENST00000621162',\n",
       " 64116: 'ENST00000502903',\n",
       " 147199: 'ENST00000342878',\n",
       " 337880: 'ENST00000332378',\n",
       " 91057: 'ENST00000317945',\n",
       " 29081: 'ENST00000471560',\n",
       " 728299: 'ENST00000382822',\n",
       " 113612: 'ENST00000508453',\n",
       " 102157402: 'ENST00000618980',\n",
       " 6150: 'ENST00000673399',\n",
       " 107987373: 'ENST00000673399',\n",
       " 9786: 'ENST00000619722',\n",
       " 5159: 'ENST00000523456',\n",
       " 107984345: 'ENST00000641280',\n",
       " 2794: 'ENST00000478986',\n",
       " 390174: 'ENST00000312153',\n",
       " 107986298: 'ENST00000668384',\n",
       " 219931: 'ENST00000442692',\n",
       " 504191: 'ENST00000571130',\n",
       " 3268: 'ENST00000429987',\n",
       " 51351: 'ENST00000610793',\n",
       " 109504726: 'ENST00000610793',\n",
       " 9326: 'ENST00000634200',\n",
       " 148103: 'ENST00000588788',\n",
       " 116449: 'ENST00000442825',\n",
       " 283189: 'ENST00000302957',\n",
       " 2086: 'ENST00000535538',\n",
       " 6880: 'ENST00000328663',\n",
       " 2618: 'ENST00000441403',\n",
       " 64236: 'ENST00000622702',\n",
       " 150165: 'ENST00000331428',\n",
       " 55677: 'ENST00000436740',\n",
       " 101927322: 'ENST00000588144',\n",
       " 85379: 'ENST00000358431',\n",
       " 3207: 'ENST00000006015',\n",
       " 440270: 'ENST00000438958',\n",
       " 124842: 'ENST00000577271',\n",
       " 7384: 'ENST00000493806',\n",
       " 353149: 'ENST00000612583',\n",
       " 25907: 'ENST00000503771',\n",
       " 5884: 'ENST00000616683',\n",
       " 57584: 'ENST00000320481',\n",
       " 100505573: 'ENST00000560415',\n",
       " 107986888: 'ENST00000653093',\n",
       " 9131: 'ENST00000535724',\n",
       " 102724398: 'ENST00000624872',\n",
       " 154007: 'ENST00000496946',\n",
       " 84067: 'ENST00000532797',\n",
       " 115653: 'ENST00000638267',\n",
       " 344387: 'ENST00000419111',\n",
       " 4102: 'ENST00000598245',\n",
       " 256364: 'ENST00000419857',\n",
       " 284098: 'ENST00000616581',\n",
       " 79977: 'ENST00000474338',\n",
       " 492311: 'ENST00000333305',\n",
       " 128954: 'ENST00000523144',\n",
       " 8760: 'ENST00000379070',\n",
       " 3788: 'ENST00000521839',\n",
       " 57050: 'ENST00000254803',\n",
       " 102723547: 'ENST00000535353',\n",
       " 9439: 'ENST00000539158',\n",
       " 3804: 'ENST00000640389',\n",
       " 10666: 'ENST00000580335',\n",
       " 143425: 'ENST00000532592',\n",
       " 57524: 'ENST00000562055',\n",
       " 6788: 'ENST00000617590',\n",
       " 11102: 'ENST00000462046',\n",
       " 90362: 'ENST00000521413',\n",
       " 197131: 'ENST00000627960',\n",
       " 283298: 'ENST00000528758',\n",
       " 81622: 'ENST00000531152',\n",
       " 1617: 'ENST00000426000',\n",
       " 84174: 'ENST00000360672',\n",
       " 22902: 'ENST00000513593',\n",
       " 91442: 'ENST00000591791',\n",
       " 9275: 'ENST00000481667',\n",
       " 55300: 'ENST00000264864',\n",
       " 109703458: 'ENST00000476007',\n",
       " 84231: 'ENST00000570169',\n",
       " 10331: 'ENST00000595387',\n",
       " 60436: 'ENST00000611732',\n",
       " 220164: 'ENST00000577609',\n",
       " 3150: 'ENST00000463631',\n",
       " 84196: 'ENST00000400301',\n",
       " 150051: 'ENST00000433071',\n",
       " 79133: 'ENST00000486772',\n",
       " 7074: 'ENST00000541036',\n",
       " 57335: 'ENST00000593105',\n",
       " 23365: 'ENST00000528681',\n",
       " 27000: 'ENST00000412522',\n",
       " 7179: 'ENST00000612746',\n",
       " 1830: 'ENST00000257189',\n",
       " 257062: 'ENST00000448307',\n",
       " 101927120: 'ENST00000526436',\n",
       " 1946: 'ENST00000611503',\n",
       " 102724428: 'ENST00000624077',\n",
       " 171586: 'ENST00000579982',\n",
       " 1387: 'ENST00000636895',\n",
       " 100131897: 'ENST00000523970',\n",
       " 3811: 'ENST00000640111',\n",
       " 85460: 'ENST00000503068',\n",
       " 653486: 'ENST00000613221',\n",
       " 55969: 'ENST00000639924',\n",
       " 5089: 'ENST00000466640',\n",
       " 54984: 'ENST00000647329',\n",
       " 51614: 'ENST00000640748',\n",
       " 23008: 'ENST00000495724',\n",
       " 8447: 'ENST00000633635',\n",
       " 55284: 'ENST00000650817',\n",
       " 54919: 'ENST00000461576',\n",
       " 157724: 'ENST00000520624',\n",
       " 10910: 'ENST00000609175',\n",
       " 54927: 'ENST00000457942',\n",
       " 163154: 'ENST00000419421',\n",
       " 9656: 'ENST00000550500',\n",
       " 64172: 'ENST00000521630',\n",
       " 23762: 'ENST00000403222',\n",
       " 1302: 'ENST00000486982',\n",
       " 83595: 'ENST00000646538',\n",
       " 10497: 'ENST00000617908',\n",
       " 283726: 'ENST00000565006',\n",
       " 56931: 'ENST00000589854',\n",
       " 23258: 'ENST00000530780',\n",
       " 100527943: 'ENST00000558530',\n",
       " 132884: 'ENST00000509670',\n",
       " 5575: 'ENST00000544935',\n",
       " 347549: 'ENST00000417339',\n",
       " 29063: 'ENST00000612982',\n",
       " 150709: 'ENST00000313581',\n",
       " 115548: 'ENST00000508431',\n",
       " 6094: 'ENST00000525947',\n",
       " 375611: 'ENST00000354356',\n",
       " 115939: 'ENST00000566296',\n",
       " 654: 'ENST00000283147',\n",
       " 105375624: 'ENST00000668794',\n",
       " 22941: 'ENST00000449833',\n",
       " 51115: 'ENST00000518772',\n",
       " 3802: 'ENST00000639974',\n",
       " 100129654: 'ENST00000566129',\n",
       " 63915: 'ENST00000543936',\n",
       " 7060: 'ENST00000511888',\n",
       " 253650: 'ENST00000475234',\n",
       " 116969: 'ENST00000425767',\n",
       " 81567: 'ENST00000469459',\n",
       " 3809: 'ENST00000640257',\n",
       " 118670: 'ENST00000368894',\n",
       " 4034: 'ENST00000619071',\n",
       " 5201: 'ENST00000507185',\n",
       " 3805: 'ENST00000638428',\n",
       " 112268354: 'ENST00000638428',\n",
       " 3613: 'ENST00000625802',\n",
       " 4784: 'ENST00000358552',\n",
       " 259217: 'ENST00000434227',\n",
       " 79664: 'ENST00000560520',\n",
       " 27067: 'ENST00000523160',\n",
       " ...}"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl_entrez_gene_ids.dropna(inplace=True)\n",
    "ensmbl_entrez_gene_ids['NCBI gene ID'] = ensmbl_entrez_gene_ids['NCBI gene ID'].astype(int)\n",
    "ensmbl_entrez_gene_ids = ensmbl_entrez_gene_ids.set_index('NCBI gene ID').to_dict()['Transcript stable ID']\n",
    "ensmbl_entrez_gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensmbl_ids = pd.Series(lihc.columns.values[1:]).apply(lambda x: x.split('|')[1]).astype(int).map(ensmbl_entrez_gene_ids).dropna()\n",
    "\n",
    "# Had to drop ~2k that didn't align, is there a better way?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'link_so_mini_closure': <biomart.Filter name='link_so_mini_closure', type='list'>,\n",
       " 'link_go_closure': <biomart.Filter name='link_go_closure', type='text'>,\n",
       " 'link_ensembl_transcript_stable_id': <biomart.Filter name='link_ensembl_transcript_stable_id', type='text'>,\n",
       " 'gene_id': <biomart.Filter name='gene_id', type='text'>,\n",
       " 'transcript_id': <biomart.Filter name='transcript_id', type='text'>,\n",
       " 'link_ensembl_gene_id': <biomart.Filter name='link_ensembl_gene_id', type='text'>,\n",
       " 'chromosome_name': <biomart.Filter name='chromosome_name', type='text'>,\n",
       " 'start': <biomart.Filter name='start', type='text'>,\n",
       " 'end': <biomart.Filter name='end', type='text'>,\n",
       " 'band_start': <biomart.Filter name='band_start', type='drop_down_basic_filter'>,\n",
       " 'band_end': <biomart.Filter name='band_end', type='drop_down_basic_filter'>,\n",
       " 'marker_start': <biomart.Filter name='marker_start', type='drop_down_basic_filter'>,\n",
       " 'marker_end': <biomart.Filter name='marker_end', type=''>,\n",
       " 'hsapiens_encode.type': <biomart.Filter name='hsapiens_encode.type', type='text'>,\n",
       " 'hsapiens_encode.encode_region': <biomart.Filter name='hsapiens_encode.encode_region', type=''>,\n",
       " 'strand': <biomart.Filter name='strand', type='text'>,\n",
       " 'chromosomal_region': <biomart.Filter name='chromosomal_region', type='text'>,\n",
       " 'id_list_xrefs_filters': <biomart.Filter name='id_list_xrefs_filters', type='boolean_list'>,\n",
       " 'id_list_limit_xrefs_filters': <biomart.Filter name='id_list_limit_xrefs_filters', type='id_list'>,\n",
       " 'id_list_microarray_filters': <biomart.Filter name='id_list_microarray_filters', type='boolean_list'>,\n",
       " 'id_list_limit_microarray_filters': <biomart.Filter name='id_list_limit_microarray_filters', type='id_list'>,\n",
       " 'transcript_count_greater_than': <biomart.Filter name='transcript_count_greater_than', type='text'>,\n",
       " 'transcript_count_less_than': <biomart.Filter name='transcript_count_less_than', type='text'>,\n",
       " 'biotype': <biomart.Filter name='biotype', type='list'>,\n",
       " 'transcript_biotype': <biomart.Filter name='transcript_biotype', type='list'>,\n",
       " 'source': <biomart.Filter name='source', type='list'>,\n",
       " 'transcript_source': <biomart.Filter name='transcript_source', type='list'>,\n",
       " 'transcript_tsl': <biomart.Filter name='transcript_tsl', type='boolean'>,\n",
       " 'transcript_gencode_basic': <biomart.Filter name='transcript_gencode_basic', type='boolean'>,\n",
       " 'transcript_appris': <biomart.Filter name='transcript_appris', type='boolean'>,\n",
       " 'mane_select': <biomart.Filter name='mane_select', type='boolean'>,\n",
       " 'phenotype_description': <biomart.Filter name='phenotype_description', type='list'>,\n",
       " 'phenotype_source': <biomart.Filter name='phenotype_source', type='list'>,\n",
       " 'go_clos': <biomart.Filter name='go_clos', type=''>,\n",
       " 'go_name': <biomart.Filter name='go_name', type=''>,\n",
       " 'go_evidence_code': <biomart.Filter name='go_evidence_code', type='list'>,\n",
       " 'homolog_filters': <biomart.Filter name='homolog_filters', type='boolean_list'>,\n",
       " 'id_list_protein_domain_and_feature_filters': <biomart.Filter name='id_list_protein_domain_and_feature_filters', type='boolean_list'>,\n",
       " 'id_list_limit_protein_domain_filters': <biomart.Filter name='id_list_limit_protein_domain_filters', type='id_list'>,\n",
       " 'germ_line_variation_source': <biomart.Filter name='germ_line_variation_source', type='list'>,\n",
       " 'somatic_variation_source': <biomart.Filter name='somatic_variation_source', type='list'>,\n",
       " 'with_validated_snp': <biomart.Filter name='with_validated_snp', type='boolean'>,\n",
       " 'so_consequence_name': <biomart.Filter name='so_consequence_name', type='list'>}"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find a faster way to do this\n",
    "\n",
    "attributes = [\n",
    "#     'gene_id',\n",
    "    'entrezgene_id',\n",
    "    'ensembl_gene_id',\n",
    "    'ensembl_transcript_id',\n",
    "    'go_id',\n",
    "    'name_1006',\n",
    "    'definition_1006',\n",
    "    'go_linkage_type',\n",
    "    'hgnc_id',\n",
    "    'hgnc_symbol',\n",
    "#     'hgnc_trans_name',\n",
    "]\n",
    "go_mappings = dataset.query(attributes=attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "go_mappings.to_csv('data/go_mappings.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI gene ID</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term definition</th>\n",
       "      <th>GO term evidence code</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [NCBI gene ID, Gene stable ID, Transcript stable ID, GO term accession, GO term name, GO term definition, GO term evidence code, HGNC ID, HGNC symbol]\n",
       "Index: []"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Figure out better way to do this\n",
    "attributes = [\n",
    "#     'gene_id',\n",
    "    'entrezgene_id',\n",
    "    'ensembl_gene_id',\n",
    "    'ensembl_transcript_id',\n",
    "    'go_id',\n",
    "    'name_1006',\n",
    "    'definition_1006',\n",
    "    'go_linkage_type',\n",
    "    'hgnc_id',\n",
    "    'hgnc_symbol',\n",
    "#     'hgnc_trans_name',\n",
    "]\n",
    "dataset.query(attributes=attributes,\n",
    "              filters={'transcript_id': ensmbl_ids.values}\n",
    "             )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examine Gene Counts\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mattheweng/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1392800, 9)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_mappings = pd.read_csv('data/go_mappings.csv')\n",
    "\n",
    "go_mappings.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI gene ID</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term definition</th>\n",
       "      <th>GO term evidence code</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7481</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7470</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7500</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7471</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7490</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  NCBI gene ID   Gene stable ID Transcript stable ID GO term accession  \\\n",
       "0          NaN  ENSG00000210049      ENST00000387314               NaN   \n",
       "1          NaN  ENSG00000211459      ENST00000389680               NaN   \n",
       "2          NaN  ENSG00000210077      ENST00000387342               NaN   \n",
       "3          NaN  ENSG00000210082      ENST00000387347               NaN   \n",
       "4          NaN  ENSG00000209082      ENST00000386347               NaN   \n",
       "\n",
       "  GO term name GO term definition GO term evidence code    HGNC ID HGNC symbol  \n",
       "0          NaN                NaN                   NaN  HGNC:7481       MT-TF  \n",
       "1          NaN                NaN                   NaN  HGNC:7470     MT-RNR1  \n",
       "2          NaN                NaN                   NaN  HGNC:7500       MT-TV  \n",
       "3          NaN                NaN                   NaN  HGNC:7471     MT-RNR2  \n",
       "4          NaN                NaN                   NaN  HGNC:7490      MT-TL1  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_mappings.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI gene ID</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term definition</th>\n",
       "      <th>GO term evidence code</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>4535</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>A lipid bilayer along with all the proteins an...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>HGNC:7455</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>4535</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>GO:0016021</td>\n",
       "      <td>integral component of membrane</td>\n",
       "      <td>The component of a membrane consisting of the ...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>HGNC:7455</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>4535</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>GO:0055114</td>\n",
       "      <td>oxidation-reduction process</td>\n",
       "      <td>A metabolic process that results in the remova...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>HGNC:7455</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>4535</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>GO:0005743</td>\n",
       "      <td>mitochondrial inner membrane</td>\n",
       "      <td>The inner, i.e. lumen-facing, lipid bilayer of...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>HGNC:7455</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>4535</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>A semiautonomous, self replicating organelle t...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>HGNC:7455</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1392787</td>\n",
       "      <td>60678</td>\n",
       "      <td>ENSG00000284869</td>\n",
       "      <td>ENST00000646013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24614</td>\n",
       "      <td>EEFSEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1392796</td>\n",
       "      <td>56169</td>\n",
       "      <td>ENSG00000285114</td>\n",
       "      <td>ENST00000645112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7151</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1392797</td>\n",
       "      <td>56169</td>\n",
       "      <td>ENSG00000285114</td>\n",
       "      <td>ENST00000642712</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7151</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1392798</td>\n",
       "      <td>56169</td>\n",
       "      <td>ENSG00000285114</td>\n",
       "      <td>ENST00000646090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7151</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1392799</td>\n",
       "      <td>56169</td>\n",
       "      <td>ENSG00000285114</td>\n",
       "      <td>ENST00000643960</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7151</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1306714 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        NCBI gene ID   Gene stable ID Transcript stable ID GO term accession  \\\n",
       "5               4535  ENSG00000198888      ENST00000361390        GO:0016020   \n",
       "6               4535  ENSG00000198888      ENST00000361390        GO:0016021   \n",
       "7               4535  ENSG00000198888      ENST00000361390        GO:0055114   \n",
       "8               4535  ENSG00000198888      ENST00000361390        GO:0005743   \n",
       "9               4535  ENSG00000198888      ENST00000361390        GO:0005739   \n",
       "...              ...              ...                  ...               ...   \n",
       "1392787        60678  ENSG00000284869      ENST00000646013               NaN   \n",
       "1392796        56169  ENSG00000285114      ENST00000645112               NaN   \n",
       "1392797        56169  ENSG00000285114      ENST00000642712               NaN   \n",
       "1392798        56169  ENSG00000285114      ENST00000646090               NaN   \n",
       "1392799        56169  ENSG00000285114      ENST00000643960               NaN   \n",
       "\n",
       "                           GO term name  \\\n",
       "5                              membrane   \n",
       "6        integral component of membrane   \n",
       "7           oxidation-reduction process   \n",
       "8          mitochondrial inner membrane   \n",
       "9                         mitochondrion   \n",
       "...                                 ...   \n",
       "1392787                             NaN   \n",
       "1392796                             NaN   \n",
       "1392797                             NaN   \n",
       "1392798                             NaN   \n",
       "1392799                             NaN   \n",
       "\n",
       "                                        GO term definition  \\\n",
       "5        A lipid bilayer along with all the proteins an...   \n",
       "6        The component of a membrane consisting of the ...   \n",
       "7        A metabolic process that results in the remova...   \n",
       "8        The inner, i.e. lumen-facing, lipid bilayer of...   \n",
       "9        A semiautonomous, self replicating organelle t...   \n",
       "...                                                    ...   \n",
       "1392787                                                NaN   \n",
       "1392796                                                NaN   \n",
       "1392797                                                NaN   \n",
       "1392798                                                NaN   \n",
       "1392799                                                NaN   \n",
       "\n",
       "        GO term evidence code     HGNC ID HGNC symbol  \n",
       "5                         IEA   HGNC:7455      MT-ND1  \n",
       "6                         IEA   HGNC:7455      MT-ND1  \n",
       "7                         IEA   HGNC:7455      MT-ND1  \n",
       "8                         IEA   HGNC:7455      MT-ND1  \n",
       "9                         IEA   HGNC:7455      MT-ND1  \n",
       "...                       ...         ...         ...  \n",
       "1392787                   NaN  HGNC:24614      EEFSEC  \n",
       "1392796                   NaN   HGNC:7151       GSDMC  \n",
       "1392797                   NaN   HGNC:7151       GSDMC  \n",
       "1392798                   NaN   HGNC:7151       GSDMC  \n",
       "1392799                   NaN   HGNC:7151       GSDMC  \n",
       "\n",
       "[1306714 rows x 9 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_mappings.loc[~go_mappings['NCBI gene ID'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'4535': 'MT-ND1',\n",
       " '4536': 'MT-ND2',\n",
       " '4512': 'MT-CO1',\n",
       " '113219467': 'MT-TS1',\n",
       " '4513': 'MT-CO2',\n",
       " '4509': 'MT-ATP8',\n",
       " '4508': 'MT-ATP6',\n",
       " '4514': 'MT-CO3',\n",
       " '4537': 'MT-ND3',\n",
       " '4539': 'MT-ND4L',\n",
       " '4538': 'MT-ND4',\n",
       " '4540': 'MT-ND5',\n",
       " '4541': 'MT-ND6',\n",
       " '4519': 'MT-CYB',\n",
       " '1028': 'CDKN1C',\n",
       " '51621': 'KLF13',\n",
       " '255027': 'MPV17L',\n",
       " '9665': 'MARF1',\n",
       " '4849': 'CNOT3',\n",
       " '102723475': 'KCNE1B',\n",
       " '9093': 'DNAJA3',\n",
       " '79165': 'LENG1',\n",
       " '57348': 'TTYH1',\n",
       " '1113': 'CHGA',\n",
       " '5165': 'PDK3',\n",
       " '89927': 'BMERB1',\n",
       " '94104': 'PAXBP1',\n",
       " '1007': 'CDH9',\n",
       " '147798': 'TMC4',\n",
       " '23532': 'PRAME',\n",
       " '338755': 'OR2AG2',\n",
       " '144125': 'OR2AG1',\n",
       " '54664': 'TMEM106B',\n",
       " '51277': 'DNAJC27',\n",
       " '440243': 'GOLGA6L22',\n",
       " '161725': 'OTUD7A',\n",
       " '2558': 'GABRA5',\n",
       " '285148': 'IAH1',\n",
       " '83787': 'ARMC10',\n",
       " '643401': 'PURPL',\n",
       " '7726': 'TRIM26',\n",
       " '2533': 'FYB1',\n",
       " '56245': 'C21orf62',\n",
       " '26531': 'OR11A1',\n",
       " '442194': 'OR10C1',\n",
       " '222236': 'NAPEPLD',\n",
       " '9962': 'SLC23A2',\n",
       " '81696': 'OR5V1',\n",
       " '26530': 'OR12D1',\n",
       " '81797': 'OR12D3',\n",
       " '2562': 'GABRB3',\n",
       " '26529': 'OR12D2',\n",
       " '8590': 'OR6A2',\n",
       " '6868': 'ADAM17',\n",
       " '51314': 'NME8',\n",
       " '7761': 'ZNF214',\n",
       " '102724023': 'GATD3A',\n",
       " '120775': 'OR2D3',\n",
       " '5002': 'SLC22A18',\n",
       " '7762': 'ZNF215',\n",
       " '90527': 'DUOXA1',\n",
       " '6424': 'SFRP4',\n",
       " '144124': 'OR10A5',\n",
       " '120776': 'OR2D2',\n",
       " '283297': 'OR10A4',\n",
       " '341276': 'OR10A2',\n",
       " '54749': 'EPDR1',\n",
       " '5003': 'SLC22A18AS',\n",
       " '8340': 'HIST1H2BL',\n",
       " '51439': 'FAM8A1',\n",
       " '8329': 'HIST1H2AI',\n",
       " '8357': 'HIST1H3H',\n",
       " '353345': 'GPR141',\n",
       " '8227': 'AKAP17A',\n",
       " '3009': 'HIST1H1B',\n",
       " '55975': 'KLHL7',\n",
       " '8332': 'HIST1H2AL',\n",
       " '221806': 'VWDE',\n",
       " '9104': 'RGN',\n",
       " '8362': 'HIST1H4K',\n",
       " '8354': 'HIST1H3I',\n",
       " '3705': 'ITPK1',\n",
       " '8330': 'HIST1H2AK',\n",
       " '8342': 'HIST1H2BM',\n",
       " '56265': 'CPXM1',\n",
       " '57718': 'PPP4R4',\n",
       " '8336': 'HIST1H2AM',\n",
       " '8363': 'HIST1H4J',\n",
       " '8356': 'HIST1H3J',\n",
       " '8368': 'HIST1H4L',\n",
       " '8331': 'HIST1H2AJ',\n",
       " '8341': 'HIST1H2BN',\n",
       " '90525': 'SHF',\n",
       " '223075': 'ITPRID1',\n",
       " '1022': 'CDK7',\n",
       " '10669': 'CGREF1',\n",
       " '405753': 'DUOXA2',\n",
       " '57407': 'NMRAL1',\n",
       " '26716': 'OR2H1',\n",
       " '10398': 'MYL9',\n",
       " '3795': 'KHK',\n",
       " '50506': 'DUOX2',\n",
       " '26015': 'RPAP1',\n",
       " '22839': 'DLGAP4',\n",
       " '51156': 'SERPINA10',\n",
       " '79143': 'MBOAT7',\n",
       " '8100': 'IFT88',\n",
       " '59307': 'SIGIRR',\n",
       " '53905': 'DUOX1',\n",
       " '3163': 'HMOX2',\n",
       " '116511': 'MAS1L',\n",
       " '8468': 'FKBP6',\n",
       " '105375009': 'LINC02829',\n",
       " '8348': 'HIST1H2BO',\n",
       " '81697': 'OR2B2',\n",
       " '10842': 'PPP1R17',\n",
       " '11215': 'AKAP11',\n",
       " '651746': 'ANKRD33B',\n",
       " '64112': 'MOAP1',\n",
       " '219771': 'CCNY',\n",
       " '54765': 'TRIM44',\n",
       " '80831': 'APOL5',\n",
       " '26212': 'OR2B6',\n",
       " '55644': 'OSGEP',\n",
       " '202658': 'RPP21',\n",
       " '57477': 'SHROOM4',\n",
       " '101930100': 'LINC01666',\n",
       " '56658': 'TRIM39',\n",
       " '438': 'ASMT',\n",
       " '3460': 'IFNGR2',\n",
       " '160287': 'LDHAL6A',\n",
       " '283080': 'LINC01517',\n",
       " '56243': 'KIAA1217',\n",
       " '7932': 'OR2H2',\n",
       " '10537': 'UBD',\n",
       " '9246': 'UBE2L6',\n",
       " '100288797': 'TMEM239',\n",
       " '390598': 'SKOR1',\n",
       " '128653': 'C20orf141',\n",
       " '79944': 'L2HGDH',\n",
       " '2550': 'GABBR1',\n",
       " '3092': 'HIP1',\n",
       " '7251': 'TSG101',\n",
       " '100131244': 'ANKRD63',\n",
       " '5330': 'PLCB2',\n",
       " '866': 'SERPINA6',\n",
       " '757': 'TMEM50B',\n",
       " '219770': 'GJD4',\n",
       " '79897': 'RPP21',\n",
       " '8325': 'FZD8',\n",
       " '390502': 'SERPINA2',\n",
       " '8326': 'FZD9',\n",
       " '3812': 'KIR3DL2',\n",
       " '170954': 'PPP1R18',\n",
       " '107987457': 'PPP1R18',\n",
       " '7140': 'TNNT3',\n",
       " '167153': 'TENT2',\n",
       " '11033': 'ADAP1',\n",
       " '200958': 'MUC20',\n",
       " '80741': 'LY6G5C',\n",
       " '55148': 'UBR7',\n",
       " '3106': 'HLA-B',\n",
       " '3120': 'HLA-DQB2',\n",
       " '51604': 'PIGT',\n",
       " '29947': 'DNMT3L',\n",
       " '5269': 'SERPINB6',\n",
       " '140831': 'ZSWIM3',\n",
       " '84520': 'GON7',\n",
       " '8600': 'TNFSF11',\n",
       " '26175': 'TMEM251',\n",
       " '9210': 'BMP15',\n",
       " '140873': 'C20orf173',\n",
       " '387357': 'THEMIS',\n",
       " '2492': 'FSHR',\n",
       " '140890': 'SREK1',\n",
       " '56975': 'FAM20C',\n",
       " '90204': 'ZSWIM1',\n",
       " '23564': 'DDAH2',\n",
       " '79042': 'TSEN34',\n",
       " '255967': 'PAN3',\n",
       " '7407': 'VARS',\n",
       " '100507170': 'CT47A12',\n",
       " '6227': 'RPS21',\n",
       " '6725': 'SRMS',\n",
       " '55388': 'MCM10',\n",
       " '6048': 'RNF5',\n",
       " '54962': 'TIPIN',\n",
       " '64795': 'RMND5A',\n",
       " '245938': 'DEFB125',\n",
       " '55611': 'OTUB1',\n",
       " '30818': 'KCNIP3',\n",
       " '4585': 'MUC4',\n",
       " '255313': 'CT47A11',\n",
       " '84569': 'LYZL1',\n",
       " '2000': 'ELF4',\n",
       " '7844': 'RNF103',\n",
       " '81623': 'DEFB126',\n",
       " '728036': 'CT47A10',\n",
       " '6535': 'SLC6A8',\n",
       " '153478': 'PLEKHG4B',\n",
       " '3196': 'TLX2',\n",
       " '2019': 'EN1',\n",
       " '90639': 'COX19',\n",
       " '140688': 'NOL4L',\n",
       " '9031': 'BAZ1B',\n",
       " '896': 'CCND3',\n",
       " '81850': 'KRTAP1-3',\n",
       " '338323': 'NLRP14',\n",
       " '284428': 'MBD3L5',\n",
       " '79644': 'SRD5A3',\n",
       " '81851': 'KRTAP1-1',\n",
       " '10382': 'TUBB4A',\n",
       " '3133': 'HLA-E',\n",
       " '81872': 'KRTAP2-1',\n",
       " '132001': 'TAMM41',\n",
       " '728042': 'CT47A9',\n",
       " '6538': 'SLC6A11',\n",
       " '79591': 'ARMH3',\n",
       " '1237': 'CCR8',\n",
       " '653656': 'MBD3L4',\n",
       " '100526767': 'RNF103-CHMP3',\n",
       " '143458': 'LDLRAD3',\n",
       " '114823': 'LENG8',\n",
       " '730755': 'KRTAP2-3',\n",
       " '125997': 'MBD3L2',\n",
       " '51652': 'CHMP3',\n",
       " '401138': 'AMTN',\n",
       " '122258': 'SPACA7',\n",
       " '140850': 'DEFB127',\n",
       " '54977': 'SLC25A38',\n",
       " '442179': 'OR1F12',\n",
       " '245939': 'DEFB128',\n",
       " '51617': 'NSG2',\n",
       " '85461': 'TANC1',\n",
       " '84993': 'UBL7',\n",
       " '128497': 'SPATA25',\n",
       " '80345': 'ZSCAN16',\n",
       " '401251': 'SAPCD1',\n",
       " '4987': 'OPRL1',\n",
       " '23769': 'FLRT1',\n",
       " '7718': 'ZNF165',\n",
       " '53335': 'BCL11A',\n",
       " '29928': 'TIMM22',\n",
       " '28992': 'MACROD1',\n",
       " '6203': 'RPS9',\n",
       " '23423': 'TMED3',\n",
       " '29': 'ABR',\n",
       " '54939': 'COMMD4',\n",
       " '222': 'ALDH3B2',\n",
       " '85294': 'KRTAP2-4',\n",
       " '94059': 'LENG9',\n",
       " '2683': 'B4GALT1',\n",
       " '55667': 'DENND4C',\n",
       " '811': 'CALR',\n",
       " '10771': 'ZMYND11',\n",
       " '139728': 'PNCK',\n",
       " '23251': 'MINAR1',\n",
       " '10949': 'HNRNPA0',\n",
       " '8685': 'MARCO',\n",
       " '729020': 'RPEL1',\n",
       " '57189': 'DENND11',\n",
       " '8814': 'CDKL1',\n",
       " '3601': 'IL15RA',\n",
       " '5306': 'PITPNA',\n",
       " '1351': 'COX8A',\n",
       " '9159': 'PCSK7',\n",
       " '9118': 'INA',\n",
       " '5407': 'PNLIPRP1',\n",
       " '55750': 'AGK',\n",
       " '10071': 'MUC12',\n",
       " '5624': 'PROC',\n",
       " '55553': 'SOX6',\n",
       " '2832': 'NPBWR2',\n",
       " '10371': 'SEMA3A',\n",
       " '11025': 'LILRB3',\n",
       " '102725035': 'LILRB3',\n",
       " '107987425': 'LILRB3',\n",
       " '107987462': 'LILRB3',\n",
       " '92521': 'SPECC1',\n",
       " '9499': 'MYOT',\n",
       " '79803': 'HPS6',\n",
       " '168090': 'C6orf118',\n",
       " '9542': 'NRG2',\n",
       " '3107': 'HLA-C',\n",
       " '100233156': 'TEKT4P2',\n",
       " '64210': 'MMS19',\n",
       " '27109': 'DMAC2L',\n",
       " '129881': 'CCDC173',\n",
       " '10752': 'CHL1',\n",
       " '114799': 'ESCO1',\n",
       " '123606': 'NIPA1',\n",
       " '100132386': 'KRTAP4-9',\n",
       " '653240': 'KRTAP4-11',\n",
       " '83755': 'KRTAP4-12',\n",
       " '3908': 'LAMA2',\n",
       " '100533177': 'KRTAP29-1',\n",
       " '2334': 'AFF2',\n",
       " '399814': 'C10orf120',\n",
       " '1192': 'CLIC1',\n",
       " '157506': 'RDH10',\n",
       " '51306': 'FAM13B',\n",
       " '27290': 'SPINK4',\n",
       " '202333': 'CMYA5',\n",
       " '116985': 'ARAP1',\n",
       " '6129': 'RPL7',\n",
       " '7402': 'UTRN',\n",
       " '3111': 'HLA-DOA',\n",
       " '573': 'BAG1',\n",
       " '11270': 'NRM',\n",
       " '79616': 'CCNJL',\n",
       " '83732': 'RIOK1',\n",
       " '51092': 'SIDT2',\n",
       " '51274': 'KLF3',\n",
       " '5265': 'SERPINA1',\n",
       " '10247': 'RIDA',\n",
       " '8722': 'CTSF',\n",
       " '126917': 'IFFO2',\n",
       " '55900': 'ZNF302',\n",
       " '57695': 'USP37',\n",
       " '5476': 'CTSA',\n",
       " '112936': 'VPS26B',\n",
       " '80036': 'TRPM3',\n",
       " '9991': 'PTBP3',\n",
       " '91801': 'ALKBH8',\n",
       " '26285': 'CLDN17',\n",
       " '6347': 'CCL2',\n",
       " '23352': 'UBR4',\n",
       " '7328': 'UBE2H',\n",
       " '9073': 'CLDN8',\n",
       " '10944': 'C11orf58',\n",
       " '23310': 'NCAPD3',\n",
       " '339665': 'SLC35E4',\n",
       " '23095': 'KIF1B',\n",
       " '643803': 'KRTAP24-1',\n",
       " '150290': 'DUSP18',\n",
       " '100131902': 'KRTAP25-1',\n",
       " '340252': 'ZNF680',\n",
       " '6840': 'SVIL',\n",
       " '85477': 'SCIN',\n",
       " '3113': 'HLA-DPA1',\n",
       " '6354': 'CCL7',\n",
       " '326': 'AIRE',\n",
       " '25987': 'TSKU',\n",
       " '110599588': 'ASDURF',\n",
       " '4439': 'MSH5',\n",
       " '10940': 'POP1',\n",
       " '54529': 'ASNSD1',\n",
       " '3115': 'HLA-DPB1',\n",
       " '148170': 'CDC42EP5',\n",
       " '151525': 'WDSUB1',\n",
       " '57583': 'TMEM181',\n",
       " '10075': 'HUWE1',\n",
       " '207063': 'DHRSX',\n",
       " '55735': 'DNAJC11',\n",
       " '79168': 'LILRA6',\n",
       " '1776': 'DNASE1L3',\n",
       " '9437': 'NCR1',\n",
       " '84893': 'FBH1',\n",
       " '6632': 'SNRPD1',\n",
       " '1643': 'DDB2',\n",
       " '27039': 'PKD2L2',\n",
       " '9189': 'ZBED1',\n",
       " '388818': 'KRTAP26-1',\n",
       " '100528021': 'ST20-MTHFS',\n",
       " '10588': 'MTHFS',\n",
       " '2563': 'GABRD',\n",
       " '29058': 'TMEM230',\n",
       " '221044': 'UCMA',\n",
       " '1734': 'DIO2',\n",
       " '338440': 'ANO9',\n",
       " '8861': 'LDB1',\n",
       " '64318': 'NOC3L',\n",
       " '79025': 'FNDC11',\n",
       " '25837': 'RAB26',\n",
       " '64419': 'MTMR14',\n",
       " '2204': 'FCAR',\n",
       " '6356': 'CCL11',\n",
       " '30012': 'TLX3',\n",
       " '5368': 'PNOC',\n",
       " '57176': 'VARS2',\n",
       " '79603': 'CERS4',\n",
       " '728049': 'CT47A8',\n",
       " '6355': 'CCL8',\n",
       " '100996455': 'LINC02724',\n",
       " '81793': 'TLR10',\n",
       " '653282': 'CT47A7',\n",
       " '155051': 'CRYGN',\n",
       " '3904': 'LAIR2',\n",
       " '4798': 'NFRKB',\n",
       " '85280': 'KRTAP9-4',\n",
       " '728062': 'CT47A6',\n",
       " '83700': 'JAM3',\n",
       " '151230': 'KLHL23',\n",
       " '100526832': 'KLHL23',\n",
       " '9081': 'PRYP3',\n",
       " '5228': 'PGF',\n",
       " '6876': 'TAGLN',\n",
       " '221527': 'ZBTB12',\n",
       " '222950': 'NYAP1',\n",
       " '90326': 'THAP3',\n",
       " '629': 'CFB',\n",
       " '255426': 'RASGEF1C',\n",
       " '4650': 'MYO9B',\n",
       " '389257': 'LRRC14B',\n",
       " '7096': 'TLR1',\n",
       " '493911': 'PHOSPHO2',\n",
       " '140881': 'DEFB129',\n",
       " '7262': 'PHLDA2',\n",
       " '1021': 'CDK6',\n",
       " '4123': 'MAN2C1',\n",
       " '123036': 'TC2N',\n",
       " '9782': 'MATR3',\n",
       " '9061': 'PAPSS1',\n",
       " '219464': 'OR5T2',\n",
       " '390154': 'OR5T3',\n",
       " '400830': 'DEFB132',\n",
       " '83900': 'KRTAP9-3',\n",
       " '83899': 'KRTAP9-2',\n",
       " '7429': 'VIL1',\n",
       " '10333': 'TLR6',\n",
       " '653657': 'MBD3L3',\n",
       " '55727': 'BTBD7',\n",
       " '5886': 'RAD23A',\n",
       " '92092': 'ZC3HAV1L',\n",
       " '196996': 'GRAMD2A',\n",
       " '6993': 'DYNLT1',\n",
       " '55296': 'TBC1D19',\n",
       " '255725': 'OR52B2',\n",
       " '4773': 'NFATC2',\n",
       " '81847': 'RNF146',\n",
       " '196335': 'OR56B4',\n",
       " '4999': 'ORC2',\n",
       " '120787': 'OR52W1',\n",
       " '5360': 'PLTP',\n",
       " '81928': 'CABLES2',\n",
       " '94120': 'SYTL3',\n",
       " '100507632': 'LINC00968',\n",
       " '643812': 'KRTAP27-1',\n",
       " '165257': 'C1QL2',\n",
       " '337963': 'KRTAP23-1',\n",
       " '729597': 'SPDYE6',\n",
       " '6742': 'SSBP1',\n",
       " '337959': 'KRTAP13-2',\n",
       " '5179': 'PENK',\n",
       " '140258': 'KRTAP13-1',\n",
       " '55312': 'RFK',\n",
       " '9575': 'CLOCK',\n",
       " '337960': 'KRTAP13-3',\n",
       " '728072': 'CT47A5',\n",
       " '133957': 'CCDC127',\n",
       " '284827': 'KRTAP13-4',\n",
       " '254950': 'KRTAP15-1',\n",
       " '3206': 'HOXA10',\n",
       " '165545': 'DQX1',\n",
       " '5154': 'PDGFA',\n",
       " '337882': 'KRTAP19-1',\n",
       " '79820': 'CATSPERB',\n",
       " '2556': 'GABRA3',\n",
       " '90480': 'GADD45GIP1',\n",
       " '337969': 'KRTAP19-2',\n",
       " '337970': 'KRTAP19-3',\n",
       " '100505724': 'KRTAP9-7',\n",
       " '10869': 'USP19',\n",
       " '54809': 'SAMD9',\n",
       " '337971': 'KRTAP19-4',\n",
       " '258': 'AMBN',\n",
       " '23244': 'PDS5A',\n",
       " '337972': 'KRTAP19-5',\n",
       " '23016': 'EXOSC7',\n",
       " '710': 'SERPING1',\n",
       " '1004': 'CDH6',\n",
       " '4709': 'NDUFB3',\n",
       " '717': 'C2',\n",
       " '7145': 'TNS1',\n",
       " '390155': 'OR5T1',\n",
       " '65125': 'WNK1',\n",
       " '56729': 'RETN',\n",
       " '728075': 'CT47A4',\n",
       " '219469': 'OR8H1',\n",
       " '79706': 'PRKRIP1',\n",
       " '105379487': 'LINC01670',\n",
       " '84071': 'ARMC2',\n",
       " '10942': 'PRSS21',\n",
       " '6357': 'CCL13',\n",
       " '285172': 'FAM126B',\n",
       " '7174': 'TPP2',\n",
       " '3205': 'HOXA9',\n",
       " '92689': 'FAM114A1',\n",
       " '337973': 'KRTAP19-6',\n",
       " '337974': 'KRTAP19-7',\n",
       " '100288287': 'KRTAP22-2',\n",
       " '50809': 'HP1BP3',\n",
       " '3921': 'RPSA',\n",
       " '337968': 'KRTAP6-3',\n",
       " '337967': 'KRTAP6-2',\n",
       " '337979': 'KRTAP22-1',\n",
       " '494551': 'WEE2',\n",
       " '337966': 'KRTAP6-1',\n",
       " '337975': 'KRTAP20-1',\n",
       " '5111': 'PCNA',\n",
       " '51304': 'ZDHHC3',\n",
       " '5813': 'PURA',\n",
       " '345757': 'FAM174A',\n",
       " '55251': 'PCMTD2',\n",
       " '56829': 'ZC3HAV1',\n",
       " '100507608': 'KRTAP9-6',\n",
       " '54539': 'NDUFB11',\n",
       " '81870': 'KRTAP9-9',\n",
       " '57491': 'AHRR',\n",
       " '80017': 'DGLUCY',\n",
       " '140825': 'NEURL2',\n",
       " '219285': 'SAMD9L',\n",
       " '177': 'AGER',\n",
       " '23287': 'AGTPBP1',\n",
       " '144809': 'FAM216B',\n",
       " '160298': 'C11orf42',\n",
       " '8241': 'RBM10',\n",
       " '5077': 'PAX3',\n",
       " '55531': 'ELMOD1',\n",
       " '9319': 'TRIP13',\n",
       " '9083': 'BPY2',\n",
       " '442867': 'BPY2',\n",
       " '442868': 'BPY2',\n",
       " '10124': 'ARL4A',\n",
       " '27288': 'RBMXL2',\n",
       " '22907': 'DHX30',\n",
       " '10344': 'CCL26',\n",
       " '4661': 'MYT1',\n",
       " '151835': 'CPNE9',\n",
       " '55858': 'TMEM165',\n",
       " '100151643': 'KRTAP20-4',\n",
       " '22827': 'PUF60',\n",
       " '7301': 'TYRO3',\n",
       " '337976': 'KRTAP20-2',\n",
       " '79631': 'EFL1',\n",
       " '337985': 'KRTAP20-3',\n",
       " '100288323': 'KRTAP21-3',\n",
       " '337977': 'KRTAP21-1',\n",
       " '337978': 'KRTAP21-2',\n",
       " '10016': 'PDCD6',\n",
       " '116535': 'MRGPRF',\n",
       " '7123': 'CLEC3B',\n",
       " '5634': 'PRPS2',\n",
       " '6194': 'RPS6',\n",
       " '1139': 'CHRNA7',\n",
       " '102724631': 'POTEB3',\n",
       " '79815': 'NIPAL2',\n",
       " '199675': 'MCEMP1',\n",
       " '5926': 'ARID4A',\n",
       " '79230': 'ZNF557',\n",
       " '619279': 'ZNF704',\n",
       " '389610': 'XKR5',\n",
       " '6892': 'TAPBP',\n",
       " '401': 'PHOX2A',\n",
       " '6138': 'RPL15',\n",
       " '348654': 'GEN1',\n",
       " '126003': 'TRAPPC5',\n",
       " '1672': 'DEFB1',\n",
       " '3636': 'INPPL1',\n",
       " '1823': 'DSC1',\n",
       " '2968': 'GTF2H4',\n",
       " '55203': 'LGI2',\n",
       " '55824': 'PAG1',\n",
       " '124935': 'SLC43A2',\n",
       " '4122': 'MAN2A2',\n",
       " '2173': 'FABP7',\n",
       " '6588': 'SLN',\n",
       " '22906': 'TRAK1',\n",
       " '115701': 'ALPK2',\n",
       " '54790': 'TET2',\n",
       " '54476': 'RNF216',\n",
       " '513': 'ATP5F1D',\n",
       " '149775': 'GNAS-AS1',\n",
       " '151278': 'CCDC140',\n",
       " '344752': 'AADACL2',\n",
       " '54784': 'ALKBH4',\n",
       " '81490': 'PTDSS2',\n",
       " '2517': 'FUCA1',\n",
       " '6346': 'CCL1',\n",
       " '6389': 'SDHA',\n",
       " '1671': 'DEFA6',\n",
       " '55620': 'STAP2',\n",
       " '29062': 'WDR91',\n",
       " '54510': 'PCDH18',\n",
       " '1824': 'DSC2',\n",
       " '80228': 'ORAI2',\n",
       " '55655': 'NLRP2',\n",
       " '2778': 'GNAS',\n",
       " '5834': 'PYGB',\n",
       " '79839': 'CCDC102B',\n",
       " '6850': 'SYK',\n",
       " '129049': 'SGSM1',\n",
       " '114898': 'C1QTNF2',\n",
       " '728082': 'CT47A3',\n",
       " '63920': 'ZBED8',\n",
       " '728090': 'CT47A2',\n",
       " '285761': 'DCBLD1',\n",
       " '728096': 'CT47A1',\n",
       " '11027': 'LILRA2',\n",
       " '6598': 'SMARCB1',\n",
       " '79677': 'SMC6',\n",
       " '51530': 'ZC3HC1',\n",
       " '140893': 'RBBP8NL',\n",
       " '55879': 'GABRQ',\n",
       " '157777': 'MCMDC2',\n",
       " '6654': 'SOS1',\n",
       " '9975': 'NR1D2',\n",
       " '57406': 'ABHD6',\n",
       " '1611': 'DAP',\n",
       " '80008': 'TMEM156',\n",
       " '2747': 'GLUD2',\n",
       " '11046': 'SLC35D2',\n",
       " '8892': 'EIF2B2',\n",
       " '26090': 'ABHD12',\n",
       " '4488': 'MSX2',\n",
       " '26154': 'ABCA12',\n",
       " '57120': 'GOPC',\n",
       " '7920': 'ABHD16A',\n",
       " '2208': 'FCER2',\n",
       " '54872': 'PIGG',\n",
       " '23082': 'PPRC1',\n",
       " '79290': 'OR13A1',\n",
       " '202243': 'CCDC125',\n",
       " '7033': 'TFF3',\n",
       " '29099': 'COMMD9',\n",
       " '25942': 'SIN3A',\n",
       " '115273': 'RAB42',\n",
       " '10924': 'SMPDL3A',\n",
       " '6883': 'TAF12',\n",
       " '3605': 'IL17A',\n",
       " '28512': 'NKIRAS1',\n",
       " '219473': 'OR8K3',\n",
       " '8449': 'DHX16',\n",
       " '122553': 'TRAPPC6B',\n",
       " '51015': 'ISOC1',\n",
       " '3293': 'HSD17B3',\n",
       " '6369': 'CCL24',\n",
       " '90342': 'FER1L5',\n",
       " '3005': 'H1F0',\n",
       " '6098': 'ROS1',\n",
       " '64374': 'SIL1',\n",
       " '1244': 'ABCC2',\n",
       " '8225': 'GTPBP6',\n",
       " '219': 'ALDH1B1',\n",
       " '55607': 'PPP1R9A',\n",
       " '729288': 'ZNF286B',\n",
       " '51175': 'TUBE1',\n",
       " '51284': 'TLR7',\n",
       " '157574': 'FBXO16',\n",
       " '254272': 'TBC1D28',\n",
       " '347252': 'IGFBPL1',\n",
       " '100507303': 'SNHG19',\n",
       " '6854': 'SYN2',\n",
       " '240': 'ALOX5',\n",
       " '51311': 'TLR8',\n",
       " '3112': 'HLA-DOB',\n",
       " '55344': 'PLCXD1',\n",
       " '221143': 'EEF1AKMT1',\n",
       " '158405': 'KIAA1958',\n",
       " '4869': 'NPM1',\n",
       " '10809': 'STARD10',\n",
       " '81614': 'NIPA2',\n",
       " '91319': 'DERL3',\n",
       " '158471': 'PRUNE2',\n",
       " '2310': 'FOXO3B',\n",
       " '53342': 'IL17D',\n",
       " '84263': 'HSDL2',\n",
       " '101926971': 'LINC01284',\n",
       " '54928': 'IMPAD1',\n",
       " '8938': 'BAIAP3',\n",
       " '1269': 'CNR2',\n",
       " '199713': 'NLRP7',\n",
       " '124944': 'C17orf49',\n",
       " '7697': 'ZNF138',\n",
       " '55629': 'PNRC2',\n",
       " '340602': 'EZHIP',\n",
       " '150160': 'CCT8L2',\n",
       " '170685': 'NUDT10',\n",
       " '55893': 'ZNF395',\n",
       " '8838': 'CCN6',\n",
       " '79623': 'GALNT14',\n",
       " '9938': 'ARHGAP25',\n",
       " '221545': 'C6orf136',\n",
       " '84902': 'CEP89',\n",
       " '203611': 'CDY2B',\n",
       " '51011': 'FAHD2A',\n",
       " '23191': 'CYFIP1',\n",
       " '107983955': 'ELOA3C',\n",
       " '134829': 'CLVS2',\n",
       " '9063': 'PIAS2',\n",
       " '51427': 'ZNF107',\n",
       " '23657': 'SLC7A11',\n",
       " '203074': 'PRSS55',\n",
       " '7079': 'TIMP4',\n",
       " '79661': 'NEIL1',\n",
       " '199699': 'DAND5',\n",
       " '390157': 'OR8K1',\n",
       " '8744': 'TNFSF9',\n",
       " '80179': 'MYO19',\n",
       " '55190': 'NUDT11',\n",
       " '140680': 'C20orf96',\n",
       " '7032': 'TFF2',\n",
       " '100507436': 'MICA',\n",
       " '1669': 'DEFA4',\n",
       " '440400': 'RNASEK',\n",
       " '2650': 'GCNT1',\n",
       " '55683': 'KANSL3',\n",
       " '10117': 'ENAM',\n",
       " '3643': 'INSR',\n",
       " '94240': 'EPSTI1',\n",
       " '5411': 'PNN',\n",
       " '4771': 'NF2',\n",
       " '11177': 'BAZ1A',\n",
       " '149951': 'COMMD7',\n",
       " '27430': 'MAT2B',\n",
       " '140453': 'MUC17',\n",
       " '339318': 'ZNF181',\n",
       " '4340': 'MOG',\n",
       " '7127': 'TNFAIP2',\n",
       " '3512': 'JCHAIN',\n",
       " '2171': 'FABP5',\n",
       " '8495': 'PPFIBP2',\n",
       " '619208': 'FAM229B',\n",
       " '162699': 'ELOA3D',\n",
       " '51224': 'ELOA2',\n",
       " '1828': 'DSG1',\n",
       " '284805': 'C20orf203',\n",
       " '91404': 'SESTD1',\n",
       " '728929': 'ELOA3B',\n",
       " '7031': 'TFF1',\n",
       " '100506888': 'ELOA3',\n",
       " '10569': 'SLU7',\n",
       " '83473': 'KATNAL2',\n",
       " '9618': 'TRAF4',\n",
       " '54943': 'DNAJC28',\n",
       " '105375622': 'LINC02849',\n",
       " '159963': 'SLC5A12',\n",
       " '10630': 'PDPN',\n",
       " '81844': 'TRIM56',\n",
       " '79148': 'MMP28',\n",
       " '219477': 'OR8J1',\n",
       " '345778': 'MTX3',\n",
       " '147409': 'DSG4',\n",
       " '504189': 'OR8U8',\n",
       " '85364': 'ZCCHC3',\n",
       " '504190': 'OR8U9',\n",
       " '1667': 'DEFA1',\n",
       " '440934': 'CT75',\n",
       " '80023': 'NRSN2',\n",
       " '5725': 'PTBP1',\n",
       " '970': 'CD70',\n",
       " '57819': 'LSM2',\n",
       " '23066': 'CAND2',\n",
       " '79722': 'ANKRD55',\n",
       " '7430': 'EZR',\n",
       " '387758': 'FIBIN',\n",
       " '23514': 'SPIDR',\n",
       " '51091': 'SEPSECS',\n",
       " '92799': 'SHKBP1',\n",
       " '441308': 'OR4F21',\n",
       " '7253': 'TSHR',\n",
       " '55240': 'STEAP3',\n",
       " '54905': 'CYP2W1',\n",
       " '2321': 'FLT1',\n",
       " '8424': 'BBOX1',\n",
       " '9468': 'PCYT1B',\n",
       " '57414': 'RHBDD2',\n",
       " '239': 'ALOX12',\n",
       " '169270': 'ZNF596',\n",
       " '3572': 'IL6ST',\n",
       " '346673': 'STRA8',\n",
       " '401027': 'C2orf66',\n",
       " '27346': 'TMEM97',\n",
       " '1832': 'DSP',\n",
       " '5375': 'PMP2',\n",
       " '196792': 'FAM24B',\n",
       " '55328': 'RNLS',\n",
       " '130074': 'FAM168B',\n",
       " '6666': 'SOX12',\n",
       " '170949': 'DEFT1P',\n",
       " '100287083': 'DEFT1P',\n",
       " '90410': 'IFT20',\n",
       " '54733': 'SLC35F2',\n",
       " '728358': 'DEFA1B',\n",
       " '646486': 'FABP12',\n",
       " '5408': 'PNLIPRP2',\n",
       " '50624': 'CUZD1',\n",
       " '646480': 'FABP9',\n",
       " '64328': 'XPO4',\n",
       " '91584': 'PLXNA4',\n",
       " '7114': 'TMSB4X',\n",
       " '818': 'CAMK2G',\n",
       " '6161': 'RPL32',\n",
       " '123720': 'WHAMM',\n",
       " '2167': 'FABP4',\n",
       " '51272': 'BET1L',\n",
       " '9252': 'RPS6KA5',\n",
       " '4967': 'OGDH',\n",
       " '255349': 'TMEM211',\n",
       " '64866': 'CDCP1',\n",
       " '203076': 'C8orf74',\n",
       " '142910': 'LIPJ',\n",
       " '6891': 'TAP2',\n",
       " '85481': 'PSKH2',\n",
       " '9972': 'NUP153',\n",
       " '55231': 'CCDC87',\n",
       " '4336': 'MOBP',\n",
       " '94137': 'RP1L1',\n",
       " '245972': 'ATP6V0D2',\n",
       " '10892': 'MALT1',\n",
       " '117583': 'PARD3B',\n",
       " '1668': 'DEFA3',\n",
       " '147166': 'TRIM16L',\n",
       " '100463482': 'MTRNR2L6',\n",
       " '6741': 'SSB',\n",
       " '113746': 'ODF3',\n",
       " '166929': 'SGMS2',\n",
       " '8740': 'TNFSF14',\n",
       " '145508': 'CEP128',\n",
       " '10793': 'ZNF273',\n",
       " '1794': 'DOCK2',\n",
       " '80742': 'PRR3',\n",
       " '541565': 'C8orf58',\n",
       " '64711': 'HS3ST6',\n",
       " '337879': 'KRTAP8-1',\n",
       " '1670': 'DEFA5',\n",
       " '337878': 'KRTAP7-1',\n",
       " '64116': 'SLC39A8',\n",
       " '147199': 'SCGB1C1',\n",
       " '337880': 'KRTAP11-1',\n",
       " '91057': 'CCDC34',\n",
       " '29081': 'METTL5',\n",
       " '728299': 'KRTAP19-8',\n",
       " '113612': 'CYP2U1',\n",
       " '102157402': 'AK6',\n",
       " '6150': 'MRPL23',\n",
       " '107987373': 'MRPL23',\n",
       " '9786': 'KIAA0586',\n",
       " '5159': 'PDGFRB',\n",
       " '107984345': 'SMIM38',\n",
       " '2794': 'GNL1',\n",
       " '390174': 'OR9G1',\n",
       " '219931': 'TPCN2',\n",
       " '504191': 'OR9G9',\n",
       " '3268': 'AGFG2',\n",
       " '51351': 'ZNF117',\n",
       " '109504726': 'ZNF117',\n",
       " '9326': 'ZNHIT3',\n",
       " '148103': 'ZNF599',\n",
       " '116449': 'CLNK',\n",
       " '283189': 'OR9G4',\n",
       " '2086': 'ERV3-1',\n",
       " '6880': 'TAF9',\n",
       " '2618': 'GART',\n",
       " '64236': 'PDLIM2',\n",
       " '150165': 'XKR3',\n",
       " '55677': 'IWS1',\n",
       " '85379': 'KIAA1671',\n",
       " '3207': 'HOXA11',\n",
       " '440270': 'GOLGA8B',\n",
       " '124842': 'TMEM132E',\n",
       " '7384': 'UQCRC1',\n",
       " '353149': 'TBC1D26',\n",
       " '25907': 'TMEM158',\n",
       " '5884': 'RAD17',\n",
       " '57584': 'ARHGAP21',\n",
       " '100505573': 'INAFM2',\n",
       " '9131': 'AIFM1',\n",
       " '154007': 'SNRNP48',\n",
       " '84067': 'FAM160A2',\n",
       " '115653': 'KIR3DL3',\n",
       " '344387': 'CDKL4',\n",
       " '4102': 'MAGEA3',\n",
       " '256364': 'EML3',\n",
       " '284098': 'PIGW',\n",
       " '79977': 'GRHL2',\n",
       " '492311': 'IGIP',\n",
       " '128954': 'GAB4',\n",
       " '8760': 'CDS2',\n",
       " '3788': 'KCNS2',\n",
       " '57050': 'UTP3',\n",
       " '102723547': 'CSAG2',\n",
       " '9439': 'MED23',\n",
       " '3804': 'KIR2DL3',\n",
       " '10666': 'CD226',\n",
       " '143425': 'SYT9',\n",
       " '57524': 'CASKIN1',\n",
       " '6788': 'STK3',\n",
       " '11102': 'RPP14',\n",
       " '90362': 'FAM110B',\n",
       " '197131': 'UBR1',\n",
       " '283298': 'OLFML1',\n",
       " '81622': 'UNC93B1',\n",
       " '1617': 'DAZ1',\n",
       " '84174': 'SLA2',\n",
       " '22902': 'RUFY3',\n",
       " '91442': 'FAAP24',\n",
       " '9275': 'BCL7B',\n",
       " '55300': 'PI4K2B',\n",
       " '109703458': 'HTD2',\n",
       " '84231': 'TRAF7',\n",
       " '10331': 'B3GNT3',\n",
       " '60436': 'TGIF2',\n",
       " '220164': 'DOK6',\n",
       " '3150': 'HMGN1',\n",
       " '84196': 'USP48',\n",
       " '79133': 'NDUFAF5',\n",
       " '7074': 'TIAM1',\n",
       " '57335': 'ZNF286A',\n",
       " '23365': 'ARHGEF12',\n",
       " '27000': 'DNAJC2',\n",
       " '7179': 'TPTE',\n",
       " '1830': 'DSG3',\n",
       " '257062': 'CATSPERD',\n",
       " '101927120': 'LINC02735',\n",
       " '1946': 'EFNA5',\n",
       " '102724428': 'SIK1B',\n",
       " '171586': 'ABHD3',\n",
       " '1387': 'CREBBP',\n",
       " '100131897': 'INSYN2B',\n",
       " '3811': 'KIR3DL1',\n",
       " '85460': 'ZNF518B',\n",
       " '653486': 'SCGB1C2',\n",
       " '55969': 'RAB5IF',\n",
       " '5089': 'PBX2',\n",
       " '54984': 'PINX1',\n",
       " '51614': 'ERGIC3',\n",
       " '23008': 'KLHDC10',\n",
       " '8447': 'DOC2B',\n",
       " '55284': 'UBE2W',\n",
       " '54919': 'DNAAF5',\n",
       " '157724': 'SLC7A13',\n",
       " '10910': 'SUGT1',\n",
       " '54927': 'CHCHD3',\n",
       " '163154': 'PRR22',\n",
       " '9656': 'MDC1',\n",
       " '64172': 'OSGEPL1',\n",
       " '23762': 'OSBP2',\n",
       " '1302': 'COL11A2',\n",
       " '83595': 'SOX7',\n",
       " '10497': 'UNC13B',\n",
       " '283726': 'SAXO2',\n",
       " '56931': 'DUS3L',\n",
       " '23258': 'DENND5A',\n",
       " '100527943': 'TGIF2-RAB5IF',\n",
       " '132884': 'EVC2',\n",
       " '5575': 'PRKAR1B',\n",
       " '347549': 'CENPVL3',\n",
       " '29063': 'ZCCHC4',\n",
       " '150709': 'ANKAR',\n",
       " '115548': 'FCHO2',\n",
       " '6094': 'ROM1',\n",
       " '375611': 'SLC26A5',\n",
       " '115939': 'TSR3',\n",
       " '654': 'BMP6',\n",
       " '22941': 'SHANK2',\n",
       " '51115': 'RMDN1',\n",
       " '3802': 'KIR2DL1',\n",
       " '100129654': 'TCF24',\n",
       " '63915': 'BLOC1S5',\n",
       " '7060': 'THBS4',\n",
       " '253650': 'ANKRD18A',\n",
       " '116969': 'ART5',\n",
       " '81567': 'TXNDC5',\n",
       " '3809': 'KIR2DS4',\n",
       " '118670': 'FAM24A',\n",
       " '4034': 'LRCH4',\n",
       " '5201': 'PFDN1',\n",
       " '3805': 'KIR2DL4',\n",
       " '112268354': 'KIR2DL4',\n",
       " '3613': 'IMPA2',\n",
       " '4784': 'NFIX',\n",
       " '259217': 'HSPA12A',\n",
       " '79664': 'ICE2',\n",
       " '27067': 'STAU2',\n",
       " '10846': 'PDE10A',\n",
       " '90632': 'PDE10A',\n",
       " '55366': 'LGR4',\n",
       " '51371': 'POMP',\n",
       " '417': 'ART1',\n",
       " '84418': 'CYSTM1',\n",
       " '383': 'ARG1',\n",
       " '85319': 'BAGE2',\n",
       " '9521': 'EEF1E1',\n",
       " '57698': 'SHTN1',\n",
       " '283537': 'SLC46A3',\n",
       " '658': 'BMPR1B',\n",
       " '130367': 'SGPP2',\n",
       " '26579': 'MYEOV',\n",
       " '735': 'C9',\n",
       " '100316904': 'SAP25',\n",
       " '441495': 'CENPVL2',\n",
       " '389857': 'CENPVL1',\n",
       " '266740': 'MAGEA2B',\n",
       " '28227': 'PPP2R3B',\n",
       " '23708': 'GSPT2',\n",
       " '203068': 'TUBB',\n",
       " '203328': 'SUSD3',\n",
       " '23543': 'RBFOX2',\n",
       " '597': 'BCL2A1',\n",
       " ...}"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create mapping of NCBI gene id to gene name/symbol\n",
    "\n",
    "map_geneid_to_hfnc_symbol = go_mappings[['NCBI gene ID', 'HGNC symbol']] \\\n",
    "    .drop_duplicates() \\\n",
    "    .dropna()\n",
    "# map_geneid_to_hfnc_symbol['NCBI gene ID'] = map_geneid_to_hfnc_symbol['NCBI gene ID'].astype(int)\n",
    "map_geneid_to_hfnc_symbol = map_geneid_to_hfnc_symbol.set_index('NCBI gene ID') \\\n",
    "    .to_dict()['HGNC symbol']\n",
    "map_geneid_to_hfnc_symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map NCBI gene id to gene name/symbol\n",
    "lihc_processed = lihc.copy(deep=True)\n",
    "\n",
    "lihc_processed.columns = list(map(lambda x: x.split('|')[1], lihc_processed.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1028</th>\n",
       "      <th>51621</th>\n",
       "      <th>255027</th>\n",
       "      <th>9665</th>\n",
       "      <th>4849</th>\n",
       "      <th>9093</th>\n",
       "      <th>79165</th>\n",
       "      <th>57348</th>\n",
       "      <th>1113</th>\n",
       "      <th>5165</th>\n",
       "      <th>...</th>\n",
       "      <th>22826</th>\n",
       "      <th>93974</th>\n",
       "      <th>83667</th>\n",
       "      <th>54797</th>\n",
       "      <th>389072</th>\n",
       "      <th>1421</th>\n",
       "      <th>1420</th>\n",
       "      <th>219990</th>\n",
       "      <th>79669</th>\n",
       "      <th>257144</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>72.3245</td>\n",
       "      <td>768.8549</td>\n",
       "      <td>13.0314</td>\n",
       "      <td>1048.3792</td>\n",
       "      <td>1090.0798</td>\n",
       "      <td>1720.8014</td>\n",
       "      <td>289.9495</td>\n",
       "      <td>33.8817</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>104.9031</td>\n",
       "      <td>...</td>\n",
       "      <td>1751.4253</td>\n",
       "      <td>3309.3338</td>\n",
       "      <td>1008.6333</td>\n",
       "      <td>533.6374</td>\n",
       "      <td>187.6527</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.8189</td>\n",
       "      <td>29.9723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>41.3997</td>\n",
       "      <td>4116.8063</td>\n",
       "      <td>690.4879</td>\n",
       "      <td>1209.9556</td>\n",
       "      <td>736.8162</td>\n",
       "      <td>2790.0444</td>\n",
       "      <td>183.8344</td>\n",
       "      <td>25.6284</td>\n",
       "      <td>2.9571</td>\n",
       "      <td>134.5490</td>\n",
       "      <td>...</td>\n",
       "      <td>1984.2287</td>\n",
       "      <td>3247.4125</td>\n",
       "      <td>1559.8817</td>\n",
       "      <td>565.3031</td>\n",
       "      <td>324.7905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.8857</td>\n",
       "      <td>8.8714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GT-01A-11R-A38B-07</td>\n",
       "      <td>48.4196</td>\n",
       "      <td>3557.0500</td>\n",
       "      <td>642.0085</td>\n",
       "      <td>1310.0202</td>\n",
       "      <td>954.0462</td>\n",
       "      <td>1407.7561</td>\n",
       "      <td>115.6691</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>112.0825</td>\n",
       "      <td>...</td>\n",
       "      <td>1500.1121</td>\n",
       "      <td>3359.7848</td>\n",
       "      <td>720.9146</td>\n",
       "      <td>494.0596</td>\n",
       "      <td>633.0419</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.3800</td>\n",
       "      <td>17.0365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GU-01A-11R-A38B-07</td>\n",
       "      <td>54.2191</td>\n",
       "      <td>2182.9435</td>\n",
       "      <td>862.5004</td>\n",
       "      <td>1270.3947</td>\n",
       "      <td>2207.9677</td>\n",
       "      <td>2332.2545</td>\n",
       "      <td>241.0664</td>\n",
       "      <td>10.0097</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>40.0387</td>\n",
       "      <td>...</td>\n",
       "      <td>1516.4659</td>\n",
       "      <td>3189.7501</td>\n",
       "      <td>786.5937</td>\n",
       "      <td>320.3096</td>\n",
       "      <td>69.2336</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.5024</td>\n",
       "      <td>3.3366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GV-01A-11R-A38B-07</td>\n",
       "      <td>45.6853</td>\n",
       "      <td>2036.3790</td>\n",
       "      <td>296.9543</td>\n",
       "      <td>1212.3519</td>\n",
       "      <td>833.3333</td>\n",
       "      <td>1542.3012</td>\n",
       "      <td>120.1354</td>\n",
       "      <td>1.6920</td>\n",
       "      <td>38.0711</td>\n",
       "      <td>65.9898</td>\n",
       "      <td>...</td>\n",
       "      <td>1544.8393</td>\n",
       "      <td>2682.7411</td>\n",
       "      <td>963.6210</td>\n",
       "      <td>438.2403</td>\n",
       "      <td>282.5719</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8460</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-ZS-A9CD-01A-11R-A37K-07</td>\n",
       "      <td>40.5710</td>\n",
       "      <td>1616.8295</td>\n",
       "      <td>468.0691</td>\n",
       "      <td>942.9001</td>\n",
       "      <td>670.9241</td>\n",
       "      <td>2818.1818</td>\n",
       "      <td>222.3892</td>\n",
       "      <td>21.7881</td>\n",
       "      <td>12.0210</td>\n",
       "      <td>85.6499</td>\n",
       "      <td>...</td>\n",
       "      <td>1625.0939</td>\n",
       "      <td>3519.9098</td>\n",
       "      <td>1054.0947</td>\n",
       "      <td>293.0128</td>\n",
       "      <td>178.0616</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7513</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-ZS-A9CE-01A-11R-A37K-07</td>\n",
       "      <td>20.7877</td>\n",
       "      <td>4619.2560</td>\n",
       "      <td>695.8425</td>\n",
       "      <td>1242.8884</td>\n",
       "      <td>699.1247</td>\n",
       "      <td>1862.1444</td>\n",
       "      <td>237.9650</td>\n",
       "      <td>3.2823</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>101.7505</td>\n",
       "      <td>...</td>\n",
       "      <td>1514.2232</td>\n",
       "      <td>4958.4245</td>\n",
       "      <td>471.0066</td>\n",
       "      <td>439.2779</td>\n",
       "      <td>235.2298</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5470</td>\n",
       "      <td>1.0941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-ZS-A9CF-01A-11R-A38B-07</td>\n",
       "      <td>63.4045</td>\n",
       "      <td>793.9352</td>\n",
       "      <td>504.4797</td>\n",
       "      <td>1274.2936</td>\n",
       "      <td>671.2612</td>\n",
       "      <td>2253.6182</td>\n",
       "      <td>132.3225</td>\n",
       "      <td>7.5810</td>\n",
       "      <td>5.5134</td>\n",
       "      <td>94.4176</td>\n",
       "      <td>...</td>\n",
       "      <td>2202.6189</td>\n",
       "      <td>5642.3156</td>\n",
       "      <td>1602.3432</td>\n",
       "      <td>809.0972</td>\n",
       "      <td>474.1558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.9593</td>\n",
       "      <td>6.2026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-ZS-A9CF-02A-11R-A38B-07</td>\n",
       "      <td>33.3745</td>\n",
       "      <td>2132.2621</td>\n",
       "      <td>632.8801</td>\n",
       "      <td>1221.8789</td>\n",
       "      <td>597.6514</td>\n",
       "      <td>1804.0791</td>\n",
       "      <td>192.8307</td>\n",
       "      <td>0.6180</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>84.6724</td>\n",
       "      <td>...</td>\n",
       "      <td>1734.2398</td>\n",
       "      <td>4358.4672</td>\n",
       "      <td>1465.3894</td>\n",
       "      <td>851.0507</td>\n",
       "      <td>768.8504</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2361</td>\n",
       "      <td>6.7985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-ZS-A9CG-01A-11R-A37K-07</td>\n",
       "      <td>49.3573</td>\n",
       "      <td>2696.1440</td>\n",
       "      <td>256.3839</td>\n",
       "      <td>1244.9015</td>\n",
       "      <td>802.0566</td>\n",
       "      <td>1883.1191</td>\n",
       "      <td>235.8183</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>46.6153</td>\n",
       "      <td>...</td>\n",
       "      <td>1417.6521</td>\n",
       "      <td>3505.7412</td>\n",
       "      <td>1863.9246</td>\n",
       "      <td>401.0283</td>\n",
       "      <td>228.2776</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6855</td>\n",
       "      <td>1.3710</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>423 rows Ã— 19228 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 1028      51621    255027       9665  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  72.3245   768.8549   13.0314  1048.3792   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  41.3997  4116.8063  690.4879  1209.9556   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07  48.4196  3557.0500  642.0085  1310.0202   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07  54.2191  2182.9435  862.5004  1270.3947   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  45.6853  2036.3790  296.9543  1212.3519   \n",
       "...                               ...        ...       ...        ...   \n",
       "TCGA-ZS-A9CD-01A-11R-A37K-07  40.5710  1616.8295  468.0691   942.9001   \n",
       "TCGA-ZS-A9CE-01A-11R-A37K-07  20.7877  4619.2560  695.8425  1242.8884   \n",
       "TCGA-ZS-A9CF-01A-11R-A38B-07  63.4045   793.9352  504.4797  1274.2936   \n",
       "TCGA-ZS-A9CF-02A-11R-A38B-07  33.3745  2132.2621  632.8801  1221.8789   \n",
       "TCGA-ZS-A9CG-01A-11R-A37K-07  49.3573  2696.1440  256.3839  1244.9015   \n",
       "\n",
       "                                   4849       9093     79165    57348  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1090.0798  1720.8014  289.9495  33.8817   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07   736.8162  2790.0444  183.8344  25.6284   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   954.0462  1407.7561  115.6691   0.0000   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07  2207.9677  2332.2545  241.0664  10.0097   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07   833.3333  1542.3012  120.1354   1.6920   \n",
       "...                                 ...        ...       ...      ...   \n",
       "TCGA-ZS-A9CD-01A-11R-A37K-07   670.9241  2818.1818  222.3892  21.7881   \n",
       "TCGA-ZS-A9CE-01A-11R-A37K-07   699.1247  1862.1444  237.9650   3.2823   \n",
       "TCGA-ZS-A9CF-01A-11R-A38B-07   671.2612  2253.6182  132.3225   7.5810   \n",
       "TCGA-ZS-A9CF-02A-11R-A38B-07   597.6514  1804.0791  192.8307   0.6180   \n",
       "TCGA-ZS-A9CG-01A-11R-A37K-07   802.0566  1883.1191  235.8183   0.0000   \n",
       "\n",
       "                                 1113      5165  ...      22826      93974  \\\n",
       "bcr_patient_barcode                              ...                         \n",
       "TCGA-2V-A95S-01A-11R-A37K-07   0.0000  104.9031  ...  1751.4253  3309.3338   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07   2.9571  134.5490  ...  1984.2287  3247.4125   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   0.0000  112.0825  ...  1500.1121  3359.7848   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07   0.0000   40.0387  ...  1516.4659  3189.7501   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  38.0711   65.9898  ...  1544.8393  2682.7411   \n",
       "...                               ...       ...  ...        ...        ...   \n",
       "TCGA-ZS-A9CD-01A-11R-A37K-07  12.0210   85.6499  ...  1625.0939  3519.9098   \n",
       "TCGA-ZS-A9CE-01A-11R-A37K-07   0.0000  101.7505  ...  1514.2232  4958.4245   \n",
       "TCGA-ZS-A9CF-01A-11R-A38B-07   5.5134   94.4176  ...  2202.6189  5642.3156   \n",
       "TCGA-ZS-A9CF-02A-11R-A38B-07   0.0000   84.6724  ...  1734.2398  4358.4672   \n",
       "TCGA-ZS-A9CG-01A-11R-A37K-07   0.0000   46.6153  ...  1417.6521  3505.7412   \n",
       "\n",
       "                                  83667     54797    389072  1421  1420  \\\n",
       "bcr_patient_barcode                                                       \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1008.6333  533.6374  187.6527   0.0   0.0   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  1559.8817  565.3031  324.7905   0.0   0.0   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   720.9146  494.0596  633.0419   0.0   0.0   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07   786.5937  320.3096   69.2336   0.0   0.0   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07   963.6210  438.2403  282.5719   0.0   0.0   \n",
       "...                                 ...       ...       ...   ...   ...   \n",
       "TCGA-ZS-A9CD-01A-11R-A37K-07  1054.0947  293.0128  178.0616   0.0   0.0   \n",
       "TCGA-ZS-A9CE-01A-11R-A37K-07   471.0066  439.2779  235.2298   0.0   0.0   \n",
       "TCGA-ZS-A9CF-01A-11R-A38B-07  1602.3432  809.0972  474.1558   0.0   0.0   \n",
       "TCGA-ZS-A9CF-02A-11R-A38B-07  1465.3894  851.0507  768.8504   0.0   0.0   \n",
       "TCGA-ZS-A9CG-01A-11R-A37K-07  1863.9246  401.0283  228.2776   0.0   0.0   \n",
       "\n",
       "                              219990   79669   257144  \n",
       "bcr_patient_barcode                                    \n",
       "TCGA-2V-A95S-01A-11R-A37K-07     0.0  7.8189  29.9723  \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07     0.0  7.8857   8.8714  \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07     0.0  5.3800  17.0365  \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07     0.0  2.5024   3.3366  \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07     0.0  0.8460   0.0000  \n",
       "...                              ...     ...      ...  \n",
       "TCGA-ZS-A9CD-01A-11R-A37K-07     0.0  0.7513   0.0000  \n",
       "TCGA-ZS-A9CE-01A-11R-A37K-07     0.0  0.5470   1.0941  \n",
       "TCGA-ZS-A9CF-01A-11R-A38B-07     0.0  8.9593   6.2026  \n",
       "TCGA-ZS-A9CF-02A-11R-A38B-07     0.0  1.2361   6.7985  \n",
       "TCGA-ZS-A9CG-01A-11R-A37K-07     0.0  0.6855   1.3710  \n",
       "\n",
       "[423 rows x 19228 columns]"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids = list(map(lambda x: str(int(x)) if type(x) is float else x, list(map_geneid_to_hfnc_symbol.keys())))\n",
    "gene_ids = [gene for gene in gene_ids if gene in lihc_processed.columns] # get intersection of gene ids\n",
    "lihc_processed = lihc_processed[gene_ids]\n",
    "lihc_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDKN1C</th>\n",
       "      <th>KLF13</th>\n",
       "      <th>MPV17L</th>\n",
       "      <th>MARF1</th>\n",
       "      <th>CNOT3</th>\n",
       "      <th>DNAJA3</th>\n",
       "      <th>LENG1</th>\n",
       "      <th>TTYH1</th>\n",
       "      <th>CHGA</th>\n",
       "      <th>PDK3</th>\n",
       "      <th>...</th>\n",
       "      <th>DNAJC8</th>\n",
       "      <th>ATP5IF1</th>\n",
       "      <th>SESN2</th>\n",
       "      <th>MED18</th>\n",
       "      <th>PLEKHM3</th>\n",
       "      <th>CRYGD</th>\n",
       "      <th>CRYGC</th>\n",
       "      <th>OOSP2</th>\n",
       "      <th>C3orf52</th>\n",
       "      <th>GCSAM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>72.3245</td>\n",
       "      <td>768.8549</td>\n",
       "      <td>13.0314</td>\n",
       "      <td>1048.3792</td>\n",
       "      <td>1090.0798</td>\n",
       "      <td>1720.8014</td>\n",
       "      <td>289.9495</td>\n",
       "      <td>33.8817</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>104.9031</td>\n",
       "      <td>...</td>\n",
       "      <td>1751.4253</td>\n",
       "      <td>3309.3338</td>\n",
       "      <td>1008.6333</td>\n",
       "      <td>533.6374</td>\n",
       "      <td>187.6527</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.8189</td>\n",
       "      <td>29.9723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>41.3997</td>\n",
       "      <td>4116.8063</td>\n",
       "      <td>690.4879</td>\n",
       "      <td>1209.9556</td>\n",
       "      <td>736.8162</td>\n",
       "      <td>2790.0444</td>\n",
       "      <td>183.8344</td>\n",
       "      <td>25.6284</td>\n",
       "      <td>2.9571</td>\n",
       "      <td>134.5490</td>\n",
       "      <td>...</td>\n",
       "      <td>1984.2287</td>\n",
       "      <td>3247.4125</td>\n",
       "      <td>1559.8817</td>\n",
       "      <td>565.3031</td>\n",
       "      <td>324.7905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.8857</td>\n",
       "      <td>8.8714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GT-01A-11R-A38B-07</td>\n",
       "      <td>48.4196</td>\n",
       "      <td>3557.0500</td>\n",
       "      <td>642.0085</td>\n",
       "      <td>1310.0202</td>\n",
       "      <td>954.0462</td>\n",
       "      <td>1407.7561</td>\n",
       "      <td>115.6691</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>112.0825</td>\n",
       "      <td>...</td>\n",
       "      <td>1500.1121</td>\n",
       "      <td>3359.7848</td>\n",
       "      <td>720.9146</td>\n",
       "      <td>494.0596</td>\n",
       "      <td>633.0419</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.3800</td>\n",
       "      <td>17.0365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GU-01A-11R-A38B-07</td>\n",
       "      <td>54.2191</td>\n",
       "      <td>2182.9435</td>\n",
       "      <td>862.5004</td>\n",
       "      <td>1270.3947</td>\n",
       "      <td>2207.9677</td>\n",
       "      <td>2332.2545</td>\n",
       "      <td>241.0664</td>\n",
       "      <td>10.0097</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>40.0387</td>\n",
       "      <td>...</td>\n",
       "      <td>1516.4659</td>\n",
       "      <td>3189.7501</td>\n",
       "      <td>786.5937</td>\n",
       "      <td>320.3096</td>\n",
       "      <td>69.2336</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.5024</td>\n",
       "      <td>3.3366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GV-01A-11R-A38B-07</td>\n",
       "      <td>45.6853</td>\n",
       "      <td>2036.3790</td>\n",
       "      <td>296.9543</td>\n",
       "      <td>1212.3519</td>\n",
       "      <td>833.3333</td>\n",
       "      <td>1542.3012</td>\n",
       "      <td>120.1354</td>\n",
       "      <td>1.6920</td>\n",
       "      <td>38.0711</td>\n",
       "      <td>65.9898</td>\n",
       "      <td>...</td>\n",
       "      <td>1544.8393</td>\n",
       "      <td>2682.7411</td>\n",
       "      <td>963.6210</td>\n",
       "      <td>438.2403</td>\n",
       "      <td>282.5719</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8460</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 19228 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CDKN1C      KLF13    MPV17L      MARF1  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  72.3245   768.8549   13.0314  1048.3792   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  41.3997  4116.8063  690.4879  1209.9556   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07  48.4196  3557.0500  642.0085  1310.0202   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07  54.2191  2182.9435  862.5004  1270.3947   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  45.6853  2036.3790  296.9543  1212.3519   \n",
       "\n",
       "                                  CNOT3     DNAJA3     LENG1    TTYH1  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1090.0798  1720.8014  289.9495  33.8817   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07   736.8162  2790.0444  183.8344  25.6284   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   954.0462  1407.7561  115.6691   0.0000   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07  2207.9677  2332.2545  241.0664  10.0097   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07   833.3333  1542.3012  120.1354   1.6920   \n",
       "\n",
       "                                 CHGA      PDK3  ...     DNAJC8    ATP5IF1  \\\n",
       "bcr_patient_barcode                              ...                         \n",
       "TCGA-2V-A95S-01A-11R-A37K-07   0.0000  104.9031  ...  1751.4253  3309.3338   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07   2.9571  134.5490  ...  1984.2287  3247.4125   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   0.0000  112.0825  ...  1500.1121  3359.7848   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07   0.0000   40.0387  ...  1516.4659  3189.7501   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  38.0711   65.9898  ...  1544.8393  2682.7411   \n",
       "\n",
       "                                  SESN2     MED18   PLEKHM3  CRYGD  CRYGC  \\\n",
       "bcr_patient_barcode                                                         \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1008.6333  533.6374  187.6527    0.0    0.0   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  1559.8817  565.3031  324.7905    0.0    0.0   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07   720.9146  494.0596  633.0419    0.0    0.0   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07   786.5937  320.3096   69.2336    0.0    0.0   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07   963.6210  438.2403  282.5719    0.0    0.0   \n",
       "\n",
       "                              OOSP2  C3orf52    GCSAM  \n",
       "bcr_patient_barcode                                    \n",
       "TCGA-2V-A95S-01A-11R-A37K-07    0.0   7.8189  29.9723  \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07    0.0   7.8857   8.8714  \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07    0.0   5.3800  17.0365  \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07    0.0   2.5024   3.3366  \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07    0.0   0.8460   0.0000  \n",
       "\n",
       "[5 rows x 19228 columns]"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed.columns = pd.Series(lihc_processed.T.index).map(map_geneid_to_hfnc_symbol).values\n",
    "lihc_processed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre-filtering RNA-seq data\n",
    "\n",
    "Our count matrix contains many rows with only zeros, and additionally many rows with only a few fragments total. In order to reduce the size of the object, and to increase the speed of our functions, we can remove the rows that have no or nearly no information about the amount of gene expression. Here we remove columns of the count matrix that have no counts, or only a single count across all samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(423, 17180)"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed = lihc_processed.T[lihc_processed.T.sum(axis=1) > len(lihc_processed)].T\n",
    "lihc_processed.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Survey of clinical characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## TODO: Conduct survey of clinical characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['bcr_patient_uuid', 'bcr_patient_barcode', 'form_completion_date',\n",
       "       'prospective_collection', 'retrospective_collection', 'gender',\n",
       "       'height_cm_at_diagnosis', 'weight_kg_at_diagnosis', 'race', 'ethnicity',\n",
       "       'history_other_malignancy', 'history_neoadjuvant_treatment',\n",
       "       'tumor_status', 'vital_status', 'family_history_cancer_indicator',\n",
       "       'family_history_cancer_number_relatives',\n",
       "       'history_hepato_carcinoma_risk_factors',\n",
       "       'other_hepato_carcinoma_risk_factors', 'viral_hepatitis_serology',\n",
       "       'radiation_treatment_adjuvant', 'pharmaceutical_tx_adjuvant',\n",
       "       'ablation_embolization_tx_adjuvant', 'histologic_diagnosis',\n",
       "       'definitive_surgical_procedure', 'tumor_grade', 'residual_tumor',\n",
       "       'ajcc_staging_edition', 'ajcc_tumor_pathologic_pt',\n",
       "       'ajcc_nodes_pathologic_pn', 'ajcc_metastasis_pathologic_pm',\n",
       "       'ajcc_pathologic_tumor_stage', 'vascular_invasion',\n",
       "       'child_pugh_classification', 'alpha_fetoprotien_at_procurement',\n",
       "       'alpha_fetoprotien_norm_range_lower',\n",
       "       'alpha_fetoprotien_norm_range_upper', 'platelet_count_preresection',\n",
       "       'platelet_norm_range_lower', 'platelet_norm_range_upper',\n",
       "       'prothrombin_time_INR_at_procurement',\n",
       "       'prothrom_time_INR_norm_range_lower',\n",
       "       'prothrombin_time_norm_range_upper', 'serum_albumin_preresection',\n",
       "       'serum_albumin_norm_range_lower', 'serum_albumin_norm_range_upper',\n",
       "       'bilirubin_total_norm_range_upper', 'bilirubin_total_norm_range_lower',\n",
       "       'bilirubin_total', 'creatinine_level_preresection',\n",
       "       'creatinine_norm_range_lower', 'creatinine_norm_range_upper',\n",
       "       'ishak_fibrosis_score', 'hepatic_inflammation_adj_tissue', 'ecog_score',\n",
       "       'new_tumor_event_dx_indicator', 'age_at_diagnosis', 'birth_days_to',\n",
       "       'clinical_M', 'clinical_N', 'clinical_T', 'clinical_stage',\n",
       "       'days_to_initial_pathologic_diagnosis', 'death_days_to', 'disease_code',\n",
       "       'extranodal_involvement', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site',\n",
       "       'informed_consent_verified', 'last_contact_days_to', 'patient_id',\n",
       "       'project_code', 'stage_other', 'tissue_source_site',\n",
       "       'tumor_tissue_site', 'year_of_initial_pathologic_diagnosis'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinpat = pd.read_csv('/Users/mattheweng/Downloads/TCGA-LIHC Clinical Data 1516/nationwidechildrens.org_clinical_patient_lihc.txt', sep='\\t')\n",
    "clinpat.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of these clinical features will be more relevant than others. Try exploring some of these to see which might be worth further delving into.\n",
    "\n",
    "We can get a better idea of our patient population by viewing demographic characteristics and features representative of the health of the patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>tumor_grade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-BC-4073</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-BC-A10W</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-BC-A10X</td>\n",
       "      <td>G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>TCGA-BC-A10Y</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>TCGA-BC-A10Z</td>\n",
       "      <td>G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>62</td>\n",
       "      <td>TCGA-DD-A4NP</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63</td>\n",
       "      <td>TCGA-DD-A4NQ</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>64</td>\n",
       "      <td>TCGA-DD-A4NR</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>65</td>\n",
       "      <td>TCGA-DD-A4NS</td>\n",
       "      <td>G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>66</td>\n",
       "      <td>TCGA-DD-A4NV</td>\n",
       "      <td>G1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bcr_patient_barcode tumor_grade\n",
       "2         TCGA-BC-4073          G3\n",
       "3         TCGA-BC-A10W          G3\n",
       "4         TCGA-BC-A10X          G2\n",
       "5         TCGA-BC-A10Y          G3\n",
       "6         TCGA-BC-A10Z          G2\n",
       "..                 ...         ...\n",
       "62        TCGA-DD-A4NP          G3\n",
       "63        TCGA-DD-A4NQ          G3\n",
       "64        TCGA-DD-A4NR          G3\n",
       "65        TCGA-DD-A4NS          G2\n",
       "66        TCGA-DD-A4NV          G1\n",
       "\n",
       "[65 rows x 2 columns]"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinpat.loc[2:, ['bcr_patient_barcode', 'tumor_grade']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# TODO: plot histogram"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The patient barcodes need to be matched in order to make comparisons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDKN1C</th>\n",
       "      <th>KLF13</th>\n",
       "      <th>MPV17L</th>\n",
       "      <th>MARF1</th>\n",
       "      <th>CNOT3</th>\n",
       "      <th>DNAJA3</th>\n",
       "      <th>LENG1</th>\n",
       "      <th>TTYH1</th>\n",
       "      <th>CHGA</th>\n",
       "      <th>PDK3</th>\n",
       "      <th>...</th>\n",
       "      <th>PRSS21</th>\n",
       "      <th>CTAGE1</th>\n",
       "      <th>FLII</th>\n",
       "      <th>DNAJC8</th>\n",
       "      <th>ATP5IF1</th>\n",
       "      <th>SESN2</th>\n",
       "      <th>MED18</th>\n",
       "      <th>PLEKHM3</th>\n",
       "      <th>C3orf52</th>\n",
       "      <th>GCSAM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>72.3245</td>\n",
       "      <td>768.8549</td>\n",
       "      <td>13.0314</td>\n",
       "      <td>1048.3792</td>\n",
       "      <td>1090.0798</td>\n",
       "      <td>1720.8014</td>\n",
       "      <td>289.9495</td>\n",
       "      <td>33.8817</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>104.9031</td>\n",
       "      <td>...</td>\n",
       "      <td>14.3346</td>\n",
       "      <td>12.3799</td>\n",
       "      <td>2804.3655</td>\n",
       "      <td>1751.4253</td>\n",
       "      <td>3309.3338</td>\n",
       "      <td>1008.6333</td>\n",
       "      <td>533.6374</td>\n",
       "      <td>187.6527</td>\n",
       "      <td>7.8189</td>\n",
       "      <td>29.9723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>41.3997</td>\n",
       "      <td>4116.8063</td>\n",
       "      <td>690.4879</td>\n",
       "      <td>1209.9556</td>\n",
       "      <td>736.8162</td>\n",
       "      <td>2790.0444</td>\n",
       "      <td>183.8344</td>\n",
       "      <td>25.6284</td>\n",
       "      <td>2.9571</td>\n",
       "      <td>134.5490</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>14.2928</td>\n",
       "      <td>2189.7486</td>\n",
       "      <td>1984.2287</td>\n",
       "      <td>3247.4125</td>\n",
       "      <td>1559.8817</td>\n",
       "      <td>565.3031</td>\n",
       "      <td>324.7905</td>\n",
       "      <td>7.8857</td>\n",
       "      <td>8.8714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 17180 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CDKN1C      KLF13    MPV17L      MARF1  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  72.3245   768.8549   13.0314  1048.3792   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  41.3997  4116.8063  690.4879  1209.9556   \n",
       "\n",
       "                                  CNOT3     DNAJA3     LENG1    TTYH1    CHGA  \\\n",
       "bcr_patient_barcode                                                             \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1090.0798  1720.8014  289.9495  33.8817  0.0000   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07   736.8162  2790.0444  183.8344  25.6284  2.9571   \n",
       "\n",
       "                                  PDK3  ...   PRSS21   CTAGE1       FLII  \\\n",
       "bcr_patient_barcode                     ...                                \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  104.9031  ...  14.3346  12.3799  2804.3655   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  134.5490  ...   0.0000  14.2928  2189.7486   \n",
       "\n",
       "                                 DNAJC8    ATP5IF1      SESN2     MED18  \\\n",
       "bcr_patient_barcode                                                       \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  1751.4253  3309.3338  1008.6333  533.6374   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  1984.2287  3247.4125  1559.8817  565.3031   \n",
       "\n",
       "                               PLEKHM3  C3orf52    GCSAM  \n",
       "bcr_patient_barcode                                       \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  187.6527   7.8189  29.9723  \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  324.7905   7.8857   8.8714  \n",
       "\n",
       "[2 rows x 17180 columns]"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_uuid</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>form_completion_date</th>\n",
       "      <th>prospective_collection</th>\n",
       "      <th>retrospective_collection</th>\n",
       "      <th>gender</th>\n",
       "      <th>height_cm_at_diagnosis</th>\n",
       "      <th>weight_kg_at_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>icd_o_3_histology</th>\n",
       "      <th>icd_o_3_site</th>\n",
       "      <th>informed_consent_verified</th>\n",
       "      <th>last_contact_days_to</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>project_code</th>\n",
       "      <th>stage_other</th>\n",
       "      <th>tissue_source_site</th>\n",
       "      <th>tumor_tissue_site</th>\n",
       "      <th>year_of_initial_pathologic_diagnosis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>bcr_patient_uuid</td>\n",
       "      <td>bcr_patient_barcode</td>\n",
       "      <td>form_completion_date</td>\n",
       "      <td>tissue_prospective_collection_indicator</td>\n",
       "      <td>tissue_retrospective_collection_indicator</td>\n",
       "      <td>gender</td>\n",
       "      <td>height</td>\n",
       "      <td>weight</td>\n",
       "      <td>race</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>...</td>\n",
       "      <td>icd_o_3_histology</td>\n",
       "      <td>icd_o_3_site</td>\n",
       "      <td>informed_consent_verified</td>\n",
       "      <td>days_to_last_followup</td>\n",
       "      <td>patient_id</td>\n",
       "      <td>project_code</td>\n",
       "      <td>stage_other</td>\n",
       "      <td>tissue_source_site</td>\n",
       "      <td>tumor_tissue_site</td>\n",
       "      <td>year_of_initial_pathologic_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:2673794</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:3088492</td>\n",
       "      <td>CDE_ID:3088528</td>\n",
       "      <td>CDE_ID:2200604</td>\n",
       "      <td>CDE_ID:649</td>\n",
       "      <td>CDE_ID:651</td>\n",
       "      <td>CDE_ID:2192199</td>\n",
       "      <td>CDE_ID:2192217</td>\n",
       "      <td>...</td>\n",
       "      <td>CDE_ID:3226275</td>\n",
       "      <td>CDE_ID:3226281</td>\n",
       "      <td>CDE_ID:3288361</td>\n",
       "      <td>CDE_ID:3008273</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:2007104</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:3427536</td>\n",
       "      <td>CDE_ID:2896960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bcr_patient_uuid  bcr_patient_barcode  form_completion_date  \\\n",
       "0  bcr_patient_uuid  bcr_patient_barcode  form_completion_date   \n",
       "1           CDE_ID:       CDE_ID:2673794               CDE_ID:   \n",
       "\n",
       "                    prospective_collection  \\\n",
       "0  tissue_prospective_collection_indicator   \n",
       "1                           CDE_ID:3088492   \n",
       "\n",
       "                    retrospective_collection          gender  \\\n",
       "0  tissue_retrospective_collection_indicator          gender   \n",
       "1                             CDE_ID:3088528  CDE_ID:2200604   \n",
       "\n",
       "  height_cm_at_diagnosis weight_kg_at_diagnosis            race  \\\n",
       "0                 height                 weight            race   \n",
       "1             CDE_ID:649             CDE_ID:651  CDE_ID:2192199   \n",
       "\n",
       "        ethnicity  ...  icd_o_3_histology    icd_o_3_site  \\\n",
       "0       ethnicity  ...  icd_o_3_histology    icd_o_3_site   \n",
       "1  CDE_ID:2192217  ...     CDE_ID:3226275  CDE_ID:3226281   \n",
       "\n",
       "   informed_consent_verified   last_contact_days_to  patient_id  project_code  \\\n",
       "0  informed_consent_verified  days_to_last_followup  patient_id  project_code   \n",
       "1             CDE_ID:3288361         CDE_ID:3008273     CDE_ID:       CDE_ID:   \n",
       "\n",
       "      stage_other  tissue_source_site  tumor_tissue_site  \\\n",
       "0     stage_other  tissue_source_site  tumor_tissue_site   \n",
       "1  CDE_ID:2007104             CDE_ID:     CDE_ID:3427536   \n",
       "\n",
       "   year_of_initial_pathologic_diagnosis  \n",
       "0  year_of_initial_pathologic_diagnosis  \n",
       "1                        CDE_ID:2896960  \n",
       "\n",
       "[2 rows x 76 columns]"
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinpat.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More information about how patient barcodes are defined can be found [here](https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/). For simplification, the alphanumeric combination to the right of the 2nd hyphen is the code that we're interested in. We'll need to transform these indices for both dataframes to combine these together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform clinical dataset\n",
    "clinpat.loc[2:, 'bcr_patient_barcode'] = clinpat.loc[2:, 'bcr_patient_barcode'].str.split('-').apply(lambda x: x[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform LIHC RNA-seq dataset\n",
    "lihc_processed.index = list(map(lambda x: x[2], lihc.index.str.split('-')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_uuid</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>form_completion_date</th>\n",
       "      <th>prospective_collection</th>\n",
       "      <th>retrospective_collection</th>\n",
       "      <th>gender</th>\n",
       "      <th>height_cm_at_diagnosis</th>\n",
       "      <th>weight_kg_at_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>icd_o_3_histology</th>\n",
       "      <th>icd_o_3_site</th>\n",
       "      <th>informed_consent_verified</th>\n",
       "      <th>last_contact_days_to</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>project_code</th>\n",
       "      <th>stage_other</th>\n",
       "      <th>tissue_source_site</th>\n",
       "      <th>tumor_tissue_site</th>\n",
       "      <th>year_of_initial_pathologic_diagnosis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>bcr_patient_uuid</td>\n",
       "      <td>bcr_patient_barcode</td>\n",
       "      <td>form_completion_date</td>\n",
       "      <td>tissue_prospective_collection_indicator</td>\n",
       "      <td>tissue_retrospective_collection_indicator</td>\n",
       "      <td>gender</td>\n",
       "      <td>height</td>\n",
       "      <td>weight</td>\n",
       "      <td>race</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>...</td>\n",
       "      <td>icd_o_3_histology</td>\n",
       "      <td>icd_o_3_site</td>\n",
       "      <td>informed_consent_verified</td>\n",
       "      <td>days_to_last_followup</td>\n",
       "      <td>patient_id</td>\n",
       "      <td>project_code</td>\n",
       "      <td>stage_other</td>\n",
       "      <td>tissue_source_site</td>\n",
       "      <td>tumor_tissue_site</td>\n",
       "      <td>year_of_initial_pathologic_diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:2673794</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:3088492</td>\n",
       "      <td>CDE_ID:3088528</td>\n",
       "      <td>CDE_ID:2200604</td>\n",
       "      <td>CDE_ID:649</td>\n",
       "      <td>CDE_ID:651</td>\n",
       "      <td>CDE_ID:2192199</td>\n",
       "      <td>CDE_ID:2192217</td>\n",
       "      <td>...</td>\n",
       "      <td>CDE_ID:3226275</td>\n",
       "      <td>CDE_ID:3226281</td>\n",
       "      <td>CDE_ID:3288361</td>\n",
       "      <td>CDE_ID:3008273</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:2007104</td>\n",
       "      <td>CDE_ID:</td>\n",
       "      <td>CDE_ID:3427536</td>\n",
       "      <td>CDE_ID:2896960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>00f209c3-041a-4c6e-8b0f-6383eb3c85fc</td>\n",
       "      <td>4073</td>\n",
       "      <td>2010-12-20</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>MALE</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>...</td>\n",
       "      <td>8170/3</td>\n",
       "      <td>C22.0</td>\n",
       "      <td>YES</td>\n",
       "      <td>352</td>\n",
       "      <td>4073</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>BC</td>\n",
       "      <td>Liver</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>39baf24a-4032-4c23-b3fe-2c50df1881f6</td>\n",
       "      <td>A10W</td>\n",
       "      <td>2011-3-31</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>MALE</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>71</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>...</td>\n",
       "      <td>8170/3</td>\n",
       "      <td>C22.0</td>\n",
       "      <td>YES</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>A10W</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>BC</td>\n",
       "      <td>Liver</td>\n",
       "      <td>2002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>3dc48ea9-a3c6-437f-af5a-88feb300cc5f</td>\n",
       "      <td>A10X</td>\n",
       "      <td>2011-3-31</td>\n",
       "      <td>NO</td>\n",
       "      <td>YES</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>65</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>...</td>\n",
       "      <td>8170/3</td>\n",
       "      <td>C22.0</td>\n",
       "      <td>YES</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>A10X</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>BC</td>\n",
       "      <td>Liver</td>\n",
       "      <td>2002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       bcr_patient_uuid  bcr_patient_barcode  \\\n",
       "0                      bcr_patient_uuid  bcr_patient_barcode   \n",
       "1                               CDE_ID:       CDE_ID:2673794   \n",
       "2  00f209c3-041a-4c6e-8b0f-6383eb3c85fc                 4073   \n",
       "3  39baf24a-4032-4c23-b3fe-2c50df1881f6                 A10W   \n",
       "4  3dc48ea9-a3c6-437f-af5a-88feb300cc5f                 A10X   \n",
       "\n",
       "   form_completion_date                   prospective_collection  \\\n",
       "0  form_completion_date  tissue_prospective_collection_indicator   \n",
       "1               CDE_ID:                           CDE_ID:3088492   \n",
       "2            2010-12-20                                       NO   \n",
       "3             2011-3-31                                       NO   \n",
       "4             2011-3-31                                       NO   \n",
       "\n",
       "                    retrospective_collection          gender  \\\n",
       "0  tissue_retrospective_collection_indicator          gender   \n",
       "1                             CDE_ID:3088528  CDE_ID:2200604   \n",
       "2                                        YES            MALE   \n",
       "3                                        YES            MALE   \n",
       "4                                        YES          FEMALE   \n",
       "\n",
       "  height_cm_at_diagnosis weight_kg_at_diagnosis            race  \\\n",
       "0                 height                 weight            race   \n",
       "1             CDE_ID:649             CDE_ID:651  CDE_ID:2192199   \n",
       "2        [Not Available]        [Not Available]           WHITE   \n",
       "3        [Not Available]                     71           ASIAN   \n",
       "4        [Not Available]                     65           WHITE   \n",
       "\n",
       "                ethnicity  ...  icd_o_3_histology    icd_o_3_site  \\\n",
       "0               ethnicity  ...  icd_o_3_histology    icd_o_3_site   \n",
       "1          CDE_ID:2192217  ...     CDE_ID:3226275  CDE_ID:3226281   \n",
       "2  NOT HISPANIC OR LATINO  ...             8170/3           C22.0   \n",
       "3  NOT HISPANIC OR LATINO  ...             8170/3           C22.0   \n",
       "4  NOT HISPANIC OR LATINO  ...             8170/3           C22.0   \n",
       "\n",
       "   informed_consent_verified   last_contact_days_to  patient_id  \\\n",
       "0  informed_consent_verified  days_to_last_followup  patient_id   \n",
       "1             CDE_ID:3288361         CDE_ID:3008273     CDE_ID:   \n",
       "2                        YES                    352        4073   \n",
       "3                        YES        [Not Available]        A10W   \n",
       "4                        YES        [Not Available]        A10X   \n",
       "\n",
       "      project_code      stage_other  tissue_source_site  tumor_tissue_site  \\\n",
       "0     project_code      stage_other  tissue_source_site  tumor_tissue_site   \n",
       "1          CDE_ID:   CDE_ID:2007104             CDE_ID:     CDE_ID:3427536   \n",
       "2  [Not Available]  [Not Available]                  BC              Liver   \n",
       "3  [Not Available]  [Not Available]                  BC              Liver   \n",
       "4  [Not Available]  [Not Available]                  BC              Liver   \n",
       "\n",
       "   year_of_initial_pathologic_diagnosis  \n",
       "0  year_of_initial_pathologic_diagnosis  \n",
       "1                        CDE_ID:2896960  \n",
       "2                                  2009  \n",
       "3                                  2002  \n",
       "4                                  2002  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinpat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDKN1C</th>\n",
       "      <th>KLF13</th>\n",
       "      <th>MPV17L</th>\n",
       "      <th>MARF1</th>\n",
       "      <th>CNOT3</th>\n",
       "      <th>DNAJA3</th>\n",
       "      <th>LENG1</th>\n",
       "      <th>TTYH1</th>\n",
       "      <th>CHGA</th>\n",
       "      <th>PDK3</th>\n",
       "      <th>...</th>\n",
       "      <th>CTAGE1</th>\n",
       "      <th>FLII</th>\n",
       "      <th>DNAJC8</th>\n",
       "      <th>ATP5IF1</th>\n",
       "      <th>SESN2</th>\n",
       "      <th>MED18</th>\n",
       "      <th>PLEKHM3</th>\n",
       "      <th>C3orf52</th>\n",
       "      <th>GCSAM</th>\n",
       "      <th>tumor_grade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4073</td>\n",
       "      <td>32.0220</td>\n",
       "      <td>1742.4520</td>\n",
       "      <td>487.6487</td>\n",
       "      <td>2077.3102</td>\n",
       "      <td>865.5078</td>\n",
       "      <td>1749.7713</td>\n",
       "      <td>123.0558</td>\n",
       "      <td>0.4575</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>237.4199</td>\n",
       "      <td>...</td>\n",
       "      <td>11.8939</td>\n",
       "      <td>2357.2736</td>\n",
       "      <td>2216.8344</td>\n",
       "      <td>1673.8335</td>\n",
       "      <td>663.7694</td>\n",
       "      <td>564.0439</td>\n",
       "      <td>222.3239</td>\n",
       "      <td>111.6194</td>\n",
       "      <td>13.2662</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A10W</td>\n",
       "      <td>186.4261</td>\n",
       "      <td>390.8935</td>\n",
       "      <td>268.0412</td>\n",
       "      <td>1195.8763</td>\n",
       "      <td>457.9038</td>\n",
       "      <td>2292.0962</td>\n",
       "      <td>224.2268</td>\n",
       "      <td>34.3643</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>167.5258</td>\n",
       "      <td>...</td>\n",
       "      <td>5.1546</td>\n",
       "      <td>1908.0756</td>\n",
       "      <td>1440.7216</td>\n",
       "      <td>4452.7491</td>\n",
       "      <td>414.9485</td>\n",
       "      <td>268.9003</td>\n",
       "      <td>207.9038</td>\n",
       "      <td>29.2096</td>\n",
       "      <td>12.8866</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A10W</td>\n",
       "      <td>18.8844</td>\n",
       "      <td>1219.3383</td>\n",
       "      <td>178.1056</td>\n",
       "      <td>1268.2156</td>\n",
       "      <td>839.0588</td>\n",
       "      <td>2781.1875</td>\n",
       "      <td>91.4596</td>\n",
       "      <td>0.3703</td>\n",
       "      <td>2.2217</td>\n",
       "      <td>115.1577</td>\n",
       "      <td>...</td>\n",
       "      <td>5.5542</td>\n",
       "      <td>3869.0757</td>\n",
       "      <td>1061.2280</td>\n",
       "      <td>718.7172</td>\n",
       "      <td>269.9355</td>\n",
       "      <td>139.5963</td>\n",
       "      <td>352.1381</td>\n",
       "      <td>106.6412</td>\n",
       "      <td>8.8868</td>\n",
       "      <td>G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A10X</td>\n",
       "      <td>163.9851</td>\n",
       "      <td>324.8762</td>\n",
       "      <td>563.1188</td>\n",
       "      <td>1310.6436</td>\n",
       "      <td>532.7970</td>\n",
       "      <td>1969.6782</td>\n",
       "      <td>175.1238</td>\n",
       "      <td>12.9950</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>129.9505</td>\n",
       "      <td>...</td>\n",
       "      <td>9.2822</td>\n",
       "      <td>1659.0347</td>\n",
       "      <td>1670.1733</td>\n",
       "      <td>3089.1089</td>\n",
       "      <td>613.8614</td>\n",
       "      <td>345.9158</td>\n",
       "      <td>290.8416</td>\n",
       "      <td>25.3713</td>\n",
       "      <td>9.2822</td>\n",
       "      <td>G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A10X</td>\n",
       "      <td>268.0892</td>\n",
       "      <td>598.5501</td>\n",
       "      <td>560.2517</td>\n",
       "      <td>1238.6814</td>\n",
       "      <td>589.7962</td>\n",
       "      <td>1952.6741</td>\n",
       "      <td>204.6232</td>\n",
       "      <td>30.6388</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>99.5760</td>\n",
       "      <td>...</td>\n",
       "      <td>10.9424</td>\n",
       "      <td>2228.9700</td>\n",
       "      <td>1414.3072</td>\n",
       "      <td>2164.9569</td>\n",
       "      <td>568.4585</td>\n",
       "      <td>277.9374</td>\n",
       "      <td>200.7933</td>\n",
       "      <td>10.9424</td>\n",
       "      <td>9.8482</td>\n",
       "      <td>G2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 17181 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CDKN1C      KLF13    MPV17L      MARF1     CNOT3     DNAJA3     LENG1  \\\n",
       "4073   32.0220  1742.4520  487.6487  2077.3102  865.5078  1749.7713  123.0558   \n",
       "A10W  186.4261   390.8935  268.0412  1195.8763  457.9038  2292.0962  224.2268   \n",
       "A10W   18.8844  1219.3383  178.1056  1268.2156  839.0588  2781.1875   91.4596   \n",
       "A10X  163.9851   324.8762  563.1188  1310.6436  532.7970  1969.6782  175.1238   \n",
       "A10X  268.0892   598.5501  560.2517  1238.6814  589.7962  1952.6741  204.6232   \n",
       "\n",
       "        TTYH1    CHGA      PDK3  ...   CTAGE1       FLII     DNAJC8  \\\n",
       "4073   0.4575  0.0000  237.4199  ...  11.8939  2357.2736  2216.8344   \n",
       "A10W  34.3643  0.0000  167.5258  ...   5.1546  1908.0756  1440.7216   \n",
       "A10W   0.3703  2.2217  115.1577  ...   5.5542  3869.0757  1061.2280   \n",
       "A10X  12.9950  0.0000  129.9505  ...   9.2822  1659.0347  1670.1733   \n",
       "A10X  30.6388  0.0000   99.5760  ...  10.9424  2228.9700  1414.3072   \n",
       "\n",
       "        ATP5IF1     SESN2     MED18   PLEKHM3   C3orf52    GCSAM  tumor_grade  \n",
       "4073  1673.8335  663.7694  564.0439  222.3239  111.6194  13.2662           G3  \n",
       "A10W  4452.7491  414.9485  268.9003  207.9038   29.2096  12.8866           G3  \n",
       "A10W   718.7172  269.9355  139.5963  352.1381  106.6412   8.8868           G3  \n",
       "A10X  3089.1089  613.8614  345.9158  290.8416   25.3713   9.2822           G2  \n",
       "A10X  2164.9569  568.4585  277.9374  200.7933   10.9424   9.8482           G2  \n",
       "\n",
       "[5 rows x 17181 columns]"
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge data for clincal and RNA-seq dataset\n",
    "\n",
    "lihc_processed = lihc_processed.join(clinpat[['bcr_patient_barcode', 'tumor_grade']].set_index('bcr_patient_barcode')).dropna()\n",
    "lihc_processed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Label and Split data based on recurrence? drug?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Days to death"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clinpat['death_days_to'].loc[clinpat['death_days_to'] == '[Not Applicable]'] = '-1'\n",
    "clinpat.loc[2:,'death_days_to'] = clinpat.loc[2:,'death_days_to'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAr8AAAF1CAYAAADhgoKhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAb50lEQVR4nO3df7htdV0n8PdHwB8jjqCcCAW7RjZmlujcSLOnMU1DrQFnnIIxwxlnKM3Gfoc+TWE/THtKs6eywTApfyCDmo7aD0ZxfCzDLooIogMqDtIVriICmRbwmT/2uno63nPPuefsfc/lfl+v59nPXj/3+qzvXnff91n7u9eq7g4AAIzgLltdAAAA7C/CLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhF9hSVXVFVT1mq+vYSlX1lKq6tqpuraqHr2P5d1XVf9kftS1CVT2jqt5zZ3tt4OAg/AILU1XXVNX3rJj2z8JJd39zd79rjdfZVlVdVYcuqNSt9ptJntPdh3f3B7a6mJWq6qyqevVW17HSAMcFsADCLzC8AyA8fV2SK7a4BoAhCL/Allp+driqTqyqHVV1c1VdX1UvmRZ79/R809Q14FFVdZeq+oWq+mRV3VBVf1xV9172uj88zftsVf33Fds5q6ouqKpXV9XNSZ4xbfu9VXVTVe2sqt+tqrsue72uqmdX1VVVdUtV/UpVHV9Vfz3Ve/7y5Vfs4x5rraq7VdWtSQ5J8sGq+tgq6z++qj5SVZ+vqt9NUsvmHV9V75z28zNV9ZqqOmKa97NV9YYVr/U7VfWyafgZVfXxaX8+UVVP28O2T0ry/CQ/OLX9B6fp96uqt1TVjVV1dVX91728x/edlr25qt6X5PgV8x9cVRdOr/XRqvqBZfOeXFUfmNa9tqrOWrbqVx0Xy9b7zar63LRfT1ytNmA8wi9wIHlZkpd197/MLCCdP03/run5iKlrwHuTPGN6fHeSr09yeJLfTZKqekiS30/ytCTHJLl3kvuv2NbJSS5IckSS1yS5PclPJjkqyaOSPC7Js1es871J/nWSRyb5uSRnJ/mhJMcleWiS01bZrz3W2t1f6u7Dp2Ue1t3Hr1yxqo5K8sYkvzDV9rEkj16+SJJfT3K/JN801XLWNO/VSU5aFoYPTXJqkj+uqnsm+Z0kT+zueyX5jiSXrtx+d/95khcmef3U9g+bZp2X5FPTdp+a5IVV9dhV9v/3knwxs/fiP0+P3ft3zyQXJnltkq+Z6vv96T1Mkr9P8sOZvU9PTvKsqjplmren4yJJvj3JR6f2+o0k51TVl/9gAMYm/AKL9qfT2dSbquqmzELpav4pyTdU1VHdfWt3/81eln1akpd098e7+9Ykz0ty6hTwnprkf3X3e7r7H5P8YpJesf57u/tPu/uO7v6H7r6ku/+mu2/r7muS/I8k/2bFOr/R3Td39xVJLk/yl9P2P5/kz5Ks9mO1vdW6licluaK7L+juf0ry20k+vXtmd1/d3RdOQXpXkpfsrru7d2Z2dvQ/TIuflOQz3X3JNH5HkodW1T26e+e0X2uqquMyC+A/391f7O5Lk/xhZiF15bKHJPn3SX6xu/++uy9Pcu6yRb4vyTXd/UdT238gyRt219zd7+ruD03v02VJXpevfl9W+mR3v6K7b5+2dUySo9ezb8DBT/gFFu2U7j5i9yNffTZ1uWcm+cYkH6mqv62q79vLsvdL8sll459McmhmIed+Sa7dPaO7v5DksyvWv3b5SFV9Y1W9tao+PXWFeGFmZw6Xu37Z8D/sYfzw7Nneal3Lyn3p5eNVdXRVnVdV1011v3pF3edmdnY60/OfTK/z90l+MMmPJtlZVW+rqgevo57dNd3Y3bes2KeVZ9eTZCmzfb12xbK7fV2Sb1/xB9LTknzttH/fXlUXVdWuqvr8VO/K92Wl5X8cfGEaXO29AQYj/AIHjO6+qrtPy+zr7xcnuWD6WnzlWdsk+bvMgtNuD0hyW2aBdGeSY3fPqKp7JLnvys2tGH95ko8kedDU7eL5Wda3dpP2VutadmbWlSFJMn19f9yy+S/MbF++Zar7h/LP6/7TJN9aVQ/N7Czra3bP6O6/6O7HZ3Zm9CNJXrFKDSvb6u+S3Keq7rVin67bw7q7MtvX41Ysu9u1Sf7P8j+Qpi4Mz5rmvzbJW5Ic1933TvIHy/ZvT8cFwF4Jv8ABo6p+qKqWuvuOJDdNk+/ILEDdkVl/2d1el+Qnq+qBVXV4vtIv9bbM+vJ+f1V9x/QjtLOydpC9V5Kbk9w6nQF91hrL74u91bqWtyX55qr6d1M3if+W6azosrpvTfL5qrp/kp9dvnJ3fzGz9nhtkvd19/9LvnzG+OTpj4svTa9xxyo1XJ9kW1XdZXrNa5P8dZJfr6q7V9W3ZnbW/qsuhzZ1PXhjkrOq6l9MfXlPX7bIW5N8Y1U9vaoOmx7fVlXftGz/buzuL1bViUn+47J193RcAOyV8AscSE5KcsV0BYSXJTl16o/7hSS/luSvpq/GH5nklZl9hf/uJJ/I7AdVP54kU9/VH8/sR1k7Mwt2N2QW8lbzM5kFq1syOwP6+jnu16q1rqW7P5NZ/9cXZdZ140FJ/mrZIi9I8ogkn88sKL9xDy9zbpJvmWrY7S5Jfiqzs7g3ZtaPdrXA/z+n589W1fun4dOSbJvWf1OSX+ru/73K+s/JrNvBp5O8KskfLdu/W5I8IbMfuv3dtMyLk9xtWuTZSX65qm7JrO/2+cvW3dNxAbBXNes+BnDwms623pRZl4ZPbHU9+1tVPSCzbg1f2903b3U9AFvJmV/goFRV3z99zX7PzO6g9qEk12xtVfvf1FXhp5KcJ/gCzH6BC3AwOjmzr/kryY7MulAM9VXXFPyvz+zqCidtcTkABwTdHgAAGIZuDwAADEP4BQBgGPu1z+9RRx3V27Zt25+bBABgQJdccslnuntp5fT9Gn63bduWHTt27M9NAgAwoKr65J6m6/YAAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADOPQrS5gf9h25tu2uoT97poXPXmrSwAAOOA48wsAwDCEXwAAhiH8AgAwjHWH36o6pKo+UFVvncYfWFUXV9XVVfX6qrrr4soEAIDN25czv89NcuWy8RcneWl3f0OSzyV55jwLAwCAeVtX+K2qY5M8OckfTuOV5LFJLpgWOTfJKYsoEAAA5mW9Z35/O8nPJbljGr9vkpu6+7Zp/FNJ7j/n2gAAYK7WDL9V9X1JbujuSzaygao6o6p2VNWOXbt2beQlAABgLtZz5vfRSf5tVV2T5LzMuju8LMkRVbX7JhnHJrluTyt399ndvb27ty8tLc2hZAAA2Jg1w293P6+7j+3ubUlOTfLO7n5akouSPHVa7PQkb15YlQAAMAebuc7vzyf5qaq6OrM+wOfMpyQAAFiMQ9de5Cu6+11J3jUNfzzJifMvCQAAFsMd3gAAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGGsGX6r6u5V9b6q+mBVXVFVL5imv6qqPlFVl06PExZfLgAAbNyh61jmS0ke2923VtVhSd5TVX82zfvZ7r5gceUBAMD8rBl+u7uT3DqNHjY9epFFAQDAIqyrz29VHVJVlya5IcmF3X3xNOvXquqyqnppVd1tYVUCAMAcrCv8dvft3X1CkmOTnFhVD03yvCQPTvJtSe6T5Of3tG5VnVFVO6pqx65du+ZUNgAA7Lt9utpDd9+U5KIkJ3X3zp75UpI/SnLiKuuc3d3bu3v70tLS5isGAIANWs/VHpaq6ohp+B5JHp/kI1V1zDStkpyS5PJFFgoAAJu1nqs9HJPk3Ko6JLOwfH53v7Wq3llVS0kqyaVJfnSBdQIAwKat52oPlyV5+B6mP3YhFQEAwIK4wxsAAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADCMNcNvVd29qt5XVR+sqiuq6gXT9AdW1cVVdXVVvb6q7rr4cgEAYOPWc+b3S0ke290PS3JCkpOq6pFJXpzkpd39DUk+l+SZiysTAAA2b83w2zO3TqOHTY9O8tgkF0zTz01yykIqBACAOVlXn9+qOqSqLk1yQ5ILk3wsyU3dfdu0yKeS3H8xJQIAwHysK/x29+3dfUKSY5OcmOTB691AVZ1RVTuqaseuXbs2WCYAAGzePl3tobtvSnJRkkclOaKqDp1mHZvkulXWObu7t3f39qWlpU0VCwAAm7Geqz0sVdUR0/A9kjw+yZWZheCnToudnuTNiyoSAADm4dC1F8kxSc6tqkMyC8vnd/dbq+rDSc6rql9N8oEk5yywTgAA2LQ1w293X5bk4XuY/vHM+v8CAMCdgju8AQAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAw1gz/FbVcVV1UVV9uKquqKrnTtPPqqrrqurS6fGkxZcLAAAbd+g6lrktyU939/ur6l5JLqmqC6d5L+3u31xceQAAMD9rht/u3plk5zR8S1VdmeT+iy4MAADmbZ/6/FbVtiQPT3LxNOk5VXVZVb2yqo5cZZ0zqmpHVe3YtWvXpooFAIDNWHf4rarDk7whyU90981JXp7k+CQnZHZm+Lf2tF53n93d27t7+9LS0hxKBgCAjVlX+K2qwzILvq/p7jcmSXdf3923d/cdSV6R5MTFlQkAAJu3nqs9VJJzklzZ3S9ZNv2YZYs9Jcnl8y8PAADmZz1Xe3h0kqcn+VBVXTpNe36S06rqhCSd5JokP7KQCgEAYE7Wc7WH9ySpPcx6+/zLAQCAxXGHNwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBhrht+qOq6qLqqqD1fVFVX13Gn6farqwqq6ano+cvHlAgDAxq3nzO9tSX66ux+S5JFJfqyqHpLkzCTv6O4HJXnHNA4AAAesNcNvd+/s7vdPw7ckuTLJ/ZOcnOTcabFzk5yyqCIBAGAe9qnPb1VtS/LwJBcnObq7d06zPp3k6FXWOaOqdlTVjl27dm2iVAAA2Jx1h9+qOjzJG5L8RHffvHxed3eS3tN63X12d2/v7u1LS0ubKhYAADZjXeG3qg7LLPi+prvfOE2+vqqOmeYfk+SGxZQIAADzsZ6rPVSSc5Jc2d0vWTbrLUlOn4ZPT/Lm+ZcHAADzc+g6lnl0kqcn+VBVXTpNe36SFyU5v6qemeSTSX5gMSUCAMB8rBl+u/s9SWqV2Y+bbzkAALA47vAGAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMY83wW1WvrKobquryZdPOqqrrqurS6fGkxZYJAACbt54zv69KctIepr+0u0+YHm+fb1kAADB/a4bf7n53khv3Qy0AALBQm+nz+5yqumzqFnHkagtV1RlVtaOqduzatWsTmwMAgM3ZaPh9eZLjk5yQZGeS31ptwe4+u7u3d/f2paWlDW4OAAA2b0Pht7uv7+7bu/uOJK9IcuJ8ywIAgPnbUPitqmOWjT4lyeWrLQsAAAeKQ9daoKpel+QxSY6qqk8l+aUkj6mqE5J0kmuS/MgCawQAgLlYM/x292l7mHzOAmoBAICFcoc3AACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBjCLwAAwxB+AQAYhvALAMAwhF8AAIYh/AIAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGGuG36p6ZVXdUFWXL5t2n6q6sKqump6PXGyZAACwees58/uqJCetmHZmknd094OSvGMaBwCAA9qa4be7353kxhWTT05y7jR8bpJT5lwXAADM3Ub7/B7d3Tun4U8nOXq1BavqjKraUVU7du3atcHNAQDA5m36B2/d3Ul6L/PP7u7t3b19aWlps5sDAIAN22j4vb6qjkmS6fmG+ZUEAACLsdHw+5Ykp0/Dpyd583zKAQCAxVnPpc5el+S9Sf5VVX2qqp6Z5EVJHl9VVyX5nmkcAAAOaIeutUB3n7bKrMfNuRYAAFgod3gDAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBiHbnUBLMa2M9+21SXsd9e86MlbXQIAcIBz5hcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDBc6oyDhsu7AQBrceYXAIBhCL8AAAxD+AUAYBib6vNbVdckuSXJ7Ulu6+7t8ygKAAAWYR4/ePvu7v7MHF4HAAAWSrcHAACGsdnw20n+sqouqaoz5lEQAAAsyma7PXxnd19XVV+T5MKq+kh3v3v5AlMoPiNJHvCAB2xycwAAsHGbOvPb3ddNzzckeVOSE/ewzNndvb27ty8tLW1mcwAAsCkbDr9Vdc+qutfu4SRPSHL5vAoDAIB520y3h6OTvKmqdr/Oa7v7z+dSFQAALMCGw293fzzJw+ZYCwAALJRLnQEAMAzhFwCAYQi/AAAMQ/gFAGAYwi8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGFs+PbGAFth25lv2+oS9rtrXvTkrS4B4KDhzC8AAMMQfgEAGIbwCwDAMIRfAACGIfwCADAM4RcAgGEIvwAADMN1fuFObMRr3sLBaMR/y65fzVZx5hcAgGEIvwAADEP4BQBgGMIvAADDEH4BABiG8AsAwDBc6gzgADfiZbA4+Dmux3AgXtLOmV8AAIYh/AIAMAzhFwCAYWwq/FbVSVX10aq6uqrOnFdRAACwCBsOv1V1SJLfS/LEJA9JclpVPWRehQEAwLxt5szviUmu7u6Pd/c/JjkvycnzKQsAAOZvM+H3/kmuXTb+qWkaAAAckBZ+nd+qOiPJGdPorVX10UVvcxVHJfnMFm17VNp8/9PmW0O773/afGto9/3vTt3m9eIt3fzX7WniZsLvdUmOWzZ+7DTtn+nus5OcvYntzEVV7eju7Vtdx0i0+f6nzbeGdt//tPnW0O77nzafv810e/jbJA+qqgdW1V2TnJrkLfMpCwAA5m/DZ367+7aqek6Sv0hySJJXdvcVc6sMAADmbFN9frv77UnePqdaFm3Lu14MSJvvf9p8a2j3/U+bbw3tvv9p8zmr7t7qGgAAYL9we2MAAIZx0Idft2BenKq6pqo+VFWXVtWOadp9qurCqrpqej5yml5V9TvT+3BZVT1ia6u/86iqV1bVDVV1+bJp+9zOVXX6tPxVVXX6VuzLncUqbX5WVV03He+XVtWTls173tTmH62q71023efPOlXVcVV1UVV9uKquqKrnTtMd6wu0l3Z3vC9IVd29qt5XVR+c2vwF0/QHVtXFU/u9frqYQKrqbtP41dP8bctea4/vBWvo7oP2kdkP8T6W5OuT3DXJB5M8ZKvrOlgeSa5JctSKab+R5Mxp+MwkL56Gn5Tkz5JUkkcmuXir67+zPJJ8V5JHJLl8o+2c5D5JPj49HzkNH7nV+3agPlZp87OS/Mweln3I9NlytyQPnD5zDvH5s89tfkySR0zD90ryf6e2daxvTbs73hfX5pXk8Gn4sCQXT8fw+UlOnab/QZJnTcPPTvIH0/CpSV6/t/diq/fvzvA42M/8ugXz/ndyknOn4XOTnLJs+h/3zN8kOaKqjtmKAu9suvvdSW5cMXlf2/l7k1zY3Td29+eSXJjkpMVXf+e0Spuv5uQk53X3l7r7E0muzuyzx+fPPujund39/mn4liRXZnbXUMf6Au2l3VfjeN+k6Zi9dRo9bHp0kscmuWCavvJY3/1v4IIkj6uqyurvBWs42MOvWzAvVif5y6q6pGZ38kuSo7t75zT86SRHT8Pei/na13bW/vPxnOkr9lfu/vo92nzupq91H57ZGTHH+n6yot0Tx/vCVNUhVXVpkhsy+wPtY0lu6u7bpkWWt9+X23aa//kk940237CDPfyyWN/Z3Y9I8sQkP1ZV37V8Zs++l3E5kQXTzvvNy5Mcn+SEJDuT/NbWlnNwqqrDk7whyU90983L5znWF2cP7e54X6Duvr27T8js7rgnJnnwFpc0lIM9/K7rFsxsTHdfNz3fkORNmf0Dvn53d4bp+YZpce/FfO1rO2v/Teru66f/sO5I8op85etFbT4nVXVYZgHsNd39xmmyY33B9tTujvf9o7tvSnJRkkdl1nVn9/0Xlrffl9t2mn/vJJ+NNt+wgz38ugXzglTVPavqXruHkzwhyeWZte/uX1efnuTN0/Bbkvzw9AvtRyb5/LKvMtl3+9rOf5HkCVV15PT15ROmaazTij7qT8nseE9mbX7q9IvsByZ5UJL3xefPPpn6MJ6T5MrufsmyWY71BVqt3R3vi1NVS1V1xDR8jySPz6yv9UVJnjottvJY3/1v4KlJ3jl9C7Lae8EaNnWHtwNduwXzIh2d5E2zz80cmuS13f3nVfW3Sc6vqmcm+WSSH5iWf3tmv86+OskXkvyn/V/ynVNVvS7JY5IcVVWfSvJLSV6UfWjn7r6xqn4ls/+gkuSXu3u9P+gazipt/piqOiGzr92vSfIjSdLdV1TV+Uk+nOS2JD/W3bdPr+PzZ/0eneTpST409YVMkufHsb5oq7X7aY73hTkmyblVdUhmJyHP7+63VtWHk5xXVb+a5AOZ/VGS6flPqurqzH6Ie2qy9/eCvXOHNwAAhnGwd3sAAIAvE34BABiG8AsAwDCEXwAAhiH8AgAwDOEXAIBhCL8AAAxD+AUAYBj/H/JU2FKh4166AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 864x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "f = plt.figure(figsize=(12,6))\n",
    "plt.hist(clinpat.loc[2:,'death_days_to'], bins=10);\n",
    "plt.title('Histogram of days to death');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Add new deceased variables\n",
    "mortality = clinpat.loc[2:,['death_days_to','bcr_patient_barcode']]\n",
    "\n",
    "mortality['deceased'] = mortality['death_days_to'].apply(lambda x: 1 if x > 0 else 0)\n",
    "\n",
    "mortality['deceased_1yr'] = mortality['death_days_to'].apply(lambda x: 1 if x <= 365 else 0)\n",
    "mortality['deceased_3yr'] = mortality['death_days_to'].apply(lambda x: 1 if x <= 1095 else 0)\n",
    "mortality['deceased_5yr'] = mortality['death_days_to'].apply(lambda x: 1 if x <= 1825 else 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking for differentially expressed gene between clinical conditions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## R-log transformation\n",
    "\n",
    "https://bioconductor.org/help/course-materials/2017/CSAMA/labs/2-tuesday/lab-03-rnaseq/rnaseqGene_CSAMA2017.html\n",
    "\n",
    "Many common statistical methods for exploratory analysis of multidimensional data, for example clustering and principal components analysis (PCA), work best for data that generally has the same range of variance at different ranges of the mean values. When the expected amount of variance is approximately the same across different mean values, the data is said to be homoskedastic. For RNA-seq raw counts, however, the variance grows with the mean. For example, if one performs PCA directly on a matrix of size-factor-normalized read counts, the result typically depends only on the few most strongly expressed genes because they show the largest absolute differences between samples. A simple and often used strategy to avoid this is to take the logarithm of the normalized count values plus a small pseudocount; however, now the genes with the very lowest counts will tend to dominate the results because, due to the strong Poisson noise inherent to small count values, and the fact that the logarithm amplifies differences for the smallest values, these low count genes will show the strongest relative differences between samples.\n",
    "\n",
    "As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean. One such transformation is the regularized-logarithm transformation or rlog 2. For genes with high counts, the rlog transformation will give similar result to the ordinary log2 transformation of normalized counts. For genes with lower counts, however, the values are shrunken towards the genesâ€™ averages across all samples. Using an empirical Bayesian prior on inter-sample differences in the form of a ridge penalty, the rlog-transformed data then becomes approximately homoskedastic, and can be used directly for computing distances between samples and making PCA plots. Another transformation, the variance stabilizing transformation 17, is discussed alongside the rlog in the DESeq2 vignette."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PCA Plots\n",
    "Now we have everything setup, the first thing to do is to generate PCA plots to observe whether the samples cluster as expected: controls with controls, and treatments with treatments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another way to visualize sample-to-sample distances is a principal components analysis (PCA). In this ordination method, the data points (here, the samples) are projected onto the 2D plane such that they spread out in the two directions that explain most of the differences (figure below). The x-axis is the direction that separates the data points the most. The values of the samples in this direction are written PC1. The y-axis is a direction (it must be orthogonal to the first direction) that separates the data the second most. The values of the samples in this direction are written PC2. The percent of the total variance that is contained in the direction is printed in the axis label. Note that these percentages do not add to 100%, because there are more dimensions that contain the remaining variance (although each of these remaining dimensions will explain less than the two that we see)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Use DESEQ2 R library to rlog transform\n",
    "\n",
    "from rpy2.robjects.packages import importr\n",
    "deseq = importr('DESeq2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from rpy2.robjects import pandas2ri, Formula\n",
    "pandas2ri.activate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sampleA</th>\n",
       "      <th>sampleB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>geneA</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>geneB</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>geneC</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id  sampleA  sampleB\n",
       "0  geneA        5        1\n",
       "1  geneB        4        5\n",
       "2  geneC        1        2"
      ]
     },
     "execution_count": 461,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({'id': ['geneA', 'geneB', 'geneC'], 'sampleA1': [5,4,1], 'sampleA2': [5,4,1], 'sampleB1': [1,5,2], 'sampleB2': [1,5,2]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samplename</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>sampleA1</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>sampleA2</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>sampleB1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>sampleB2</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  samplename treatment\n",
       "0   sampleA1         A\n",
       "1   sampleA2         A\n",
       "2   sampleB1         B\n",
       "3   sampleB2         B"
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({'samplename': ['sampleA1', 'sampleA2', 'sampleB1', 'sampleB2'], 'treatment': ['A','A','B','B']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R object with classes: ('formula',) mapped to:"
      ]
     },
     "execution_count": 464,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# binary matrix of group that the row belongs to\n",
    "design_formula"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NICK to create wrapper in python for DEseq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in estimateDispersionsFit(object, fitType, quiet = TRUE) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "Calls: <Anonymous> -> estimateDispersionsFit\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in estimateDispersionsFit(object, fitType, quiet = TRUE) : \n  all gene-wise dispersion estimates are within 2 orders of magnitude\n  from the minimum value, and so the standard curve fitting techniques will not work.\n  One can instead use the gene-wise estimates as final estimates:\n  dds <- estimateDispersionsGeneEst(dds)\n  dispersions(dds) <- mcols(dds)$dispGeneEst\n  ...then continue with testing using nbinomWaldTest or nbinomLRT\nCalls: <Anonymous> -> estimateDispersionsFit\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-465-4c86b54ea56d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     36\u001b[0m                                         design=design_formula)\n\u001b[1;32m     37\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m \u001b[0mdeseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mblind\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    190\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mr_k\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    191\u001b[0m         return (super(SignatureTranslatedFunction, self)\n\u001b[0;32m--> 192\u001b[0;31m                 .__call__(*args, **kwargs))\n\u001b[0m\u001b[1;32m    193\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    119\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m                 \u001b[0mnew_kwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 121\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mFunction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mnew_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mnew_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    122\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrpy2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    123\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_cdata_res_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunction\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0mcdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m         \u001b[0;31m# TODO: test cdata is of the expected CType\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_cdata_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    787\u001b[0m                     error_occured))\n\u001b[1;32m    788\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merror_occured\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 789\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0membedded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_rinterface\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_geterrmessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    790\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    791\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in estimateDispersionsFit(object, fitType, quiet = TRUE) : \n  all gene-wise dispersion estimates are within 2 orders of magnitude\n  from the minimum value, and so the standard curve fitting techniques will not work.\n  One can instead use the gene-wise estimates as final estimates:\n  dds <- estimateDispersionsGeneEst(dds)\n  dispersions(dds) <- mcols(dds)$dispGeneEst\n  ...then continue with testing using nbinomWaldTest or nbinomLRT\nCalls: <Anonymous> -> estimateDispersionsFit\n"
     ]
    }
   ],
   "source": [
    "# Create DESeq dataset\n",
    "\n",
    "'''\n",
    "DESeq2 object through rpy2\n",
    "    input:\n",
    "    count_matrix: should be a pandas dataframe with each column as count, and a id column for gene id\n",
    "        example:\n",
    "        id    sampleA    sampleB\n",
    "        geneA    5    1\n",
    "        geneB    4    5\n",
    "        geneC    1    2\n",
    "    design_matrix: an design matrix in the form of pandas dataframe, see DESeq2 manual, samplenames as rownames\n",
    "                treatment\n",
    "    sampleA1        A\n",
    "    sampleA2        A\n",
    "    sampleB1        B\n",
    "    sampleB2        B\n",
    "    design_formula: see DESeq2 manual, example: \"~ treatment\"\"\n",
    "    gene_column: column name of gene id columns, example \"id\"\n",
    "\n",
    "'''\n",
    "\n",
    "count_matrix = pd.DataFrame({'id': ['geneA', 'geneB', 'geneC'], 'sampleA1': [5,4,1], 'sampleA2': [5,4,1], 'sampleB1': [1,5,2], 'sampleB2': [1,5,2]})\n",
    "design_matrix = pd.DataFrame({'samplename': ['sampleA1', 'sampleA2', 'sampleB1', 'sampleB2'], 'treatment': ['A','A','B','B']})\n",
    "\n",
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    count_matrix = ro.conversion.py2rpy(count_matrix.set_index('id'))\n",
    "    design_matrix = ro.conversion.py2rpy(design_matrix.set_index('samplename'))\n",
    "\n",
    "# count_matrix = pandas2ri.py2ri(count_matrix)\n",
    "# design_matrix = pandas2ri.py2ri(lihc_processed['tumor_grade'])\n",
    "design_formula = Formula(\"~ treatment\")\n",
    "        \n",
    "dds = deseq.DESeqDataSetFromMatrix(countData=count_matrix,\n",
    "                                        colData=design_matrix,\n",
    "                                        design=design_formula)\n",
    "\n",
    "deseq.rlog(dds,blind=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R object with classes: ('DESeqDataSet',) mapped to:"
      ]
     },
     "execution_count": 466,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mattheweng/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py:60: UserWarning: Error while trying to convert the column \"0\". Fall back to string conversion. The error is: Series can only be of one type, or None.\n",
      "  % (name, str(e)))\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'float' object has no attribute 'encode'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mpy2rpy_pandasdataframe\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m     54\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 55\u001b[0;31m             \u001b[0mod\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     56\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    806\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 807\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    808\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mpy2rpy_pandasseries\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    125\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mhomogeneous_type\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 126\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Series can only be of one type, or None.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    127\u001b[0m         \u001b[0;31m# TODO: Could this be merged with obj.type.name == 'O' case above ?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Series can only be of one type, or None.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36mfrom_object\u001b[0;34m(cls, obj)\u001b[0m\n\u001b[1;32m    368\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 369\u001b[0;31m             \u001b[0mmv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmemoryview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    370\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_memoryview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: memoryview: a bytes-like object is required, not 'Series'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-407-995faf89e015>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0mcount_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlihc_processed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tumor_grade'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m     \u001b[0mcount_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount_matrix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m     \u001b[0mdesign_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlihc_processed\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tumor_grade'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    805\u001b[0m                             '1 positional argument')\n\u001b[1;32m    806\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 807\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    808\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    809\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mpy2rpy_pandasdataframe\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m     59\u001b[0m                           \u001b[0;34m'The error is: %s'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m                           % (name, str(e)))\n\u001b[0;32m---> 61\u001b[0;31m             \u001b[0mod\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mStrVector\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mod\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/robjects/vectors.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    382\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    383\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_add_rops\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    386\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    287\u001b[0m             \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcollections\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSized\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 289\u001b[0;31m             \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__sexp__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    290\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    291\u001b[0m             raise TypeError('The constructor must be called '\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36mfrom_object\u001b[0;34m(cls, obj)\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mTypeError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 373\u001b[0;31m                 \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_iterable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    374\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    375\u001b[0m                 msg = ('The class methods from_memoryview() and '\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_cdata_res_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunction\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0mcdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m         \u001b[0;31m# TODO: test cdata is of the expected CType\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_cdata_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36mfrom_iterable\u001b[0;34m(cls, iterable, populate_func)\u001b[0m\n\u001b[1;32m    318\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mpopulate_func\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    319\u001b[0m                 cls._populate_r_vector(iterable,\n\u001b[0;32m--> 320\u001b[0;31m                                        r_vector)\n\u001b[0m\u001b[1;32m    321\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    322\u001b[0m                 \u001b[0mpopulate_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mr_vector\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36m_populate_r_vector\u001b[0;34m(cls, iterable, r_vector)\u001b[0m\n\u001b[1;32m    301\u001b[0m                                   \u001b[0mr_vector\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                                   \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_R_SET_VECTOR_ELT\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 303\u001b[0;31m                                   cls._CAST_IN)\n\u001b[0m\u001b[1;32m    304\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36m_populate_r_vector\u001b[0;34m(iterable, r_vector, set_elt, cast_value)\u001b[0m\n\u001b[1;32m    238\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_populate_r_vector\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mr_vector\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mset_elt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcast_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 240\u001b[0;31m         \u001b[0mset_elt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr_vector\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcast_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    241\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py\u001b[0m in \u001b[0;36m_as_charsxp_cdata\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    431\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__sexp__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    432\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 433\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_str_to_charsxp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    434\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    435\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_str_to_charsxp\u001b[0;34m(val)\u001b[0m\n\u001b[1;32m    133\u001b[0m         \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mR_NaString\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    134\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 135\u001b[0;31m         \u001b[0mcchar\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_str_to_cchar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    136\u001b[0m         \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRf_mkCharCE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcchar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_CE_DEFAULT_VALUE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    137\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/Galvanize/Projects/hacknights/env-tcga/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_str_to_cchar\u001b[0;34m(s, encoding)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_str_to_cchar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m     \u001b[0;31m# TODO: use isStrinb and installTrChar\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m     \u001b[0mb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mffi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'char[]'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'float' object has no attribute 'encode'"
     ]
    }
   ],
   "source": [
    "# Create DESeq dataset\n",
    "\n",
    "'''\n",
    "DESeq2 object through rpy2\n",
    "    input:\n",
    "    count_matrix: should be a pandas dataframe with each column as count, and a id column for gene id\n",
    "        example:\n",
    "        id    sampleA    sampleB\n",
    "        geneA    5    1\n",
    "        geneB    4    5\n",
    "        geneC    1    2\n",
    "    design_matrix: an design matrix in the form of pandas dataframe, see DESeq2 manual, samplenames as rownames\n",
    "                treatment\n",
    "    sampleA1        A\n",
    "    sampleA2        A\n",
    "    sampleB1        B\n",
    "    sampleB2        B\n",
    "    design_formula: see DESeq2 manual, example: \"~ treatment\"\"\n",
    "    gene_column: column name of gene id columns, example \"id\"\n",
    "\n",
    "'''\n",
    "\n",
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    count_matrix = lihc_processed.reset_index().drop(['tumor_grade'], axis=1).T\n",
    "    count_matrix = ro.conversion.py2rpy(count_matrix.reset_index(drop=True))\n",
    "    design_matrix = ro.conversion.py2rpy(lihc_processed['tumor_grade'].reset_index(drop=True))\n",
    "\n",
    "# count_matrix = pandas2ri.py2ri(count_matrix)\n",
    "# design_matrix = pandas2ri.py2ri(lihc_processed['tumor_grade'])\n",
    "design_formula = Formula(\"~ treatment\")\n",
    "        \n",
    "dds = deseq.DESeqDataSetFromMatrix(countData=count_matrix,\n",
    "                                        colData=design_matrix,\n",
    "                                        design=design_formula)\n",
    "\n",
    "deseq.rlog(dds,blind=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     G3\n",
       "1     G3\n",
       "2     G3\n",
       "3     G2\n",
       "4     G2\n",
       "      ..\n",
       "87    G2\n",
       "88    G1\n",
       "89    G3\n",
       "90    G3\n",
       "91    G1\n",
       "Name: tumor_grade, Length: 92, dtype: object"
      ]
     },
     "execution_count": 396,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed['tumor_grade'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>82</th>\n",
       "      <th>83</th>\n",
       "      <th>84</th>\n",
       "      <th>85</th>\n",
       "      <th>86</th>\n",
       "      <th>87</th>\n",
       "      <th>88</th>\n",
       "      <th>89</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>4073</td>\n",
       "      <td>A10W</td>\n",
       "      <td>A10W</td>\n",
       "      <td>A10X</td>\n",
       "      <td>A10X</td>\n",
       "      <td>A10Y</td>\n",
       "      <td>A10Y</td>\n",
       "      <td>A10Z</td>\n",
       "      <td>A10Z</td>\n",
       "      <td>A110</td>\n",
       "      <td>...</td>\n",
       "      <td>A4NN</td>\n",
       "      <td>A4NO</td>\n",
       "      <td>A4NP</td>\n",
       "      <td>A4NQ</td>\n",
       "      <td>A4NR</td>\n",
       "      <td>A4NS</td>\n",
       "      <td>A4NV</td>\n",
       "      <td>A5W4</td>\n",
       "      <td>A69H</td>\n",
       "      <td>A69I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>32.022</td>\n",
       "      <td>186.426</td>\n",
       "      <td>18.8844</td>\n",
       "      <td>163.985</td>\n",
       "      <td>268.089</td>\n",
       "      <td>244.851</td>\n",
       "      <td>48.2907</td>\n",
       "      <td>185.821</td>\n",
       "      <td>34.0774</td>\n",
       "      <td>62.2372</td>\n",
       "      <td>...</td>\n",
       "      <td>32.811</td>\n",
       "      <td>48.6618</td>\n",
       "      <td>81.6327</td>\n",
       "      <td>84.3673</td>\n",
       "      <td>52.5176</td>\n",
       "      <td>342.282</td>\n",
       "      <td>57.0324</td>\n",
       "      <td>35.6486</td>\n",
       "      <td>12506.6</td>\n",
       "      <td>42.4645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1742.45</td>\n",
       "      <td>390.894</td>\n",
       "      <td>1219.34</td>\n",
       "      <td>324.876</td>\n",
       "      <td>598.55</td>\n",
       "      <td>998.474</td>\n",
       "      <td>1978.96</td>\n",
       "      <td>333.582</td>\n",
       "      <td>1967.86</td>\n",
       "      <td>227.923</td>\n",
       "      <td>...</td>\n",
       "      <td>2280.12</td>\n",
       "      <td>3627.74</td>\n",
       "      <td>385.9</td>\n",
       "      <td>617.534</td>\n",
       "      <td>1362.21</td>\n",
       "      <td>1207.54</td>\n",
       "      <td>5304.01</td>\n",
       "      <td>2179.54</td>\n",
       "      <td>897.81</td>\n",
       "      <td>595.261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>487.649</td>\n",
       "      <td>268.041</td>\n",
       "      <td>178.106</td>\n",
       "      <td>563.119</td>\n",
       "      <td>560.252</td>\n",
       "      <td>360.031</td>\n",
       "      <td>461.87</td>\n",
       "      <td>170.895</td>\n",
       "      <td>1300.55</td>\n",
       "      <td>266.611</td>\n",
       "      <td>...</td>\n",
       "      <td>441.724</td>\n",
       "      <td>306.569</td>\n",
       "      <td>150.013</td>\n",
       "      <td>472.283</td>\n",
       "      <td>844.072</td>\n",
       "      <td>196.696</td>\n",
       "      <td>159.981</td>\n",
       "      <td>250.369</td>\n",
       "      <td>262.774</td>\n",
       "      <td>226.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2077.31</td>\n",
       "      <td>1195.88</td>\n",
       "      <td>1268.22</td>\n",
       "      <td>1310.64</td>\n",
       "      <td>1238.68</td>\n",
       "      <td>1904.65</td>\n",
       "      <td>1061.92</td>\n",
       "      <td>932.836</td>\n",
       "      <td>1595.17</td>\n",
       "      <td>742.641</td>\n",
       "      <td>...</td>\n",
       "      <td>743.879</td>\n",
       "      <td>1215.33</td>\n",
       "      <td>1030.48</td>\n",
       "      <td>819.319</td>\n",
       "      <td>1144.02</td>\n",
       "      <td>1516.78</td>\n",
       "      <td>962.784</td>\n",
       "      <td>547.165</td>\n",
       "      <td>667.153</td>\n",
       "      <td>803.033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17176</td>\n",
       "      <td>663.769</td>\n",
       "      <td>414.949</td>\n",
       "      <td>269.935</td>\n",
       "      <td>613.861</td>\n",
       "      <td>568.458</td>\n",
       "      <td>544.622</td>\n",
       "      <td>531.676</td>\n",
       "      <td>707.463</td>\n",
       "      <td>1291.49</td>\n",
       "      <td>365.854</td>\n",
       "      <td>...</td>\n",
       "      <td>397.649</td>\n",
       "      <td>373.479</td>\n",
       "      <td>791.413</td>\n",
       "      <td>634.059</td>\n",
       "      <td>1095.83</td>\n",
       "      <td>438.823</td>\n",
       "      <td>621.073</td>\n",
       "      <td>1427.6</td>\n",
       "      <td>459.854</td>\n",
       "      <td>932.701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17177</td>\n",
       "      <td>564.044</td>\n",
       "      <td>268.9</td>\n",
       "      <td>139.596</td>\n",
       "      <td>345.916</td>\n",
       "      <td>277.937</td>\n",
       "      <td>456.903</td>\n",
       "      <td>581.879</td>\n",
       "      <td>438.806</td>\n",
       "      <td>421.87</td>\n",
       "      <td>310.345</td>\n",
       "      <td>...</td>\n",
       "      <td>295.299</td>\n",
       "      <td>264.599</td>\n",
       "      <td>542.804</td>\n",
       "      <td>273.106</td>\n",
       "      <td>443.422</td>\n",
       "      <td>287.042</td>\n",
       "      <td>410.827</td>\n",
       "      <td>508.2</td>\n",
       "      <td>636.496</td>\n",
       "      <td>498.957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17178</td>\n",
       "      <td>222.324</td>\n",
       "      <td>207.904</td>\n",
       "      <td>352.138</td>\n",
       "      <td>290.842</td>\n",
       "      <td>200.793</td>\n",
       "      <td>510.298</td>\n",
       "      <td>235.716</td>\n",
       "      <td>126.866</td>\n",
       "      <td>472.77</td>\n",
       "      <td>96.7199</td>\n",
       "      <td>...</td>\n",
       "      <td>133.203</td>\n",
       "      <td>324.209</td>\n",
       "      <td>268.222</td>\n",
       "      <td>159.167</td>\n",
       "      <td>216.026</td>\n",
       "      <td>374.806</td>\n",
       "      <td>594.007</td>\n",
       "      <td>156.688</td>\n",
       "      <td>51.0949</td>\n",
       "      <td>142.559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17179</td>\n",
       "      <td>111.619</td>\n",
       "      <td>29.2096</td>\n",
       "      <td>106.641</td>\n",
       "      <td>25.3713</td>\n",
       "      <td>10.9424</td>\n",
       "      <td>16.0183</td>\n",
       "      <td>0.4781</td>\n",
       "      <td>124.627</td>\n",
       "      <td>0</td>\n",
       "      <td>10.9336</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4691</td>\n",
       "      <td>1.2165</td>\n",
       "      <td>1.0602</td>\n",
       "      <td>59.1441</td>\n",
       "      <td>64.4288</td>\n",
       "      <td>94.476</td>\n",
       "      <td>1.45</td>\n",
       "      <td>23.2131</td>\n",
       "      <td>10.219</td>\n",
       "      <td>1.5166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17180</td>\n",
       "      <td>13.2662</td>\n",
       "      <td>12.8866</td>\n",
       "      <td>8.8868</td>\n",
       "      <td>9.2822</td>\n",
       "      <td>9.8482</td>\n",
       "      <td>15.2555</td>\n",
       "      <td>1.4344</td>\n",
       "      <td>11.194</td>\n",
       "      <td>0</td>\n",
       "      <td>6.7283</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4691</td>\n",
       "      <td>4.2579</td>\n",
       "      <td>1.0602</td>\n",
       "      <td>5.2186</td>\n",
       "      <td>21.1153</td>\n",
       "      <td>17.5529</td>\n",
       "      <td>3.8666</td>\n",
       "      <td>0.829</td>\n",
       "      <td>4.3796</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17181 rows Ã— 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        1        2        3        4        5        6        7   \\\n",
       "0         4073     A10W     A10W     A10X     A10X     A10Y     A10Y     A10Z   \n",
       "1       32.022  186.426  18.8844  163.985  268.089  244.851  48.2907  185.821   \n",
       "2      1742.45  390.894  1219.34  324.876   598.55  998.474  1978.96  333.582   \n",
       "3      487.649  268.041  178.106  563.119  560.252  360.031   461.87  170.895   \n",
       "4      2077.31  1195.88  1268.22  1310.64  1238.68  1904.65  1061.92  932.836   \n",
       "...        ...      ...      ...      ...      ...      ...      ...      ...   \n",
       "17176  663.769  414.949  269.935  613.861  568.458  544.622  531.676  707.463   \n",
       "17177  564.044    268.9  139.596  345.916  277.937  456.903  581.879  438.806   \n",
       "17178  222.324  207.904  352.138  290.842  200.793  510.298  235.716  126.866   \n",
       "17179  111.619  29.2096  106.641  25.3713  10.9424  16.0183   0.4781  124.627   \n",
       "17180  13.2662  12.8866   8.8868   9.2822   9.8482  15.2555   1.4344   11.194   \n",
       "\n",
       "            8        9   ...       82       83       84       85       86  \\\n",
       "0         A10Z     A110  ...     A4NN     A4NO     A4NP     A4NQ     A4NR   \n",
       "1      34.0774  62.2372  ...   32.811  48.6618  81.6327  84.3673  52.5176   \n",
       "2      1967.86  227.923  ...  2280.12  3627.74    385.9  617.534  1362.21   \n",
       "3      1300.55  266.611  ...  441.724  306.569  150.013  472.283  844.072   \n",
       "4      1595.17  742.641  ...  743.879  1215.33  1030.48  819.319  1144.02   \n",
       "...        ...      ...  ...      ...      ...      ...      ...      ...   \n",
       "17176  1291.49  365.854  ...  397.649  373.479  791.413  634.059  1095.83   \n",
       "17177   421.87  310.345  ...  295.299  264.599  542.804  273.106  443.422   \n",
       "17178   472.77  96.7199  ...  133.203  324.209  268.222  159.167  216.026   \n",
       "17179        0  10.9336  ...   1.4691   1.2165   1.0602  59.1441  64.4288   \n",
       "17180        0   6.7283  ...   1.4691   4.2579   1.0602   5.2186  21.1153   \n",
       "\n",
       "            87       88       89       90       91  \n",
       "0         A4NS     A4NV     A5W4     A69H     A69I  \n",
       "1      342.282  57.0324  35.6486  12506.6  42.4645  \n",
       "2      1207.54  5304.01  2179.54   897.81  595.261  \n",
       "3      196.696  159.981  250.369  262.774   226.73  \n",
       "4      1516.78  962.784  547.165  667.153  803.033  \n",
       "...        ...      ...      ...      ...      ...  \n",
       "17176  438.823  621.073   1427.6  459.854  932.701  \n",
       "17177  287.042  410.827    508.2  636.496  498.957  \n",
       "17178  374.806  594.007  156.688  51.0949  142.559  \n",
       "17179   94.476     1.45  23.2131   10.219   1.5166  \n",
       "17180  17.5529   3.8666    0.829   4.3796        0  \n",
       "\n",
       "[17181 rows x 92 columns]"
      ]
     },
     "execution_count": 406,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_matrix = lihc_processed.reset_index().drop(['tumor_grade'], axis=1).T\n",
    "count_matrix.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <span>R/rpy2 DataFrame (5 x 20531)</span>\n",
       "        <table>\n",
       "          <thead>\n",
       "            <tr>\n",
       "              \n",
       "              <th>X..100130426</th>\n",
       "              \n",
       "              <th>X..100133144</th>\n",
       "              \n",
       "              <th>X..100134869</th>\n",
       "              \n",
       "              <th>...</th>\n",
       "              \n",
       "              <th>ZZZ3.26009</th>\n",
       "              \n",
       "              <th>psiTPTE22.387590</th>\n",
       "              \n",
       "              <th>tAKR.389932</th>\n",
       "              \n",
       "            </tr>\n",
       "          </thead>\n",
       "          <tbody>\n",
       "          \n",
       "          <tr>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "          </tr>\n",
       "          \n",
       "          </tbody>\n",
       "        </table>\n",
       "    "
      ],
      "text/plain": [
       "R object with classes: ('data.frame',) mapped to:\n",
       "[FloatSexp..., FloatSexp..., FloatSexp..., FloatSexp..., ..., FloatSexp..., FloatSexp..., FloatSexp..., FloatSexp...]\n",
       "  X..100130426: <class 'numpy.ndarray'>\n",
       "  array([0., 0., 0., 0., 0.])\n",
       "  X..100133144: <class 'numpy.ndarray'>\n",
       "  array([ 1.5051, 26.412 ,  0.    ,  5.7222, 11.4975])\n",
       "  X..100134869: <class 'numpy.ndarray'>\n",
       "  array([3.7074, 2.6663, 4.4833, 5.1216, 5.423 ])\n",
       "  X..10357: <class 'numpy.ndarray'>\n",
       "  array([ 90.1124,  71.0054,  95.5122,  61.6679, 104.467 ])\n",
       "...\n",
       "  X..136542: <class 'numpy.ndarray'>\n",
       "  array([ 890.0472,  755.5446,  860.7935, 1339.6283,  575.2961])\n",
       "  X..155060: <class 'numpy.ndarray'>\n",
       "  array([510.1808, 860.5224, 523.6494, 343.6655, 665.8206])\n",
       "  X..26823: <class 'numpy.ndarray'>\n",
       "  array([ 3.9094,  6.4071, 14.3466,  2.5024,  2.5381])\n",
       "  X..280660: <class 'numpy.ndarray'>\n",
       "  array([  6.5157, 482.9966,  83.3894,   2.5024, 119.2893])"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    r_from_pd_df = ro.conversion.py2rpy(lihc.head())\n",
    "r_from_pd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>4073</th>\n",
       "      <th>A10W</th>\n",
       "      <th>A10W</th>\n",
       "      <th>A10X</th>\n",
       "      <th>A10X</th>\n",
       "      <th>A10Y</th>\n",
       "      <th>A10Y</th>\n",
       "      <th>A10Z</th>\n",
       "      <th>A10Z</th>\n",
       "      <th>A110</th>\n",
       "      <th>...</th>\n",
       "      <th>A4NN</th>\n",
       "      <th>A4NO</th>\n",
       "      <th>A4NP</th>\n",
       "      <th>A4NQ</th>\n",
       "      <th>A4NR</th>\n",
       "      <th>A4NS</th>\n",
       "      <th>A4NV</th>\n",
       "      <th>A5W4</th>\n",
       "      <th>A69H</th>\n",
       "      <th>A69I</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>32.0220</td>\n",
       "      <td>186.4261</td>\n",
       "      <td>18.8844</td>\n",
       "      <td>163.9851</td>\n",
       "      <td>268.0892</td>\n",
       "      <td>244.8513</td>\n",
       "      <td>48.2907</td>\n",
       "      <td>185.8209</td>\n",
       "      <td>34.0774</td>\n",
       "      <td>62.2372</td>\n",
       "      <td>...</td>\n",
       "      <td>32.8110</td>\n",
       "      <td>48.6618</td>\n",
       "      <td>81.6327</td>\n",
       "      <td>84.3673</td>\n",
       "      <td>52.5176</td>\n",
       "      <td>342.2819</td>\n",
       "      <td>57.0324</td>\n",
       "      <td>35.6486</td>\n",
       "      <td>12506.5693</td>\n",
       "      <td>42.4645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KLF13</td>\n",
       "      <td>1742.4520</td>\n",
       "      <td>390.8935</td>\n",
       "      <td>1219.3383</td>\n",
       "      <td>324.8762</td>\n",
       "      <td>598.5501</td>\n",
       "      <td>998.4744</td>\n",
       "      <td>1978.9625</td>\n",
       "      <td>333.5821</td>\n",
       "      <td>1967.8637</td>\n",
       "      <td>227.9226</td>\n",
       "      <td>...</td>\n",
       "      <td>2280.1175</td>\n",
       "      <td>3627.7372</td>\n",
       "      <td>385.8998</td>\n",
       "      <td>617.5336</td>\n",
       "      <td>1362.2090</td>\n",
       "      <td>1207.5374</td>\n",
       "      <td>5304.0116</td>\n",
       "      <td>2179.5406</td>\n",
       "      <td>897.8102</td>\n",
       "      <td>595.2607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MPV17L</td>\n",
       "      <td>487.6487</td>\n",
       "      <td>268.0412</td>\n",
       "      <td>178.1056</td>\n",
       "      <td>563.1188</td>\n",
       "      <td>560.2517</td>\n",
       "      <td>360.0305</td>\n",
       "      <td>461.8695</td>\n",
       "      <td>170.8955</td>\n",
       "      <td>1300.5500</td>\n",
       "      <td>266.6106</td>\n",
       "      <td>...</td>\n",
       "      <td>441.7238</td>\n",
       "      <td>306.5693</td>\n",
       "      <td>150.0133</td>\n",
       "      <td>472.2827</td>\n",
       "      <td>844.0715</td>\n",
       "      <td>196.6959</td>\n",
       "      <td>159.9807</td>\n",
       "      <td>250.3694</td>\n",
       "      <td>262.7737</td>\n",
       "      <td>226.7299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MARF1</td>\n",
       "      <td>2077.3102</td>\n",
       "      <td>1195.8763</td>\n",
       "      <td>1268.2156</td>\n",
       "      <td>1310.6436</td>\n",
       "      <td>1238.6814</td>\n",
       "      <td>1904.6529</td>\n",
       "      <td>1061.9173</td>\n",
       "      <td>932.8358</td>\n",
       "      <td>1595.1688</td>\n",
       "      <td>742.6409</td>\n",
       "      <td>...</td>\n",
       "      <td>743.8786</td>\n",
       "      <td>1215.3285</td>\n",
       "      <td>1030.4797</td>\n",
       "      <td>819.3192</td>\n",
       "      <td>1144.0173</td>\n",
       "      <td>1516.7785</td>\n",
       "      <td>962.7840</td>\n",
       "      <td>547.1650</td>\n",
       "      <td>667.1533</td>\n",
       "      <td>803.0332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CNOT3</td>\n",
       "      <td>865.5078</td>\n",
       "      <td>457.9038</td>\n",
       "      <td>839.0588</td>\n",
       "      <td>532.7970</td>\n",
       "      <td>589.7962</td>\n",
       "      <td>558.3524</td>\n",
       "      <td>658.8573</td>\n",
       "      <td>632.0896</td>\n",
       "      <td>941.6586</td>\n",
       "      <td>632.4643</td>\n",
       "      <td>...</td>\n",
       "      <td>1544.0744</td>\n",
       "      <td>781.0219</td>\n",
       "      <td>621.2563</td>\n",
       "      <td>1290.7322</td>\n",
       "      <td>809.4207</td>\n",
       "      <td>1020.1342</td>\n",
       "      <td>1011.5998</td>\n",
       "      <td>849.7638</td>\n",
       "      <td>1502.1898</td>\n",
       "      <td>649.8578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SESN2</td>\n",
       "      <td>663.7694</td>\n",
       "      <td>414.9485</td>\n",
       "      <td>269.9355</td>\n",
       "      <td>613.8614</td>\n",
       "      <td>568.4585</td>\n",
       "      <td>544.6224</td>\n",
       "      <td>531.6758</td>\n",
       "      <td>707.4627</td>\n",
       "      <td>1291.4914</td>\n",
       "      <td>365.8537</td>\n",
       "      <td>...</td>\n",
       "      <td>397.6494</td>\n",
       "      <td>373.4793</td>\n",
       "      <td>791.4127</td>\n",
       "      <td>634.0592</td>\n",
       "      <td>1095.8311</td>\n",
       "      <td>438.8229</td>\n",
       "      <td>621.0730</td>\n",
       "      <td>1427.6032</td>\n",
       "      <td>459.8540</td>\n",
       "      <td>932.7014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MED18</td>\n",
       "      <td>564.0439</td>\n",
       "      <td>268.9003</td>\n",
       "      <td>139.5963</td>\n",
       "      <td>345.9158</td>\n",
       "      <td>277.9374</td>\n",
       "      <td>456.9031</td>\n",
       "      <td>581.8790</td>\n",
       "      <td>438.8060</td>\n",
       "      <td>421.8699</td>\n",
       "      <td>310.3448</td>\n",
       "      <td>...</td>\n",
       "      <td>295.2987</td>\n",
       "      <td>264.5985</td>\n",
       "      <td>542.8041</td>\n",
       "      <td>273.1064</td>\n",
       "      <td>443.4218</td>\n",
       "      <td>287.0418</td>\n",
       "      <td>410.8265</td>\n",
       "      <td>508.2002</td>\n",
       "      <td>636.4964</td>\n",
       "      <td>498.9573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PLEKHM3</td>\n",
       "      <td>222.3239</td>\n",
       "      <td>207.9038</td>\n",
       "      <td>352.1381</td>\n",
       "      <td>290.8416</td>\n",
       "      <td>200.7933</td>\n",
       "      <td>510.2975</td>\n",
       "      <td>235.7160</td>\n",
       "      <td>126.8657</td>\n",
       "      <td>472.7704</td>\n",
       "      <td>96.7199</td>\n",
       "      <td>...</td>\n",
       "      <td>133.2027</td>\n",
       "      <td>324.2092</td>\n",
       "      <td>268.2216</td>\n",
       "      <td>159.1671</td>\n",
       "      <td>216.0260</td>\n",
       "      <td>374.8064</td>\n",
       "      <td>594.0068</td>\n",
       "      <td>156.6882</td>\n",
       "      <td>51.0949</td>\n",
       "      <td>142.5592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>C3orf52</td>\n",
       "      <td>111.6194</td>\n",
       "      <td>29.2096</td>\n",
       "      <td>106.6412</td>\n",
       "      <td>25.3713</td>\n",
       "      <td>10.9424</td>\n",
       "      <td>16.0183</td>\n",
       "      <td>0.4781</td>\n",
       "      <td>124.6269</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>10.9336</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4691</td>\n",
       "      <td>1.2165</td>\n",
       "      <td>1.0602</td>\n",
       "      <td>59.1441</td>\n",
       "      <td>64.4288</td>\n",
       "      <td>94.4760</td>\n",
       "      <td>1.4500</td>\n",
       "      <td>23.2131</td>\n",
       "      <td>10.2190</td>\n",
       "      <td>1.5166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GCSAM</td>\n",
       "      <td>13.2662</td>\n",
       "      <td>12.8866</td>\n",
       "      <td>8.8868</td>\n",
       "      <td>9.2822</td>\n",
       "      <td>9.8482</td>\n",
       "      <td>15.2555</td>\n",
       "      <td>1.4344</td>\n",
       "      <td>11.1940</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>6.7283</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4691</td>\n",
       "      <td>4.2579</td>\n",
       "      <td>1.0602</td>\n",
       "      <td>5.2186</td>\n",
       "      <td>21.1153</td>\n",
       "      <td>17.5529</td>\n",
       "      <td>3.8666</td>\n",
       "      <td>0.8290</td>\n",
       "      <td>4.3796</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17180 rows Ã— 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              4073       A10W       A10W       A10X       A10X       A10Y  \\\n",
       "CDKN1C     32.0220   186.4261    18.8844   163.9851   268.0892   244.8513   \n",
       "KLF13    1742.4520   390.8935  1219.3383   324.8762   598.5501   998.4744   \n",
       "MPV17L    487.6487   268.0412   178.1056   563.1188   560.2517   360.0305   \n",
       "MARF1    2077.3102  1195.8763  1268.2156  1310.6436  1238.6814  1904.6529   \n",
       "CNOT3     865.5078   457.9038   839.0588   532.7970   589.7962   558.3524   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "SESN2     663.7694   414.9485   269.9355   613.8614   568.4585   544.6224   \n",
       "MED18     564.0439   268.9003   139.5963   345.9158   277.9374   456.9031   \n",
       "PLEKHM3   222.3239   207.9038   352.1381   290.8416   200.7933   510.2975   \n",
       "C3orf52   111.6194    29.2096   106.6412    25.3713    10.9424    16.0183   \n",
       "GCSAM      13.2662    12.8866     8.8868     9.2822     9.8482    15.2555   \n",
       "\n",
       "              A10Y      A10Z       A10Z      A110  ...       A4NN       A4NO  \\\n",
       "CDKN1C     48.2907  185.8209    34.0774   62.2372  ...    32.8110    48.6618   \n",
       "KLF13    1978.9625  333.5821  1967.8637  227.9226  ...  2280.1175  3627.7372   \n",
       "MPV17L    461.8695  170.8955  1300.5500  266.6106  ...   441.7238   306.5693   \n",
       "MARF1    1061.9173  932.8358  1595.1688  742.6409  ...   743.8786  1215.3285   \n",
       "CNOT3     658.8573  632.0896   941.6586  632.4643  ...  1544.0744   781.0219   \n",
       "...            ...       ...        ...       ...  ...        ...        ...   \n",
       "SESN2     531.6758  707.4627  1291.4914  365.8537  ...   397.6494   373.4793   \n",
       "MED18     581.8790  438.8060   421.8699  310.3448  ...   295.2987   264.5985   \n",
       "PLEKHM3   235.7160  126.8657   472.7704   96.7199  ...   133.2027   324.2092   \n",
       "C3orf52     0.4781  124.6269     0.0000   10.9336  ...     1.4691     1.2165   \n",
       "GCSAM       1.4344   11.1940     0.0000    6.7283  ...     1.4691     4.2579   \n",
       "\n",
       "              A4NP       A4NQ       A4NR       A4NS       A4NV       A5W4  \\\n",
       "CDKN1C     81.6327    84.3673    52.5176   342.2819    57.0324    35.6486   \n",
       "KLF13     385.8998   617.5336  1362.2090  1207.5374  5304.0116  2179.5406   \n",
       "MPV17L    150.0133   472.2827   844.0715   196.6959   159.9807   250.3694   \n",
       "MARF1    1030.4797   819.3192  1144.0173  1516.7785   962.7840   547.1650   \n",
       "CNOT3     621.2563  1290.7322   809.4207  1020.1342  1011.5998   849.7638   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "SESN2     791.4127   634.0592  1095.8311   438.8229   621.0730  1427.6032   \n",
       "MED18     542.8041   273.1064   443.4218   287.0418   410.8265   508.2002   \n",
       "PLEKHM3   268.2216   159.1671   216.0260   374.8064   594.0068   156.6882   \n",
       "C3orf52     1.0602    59.1441    64.4288    94.4760     1.4500    23.2131   \n",
       "GCSAM       1.0602     5.2186    21.1153    17.5529     3.8666     0.8290   \n",
       "\n",
       "               A69H      A69I  \n",
       "CDKN1C   12506.5693   42.4645  \n",
       "KLF13      897.8102  595.2607  \n",
       "MPV17L     262.7737  226.7299  \n",
       "MARF1      667.1533  803.0332  \n",
       "CNOT3     1502.1898  649.8578  \n",
       "...             ...       ...  \n",
       "SESN2      459.8540  932.7014  \n",
       "MED18      636.4964  498.9573  \n",
       "PLEKHM3     51.0949  142.5592  \n",
       "C3orf52     10.2190    1.5166  \n",
       "GCSAM        4.3796    0.0000  \n",
       "\n",
       "[17180 rows x 92 columns]"
      ]
     },
     "execution_count": 372,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed.drop('tumor_grade', axis=1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4073    G3\n",
       "A10W    G3\n",
       "A10W    G3\n",
       "A10X    G2\n",
       "A10X    G2\n",
       "        ..\n",
       "A4NS    G2\n",
       "A4NV    G1\n",
       "A5W4    G3\n",
       "A69H    G3\n",
       "A69I    G1\n",
       "Name: tumor_grade, Length: 92, dtype: object"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lihc_processed['tumor_grade']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential Gene Expression\n",
    "Now we are ready to identify the differentially expressed genes between the two sets of samples: control vs. treatment. We will achieve this using the Characteristic Direction method[6](#ref6) that we developed and published in BMC Bioinformatics in 2014.\n",
    "\n",
    "An implementation in Python of the Characteristic Direction method can be downloaded and installed from here: https://github.com/wangz10/geode.\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [],
   "source": [
    "import src.geode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import geode\n",
    "d_platform_cd = {} # to top up/down genes\n",
    "cd_results = pd.DataFrame(index=expr_df.index)\n",
    "\n",
    "sample_classes = {}\n",
    "for layout in meta_df['LibraryLayout_s'].unique():\n",
    "    ## make sample_class \n",
    "    sample_class = np.zeros(expr_df.shape[1], dtype=np.int32)\n",
    "    sample_class[meta_df['LibraryLayout_s'].values == layout] = 1\n",
    "    sample_class[(meta_df['LibraryLayout_s'].values == layout) & \n",
    "                 (meta_df['infection_status_s'].values == 'Zika infected')] = 2\n",
    "    platform = d_layout_platform[layout]\n",
    "    sample_classes[platform] = sample_class\n",
    "\n",
    "sample_classes['combined'] = sample_classes['MiSeq'] + sample_classes['NextSeq 500']\n",
    "print sample_classes\n",
    "\n",
    "for platform, sample_class in sample_classes.items():\n",
    "    cd_res = geode.chdir(expr_df.values, sample_class, expr_df.index, \n",
    "                      gamma=.5, sort=False, calculate_sig=False)\n",
    "    cd_coefs = np.array(map(lambda x: x[0], cd_res))\n",
    "    cd_results[platform] = cd_coefs\n",
    "    \n",
    "    # sort CD in by absolute values in descending order\n",
    "    srt_idx = np.abs(cd_coefs).argsort()[::-1]\n",
    "    cd_coefs = cd_coefs[srt_idx][:600]\n",
    "    sorted_DEGs = expr_df.index[srt_idx][:600]\n",
    "    # split up and down\n",
    "    up_genes = dict(zip(sorted_DEGs[cd_coefs > 0], cd_coefs[cd_coefs > 0]))\n",
    "    dn_genes = dict(zip(sorted_DEGs[cd_coefs < 0], cd_coefs[cd_coefs < 0]))\n",
    "    d_platform_cd[platform+'-up'] = up_genes\n",
    "    d_platform_cd[platform+'-dn'] = dn_genes\n",
    "\n",
    "print cd_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Check the cosine distance between the two signatures\n",
    "from scipy.spatial.distance import cosine\n",
    "from itertools import combinations\n",
    "for col1, col2 in combinations(cd_results.columns, 2):\n",
    "    print col1, col2, cosine(cd_results[col1], cd_results[col2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EXTRA: DELETE  later"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Prepare count matrices\n",
    "expect input data as obtained, e.g., from RNA-seq or another high-throughput sequencing experiment, in the form of a matrix of integer values. The value in the i-th row and the j-th column of the matrix tells how many reads (or fragments, for paired-end RNA-seq) have been assigned to gene i in sample j. Analogously, for other types of assays, the rows of the matrix might correspond e.g., to binding regions (with ChIP-Seq), species of bacteria (with metagenomic datasets), or peptide sequences (with quantitative mass spectrometry).\n",
    "\n",
    "The values in the matrix should be counts of sequencing reads/fragments. This is important for DESeq2â€™s statistical model to hold, as only counts allow assessing the measurement precision correctly. It is important to never provide counts that were pre-normalized for sequencing depth/library size, as the statistical model is most powerful when applied to un-normalized counts, and is designed to account for library size differences internally.\n",
    "\n",
    "## Align Reads to reference genome\n",
    "The computational analysis of an RNA-seq experiment begins from the FASTQ files that contain the nucleotide sequence of each read and a quality score at each position. These reads must first be aligned to a reference genome or transcriptome, or the abundances and estimated counts per transcript can be estimated without alignment, as described above. In either case, it is important to know if the sequencing experiment was single-end or paired-end, as the alignment software will require the user to specify both FASTQ files for a paired-end experiment. The output of this alignment step is commonly stored in a file format called SAM/BAM.\n",
    "\n",
    "\n",
    "## Define gene models\n",
    "\n",
    "## Plot counts\n",
    "\n",
    "## PCA Plot \n",
    "\n",
    "## Differential Expression Analysis\n",
    "## Gene Clustering\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from diffexp.py_deseq import py_DESeq2\n",
    "\n",
    "dds = py_DESeq2(count_matrix = df,\n",
    "               design_matrix = sample_df,\n",
    "               design_formula = '~ sample',\n",
    "               gene_column = 'id') # <- telling DESeq2 this should be the gene ID column\n",
    "    \n",
    "dds.run_deseq() \n",
    "dds.get_deseq_result()\n",
    "res = dds.deseq_result \n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.scatter(res.log2FoldChange, -np.log2(res.padj))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'dgeclust'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-8d6855043f8a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mdgeclust\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mCountData\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSimulationManager\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'dgeclust'"
     ]
    }
   ],
   "source": [
    "from dgeclust import CountData, SimulationManager"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env-tcga",
   "language": "python",
   "name": "env-tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
