{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Python for Bioinformatics\n",
    "-----------------------------\n",
    "\n",
    "![title](https://s3.amazonaws.com/py4bio/tapabiosmall.png)\n",
    "\n",
    "This Jupyter notebook is intented to be used alongside the book [Python for Bioinformatics](http://py3.us/)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Chapter 22: Drawing Marker Positions Using Data Stored in a Database\n",
    "-----------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** Before opening the file, this file should be accesible from this Jupyter notebook. In order to do so, the following commands will download these files from Github and extract them into a directory called samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 16.5M  100 16.5M    0     0  15.6M      0  0:00:01  0:00:01 --:--:-- 15.6M\n",
      "mkdir: cannot create directory 'samples': File exists\n",
      "BLAST_output.xml\n",
      "TAIR7_Transcripts_by_map_position.gz\n",
      "pMOSBlue.txt\n",
      "fishbacteria.csv\n",
      "UniVec_Core.nsq\n",
      "t3beta.fasta\n",
      "PythonU.db\n",
      "input4align.dnd\n",
      "pdb1apk.ent.gz\n",
      "readme.txt\n",
      "contig1.ace\n",
      "example.aln\n",
      "hsc1.fasta\n",
      "bioinfo/seqs/15721870.fasta\n",
      "primers.txt\n",
      "bioinfo/seqs/4586830.fasta\n",
      "bioinfo/seqs/7638455.fasta\n",
      "GSM188012.CEL\n",
      "3seqs.fas\n",
      "sampleX.fas\n",
      "sampleXblast.xml\n",
      "B1.csv\n",
      "phd1\n",
      "conglycinin.phy\n",
      "bioinfo/seqs/218744616.fasta\n",
      "spfile.txt\n",
      "bioinfo/seqs/513419.fasta\n",
      "bioinfo/seqs/513710.fasta\n",
      "prot.fas\n",
      "cas9align.fasta\n",
      "seqA.fas\n",
      "bioinfo/seqs/\n",
      "bioinfo/\n",
      "pdbaa\n",
      "other.xml\n",
      "vectorssmall.fasta\n",
      "t3.fasta\n",
      "a19.gp\n",
      "data.csv\n",
      "input4align.fasta\n",
      "B1IXL9.txt\n",
      "fasta22.fas\n",
      "bioinfo/seqs/7415878.fasta\n",
      "bioinfo/seqs/513718.fasta\n",
      "bioinfo/seqs/513719.fasta\n",
      "bioinfo/seqs/6598312.fasta\n",
      "UniVec_Core.nin\n",
      "Q5R5X8.fas\n",
      "bioinfo/seqs/513717.fasta\n",
      "BcrA.gp\n",
      "bioinfo/seqs/2623545.fasta\n",
      "bioinfo/seqs/63108399.fasta\n",
      "conglycinin.dnd\n",
      "NC2033.txt\n",
      "fishdata.csv\n",
      "uniprotrecord.xml\n",
      "BLAST_output.html\n",
      "Q9JJE1.xml\n",
      "test3.csv\n",
      "UniVec_Core.nhr\n",
      "sampledata.xlsx\n",
      "UniVec_Core\n",
      "NC_006581.gb\n",
      "conglycinin.multiple.phy\n",
      "conglycinin.fasta\n"
     ]
    }
   ],
   "source": [
    "!curl https://raw.githubusercontent.com/Serulab/Py4Bio/master/samples/samples.tar.bz2 -o samples.tar.bz2\n",
    "!mkdir samples\n",
    "!tar xvfj samples.tar.bz2 -C samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following programs needs a MongoDB database in order to run. Enter your DB parameters in the line starting with CONNECTION_STRING."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Listing 22.1:** createdb.py: Convert the data from a CSV file to insert it into\n",
    "MongoDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ServerSelectionTimeoutError",
     "evalue": "localhost:27017: [Errno 111] Connection refused",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mServerSelectionTimeoutError\u001b[0m               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-f976de1277d9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0mdb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpr4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[0mcollection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmarkers_map4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m \u001b[0mcollection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'marker_id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTEXT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m \u001b[0mcollection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert_many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmarkers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/pymongo/collection.py\u001b[0m in \u001b[0;36mcreate_index\u001b[0;34m(self, keys, **kwargs)\u001b[0m\n\u001b[1;32m   1385\u001b[0m         \u001b[0mkeys\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhelpers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_index_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1386\u001b[0m         \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefault\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"name\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhelpers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_gen_index_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1387\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__create_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1388\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/pymongo/collection.py\u001b[0m in \u001b[0;36m__create_index\u001b[0;34m(self, keys, index_options)\u001b[0m\n\u001b[1;32m   1292\u001b[0m         \u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1293\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1294\u001b[0;31m         \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_socket_for_writes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msock_info\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1295\u001b[0m             \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSON\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'createIndexes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'indexes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1296\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/contextlib.py\u001b[0m in \u001b[0;36m__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     80\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__enter__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 82\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     83\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"generator didn't yield\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/pymongo/mongo_client.py\u001b[0m in \u001b[0;36m_get_socket\u001b[0;34m(self, selector)\u001b[0m\n\u001b[1;32m    760\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mcontextlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontextmanager\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    761\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mselector\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 762\u001b[0;31m         \u001b[0mserver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_topology\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mselect_server\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mselector\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    763\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    764\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__all_credentials\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msock_info\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/pymongo/topology.py\u001b[0m in \u001b[0;36mselect_server\u001b[0;34m(self, selector, server_selection_timeout, address)\u001b[0m\n\u001b[1;32m    208\u001b[0m         return random.choice(self.select_servers(selector,\n\u001b[1;32m    209\u001b[0m                                                  \u001b[0mserver_selection_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 210\u001b[0;31m                                                  address))\n\u001b[0m\u001b[1;32m    211\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m     def select_server_by_address(self, address,\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/pymongo/topology.py\u001b[0m in \u001b[0;36mselect_servers\u001b[0;34m(self, selector, server_selection_timeout, address)\u001b[0m\n\u001b[1;32m    184\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mserver_timeout\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mnow\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mend_time\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m                     raise ServerSelectionTimeoutError(\n\u001b[0;32m--> 186\u001b[0;31m                         self._error_message(selector))\n\u001b[0m\u001b[1;32m    187\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    188\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ensure_opened\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mServerSelectionTimeoutError\u001b[0m: localhost:27017: [Errno 111] Connection refused"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import gzip\n",
    "import os\n",
    "\n",
    "from pymongo import MongoClient, TEXT\n",
    "\n",
    "FILE_NAME = 'samples/TAIR7_Transcripts_by_map_position.gz'\n",
    "CONNECTION_STRING = os.getenv('MONGODB_CS', 'localhost:27017')\n",
    "\n",
    "# Get a file handler of an uncompressed file:\n",
    "with gzip.open(FILE_NAME, \"rt\", newline=\"\") as f_unzip:\n",
    "    rows = csv.reader(f_unzip, delimiter='\\t')\n",
    "    next(rows) # Skip the header\n",
    "    # Dictionary for storing markers and associated information:\n",
    "    at_d = {}\n",
    "    # Load the dictionary using the data in the file:\n",
    "    for row in rows:\n",
    "        if row[0] in at_d:\n",
    "            chromosome, left_val, right_val = at_d[row[0]]\n",
    "            c7 = int(row[7])\n",
    "            left = c7 if c7<int(left_val) else left_val\n",
    "            c8 = int(row[8])\n",
    "            right = c8 if c8>int(right_val) else right_val\n",
    "            at_d[row[0]] = (int(chromosome), left, right)\n",
    "        else:\n",
    "            at_d[row[0]] = (int(row[5]), int(row[7]), int(row[8]))\n",
    "# Make a list with dictionaries to be stored as documents in\n",
    "# MongoDB\n",
    "markers = []\n",
    "for marker in at_d:\n",
    "    markers.append({'marker_id': marker, 'chromosome':\n",
    "                    at_d[marker][0], 'start': at_d[marker][1],\n",
    "                    'end': at_d[marker][2]})\n",
    "client = MongoClient(CONNECTION_STRING)\n",
    "db = client.pr4\n",
    "collection = db.markers_map4\n",
    "collection.create_index([('marker_id', TEXT)])\n",
    "collection.insert_many(markers)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Listing 22.2:** drawmarker.py: Draw markers in chromosomes from data extracted\n",
    "from a MongoDB database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "MissingPythonDependencyError",
     "evalue": "Please install ReportLab if you want to use Bio.Graphics. You can find ReportLab at http://www.reportlab.com/software/opensource/",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/Bio/Graphics/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mimport\u001b[0m \u001b[0mreportlab\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0;32mdel\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'reportlab'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMissingPythonDependencyError\u001b[0m              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-5319f3a06955>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpymongo\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mMongoClient\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGraphics\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mBasicChromosome\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mreportlab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlib\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcolors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mreportlab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munits\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcm\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3_431/lib/python3.6/site-packages/Bio/Graphics/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mMissingPythonDependencyError\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     raise MissingPythonDependencyError(\n\u001b[0;32m---> 16\u001b[0;31m         \u001b[0;34m\"Please install ReportLab if you want \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m         \u001b[0;34m\"to use Bio.Graphics. You can find ReportLab at \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m         \"http://www.reportlab.com/software/opensource/\")\n",
      "\u001b[0;31mMissingPythonDependencyError\u001b[0m: Please install ReportLab if you want to use Bio.Graphics. You can find ReportLab at http://www.reportlab.com/software/opensource/"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "from pymongo import MongoClient\n",
    "from Bio.Graphics import BasicChromosome\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "\n",
    "CONNECTION_STRING = os.getenv('MONGODB_CS', 'localhost:27017')\n",
    "\n",
    "def sortmarkers(crms,end):\n",
    "    \"\"\" Sort markers into chromosomes \"\"\"\n",
    "    i = 0\n",
    "    crms_o = [[] for r in range(len(end))]\n",
    "    crms_fo = [[] for r in range(len(end))]\n",
    "    for crm in crms:\n",
    "        for marker in crm:\n",
    "            # add the marker start position at each chromosome.\n",
    "            crms_fo[i].append(marker[1])\n",
    "        crms_fo[i].sort() # Sort the marker positions.\n",
    "        i += 1\n",
    "    i = 0\n",
    "    for order in crms_fo:\n",
    "        # Using the marker order set in crms_fo, fill crms_o\n",
    "        # with all the marker information\n",
    "        for pos in order:\n",
    "            for mark in crms[i]:\n",
    "                if pos==mark[1]:\n",
    "                    crms_o[i].append(mark)\n",
    "        i += 1\n",
    "    return crms_o\n",
    "\n",
    "def getchromo(crms_o, end):\n",
    "    \"\"\" From an ordered list of markers, generate chromosomes.\n",
    "    \"\"\"\n",
    "    chromo = [[] for r in range(len(end))]\n",
    "    i = 0\n",
    "    for crm_o in crms_o:\n",
    "        j = 0\n",
    "        if len(crm_o)>1:\n",
    "            for mark in crm_o:\n",
    "                if mark==crm_o[0]: #first marker\n",
    "                    chromo[i].append(('', None, mark[1]))\n",
    "                    chromo[i].append((mark[0], colors.red,\n",
    "                                      mark[2]-mark[1]))\n",
    "                    ant = mark[2]\n",
    "                elif mark==crm_o[-1]: #last marker\n",
    "                    chromo[i].append(('', None, mark[1]-ant))\n",
    "                    chromo[i].append((mark[0], colors.red,\n",
    "                                      mark[2]-mark[1]))\n",
    "                    chromo[i].append(('', None, end[i]-mark[2]))\n",
    "                else:\n",
    "                    chromo[i].append(('', None, mark[1]-ant))\n",
    "                    chromo[i].append((mark[0], colors.red,\n",
    "                                      mark[2]-mark[1]))\n",
    "                    ant=mark[2]\n",
    "        elif len(crm_o)==1: # For chromosomes with one marker\n",
    "            chromo[i].append(('', None, crm_o[0][1]))\n",
    "            chromo[i].append((crm_o[0][0], colors.red,\n",
    "                              crm_o[0][2]-crm_o[0][1]))\n",
    "            chromo[i].append(('', None, end[i]-crm_o[0][2]))\n",
    "        else:\n",
    "            # For chromosomes without markers\n",
    "            # Add 3% of each chromosome.\n",
    "            chromo[i].append(('', None, int(0.03*end[i])))\n",
    "            chromo[i].append(('', None, end[i]))\n",
    "            chromo[i].append(('', None, int(0.03*end[i])))\n",
    "        i += 1\n",
    "        j += 1\n",
    "    return chromo\n",
    "\n",
    "def addends(chromo):\n",
    "    \"\"\" Adds a 3% of blank region at both ends for better\n",
    "        graphic output.\n",
    "    \"\"\"\n",
    "    size = 0\n",
    "    for x in chromo:\n",
    "        size += x[2]\n",
    "    # get 3% of size of each chromosome:\n",
    "    endsize = int(float(size)*.03)\n",
    "    # add this size to both ends in chromo:\n",
    "    chromo.insert(0,('', None, endsize))\n",
    "    chromo.append(('', None, endsize))\n",
    "    return chromo\n",
    "\n",
    "def load_chrom(chr_name):\n",
    "    \"\"\" Generate a chromosome with information\n",
    "    \"\"\"\n",
    "    cur_chromosome = BasicChromosome.Chromosome(chr_name[0])\n",
    "    chr_segment_info = chr_name[1]\n",
    "\n",
    "    for seg_info_num in range(len(chr_segment_info)):\n",
    "        label, color, scale = chr_segment_info[seg_info_num]\n",
    "        # make the top and bottom telomeres\n",
    "        if seg_info_num == 0:\n",
    "            cur_segment = BasicChromosome.TelomereSegment()\n",
    "        elif seg_info_num == len(chr_segment_info) - 1:\n",
    "            cur_segment = BasicChromosome.TelomereSegment(1)\n",
    "        # otherwise, they are just regular segments\n",
    "        else:\n",
    "            cur_segment = BasicChromosome.ChromosomeSegment()\n",
    "        cur_segment.label = label\n",
    "        cur_segment.label_size = 12\n",
    "        cur_segment.fill_color = color\n",
    "        cur_segment.scale = scale\n",
    "        cur_chromosome.add(cur_segment)\n",
    "\n",
    "    cur_chromosome.scale_num = max(END) + (max(END)*.04)\n",
    "    return cur_chromosome\n",
    "\n",
    "def dblookup(atgids):\n",
    "    \"\"\" Code to retrieve all marker data from name using MongoDB\n",
    "    \"\"\"\n",
    "    client = MongoClient(CONNECTION_STRING)\n",
    "    db = client.pr4\n",
    "    collection = db.markers_map4\n",
    "    markers = []\n",
    "    for marker in atgids:\n",
    "        mrk = collection.find_one({'marker_id': marker})\n",
    "        if mrk:\n",
    "            markers.append((marker, (mrk['chromosome'],\n",
    "                            mrk['start'], mrk['end'])))\n",
    "        else:\n",
    "            print('Marker {0} is not in the DB'.format(marker))\n",
    "    return markers\n",
    "\n",
    "# Size of each chromosome:\n",
    "END = (30427563, 19696817, 23467989, 18581571, 26986107)\n",
    "gids = []\n",
    "rx_rid = re.compile('^AT[1-5]G\\d{5}$')\n",
    "print('''Enter AT ID or press 'enter' to stop entering IDs.\n",
    "Valid IDs:\n",
    "AT2G28000\n",
    "AT3G03020\n",
    "Also you can enter DBDEMO to use predefined set of markers\n",
    "fetched from a MongoDB database. Enter NODBDEMO to use a\n",
    "predefined set of markers without database access.''')\n",
    "while True:\n",
    "    rid = input('Enter Gene ID: ')\n",
    "    if not rid:\n",
    "        break\n",
    "    if rid=='DBDEMO':\n",
    "        gids = ['AT3G14890','AT1G66160','AT3G55260','AT5G59570',\n",
    "                'AT4G32551','AT1G01430','AT4G26000','AT2G28000',\n",
    "                'AT5G21090','AT5G10470']\n",
    "        break\n",
    "    elif rid=='NODBDEMO':\n",
    "        samplemarkers=[('AT3G14890', (3, 5008749, 5013275)),\n",
    "                       ('AT1G66160', (1, 24640827, 24642411)),\n",
    "                       ('AT3G55260', (3, 20500225, 20504056)),\n",
    "                       ('AT1G10960', (1, 3664385, 3665040)),\n",
    "                       ('AT5G23350', (5, 7857646, 7859280)),\n",
    "                       ('AT5G15250', (5, 4950414, 4952780)),\n",
    "                       ('AT1G55700', (1, 20825263, 20827306)),\n",
    "                       ('AT5G21090', (5, 7164583, 7167257)),\n",
    "                       ('AT5G10470', (5, 3289228, 3297249)),\n",
    "                       ('AT2G28000', (2, 11933524, 11936523)),\n",
    "                       ('AT3G03020', (3, 680920, 682009)),\n",
    "                       ('AT4G26000', (4, 13197255, 13199845)),\n",
    "                       ('AT4G32551', (4, 15707516, 15713587))]\n",
    "        break\n",
    "    if rx_rid.match(rid):\n",
    "        gids.append(rid)\n",
    "    else:\n",
    "        print(\"Bad format, please enter it again\")\n",
    "\n",
    "if rid!='NODBDEMO':\n",
    "    samplemarkers = dblookup(gids)\n",
    "\n",
    "crms = [[] for r in range(len(END))]\n",
    "for x in samplemarkers:\n",
    "    crms[int(x[1][0])-1].append((x[0], x[1][1], x[1][2]))\n",
    "\n",
    "crms_o = sortmarkers(crms, END)\n",
    "chromo = getchromo(crms_o, END)\n",
    "all_chr_info = [('Chr I', chromo[0]), ('Chr II', chromo[1]),\n",
    "                ('Chr III', chromo[2]), ('Chr IV', chromo[3]),\n",
    "                ('Chr V', chromo[4])]\n",
    "\n",
    "organism = BasicChromosome.Organism()\n",
    "organism.page_size = (29.7*cm, 21*cm) #A4 landscape\n",
    "for chr_info in all_chr_info:\n",
    "    newcrom = (chr_info[0], addends(chr_info[1]))\n",
    "    organism.add(load_chrom(newcrom))\n",
    "\n",
    "organism.draw('at.pdf','Arabidopsis thaliana')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6",
   "language": "python",
   "name": "python36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
