{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do not use the tree house version https://github.com/rcurrie/tumornormal/blob/master/ingest.ipynb\n",
    "\n",
    "The only file you need to run other notesbooks is ../data/tcga_target_gtex.h5\n",
    "\n",
    "you can delete the following files to save a little disk space\n",
    "\n",
    "```\n",
    "cd data\n",
    "rm TcgaTargetGTEX_phenotype.txt.gz\n",
    "rm TcgaTargetGtex_rsem_gene_tpm.gz \n",
    "rm TcgaTargetGtex_rsem_gene_tpm.h5\n",
    "```\n",
    "\n",
    "remove sample if category or target is missing\n",
    "\n",
    "you might want drop additional gene expression features if there are a large number of log2(0.001) values. It means the original data had zero TPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO <ipython-input-1-27500482ad0e>:8 - <module>()] using logging configuration file:../src/logging.test.ini.json\n"
     ]
    }
   ],
   "source": [
    "import pathlib as pl\n",
    "\n",
    "import logging\n",
    "from   setupLogging import setupLogging\n",
    "configFilePath = setupLogging( default_path=pl.Path('../src/logging.ini.json'))\n",
    "configFilePath = setupLogging( default_path=pl.Path('../src/logging.test.ini.json'))\n",
    "logger = logging.getLogger(\"notebook\")\n",
    "logger.info(\"using logging configuration file:{}\".format(configFilePath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pathlib as pl\n",
    "import h5py\n",
    "import sys\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "\n",
    "dataDir = pl.Path(\"../data\")\n",
    "if not dataDir.exists() :\n",
    "    dataDir.mkdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "see src/dataUtilities for python script version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Down load TCGA-Target-GTex expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\n",
      "url:https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\n",
      "../data/TcgaTargetGtex_rsem_gene_tpm.gz\n",
      "CPU times: user 3.25 s, sys: 2.99 s, total: 6.24 s\n",
      "Wall time: 2min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def downLoadGeneExpressionFromXenaHub(dataDir):\n",
    "    '''\n",
    "    will down load data if it is not on local disk . This can take a while\n",
    "    \n",
    "    arguments:\n",
    "        dataDir:\n",
    "            path to local data directory. type pathlib\n",
    "            example: '../data'\n",
    "            \n",
    "    returns:\n",
    "        rawDataPath, path to file on local system. type pathlib\n",
    "    '''\n",
    "    rawDataPath = dataDir.joinpath(\"TcgaTargetGtex_rsem_gene_tpm.gz\")\n",
    "    if not rawDataPath.exists():\n",
    "        print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "        url = \"https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\"\n",
    "        print(\"url:{}\".format(url))\n",
    "        response = requests.get(\n",
    "                url, \n",
    "                stream=True,\n",
    "                # ERROR: cannot verify toil.xenahubs.net's certificate, issued by `/C=US/O=Amazon/OU=Server CA 1B/CN=Amazon':\n",
    "                #verify=False \n",
    "                ) \n",
    "        response.raise_for_status()\n",
    "        #totalLengthStr = response.headers.get('content-length')\n",
    "        #totalLength = int(totalLengthStr)\n",
    "\n",
    "        with open(rawDataPath, \"wb\") as f:\n",
    "            dataLength = 0\n",
    "            for chunk in response.iter_content(chunk_size=32768):\n",
    "                dataLength += len(chunk)\n",
    "                f.write(chunk)\n",
    "                \n",
    "    return rawDataPath\n",
    "\n",
    "rawDataPath = downLoadGeneExpressionFromXenaHub(dataDir)\n",
    "print(rawDataPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sourceFile:../data/TcgaTargetGtex_rsem_gene_tpm.h5\n",
      "[INFO utils.py:129 - _init_num_threads()] Note: NumExpr detected 16 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "[INFO utils.py:141 - _init_num_threads()] NumExpr defaulting to 8 threads.\n",
      "CPU times: user 14.1 s, sys: 3.61 s, total: 17.7 s\n",
      "Wall time: 17.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def loadRawExpressionDF(dataDir, rawDataPath):\n",
    "    '''\n",
    "    This can be slow if h5 verision does not exist locally, will create data from gz\n",
    "    \n",
    "    arguments:\n",
    "        dataDir:\n",
    "            path to local data directory. type pathlib\n",
    "            example: ../data/TcgaTargetGtex_rsem_gene_tpm.gz\n",
    "            \n",
    "    returns\n",
    "        expressionDF: type pandas data frame\n",
    "        The shape of the data frame matches the xena hub data\n",
    "        E.G. geneExpression x samples\n",
    "    '''    \n",
    "    # AEDWIP load gene expresion pandas data frame\n",
    "    pandasDataSourcePath = dataDir.joinpath(\"TcgaTargetGtex_rsem_gene_tpm.h5\")\n",
    "    print(\"sourceFile:{}\".format(pandasDataSourcePath))\n",
    "    if not pandasDataSourcePath.exists():\n",
    "        print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "        expressionDF = pd.read_csv(rawDataPath, sep=\"\\t\", index_col=0) \\\n",
    "                                    .astype(np.float32)\n",
    "        \n",
    "        expressionDF.to_hdf(pandasDataSourcePath, \n",
    "                                       \"expression\", mode=\"w\", format=\"fixed\")\n",
    "            \n",
    "    else : \n",
    "        expressionDF = pd.read_hdf(pandasDataSourcePath, \"expression\") \n",
    "    \n",
    "    return expressionDF\n",
    "\n",
    "expressionDF = loadRawExpressionDF(dataDir, rawDataPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sanity check all values are supposed to be log2(TMP + 0.001)\n",
    "result = (expressionDF.isnull().sum().sum()) \n",
    "assert (result == 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numGeneExpr:60498, numSamples:19131 \n"
     ]
    }
   ],
   "source": [
    "numGeneExpr, numSamples,  = expressionDF.shape\n",
    "print(\"numGeneExpr:{}, numSamples:{} \".format(numGeneExpr, numSamples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-S4Q7-0003-SM-3NM8M</th>\n",
       "      <th>TCGA-19-1787-01</th>\n",
       "      <th>TCGA-S9-A7J2-01</th>\n",
       "      <th>GTEX-QV31-1626-SM-2S1QC</th>\n",
       "      <th>TCGA-G3-A3CH-11</th>\n",
       "      <th>TCGA-B5-A5OE-01</th>\n",
       "      <th>GTEX-13QIC-0011-R1a-SM-5O9CJ</th>\n",
       "      <th>TCGA-B2-5641-11</th>\n",
       "      <th>GTEX-ZPCL-0126-SM-4WWC8</th>\n",
       "      <th>TARGET-20-PANGDN-09</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-FI-A2EY-01</th>\n",
       "      <th>TCGA-55-6985-11</th>\n",
       "      <th>TCGA-EJ-5527-01</th>\n",
       "      <th>TCGA-G3-A25X-01</th>\n",
       "      <th>TCGA-24-2254-01</th>\n",
       "      <th>GTEX-11ZTS-3326-SM-5LU9Y</th>\n",
       "      <th>GTEX-VJYA-0726-SM-4KL1T</th>\n",
       "      <th>GTEX-ZA64-2126-SM-5Q5A8</th>\n",
       "      <th>GTEX-Q2AG-2826-SM-2HMJQ</th>\n",
       "      <th>GTEX-XV7Q-0426-SM-4BRVN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <td>-3.4580</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.2998</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0350</td>\n",
       "      <td>-2.0529</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.9379</td>\n",
       "      <td>1.5165</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.3884</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>-3.3076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000259041.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-3.0469</td>\n",
       "      <td>-2.1779</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.8314</td>\n",
       "      <td>-2.4659</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.1714</td>\n",
       "      <td>-4.2934</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>-2.9324</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>2.5852</td>\n",
       "      <td>-3.0469</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19131 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GTEX-S4Q7-0003-SM-3NM8M  TCGA-19-1787-01  TCGA-S9-A7J2-01  \\\n",
       "sample                                                                         \n",
       "ENSG00000242268.2                  -3.4580          -9.9658           0.2998   \n",
       "ENSG00000259041.1                  -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000270112.3                  -3.6259          -3.8160          -3.0469   \n",
       "\n",
       "                   GTEX-QV31-1626-SM-2S1QC  TCGA-G3-A3CH-11  TCGA-B5-A5OE-01  \\\n",
       "sample                                                                         \n",
       "ENSG00000242268.2                  -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000259041.1                  -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000270112.3                  -2.1779          -9.9658          -9.9658   \n",
       "\n",
       "                   GTEX-13QIC-0011-R1a-SM-5O9CJ  TCGA-B2-5641-11  \\\n",
       "sample                                                             \n",
       "ENSG00000242268.2                       -9.9658          -9.9658   \n",
       "ENSG00000259041.1                       -9.9658          -9.9658   \n",
       "ENSG00000270112.3                       -1.8314          -2.4659   \n",
       "\n",
       "                   GTEX-ZPCL-0126-SM-4WWC8  TARGET-20-PANGDN-09  ...  \\\n",
       "sample                                                           ...   \n",
       "ENSG00000242268.2                  -9.9658              -9.9658  ...   \n",
       "ENSG00000259041.1                  -9.9658              -9.9658  ...   \n",
       "ENSG00000270112.3                  -9.9658              -9.9658  ...   \n",
       "\n",
       "                   TCGA-FI-A2EY-01  TCGA-55-6985-11  TCGA-EJ-5527-01  \\\n",
       "sample                                                                 \n",
       "ENSG00000242268.2          -4.0350          -2.0529          -9.9658   \n",
       "ENSG00000259041.1          -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000270112.3          -3.1714          -4.2934          -9.9658   \n",
       "\n",
       "                   TCGA-G3-A25X-01  TCGA-24-2254-01  GTEX-11ZTS-3326-SM-5LU9Y  \\\n",
       "sample                                                                          \n",
       "ENSG00000242268.2          -9.9658          -1.9379                    1.5165   \n",
       "ENSG00000259041.1          -9.9658          -9.9658                   -9.9658   \n",
       "ENSG00000270112.3          -9.9658          -5.5735                   -2.9324   \n",
       "\n",
       "                   GTEX-VJYA-0726-SM-4KL1T  GTEX-ZA64-2126-SM-5Q5A8  \\\n",
       "sample                                                                \n",
       "ENSG00000242268.2                  -9.9658                  -2.3884   \n",
       "ENSG00000259041.1                  -9.9658                  -9.9658   \n",
       "ENSG00000270112.3                  -9.9658                  -9.9658   \n",
       "\n",
       "                   GTEX-Q2AG-2826-SM-2HMJQ  GTEX-XV7Q-0426-SM-4BRVN  \n",
       "sample                                                               \n",
       "ENSG00000242268.2                   0.0440                  -3.3076  \n",
       "ENSG00000259041.1                  -9.9658                  -9.9658  \n",
       "ENSG00000270112.3                   2.5852                  -3.0469  \n",
       "\n",
       "[3 rows x 19131 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expressionDF.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Down load meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/TcgaTargetGTEX_phenotype.txt.gz\n",
      "CPU times: user 206 µs, sys: 90 µs, total: 296 µs\n",
      "Wall time: 238 µs\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def downLoadPhenotypeFromXenaHub(dataDir):\n",
    "    '''\n",
    "    will down load data if it is not on local disk . This can take a while\n",
    "    \n",
    "    arguments:\n",
    "        dataDir:\n",
    "            path to local data directory. type pathlib\n",
    "            example: '../data'\n",
    "            \n",
    "    returns:\n",
    "        rawDataPath, path to file on local system. type pathlib\n",
    "    '''\n",
    "    rawDataPath = dataDir.joinpath(\"TcgaTargetGTEX_phenotype.txt.gz\")\n",
    "    if not rawDataPath.exists():\n",
    "        url = \"https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz\"\n",
    "        print(\"Downloading {}\".format(url))\n",
    "        with open(rawDataPath, \"wb\") as f:\n",
    "            f.write(requests.get(url).content)\n",
    "                \n",
    "    return rawDataPath\n",
    "\n",
    "phenoTypePath = downLoadPhenotypeFromXenaHub(dataDir)\n",
    "print(phenoTypePath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 244 ms, sys: 5.21 ms, total: 249 ms\n",
      "Wall time: 250 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def loadRawPhenotypeDF(dataDir, rawDataPath):\n",
    "    '''\n",
    "    This can be slow if h5 verision does not exist locally, will create data from gz\n",
    "    \n",
    "    arguments:\n",
    "        dataDir:\n",
    "            path to local data directory. type pathlib\n",
    "            example: .../data/TcgaTargetGTEX_phenotype.txt.gz\n",
    "    '''   \n",
    "    \n",
    "    # rename original column names\n",
    "    # Index(['detailed_category', 'primary disease or tissue', '_primary_site',\n",
    "    #  '_sample_type', '_gender', '_study'],\n",
    "    # dtype='object'\n",
    "    \n",
    "    colNames = [\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"]\n",
    "    phenotypeDF = pd.read_table(\n",
    "                        rawDataPath, compression=\"gzip\", \n",
    "                        header=0, \n",
    "                        names=colNames,\n",
    "                        sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, \n",
    "                        dtype=\"str\").sort_index(axis=\"index\")\n",
    "    \n",
    "    # Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "    phenotypeDF[\"tumor_normal\"] = phenotypeDF.apply(\n",
    "        lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)  \n",
    "    \n",
    "    # debug\n",
    "    # for colName in phenoTypeDF.columns:\n",
    "        # print(\"\\n:{}\".format(colName))\n",
    "        # print( phenoTypeDF.loc[:,[colName]].isnull().sum() )\n",
    "    \n",
    "    return phenotypeDF\n",
    "\n",
    "  \n",
    "phenoTypeDF = loadRawPhenotypeDF(dataDir, phenoTypePath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['category', 'disease', 'primary_site', 'sample_type', 'gender', 'study',\n",
      "       'tumor_normal'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(phenoTypeDF.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>Artery - Tibial</td>\n",
       "      <td>Artery - Tibial</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>Artery - Coronary</td>\n",
       "      <td>Artery - Coronary</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>Heart - Atrial Appendage</td>\n",
       "      <td>Heart - Atrial Appendage</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          category                   disease  \\\n",
       "id                                                                             \n",
       "GTEX-1117F-0226-SM-5GZZ7    Adipose - Subcutaneous    Adipose - Subcutaneous   \n",
       "GTEX-1117F-0426-SM-5EGHI         Muscle - Skeletal         Muscle - Skeletal   \n",
       "GTEX-1117F-0526-SM-5EGHJ           Artery - Tibial           Artery - Tibial   \n",
       "GTEX-1117F-0626-SM-5N9CS         Artery - Coronary         Artery - Coronary   \n",
       "GTEX-1117F-0726-SM-5GIEN  Heart - Atrial Appendage  Heart - Atrial Appendage   \n",
       "\n",
       "                            primary_site    sample_type  gender study  \\\n",
       "id                                                                      \n",
       "GTEX-1117F-0226-SM-5GZZ7  Adipose Tissue  Normal Tissue  Female  GTEX   \n",
       "GTEX-1117F-0426-SM-5EGHI          Muscle  Normal Tissue  Female  GTEX   \n",
       "GTEX-1117F-0526-SM-5EGHJ    Blood Vessel  Normal Tissue  Female  GTEX   \n",
       "GTEX-1117F-0626-SM-5N9CS    Blood Vessel  Normal Tissue  Female  GTEX   \n",
       "GTEX-1117F-0726-SM-5GIEN           Heart  Normal Tissue  Female  GTEX   \n",
       "\n",
       "                         tumor_normal  \n",
       "id                                     \n",
       "GTEX-1117F-0226-SM-5GZZ7       Normal  \n",
       "GTEX-1117F-0426-SM-5EGHI       Normal  \n",
       "GTEX-1117F-0526-SM-5EGHJ       Normal  \n",
       "GTEX-1117F-0626-SM-5N9CS       Normal  \n",
       "GTEX-1117F-0726-SM-5GIEN       Normal  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenoTypeDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ":category\n",
      "category    1\n",
      "dtype: int64\n",
      "\n",
      ":disease\n",
      "disease    1\n",
      "dtype: int64\n",
      "\n",
      ":primary_site\n",
      "primary_site    5\n",
      "dtype: int64\n",
      "\n",
      ":sample_type\n",
      "sample_type    0\n",
      "dtype: int64\n",
      "\n",
      ":gender\n",
      "gender    159\n",
      "dtype: int64\n",
      "\n",
      ":study\n",
      "study    0\n",
      "dtype: int64\n",
      "\n",
      ":tumor_normal\n",
      "tumor_normal    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for colName in phenoTypeDF.columns:\n",
    "    print(\"\\n:{}\".format(colName))\n",
    "    print( phenoTypeDF.loc[:,[colName]].isnull().sum() )\n",
    "    #print(phenoTypeDF.isnull().sum().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dropRowsIfMissingMetadata(geneExprDF, phenoDF): \n",
    "    '''\n",
    "    drops rows from the data frames where 'category','disease' is null\n",
    "    Note, drop is not done 'inplace'\n",
    "    \n",
    "    arguments\n",
    "        geneExprDF, phenoDF:\n",
    "            pandas data frames. Assume the rows are samples\n",
    "            geneExprDF has gene expresion data\n",
    "            phenoDF has meta data. E.G. labels\n",
    "            \n",
    "    returns (cleanGeneExprDF, cleanPhenoDF [null row idx])\n",
    "        bad row idx are integer\n",
    "    '''\n",
    "\n",
    "    tmpDF = phenoDF.loc[:, ['category','disease']]\n",
    "    # https://stackoverflow.com/a/14033137\n",
    "    nullRowIdx = pd.isnull(tmpDF).any(1).to_numpy().nonzero()[0]\n",
    "    \n",
    "    cleanGeneExprDF = geneExprDF.drop( geneExprDF.index[ nullRowIdx ], inplace=False )\n",
    "    cleanPhenoDF    =    phenoDF.drop( phenoDF.index[ nullRowIdx ],    inplace=False )\n",
    "    \n",
    "    return (cleanGeneExprDF, cleanPhenoDF, nullRowIdx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def testDropRowsIfMissingMetadata():\n",
    "    v = np.reshape( np.array( np.arange(1,18 + 1)), (6,3) )\n",
    "    testExpDF = pd.DataFrame(v\n",
    "                       ,columns=['a', 'b', 'c' ],\n",
    "                       index=['m', 'n', 'o', 'p', 'q', 'r' ]\n",
    "                      )\n",
    "    #print(testExpDF)\n",
    "    display(HTML(testExpDF.to_html()))\n",
    "\n",
    "\n",
    "    # testPheDF = pd.DataFrame(aedwp,\n",
    "    #                    columns=['category', 'disease', 'foo' ],\n",
    "    #                    index=['m', 'n', 'o', 'p', ]\n",
    "    #                   )\n",
    "\n",
    "    testPheDF = phenoTypeDF.iloc[0:6, [0, 1, 2] ].copy(deep=True)\n",
    "    # display(HTML(testPheDF.to_html()))\n",
    "\n",
    "    testPheDF.iloc[[1,3], 0] = np.NaN\n",
    "    testPheDF.iloc[[2], 1] = np.NaN\n",
    "    display(HTML(testPheDF.to_html()))\n",
    "\n",
    "    print('\\n\\n\\n\\n ********** BEGIN test ******')\n",
    "    cleanGeneExprDF, cleanPhenoDF, nullRowIdx = dropRowsIfMissingMetadata(testExpDF, testPheDF)\n",
    "    display(HTML(cleanGeneExprDF.to_html()))\n",
    "    display(HTML(cleanPhenoDF.to_html()))\n",
    "    print('nullRowIdx:{}'.format(nullRowIdx))\n",
    "    \n",
    "# testDropRowsIfMissingMetadata()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before transpose numGeneExpr:60498, numSamples:19131 \n",
      "after  transpose numGeneExpr:19131, numSamples:60498 \n",
      "\n",
      "phenoTypeDF.shape before drop:(19131, 7)\n",
      "expressionDF after drop numGeneExpr:19130, numSamples:60498 \n",
      "phenoTypeDF.shape after   drop:(19130, 7)\n",
      "nullRowIdx:[8727]\n"
     ]
    }
   ],
   "source": [
    "# we transpose the expression data so that it is in a machine learning\n",
    "# friendly shape. \n",
    "numGeneExpr, numSamples,= expressionDF.shape\n",
    "print(\"before transpose numGeneExpr:{}, numSamples:{} \".format(numGeneExpr, numSamples))\n",
    "\n",
    "expressionDF = expressionDF.transpose()\n",
    "numGeneExpr, numSamples, = expressionDF.shape\n",
    "print(\"after  transpose numGeneExpr:{}, numSamples:{} \".format(numGeneExpr, numSamples))\n",
    "\n",
    "print()\n",
    "print(\"phenoTypeDF.shape before drop:{}\".format(phenoTypeDF.shape))\n",
    "expressionDF, phenoTypeDF, nullRowIdx = dropRowsIfMissingMetadata(expressionDF, phenoTypeDF)\n",
    "numGeneExpr, numSamples, = expressionDF.shape\n",
    "print(\"expressionDF after drop numGeneExpr:{}, numSamples:{} \".format(numGeneExpr, numSamples))\n",
    "print(\"phenoTypeDF.shape after   drop:{}\".format(phenoTypeDF.shape))\n",
    "print(\"nullRowIdx:{}\".format(nullRowIdx))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/tcga_target_gtex.h5\n"
     ]
    }
   ],
   "source": [
    "def saveTCGA_Target_GtextDataFrames(dataDir, exprDF, labelDF):\n",
    "    '''\n",
    "    export data in h5 file format. exprDF will be transposed before write\n",
    "    arguments:\n",
    "        dataDir:\n",
    "            directory on local system to write file to\n",
    "            type: Path\n",
    "            example \"../data\"\n",
    "            \n",
    "        exprDF, labelDF\n",
    "            type: pandas data frames\n",
    "            \n",
    "    returns:\n",
    "        tcgaTargetGtexPath:\n",
    "            type: Path\n",
    "    '''\n",
    "    tcgaTargetGtexPath = dataDir.joinpath(\"tcga_target_gtex.h5\")\n",
    "\n",
    "    with pd.HDFStore(tcgaTargetGtexPath, \"w\") as store:\n",
    "        store[\"expression\"] = exprDF.sort_index(axis=\"columns\")\n",
    "        store[\"labels\"] = labelDF.astype(str)\n",
    "\n",
    "    return tcgaTargetGtexPath\n",
    "\n",
    "tcgaTargetGtexPath = saveTCGA_Target_GtextDataFrames(dataDir, expressionDF, phenoTypeDF)\n",
    "print(tcgaTargetGtexPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test end to end load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadTCGA_Target_GTex(dataDir):\n",
    "    '''\n",
    "    Download gene expression and clinical data from the UCSC Xena Toil re-compute \n",
    "    dataset wrangle, and store in an hdf5 file for quick loading machine learning. \n",
    "    This dataset comprises gene expression data for twenty thousand tumor and \n",
    "    normal samples processed using the exact same genomics pipeline and therefore \n",
    "    can be compared to each other.\n",
    "\n",
    "    Each of the source data set consists of a float vector, log2(TPM+0.001) \n",
    "\n",
    "    samples missing either the 'category' or 'disease'labels are dropped. \n",
    "    There should not be an NaN values\n",
    "\n",
    "    The first time you run this function will be slow. On a fast machine it may take\n",
    "    20 min. It will down load data files from xena and run pipe like \n",
    "    \n",
    "    The only file you need locally is tcga_target_gtex.h5\n",
    "    \n",
    "    arguments:\n",
    "        dataDir: a Path object. E.G ../data\n",
    "        \n",
    "    returns:XDF, YDF objects of type pandas DataFrame\n",
    "        XDF: \n",
    "            numberOfSamples = 19,130\n",
    "            number of gene expression features = 60,498\n",
    "            \n",
    "        YDF:\n",
    "            numberOfSamples = 19,130\n",
    "            number of label features = 7\n",
    "        \n",
    "\n",
    "    you can delete the following files to save a little disk space\n",
    "\n",
    "    ```\n",
    "    cd data\n",
    "    rm TcgaTargetGTEX_phenotype.txt.gz\n",
    "    rm TcgaTargetGtex_rsem_gene_tpm.gz \n",
    "    rm TcgaTargetGtex_rsem_gene_tpm.h5\n",
    "    ```\n",
    "\n",
    "    TODO: \n",
    "    explore gene expression values. \n",
    "    if a features has a large number of log2(0.001) values it means the original \n",
    "    data had zero TPM\n",
    "    \n",
    "   refferences:\n",
    "        - https://xenabrowser.net/datapages/?host=https://toil.xenahubs.net\n",
    "        - https://github.com/rcurrie/tumornormal/blob/master/ingest.ipynb\n",
    "    '''\n",
    "    \n",
    "    tcgaTargetGtexPath = dataDir.joinpath(\"tcga_target_gtex.h5\")\n",
    "    if tcgaTargetGtexPath.exists() :\n",
    "        XDF = pd.read_hdf(tcgaTargetGtexPath, \"expression\")\n",
    "        YDF = pd.read_hdf(tcgaTargetGtexPath, \"labels\")\n",
    "        \n",
    "    else:\n",
    "        rawDataPath = downLoadGeneExpressionFromXenaHub(dataDir)\n",
    "        print(rawDataPath)\n",
    "\n",
    "        XDF= loadRawExpressionDF(dataDir, rawDataPath)\n",
    "\n",
    "        phenoTypePath = downLoadPhenotypeFromXenaHub(dataDir)\n",
    "        print(phenoTypePath)\n",
    "\n",
    "        YDF = loadRawPhenotypeDF(dataDir, phenoTypePath)\n",
    "        \n",
    "        # shape must match Y, we also want in a ML friend format\n",
    "        # i.e. rows are samples\n",
    "        XDF = XDF.transpose()\n",
    "        \n",
    "        XDF, YDF, nullRowIdx = dropRowsIfMissingMetadata(XDF, YDF)\n",
    "        print(\"nullRowIdx:{}\".format(nullRowIdx))\n",
    "\n",
    "        tcgaTargetGtexPath = saveTCGA_Target_GtextDataFrames(dataDir, XDF, YDF)\n",
    "        print(tcgaTargetGtexPath)\n",
    "        \n",
    "    return(XDF, YDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\n",
      "url:https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\n",
      "../data/TcgaTargetGtex_rsem_gene_tpm.gz\n",
      "sourceFile:../data/TcgaTargetGtex_rsem_gene_tpm.h5\n",
      "Converting expression to dataframe and storing in hdf5 file\n",
      "Downloading https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz\n",
      "../data/TcgaTargetGTEX_phenotype.txt.gz\n",
      "nullRowIdx:[8727]\n",
      "../data/tcga_target_gtex.h5\n",
      "XDF after drop numGeneExpr:19130, numSamples:60498 \n",
      "YDF.shape after   drop:(19130, 7)\n",
      "CPU times: user 11min 50s, sys: 50.8 s, total: 12min 41s\n",
      "Wall time: 14min 35s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "XDF, YDF = loadTCGA_Target_GTex(dataDir)\n",
    "numGeneExpr, numSamples, = XDF.shape\n",
    "print(\"XDF after drop numGeneExpr:{}, numSamples:{} \".format(numGeneExpr, numSamples))\n",
    "print(\"YDF.shape after   drop:{}\".format(YDF.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
