{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ec2-user/SageMaker/projects/gnn/GNN-material/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'src'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[25], line 16\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# Make sure everything is on the path.\u001b[39;00m\n\u001b[1;32m     14\u001b[0m sys\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39minsert(\u001b[38;5;241m1\u001b[39m, os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(sys\u001b[38;5;241m.\u001b[39mpath[\u001b[38;5;241m0\u001b[39m], \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../src\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m---> 16\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msrc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpreprocess\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbuild_drug_response_matrix\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m get_gdsc_gene_expression\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msrc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mglobal_vars\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m PlottingParameters\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'src'"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "import copy\n",
    "\n",
    "sns.set_theme(style=\"white\")\n",
    "\n",
    "import sys\n",
    "import os\n",
    "# Make sure everything is on the path.\n",
    "sys.path.insert(1, os.path.join(sys.path[0], '../src'))\n",
    "\n",
    "from src.preprocess.build_drug_response_matrix import get_gdsc_gene_expression\n",
    "from src.utils.global_vars import PlottingParameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conda 4.8.4\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%conda --version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we are going to create all the datasets.\n",
    "\n",
    "Datasets used for the __drug response matrix__ (`../data/processed/drm_full.pkl`):\n",
    "- `../data/raw/GDSC2_fitted_dose_response_25Feb20.xlsx`\n",
    "- `../data/raw/GDSC1_fitted_dose_response_25Feb20.xlsx`\n",
    "- `../data/raw/GDSC1_public_raw_data_25Feb20.csv`\n",
    "- `../data/raw/GDSC2_public_raw_data_25Feb20.csv`\n",
    "\n",
    "Datasets used for the __gene expression dataset__ (`../data/processed/gexpr_full.pkl`):\n",
    "- `../data/raw/Cell_line_RMA_proc_basalExp.txt`\n",
    "- `../data/raw/Cell_Lines_Details.xlsx`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "Datasets used for the __copy number picnic dataset__ (`../data/processed/cnvp_full.pkl`): \n",
    "- `../data/raw/cnv_abs_copy_number_picnic_20191101.csv`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "Datasets used for the __copy number gistic dataset__ (`../data/processed/cnvg_full.pkl`): \n",
    "- `../data/raw/cnv_gistic_20191101.csv`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "\n",
    "Datasets used for the __mutation dataset__ (`../data/processed/mut_full.pkl`): \n",
    "- `../data/raw/mutations_all_20220315.zip`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW_DATA_PATH = '../../data/raw/'\n",
    "PROCESSED_DATA_PATH = '../../data/processed/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Raw and Processed Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IC50's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `GDSC1_fitted_dose_response_24Jul22.xlsx` took 41.02563 seconds to import. It has shape (333161, 19)\n",
      "File `GDSC2_fitted_dose_response_24Jul22.xlsx` took 29.60134 seconds to import. It has shape (242036, 19)\n"
     ]
    }
   ],
   "source": [
    "GDSC1_IC50_FILE = 'GDSC1_fitted_dose_response_24Jul22.xlsx'\n",
    "GDSC2_IC50_FILE = 'GDSC2_fitted_dose_response_24Jul22.xlsx'\n",
    "\n",
    "# Read the IC50 files.\n",
    "# GDSC1\n",
    "start = time.time()\n",
    "gdsc1_ic50s = pd.read_excel(f'{RAW_DATA_PATH}{GDSC1_IC50_FILE}', header=0)\n",
    "print(f\"File `{GDSC1_IC50_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc1_ic50s.shape}\")\n",
    "\n",
    "# GDSC2\n",
    "start = time.time()\n",
    "gdsc2_ic50s = pd.read_excel(f'{RAW_DATA_PATH}{GDSC2_IC50_FILE}', header=0)\n",
    "print(f\"File `{GDSC2_IC50_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc2_ic50s.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>361</td>\n",
       "      <td>17635802</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.966813</td>\n",
       "      <td>0.985678</td>\n",
       "      <td>0.026081</td>\n",
       "      <td>1.299144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>361</td>\n",
       "      <td>17636176</td>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.692090</td>\n",
       "      <td>0.972690</td>\n",
       "      <td>0.110059</td>\n",
       "      <td>0.156076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>361</td>\n",
       "      <td>17636568</td>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.477990</td>\n",
       "      <td>0.944459</td>\n",
       "      <td>0.087019</td>\n",
       "      <td>-0.035912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0   GDSC1             361       17635802     684057            ES5   \n",
       "1   GDSC1             361       17636176     684059            ES7   \n",
       "2   GDSC1             361       17636568     684062          EW-11   \n",
       "\n",
       "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID  DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0       SIDM00263  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "1       SIDM00269  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "2       SIDM00203  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "\n",
       "     PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
       "0  EGFR signaling        1045          Y  0.007813       2.0  3.966813   \n",
       "1  EGFR signaling        1045          Y  0.007813       2.0  2.692090   \n",
       "2  EGFR signaling        1045          Y  0.007813       2.0  2.477990   \n",
       "\n",
       "        AUC      RMSE   Z_SCORE  \n",
       "0  0.985678  0.026081  1.299144  \n",
       "1  0.972690  0.110059  0.156076  \n",
       "2  0.944459  0.087019 -0.035912  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc1_ic50s.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | DATASET   |   NLME_RESULT_ID |   NLME_CURVE_ID |   COSMIC_ID | CELL_LINE_NAME   | SANGER_MODEL_ID   | TCGA_DESC    |   DRUG_ID | DRUG_NAME   | PUTATIVE_TARGET   | PATHWAY_NAME   |   COMPANY_ID | WEBRELEASE   |   MIN_CONC |   MAX_CONC |   LN_IC50 |      AUC |     RMSE |   Z_SCORE |\n",
      "|---:|:----------|-----------------:|----------------:|------------:|:-----------------|:------------------|:-------------|----------:|:------------|:------------------|:---------------|-------------:|:-------------|-----------:|-----------:|----------:|---------:|---------:|----------:|\n",
      "|  0 | GDSC1     |              361 |        17635802 |      684057 | ES5              | SIDM00263         | UNCLASSIFIED |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   3.96681 | 0.985678 | 0.026081 |  1.29914  |\n",
      "|  1 | GDSC1     |              361 |        17636176 |      684059 | ES7              | SIDM00269         | UNCLASSIFIED |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   2.69209 | 0.97269  | 0.110059 |  0.156076 |\n",
      "|  2 | GDSC1     |              361 |        17636568 |      684062 | EW-11            | SIDM00203         | UNCLASSIFIED |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   2.47799 | 0.944459 | 0.087019 | -0.035912 |\n",
      "\n",
      "|    | DATASET   |   NLME_RESULT_ID |   NLME_CURVE_ID |   COSMIC_ID | CELL_LINE_NAME   | SANGER_MODEL_ID   | TCGA_DESC    |   DRUG_ID | DRUG_NAME    | PUTATIVE_TARGET   | PATHWAY_NAME    |   COMPANY_ID | WEBRELEASE   |   MIN_CONC |   MAX_CONC |   LN_IC50 |      AUC |     RMSE |   Z_SCORE |\n",
      "|---:|:----------|-----------------:|----------------:|------------:|:-----------------|:------------------|:-------------|----------:|:-------------|:------------------|:----------------|-------------:|:-------------|-----------:|-----------:|----------:|---------:|---------:|----------:|\n",
      "|  0 | GDSC2     |              401 |        18945558 |      683667 | PFSK-1           | SIDM01132         | MB           |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |     0.0001 |        0.1 |  -1.46215 | 0.930105 | 0.088999 |  0.432482 |\n",
      "|  1 | GDSC2     |              401 |        18945796 |      684052 | A673             | SIDM00848         | UNCLASSIFIED |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |     0.0001 |        0.1 |  -4.86945 | 0.614932 | 0.111423 | -1.42032  |\n",
      "|  2 | GDSC2     |              401 |        18946078 |      684057 | ES5              | SIDM00263         | UNCLASSIFIED |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |     0.0001 |        0.1 |  -3.36068 | 0.790953 | 0.142754 | -0.599894 |\n"
     ]
    }
   ],
   "source": [
    "print(gdsc1_ic50s.head(3).to_markdown())\n",
    "print()\n",
    "print(gdsc2_ic50s.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "## Raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/kv/l0jzxgbj1kggff_kd35bfzkw0000gn/T/ipykernel_10142/1587790020.py:7: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gdsc1_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC1_RAW_FILE}', header=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `GDSC1_public_raw_data_25Feb20.csv` took 14.88413 seconds to import. It has shape (5837703, 18)\n",
      "File `GDSC2_public_raw_data_25Feb20.csv` took 15.84662 seconds to import. It has shape (6646430, 18)\n"
     ]
    }
   ],
   "source": [
    "GDSC1_RAW_FILE = 'GDSC1_public_raw_data_25Feb20.csv'\n",
    "GDSC2_RAW_FILE = 'GDSC2_public_raw_data_25Feb20.csv'\n",
    "\n",
    "# Read the raw files.\n",
    "# GDSC1\n",
    "start = time.time()\n",
    "gdsc1_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC1_RAW_FILE}', header=0)\n",
    "print(f\"File `{GDSC1_RAW_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc1_raw.shape}\")\n",
    "\n",
    "# GDSC2\n",
    "start = time.time()\n",
    "gdsc2_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC2_RAW_FILE}', header=0)\n",
    "print(f\"File `{GDSC2_RAW_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc2_raw.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | RESEARCH_PROJECT   |   BARCODE |   SCAN_ID | DATE_CREATED         |   SCAN_DATE |   CELL_ID |   MASTER_CELL_ID |   COSMIC_ID | CELL_LINE_NAME   |   SEEDING_DENSITY | DRUGSET_ID   | ASSAY   |   DURATION |   POSITION | TAG     |   DRUG_ID |   CONC |   INTENSITY |\n",
      "|---:|:-------------------|----------:|----------:|:---------------------|------------:|----------:|-----------------:|------------:|:-----------------|------------------:|:-------------|:--------|-----------:|-----------:|:--------|----------:|-------:|------------:|\n",
      "|  0 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          1 | B       |       nan |    nan |       26022 |\n",
      "|  1 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          2 | B       |       nan |    nan |       20491 |\n",
      "|  2 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          3 | UN-USED |       nan |    nan |      181946 |\n",
      "\n",
      "|    | RESEARCH_PROJECT   |   BARCODE |   SCAN_ID | DATE_CREATED         | SCAN_DATE            |   CELL_ID |   MASTER_CELL_ID |   COSMIC_ID | CELL_LINE_NAME   |   SEEDING_DENSITY |   DRUGSET_ID | ASSAY   |   DURATION |   POSITION | TAG     |   DRUG_ID |   CONC |   INTENSITY |\n",
      "|---:|:-------------------|----------:|----------:|:---------------------|:---------------------|----------:|-----------------:|------------:|:-----------------|------------------:|-------------:|:--------|-----------:|-----------:|:--------|----------:|-------:|------------:|\n",
      "|  0 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          1 | UN-USED |       nan |    nan |           0 |\n",
      "|  1 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          2 | UN-USED |       nan |    nan |           0 |\n",
      "|  2 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          3 | UN-USED |       nan |    nan |         398 |\n"
     ]
    }
   ],
   "source": [
    "print(gdsc1_raw.head(3).to_markdown())\n",
    "print()\n",
    "print(gdsc2_raw.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12484133, 18)\n"
     ]
    }
   ],
   "source": [
    "# Join both datasets for analysis purposes.\n",
    "gdsc_raw_join = pd.concat([gdsc1_raw, gdsc2_raw], ignore_index=True)\n",
    "print(gdsc_raw_join.shape)\n",
    "assert gdsc_raw_join[gdsc_raw_join.index.duplicated()].shape[0] == 0\n",
    "assert gdsc_raw_join.shape[0] == gdsc1_raw.shape[0] + gdsc2_raw.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(242036, 19)\n",
      "Index(['DATASET', 'NLME_RESULT_ID', 'NLME_CURVE_ID', 'COSMIC_ID',\n",
      "       'CELL_LINE_NAME', 'SANGER_MODEL_ID', 'TCGA_DESC', 'DRUG_ID',\n",
      "       'DRUG_NAME', 'PUTATIVE_TARGET', 'PATHWAY_NAME', 'COMPANY_ID',\n",
      "       'WEBRELEASE', 'MIN_CONC', 'MAX_CONC', 'LN_IC50', 'AUC', 'RMSE',\n",
      "       'Z_SCORE'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945558</td>\n",
       "      <td>683667</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>MB</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-1.462148</td>\n",
       "      <td>0.930105</td>\n",
       "      <td>0.088999</td>\n",
       "      <td>0.432482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945796</td>\n",
       "      <td>684052</td>\n",
       "      <td>A673</td>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-4.869447</td>\n",
       "      <td>0.614932</td>\n",
       "      <td>0.111423</td>\n",
       "      <td>-1.420322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946078</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-3.360684</td>\n",
       "      <td>0.790953</td>\n",
       "      <td>0.142754</td>\n",
       "      <td>-0.599894</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0   GDSC2             401       18945558     683667         PFSK-1   \n",
       "1   GDSC2             401       18945796     684052           A673   \n",
       "2   GDSC2             401       18946078     684057            ES5   \n",
       "\n",
       "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID     DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0       SIDM01132            MB     1003  Camptothecin            TOP1   \n",
       "1       SIDM00848  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "2       SIDM00263  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "\n",
       "      PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
       "0  DNA replication        1046          Y    0.0001       0.1 -1.462148   \n",
       "1  DNA replication        1046          Y    0.0001       0.1 -4.869447   \n",
       "2  DNA replication        1046          Y    0.0001       0.1 -3.360684   \n",
       "\n",
       "        AUC      RMSE   Z_SCORE  \n",
       "0  0.930105  0.088999  0.432482  \n",
       "1  0.614932  0.111423 -1.420322  \n",
       "2  0.790953  0.142754 -0.599894  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gdsc2_ic50s.shape)\n",
    "print(gdsc2_ic50s.columns)\n",
    "gdsc2_ic50s.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(242036, 9)\n",
      "Missing rates:\n",
      "==============\n",
      "DATASET            0.0\n",
      "CELL_LINE_NAME     0.0\n",
      "DRUG_NAME          0.0\n",
      "DRUG_ID            0.0\n",
      "SANGER_MODEL_ID    0.0\n",
      "AUC                0.0\n",
      "RMSE               0.0\n",
      "Z_SCORE            0.0\n",
      "LN_IC50            0.0\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>0.930105</td>\n",
       "      <td>0.088999</td>\n",
       "      <td>0.432482</td>\n",
       "      <td>-1.462148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>A673</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>0.614932</td>\n",
       "      <td>0.111423</td>\n",
       "      <td>-1.420322</td>\n",
       "      <td>-4.869447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>0.790953</td>\n",
       "      <td>0.142754</td>\n",
       "      <td>-0.599894</td>\n",
       "      <td>-3.360684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME     DRUG_NAME  DRUG_ID SANGER_MODEL_ID       AUC  \\\n",
       "0   GDSC2         PFSK-1  Camptothecin     1003       SIDM01132  0.930105   \n",
       "1   GDSC2           A673  Camptothecin     1003       SIDM00848  0.614932   \n",
       "2   GDSC2            ES5  Camptothecin     1003       SIDM00263  0.790953   \n",
       "\n",
       "       RMSE   Z_SCORE   LN_IC50  \n",
       "0  0.088999  0.432482 -1.462148  \n",
       "1  0.111423 -1.420322 -4.869447  \n",
       "2  0.142754 -0.599894 -3.360684  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COLS_TO_KEEP = ['DATASET', 'CELL_LINE_NAME', 'DRUG_NAME', 'DRUG_ID', \n",
    "                'SANGER_MODEL_ID', 'AUC', 'RMSE', 'Z_SCORE', 'LN_IC50']\n",
    "gdsc_base = gdsc2_ic50s[COLS_TO_KEEP].drop_duplicates()\n",
    "print(gdsc_base.shape)\n",
    "# Percent of NaN values per column.\n",
    "print(\"Missing rates:\")\n",
    "print(\"==============\")\n",
    "print(100 * gdsc_base.isna().sum() / gdsc_base.shape[0])\n",
    "gdsc_base.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC2    100.0\n",
       "Name: DATASET, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Occurence rate for each GDSC dataset.\n",
    "round(100 * gdsc_base.DATASET.value_counts() / gdsc_base.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save drug response matrix.\n",
    "gdsc_base.to_pickle(PROCESSED_DATA_PATH + 'gdsc2/drm_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1002, 13)\n",
      "|    | Sample Name   |   COSMIC identifier | Whole Exome Sequencing (WES)   | Copy Number Alterations (CNA)   | Gene Expression   | Methylation   | Drug       | GDSC                  | GDSC           | Cancer Type             | Microsatellite             | Screen Medium   | Growth Properties   |\n",
      "|    |               |                     |                                |                                 |                   |               | Response   | Tissue descriptor 1   | Tissue         | (matching TCGA label)   | instability Status (MSI)   |                 |                     |\n",
      "|    |               |                     |                                |                                 |                   |               |            |                       | descriptor 2   |                         |                            |                 |                     |\n",
      "|---:|:--------------|--------------------:|:-------------------------------|:--------------------------------|:------------------|:--------------|:-----------|:----------------------|:---------------|:------------------------|:---------------------------|:----------------|:--------------------|\n",
      "|  0 | A253          |              906794 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | nan                     | MSS/MSI-L                  | D/F12           | Adherent            |\n",
      "|  1 | BB30-HNC      |              753531 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | HNSC                    | MSS/MSI-L                  | D/F12           | Adherent            |\n",
      "|  2 | BB49-HNC      |              753532 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | HNSC                    | MSS/MSI-L                  | D/F12           | Adherent            |\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/gnn/lib/python3.10/site-packages/openpyxl/worksheet/_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n"
     ]
    }
   ],
   "source": [
    "cl_details = pd.read_excel(RAW_DATA_PATH + 'Cell_Lines_Details.xlsx')\n",
    "print(cl_details.shape)\n",
    "print(cl_details.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17737, 1020)\n",
      "|    | GENE_SYMBOLS   | GENE_title                                   |   DATA.906826 |   DATA.687983 |   DATA.910927 |   DATA.1240138 |   DATA.1240139 |   DATA.906792 |   DATA.910688 |   DATA.1240135 |\n",
      "|---:|:---------------|:---------------------------------------------|--------------:|--------------:|--------------:|---------------:|---------------:|--------------:|--------------:|---------------:|\n",
      "|  0 | TSPAN6         | tetraspanin 6 [Source:HGNC Symbol;Acc:11858] |       7.63202 |       7.54867 |       8.71234 |        7.79714 |        7.72927 |       7.07453 |        3.2852 |        6.96161 |\n",
      "|  1 | TNMD           | tenomodulin [Source:HGNC Symbol;Acc:17757]   |       2.96459 |       2.77772 |       2.64351 |        2.81792 |        2.95774 |       2.88968 |        2.8282 |        2.87475 |\n"
     ]
    }
   ],
   "source": [
    "gene_expression = pd.read_csv(RAW_DATA_PATH + 'Cell_line_RMA_proc_basalExp.txt', sep=\"\\t\")\n",
    "print(gene_expression.shape)\n",
    "print(gene_expression.iloc[:2, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'get_gdsc_gene_expression' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 10\u001b[0m\n\u001b[1;32m      6\u001b[0m GENE_EXPRESSION_RAW_FILE \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCell_line_RMA_proc_basalExp.txt\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m# Return the gene expression dataframe(n_cells x n_genes) for a set of gene symbols for all cell_lines of the GDSC cell line annotation file.\u001b[39;00m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# If the genes are None, return the data for all genes.\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m gene_expr \u001b[38;5;241m=\u001b[39m \u001b[43mget_gdsc_gene_expression\u001b[49m(\n\u001b[1;32m     11\u001b[0m     path_cell_annotations\u001b[38;5;241m=\u001b[39mRAW_DATA_PATH \u001b[38;5;241m+\u001b[39m CELL_LINE_DETAILS_FILE,\n\u001b[1;32m     12\u001b[0m     path_gene_expression\u001b[38;5;241m=\u001b[39mRAW_DATA_PATH \u001b[38;5;241m+\u001b[39m GENE_EXPRESSION_RAW_FILE \n\u001b[1;32m     13\u001b[0m )\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28mprint\u001b[39m(gene_expr\u001b[38;5;241m.\u001b[39mshape)\n\u001b[1;32m     15\u001b[0m gene_expr\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m3\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'get_gdsc_gene_expression' is not defined"
     ]
    }
   ],
   "source": [
    "# from src.preprocess.build_drug_response_matrix import (\n",
    "#     get_gdsc_gene_expression, cosmic_ids_to_cell_line_names\n",
    "# )\n",
    "\n",
    "CELL_LINE_DETAILS_FILE = 'Cell_Lines_Details.xlsx'\n",
    "GENE_EXPRESSION_RAW_FILE = 'Cell_line_RMA_proc_basalExp.txt'\n",
    "\n",
    "# Return the gene expression dataframe(n_cells x n_genes) for a set of gene symbols for all cell_lines of the GDSC cell line annotation file.\n",
    "# If the genes are None, return the data for all genes.\n",
    "gene_expr = get_gdsc_gene_expression(\n",
    "    path_cell_annotations=RAW_DATA_PATH + CELL_LINE_DETAILS_FILE,\n",
    "    path_gene_expression=RAW_DATA_PATH + GENE_EXPRESSION_RAW_FILE \n",
    ")\n",
    "print(gene_expr.shape)\n",
    "gene_expr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "print(landmark_genes.head(3).to_markdown())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose only the cell-line columns of the gene expressions table that are in the landmark gene file.\n",
    "inter_cols = list(set(gene_expr.columns).intersection(set(landmark_genes.Symbol)))\n",
    "gene_expr_sparse = gene_expr[inter_cols]\n",
    "gene_expr_sparse.columns.rename('CELL_LINE_NAME', inplace=True)\n",
    "print(gene_expr_sparse.shape)\n",
    "gene_expr_sparse.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 916)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>LPAR2</th>\n",
       "      <th>PEX11A</th>\n",
       "      <th>...</th>\n",
       "      <th>CXCR4</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>TBX2</th>\n",
       "      <th>RRP1B</th>\n",
       "      <th>MLLT11</th>\n",
       "      <th>UBE3B</th>\n",
       "      <th>VAPB</th>\n",
       "      <th>CDK5R1</th>\n",
       "      <th>STAT5B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>5.377255</td>\n",
       "      <td>4.000075</td>\n",
       "      <td>...</td>\n",
       "      <td>9.345843</td>\n",
       "      <td>3.631735</td>\n",
       "      <td>6.312806</td>\n",
       "      <td>3.169560</td>\n",
       "      <td>8.001636</td>\n",
       "      <td>5.106525</td>\n",
       "      <td>3.911568</td>\n",
       "      <td>5.059999</td>\n",
       "      <td>6.212202</td>\n",
       "      <td>5.440448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>6.288406</td>\n",
       "      <td>5.435063</td>\n",
       "      <td>...</td>\n",
       "      <td>3.005560</td>\n",
       "      <td>3.435098</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>2.935917</td>\n",
       "      <td>8.226306</td>\n",
       "      <td>7.938137</td>\n",
       "      <td>4.489525</td>\n",
       "      <td>5.751381</td>\n",
       "      <td>4.490407</td>\n",
       "      <td>3.992091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>6.166101</td>\n",
       "      <td>6.351613</td>\n",
       "      <td>...</td>\n",
       "      <td>3.852866</td>\n",
       "      <td>3.490747</td>\n",
       "      <td>6.345090</td>\n",
       "      <td>3.025210</td>\n",
       "      <td>8.219939</td>\n",
       "      <td>8.871089</td>\n",
       "      <td>4.315447</td>\n",
       "      <td>5.074429</td>\n",
       "      <td>5.300401</td>\n",
       "      <td>4.583204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 916 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50     LPAR2    PEX11A  ...     CXCR4    NOTCH1   PLEKHM1      TBX2  \\\n",
       "0  2.395685  5.377255  4.000075  ...  9.345843  3.631735  6.312806  3.169560   \n",
       "1  3.140923  6.288406  5.435063  ...  3.005560  3.435098  6.188079  2.935917   \n",
       "2  3.968757  6.166101  6.351613  ...  3.852866  3.490747  6.345090  3.025210   \n",
       "\n",
       "      RRP1B    MLLT11     UBE3B      VAPB    CDK5R1    STAT5B  \n",
       "0  8.001636  5.106525  3.911568  5.059999  6.212202  5.440448  \n",
       "1  8.226306  7.938137  4.489525  5.751381  4.490407  3.992091  \n",
       "2  8.219939  8.871089  4.315447  5.074429  5.300401  4.583204  \n",
       "\n",
       "[3 rows x 916 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_geneexpr = gdsc_base.merge(right=gene_expr_sparse,\n",
    "                                     left_on=['CELL_LINE_NAME'],\n",
    "                                     right_index=True,\n",
    "                                     how='left',\n",
    "                                     suffixes=['_gdsc', '_geneexpr'])\n",
    "print(join_gdsc_geneexpr.shape)\n",
    "join_gdsc_geneexpr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the gene expression information to a file.\n",
    "join_gdsc_geneexpr.to_pickle(PROCESSED_DATA_PATH + 'gexpr_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy number variation - picnic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `cnv_abs_copy_number_picnic_20191101.csv` took 3.31045 seconds to import. \n",
      "Shape: (24502, 988)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>M14</th>\n",
       "      <th>TE-12</th>\n",
       "      <th>TMK-1</th>\n",
       "      <th>STS-0421</th>\n",
       "      <th>PL4</th>\n",
       "      <th>PCI-4B</th>\n",
       "      <th>PCI-30</th>\n",
       "      <th>HSC-39</th>\n",
       "      <th>...</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>MMAc-SF</th>\n",
       "      <th>BE-13</th>\n",
       "      <th>MC-IXC</th>\n",
       "      <th>Ramos-2G6-4C10</th>\n",
       "      <th>CGTH-W-1</th>\n",
       "      <th>H9</th>\n",
       "      <th>GR-ST</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>MM1S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG00001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG00003</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG00004</td>\n",
       "      <td>A2M</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 988 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_ID GENE_SYMBOL  M14  TE-12  TMK-1  STS-0421  PL4  PCI-4B  PCI-30  \\\n",
       "1  SIDG00001        A1BG  3.0    3.0    3.0       4.0  6.0     3.0     2.0   \n",
       "2  SIDG00003        A1CF  3.0    3.0    3.0       4.0  3.0     3.0     3.0   \n",
       "3  SIDG00004         A2M  3.0    3.0    2.0       4.0  9.0     3.0     3.0   \n",
       "\n",
       "   HSC-39  ...  451Lu  MMAc-SF  BE-13  MC-IXC  Ramos-2G6-4C10  CGTH-W-1   H9  \\\n",
       "1     3.0  ...    4.0      3.0    4.0     2.0             2.0       4.0  4.0   \n",
       "2     3.0  ...    4.0      2.0    3.0     1.0             2.0       3.0  4.0   \n",
       "3     3.0  ...    5.0      3.0    4.0     2.0             2.0       4.0  3.0   \n",
       "\n",
       "   GR-ST  YMB-1-E  MM1S  \n",
       "1    2.0      3.0   2.0  \n",
       "2    2.0      3.0   2.0  \n",
       "3    2.0      5.0   2.0  \n",
       "\n",
       "[3 rows x 988 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "cnv_picnic_v0 = pd.read_csv(RAW_DATA_PATH + 'cnv_abs_copy_number_picnic_20191101.csv', sep=\",\", header=1)\n",
    "cnv_picnic_v0.rename(columns={\n",
    "    'Unnamed: 1': 'GENE_SYMBOL',\n",
    "    'model_name': 'GENE_ID'}, inplace=True)\n",
    "cnv_picnic_v0 = cnv_picnic_v0.iloc[1:, :]\n",
    "print(f\"File `cnv_abs_copy_number_picnic_20191101.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {cnv_picnic_v0.shape}\")\n",
    "cnv_picnic_v0.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | GENE_ID   | GENE_SYMBOL   |   M14 |   TE-12 |   TMK-1 |   STS-0421 |   PL4 |   PCI-4B |   PCI-30 |   HSC-39 |\n",
      "|---:|:----------|:--------------|------:|--------:|--------:|-----------:|------:|---------:|---------:|---------:|\n",
      "|  1 | SIDG00001 | A1BG          |     3 |       3 |       3 |          4 |     6 |        3 |        2 |        3 |\n",
      "|  2 | SIDG00003 | A1CF          |     3 |       3 |       3 |          4 |     3 |        3 |        3 |        3 |\n",
      "|  3 | SIDG00004 | A2M           |     3 |       3 |       2 |          4 |     9 |        3 |        3 |        3 |\n"
     ]
    }
   ],
   "source": [
    "print(cnv_picnic_v0.iloc[:3, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that here the columns are the cell-line names and the row are indentified by the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(986, 24502)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11AP1</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24502 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      A1BG A1CF  A2M A2ML1 A2ML1-AS1 A2ML1-AS2 A3GALT2 A4GALT A4GNT AAAS  ...  \\\n",
       "M14    3.0  3.0  3.0   3.0       3.0       3.0     4.0    4.0   4.0  3.0  ...   \n",
       "TE-12  3.0  3.0  3.0   3.0       3.0       3.0     2.0    3.0   3.0  3.0  ...   \n",
       "TMK-1  3.0  3.0  2.0   2.0       2.0       2.0     2.0    2.0   3.0  3.0  ...   \n",
       "\n",
       "      ZWINT ZXDA ZXDB ZXDC ZYG11A ZYG11AP1 ZYG11B  ZYX ZZEF1 ZZZ3  \n",
       "M14     3.0  2.0  2.0  4.0    4.0      3.0    4.0  4.0   4.0  4.0  \n",
       "TE-12   3.0  2.0  2.0  3.0    3.0      3.0    3.0  3.0   3.0  3.0  \n",
       "TMK-1   3.0  2.0  2.0  1.0    3.0      3.0    3.0  3.0   2.0  3.0  \n",
       "\n",
       "[3 rows x 24502 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_picnic_v1 = cnv_picnic_v0.iloc[:, 1:].T\n",
    "cnv_picnic_v1 = cnv_picnic_v1.rename(columns=cnv_picnic_v1.iloc[0]).drop(cnv_picnic_v1.index[0])\n",
    "print(cnv_picnic_v1.shape)\n",
    "cnv_picnic_v1.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 7)\n",
      "(986, 966)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 966 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AARS ABCB6 ABCC5 ABCF1 ABCF3 ABHD4 ABHD6 ABL1 ACAA1 ACAT2  ... ZMIZ1  \\\n",
       "M14    3.0   3.0   4.0   3.0   4.0   2.0   3.0  3.0   3.0   2.0  ...   3.0   \n",
       "TE-12  4.0   3.0   3.0   2.0   3.0   4.0   2.0  4.0   2.0   2.0  ...   3.0   \n",
       "TMK-1  3.0   2.0   3.0   3.0   3.0   3.0   2.0  3.0   2.0   2.0  ...   2.0   \n",
       "\n",
       "      ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451 ZNF586 ZNF589 ZW10  \n",
       "M14     3.0    3.0    3.0    3.0    2.0    3.0    3.0    3.0  3.0  \n",
       "TE-12   4.0    5.0    3.0    2.0    3.0    2.0    3.0    2.0  3.0  \n",
       "TMK-1   2.0    5.0    3.0    3.0    2.0    3.0    3.0    2.0  4.0  \n",
       "\n",
       "[3 rows x 966 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------------- #\n",
    "# Sparse down the gene column dimension by only using landmark genes. #\n",
    "# ------------------------------------------------------------------- #\n",
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "\n",
    "# Check how many cell line columns of the cnv picnic table are in the landmark gene file.\n",
    "count, cols_to_keep = 0, []\n",
    "for c in cnv_picnic_v1.columns[cnv_picnic_v1.columns != 'nan']:\n",
    "    if c in landmark_genes.Symbol.tolist(): \n",
    "        count += 1\n",
    "        cols_to_keep.append(c)\n",
    "\n",
    "cnv_picnic_v2 = cnv_picnic_v1[cols_to_keep]\n",
    "assert cnv_picnic_v2.shape[1] == len(cols_to_keep)\n",
    "print(cnv_picnic_v2.shape)\n",
    "cnv_picnic_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 974)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 974 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50 AARS ABCB6  ... ZMIZ1 ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451  \\\n",
       "0  2.395685  2.0   2.0  ...   2.0   2.0    2.0    2.0    2.0    2.0    2.0   \n",
       "1  3.140923  2.0   2.0  ...   1.0   1.0    2.0    2.0    2.0    3.0    2.0   \n",
       "2  3.968757  3.0   4.0  ...   2.0   3.0    3.0    3.0    4.0    6.0    4.0   \n",
       "\n",
       "  ZNF586 ZNF589 ZW10  \n",
       "0    2.0    2.0  1.0  \n",
       "1    2.0    3.0  2.0  \n",
       "2    3.0    4.0  3.0  \n",
       "\n",
       "[3 rows x 974 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_cnv_picnic = gdsc_base.merge(right=cnv_picnic_v2,\n",
    "                                       left_on=['CELL_LINE_NAME'],\n",
    "                                       right_index=True,\n",
    "                                       how='left',\n",
    "                                       suffixes=['_gdsc', '_cnvp'])\n",
    "print(join_gdsc_cnv_picnic.shape)\n",
    "join_gdsc_cnv_picnic.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the copy number picnic information to a file.\n",
    "join_gdsc_cnv_picnic.to_pickle(PROCESSED_DATA_PATH + 'cnvp_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy number variation - gistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `cnv_gistic_20191101.csv` took 3.35268 seconds to import. \n",
      "Shape: (20669, 980)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>M14</th>\n",
       "      <th>TE-12</th>\n",
       "      <th>TMK-1</th>\n",
       "      <th>STS-0421</th>\n",
       "      <th>PL4</th>\n",
       "      <th>PCI-4B</th>\n",
       "      <th>PCI-30</th>\n",
       "      <th>HSC-39</th>\n",
       "      <th>...</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>MMAc-SF</th>\n",
       "      <th>BE-13</th>\n",
       "      <th>MC-IXC</th>\n",
       "      <th>Ramos-2G6-4C10</th>\n",
       "      <th>CGTH-W-1</th>\n",
       "      <th>H9</th>\n",
       "      <th>GR-ST</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>MM1S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG00001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG00002</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG00003</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 980 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_ID GENE_SYMBOL  M14  TE-12  TMK-1  STS-0421  PL4  PCI-4B  PCI-30  \\\n",
       "1  SIDG00001        A1BG  0.0   -1.0    0.0       0.0  1.0    -1.0     0.0   \n",
       "2  SIDG00002    A1BG-AS1  0.0   -1.0    0.0       0.0  1.0    -1.0     0.0   \n",
       "3  SIDG00003        A1CF -1.0    0.0    0.0       1.0 -1.0     0.0     1.0   \n",
       "\n",
       "   HSC-39  ...  451Lu  MMAc-SF  BE-13  MC-IXC  Ramos-2G6-4C10  CGTH-W-1   H9  \\\n",
       "1     0.0  ...    0.0     -1.0    2.0     0.0             0.0       1.0  1.0   \n",
       "2     0.0  ...    0.0     -1.0    2.0     0.0             0.0       1.0  1.0   \n",
       "3     0.0  ...    0.0     -1.0   -1.0    -1.0             0.0       0.0  1.0   \n",
       "\n",
       "   GR-ST  YMB-1-E  MM1S  \n",
       "1    0.0      0.0   0.0  \n",
       "2    0.0      0.0   0.0  \n",
       "3    0.0      0.0   0.0  \n",
       "\n",
       "[3 rows x 980 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "cnv_gistic_v0 = pd.read_csv(RAW_DATA_PATH + 'cnv_gistic_20191101.csv', sep=\",\", header=1)\n",
    "cnv_gistic_v0.rename(columns={\n",
    "    'Unnamed: 1': 'GENE_SYMBOL',\n",
    "    'model_name': 'GENE_ID'}, inplace=True)\n",
    "cnv_gistic_v0 = cnv_gistic_v0.iloc[1:, :]\n",
    "print(f\"File `cnv_gistic_20191101.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {cnv_gistic_v0.shape}\")\n",
    "cnv_gistic_v0.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | GENE_ID   | GENE_SYMBOL   |   M14 |   TE-12 |   TMK-1 |   STS-0421 |   PL4 |   PCI-4B |   PCI-30 |   HSC-39 |\n",
      "|---:|:----------|:--------------|------:|--------:|--------:|-----------:|------:|---------:|---------:|---------:|\n",
      "|  1 | SIDG00001 | A1BG          |     0 |      -1 |       0 |          0 |     1 |       -1 |        0 |        0 |\n",
      "|  2 | SIDG00002 | A1BG-AS1      |     0 |      -1 |       0 |          0 |     1 |       -1 |        0 |        0 |\n",
      "|  3 | SIDG00003 | A1CF          |    -1 |       0 |       0 |          1 |    -1 |        0 |        1 |        0 |\n"
     ]
    }
   ],
   "source": [
    "print(cnv_gistic_v0.iloc[:3, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that here the columns are the cell-line names and the row are indentified by the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 20669)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZUP1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 20669 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      A1BG A1BG-AS1 A1CF  A2M A2ML1 A2MP1 A4GALT A4GNT AAAS AACS  ... ZUP1  \\\n",
       "M14    0.0      0.0 -1.0  0.0   0.0   0.0    1.0   1.0  0.0  0.0  ... -1.0   \n",
       "TE-12 -1.0     -1.0  0.0  1.0   1.0   1.0   -1.0   0.0  0.0  0.0  ... -1.0   \n",
       "TMK-1  0.0      0.0  0.0 -1.0  -1.0  -1.0   -1.0   0.0  0.0  0.0  ... -1.0   \n",
       "\n",
       "      ZW10 ZWILCH ZWINT ZXDC ZYG11A ZYG11B  ZYX ZZEF1 ZZZ3  \n",
       "M14    0.0    1.0  -1.0  1.0    1.0    1.0  2.0   2.0  2.0  \n",
       "TE-12  1.0    1.0   1.0  0.0    0.0    0.0  1.0   0.0  0.0  \n",
       "TMK-1  1.0    0.0   1.0 -2.0    0.0    0.0  2.0  -1.0  1.0  \n",
       "\n",
       "[3 rows x 20669 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_gistic_v1 = cnv_gistic_v0.iloc[:, 1:].T\n",
    "cnv_gistic_v1 = cnv_gistic_v1.rename(columns=cnv_gistic_v1.iloc[0]).drop(cnv_gistic_v1.index[0])\n",
    "print(cnv_gistic_v1.shape)\n",
    "cnv_gistic_v1.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 7)\n",
      "(978, 938)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 938 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AARS ABCB6 ABCC5 ABCF1 ABCF3 ABHD4 ABHD6 ABL1 ACAA1 ACAT2  ... ZMIZ1  \\\n",
       "M14    0.0   0.0   1.0   0.0   1.0  -1.0   0.0  0.0   0.0  -1.0  ...  -1.0   \n",
       "TE-12  1.0  -1.0   0.0  -1.0   0.0   1.0  -1.0  1.0  -1.0  -1.0  ...  -1.0   \n",
       "TMK-1  0.0  -1.0   0.0   0.0   0.0   0.0  -1.0  0.0  -1.0  -1.0  ...  -1.0   \n",
       "\n",
       "      ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451 ZNF586 ZNF589 ZW10  \n",
       "M14     0.0    0.0    0.0    0.0   -1.0    0.0    0.0    0.0  0.0  \n",
       "TE-12   1.0    1.0   -1.0   -1.0    0.0   -1.0   -1.0   -1.0  1.0  \n",
       "TMK-1  -1.0    1.0    0.0    0.0   -1.0    1.0    0.0   -1.0  1.0  \n",
       "\n",
       "[3 rows x 938 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------------- #\n",
    "# Sparse down the gene column dimension by only using landmark genes. #\n",
    "# ------------------------------------------------------------------- #\n",
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "\n",
    "# Check how many cell line columns of the cnv picnic table are in the landmark gene file.\n",
    "count, cols_to_keep = 0, []\n",
    "for c in cnv_gistic_v1.columns[cnv_gistic_v1.columns != 'nan']:\n",
    "    if c in landmark_genes.Symbol.tolist(): \n",
    "        count += 1\n",
    "        cols_to_keep.append(c)\n",
    "\n",
    "cnv_gistic_v2 = cnv_gistic_v1[cols_to_keep]\n",
    "assert cnv_gistic_v2.shape[1] == len(cols_to_keep)\n",
    "print(cnv_gistic_v2.shape)\n",
    "cnv_gistic_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 946)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 946 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50 AARS ABCB6  ... ZMIZ1 ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451  \\\n",
       "0  2.395685  0.0   0.0  ...   0.0   0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "1  3.140923  0.0   0.0  ...  -1.0  -1.0    0.0    0.0    0.0    1.0    0.0   \n",
       "2  3.968757 -1.0   1.0  ...  -1.0  -1.0   -1.0   -1.0    0.0    1.0    1.0   \n",
       "\n",
       "  ZNF586 ZNF589 ZW10  \n",
       "0    0.0    0.0 -2.0  \n",
       "1    0.0    1.0  0.0  \n",
       "2   -1.0    0.0 -1.0  \n",
       "\n",
       "[3 rows x 946 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_cnv_gistic = gdsc_base.merge(right=cnv_gistic_v2,\n",
    "                                       left_on=['CELL_LINE_NAME'],\n",
    "                                       right_index=True,\n",
    "                                       how='left',\n",
    "                                       suffixes=['_gdsc', '_cnvg'])\n",
    "print(join_gdsc_cnv_gistic.shape)\n",
    "join_gdsc_cnv_gistic.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the copy number gistic information to a file.\n",
    "join_gdsc_cnv_gistic.to_pickle(PROCESSED_DATA_PATH + 'cnvg_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `mutations_all_20220315.csv` took 20.57474 seconds to import. \n",
      "Shape: (8322616, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15  \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15  \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15  \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15  \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "mutations_v0 = pd.read_csv(RAW_DATA_PATH + 'mutations_all_20220315.csv', sep=\",\", header=0)\n",
    "print(f\"File `mutations_all_20220315.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {mutations_v0.shape}\")\n",
    "mutations_v0.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | gene_id   | gene_symbol   | model_id   | protein_mutation   | rna_mutation   | cdna_mutation   | cancer_driver   | cancer_predisposition_variant   | effect   |    vaf | coding   | source   | model_name        |\n",
      "|---:|:----------|:--------------|:-----------|:-------------------|:---------------|:----------------|:----------------|:--------------------------------|:---------|-------:|:---------|:---------|:------------------|\n",
      "|  0 | SIDG36104 | SOX30         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.2703 | False    | Sanger   | HCM-SANG-1082-C15 |\n",
      "|  1 | SIDG37647 | THG1L         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.75   | False    | Sanger   | HCM-SANG-1082-C15 |\n",
      "|  2 | SIDG16920 | LSM11         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.6452 | False    | Sanger   | HCM-SANG-1082-C15 |\n"
     ]
    }
   ],
   "source": [
    "print(mutations_v0.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\n",
      "Shape: (8322616, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \\\n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15   \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15   \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15   \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15   \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15   \n",
       "\n",
       "  SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "0             NaN            NaN  \n",
       "1             NaN            NaN  \n",
       "2             NaN            NaN  \n",
       "3             NaN            NaN  \n",
       "4             NaN            NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Find the CELL_LINE_NAME's per SANGER_MODEL_ID.\n",
    "celllines_per_sangermodelid = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']].groupby('SANGER_MODEL_ID')['CELL_LINE_NAME'].nunique()\n",
    "counts_per_sangermodelid = celllines_per_sangermodelid.values\n",
    "\n",
    "assert (counts_per_sangermodelid != 1).any() == False\n",
    "assert (counts_per_sangermodelid == 1).all()\n",
    "\n",
    "# Only take the interested columns for the mapping.\n",
    "gdsc_mapping_subset = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']]\n",
    "\n",
    "# Only take the unique SANGER_MODEL_ID's, since these have a 1-to-1 relationship to the CELL_LINE_NAME's anyways.\n",
    "gdsc_mapping_subset = gdsc_mapping_subset.groupby('SANGER_MODEL_ID').first().reset_index(level=0)\n",
    "\n",
    "# Join the CELL_LINE_NAME's onto the mutations_all dataset, based on the model_id.\n",
    "mutations_v1 = mutations_v0.merge(right=gdsc_mapping_subset,\n",
    "                                  left_on='model_id',\n",
    "                                  right_on='SANGER_MODEL_ID',\n",
    "                                  how='left')\n",
    "print(\"We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\")                                  \n",
    "print(f\"Shape: {mutations_v1.shape}\")\n",
    "mutations_v1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Each `CELL_LINE_NAME` has only exactly one `SANGER_MODEL_ID`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We excluded all rows which have CELL_LINE_NAME as NaN.\n",
      "Number of unique `gene_symbol`s in the mutations dataset: 20450\n",
      "Shape: (4574963, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SIDG40054</td>\n",
       "      <td>UAP1</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2857</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>JHH-6</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>JHH-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SIDG13175</td>\n",
       "      <td>IPO11</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.3511</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>NCI-H727</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>NCI-H727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SIDG14638</td>\n",
       "      <td>LIN9</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SIDG07173</td>\n",
       "      <td>DYM</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.5294</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SIDG08926</td>\n",
       "      <td>FCAMR</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>D-336MG</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>D-336MG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "19  SIDG40054        UAP1  SIDM00615                -          r.?   \n",
       "20  SIDG13175       IPO11  SIDM01119                -          r.?   \n",
       "21  SIDG14638        LIN9  SIDM00933                -          r.?   \n",
       "22  SIDG07173         DYM  SIDM00933                -          r.?   \n",
       "23  SIDG08926       FCAMR  SIDM00620                -          r.?   \n",
       "\n",
       "   cdna_mutation  cancer_driver  cancer_predisposition_variant      effect  \\\n",
       "19             -          False                          False    intronic   \n",
       "20             -          False                          False    intronic   \n",
       "21             -          False                          False  downstream   \n",
       "22             -          False                          False    intronic   \n",
       "23             -          False                          False    intronic   \n",
       "\n",
       "       vaf  coding  source model_name SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "19  0.2857   False  Sanger      JHH-6       SIDM00615          JHH-6  \n",
       "20  0.3511   False  Sanger   NCI-H727       SIDM01119       NCI-H727  \n",
       "21  1.0000   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "22  0.5294   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "23  0.6250   False  Sanger    D-336MG       SIDM00620        D-336MG  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " # Only taking the ones with a CELL_LINE_NAME.\n",
    "mutations_v2 = mutations_v1[mutations_v1['CELL_LINE_NAME'].notna()]\n",
    "print(\"We excluded all rows which have CELL_LINE_NAME as NaN.\")\n",
    "print(f\"Number of unique `gene_symbol`s in the mutations dataset: {len(np.unique(mutations_v2.gene_symbol))}\")\n",
    "print(f\"Shape: {mutations_v2.shape}\")\n",
    "mutations_v2.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We use the landmark genes to sparse down the dimension.\n",
      "We merged the most recent dataset mutations_v2 with the landmark genes to sparse it down.\n",
      "Shape: (235906, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>splice_region</td>\n",
       "      <td>0.6190</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5prime_UTR_variant</td>\n",
       "      <td>0.4118</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG40702       VGLL4  SIDM00330                -          r.?   \n",
       "1  SIDG40702       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2  SIDG40702       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1           c.?          False                          False   \n",
       "2           c.?          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source model_name SANGER_MODEL_ID  \\\n",
       "0            intronic  0.9926   False  Sanger      LAN-6       SIDM00330   \n",
       "1       splice_region  0.6190   False   Broad      CL-11       SIDM00944   \n",
       "2  5prime_UTR_variant  0.4118   False   Broad      SNG-M       SIDM00373   \n",
       "\n",
       "  CELL_LINE_NAME Symbol  \n",
       "0          LAN-6  VGLL4  \n",
       "1          CL-11  VGLL4  \n",
       "2          SNG-M  VGLL4  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(\"We use the landmark genes to sparse down the dimension.\")\n",
    "\n",
    "# Take only the rows which have a `gene_symbol` which is also present in the landmark genes table.\n",
    "mutations_v3 = mutations_v2.merge(right=landmark_genes['Symbol'],\n",
    "                                  left_on='gene_symbol',\n",
    "                                  right_on='Symbol')\n",
    "print(\"We merged the most recent dataset mutations_v2 with the landmark genes to sparse it down.\")                                  \n",
    "print(f\"Shape: {mutations_v3.shape}\")\n",
    "mutations_v3.head(3)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We only took a chosen set of columns.\n",
      "Number of unique ...\n",
      "    CELL_LINE_NAME's   :    983 (0.42% out of all rows)\n",
      "    gene_symbol's      :    956 (0.41% out of all rows)\n",
      "    protein_mutation's :  25113 (10.65% out of all rows)\n",
      "    rna_mutation's     :  28392 (12.04% out of all rows)\n",
      "    cdna_mutation's    :  22196 (9.41% out of all rows)\n",
      "    cancer_driver's    :      2 ([False  True])\n",
      "\n",
      "Shape: (235906, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JHOS-2</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00305</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCI-H1436</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00697</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9677</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0          LAN-6       VGLL4  SIDM00330                -          r.?   \n",
       "1          CL-11       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2          SNG-M       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "3         JHOS-2       VGLL4  SIDM00305                -          r.?   \n",
       "4      NCI-H1436       VGLL4  SIDM00697                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver     vaf  \n",
       "0             -          False  0.9926  \n",
       "1           c.?          False  0.6190  \n",
       "2           c.?          False  0.4118  \n",
       "3             -          False  0.5698  \n",
       "4             -          False  0.9677  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v4 = mutations_v3[[\n",
    "    'CELL_LINE_NAME',\n",
    "    'gene_symbol',\n",
    "    'model_id',\n",
    "    'protein_mutation',\n",
    "    'rna_mutation',\n",
    "    'cdna_mutation',\n",
    "    'cancer_driver',\n",
    "    'vaf'\n",
    "]]\n",
    "print(\"We only took a chosen set of columns.\")\n",
    "n_v4 = mutations_v4.shape[0]\n",
    "uniq_cl = len(np.unique(mutations_v4.CELL_LINE_NAME))\n",
    "uniq_gs = len(np.unique(mutations_v4.gene_symbol))\n",
    "uniq_pm = len(np.unique(mutations_v4.protein_mutation))\n",
    "uniq_rna = len(np.unique(mutations_v4.rna_mutation))\n",
    "uniq_cdna = len(np.unique(mutations_v4.cdna_mutation))\n",
    "\n",
    "print(f\"\"\"Number of unique ...\n",
    "    CELL_LINE_NAME's   : {uniq_cl:6.0f} ({100*uniq_cl/n_v4:2.2f}% out of all rows)\n",
    "    gene_symbol's      : {uniq_gs:6.0f} ({100*uniq_gs/n_v4:2.2f}% out of all rows)\n",
    "    protein_mutation's : {uniq_pm:6.0f} ({100*uniq_pm/n_v4:2.2f}% out of all rows)\n",
    "    rna_mutation's     : {uniq_rna:6.0f} ({100*uniq_rna/n_v4:2.2f}% out of all rows)\n",
    "    cdna_mutation's    : {uniq_cdna:6.0f} ({100*uniq_cdna/n_v4:2.2f}% out of all rows)\n",
    "    cancer_driver's    : {len(np.unique(mutations_v4.cancer_driver)):6.0f} ({np.unique(mutations_v4.cancer_driver)})\n",
    "\"\"\")\n",
    "print(f\"Shape: {mutations_v4.shape}\")\n",
    "mutations_v4.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (983, 957)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>451Lu</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 957 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GENE_SYMBOL CELL_LINE_NAME  ABCB6  ABCC5  ABCF1  ABCF3  ABHD4  ABHD6  ABL1  \\\n",
       "0                    22RV1    0.0    0.0    1.0    1.0    1.0    0.0   1.0   \n",
       "1                 23132-87    0.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "2                 42-MG-BA    1.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "3                    451Lu    1.0    0.0    1.0    0.0    0.0    0.0   0.0   \n",
       "4                     5637    0.0    1.0    1.0    1.0    1.0    0.0   1.0   \n",
       "\n",
       "GENE_SYMBOL  ACAA1  ACAT2  ...  ZMIZ1  ZMYM2  ZNF131  ZNF274  ZNF318  ZNF395  \\\n",
       "0              0.0    0.0  ...    1.0    1.0     0.0     0.0     1.0     0.0   \n",
       "1              0.0    0.0  ...    0.0    1.0     1.0     0.0     0.0     0.0   \n",
       "2              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "3              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "4              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "GENE_SYMBOL  ZNF451  ZNF586  ZNF589  ZW10  \n",
       "0               0.0     0.0     0.0   1.0  \n",
       "1               0.0     0.0     1.0   0.0  \n",
       "2               0.0     0.0     0.0   0.0  \n",
       "3               1.0     0.0     0.0   0.0  \n",
       "4               0.0     0.0     0.0   0.0  \n",
       "\n",
       "[5 rows x 957 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v5 = pd.pivot_table(\n",
    "    data    = mutations_v4,\n",
    "    values  = 'cancer_driver',\n",
    "    index   = ['CELL_LINE_NAME'],\n",
    "    columns = ['gene_symbol'],\n",
    "    aggfunc = np.sum,\n",
    "    dropna  = False\n",
    ")\n",
    "\n",
    "# Set mutation values: 1.0=mutation, 0.0=no_mutation\n",
    "mutations_v5[mutations_v5==0.0] = 1.0\n",
    "mutations_v5[np.isnan(mutations_v5)] = 0.0\n",
    "\n",
    "mutations_v5['CELL_LINE_NAME'] = mutations_v5.index\n",
    "mutations_v5.insert(0, 'CELL_LINE_NAME', mutations_v5.pop('CELL_LINE_NAME'))\n",
    "mutations_v5.reset_index(drop=True, inplace=True)\n",
    "mutations_v5.columns.name = 'GENE_SYMBOL'\n",
    "\n",
    "print(f\"Shape: {mutations_v5.shape}\")\n",
    "mutations_v5.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(444205, 965)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Rapamycin</td>\n",
       "      <td>3</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>0.980891</td>\n",
       "      <td>0.032311</td>\n",
       "      <td>0.806489</td>\n",
       "      <td>-0.658244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Sunitinib</td>\n",
       "      <td>5</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>0.903979</td>\n",
       "      <td>0.023236</td>\n",
       "      <td>-0.206663</td>\n",
       "      <td>2.161095</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 965 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID SANGER_MODEL_ID       AUC  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1       SIDM00636  0.982114   \n",
       "1   GDSC1         MC-CAR  Rapamycin        3       SIDM00636  0.980891   \n",
       "2   GDSC1         MC-CAR  Sunitinib        5       SIDM00636  0.903979   \n",
       "\n",
       "       RMSE   Z_SCORE   LN_IC50  ABCB6  ...  ZMIZ1  ZMYM2  ZNF131  ZNF274  \\\n",
       "0  0.022521 -0.189576  2.395685    0.0  ...    0.0    1.0     0.0     0.0   \n",
       "1  0.032311  0.806489 -0.658244    0.0  ...    0.0    1.0     0.0     0.0   \n",
       "2  0.023236 -0.206663  2.161095    0.0  ...    0.0    1.0     0.0     0.0   \n",
       "\n",
       "   ZNF318  ZNF395  ZNF451  ZNF586  ZNF589  ZW10  \n",
       "0     0.0     0.0     1.0     0.0     0.0   0.0  \n",
       "1     0.0     0.0     1.0     0.0     0.0   0.0  \n",
       "2     0.0     0.0     1.0     0.0     0.0   0.0  \n",
       "\n",
       "[3 rows x 965 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "join_gdsc_mut = gdsc_base.merge(right=mutations_v5,\n",
    "                               on=['CELL_LINE_NAME'],\n",
    "                               how='inner',\n",
    "                               suffixes=['_gdsc', '_mut'])\n",
    "                            \n",
    "print(join_gdsc_mut.shape)\n",
    "join_gdsc_mut.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the mutational information to a file.\n",
    "join_gdsc_mut.to_pickle(PROCESSED_DATA_PATH + 'mut_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intersecting Cell lines and Genes\n",
    "\n",
    "We need a list of intersecting genes symbols and generate indices for them. This is due to the reason that we don't want any missing values in our final dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "drm = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "gexpr = pd.read_pickle(PROCESSED_DATA_PATH + 'gexpr_full.pkl')\n",
    "cnvg = pd.read_pickle(PROCESSED_DATA_PATH + 'cnvg_full.pkl')\n",
    "cnvp = pd.read_pickle(PROCESSED_DATA_PATH + 'cnvp_full.pkl')\n",
    "mut = pd.read_pickle(PROCESSED_DATA_PATH + 'mut_full.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 9)\n",
      "(423663, 916)\n",
      "(437781, 946)\n",
      "(423209, 974)\n",
      "(444205, 965)\n"
     ]
    }
   ],
   "source": [
    "print(drm.dropna().shape)\n",
    "print(gexpr.dropna().shape)\n",
    "print(cnvg.dropna().shape)\n",
    "print(cnvp.dropna().shape)\n",
    "print(mut.dropna().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We don't want any missing values for our features.\n",
    "# This would destroy the predictions of the network.\n",
    "gexpr.dropna(inplace=True)\n",
    "cnvg.dropna(inplace=True)\n",
    "cnvp.dropna(inplace=True)\n",
    "mut.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of intersecting cell-lines: 870\n",
      "Number of intersecting genes: 858\n"
     ]
    }
   ],
   "source": [
    "# ----------------------- #\n",
    "# Intersecting cell-lines #\n",
    "# ----------------------- #\n",
    "uniq_cl_drm = list(drm.CELL_LINE_NAME.unique())\n",
    "uniq_cl_gexpr = list(gexpr.CELL_LINE_NAME.unique())\n",
    "uniq_cl_cnvg = list(cnvg.CELL_LINE_NAME.unique())\n",
    "uniq_cl_cnvp = list(cnvp.CELL_LINE_NAME.unique())\n",
    "uniq_cl_mut = list(mut.CELL_LINE_NAME.unique()) \n",
    "\n",
    "# Intersecting cell-lines between all raw datasets.\n",
    "INTER_CLS = set(uniq_cl_drm)\\\n",
    "    .intersection(set(uniq_cl_gexpr))\\\n",
    "    .intersection(set(uniq_cl_cnvg))\\\n",
    "    .intersection(set(uniq_cl_cnvp))\\\n",
    "    .intersection(set(uniq_cl_mut))\n",
    " \n",
    "# ------------------ #\n",
    "# Intersecting genes #\n",
    "# ------------------ #\n",
    "ignore = ['DATASET', 'CELL_LINE_NAME', 'DRUG_NAME', 'DRUG_ID', \n",
    "          'SANGER_MODEL_ID', 'AUC', 'RMSE', 'Z_SCORE', 'LN_IC50']\n",
    "uniq_genes_gexpr = gexpr.columns[~gexpr.columns.isin(ignore)]\n",
    "uniq_genes_cnvg = cnvg.columns[~cnvg.columns.isin(ignore)]\n",
    "uniq_genes_cnvp = cnvp.columns[~cnvp.columns.isin(ignore)]\n",
    "uniq_genes_mut = mut.columns[~mut.columns.isin(ignore)]\n",
    "\n",
    "# Intersecting genes between all raw datasets.\n",
    "INTER_GENES = set(uniq_genes_gexpr)\\\n",
    "    .intersection(set(uniq_genes_cnvg))\\\n",
    "    .intersection(set(uniq_genes_cnvp))\\\n",
    "    .intersection(set(uniq_genes_mut))\n",
    "\n",
    "print(\"Number of intersecting cell-lines:\", len(INTER_CLS))\n",
    "print(\"Number of intersecting genes:\", len(INTER_GENES))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of intersecting genes: 858\n",
      "Number of intersecting cell-lines: 870\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of intersecting genes:\", len(INTER_GENES))\n",
    "# Assign unique index to each of the genes.\n",
    "inter_genes_df = pd.DataFrame({'GENE_SYMBOL': list(INTER_GENES)})\n",
    "inter_genes_df['GENE_INDEX'] = inter_genes_df.index\n",
    "inter_genes_df.to_csv(PROCESSED_DATA_PATH + 'sparse_inter_genes.csv', header=True, index=False)\n",
    "\n",
    "print(\"Number of intersecting cell-lines:\", len(INTER_CLS))\n",
    "pd.DataFrame({'CELL_LINE_NAME': list(INTER_CLS)})\\\n",
    "    .to_csv(PROCESSED_DATA_PATH + 'sparse_inter_cls.csv', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Expression\n",
      "===============\n",
      "Shape total: (398531, 861)\n",
      "Shape GDSC1: (276262, 861)\n",
      "Shape GDSC2: (122269, 861)\n",
      "CNV Gistic\n",
      "==========\n",
      "Shape total: (398531, 861)\n",
      "Shape GDSC1: (276262, 861)\n",
      "Shape GDSC2: (122269, 861)\n",
      "CNV Picnic\n",
      "==========\n",
      "Shape total: (398531, 861)\n",
      "Shape GDSC1: (276262, 861)\n",
      "Shape GDSC2: (122269, 861)\n",
      "Mutation\n",
      "========\n",
      "Shape total: (398531, 861)\n",
      "Shape GDSC1: (276262, 861)\n",
      "Shape GDSC2: (122269, 861)\n",
      "(870, 859)\n"
     ]
    }
   ],
   "source": [
    "def get_intersecting_subset(df: pd.DataFrame, title: str, inter_genes, inter_cls):\n",
    "    print(f\"{title}\\n{len(title)*'='}\")\n",
    "    df_inter = df[['DRUG_ID', 'DATASET', 'CELL_LINE_NAME'] + list(inter_genes)]\n",
    "    df_inter = df_inter[df_inter.CELL_LINE_NAME.isin(list(inter_cls))]\n",
    "    print(f\"Shape total: {df_inter.shape}\")\n",
    "    print(f\"Shape GDSC1: {df_inter[df_inter.DATASET=='GDSC1'].shape}\")\n",
    "    print(f\"Shape GDSC2: {df_inter[df_inter.DATASET=='GDSC2'].shape}\")\n",
    "    return df_inter\n",
    "\n",
    "gexpr2 = get_intersecting_subset(gexpr, 'Gene Expression', INTER_GENES, INTER_CLS)\n",
    "cnvg2 = get_intersecting_subset(cnvg, 'CNV Gistic', INTER_GENES, INTER_CLS)\n",
    "cnvp2 = get_intersecting_subset(cnvp, 'CNV Picnic', INTER_GENES, INTER_CLS)\n",
    "mut2 = get_intersecting_subset(mut, 'Mutation', INTER_GENES, INTER_CLS)\n",
    "\n",
    "def get_only_first_row_per_cell_line(df: pd.DataFrame):\n",
    "    res = df.groupby('CELL_LINE_NAME').first().reset_index()\n",
    "    return res.loc[:, ~res.columns.isin(['DRUG_ID', 'DATASET'])]\n",
    "\n",
    "gexpr3 = get_only_first_row_per_cell_line(gexpr2)\n",
    "cnvg3 = get_only_first_row_per_cell_line(cnvg2)\n",
    "cnvp3 = get_only_first_row_per_cell_line(cnvp2)\n",
    "mut3 = get_only_first_row_per_cell_line(mut2)\n",
    "assert gexpr3.shape == cnvg3.shape == cnvp3.shape == mut3.shape,\\\n",
    "    \"ERROR: Not all sparsed feature datasets have the same shape.\"\n",
    "print(gexpr3.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save all new feature dataframes.\n",
    "gexpr3.to_pickle(PROCESSED_DATA_PATH + 'sparse_gexpr.pkl')\n",
    "cnvg3.to_pickle(PROCESSED_DATA_PATH + 'sparse_cnvg.pkl')\n",
    "cnvp3.to_pickle(PROCESSED_DATA_PATH + 'sparse_cnvp.pkl')\n",
    "mut3.to_pickle(PROCESSED_DATA_PATH + 'sparse_mut.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<ins>__Summary:__</ins>\n",
    "- Each feature has a shape of (870, 859)\n",
    "  - There are 870 unique cell-line name rows \n",
    "  - There are 858 gene symbols as columns, since the first columns is the `CELL_LINE_NAME`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein-Protein Interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import Data\n",
    "from io import BytesIO\n",
    "from typing import List\n",
    "import gzip\n",
    "\n",
    "\n",
    "PATH_TO_LINKING_DATA = '../../datasets/string/'\n",
    "PATH_TO_SAVE_DATA_TO = '../../datasets/string/my_datasets/'\n",
    "PROTEIN_LINKS_FILE = '9606.protein.links.detailed.v11.5.txt.gz'\n",
    "PROTEIN_INFO_FILE = '9606.protein.info.v11.5.txt.gz'\n",
    "\n",
    "def read_protein_links(path: str):\n",
    "    print(f\"Start reading `{path}`...\")\n",
    "    contents = gzip.open(path, \"rb\").read()\n",
    "    data = BytesIO(contents)\n",
    "    protein_links = pd.read_csv(data, sep=' ')\n",
    "    print(\"Finished reading.\")\n",
    "\n",
    "    # Exclude the Homo Sapiens taxonomy ID from the protein columns.\n",
    "    protein_links.protein1 = protein_links.protein1.str[5:]\n",
    "    protein_links.protein2 = protein_links.protein2.str[5:]\n",
    "\n",
    "    return protein_links\n",
    "\n",
    "def read_protein_info(path: str):\n",
    "    print(f\"Start reading `{path}`...\")\n",
    "    contents = gzip.open(path, \"rb\").read()\n",
    "    data = BytesIO(contents)\n",
    "    protein_info_v1 = pd.read_csv(data, sep='\\t')\n",
    "    print(\"Finished reading.\")\n",
    "    protein_info_v2 = protein_info_v1.rename(columns={'#string_protein_id': 'string_protein_id'}, \n",
    "                                             inplace=False)\n",
    "\n",
    "    # Exclude the Homo Sapiens taxonomy ID from the protein columns.\n",
    "    protein_info_v2.string_protein_id = protein_info_v2.string_protein_id.str[5:]\n",
    "\n",
    "    return protein_info_v2\n",
    "\n",
    "def combine_protein_links_with_info(links: pd.DataFrame, infos: pd.DataFrame):\n",
    "    \"\"\"\n",
    "    Maps the corresponding gene symbols to the proteins in the protein link dataset.\n",
    "\n",
    "    Args:\n",
    "        links (pd.DataFrame): DataFrame containing at least the columns `protein1` and `protein2`.\n",
    "        infos (pd.DataFrame): DataFrame containing at least the columns `string_protein_id` and `preferred_name`.\n",
    "    Returns:\n",
    "        (pd.DataFrame): DataFrame containing the corresponding `gene_symbol` per `protein1` and `protein2`.\n",
    "    \"\"\"\n",
    "    # Get the gene symbols for the protein1 column.\n",
    "    res = links.merge(right=infos[['string_protein_id', 'preferred_name']],\n",
    "                      how='left',\n",
    "                      left_on='protein1',\n",
    "                      right_on='string_protein_id')\n",
    "    res.rename(columns={'preferred_name': 'gene_symbol1'}, inplace=True)\n",
    "    res.drop(['string_protein_id'], axis=1, inplace=True)\n",
    "\n",
    "    # Get the gene symbols for the protein2 column.\n",
    "    res = res.merge(right=infos[['string_protein_id', 'preferred_name']],\n",
    "                    how='left',\n",
    "                    left_on='protein2',\n",
    "                    right_on='string_protein_id')\n",
    "    res.rename(columns={'preferred_name': 'gene_symbol2'}, inplace=True)\n",
    "    res.drop(['string_protein_id'], axis=1, inplace=True)\n",
    "\n",
    "    # Drop all rows where the gene symbol has not been found.\n",
    "    res.dropna(subset=['gene_symbol1', 'gene_symbol2'], inplace=True)\n",
    "\n",
    "    assert not res[['gene_symbol1', 'gene_symbol2']].isna().sum().any(),\\\n",
    "        \"Some gene_symbol columns are missing!\"\n",
    "\n",
    "    return res \n",
    "\n",
    "def add_gene_index(proteins: pd.DataFrame, gene_idxs: pd.DataFrame):\n",
    "    \"\"\"Append corresponding index for each gene symbol column as new column(s).\n",
    "\n",
    "    Args\n",
    "        proteins (pd.DataFrame): Dataframe containing protein-protein interaction\n",
    "            information. Needs to contain `gene_symbol1` and `gene_symbol2` as \n",
    "            columns.\n",
    "        gene_idxs (pd.DataFrame): Datframe containing gene symbols and a \n",
    "            corresponding index. Needs to contain `GENE_SYMBOL` as a column name.\n",
    "    \"\"\"\n",
    "    temp = proteins.merge(right=gene_idxs,\n",
    "                          how='left',\n",
    "                          left_on=['gene_symbol1'],\n",
    "                          right_on=['GENE_SYMBOL'])\n",
    "    temp.rename(columns={'GENE_INDEX': 'index_gene_symbol1'}, inplace=True)\n",
    "    temp.drop(['GENE_SYMBOL'], axis=1, inplace=True)\n",
    "\n",
    "    res = temp.merge(right=gene_idxs,\n",
    "                     how='left',\n",
    "                     left_on=['gene_symbol2'],\n",
    "                     right_on=['GENE_SYMBOL'])\n",
    "    res.rename(columns={'GENE_INDEX': 'index_gene_symbol2'}, inplace=True)\n",
    "    res.drop(['GENE_SYMBOL'], axis=1, inplace=True) \n",
    "\n",
    "    del temp\n",
    "    return res    \n",
    "\n",
    "def shrink_to_only_singely_occuring_proteins(pinfo: pd.DataFrame):\n",
    "    freq_per_gene = pinfos\\\n",
    "        .groupby(['preferred_name']).size()\\\n",
    "        .reset_index(name='freq', inplace=False)\\\n",
    "        .sort_values(['freq'], ascending=False)\n",
    "    print(f\"There are {freq_per_gene[freq_per_gene.freq>1].shape[0]} proteins with ID frequency > 2. \"\\\n",
    "        + \"They will be deleted.\")\n",
    "\n",
    "    # Remove the gene symbols which have a frequency higher than 1.\n",
    "    return pinfos[~pinfos.preferred_name.isin(freq_per_gene[freq_per_gene.freq>1].preferred_name.tolist())] \n",
    "\n",
    "def get_neighbor_tuples(proteins: pd.DataFrame, genes: pd.DataFrame):\n",
    "    # Create a list of tuples which hold neighbor nodes.\n",
    "    gene_symbol_tuples = list(set(list(zip(proteins.gene_symbol1, proteins.gene_symbol2))))\n",
    "    print(gene_symbol_tuples[:5])\n",
    "\n",
    "    # Map the list of tuples of neighbor gene nodes to the corresponding int. \n",
    "    transform_gene_tuple_to_index_tuple = lambda x : (\n",
    "        genes[genes['GENE_SYMBOL']==x[0]]['GENE_INDEX'].values[0], \n",
    "        genes[genes['GENE_SYMBOL']==x[1]]['GENE_INDEX'].values[0])\n",
    "        \n",
    "    return [transform_gene_tuple_to_index_tuple(tup) for tup in gene_symbol_tuples]\n",
    "\n",
    "def is_undirected_list_of_tuples(neighbors: List[tuple[int]]):\n",
    "    for neigh in neighbors:\n",
    "        a, b = neigh[0], neigh[1]\n",
    "        found = False\n",
    "        for n in neighbors:\n",
    "            if n == (b, a):\n",
    "                found = True\n",
    "                break\n",
    "        if not found:\n",
    "            print(f\"ERROR: the list is directed since couldn't find ({b}, {a}) for {neigh}!\")\n",
    "            return False\n",
    "        # assert found, f\"ERROR: the list is directed since couldn't find ({b}, {a}) for {neigh}!\"\n",
    "    print(\"SUCCESS: The given list of tuples is undirected!\")\n",
    "    return True\n",
    "\n",
    "def create_graph_dict_with_indices(proteins: pd.DataFrame, nodes_as_indeces: List):\n",
    "    \"\"\"Creates a dictionary holidng as key the unique index of a gene symbol\n",
    "    and as value the list of gene symbol indices which are neighbors of the \n",
    "    gene symbol index of the key.\"\"\"\n",
    "\n",
    "    # Find neighbors using the indices.\n",
    "    dict_as_indices = {gene_node: [] for gene_node in NODES_AS_INDECES}  # or dict(zip(NODES_AS_INDECES, [[] for _ in range(len(NODES_AS_INDECES))]))\n",
    "\n",
    "    # Get the edges.\n",
    "    for gene in dict_as_indices:\n",
    "        exists_in_gene_symbol1 = not proteins.loc[proteins.index_gene_symbol1==gene].empty\n",
    "        exists_in_gene_symbol2 = not proteins.loc[proteins.index_gene_symbol2==gene].empty\n",
    "\n",
    "        neighbor_nodes = None \n",
    "        if exists_in_gene_symbol1:\n",
    "            neighbor_nodes = list(set(proteins.loc[proteins.index_gene_symbol1==gene].index_gene_symbol2))\n",
    "        elif exists_in_gene_symbol2: \n",
    "            neighbor_nodes = list(set(proteins.loc[proteins.index_gene_symbol2==gene].index_gene_symbol1))\n",
    "        else: \n",
    "            print(f\"The gene {gene} couldn't be found in the dataset!\") \n",
    "\n",
    "        # Set neighbors for the gene.\n",
    "        dict_as_indices[gene] += [neighbor_node for neighbor_node in neighbor_nodes if neighbor_node not in dict_as_indices[gene]]\n",
    "    \n",
    "    return dict_as_indices\n",
    "\n",
    "def get_proteins_and_genes_above_thresh(proteins: pd.DataFrame,\n",
    "                                        genes: pd.DataFrame,\n",
    "                                        thresh: float):\n",
    "    \"\"\"Return new proteins dataframe and gene dataframe by only selecting \n",
    "    data above the chosen threshold.\"\"\"\n",
    "    p_sub = proteins[proteins['combined_score'] > thresh]\n",
    "    gene_symbols = list(set(p_sub.gene_symbol1.unique()))\n",
    "\n",
    "    # Select only the gene symbols with a combined_score > SCORE_THRESH.\n",
    "    gene_sub = genes[genes.GENE_SYMBOL.isin(gene_symbols)].reset_index(drop=True)\n",
    "    gene_sub.loc[:, 'GENE_INDEX'] = gene_sub.index\n",
    "\n",
    "    return p_sub, gene_sub     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start reading `../../data/raw/9606.protein.links.detailed.v11.5.txt.gz`...\n",
      "Finished reading.\n",
      "Start reading `../../data/raw/9606.protein.info.v11.5.txt.gz`...\n",
      "Finished reading.\n",
      "(11938498, 10)\n",
      "(19566, 4)\n",
      "|    | protein1        | protein2        |   neighborhood |   fusion |   cooccurence |   coexpression |   experimental |   database |   textmining |   combined_score |\n",
      "|---:|:----------------|:----------------|---------------:|---------:|--------------:|---------------:|---------------:|-----------:|-------------:|-----------------:|\n",
      "|  0 | ENSP00000000233 | ENSP00000379496 |              0 |        0 |             0 |             54 |              0 |          0 |          144 |              155 |\n",
      "|  1 | ENSP00000000233 | ENSP00000314067 |              0 |        0 |             0 |              0 |            180 |          0 |           61 |              197 |\n",
      "|  2 | ENSP00000000233 | ENSP00000263116 |              0 |        0 |             0 |             62 |            152 |          0 |          101 |              222 |\n",
      "\n",
      "|    | string_protein_id   | preferred_name   |   protein_size | annotation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |\n",
      "|---:|:--------------------|:-----------------|---------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
      "|  0 | ENSP00000000233     | ARF5             |            180 | ADP-ribosylation factor 5; GTP-binding protein that functions as an allosteric activator of the cholera toxin catalytic subunit, an ADP- ribosyltransferase. Involved in protein trafficking; may modulate vesicle budding and uncoating within the Golgi apparatus; ARF GTPase family                                                                                                                                                                                                                                                                                                                                 |\n",
      "|  1 | ENSP00000000412     | M6PR             |            277 | Cation-dependent mannose-6-phosphate receptor; Transport of phosphorylated lysosomal enzymes from the Golgi complex and the cell surface to lysosomes. Lysosomal enzymes bearing phosphomannosyl residues bind specifically to mannose-6- phosphate receptors in the Golgi apparatus and the resulting receptor-ligand complex is transported to an acidic prelyosomal compartment where the low pH mediates the dissociation of the complex; MRH domain containing                                                                                                                                                    |\n",
      "|  2 | ENSP00000001008     | FKBP4            |            459 | Peptidyl-prolyl cis-trans isomerase FKBP4; Immunophilin protein with PPIase and co-chaperone activities. Component of steroid receptors heterocomplexes through interaction with heat-shock protein 90 (HSP90). May play a role in the intracellular trafficking of heterooligomeric forms of steroid hormone receptors between cytoplasm and nuclear compartments. The isomerase activity controls neuronal growth cones via regulation of TRPC1 channel opening. Acts also as a regulator of microtubule dynamics by inhibiting MAPT/TAU ability to promote microtubule assembly. May have a protective role a [...] |\n"
     ]
    }
   ],
   "source": [
    "plinks = read_protein_links(RAW_DATA_PATH + '9606.protein.links.detailed.v11.5.txt.gz')\n",
    "pinfos = read_protein_info(RAW_DATA_PATH + '9606.protein.info.v11.5.txt.gz')\n",
    "print(plinks.shape)\n",
    "print(pinfos.shape)\n",
    "print(plinks.head(3).to_markdown())\n",
    "print()\n",
    "print(pinfos.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3 proteins with ID frequency > 2. They will be deleted.\n",
      "(19560, 4)\n",
      "(11936122, 12)\n"
     ]
    }
   ],
   "source": [
    "pinfos2 = shrink_to_only_singely_occuring_proteins(pinfos)\n",
    "print(pinfos2.shape)\n",
    "\n",
    "proteins = combine_protein_links_with_info(plinks, pinfos2)\n",
    "print(proteins.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(83126, 12), thus 99.30 % got removed\n",
      "Number of unique gene_symbol1's: 858 (95.57 % got removed)\n",
      "Number of unique gene_symbol2's: 858 (95.57 % got removed)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000408295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>288</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>RFC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000306245</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>192</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>FOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000340989</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>157</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>SFN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "1940  ENSP00000001008  ENSP00000408295             0       0            0   \n",
       "1953  ENSP00000001008  ENSP00000306245             0       0            0   \n",
       "1978  ENSP00000001008  ENSP00000340989             0       0            0   \n",
       "\n",
       "      coexpression  experimental  database  textmining  combined_score  \\\n",
       "1940            81             0         0         258             288   \n",
       "1953             0            59         0         177             192   \n",
       "1978             0            78         0         123             157   \n",
       "\n",
       "     gene_symbol1 gene_symbol2  \n",
       "1940        FKBP4         RFC5  \n",
       "1953        FKBP4          FOS  \n",
       "1978        FKBP4          SFN  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inter_genes = pd.read_csv(PROCESSED_DATA_PATH + 'sparse_inter_genes.csv')\n",
    "\n",
    "# Select only the rows where both gene symbols are landmark genes\n",
    "proteins2 = proteins[(proteins.gene_symbol1.isin(inter_genes.GENE_SYMBOL)) &\n",
    "                     (proteins.gene_symbol2.isin(inter_genes.GENE_SYMBOL))]\n",
    "print(f\"{proteins2.shape}, thus {100-100*proteins2.shape[0]/proteins.shape[0]:2.2f} % got removed\")\n",
    "uniq_g1, uniq_g2 = len(proteins2.gene_symbol1.unique()), len(proteins2.gene_symbol2.unique())\n",
    "print(f\"Number of unique gene_symbol1's: {uniq_g1} ({100-100*uniq_g1/len(proteins.gene_symbol1.unique()):2.2f} % got removed)\")\n",
    "print(f\"Number of unique gene_symbol2's: {uniq_g2} ({100-100*uniq_g2/len(proteins.gene_symbol2.unique()):2.2f} % got removed)\")\n",
    "proteins2.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use `combined_score` threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAq0AAAGNCAYAAAA7GEgAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAC5XUlEQVR4nOzdd1zV1f/A8ded7D1li8iU4QBcuPe2YZmZtnfZ+LXra1baXpplaeWqHJl7pblz4EIQVFRUQFkuNhe49/cHeROBBAfX8X4+Hjzqfs7nc877c/l4ed/zOZ9zFAaDwYAQQgghhBA3MKWpAxBCCCGEEOJyJGkVQgghhBA3PElahRBCCCHEDU+SViGEEEIIccOTpFUIIYQQQtzwJGkVQgghhBA3PElabwPdunWjZ8+elJSU1CgbOXIkb7755nVrOyMjg6CgIHbu3Hnd2qiv/fv3069fP1q0aMFHH31Uo3zixIkEBQVV+wkNDaVDhw68/PLLnD592gRRV4+vZ8+e9d7/ev9uFyxYUOP9CgkJITY2lieffJITJ05ct7Yb6vz584wZM4bWrVvTs2dP1qxZc9ljXnvtNUaPHt3gsmux//Wi0+mYNm0aQ4YMoWXLlrRv354nnniCxMTEBtXTrVs3Jk+efJ2ivD7ef/99WrZsSevWrcnLy7tm9W7YsIGRI0fSsmVLoqKiGDp0KLNnz+ZGmk2yts+2S3/g+n9m1BVbQz7XanO56/HNN99k5MiRV9WGuDGoTR2AaBwnTpzg888/b/QPpBvJlClTUKvVLF++HBsbm1r38fT0ZM6cOcbXZWVlJCQk8O6773Lu3DmmTp3aWOHW8NBDDzFixIh67z9x4kTU6uv7T1ylUrFhwwbj64qKCg4cOMC4ceN48sknWbp0KQqF4rrGUB8TJkwgPT2duXPnsnTpUl588UXWr1+Po6Njo7T/5ptvotfrG6WtupSUlPDAAw9w9uxZnnvuOSIjIykqKmLGjBmMGDGC77//nrZt25o0xuslNTWVmTNn8u6779KxY0ecnZ2vSb2bNm3i6aef5uWXX2bs2LGoVCq2bt3KhAkTOHv2LM8888w1aedqPfTQQ9x7773G1x07duSdd96hV69eJoxKiIaTpPU24e3tzaxZs+jbty+tWrUydTgmUVBQQEhICD4+PnXuo1KpcHFxqbbNy8uLY8eO8fXXX1NQUFBnwnu9WVlZYWVlVe/97e3tr18wF7n0/WrSpAnnz5/n1Vdf5eDBgwQHBzdKHP8lOTmZtm3b0qxZM3r37s3kyZM5depUoyWtprpmLvbll19y7Ngxli5dipubm3H7hx9+yOnTp3nvvfdumC8Z11p+fj4AHTp0wMvL65rVO3fuXLp27VqtF93Pz4/c3FxmzJhxwySttX12WFtb1/i3K8SNToYH3CaGDh1Ky5YtefPNNykrK6t1n9pu5V+6beTIkXz99de8+uqrREVF0bFjR+bOncvOnTsZNGgQkZGRDB8+vMat4Z07d9KvXz/Cw8O57777OHr0qLFMr9fz3Xff0bVrV6Kiorjzzjur9d4tWLCA3r17M3bsWFq3bs0rr7xSa/yHDh3i0UcfJTo6mpiYGF555RXOnDkDVN0++vvvv1m4cCFBQUFkZGQ06P3TarUoFApUKlW9Y+7Tpw8LFy6kS5cuREZG8vzzz5Odnc1LL71EVFQUnTt35o8//jAec+7cOV5//XU6duxIWFgYHTt25KOPPjL20F18G+3C72XVqlUMHTqUyMhIBg8eXO2298W3+i7EM2fOHLp160bLli154IEHOHLkiHH/vLw8nn32WVq1akXHjh2ZOnUqPXv2ZMGCBQ16ry68XwAajca4bf78+QwYMICIiAh69uzJrFmzADhw4ABBQUHVYhkxYkS1XqDTp08THBxMSkoKeXl5PPPMM8TGxhIVFcXo0aNJSUn5z3g6dOjA77//zoEDB5gyZQqBgYHGW6LXQlBQEPPnz2fEiBFEREQY3+sLLh0esH37du68804iIiK4++67mT9/frV4arvdeem2NWvWMGjQIMLDw+nTpw/Tpk2rszdXp9OxYMEC7rrrrmoJ6wXvvPMOn332mTFhPXnyJC+88ALt2rWjZcuWPPXUU6Snp9dad223d2u7VtevX2+M96677iItLY2JEyfStm1bYmJieP/996sd//DDD/PNN9/QsWNHoqOjeeKJJ8jOzjbu8/3339O9e3datGhB7969mT17dq3xLViwgPvuuw+AHj168NprrwH//XkBVb/Tr776ik6dOtGpUydyc3Nr1K1UKklOTiYnJ6fa9tGjR1f7/ZeXl/PFF1/QuXNnoqKiuPfee9m7d6+xfOfOndx///3GIRvvv/++cTjXhffvu+++o127dvTt2xedTsepU6d47rnnaNWqFe3bt+eFF16o9v5cqcLCQl555RVatmxJbGwsH3zwAZWVlcb3srbP4p07d3LvvfcSERFB9+7d+eyzz6r9nVmwYAF9+/alRYsWdO3ala+//rratWowGJg8eTIdO3YkMjKSJ554otoQjoZcjwCzZ8+mW7duREZG8vLLL1NaWlqtvL7XjrjxSNJ6m1AoFIwfP56TJ08yceLEq6rrhx9+ICgoiCVLltC9e3fGjRvHu+++y1tvvcWsWbPIzs7m888/r3bMTz/9xIsvvsiCBQtwdnZm5MiRFBcXA/DZZ5+xYMECxo0bx6JFixg6dCjPPPMM27dvNx5/7NgxCgsLWbhwIY8//niNmDIyMhg+fDh2dnbMnj2byZMnc+DAAR566CEqKyuZP38+bdq0oW/fvmzevJkmTZrU61wNBgN79+5lxowZ9OjRA0tLy3rHnJGRweLFi/n+++/5+uuvjUlGREQECxYsIC4ujnfeeYfz588D8Oqrr3LkyBG+/fZbVq5cyZNPPslPP/3EX3/9VWd8H3/8MS+88ALz5s2jSZMmvPrqq8b3tbb3aMmSJXz99df8+OOPnDx5kvfeew+oSsIff/xxsrOzmT59OhMnTmTJkiX/+YehLqmpqXz77beEhYXRtGlToOr3/9577zFq1CgWL17Mww8/zMcff8yPP/5IcHAwTZo0YcuWLQAUFxeTkJDA8ePHjX+EN23aRJMmTQgJCeHdd9+loqKCX375hQULFmBlZcWzzz77nzE988wzWFhYMGTIEM6cOcO0adOu+dCJTz/9lBEjRvDHH3/Qpk0bxo4dS2ZmZo39jh07xiOPPEJkZCQLFy5k2LBhfPjhhw1qa8OGDbz88ss88MADLFu2jP/7v/9jxowZdY7rS09PJz8/n8jIyFrLvb29jT3ihYWFDB8+nPPnzzN16lRmzpxJQUEB999/PwUFBQ2K82ITJkzgrbfeYt68eZw7d4577rmHjIwMfvnlF1544QVmzpxZ7Yvf9u3bOXjwID/99BNffPEFe/bs4euvvwbgr7/+Ytq0abz//vusWrWKRx55hPfee4/4+Pga7fbr18/4vsybN48333zzsp8XF8ybN48pU6YwadKkWnslR40aRU5ODt26dePBBx/k22+/Zc+ePVhbWxuvfagaT/v777/z9ttvs2jRIkJCQnjkkUc4c+YMCQkJjB49mvDwcObPn8+ECRNYu3YtL7zwQrW2li1bxqxZs/j000+pqKhg5MiRmJmZ8dtvvzFt2jTKy8sZNWoUOp3uin9HACtXrsTX15fFixfz2muvMXPmTJYsWWIsv/SzOCUlhYcffpiePXuyZMkS3n//fdatW8fYsWOBqi+l77zzDi+88AKrV6/mjTfeYNq0aSxevNhYZ3p6OgcOHODnn39m6tSpJCYm8tlnnwENvx4XLlzIhAkTeOKJJ/jjjz9wd3dn6dKlxvKGXDvixiPDA24jfn5+PPvss3z++ef06dOHFi1aXFE9YWFhPPTQQwDcf//9/Pbbb4wePZqYmBgA+vbty/r166sdM2bMGHr06AHA+PHj6dSpE8uWLaNfv37MmDGDiRMnEhcXB4Cvry8HDhzg+++/JzY21ljHU089hbe3d60x/fLLL9ja2jJhwgRj794XX3xBv3792LRpE126dEGj0WBubv6ft8TS09Np2bKl8bVOp8Pe3p6+ffsa/4hcGAd4uZjLy8t555138PPzIzAwkODgYCwsLBg1ahQADz74IPPmzeP48eNEREQQFxdHbGwszZs3B6p6G6dOncrBgweN792lHn74YTp16gTA888/z5AhQzh8+DARERE19i0vL+fdd9+lWbNmQNXv7ssvvwRgx44dJCUlsWbNGuN7/MknnzBw4MA63yuAysrKau9XeXk5lpaWdO3alVdeeQWlUonBYGDq1KmMGjWKu+++G6i6FtPT0/nhhx948MEH6dy5M3///TcPPPAA8fHx+Pn5UVZWxo4dOxg4cCAbN26ka9euABw/fpygoCC8vLwwMzNj3LhxHD58GL1ej1JZ83t4bm4uzz//PCqVCmtraywtLXFxcaGkpAQzM7Naj7kSd955J/369QPglVdeYd68eezbtw9PT89q+82dOxd3d3feeustlEol/v7+pKWlMW3atHq39d133zF8+HDuuusuAHx8fCgqKuLtt9/mqaeeqnFOF26P29raXrbuRYsWkZ+fz+eff24cYvLVV1/RrVs3Fi9e3KBx1Rd7+OGHjZ8RF3rax40bh5mZGf7+/kycOJHU1FQ6d+4MVH1hHD9+PNbW1jRv3pxBgwbx999/A1Vj9DUaDR4eHnh6enL33Xfj5eWFv79/jXbNzc2xs7MDwNHRERsbG7799tvLfl5A1R2qkJCQOs+pVatWLFiwgB9//JH169cb4/Px8WHChAm0adOGwsJCfv/9d8aNG2f8d/zmm29ibm7OuXPn+PHHH2nRogWvvvoqAM2aNWPs2LE89thjpKamYmFhAVR9Hlz4tztv3jxKSkr48MMPjXd/Pv/8c2JjY1m9ejUDBgy4ot8RQMuWLXn66aeBqi8zP/30E0lJSQwZMsS4z8WfxS+//DKdO3fm4YcfBqo+C999913uu+8+XnjhBdLT01EoFHh4eBh/fvrpJ9zd3Y31aTQaPvzwQ2OnQN++fY0dAA29HmfPns2gQYMYNmyYMb5t27YZyxty7YgbjyStt5kHH3yQVatW8frrr1/RbV+o+lC64MIH6sXjRM3NzWt82784sbG2tsbf359Dhw4RFBSETqfj+eefr/aHtry8vNrDEgqF4j/HoqWmphIeHl7tdnSzZs1wcHDg0KFDxj9Cl9OkSRN+/vlnoCo5ev/99/H19WXMmDHGMWFHjhypV8xQ/X2xtLSsdg5mZmYAxvdq+PDhrF27lnnz5nHs2DEOHjxIVlbWfz7Ac3FvzoVxk+Xl5bXuq1Aoqv3ubG1tjfsmJyfj5ORU7UtBYGDgZZMclUrFwoULAcjJyeHjjz8G4MUXX8TJyQmAM2fOkJeXV2MsdXR0NFOnTuX06dN07dqVF198kYqKCrZt20ZsbCxFRUXs2LGDfv36sWXLFmPPy1NPPcWrr77K6tWriY6OplOnTgwZMqTW5PNCD7JWq2Xp0qXs3r2bJ554gilTppCXl8fGjRtZvXp1reemVqvrfO/1en2Nnlo/Pz/j/19432r7XRw+fJjQ0NBq8cbExDQoaU1JSSExMZHffvutWkylpaVkZmbW+HLn4OAAVA1BuZzU1FT8/f2rjYl2dHSkWbNmHDp0qN4xXurSfwuurq7GfwNQ83PD2dkZa2tr4+uLr9eBAwcyf/58evXqRWBgIB07dmTQoEHGa+5y6vt5UdeX5Is1b96cCRMmYDAYOHjwIBs3bmTGjBk8+uijrFmzhpMnT1JeXl7ti6RarTYmqRcn6he0adPGWHbhuItjSU5O5syZM8b9LigpKak2zOZKXPyZAmBnZ1ft9vqln8UpKSkcP3682mf8hZkTjhw5QlxcHJGRkdx55534+vrSsWNH+vXrh4eHh3F/V1dXY8J6oc0Lwwsaej2mpqZWS7ABoqKiOHjwIHD1144wLUlabzMqlYrx48czdOhQvvvuu8vuf/Gtsgtqu616uYc3LvQGXKDX69FqtcaxjxMnTqyWUAHV/qgrlUrjvrW5+I/fpe1c/IfpctRqtTEOX19fpk6dytChQxkzZgw//PADCoWi3jGrVKoaiVRdvXoGg4HHHnuMtLQ0Bg4cyODBg4mIiDD2ytaltnOra6odpVJZ43d3YV+VSlVrglafaXsufr9++OEHhgwZwmOPPca8efPQarV1/m4uXFtqtZp27dqh1+vZu3cvW7du5emnn6aoqIhvv/2WhIQEysvLjb10ffr0oX379mzYsIG///6byZMn8/PPPzN37twaXxoOHTrE/v37mTdvHjY2NnTu3JmXXnqJzz77DEtLS2PPb21sbW0pLCystSw/P9/Ye3dBbddnbe+fubl5je31uUYrKiqq7f/II4/U2hNe25hVHx8fnJycSEhIMPYGX2z79u389NNPxp7P2jTk39LFsV5w6bV3uR7u/3o/nZycWLx4Mbt27WLz5s1s2LCB6dOn89FHH1327gDU//Oirv2g6o7L559/zj333ENgYCAKhYLg4GCCg4Pp2bMnffr0Md41aGgsF87z4vfs4v00Gg0BAQFMmjSpxrFX+9Bfbb+Xi6/XSz+LNRoNQ4YM4dFHH61xnIuLC+bm5syaNYvExEQ2btzIpk2b+OWXX3jppZeMx1z69+HiNht6Pdb2t+ji/a722hGmJWNab0PNmzfnySefZMqUKdUemLrwD7uoqMi47dixY9ekzeTkZOP/nzt3jrS0NJo3b46vry8ajYbs7Gx8fX2NP0uWLGlQT3BAQACJiYnVerYOHz7M+fPnjbfUroSPjw+vvvoqmzZtMj5Yca1ivtjhw4fZvHkzEydO5IUXXqB///44ODiQm5vbKPM9BgUFcfbs2WrXw9GjRxs8htHe3p7333+fAwcOGMdOW1tb4+7uzq5du6rtu2vXLlxcXLCzs8PMzIy2bduyYsUKUlNTiYmJoV27dhw7dow5c+YQFxeHVquloqKCjz76iMzMTAYOHMiECRNYtmwZmZmZ7Nixo0Y8F3pvLp5j95FHHqFFixYUFhbW6OG6WIsWLThy5IhxzPEFOp2Offv2ER4e3qD35oLAwECSkpKqJXZJSUnV9tFoNNUS5sLCwmrnEBAQwLFjx6pdf4cOHeKLL76otU2lUsnQoUP5/fffazysYzAY+P7770lLS8PFxYWAgACOHj1arVf2zJkzpKWl1fpvSaPRVPvMgKq7FNfT8uXL+fXXX4mOjuaFF15g4cKFdOjQodo4yf9yLT4vLCwsWLp0KXPnzq1RdqGn3dnZGR8fH9RqdbXfsV6vp3fv3ixbtoyAgAD27NlT7fgL/1bqiqV58+ZkZGRgb29v/P07OTkxYcKEq+oNvxIBAQEcOXKk2rV45swZPvroI4qKitiyZQvffPMN4eHhPP300/z222/ce++91R5CvVz9Dbkeg4OD2b17d7VtF7/3V3vtCNOSpPU29fjjjxMQEEBWVpZxm6urK56envz8888cPXqUnTt38uWXX16TKXA++eQTNmzYwMGDB3n55ZdxdnamX79+WFhYMHr0aD777DOWL19Oeno6M2bM4JtvvqnXrbkLLgzKf/3110lNTWXnzp28/PLLBAcH065du6uK/e677yYmJoZPP/2U7OzsaxbzxWxtbVGr1axYsYKMjAz27NnDU089hU6nu+oHK+qjbdu2hIeH88orr5CUlMS+ffuMTwY39PffsWNHhgwZwo8//siBAwcAePLJJ5kxY4ZxDO/cuXOZNWsWo0ePNtbfpUsX5s6dS2BgIHZ2dri5udG0aVMWL15Mt27dgKqep/379/POO++QkJBAeno6c+bMQaPREBYWViMWHx8f4uLieP/999m6dStHjx7lo48+Mg5NuTAtV226d++Ou7s7Tz31FPHx8WRmZrJz506eeeYZNBoNd955Z4PelwtGjBhBYWEhb7/9NkeOHGHlypX88MMP1faJiopi2bJl7Nmzh9TUVF577bVqvVFPPvkky5Yt4/vvv+fYsWOsX7+ed955B3Nz8zrvSDz11FN4eXlx3333sXTpUtLT09mzZw/PPfcc8fHxfPDBBygUCgYNGoSjoyMvvvgiycnJ7N+/nxdffBFbW1v69+9fo96oqChOnz7Nzz//bHywauPGjVf03tSXTqfjo48+YvHixWRmZrJ161aSk5PrfNDsUtfi80KpVPLSSy8xa9Ys3n//fZKSkkhPT2fDhg3G2S3atGmDpaUl9913H1988QUbNmzg2LFjjBs3jvPnzxMbG8ujjz5KYmIiH330EUePHmXTpk28++67dO7cuc6kdeDAgTg4ODBmzBgSExM5dOgQL730EgkJCcYx8Y3l0UcfZd++fUyYMIEjR46wY8cOXn31VQoKCnBxcUGj0fDNN98wY8YM4zW3ffv2ev+uGno9Pvzww6xYsYLp06eTlpbG5MmTq31hvtprR5iWDA+4TanVasaPH1/t9qhCoeDjjz9m/PjxDBo0CF9fX15//XUee+yxq27vqaee4oMPPuDUqVPGsYwX/riOGTMGjUbDxx9/TF5eHt7e3owbN4477rij3vU7Ozvz448/8sknn3DnnXdiYWFBt27d+L//+78GDQ+ojUKhYNy4cQwePJj33nuPSZMmXZOYL+bm5sb48eOZOHEi06dPx83Njb59++Lm5tbg1Yqu1MSJE3n33XcZMWIENjY2PPbYYyQlJV3R+/faa6+xceNG3nrrLebMmcO9995LaWkpU6ZM4d1338Xb25vXXnvNOBURQNeuXfnf//5XbYL7du3aceLEiWo9op999hnjx4/n8ccfp6ioiObNm/PNN9/UGKpxwRdffMH48eN5/vnnKSsro2XLlsyaNQs/Pz+eeuopkpOTa53+yszMjNmzZ/P555/z0ksvcebMGezt7enQoQO//fbbFd+GdXBw4Mcff+T9999n8ODBBAQEcMcddzB9+nTjPi+++CLvvPMOo0ePxsbGhoceeqhaz2unTp34+OOPjTNTODo6MmTIkBpPnF/MysqKWbNm8cMPPzBp0iROnTqFjY0NkZGRzJkzx/jAkZmZGdOmTePDDz9kxIgRqFQq2rVrx+zZs2sd49y2bVueffZZfvjhB7744gs6derEc889d12nERoyZAinT59m4sSJnDp1CicnJ+644w6eeOKJeh1/rT4vhg0bhrOzM9OnT+fhhx+mqKgINzc3+vXrVy2W//u//0OlUvHGG29QVFREeHg406ZNw9nZGWdnZ7777ju+/PJLZs6cib29Pf3792fMmDF1tmtubs5PP/3Ehx9+yKhRo1AoFERFRTF9+vRGH5sZFBTElClT+Oqrr/jll1+wsbExPogJVeO1x48fz9SpU/n000+xtramR48edU5deKmGXo89evRgwoQJTJ48mU8//ZT27dszbNgw41jfq712hGkpDDfSWnNCCJM4c+YM+/btIy4uztijl5ubS8eOHZk9e3aNBz5uFQaD4YaYTH/RokW88sordfb6CiGEkJ5WIQRVD0I8//zzjB49mrvuuouioiK++uorfH19b+nbZjdCwiqEEKJ+JGkVQmBnZ2e8RTl9+nQ0Gg1t27blxx9/vOrhFUIIIW48idkHmLn3d7IKc/G282B0y7tp7tSUQl0R3+6YSVLOQSw1Ftwd1p9u/h2AqrtTvyYuYu3RLej1lXTya8uoqLuu2XzXlyPDA4QQQgghbiM5Rad5acU4RrW8m65N25GQlcyk7dP5vM/bTNs9B61Kw+PR93PiXCbjN07if13H4GvvxcrU9aw5spk3Oz+LAvhw02Q6+7Wlb2DXRolbZg8QQgghhLiN7D2VhI+dBz2adUSlVNHKI5zmTn5sTd9NfGYCw1oMQKvSEODkR0efaNYeqVpme+Ox7fQL7IaDhR32FnYMCenNmqObGy1uGR5Qh3PnzrF06VKaNWsmt0eFEEIIcUMrLy9n//79dOjQocbsJra2ttVmW9AbDGjV1afHU6AgITsFlVKFm/W/y5172LqxPaNqLuHMgiy8bP9dgtfDxo2M/FON9lCrJK11WLp0Ke+9956pwxBCCCGEqLdPPvmkxrZnnnmGZ5991vg60j2U2Ql/sDV9F9GeUSRlHyAp5yDNnZqiVVXvqNOqtOgqquYLL6vQYXZRsmum1mIwGCjXV9Q47nqQpLUO/v7+ALz99tsEBwdflzby8/MBLru+u7Qr7d7o7d5O5yrt3trt3k7nKu3eWu0eOHCA9957j48//pjWrVtXK7u03SY2rrzQ/hF+3beIqTt/pWWTFnT0jSGv6Ay6yvJq++oqdZhrqpbTNVNpq5WXVehQKZSNkrCCJK11ujDxfXBw8HWbo/LCsoyNPRm0tCvt3gptSrvS7q3SprQr7V5Lnp6eeHl5/ec+JeWlOFs68kmft4zb3vzzI3o370JSzkHyis7gbOUIwMn8bDxtm1TVbevOyfxsmjs1rSor+LesMciDWEIIIYQQt5ECXRFvrv2Eo2dOUFFZwarUDeQVnyXGK4o2nhH8sm8hZRU6Dp8+xuYT8cT5xgAQ5xvD4oN/crr4LOdK81mYsopOfjGNFrf0tAohhBBC3EZcrZx4tPVwPvv7ewrLimjq4M1bXZ7DXG3GE23u5/tdv/DEktcxV5txf+Qdxp7V3gGdOV+Wzxt/fkS5voI43xgGBPZotLglaRVCCCGEuM108oulk19sje3WZla82P7RWo9RKpXcGz6Ye8MHX+/wam/fJK0KIYQQQgjRAJK0CiGEEEKIG54krUIIIYQQ4oYnSasQQgghhLjhSdIqhBBCCCFueJK0CiGEEEKIG55MeWUi63el88vy3ZzNL8XGzoEH+obQpbW3qcMSQgghhLghSdJqAut3pTNpXgJUlgKQe7ak6jVI4iqEEEIIUQtJWk1gxooUysorMbtocEZZeSWTf9/HmfxSbK3MsLXWYmulxc7KDFsrLZbmahQKxVW3LT28QgghhLgZSdJqAnlnS2rdXlJWwU9Lk2stU6sU2FppqxJaK+1FP2bYWf/72s7633KNWlWtDunhFUIIIcTNSpJWE3B2sCC3lsTVxd6CSf/Xlfwi3UU/ZZwv/Pf1+cIy8ot0pJ3MJ7+ojILi8jrbsTBT/5PIViW3iUfyau3hnbEiRZJWIYQQQtzQJGk1gQf6hvzT4/nvNjONigf6hWBprsHSXIO7k1W96qqs1FNYUl4toc0v0nG+6N//zy/Uca6glDJdZa111NXzK4QQQghxo5Ck1QQu9GpeGFvq4mBxxWNLVSoldtZm2Fmb4e1m85/7PvT+6lp7eJ0dLBrcrhBCCCFEY5Kk1US6tPYm3M8SACcnp0Zps84e3r4hjdK+EEIIIcSVkqTVRDYd38HsvQs4U3oeZ0tHhkcMJs435rq22aW1N4fyk1ibuQ29uhRznTk9vHvJeFYhhBBC3PAkaTWBTcd3MCV+NrpKHQB5xWeYEj8b4LomrpuO72Dj6ZUYNDoUAGalbMxbQehxp+ueMJsiSRdCCCHEraNRk9aUtDN8v3AfmblFONqacW+vYLq08qKwWMdXc/aw73AeluYahvcKolesLwAGg4EZy1NYvf04lXoD3dp48/CgFqiUVXOWbtidwcwVKZwvLKNFM2eeuycKBxtzAI5knOOb+QmcyC7Aw9mKp+6KJNjXsTFPuVa/7ltkTFgv0FXq+HbHDFamrr9u7aadPUGFvvrDWOX6cr7bMZMtx+NRKVVVPwolKqUKtUKF8p//VpUpUSku2UepQqlQolaqLir7Z99//v9A7mFWpK6jXF8BXEjSZ1FeoaOrf4drMv9sXSRZFkIIIW4NjZa0VuoNfPDzdp68I5IOkR7sP3qaN7/dQoifIz8uScLcTM3MsX04diqfsT9spbm3PU097Fi2JY2dKdlMfLkrCmDctG0s35LGwDh/0k6eZ/LvCYx7rB1+HnZMWbCPb3/fxxujY9CVV/Lej9sZ1iOQXrG+rNuZzoSfdzD1zV5o1MrLxns9nS4+U+v2Cn0llhrz69bupQnrBeX6Cs6WnqdSr6fSUEml/p8fg55KfSUVhkr0ej0V/5TpDfqrjkVXWc53O2czZdcvWKjNsdCYY6k2x0JjgYWm5mtLjfk/+1lU/f8/ry0v7K82R6ms/ns1VY/2hbYlWRZCCCGunUZLWotKyjlfqKNSr8dgMKBQgFqtRKlQsC0piymvdUerURHo40Dnll6s3nacx++IYN2udAbF+eNoW5XM3dU9kF9XHWBgnD8bdmcQG+ZO0D+9p6MHhDFy7ErOFZRxOOMcCoWCfu2bAtAz1pdFG4+wIzmLDhEejXXatXKydCSvlsTV2dKRNzs/d93afWrJm7W262jhwEe93qh3PQaDwZjQ/meS+0/5639+WGddQ0P6UFJeSkl5KcUVJZSUl1KkKyav6IzxdWlFWb3iMlObYak2x1xjhqXaghPnM429uxfoKnVM2/Ub+aUFaFVatCoNZuqq/1b9aKt+1BrMVFrjNo1KjVJRvy87pkyWhRBCiFtVoyWttlZa+rX345NZu/jsl90YDAaeGxZFflEZaqWi2ryknq7WbE08BUBGTiHe7v9O5eTlYk16dgEGg4GMnEKCfB2qtWFlriEjp4CMnEJ8LpkCytPVmhNZBXSIqB5bfn4++fn51bbl5uZeq1OvYXjE4GpJDYBWpWV4xODr1mZd7RoqlbS0iWtQPQqFArWiamhAfTj/R5J+b/igyx6v1+sprSgzJrEl5aUUl5dS8s/r4vJSSspLKKko+6e8hJKKUsrPVtRaX3F5CdP3zq9X7BfT/JPYXkhmjUmtWmtMgLUqDfGZCbUO//h130JJWoUQQogr1GhJq15vwEyr5rUHookJc2fvoRw+nb2Ltx6KRaupnvyYaVTGifBLdZWYXVRuplWhN0B5hZ5SXQVm2kuO1aooK6+krLYyjYoyXc1EZvr06UyaNOlaneplXUhcGvv2cW3tnkv143yly3Vt92qTdKVSiaXWAkttw+aTratn2dnSgY97v4musrzqp0L3z//rKKv85/8ryv/5/3/L/mt7fmmBcXtdPcN5xWd5aeV7NLF2pYlN1Y+7tSseNq7Ymdte17G9QgghxM2u0ZLWrYmnOHTiLA8NDAMgOtSdNiFu/LrqILry6mMty8orMTerSjjNNKpq5WW6SlRKBVqNCjONGl159fGVZbpKzLVqY/J6ab0WZjVPedSoUQwdOrTatsTERMaMGXPF53s5cb4xhFo3Axpvntba2v16zh7+3neSiko9atX1GetrqiS97mR5CNba+q04diXqSpYt1Oa4WjmRmZ/FrlOJVF40xthcbWZMZt1tXKsltjZm1tctViGEEOJm0WhJa+65YsorqieRapWSZl527E87Tc7ZYlwdqibbz8wpNK7u5O1mTWZuoXHcakbuJWU5hcb6zheWUViiw9vNhpKyCpZuTqvWXmZOIZ1betWIzdbWFltb22rbsrKyrvKMbw7Roe78ueMEKWlnCA9wvm7tmCJJv9GS5UfaDDe2XamvJK/4DKcKcskqzOFUQQ6nCrI5cvYE2zL2VHvYzUprWWvvrLuNK5aa6r3P8gCYEEKIW1WjJa1Rga5MX5bCmh3H6R7tQ9LR02xNPMUHT7Yn+0wxM5al8MywSE5kFbBhTwb/e6QtAF1aebNg3WEiAlxQqRTMX3uIrq2rEs9OLb14ffJmesT40NzbnhnLU2gd7IatlZbI5s6UV+hZsukofdv7sW5nOucKy2gV5NpYp3xTiAp0QaNWsiM567omraZyoybLKqUKN2sX3KxdgNBqx1dUVpBTlMepwlxjMptVmENK7mE2Hd9RbV87M5uqnlkbV8rKy4g/uY+KalOLyQNgQgghbg2NlrT6NbHl9VHRzFqZwvcLk3BxsOCF4S1p7u3As8Oi+GZ+Ag+OW425mZoHB4QZe1b7dWjKucIyXvpqA+UVerq09mZw5wAA/D3teObuKL6es4ezBWWE+Tvx/D0tAdCoVYx9tC2T5ycwc0UyTZyseeuhWMxrGR5wO7MwUxMe4Ex8chYPD2ph6nBuGVeTLKtVajxs3fGwda9RpqvQkVWYS1ZhLicLsskqyOFUYS4Jp5I5W3q+5v6VOn5JkAfAhBBC3PwaNYOLCXMnJqzmH2IbSy2vPRBd6zEqpYKRfUMY2Tek1vK4KE/iojxrLWvqYccnz3W68oBvEzEhbnz3RyKZuYV4usj4yRuZVq3Fx94TH/ua1/ywOU/WeszpkrO8u+4LWjYJo1WTcDxt3eWhLyGEEDcd6XYUtAl1hz8SiU/OxrOzJK03q7qmFrNQm1NYVsSshD+YlfAHLlZOxgS2hWsgWrXWBNEKIYQQDSNJq8DN0RJfdxvik7MY0rmZqcMRV+hyD4DlFZ9h76n97D6ZxIa0baw+vBGNSkML1yBaNWlBS48WuFo13kwWQgghRENI0iqAqlkE/lh/mKKScqwsNKYOR1yByz0A5mzpSI9mcfRoFoeuspyU3FR2n0xi96kk9pxKgt3gZduEVh4taNmkBUHOzeq9gIQQQghxvUnSKgCIDnVj/l+p7DmUQ8fI2scIixtffR8A06o0RLqHEukeyoMM42RBNnv+SWCXHfqLxQf+xEJjTqRbKK08WhDVJAx7c9s66xNCCCGuN0laBQBBvo7YWGqIT86WpPU25GHjhkeQG/2DulNSXkpi9gFjD+y2jN0ANHPwpaVHC1o1aYG/ow9KRdViFDI3rBBCiMYgSasAqmZpaB3ixs6UbCr1BlRKebr8dmWhMSfGK4oYrygMBgPHz2Ww+1QSu08m8fv+5czfvww7MxuimoRhoTbnr7Qt6CrLAZkbVgghxPUjSaswigl1Z/2uDA4dP0tIU0dThyNuAAqFAj8Hb/wcvLkjtC/5ZYUknEpmz6kkdp1MpFBXVOMYXaWOX/ctkqRVCCHENXV9FpsXN6WWQa6olAriU26PJWxFw9maWRPnF8Nz7R5i6uCP69wvr/gMvyUuJiErmZLy0kaMUAghxK1KelqFkbWFhjB/J+KTs3mgX+jlDxC3NaVSWefcsGqlioUpq1iQvKKqt9beixDnAIJdqn7koS4hhBANJUmrqCY61I1pi/eTc6YYV0dLU4cjbnB1zQ37ePQI2nhEkHo6jQN5h0nJPcyao5tZnroOgCbWrgS5NDMmsu7WLrJKlxBCiP8kSauoJjrUnWmL9xOfkk3/Dk1NHY64wV1ubtgI9xAi3KuWYK6orCDtXDopuYc5kHeEXZn7WJ+2FQB7c1uCnQMIdmlGsHMAfvZeKJUyekkIIcS/JGkV1Xi6WOPhbEV8cpYkraJe6js3rFqlprlTU5o7NWUQPdEb9JzMzzb2xB7IPWycXstCbU6gc1OCnQMIcQkgwNGvxnKzMtWWEELcXiRpFTVEh7qz/O80SssqMDeTS0RcH0qFEi+7JnjZNaFHszig6gGuA7lHOJB3mAO5R5ibtBQDBlRKFc0cfKvGxDo342zJOabv/d04LEGm2hJCiFufZCSihuhQNxZtPMLe1Fzatmhi6nDEbcTZ0pGOvo509I0GoFBXxKG8o8ae2GWH1rL4wOpaj5WptoQQ4tYmSauoIczfCUtzNfHJ2ZK0CpOy1lrRyiOcVh7hAOgqdBw+c5yx6z6vdf/aZjIQQghxa5AnHUQNapWSVkGu7EzJQq83mDocIYy0ai2hrs1xtqx78Ytx677k7xO7qKisaMTIhBBCXG+StIpaRYe6cya/jKOZ500dihA1DI8YjFZV/cEsrUpDW69WZBfm8uXWqTyx5HVmJfxBVkGOiaIUQghxLcnwAFGr1sGuKBQQn5xFgLe9qcMRopr/mmpLr9ezLzuFNUc2s/TgGhYfWE24WxA9msUR7RGJWiUfe0IIcTOST29RKztrM4J9HdmRks3w3sGmDkeIGuqaakupVBLVJIyoJmGcKTnHuqN/s/boFr74eyp2ZjZ0adqO7s064m7tYqrQhRDC5A7mHeGn3XM5VZCDvYUtd4f1p6NvDIW6Ir7dMZOknINYaiy4O6w/3fw7AGAwGPg1cRFrj25Br6+kk19bRkXd1WjzakvSKuoUHerGjOUpnMkvxdHW3NThCNFgjhb23BnWj6EhfUjITmbNkc0sObiGRQdWE+4WTM9mcbTxiJDeVyHEbUWv1/PJ5u94pPVw2nq3IiU3lXHrviTQuRkz9/6OudqMHwZ/zIlzmYzfOIlmjr742nux6vAGdp9M4tPeb6EAPtw0mVWHN9A3sGujxC1jWkWdokPdAdiZkm3iSIS4OkqlkpZNWvB/HZ9g8oAPGNZiIKcKcvj87x94cumb/LJvIdmFuaYOUwghGkVReTH5ZYVUGioxGAwoUKJWqlEqFMRnJjCsxQC0Kg0BTn509Ilm7ZEtAGw8tp1+gd1wsLDD3sKOISG9WXN0c6PFLd0Lok6+7ja4OFgQn5xFr1hfU4cjxDXhaGnPXWH9uCOkD3uzkllzdDOLDqxmYcoqIt1D6O7fkTaekaiVKlOHKoQQDZabm0tGRka1bba2ttja2hpf25hZ0yugE19t/ZGJ237GYDDwRPT9FJQVoVKqcLto+JSHrRvbM/YAkFmQhZet+79lNm5k5J+qSnwViut8ZpK0iv+gUCiIDnFj7c50dOWVaDXyR1zcOpRKJa08WtDKowVnis/xV9oW1h7dwud//4CduS1dm7ajh39HXK2dTR2qEELU25gxY2pse+aZZ3j22WeNr/UGPWYqLS+2f5Q2HhHsy07hq20/8mrHJ9GqNNWO1aq06CqqVh8sq9BhdtGS2mZqLQaDgXJ9RY3jrgdJWsV/iglzZ/nfx0g8kkfrYDdThyPEdVHV+9qfO0L6sjdrP2uOVPW+LkpZTYR7MD2axdHaIwK1UsWm4ztqnbVACCFuBF9++SXh4eHVtl3cywqwI2Mvh88cY2TUnQBVi7g0acG8/cvQVZZX21dXqcNcYwaAmUpbrbysQodKoWyUhBUkaRWXEd7MGTOtivjkbElaxS2vqve1agWu08Vn+evoFv46+jefbfkee3NbAhz9SMhOofyfD+284jNMiZ8NIImrEOKG4OLigpeX13/uk1d8hvJLFmBRKVU0tfcmJfcweUVncLaqWsTlZH42nrZVq2N62rpzMj+b5k5Nq8oK/i1rDPIglvhPWo2KqOYuxCdnYTDI6lji9uFk6cDdLQbwzYD3eTXuKZo5+rLz5D5jwnqBrlLHr/sWmShKIYRouAi3EI6dy2Dd0b8xGAwk5xwiPiOBDr7RtPGM4Jd9Cymr0HH49DE2n4g3fimP841h8cE/OV18lnOl+SxMWUUnv8b7wi49reKyokPd2b4/ixNZBfg2sb38AULcQpRKJa09wmntEc6wOU/Wuk9e8ZlGexBBCCGulo+9Jy+2f5Q5SUv4ec88nC0deDp2FM0cfXmizf18v+sXnljyOuZqM+6PvMPYs9o7oDPny/J548+PKNdXEOcbw4DAHo0Wd6Mlret3pfPN/IRq28rKK+kV68vo/qF8NWcP+w7nYWmuYXivIOPT6gaDgRnLU1i9/TiVegPd2njz8KAWqJRVfxw27M5g5ooUzheW0aKZM8/dE4WDTdWcokcyzvHN/AROZBfg4WzFU3dFEuxb95rlonZtQlwB2JGcJUmruK05WzqSV3ym1rLnlr1DnF8snXxjcLdxbeTIhBCiYdp4RtDGM6LGdmszK15s/2itxyiVSu4NH8y94YOvd3i1t99YDXVp7c28CQOMP289GIuDjRn39gxi4ry9mJupmTm2D6+PiubnpftJO1m15v2yLWnsTMlm4std+faVbqSknWb5ljQA0k6eZ/LvCfzf/a2ZNa4vDjZmfPv7PgB05ZW89+N2esT48Nv7/RjY0Z8JP++gvELfWKd8y3CysyDAy474ZJmvVdzehkcMRqvSVtumVWno6R+Hq7Uzv+9fznPL/8dbaz5h9eENFJYVmShSIYS49ZhkTGtJWQVf/LabJ+6IxMpCw7akLEb0DkarURHo40Dnll6s3nYcgHW70hkU54+jrTkOtubc1T2QVduOAVW9rLFh7gT5OmKmUTF6QBjb92dxrqCMfYfzUCgU9GvfFLVKSc9YX2wstexIzjLFKd/0okPdOXj8DOcLy0wdihAmE+cbw+PRI3A0twOqel4fj76fR6Pv4+0uzzN54AeMiBhKSUUpU3f9xqOLX+XTzVPYkbG3xlhYIYQQDWOSMa2/r0vFt4kt7cKbcCTjHGqlAncnK2O5p6s1WxNPAZCRU4i3u42xzMvFmvTsAgwGAxk5hQT5OhjLbK20WJlryMgpICOnEB+3f4+7UO+JrAI6XNIbnp+fT35+frVtubmyOs7FokPd+HX1QXYdyKFbG29ThyOEycT5xhBq3QwAJyenamVOlg4MDunFoOCeHD+XwYZj29l8Ip4dmXux1lrR3rs1nfxiae7UVMa/CiFEAzV60lpSVsHSzWmMfbQtAKW6mpPWm2lUlOkqjeVmF5WbaVXoDVBeoadUV4GZ9pJjtSrKyispq61Mo6JMV32KB4Dp06czadKka3J+t6pmnvY42JgRn5wlSasQl6FQKPBz8MbPwZv7I4eSmH2Ajce2s/7YVlYf2Yi7tQtxvjF08outtvKMEEKIujV60rot6RSuDhbGB6LMtCp05ZXV9ikrr8TcrCrhNNNULy/TVaJSKtBqVJhp1OjKq49RLdNVYq5VG5PXS+u1MKt5yqNGjWLo0KHVtiUmJta6qsTtSqlUEB3qzuaETCoq9ahVMluaEPWhUqqIahJGVJMwistL2JGxl43HtjN//3Lm7V9GkHMzOvnG0s6nFdZaq8tXKIQQt6lGT1p37M+iY6Sn8bWHsxUVegM5Z4txdbAEIDOnEO9/bu17u1mTmVtI0D9JbkbuJWU5hca6zheWUViiw9vNxtije7HMnEI6t6w54e6la/ICZGXJ2NdLRYe6sXr7cZLTThMRIL1DQjSUpcaCLk3b0aVpO/KKz7D5eDwbj23nh12/8NOeubT2CKeTXywt3cNQq2RGQiGEuFijfyoePHGWvu39jK8tzTXEhrkzY1kKzwyL5ERWARv2ZPC/R6qGD3Rp5c2CdYeJCHBBpVIwf+0hurauSjw7tfTi9cmb6RHjQ3Nve2YsT6F1sBu2VloimztTXqFnyaaj9G3vx7qd6ZwrLKNVkExFc6Uim7ugUSuJT86WpFWIq+Rs6ciQkN4MDu5F2tkTbDy2nS0ndrI9Yw82Wiva+7Shk18sAY5+KBQKWT5WCHHba9SktVJv4PS5EuM8qhc8OyyKb+Yn8OC41ZibqXlwQJixZ7Vfh6acKyzjpa82UF6hp0trbwZ3DgDA39OOZ+6O4us5ezhbUEaYvxPP39MSAI1axdhH2zJ5fgIzVyTTxMmatx6KxbyW4QGifizM1IQHOBOfnMXDg1qYOhwhbgkKhQJ/R1/8HX25P+pO9mUls/HYdv5K+5tVhzfQxMYVHztPdp9KkuVjhRC3tUbN4FRKBYs+rTkhrY2lltceiK7zmJF9QxjZN6TW8rgoT+KiPGsta+phxyfPdbrygEUNMSFufPdHIpm5hXi6WJs6HCFuKWqlilYe4bTyCKdYV8K2jD1sPLaN7Rl7aux7YflYSVqFELcLeZpGNEibUHcAWWhAiOvMUmtBN//2jO32Yp371LU6lxBC3IokaRUN4uZoia+7DfGySIMQjcbZsu7lp99c8zHr07ZSVqFrxIiEEKLxSdIqGiw61J39R09TWCIr/AjRGOpaPrajTwzFuhIm75jB44tf48ddczhxLtNEUQohxPUlTyWJBosJdWf+X6nsOZhT53hiIcS1c2Hcam2zBxgMBlJyD7PmyCbWHN3MysPrCXTyp0ezjrTzbo2ZWnuZ2oUQ4uYgSatosEBfB2wstcQnZ0nSKkQjqWv5WIVCQahrc0JdmzO6rJCNx7ax5shmJu+YwfQ984jzi6Vnszi87TxMFboQQlwTkrSKBlMpFbQJcWVnSg6VegMqpayhLsSNwNbMmgFBPegf2J2U3FT+PLKJNUc2szJ1PUFO/vRoFkc771ZopfdVCHETkqRVXJHoUHfW7crg0PGzhDSt+yERIUTjq+p9DSTUNZD8skI2pG1jzdFNfLNjOj/vmUsnv7b0aNZRel+FEDcVSVrFFWkZ5IpKqSA+JUuSViFuYLZm1gwM7sGAoO4k56ay5sgm/jyyiRWp6whybkbPZnG09Wopva9CiBueJK3iilhbaAjzdyI+OZsH+oWaOhwhxGUoFArCXAMJcw0kv7SA9ce2sfbIZiZt/5mf9syls28sPZrF4WXXxNShCiFErSRpFVcsOtSNaYv3k3OmGFdHS1OHI4SoJ1tzGwYF92RgUA/25xxizZFNrDqykeWp6wh2bkaPi3pfNx3fUeusBUII0dgkaRVXLDrUnWmL9xOfnEX/jv6mDkcI0UAKhYIWbkG0cAv6p/d1K2su6n0NcPQjOSeVcn3VnMx5xWeYEj8bQBJXIUS9Hcg9jIetO7Zm1mw5Ec+mYzto7tSUoaF9UCrqv2SALC4grpinizUezlbsSJElXYW42VX1vvbiy35jeafL80S6hZCQlWxMWC/QVer4dd8iE0UphLjZrErdwNh1X5B+/iTHzmYwaft0DMCqwxuYm7SkQXVJ0iquSkyYO/tS8ygpqzB1KEKIa0CpUNLCLZgx7R+pc5/TxWcaMSIhxM1seepfPNZmBGGugWw+sQMfOw9e7/Q0z7Z9kI3HdjSoLklaxVWJDnWjolJPQmquqUMRQlxjzpa1zwyiVWnJKcxr5GiEEDejvKIzRLgHA5BwKpmWTVoA0MTalfNlBQ2qS5JWcVVCmzphaa4mPlmGCAhxqxkeMRitqvpUWCqFkgp9Bc+vGMvPe+ZRUFZoouiEEDcDRwt7sgvzyCrM5cT5k0S5V804dCDvMM4WDg2qSx7EEldFrVLSKsiVnSlZ6PUGlLI6lhC3jAsPW106e0CYSyBzk5awInUd69O2MiSkN/2ad5W5XoUQNXRv1pHPt3yPWqXG286DYJcAVqVuYGbC79wbPqhBdUnSKq5adKg7mxNOcjTzPAHe9qYORwhxDcX5xhBq3QwAJycn4/YnYkbSP6g7s/ct5Jd9C1mVuoF7wgfSyTcWpVJu4gkhqgwJ6Y2XbROyC3ONX4RtzKx5rM0IOvnFNqguSVrFVWsd7IpCATuSsyRpFeI24m3nwWtxT5GUfZDZCX8weccMlh1cy4jIO4hqIouOCCGqtPGMACCn6DRW+kpivKJQK1UNrke+DourZmdtRrCvI/HJWaYORQhhAi3cgvig5ys83+4hSipKGb9xIu+t/4q0s+mmDk0IYWIGg4H5+5czcv7zPLfsHfKKzzBp2098u2MmFfrKBtUlSau4JqJD3TiccZ7T50tMHYoQwgSUCiUdfKL5ou//GN3ybtLOpvPq6vFM3PYTuUWnTR2eEMJElh5cy19Ht/B49Ag0yqob/O192rD7ZCK/JS5uUF2StIprIibUHYCdKTkmjkQIYUoalYZ+gd2Y2H8cQ0J6sy1jD88vH8vMvb9TqCsydXhCiEb2V9oWHm59Lx19Y1Aoqh7WjvGK4smYkWw5Ed+guiRpFdeEj7sNrg4WMkRACAGAldaS+yKG8FW/sXT0iWbpwbU8u+wdlhxYg66y/PIVCCFuCTlFp/G2bVJju4eNG/kNnDJPklZxTSgUCqJD3dmbmouuvGFjVIQQty5nS0eein2Aj3u/QXNHP2Ym/M4Ly8ey6dgO9Aa9qcMTQlxnHjZuJOem1ti+LWMPHjZuDapLklZxzUSHulGmqyTxiKyUI4Soztfeizc6P8tbnZ/DWmvFxO0/8frqD0nMPmDq0IQQ19HdYf2ZtnsOsxL+oNKgZ9PxeCZt+5k5iYsZEtKrQXVJ0iqumfBmzphpVbI6lhCiThHuIUzo9RrPtX2QQl0R763/ivEbJnL8XIapQxNCXAcxXlG80O4RDuUdQalQsuzQWnKK8ng17mk6+EQ3qC6Zp1VcM1qNiqjmLsQnZ/H40HDjgGshhLiYUqGko28MMV4tWX14A78nr+CVVePp5BfLPeEDcbZ0ZNPxHTVW4rowMbkQ4ubSyqMFrTxaXHU9jZq05p0r4Zv5Cew/ehpLczV3dA1gUFwzCot1fDVnD/sO52FprmF4ryB6xfoCVfN7zViewurtx6nUG+jWxpuHB7VA9c9yoRt2ZzBzRQrnC8to0cyZ5+6JwsHGHIAjGef4Zn4CJ7IL8HC24qm7Ign2dWzMU77tRIe6s31/FsezCvBrYmvqcIQQNzCtSsOAoB50adqOhSmrWHFoHX+n7yLcNYiknIPGB7byis8wJX42gCSuQtyEDuYd4VBeGhX6CgwYqpXdEdq33vU0WtJqMBj44KfthAe48OaDMWTmFvLapM0093Jg4cbDmJupmTm2D8dO5TP2h60097anqYcdy7aksTMlm4kvd0UBjJu2jeVb0hgY50/ayfNM/j2BcY+1w8/DjikL9vHt7/t4Y3QMuvJK3vtxO8N6BNIr1pd1O9OZ8PMOpr7ZC41aRkVcL21CXAGIT86SpFUIUS/WWivuj7yD3gGdmZO4hI3Ht9fYR1ep49d9iyRpFeImM3//cuYlLcVSa4GlxqJamYIbNGk9eOIsZ/JLGdU/FJVSga+7LZ88F4dWrWJbUhZTXuuOVqMi0MeBzi29WL3tOI/fEcG6XekMivPH0baq9/Su7oH8uuoAA+P82bA7g9gwd4L+6T0dPSCMkWNXcq6gjMMZ51AoFPRr3xSAnrG+LNp4hB3JWXSI8Gis077tONlZEOBlR3xyNnd3DzR1OEKIm4iLlRPPtB1da9IKcLr4TCNHJIS4WhvStnJHaF/uCR941XU1WpfjkYzz+LjZ8vPS/Ywcu5LHJ6zh4PGzFBTrUCsVuDtZGff1dLXmRHYBABk5hXi72xjLvFysSc8uwGAwVJW5/Vtma6XFylxDRk4BGTmF+FxUZqw3q+A6n6mICXXnwPEznC8sM3UoQoibkLNl7cO4zNRmZBXmNnI0Qoirca40n85+sdekrkZLWguLdew7koetlZYf3+rFmHtbMeWPREp1lWg1qmr7mmlUlOmq5vos1VVidlG5mVaF3gDlFXpKdRWYaS85VquirLySstrKNCrKdBU1YsvPzycjI6PaT26ufDBeqehQdwwG2HVAVscSQjTc8IjBaFXaatuUCiVlFTqeX/Y/PtvyPamn00wUnRCiIVq4BdU6T+uVaLThARq1EhtLjfGWcUhTR9pHNOGXVQdqTEZfVl6JuVlVwmmmUVUrL9NVolIq0GpUmGnU6MqrT05dpqvEXKs2Jq+X1mthVvOUp0+fzqRJk67JeQrw97TD0daM+OQsurXxNnU4QoibzIVxq5fOHtDCNYgVqev48/BGtmfsIdi5GYOCe9LKIxylQp5VEOJGFOYayE975pGYfYAmNq6oldXzsBtyTKunizWlukoqK/WoVFUfLnq9AX9PO5KOnibnbDGuDpYAZF5029/bzZrM3ELjuNWM3EvKcv5dAux8YRmFJTq83WwoKatg6ebq38Qzcwrp3NKrRmyjRo1i6NCh1bYlJiYyZsyYa3PytxmlUkGbEHc2J2RSUalHrZI/JkKIhonzjSHUuhkATk5Oxu33RQxhaEgf1qX9zbKDa/l483d42LgxIKgHnfxi0ao0pgpZCFGLFanrsTWz5tDpNA5dcofkhn0QKyrIFWsLDT8vS2Z0/1AOnTjHtqRTjHu8PdlnipmxLIVnhkVyIquADXsy+N8jbQHo0sqbBesOExHggkqlYP7aQ3RtXZV4dmrpxeuTN9Mjxofm3vbMWJ5C62A3bK20RDZ3prxCz5JNR+nb3o91O9M5V1hGqyDXGrHZ2tpia1v9SfesrKzr/6bcwqJD3Vi9/TjJaaeJCHAxdThCiFuIhcacfoHd6B3Qma3pu1ly4E++3zmbOYmL6dO8C70COmFjZm3qMIUQwDcD3r9mdTVa0mqmUTH+qQ5MWZDI/f9biaW5mseGhBPs68izw6L4Zn4CD45bjbmZmgcHhBl7Vvt1aMq5wjJe+moD5RV6urT2ZnDnAKDqNvQzd0fx9Zw9nC0oI8zfiefvaQmARq1i7KNtmTw/gZkrkmniZM1bD8ViXsvwAHHtRTZ3QaNWsmN/tiStQojrQqVU0dE3mg4+bdifc5AlB9cwJ2kJC1NW0bVpe/oHdcPNWj5/hLgRpJ5OIzM/i1ivlpwuPksTG1dUStXlD7xIo2ZwHs7WvPtYuxrbbSy1vPZA7Ut5qZQKRvYNYWTfkFrL46I8iYvyrLWsqYcdnzzX6coDFlfMwkxNeIAz8clZPDL46lfBEEKIuigUClq4BdPCLZgT5zJZenAtfx7dxKojG4j1asmgoJ4EOPmZOkwhbhibju3g+12/VNumq9DRzb8DIyKH8O2OmSTlHMRSY8HdYf3p5t8BqJpz/9fERaw9ugW9vpJOfm0ZFXUXSmXdwwCLdSV8umUK+3MOgQJCXAKYte8PsgtyeavLczhZOtQ7bhlsKK6bmBA3TuYVkZlbePmdhRDiGvCx9+Sp2Af4pv/7DArqyb6sFN5Y8xH/++tzdp1MRG/QX74SIW5xcX4xzLzzS+PPKx2fwN7clrvC+jElfjbmajN+GPwxL7V/jFkJf3D8XAYAqw5vYPfJJD7t/RZf9P0fB/OOsOrwhv9sa9a+P6g06Pl24HjM/pkV5KGWw7DUmDNz7+8NiluSVnHdRIe6A1WrYwkhRGNytLRnRORQJg/8gAei7iSnKI+PNk3mpZXv8dfRLZT/s0SsELe70vJSvtk+nYdb34uVxoL4zASGtRiAVqUhwMmPjj7RrD2yBYCNx7bTL7AbDhZ22FvYMSSkN2uObv7P+vecTGJk5B04Wtobt7laO/NQ63tJzDnYoFglaRXXjaujJX5NbIlPzjZ1KEKI25SlxoIBQT2Y2P89no19ELVSzXfxs3h66VssSF5Boa7I1CEKYVKLDvyJj50HMV5RnCrMRaVUVRsL7mHrRnr+SQAyC7LwsnX/t8zGjYz8UxgMhjrrL9QVYa21rLFdo1Sja+CXxysa01pRqSft5Hk8XayxNL+1pxf566+/SEpKMr4OCwsjOjqa8vJyZs+eXWP/qKgooqKiKC4uZu7cuTXK27RpQ4sWLTh//jwLFiwAQKP59z1s164dQUFB5OXlsXTp0hrHd+rUCX9/f7Kysli5cmWN8u7du+Pt7U16ejpr166tUd6nTx80Gg3p6eksWbKkRvmAAQNwdnbm4MGDbN26tUb50KFDsbOzIykpiZ07d9YoHzZsGJaWluzdu5e9e/fipSwm91QJ0348hL6ygoEDq5Zxi4+PZ//+/TWOHz16NAB///03hw4dqlam0WgYMWIEABs2bCAtrfrUGZaWlgwbNgyANWvWkJFRdTujvLzqH4WTkxN33HEHACtXrqwxQ4STk5MxviVLlnD69Olq5e7u7vTp0weABQsWkJ+fX63cy8uLHj16ADB37lzOnz9vjBugadOmdO7cGYDZs2cb47ogMDCQ9u3bA/Dzzz/XeG/qe+2VlJSwYsWKatcVVL/2/vjjjxrHX821V15eTrt27XBycvrPa8/d3Z2jR4+ycePGGuVXcu1deA81Gk2Na+9SI0aMQKPRXJNr70L5hfe4rmvvAltb22t27a1evZqysuqrzV167RUXF1crv9prz8/Pj/Dw8GvyudeQa+9CnN27d7/qzz13d3c89c6EnfLET+dA3qmzbDu4mR2KLfjHBjGkdT+27Pmb+G3xVOgr0ajUuFo5Y2du0+DPvUvV99rbvXs3mZmZ1cqu9HPvgstde1ZWVnTr1g24Np979b32Lvw3LCzsmnzu1ffau/gzAxrnb667uzvp6enEx8fX+Fy+ln9z//rrLwByc3NrvQ4unSUJqnpZV6au4/VOzwBQVlFWY9o4rUqLrkL3T7kOM/W/C3+YqbUYDAbK9RV1TjcX7BLA2qNbGBFZNbWoAgV6vZ6FB1YT6NS01mPqUq+kNedMMV/P3cMD/ULxa2LLq5M2cSTzPNYWGsY91p4Ab/sGNSpuHzaWWnLPFlNYrMPSTDr2hRCmZ6W1xEprSWlFGaeLz7LleDxrTv2NbaEZjvqqJcXLKys4WSB3icTNp7Y55p955hmeffbZGtt3ZCbgbOVEoLM/8E+Ceknvp65Sh7nGDACzS8rLKnSoFMr/nB95ZOQdvLvuC/bnHKJcX8GPu+eQmZ9FYXkxb3d+vkHnpjD8V5/uPz74aTsFxeW8eF8rEg7lMm1xEuMeb8+a+BNkZBcy/qkODWr0ZrBz505GjBjB7NmzadOmzXVp48I32osnzm4Mjdlupd7AyP+tpE2IK6N6+zVauxe7Hd5nU7d7O52rtHvrtZtXfIaXVrxHSUVpjTJnS0cmD/zgurYPt/57LO1e/3Yv5C1ffvkl4eHh1crq6mn94u+p+Np7Gif4Lykv5cE/XmJS//dwtqqaevTHXXNAAQ+1uoc3/vyI3gGd6dy0ai79bem7mb9/OZ/2ees/YztTco7Vhzdw7GwGaqUaL7sm9GneBXvzmjH9l3r1tCYezuOjZ+JwdbAkPiWb1iFuBPo4YG2p4fnP1jeoQXF7USkVtAlxZWdKDvf39EWlVJg6JCGEqMbZ0pHSWhJWqEpoS8pLsdCYN3JUQlwZFxcXvLxqrv5Zm9TTafRsFmd8baExp41nBL/sW8jj0feTfv4km0/E83qnp4GqleoWH/yTFm5BqJQqFqasopNfzGXbcbSw597wwVd2QhepV9JqAMy0Kir1Bval5hrn3dSV69Go5Zav+G/Roe6s25VB2slzBHjVfz42IYRoLE6WjuQVn6m17Iklr9O9aQf6NO+Cq7VzI0cmxPWh1+s5XXIWBwu7atufaHM/3+/6hSeWvI652oz7I++g+T9jT3sHdOZ8WT5v/PkR5foK4nxjGBDYo9b6yyvLWZf2Nz2axaFUKJmw8Rt0lTpjeZhrEHeF9WtQzPVKWoN8HJj/Vyq2VlrKyiuJDnXn9PkSZixPJtjPsUENittPyyBXVEoFiYdPS9IqhLghDY8YzJT42dX+qGpVWgYH9+JkQRYrUtexLPUvoj0j6R/YjWDnABQKuXMkbl5KpZI5wybX2G5tZsWL7R+t85h7wwdftte0SFfMO399xrnSfFo2aYGLlRMpuam0bNICC405Z0vOM3//Mtp4hOPn4F3vmOuVtD5+RwSfzNpJVl4RDw9qgZ21GVP+2EdGTiFjH2lb78bE7cnaQkOYvxOJh3MZ2iXA1OEIIUQNcb5Vtzhn713AmdLzOFs6MjxisHH7/ZF3sOrwBtYc2cyOjL00dfCmX/NutPdpjeY/HkIR4na0MGUVaoWKr/u9i9VF013dFzHYOJ3W//76nJWp63kiZmS9661X0mpnpeXLF7pU23Zf72AeGxIu3zRFvTjYmnPwaBGPf7gGWzsHHugbQpfW9f92JYQQ11ucbwyh1s2Amg/MOFk6cF/EEO4M7cem49tZdugvvtkxnVn7/qB3QCd6NovDroEPlQhxq4rPTGBk1J3VEtZLDQjqzvQ98xpUb72S1vv/t5KQpo7EhjUhNsydJs5W2FhqL3+gEMD6Xels3XeSC19vcs+WMGleAoAkrkKIm4qZWkuPZnF09+/IvuwUlh/6i7lJS/kjeSUdfKPp17wbfg71ewhGiFtVbvEZfO09q20Lcw00LuMK4GfvxdnS/EsP/U/1Slo/fjaO3Qdz2JKQyU9L9+PpYkVMqDuxYU0IaSpjWsV/m7EiBV2FnounaS0rr2TGihRJWoUQNyWFQkGkeyiR7qFk5mex4tA6Nhzbxvq0rYS5BtI/sButmoSjVMrDyuL2Y6E2o7S8+iIor8Y9Ve11ka4Ya03dPbG1qVfSGujjQKCPA/f2DKKgWMfuAzlsTTrF65M3Y2OlZebYPg1qVNxe8s6WNGi7EELcTDxt3XmkzXDuDR/E2qNbWJm6no83f4ebtQt9m3eha9P2MmWWuK342nuy8+Q+vOya1LlPfGYC/o4+Daq33l8BK/UGDhw/w8qtx1gbf4JdKdloNSoCvOwb1KC4/Tg7WDRouxBC3IyszawYHNKLiQPeY0y7R7A3s+HnPfN4YsnrTN8zn5zCPFOHKESj6ObfgT+SV7IvK6XW8v05h1h84M9qc8TWR716Wt+Z8jcHjp/BYKjqdQ0PcGZ4r2ACfR1ksnhxWQ/0Dakaw1r57za1SsEDfUNMF5QQQlwnaqWK9j6tae/TmsOnj7H80F+sTF3H8tS/iPaIpF9gN0JcZMoscevq4BNNYvZBPtg4kQi3YMJcg7A1s6ZQV0RK7mH2ntpPj2ZxtPIIv3xlF6lX0pqRW0iZrpKoQFdaBrkS2dyZph52lz9QCP592OqX5bs5m1+KWqXETKukXYSHiSMTQojrK8DJj+faPWScMuvPI5vYkbmXpvbe9AusmjJrW8aeOqfaEuJm9UT0/YS7BbEydQO/JS7GYDCAAvwdfHgqZhRx9VhJ61L1Slp/fKsXJ/MKSTiUS0JqHnPXHEKhgBbNnIhs7kK/9k0b3LC4vXRp7U24X9WA61PnDLw+eQuLNx7h7u6BJo5MCCGuP0dLe4ZHDOaO0L5sOr6D5f9MmfXj7jnoKsupNFTdisorPsOU+NkAkriKm14Hn2g6+ERTqa+koKwQazNr1ErVFddXr6QVwMPZGg9na/q2b0ru2RIWbjjMyq3H2JZ4SpJW0SAtmjkTE+rO/L9S6RXri521malDEkKIRlE1ZVZHuvt3IDH7AB9t+taYsF6gq9Tx675FkrSKW4ZKqcLe4urv0NcraS0pqyDxSB57D+Wy91AumTkF+LjbMrhzM2JC3a86CHH7GdU/hGc/XcfctYd4dHDDxrQIIcTNTqFQEOEeQoW+vNbyvOIzjRyREDe+eiWt9729HJVKSQt/J/p3aEpMqDsu8uS3uAo+7rb0iPFl+ZY0Bnb0x93JytQhCSFEo3OydKwzQf0tcTGDgnpiqZW/t0JAPae8eu2BaGaP68vYR9vRv0NTSVjFNXFf7yCUSiUzV9Q+JYYQQtzqhkcMRquqvsKkRqkhwNGPBckreGbZ2yw+sBpdhc5EEQpx46hXT2tsiyZsTTzJ/L9SOZ5VgFqpwNvNhqFdAmgvT4CLK+RkZ8GQzs2Yu+YQQzsHEOBtb+qQhBCiUV0Yt1rb7AFHz5zgt8RFzEr4g+WH1nFXWD+6NG1/VQ+yCGEq50rzyczPormjHyUVpdiZ2za4jnolrZsTMvlk5k46RHrSrY0Per2B/Wmn+XjmTl4Z2UYSV3HF7uwawMqtx/hp6X7ef6K9zFsohLjtxPnGEGrdDAAnJyfjdn9HH97o/CzJOYf4Zd8ivt/5C0sOrGFY+ADaebdGqZAlYsWNr7yynKm7fmN92lYUCgVf9RvLjL2/U1JeyssdHm/Q8Jd6XfFz/jzE/X1DeGVkG/p3aMrAOH9eeyCa+/uGMHftoSs+ESEszTXc2zOIfYfz2H0wx9ThCCHEDSfUNZD3ur/MKx2fRKPS8NXWH3lt9QT2nEqqmvtSiBvYvP3LOHLmOOO6v4RGpQFgcHAvThefZVbCggbVVa+k9WRuIR0ia/amto9oQnp2YYMaFOJSfdr50cTJip+XJlOplw9gIYS4lEKhoI1nBB/3eoNnYx+kuLyECRu/Yey6zzmQe8TU4QlRp63pu3mw1TCCnJtx4V5qoLM/j0ePYOfJfQ2qq15Jq7O9BcdP5dfYnnYyH1srbS1HCFF/GrWSkf1COHYqn3U7000djhBC3LCUSiVxfjF82XcsD7e6l1MFObzz16d8uGkyx89lmDo8IWo4W3IOF0vHGtvtzW0pLi9pUF31GtPaK9aXyfP3UVBcTohfVcPJaaeZteIAvdv51ruxBetSmbkiBbXq31x57KPt8HW34as5e9h3OA9Lcw3DewXRK7aqXoPBwIzlKazefpxKvYFubbx5eFALVMqqfH3D7gxmrkjhfGEZLZo589w9UTjYmANwJOMc38xP4ER2AR7OVjx1VyTBvjXfOGF6HSM9+GO9PbNXphDX0hMzjTxoIIQQdVGr1PRu3pnOTduyMnU9i1JW8cqq8XTwacOw8IG4W7uYOkQhAGhq782OzAQGBHUHQPFPf+uaI5vxtfdqUF31SlqHdAkg73wJ3/6egF5vwACoVUqGdG7G8F7B9W7saGY+I/uGckfXgGrbJ0zfgbmZmplj+3DsVD5jf9hKc297mnrYsWxLGjtTspn4clcUwLhp26rm9ozzJ+3keSb/nsC4x9rh52HHlAX7+Pb3fbwxOgZdeSXv/bidYT0C6RXry7qd6Uz4eQdT3+yFRi2D1280CoWCBweG8cbkLSzZdJS7ujU3dUhCCHHDM1ebMSSkNz2adWTxgT9Zfugvtqbvort/R+4M64fDNViFSIircV/kEMZvmETq6TQq9ZUsTFlFZkEWqaeP8VrcUw2qq17Zm0qp4PGhEcwe15dPnuvE1y915bf3+/FAv1Bjj2d9HD15Dn/P6lMclJRVsC0pixG9g9FqVAT6ONC5pRertx0HYN2udAbF+eNoa46DrTl3dQ9k1bZjQFUva2yYO0G+jphpVIweEMb2/VmcKyhj3+E8FAoF/do3Ra1S0jPWFxtLLTuSs+odr2hc4c2ciQ51Y/7aQ+QXyZyEQghRX9ZaK+6LGMLE/u/RvVlH1h7dzLPL3mZ2wh8U6opMHZ64jYW4NOe97i+jVqpwt3bhyJnjOFs68mHP14hwD2lQXZftaS0uLUelUmKmUWFpriHQx8FYdja/lO/+2Mfroy6/PnKproLM3CIWbzrKZ7/sxtpCwx1dAvD3tEOtVFRbEcnT1ZqtiacAyMgpxNvdxljm5WJNenYBBoOBjJxCgnz/jcfWSouVuYaMnAIycgrxcfv3uAv1nsgqoEPEZcMVJjKqfyjPfbqOuWsO8cjgFqYORwghbioOFnY80no4A4J6MDdpKYsP/MmaI5sYFNyLvoFdMVebmTpEcRvyc/Dm2bYPXnU9dSat5wvL+PK3Pew+kA0KBe1aNGHM8JaYa6sOWbXtGD8vTaa8Ul+vhs4VlBHi50i/9k2JbO7CoRNneW/aNoZ0CUB7yfhFM42KMl0lAKW6ymrjG820KvQGKK/QU6qrwEx7ybFaFWXllZTVVqZRUaarqBFbfn4++fnVHzTLzc2t13mJa8vX3Zbu0T4s25LGgI5NZXlXIYS4Au7WLjzX9kEGB/fkt8TF/Jq4iOWp67gztC89/DuiVtVrdKAQV21K/Oz/LH88ekS966rzqp26OImDx89wX+9g1ColSzYfZeaKFO7vE8LHM3ey60A2LfydeebuyHo15O5kxYdPdzS+DvN3omtrb/YfOY2uvLLavmXllZibVSWcZhpVtfIyXSUqpQKtRoWZRo2uvHrSXKarxFyrNiavl9ZrYVbzlKdPn86kSZPqdR7i+hvRJ5gNezKZteIAL9/f2tThCCHETcvX3otX457iYN4Rftm3iB93z2HpwTUMazEQDAZ+2bewxkpcQlxLWYXV52Cv1FeSXZhHcXkJHXzaNKiuOpPWhEO5PHlnJHFRngCENXNi/E87OJlbREraaZ6+K4rebes/c8DhjHPsOZjD3d0Djdt0FXpcHCyoOGIg52wxrg6WAGTmFOL9z619bzdrMnMLCfrnqf+M3EvKcv6dJ/Z8YRmFJTq83WwoKatg6ea0ajFk5hTSuWXNJ9VGjRrF0KFDq21LTExkzJgx9T4/ce042VkwuJM/89amMqRLMwK87E0dkhBC3NSCnJsxtusLJGQl8+u+RUza/jMKFFQ9Wg15xWeMPWKSuIpr6X9dX6ixzWAwMG3Xbw1aDQv+40Gs/CJdtfGiwb6OFBTrOJVXyJcvdmlQwgpgYabmt9UH2ZJwEr3eQMKhXDbtzaBfh6bEhrkzY1kKpboKDp04y4Y9GXRpVZVcdmnlzYJ1h8k7V8LZglLmrz1E19ZVZZ1aevF34kn2H63qrZ2xPIXWwW7YWmmJbO5MeYWeJZuOUlGp58/txzlXWEarINcasdna2uLl5VXtx8VFpgsxpTu7NsfGUsvPS/fLii9CCHENKBQKopqEMaHXa9horYwJ6wW6Sh2/7ltkoujE7UShUDAgqDt/Hd3SoOPq7GnVGwxo1dXHhGrUSh4bGnFF4ww9Xax59YFoZq5I4YvfduNsZ87z97QiwMueZ4dF8c38BB4ctxpzMzUPDggz9qz269CUc4VlvPTVBsor9HRp7c3gzlVTZvl72vHM3VF8PWcPZwvKCPN34vl7Wv4Tq4qxj7Zl8vwEZq5IpomTNW89FIt5LcMDxI3HykLDvb0C+WFhEnsO5tIquOaXDSGEEA2nVCjrnFHgdPGZRo5G3K6yi/Io19d8zui/NDiD83C+8gdjYsLciQlzr7HdxlLLaw9E13qMSqlgZN8QRvatfVqEuChP4xCGSzX1sOOT5zpdcbzCtPq2a8qSTUf5ael+IgNdGjS9mhBCiLo5WTqSV0uCaq42p1JfiUopC7yIa6O2B7FKykvYm5VMtEf9nou6oM7hAQpAITmCMCGNWskDfUM5diqf9btkeVchhLhWhkcMRquqvgy7UqGkpKKUd9d9wenisyaKTNxqsgpzavwUlRfTL7AbD7e+t0F11dnTagBe/nojyosy1zJdJW9+9zfqS3q8przeo2FnIEQ9dYj0IGCDPbNWHqBjlCzvKoQQ18KFh61m711QbfYABQq+3zmbV1Z9wNOxo2nlIfNli6tT24NYV6rOpHV4z6Br1ogQV0qpVPDggFDe/PZvlm46yp2yvKsQQlwTcb4xhFo3A8DJycm43d/Rhy/+nsqHm75hUHBP7g0fjFqGC4grlFdU/3HSzlaO/1led9LaO7j+EQlxHUUEuNAmxI15aw/RM9YXWyvt5Q8SQghxRTxs3PigxyvM2DOfxQf+5EDuEca0e/iyCYUQtXl66VtVY07/iwFQwJxhk/9zN3mUXtwURvcP5bnP1jFv7SEeHiS3q4QQ4nrSqjQ80mY4oa6BTImfxf+t/oCnY0bRxlPWQRcN80L7R/hpz1zuDO1LqEsgaqWKw2eOM3vfH3Rt2p4It9oftK+NJK3ipuDbxJZubXxYujmNAR39cXO0NHVIQghxy2vv0xp/B2++2DqVjzd/y4DA7twXMUSWgRX1Nm//Mh5udS8xXlHGbe42rtiaWfPj7jkMazGg3nXVOXuAEDeaEX2CUSpg1ooUU4cihBC3DXcbV97v/n/0CejC0kNreeevz8gpOm3qsMRNIqcwD0/bmtOd2pvbNniWijqT1p+W7KewWFfV4NliWZVImJyzvQWDOjVj/e4MjmScM3U4Qghx29CoNDzU+h5ebP8omQVZvLrqA3Zk7DV1WOImEODkx9ykpZRV6IzbinUl/JK4iFDXwAbVVWfSunTzUYpKq1YqePSDP8kv0tW1qxCN5q5uF5Z3TTZ1KEIIcdtp692Kj3u9gbu1K59umcLPu+dSUdmwVY3E7WV0y7vZn3OQxxe/xpt/fsQbf37Ek0ve4FRB9rWbp9XNyZLxP+2gqactBuD7PxLR1jFH5vP3tmxQo0JcKSsLDff2DOSHRUnsPphDqyBZ3lUIIRqTm7UL47q/xOyEP1ieuo6DeUcZ0/5h3KxdTB2auAH52nvxZb+xbDm+k4z8U5iptfRp3oX2Pm0aPJVanUnrS/e1Zt5fqZw+X4oCOJ1fikYlQ2CF6fVt78fiTUf5eel+IpvL8q5CCNHYNCoNo1sNI9Q1kG93zOCV1eN5Mnokbb1bmTo0cQOy1lrRu3nnq66nzqS1mZc9rz0QDcDDH/zJG6NjZH5McUPQqFU80C+ET2btYsPudLq18TF1SEIIcVuK8YrCz8GbL/+eyud//0DvgM6MjLoTrUpj6tCECT2/7H980OMVrM2seG7ZOyj+Y6LWr/q/W+966zVnxbQ3e2IwGNiZks2JrHxUKiU+bjZESC+XMJGOkZ78sf4wM1ccoGOkZ51DV4QQQlxfrlZOjOv2Er/sW8jSQ2s5lHeUF9o/gruNDN+6XXX0jUbzzxeXC0sGXwv1SloLinW8PeVvjmaex9ZKi15voLCkHH9PO957vD02ltIDKxqXUqngwYFhVcu7bj7KHV1leVchhDAVtUrNAy3vItQ1kG92TOfV1RN4PPp+2vu0NnVowgTuvmju1bsbMA/r5dQraZ26KAm93sA3/9cNbzcbAE5k5fPZL7v5eWkyzw6LumYBCVFfEQEutA52Ze7aVHrG+sqXJyGEMLE2nhF80utNvtg6lS+3TmV/zkFGtbxbhgvc5g7mHeFQXhoV+goM/DuFqgIFQ0P71Lueej1ZFZ+cxRN3RBgTVgAfd1seGxLO9v2nGhC2ENfW6AFhFJeWM3fNIVOHIoQQAnC2cuTdbi8xKLgnfx7ZxJtrPuZkQbapwxImMn//ct5Z+xkLUlaw5uhm1h7dctHP5gbVVa+eVoMBrC1qfkuysdRQqqtsUINCXEt+TWzp1sZblncVQogbiFqp4v7IOwh1ac4326fz2uoJPNbmPjpew/GN4uawIW0rd4T25Z7wgVddV72S1mA/R35fd5jn7mlpfPCqUm9g/l+pBPk4XHUQQlyNEb1D2LQnk1krU3jpPhk/JYQQN4pWHuF83PtNvtw6ja+3/cT+nFQCnZoyJ3ExZ0rP42zpyPCIwdf0YR1RP6eLz/LDzl9IyT2MhcacQcE96RfYjUJdEd/umElSzkEsNRbcHdafbv4dADAYDPyauIi1R7eg11fSya8to6LuQqms+8b9udJ8OvvFXpOY65W0ju4fyquTNvHY0dMEetsDcCj9HMUl5bz3ePtrEogQV8rFwYKBcf4sWH+YIZ2a0czL3tQhCSGE+IeTpQNju77AnKQlLExZxV9HtxjHNeYVn2FK/Gzg2j5lLv6bwWDgk83fEeYayMsdn+BUQTbv/PUZzRx9WXpwLeZqM34Y/DEnzmUyfuMkmjn64mvvxarDG9h9MolPe7+FAvhw02RWHd5A38CudbbVwi2I5NzUazKbRL3GtPo2seXrl7rSMcKDUl0llXoDXVt78e2r3Qn4J4kVwpTu6h6ItYWGn5fJ8q5CCHGjUSlV3BcxBFszm2oP4gDoKnX8um+RiSK7PaWeTuNsyXlGRAxFrVThbefBB93/D0cLe+IzExjWYgBalYYAJz86+kSz9sgWADYe206/wG44WNhhb2HHkJDerLnMuNQw10B+2jOPr7ZOY27SEhYkr6j20xD16mkFcHW05MGBYQ2qXIjGYm2hYViPIKYtluVdhRDiRlVQVlDr9tPFZxo5kltXbm4uGRkZ1bbZ2tpia2trfJ12Nh0vuybMSljAphPxWKrNGRraB197L1RKVbUleT1s3diesQeAzIIsvGzd/y2zcSMj/xQGgwGFovZ5+1ekrsfWzJpDp9M4dDqtWpkCuCO0b73Prd5JqxA3uv4d/FiyuWp516jmLihl4QshhLihOFk6kldLgmpjZm2CaG5NY8aMqbHtmWee4dlnnzW+LtQVsT/nEC1cg/h2wAccOXuc8Rsm8Vqnp2pMT6ZVadFV6AAoq9Bhpv53ekkztRaDwUC5vqLOac2+GfD+NTirKpK0iluGRq1iZN8QPpu9i/W7M+jWxtvUIQkhhLjI8IjBTImfja5SZ9ymQEF+WSFT4mczKupOzDXmJozw5vfll18SHh5ebdvFvawAGpUaa62lcY7UIOdmxHq1ZG7SUnSV5dX21VXqMNeYAWCm0lYrL6vQoVIoG20eXklaxS2lU5QnCzccZtbKFDpGesjyrkIIcQO58LDV7L0LjLMHDGvRn4z8bJYc+JOk7AM8HTuaYJdmJo705uXi4oKXl9d/7uNh40ZZhY5KfSUqZdXfSb1BT1N7b1JyD5NXdAZnK0cATuZn42nbBABPW3dO5mfT3KlpVVnBv2UXe27ZOyioebdTpVRhrbWkmaMf/QK74mLl1KBzq1fS+sWvuxnWIxBPF+m+Fzc2pVLBg/3DeGvK3yzdnMYdXQNMHZIQQoiLxPnGEGpdlZQ6Of2btLT2aMGk7dP537rPGBzci2FhA1CrpG/teohwC8FKa8nsfQsZETGEw2eOsSNzL291fo6c4tP8sm8hj0ffT/r5k2w+Ec/rnZ4Gqn53iw/+SQu3IFRKFQtTVtHJr+asD3XNBGHAQKGumOTcQ6xP+5v3evwfXrUkvXWp19WwPekU9/UOrnelQphSZKALrYJdmbv2ED1jfUwdjhBCiHoIcWnOp73fYvqeeSxMWcXeU/t5JnY0Pvaepg7tlqNVaxnb9QWm7Z7DI4tewVJtzoMthxHo7M8TNvfz/a5feGLJ65irzbg/8g5jz2rvgM6cL8vnjT8/olxfQZxvDAMCe9So/+4WAy4bwzfbpzMncQkvdXis3nHXK2nt3MqL2StTuLdXEG4OlqhU9ZopSwiTGd0/lOc/X8+8takMbu9++QOEEEKYnIXGnCdiRtLGM4Lv4mfx2p8fcl/EYPoFdkOpkNzjWnK3ceXNzs/W2G5tZsWL7R+t9RilUsm94YO5N3zwVbffO6AzH23+tkHH1CtpTUjN42ReIRt2Z4BCwaUPZf/x8aB6N3i2oJRnP13Hc/e0JCbUncJiHV/N2cO+w3lYmmsY3iuIXrG+QNXktzOWp7B6+3Eq9Qa6tfHm4UEtjKtybdidwcwVKZwvLKNFM2eeuycKB5uqAdxHMs7xzfwETmQX4OFsxVN3RRLs61jvOMXNramHHV1be7Now2H+3p3K+cIybOwceKBvCF1aywNaQghxI2vjGclnTk2ZsvMXZuz9nV0nE3kq5oEGj4EUNy57c1tKyksadEy9kta7uze/ooBq8/WcvRQU/fvU4MR5ezE3UzNzbB+Oncpn7A9bae5tT1MPO5ZtSWNnSjYTX+6KAhg3bRvLt6QxMM6ftJPnmfx7AuMea4efhx1TFuzj29/38cboGHTllbz343aG9QikV6wv63amM+HnHUx9sxcatXxTu134e9nx1850zhWWAZB7toRJ8xIAJHEVQogbnJ25Lf/X4XHWp23l5z3zeHnV+zzYchid/drWOSeouHlkFmThYG7XoGPqlcF1j/b5z5/6WvF3GuZaFc72FgCUlFWwLSmLEb2D0WpUBPo40LmlF6u3HQdg3a50BsX542hrjoOtOXd1D2TVtmNAVS9rbJg7Qb6OmGlUjB4Qxvb9WZwrKGPf4TwUCgX92jdFrVLSM9YXG0stO5KzGvTmiJvbog1HamwrK69kxooUE0QjhBCioRQKBV392/NJ7zfxs/di8o4ZfLble/JLa1+kQNwcThZkM3PvAqI9Ixt0XL0fy0s6kse8v1LJyClkwpMdWBN/Ancnq3rPhXkyt5A/Nhzhs+c7Mebz9cZtaqUCdycr436ertZsTTwFQEZOId7uNsYyLxdr0rMLMBgMZOQUEuTrYCyztdJiZa4hI6eAjJxCfNz+Pe5CvSeyCugQUTO2/Px88vPzq23Lzc2t13mJG1fe2dpvO9S1XQghxI3J1dqZ/3V5gaWH1vBb4hJeWvU+T0TfT2uP8MsfLBpdXVNe6TFQXF5Coa6IUJfm3BNe/+GlUM+kdWdKNhOmx9O1tRdJh/PQGwwogK/n7EGv19Mjxvc/j6+s1PPZL7t4bEg4Npb/rqRQqqusMY+mmUZFma7SWG52UbmZVoXeAOUVekp1FZhpLzlWq6KsvJKy2so0Ksp0FbXGN336dCZNmnTZ90HcXJwdLMitJUF1drAwQTRCCCGuhlKpZFBwLyLdQ5m07Wc+2jSZ7v4dZUGCG1BdU15dmKfV38GXACe/Btdbr6T119UHeGhgGP07NGXjnqr1bIf3DsbKUsOC9Ucum7T+9uchmnrY0SbErdp2M60KXXlltW1l5ZWYm1UlnGaa6uVlukpUSgVajQozjRpdub76sbpKzLVqY/J6ab0WZrWf7qhRoxg6dGi1bYmJibUuhSZuHg/0Dakaw3rRpaBSKnigb4jpghJCCHFVfO29GN/zVeYmLWWxLEhwQ6rPlFdXol5J64msAloHu9bYHhPqzvSlyZc9ftPeTM4WlLJ5byYAxWUVfDJzJ3d2a06F3kDO2WJcHSwByMwpxPufW/vebtZk5hYS9M9T/xm5l5TlFBrbOF9YRmGJDm83G0rKKli6Oa1aDJk5hXRuWfsKEba2tjWWOMvKkvGvN7sLD1v9snw3Z/NLq77M1NK7L4QQ4uaiUWkYETmU1h7hTNr+syxIcJuo12/W1kpL1umiamNPAQ5nnMPexuyyx3/3Wvdqrx9+fzWP3xFBTKg7RzPPM2NZCs8Mi+REVgEb9mTwv0faAtCllTcL1h0mIsAFlUrB/LWH6Nq6KvHs1NKL1ydvpkeMD8297ZmxPIXWwW7YWmmJbO5MeYWeJZuO0re9H+v+eYK8VVDNxFvc2rq09ibcr+oLkY2tPW9M3sLnv+6mibMVTT0a9tSiEEKIG0uwSwCfyIIEt416zR7Qu60f3y3Yx86UbAwGOJVXxJ/bjzNlQWKDZg+ozbPDoqjQ63lw3GomTI/nwQFhxp7Vfh2aEtuiCS99tYGnP/6LkKZODO5ctSynv6cdz9wdxddz9nD//1ZyJr+U5+9pCYBGrWLso23ZuCeD+95eztLNabz1UCzmdQwPELcHrUbFGw/GYGWu4f0ft3P+n6mwhBBC3LwuLEjwSscnOFNyjtf+/JAlB9agN+gvf7C4qdR7ntbi0nIm/LyD8ko9//thKyqlkiGdm3Fvz6AGNzrtrV7G/7ex1PLaA9G17qdSKhjZN4SRdYxBjIvyJC6q9m9TTT3s+OS5Tg2OTdzaHG3NefPBGF7/ZjMTpsfz3uPtZe5eIYS4BVy8IMHMhN/ZdXIfT8eOkgUJTGBWwgKGhPTGWmtFXtEZnCwdrsncuvX6a61QKBg9IIzZ4/ry2fOd+OrFLvz2QT9G9Q9FeenyWELc4AJ9HHj2npbsP3qaHxYmmjocIYQQ18iFBQmejB5J2tl0Xl75PuvTtmIwGEwd2m1lRep6istLAXh62VsUlBVe5oj6qff98rLySjbtzeR4VgEatRJfdxviojxRqaSXStx8urTy4tjJ8/y+7jB+Hrb0a9/U1CEJIYS4Bi4sSBDmFsQ326czeccMdmbuI8ItmAXJKzhTeh5nS0eGRwyuc2omcXVcrZz4dPN3+Nl7gwF+3DMXrUpT675PxTxQ73rrlbRm5hby+jebKSuvxMvVGr0Blm05ym9/HmTso+1qPKAlxM1gZL9QjmcV8P0fiXi72hAe4GzqkIQQQlwjrlZO/K/LGJYeWsvshD/YkbnXWJZXfIYp8bOBuucUFVfuubYP8UfKSs6UnAMFnC05j1p59TP31CtpnTh3L6FNnXj+3pbGuU4LinV8OnsX3y7Yx7uPtrvqQIRobCqlgv+7vzUvf72RCdPj+XxMJ/kCJoQQt5CqBQl6svTgGs6VVl/5Ulep49d9iyRpvQ6aOnjzYvtHAXh66Vu83OExbMysr7reet3bTz1xlhF9gqtNzm9jqeXBAWEkHTl91UEIYSqW5hreejAWvcHA+z9up7i03NQhCSGEuMbOX5KwXnC6+EwjR3L7+WbA+1hrrdhzKonFB1az7OBa9mWloNc3fHaHevW0ujlZcSqvyDix/wWnz5fgbCdLp4mbm4eLNa+ObMPYH7byxa+7eX1UjDxgKIQQtxAnS0fyaklQnSwdTRDN7aWwrIj3NnzFsbMZ2JhZoTcYKNIV4+fgxdudn8farP53OOvsaU1JO2P86drai6/n7mHVtmMcP5VPenYB63dn8PWcvQzpEnBNTkoIU2oZ5MrDg1qwLSmLX1YfMHU4QgghrqHhEYPRqrTVtmlVGoZHDDZRRLePn/fOQ28w8Hnfd5g65BN+HPopn/V5Gwwwa98fDaqrzp7WV7/ZhAK4eJKIb+Yn1Njvu98T6NvOr0GNCnEjGhjnz7FT+cz58xB+TWzpGCkrqgghxK3gwrjV2XsXcKb0PAARbiEynrUR7D6ZxCsdn8TT1t24zcuuCQ+2GsanW6ZA9P31rqvOpHXqGz2vLkohbjIKhYIn74wgI6eQL37dQxMnK5p52Zs6LCGEENdAnG8ModbNAJhx8A/2nEoiv7QAW3ObyxwproYBA9ZayxrbrbVWlFXoGlRXncMDXB0t6/0jxK1Co1bx+qhobC01vP/TDs4WlJo6JCGEENfYsBYDKKvUsfDAalOHcssLdPJn4YFV1R680uv1LExZRYCTX4PqqteDWFmni5i5PIXjWfmUV9R82mvK6z0a1KgQNzIHW3PefCiWVydtZsLP8XzwZAdZ6lUIIW4hnrbudPKNZdXhDQwI6o6jhb2pQ7pljYgYwjt/fcazue8Q4OgHwOEzxyguL+Htzs81qK56Ja1f/Lqb0+dL6RjpgZnm6ieHFeJGF+Blz5h7WvLxrJ18t2Afz9wdeU3WTRZCCHFjuDusP5uP72DB/hU80ma4qcO5ZfnYe/JJ7zdZdXgD6edPoVGp6eQbS5/mnbEzt21QXfVKWo9knufDpzsSIOP7xG0krqUnaafOM29tKk09bBnQ0d/UIQkhhLhGXK2d6ebfgbVpWxgU3BNXa1kV8XpxsXLi/sg7rrqeet3z9HC2okxXedWNCXGzub9PCDGh7vywKImEQ7mmDkcIIcQ1dGdoP5QomL9/ualDEfVQr6T18aERTPljH3/tPMH+o6erzeGakiarSYhbl1Kp4KURrfByteajmfGcyisydUhCCCGuEUdLe3oFdGbD8W2czM8ydTjiMuo1PCAjp5D07EK+/G1PjTIFsOhTmZxX3LouLPX60lcbeO/H7Xz6XByW5hpThyWEEOIaGBLSizVHNzM3aSlj2j9i6nDEf6hX0vrLqgP0jPVhQIemmGvrdYgQt5Qmzla8OjKad37Yymezd/Pmg7LUqxBC3ArszG3p17wrf6SsZOi5Pvjae5k6pFvKN9unMzS0Dx42blddV72GBxSXlnNn1+b4uNvKPK3ithUZ6MKjg1uwIzmLWStTTB2OEEKIa2RgcA8sNRbMSVxi6lBuOfGZCaiV16bDs15Ja/sID7YlnbomDQpxM+vfoSm92/oyb20qm/ZkmjocIYQQ14C11opBwT3ZeXIfqafTTB3OLaWjbzRzE5eQVZBDpf7qHuqvV+rr5mjJjOUp/L3vJE2crVCrque6z9wddVVBCHGzUCgUPD40gvTsAr6cs4cmzlYEeNubOiwhhBBXqW/zriw79BdzEpfwVpeGTXov6paUfZBTBTlsOrEDBYoac57/eveketdVr6Q16chpgnwcAMg5U9KAUIW49WjUSl4fFcMLX27gg5+28/mYzjjYmps6LCGEEFfBQmPO0JDezNj7O8k5hwh1DTR1SLeEISG9r1ld9Upaxz/V4Zo1KMStwN7GjLcfiuWVSZsY//MOxj/VAY1aVosTQoibWa9mnVhycA2/JS7m3W4vyUqI10CXpu2uWV31GtN66bysMk+rEODvaceYe1ty4PhZJs/fh8FgMHVIQgghroJWreXO0L4cyDtCQlayqcO5ZSTnpDJ+w0SeXvoWuUWnmZu0lI3Htje4nnr1tL76zSYUwMV/khUACgVKBfzx8aAGNyzEraBjpCfHeuYz589DNPWwZVCnZqYOSQghxFXo1rQDiw78yW+Ji4l0D5Xe1qu051QSn235njjfWPbnpqI36FEA3+6YQaW+kq7+7etdV72S1qlv9Kz2ulJvIDO3kFkrUxjdP7RBwQtxq7mvVzDHT+UzbXESXm42tApyNXVIQgghrpBapebusP5M3jGD+MwEYryiTB3STW1e0jJGRt5J7+ad2XIiHoC7WwzASmvJkoNrGpS01mt4wKXzsjZxtqJNiBuPDg5n6qKkKzsLIW4RSqWCF+9rjY+7LR/P3MnJ3EJThySEEOIqxPnG4GHjxpzExej1elOHc1NLzz9FVJOaHZxtPCLILsprUF31SlrrYm9jJmuxCwFYmKmrVslSKHj/p+0UlZSbOiQhhBBXSKVUMazFANLzT/F3+k5Th3NTs9VakV1YMzk9cvY49mY2DaqrXsMDanvYqrisnEUbjuDTxLbejW3am8kvqw6Qd64EFwdLRvYNoV14EwqLdXw1Zw/7Dudhaa5heK8gesX6AmAwGJixPIXV249TqTfQrY03Dw9qgeqfJTQ37M5g5ooUzheW0aKZM8/dE4WDTdX0Q0cyzvHN/AROZBfg4WzFU3dFEuzrWO94hWgIdycrXh8VzdtT/ub1yZspLcrnbEEpNnYOPNA3hC6tvU0dohBCiHpq690K3+SVzE1aSlvv1qiVMkPMlejerCPTdv/G6JZ3YwCyC/PYn5PKr/sW0jOgU4PquuIHsQBcHSx56b7W9WooM7eQr+bs4b3H2hPS1JG9h3J4d+p2fn6nF5N/T8DcTM3MsX04diqfsT9spbm3PU097Fi2JY2dKdlMfLkrCmDctG0s35LGwDh/0k6eZ/LvCYx7rB1+HnZMWbCPb3/fxxujY9CVV/Lej9sZ1iOQXrG+rNuZzoSfdzD1zV5o1FfVwSxEncIDnOnS2ou18emYKUsByD1bwqR5CQCSuAohxE1CqVByT/hAPt78HRuPbaObv0z/eSWGhvShuLyUT7d8T0VlBR9smIhKqWJAUHfuCuvXoLqu6EEsALVaiWMDJlT3dLFm5tg+WJipKSmr4Ex+GRZmajRqJduSspjyWne0GhWBPg50bunF6m3HefyOCNbtSmdQnL+xrbu6B/LrqgMMjPNnw+4MYsPcCfqn93T0gDBGjl3JuYIyDmecQ6FQ0K99UwB6xvqyaOMRdiRn0SHCo95xC9FQ+w7XvA1SVl7JjBUpkrQKIcRNpLVHBAGOfszfv5w43xg0Ko2pQ7rpKBQK7o8cyl1h/cjMz0KtVNHE2hWtWtvguuqVtLo6Wja44tpYmKnJOl3E4xPWYACevDOSU3lFqJUK3J2sjPt5ulqzNfEUABk5hXi7/zvmwcvFmvTsAgwGAxk5hQT5OhjLbK20WJlryMgpICOnEB+36mMlPF2tOZFVQIeI6nHl5+eTn59fbVtubu41OWdx+8k7W/uqcXVtF0IIcWNSKBTcGz6I9zd8zZojm+kb2NXUId2UdBU6/j6xi/TzJ9Go1HjbetDepzWqBg65qDNpnTRvb70reebuqHrv62Jvwe8fDWT/0dO8/+N27ujaHK2metBmGhVlukoASnWVmF1UbqZVoTdAeYWeUl0FZtpLjtWqKCuvpKy2Mo2KMl1FjZimT5/OpEn1X/tWiP/i7GBBbi0JqpOdLPUqhBA3m3C3YEJdmrMgZSXd/DtgdgU9hLezkwXZjP3rc8oqdXjauGMwGFiZup75yct4o9MzuFm71LuuOpPWk7n/PStA1pki8s6VoFIqG5S0qlRV40kjm7vQPsKDw+nn0JVXVtunrLwSc7OqhNNMo6pWXqarRKVUoNWoMNOo0ZVXn4qiTFeJuVZtTF4vrdfCrOYpjxo1iqFDh1bblpiYyJgxY+p9XkJc8EDfkKoxrNUvP8or9KRnF+Dt1rCnJYUQQpjOhd7Wd/76jJWp6xkc0svUId1UpsTPItg5gKdiRmKuqeq8KSwr4uttPzJt12+80fnZetdVZ9I6/qnaBxxXVuqZu+YQKcfO0NTDjufvaVmvhnamZLNww2Hef+Lfeisq9Lg7W1JxwEDO2WJcHaqGIWTmFBr/sHu7WZOZW2gct5qRe0lZzr9zYp4vLKOwRIe3mw0lZRUs3ZxWLYbMnEI6t/SqEZutrS22ttVnQcjKyqrXeQlxqQvjVn9Zvpuz+aW4OFjQI9qHFX8f4/++3sgrD0TLAgRCCHETCXYJIMo9lEUHVtMzIA5LjYWpQ7ppHD5znI96vW5MWAGszay4P/IO3ljzUYPqqteY1guOZJzjqzl7yMwp5N6egdzVPdA49dTlNPO043D6Of7amU6XVl7sPpjDzgPZfPpcJ3LPljBjWQrPDIvkRFYBG/Zk8L9H2gLQpZU3C9YdJiLABZVKwfy1h+jauirx7NTSi9cnb6ZHjA/Nve2ZsTyF1sFu2FppiWzuTHmFniWbjtK3vR/rdqZzrrBMkgXRKLq09ibcr+pLmJOTEwA9on1478ftvPvDVh4ZHM6Ajk1leUAhhLhJ3Bs+iNf+/JBlB9dyd4sBpg7nqi0+sJpfExejVv6bCr7R6Wm87Tz4dsdMknIOYqmx4O6w/saZEwwGA78mLmLt0S3o9ZV08mvLqKi7UCrrnpXJzcqZrIJcvGybVNt+puQcThYOdRxVu3olreUVen5dfYAF6w4T4GXPFy90xse9/vOzAjjYmvP2w22ZuiiR7xbsw9PFijcfjMHbzYZnh0XxzfwEHhy3GnMzNQ8OCDP2rPbr0JRzhWW89NUGyiv0dGntzeDOAQD4e9rxzN1RfD1nD2cLygjzdzL2/GrUKsY+2pbJ8xOYuSKZJk7WvPVQLOa1DA8QojG4Olry8bNxfDZ7F98vTCQ9u4DHhoajVskUbEIIcaPzd/QlxjOKpYfW0qd5F2zMrE0d0lU5djaD4eGDGRRcfYaoz7Z8j7najB8Gf8yJc5mM3ziJZo6++Np7serwBnafTOLT3m+hAD7cNJlVhzfUeEDtYN4R4//H+cbwXfxM7i0dTKBTU5RKJcfOpjMzYQF3hfZvUMyXzeAOHD/D13P2knO2mAf6hTKkczOU9exdvVSYvxNfvNClxnYbSy2vPRBd6zEqpYKRfUMY2Tek1vK4KE/iojxrLfv/9u47PMoqYRv4PTOZmZTJJJNkSO+QEEloAUIL0qQjUhUXsay66guuCvKJuoqi66q7qwKi4sIqLCiKinQMiIQaqpBAIJSE9EbqpEw93x8TBoYEDQhJGO7fdeWCOed5njlnQrk5nBIe4IH3n72+jWuJbiUXpRNefqQXlm86ie92nEVeiQ4vPdwT7q6c2E9E1NZNiR2Dg1uPYd2pJPypy/jfv6ENy6zIwcDwPnZl9cZ6HMw7ho9GzYNCJkd77zD0D+mJ7ef24LH4+5GclYJRUYOhcfEAANwXMxzfntjYKLS+tv1fuHqD/88PrWrUhv8c+Qr3tE9sdpuvGVoNRjOWb0rHht3n0THMC397LAH+Pm7XupyImkkqleCRMZ0Q4ueOhd8cw6yPkvHanxMQ1I4LtIiI2rIQz0D0C+mBzWd2YHTUYHg2hLfbjd5kQEF1MTad2YGFKV9AJXfF2I73IFwTDJlUZreiP0Dti5TcowCAvOpCBKn9Lte5+yK3qgBCCLvpbovGzL8l7b5maJ35zx0ovFgDX283dIvWYtevedd8yJShUbekcUSObHCPEPh7q/D2FymY/VEy/t/0nujGOddERG3alNgx2JtzGD+kb8Wj3ae0dnMaKSkpQW5url3Z1QvOK+urEO0TiWGRA9C5b0ecKcvEu7s+wZjooVBcdYCCQqaAwWQAYA27V275pXRSQAgBo8Vkd5/WzftWdO3aodVktsBH4wqzReCnlOxrPkAChlaiGxUT7oV///VuzF+Wgnn/2Y8nx8VidP+I1m4WERFdg597OwwM74Okc7swNnoofNy8WrtJdprarnPGjBmYOfPy1lLtVD54Y/ALttcx2g4YEJqA9JIzMJiNdvcazAY4y5UAAKVMYVevNxkgk0gbBd0rFetKsSr1R+RU5sNkbrxX/kej32h2364ZWpe+yn3IiFpCOy9XvDujP/618gg+/SEV2UXVeOI+LtAiImqrJt01CslZKfju5Gb8peefWrs5dj788EPExcXZlV29ref5smwcL0rHfTHDbWUGixE+rl4wWzJQWlNmC+P5VUUIbFj5H6j2Q35VETp4h1vrqi/XXcuilC9QVleBPsHxvxlum4NL6YnaAFdnOV5+tBdWXLlAa3pPqLhAi4iozfFx88LQyP746WwyxnW8B37ubWdql1arRVBQ4z3pr+QsV+LbExvhp9KiV1BXnCjOwN7sQ5g36AXUGGux6vha/KXnNORU5mN39kHMHfB/AKw7Aaw7nYRY32jIpDKsTd+KAWG9fvO9Mstz8MbgWYjwCvnDfeNQDlEbIWtYoPXX+7vhxPmLmL0gGXklut+/kYiIWtyEmBFwksrw7YmNrd2U6xbg7ovn+zyONSc24eHvnsd/Dn+FZ3pNR4RXCJ7qMQ0mYcZT6+fiX3uXYFqXCbaR1eHt70bPwM54OeldvLD5TUT7RGJM1NDffC8/93YwmA03pd0caSVqY4b2CoG/jxv+/sUBzPooGS9N74GuUW3nX/FERAR4unhgRIdBWH8qCffFDEewR0BrN+m69AjsjB6BnRuVq5RueKHvE03eI5VK8UDcODwQN67Z7/Pn7vdj6ZHVGBM1BL4qn0aH6kT7RDb7WQytRG1Qpwhv/OuvA/DWshS8/vl+/GV8HEb1DW/tZhER0RXGdbwHSWeTsTptPWb3+0trN6dNyqsqRF5VIRYfWN64UgKsnrK42c9iaCVqo/y83fDezES8/7/D+OS748gurMYT42Ih4wItIqI2wV2pwujoIVhzYiPOl11AhFdoazepzfnmxAYMDu+LER0G2m2XdSP4tx9RG+bqLMerjyVg/MD22LgnE/P+sx+62pszN4iIiP64MVFDoFK4YXXa+tZuSptUa6zHvTHDEOThD62bd6Ov68HQStTGyaQSPDa2E56d0hVp50oxe0Ey8rlAi4ioTXBVuODejvfgaMEJnCo519rNaXN6B3XDwdxfb8qzOD2A6DZxT0JowwKtg9YFWg/3RJcO2t+/kYiIbqkRHQZiY8bPWJ22Dq8NfK7RYqM7mdbNG1+l/oiU3KPwU7WDTCqzq7+efW450kp0G4mN9MG/nxsAjdoZry/Zh837slq7SUREdzxnJyXGxwzHieIMpBWfbu3mtCnpJWfQwTscMqkMJbUXUagrtvu6HhxpJbrN+Hm74Z/PWhdoLV5zDNmFVXj8Xi7QIiJqTfdEJmL96W34+viPiB0azdHWBq8Pev6mPYuhleg2dGmB1hcbTmDtznPIK9ZhzvSeULn8sSPyiIjoxshlcky8axSWHFqJw/mpTe6Beic6Xfrb83y5TyvRHUAmleDP98Yi2Ncdi9ccw4sLkjGsdwg2J59EeVU93D00mD4yBgPjg1u7qUREd4SB4X3w46mfsDptPboHxEIq4f+Avbb9X4AEgLiiUAJIIIFEIsFXkxc1+1kMrUS3uWENC7Te+Hw/lq07CaW0HgBQUl6HRd8eAwAGVyKiFuAklWFyp9FYlPIF9uccRd+Q+NZuUqtbNGa+3WuzsKCgughfp67DnzqPv65n8Z8ARA4gLtIHbk1MDdAbzVi+Ob0VWkREdGfqH9ITQWp/fJO2HmaLubWb0+qu3pfVT6VFN/9YPNptCr78dc11PYuhlchBlFfVN1leUl7Xwi0hIrpzSaVS3B83FvnVRdh14UBrN6fNUju7o1BXcl33cHoAkYPw0bhcM6B++v1xTBrcAT6eLi3cKiKiO0+vwK4I1wRjzYmN6B/Ss7Wb06qaWohVZ6zHxoztCFEHXNezGFqJHMT0kTHWOaxX/G+UwkmK6FANtuzLwtb9FzCidygmDekAbw+GVyKiW0UikeCBuHvxTvLH+DlzL+I1nVq7Sa2myYVYALRuXpjZ+9HrehZDK5GDuLTYatWmIyivqodW42LbPaCorBbfbMvA5n1Z2JpyAcN7h2LSYIZXIqJbpatfJ0R7R2DV8bX4TroJ5fpK+Lh6YWrncUgM7dXazWsxVy/EAgAnqRM0Lh7X/SyGViIHMjA+GHFhrgAAb29vW7mvlytmTumKyUM64JttGdi0t2HktU8YJg5qz/BKRHSTSSQSdGoXhdMXz6MW1qlbpbVl+OzgSgC4Y4Kr1s379y9qJoZWojuIn7cbnr2/G6YMjcI32zKwcU8mtu7LsobXwR3gpXZu7SYSETmM5CYWYhnMBnx1/EeHDq2XgvnvkQB4suefmv1chlaiO9CV4XV1UgY27MnEln1ZGNE3DJMGdYCG4ZWI6A+7WFt2XeWOolBX/Jv1RbpSXKwth0wqY2gloubx83bDXx9oCK/bTmPD7kxs2ZuFkX3DMXFQe4ZXIqI/wNvVC6VNBFRvV69WaE3LeX3Q802Wmy1mfH9yM06XnkeYZxCe7vXQdT23RUPrifMXsWx9GnKLdVC7KTBhUAeM7BMGXa0BH60+iuNnS+HqLMfUYdEYlhAKABBCYPmmdPyUcgFmi8DgHsH4872xkEklAICdR3KxYnM6KnV6xEb64Nn7u0Ljbv2L9lxuBT5ecwzZRdUI8HHDM5O6oGOoY/9CIboR/j5ueO6B7raR1/W7zmHzviyM6huGCYPa235PERFR803tPA6fHVwJg9lgK1PIFJjaeVwrtqp1ZJbnYPGB5civLsLEu0ZifMwISKXXd1xAi4VWXa0Bby1LwV/Gx2FAtyCcz6/E3z7dC39vV2zelwVnpRNWzBuBrIIqzPt8HzoEeyI8wAMb92TiUHoRFs4eBAmAN5fux6Y9mRibGIHM/Eos/u4Y3nyyD8ICPPDZ98fxyXfH8fIjvWAwmjF/WQqmDI3CsIRQ7DiUg3e+OID/vDIMcieeqUDUlAAfFZ6f2h33D43C6m0ZWJd8Dpv2WsPrxEEd4OmubO0mEhHdNi7NW1356/coq78zdw8wmU345sQGrD+VhAhNCN69Zy6CPPxv6FktFlqLy+vQ4y5f27Y87YM8EdfeB+lZ5difVojPXhoChVyGqBAN7u4WhJ/2X8BfJnTGjsM5uDcxwrZAZNKQKHy19RTGJkZg55FcJHTyQ3TD6OkjYzrhoXlbUFGtx9ncCkgkEozqGw4AuCchFD8mn8OBk4Xo1/n6NrMlutMEaC+H16+TTmNdsnXkdXTfcIwf2J7hlYiomRJDe+EuVSQA+11d7gQZpefx6cH/oaTmIqZ2vg9joodAKrnxgcMWC60RgR6Y9WC87bWu1oCTmRcR7q+Gk1QCP283W11gOxX2pRYAAHKLdQj2c7fVBWlVyCmqhhACucU6RIdqbHVqNwXcnOXILa5GbrEOIb6X77v03OzCavTrfKt6SeRYArQqvPBgPO6/JxpfJ53G2p1nsXFvJsb0s4ZXDxXDKxER2TOYjfjq+I/YcmYHonwiMSfxafiptH/4ua2yEKumzog3l6YgsmG0dd2u83b1SrkMeoP1WJ96gxlKuexynUIGiwCMJgvqDSYoFTL7exUy6I1m6Juqk8ugN5gataeqqgpVVVV2ZSUl13ceLpEjC9SqMOvBeOu0gaQMfP/LWWzck4nR/cLRTuOKtT8fR3lVPdw9NLYDDYiI6M704pa3UFhTAl83H3Txi8He7EPXvHbCXSOb/dwWD62FF2vw5tIU+Hu7Yc70HsgpqobBaLa7Rm80w1lpDZxKucyuXm8wQyaVQCGXQSl3gsFosb/XYIazwskWXq9+rouycZe//PJLLFq06GZ1kchhBbVzx6w/xdsWbH234ywAQCmtBwCUlNdZj5IFGFyJiO5QJosJPq5eMAsLtp/fc83rJGjDofVsbgXmfb4PA7sH47GxnSCVShDg4waTRaC4vBbtNNaTfPKKdQhu+K/9YF8V8kp0tnmruSVX1RXrbM+v1OmhqzMg2NcddXoTNuzOtHv/vGId7u4W1KhdDz/8MMaPH29Xlpqaiueee+6m9Z3IkQT7umP2tHiknitBWZXerk5vNGP5pnSGViKiO9THY9++Jc9tsWX05dX1mPf5Ptx3d3s8Pi4W0oYtq1yd5Ujo5IflG9NRbzAhI7scO4/mYmB3a7gc2D0Y3+84i9KKOpRX12PN9gwMirfWDegWhL2p+Thx/iIMDX9Rxnf0hdpNgS4dfGA0WbB+13mYzBYkpVxAhU6P7tHtGrVNrVYjKCjI7kur/eNzL4gcXflVgfWSkoo6/Jh8DlU1hibriYiIrleLjbQmpWSjUmfA6qTTWJ102lY+NjECM6d0xcdrjuHRN3+Cs9IJj47pZBtZHdUvHBU6PWZ9tBNGkwUD44Mx7u72AKyLu2ZM7ooFq4+ivFqPThHe+Ov93QAAcicZ5j3RG4vXHMOKzSfh763Cq48lwLmJ6QFEdGN8NC4oKa9rVO4kk+I/P6bhiw0n0TfOH8MSQhHX3sf2j1UiIqLr1WIJbsrQKEwZGnXN+pem92yyXCaV4KGRMXhoZEyT9YldA5HYNbDJuvAAD7z/7IDrbywRNcv0kTHWOaxXTB9XymWYMbkLwgI88FPKBew4lIPkX/Pg5+2Ke3qFYkjPYHh7uLReo4mI6LbEYUciumGX5q2u2nQE5VX10Gpc7HYPePK+ODwy+i7sTS1AUsoFrNicjpVbT6FnjC+GJYQivmM7yGQ87IOIiH4fQysR/SED44MRF2ZdRNnUxtkKuQwDuwdhYPcg5JfokHQgG9sOZiPlRCG81M4Y2isE9/QKsdurmYiI6GoMrUTUYgK0Kjw8+i78aURHHDxZhJ9SLmDN9gx8sy0DXaO0GJYQit6xfpA7yX7/YUREdEdhaCWiFuckk6JPnD/6xPmjpLwO2w5mI+nABby34hDUbgoM7hGMYQmhtu3tiIiIGFqJqFVpNS6YOiwaU4ZG4VhGCX5KuYD1u85j7c5ziAnzwrCEUPTvEsCdP4iI7nD8W4CI2gSZVILuHduhe8d2qKjW4+dDOfgp5QI+Wn0Un/+Yiru7BWFY71C0D/IEAPxyOMe2AIzHxxIROT6GViJqczzdlZgwqD3GD4zEycwybN2fhe0Hs7F5XxYiAj0QHqDGrl/zIDHz+FgiojsFQysRtVkSiQSdIrzRKcIbT47vjJ2Hc7A15QK2H8wBACiv2C1LbzRj+WYeH0tE5Ki4QSIR3RZULnKM7h+Bj14YeM1rSsrroDear1lPRES3L460EtFtRSKRQHuN42MBYNprm9G9Yzv0jvVHzxhfqFwVLdxCIiK6FRhaiei2c63jY0f3D0ed3oSUtELsPV4AqVSCuEhv9I71R0Inf2g1PD6WiOh2xdBKRLed3zs+9qnxnXE2twL70wqwP60An/2Qis9+SEX7IA/0jvVH71h/hPi5QyKRtGY3iIjoOjC0EtFt6beOj5VKJYgK0SAqRIPpo+5CbnE1UtIKsT+tAP/bcgr/23IK/t5uSIj1Q+9Yf3QM84JMygBLRNSWMbQSkcMLaueOoMHumDi4A8qq6pFywhpgN+y2HmLgqVKiVyc/9I71Q5cOWijkPEaWiKitYWglojuKl9oZI/uEYWSfMNTWG3E4vRj70wqw69c8/JRyAc4KGeI7+qJ3rB96cCEXEVGbwdBKRHcsV2c5ErsFIrFbIIwmM1LPXsT+tAKknCjAnuP5kEkliIv0Qe9YPyTE+sPH04UncRGRQ6mor8LsLfPxdK/piA+Ig85Qg08OrEBa8Wm4yl0wudNoDI7oBwAQQuCr1B+x/fweWCxmDAjrjYe7ToJU2jI7qDK0EhEBkDvJbMfIPjWhM87klGN/WiH2pRbg0x9S8ekPqWjn5YKLFfVwAk/iIiLH8OmBFag21Nhef3ZwJZydlPh83HvIrsjD35MXIdIrFKGeQdh6dieO5Kfhn8NfhQTAP3YtxtazOzEyalCLtJWHCxARXUUqlSA61AsPj74Ln740BIvnDMb0UTEoq9TDbBF21+qNZixbfwIms6WVWktEdGN+OpsMpZMSPi4aAEC9sR4H845hSuwYKGRytPcOQ/+Qnth+bg8AIDkrBaOiBkPj4gFPFw/cFzMc287vbrH2MrQSEf2OYF93TB4SBfM1gml5tR4PvLoJcxfvxvJNJ3HgZCGqagwt3EoiouYrqC7GhtPb8ET81MtluhLIpDL4qrS2sgC1L3Kq8gEAedWFCFL7Xa5z90VuVQGEsP/H/K3C6QFERM3kc42TuNRuCgzsHoT0rDJ8v+OsbTQ2UKtCxzANYsK80DHMC8Ht3CHl1lpEdAuVlJQgNzfXrkytVkOtVttemy1mLNz/XzzSbQpUSjdbud6kh0Imt7tXIVPAYDI01BugdLq8OFXppIAQAkaLqdF9twJDKxFRM13rJK4nxsXa5rTWG0w4m1OBUxfKcSqrDAdPFmH7wRwAgJuzE6LDvNAx1AsxYdZ9ZF2db/0f9ER053juuecalc2YMQMzZ860vf7u5CaEeQahe0Cs3XUKmQIGs9GuzGA2wFmuBAAor6rXmwyQSaQtElgBhlYiomb7vZO4AMBZ4YTYSB/ERvoAsK62LSitwakLZUjPsgbZr346BSEAqQQI9VejY6h1JDYmzAt+3q48qYuIbtiHH36IuLg4u7IrR1kBYG/2YZTXV2JvzmEAQJ2xHh/uW4pxHYfBbDGjtKYMPm5eAID8qiIEqv0BAIFqP+RXFaGDd7i1rvpyXUtgaCUiug6/dRJXUyQSCQK0KgRoVRjcIwQAUFNnxOnscpzOKkN6Vhl2Hs3F5n1ZAABPlRIdwzS2INs+2BNKuYxbbRFRs2i1WgQFBf3mNR+Ommf3+v/Wv4LH4h9AfEAcsipysOr4Wvyl5zTkVOZjd/ZBzB3wfwCAxNBeWHc6CbG+0ZBJZVibvhUDwnrdqq40wtBKRNTC3Fzk6B7dDt2j2wEAzBaBnKJqpGeV4VTD1/60QgCAk0wCHw8XlFTUcastIrrlnuoxDUsOr8JT6+fC2UmJaV0m2EZWh7e/G5X6Kryc9C6MFhMSQ3thTNTQFmsbQysRUSuTSSUI81cjzF+NkX3CAACVOj1ONYzErtt1HmaLgNMV+73ojWZ8/N0x1OlNCPP3QKi/O+fHEtEN+Xjs27afq5RueKHvE01eJ5VK8UDcODwQN66lmmaHoZWIqA3yUCmREOuPhFh/fL/jbJPX1OvNWPzdcdtrXy9Xa/gNUCPc3wNhAWr4ebtBxh0LiMgBMLQSEbVx19pqS+vpgn/M6I+sgipk5VdZfyyoxMGThbh0BoJCLkOon3ujMOvuqmj0PCKitqxVQmtGdjne/m8Kvnx9BABAV2vAR6uP4vjZUrg6yzF1WDSGJYQCsK68Xb4pHT+lXIDZIjC4RzD+fG+sbeRg55FcrNicjkqdHrGRPnj2/q7QuDsDAM7lVuDjNceQXVSNAB83PDOpCzqGerVGl4mIbti1ttqaPioG7TSuaKdxRa+7Lm/4rTeakVNUbRdkU04UIulAtu0abw9nhPmrER7gYQu0gVoVnGT2Z85wARgRtRUtGlqFENh2IBtL16VBKr38B+PCb3+Fs9IJK+aNQFZBFeZ9vg8dgj0RHuCBjXsycSi9CAtnD4IEwJtL92PTnkyMTYxAZn4lFn93DG8+2QdhAR747Pvj+OS743j5kV4wGM2YvywFU4ZGYVhCKHYcysE7XxzAf14ZBrkTDwIjottHc7baupJSLkP7IE+0D/K0lQkhUFGtR6ZtVLYSWQVVOHamBCazdVjWSSZFiK87wgKs82srdXqs330eEjMXgBFR62vR0PrN9gzs/jUfU4ZGYc3P1jladXoT9qcV4rOXhkAhlyEqRIO7uwXhp/0X8JcJnbHjcA7uTYyAl9o6ejppSBS+2noKYxMjsPNILhI6+SG6YfT0kTGd8NC8Laio1uNsbgUkEglG9bWueLsnIRQ/Jp/DgZOF6Nc5oCW7TUT0h13vVltXk0gk0KidoVE723YtAACjyYK8Eh2y8q0hNrOgCr9mlODnQzm2a5RXLQBbtv4E+nYOgEIuu/EOERFdpxYNrff0CsWUIVFIO3fRVpZfooOTVAI/78vHiAW2U2FfagEAILdYh2A/d1tdkFaFnKJqCCGQW6xDdKjGVqd2U8DNWY7c4mrkFusQ4nv5vkvPzS6sRr/O9u2qqqpCVVWVXVlJSckf7i8RUVsnd5Ladi64UqVOj2mvb2nynvJqPSbP3YAArco2tSDMT42wAA+007jwcAQiuiVaNLReGi29Ur3B3Ohf60q5DHqD2VavvKJeqZDBIqyjA/UGE5SKq+5VyKA3mqFvqk4ug95gatSGL7/8EosWLbrhfhERORoPlRLaaywAU7spMLJvGC4UVOFsbgV2H8u31bk6OyHUT20Ls5d+7ubC7biI6I9p9d0DlAoZDEazXZneaIaz0ho4lXL7er3BDJlUAoVcBqXcCQajxf5egxnOCidbeL36uS7Kxl1++OGHMX78eLuy1NTUJs/vJSK6U1xrAdgT42Lt5rTW1huRbbfwqwrJv+bZTvkCgHYaF4Q2jOiGN+wrG6hVQSZreo0BF4AR0dVaPbQG+LjBZBEoLq9FO411vlZesQ7BDf+1H+yrQl6JzjZvNbfkqrpine1ZlTo9dHUGBPu6o05vwobdmXbvlVesw93dGh9tplarG53LW1hYePM6SUR0G2ruAjBXZ7n12NkrdmcRQqC0ot624OvS15FTxTA37Mcld5Ii2NfdNj3h0ujssYyShrDMBWBEdFmrh1ZXZzkSOvlh+cZ0zJjSBdmF1dh5NBevP94bADCwezC+33EWndtrIZNJsGZ7BgbFW4PngG5BmLt4N4b2CkGHYE8s35SO+I6+ULsp0KWDD4wmC9bvOo+RfcOw41AOKnR6uwUIRET02250AZhEIoFW4wKtxgU9r9iOy2gyI7dYh8z8KlxoCLK/ZhTbLfySSAAhGi8AW74pnaGV6A7W6qEVAGZO6YqP1xzDo2/+BGelEx4d08k2sjqqXzgqdHrM+mgnjCYLBsYHY9zd7QEAEYEemDG5KxasPoryaj06RXjjr/d3AwDInWSY90RvLF5zDCs2n4S/twqvPpYA5yamBxARUcuQO8kQHuCB8AAPu/JKnR4XCq3bcX3+Y1qT95ZU1GHmP3cgQOuGQK0KAT5uCNCqEKhVQe2m4AIwIgfXKgkurr0PVs0faXvt7qrAS9N7NnmtTCrBQyNj8NDImCbrE7sGIrFrYJN14QEeeP/ZAX+8wUREdEt5qJTo3F6Lzu21WJt8rskFYC5KJ2g1LrhQUIWUtELbNAMAcHN2QoBWhQAfFQK1bvDXWn8M8FFxERiRg+CwIxERtSnXWgD2zMTOtukBZrMFReW1yC+pQX6JDvmlNcgr0SH9QhmSf82FuJxn4alSIqAhwAZoL4/O+vu42e1Ow8VfRG0bQysREbUpzVkAJpNJrSHURwXE+NrdbzCaUXCxplGgPXK6CNsO6u2u9fF0QYCPdZ/wE5kX4SS4+IuorWJoJSKiNuePnACmkMsQ6mfdI/ZqtfVGFJRaA21eqc4aaktqkJFTDiEAp6sWf320+ij2phagncYVWo0L2mlcoPW0/pzzaIlaFkMrERHdMVyd5YgM8kRkkKdd+b2zfmzyepNZILe4GkdOF9sOvblEIZdB69kQZDWuDT9eDrXeHi6QOzW9D+2VOC2BqHkYWomI6I7nc43Tv7QaFyyeMwRCCFTXGlFSXovi8jqUVNSipLzO+lVRi8yThaiotp96IJFYT4LUel4Raj1doPVybQi7rjh4spB70hI1E0MrERHd8a61+Gt6w841EokEajcF1G6KRqO0lxiMZpRWWINscXktSioafiyvw9ncCuxLLYDJbH+KowSAQOM9aT//MQ1+3m7wdFfC010JZwX/uibi7wIiIrrjNff0r9+ikMus225pVU3WWywClTr95UBbVof/bjjR5LVVNQa8uHCX7bWL0gme7kpo3JXQuDvbfu55xetLZXInWZPPvBqnJdDthqGViIgIf2zxV3NIpRJo1M7QqJ0RHWot27DnfJPTEjTuSjx7fzdUVNejvFqPioav8mrrIQzHzuihqzM2+T5uLvJGgVZjC7nW1+lZF/HlhnROS6DbCkMrERFRK7nWtITHxnZCj6u28rqa0WRGRbUB5dX1qNDpUV6lR4WuHhVV1nBbodPjXG4FKnR61NabmnzG1dMSFn93HKWV9fBUKaFRK+GpsoZfT5USMtnvLyojupUYWomIiFrJH5mWIHeSWXcr0Lj87rV6o7lhpLYeFdV6vP3fA01eV6c34cuNJxuVSyTW0ytt0xJU9lMUrhzV9XBT/GbA5bQEulEMrURERK3oVk9LAKyjt75ervD1sr6P9rd2S3hxsP3IbcO0hIqG0dvyqnqczi5DRbUe9VdtAwZcDri2QKtyto3aFl6swfZDOZBaOC2Brh9DKxER0R3mt3ZLcFY6wU/pBD9vt999Tp3eZJtvW6G7ev6tNfCezi5DebXebp/bq6clfLT6KHb9mg+NWgnvhnm/Xg1flwIvpycQQysREdEd5mbslgBYdzVwUTrB36d5AXfKyxubrDOZBUoqapGRXY4Knb5RvVQCqFXKy0HWXQkvj0s/d4a3h/VHjVoJp2uEW05LuP0xtBIREd2BWmJawpVclE6/OS1hwaxBAACT2YKKaj3KqupRVlWP8qp6lFVd8bq6HudyK1Cp08MiGr+Ph0oBjbv9SO3FyjokH82HTHBawu2MoZWIiIhaxO8d4gAATjIpfDxd4OP52wvMzA373jYVbssbyi4UVqG8Wg9LQ7qVXTUtYen6E0jsGsipB7cJhlYiIiJqETdrWgIAyKQS22jqb7FYBMa9uK7JuopqPab+bTM6t/dB5w4+6NpBi2Bfd0gkkutuD916DK1ERETUYlp6WoJUKrnmtAS1mwJ9Owfg2JkSpJwoBAB4qZXo0kGLrlFadOmghbfH728pRi2DoZWIiIgc2rWmJTwxLtY2yltUVotjZ0pwLKMEh08VY8fhXABAsK/KGmI7aBEb6QM3F3lrdIHA0EpEREQOrjnTEny9XDEsIRTDEkJhsQhcKKzCrxkl+PVMCZIOZGPD7kxIpRJEBXuiS5Q1xEaHekHuxPmwLYWhlYiIiBze9UxLkEolCA/wQHiAB8YPbA+jyYxTF8pxrCHEfrstA6uTMqBUyNApwhtdG6YThPqpIZXaz4flVls3D0MrERER0W+QO8kQF+mDuEgfTBsZg5o6I1LPleJYRgmOnS3BsvUnAFi32+rSXmsbiT2ZebFhWgK32roZGFqJiIiIroObixy9Y/3RO9YfAHCxsg7HzpTg14wSHDtTguRf8wBYR2wtFtHoBLDlm9MZWm8AQysRERHRH+Dt4YLBPUIwuEcIhBDIKarGr2dK8PnatCavL21iJwP6fZw9TERERHSTSCQShPipcW9iJLSaprfL8rlGOf02hlYiIiKiW2D6yBgo5TK7sqtPAKPm4/QAIiIiolvgZp4ARgytRERERLdMS58A5sgcPrSey63Ax2uOIbuoGgE+bnhmUhd0DPVq7WYRERER0XVw6DmtBqMZ85elYGivEHz91iiM7R+Bd744AKPJ0tpNIyIiIqLr4NCh9fjZUkgkEozqGw4nmRT3JITC3VWBAycLW7tpRERERHQdHHp6QG6xDiG+7nZlge1UyC6sRr/Ol8uqqqpQVVVld11JSUlLNJGIiIiImsGhQ6veYIJS0XirCb3BZFf25ZdfYtGiRS3ZNCIiIiK6Dg4dWpUKGfRGs12Z3miGi9K+2w8//DDGjx9vV5aamornnnvuVjeRiIiIiJrBoUNrUDt3bNidaVeWV6zD3d2C7MrUajXUarVdWWEh570SERGRY9qbfRjfpm1AaV05tK5eeCDuXvQK6gqdoQafHFiBtOLTcJW7YHKn0Rgc0Q8AIITAV6k/Yvv5PbBYzBgQ1hsPd50EqbRllkg59EKsLh18YDRZsH7XeZjMFiSlXECFTo/u0e1au2lERERErSK/ugifHFiOp3pNw4qJH+KRbpPx4b6lqNLr8NnBlXB2UuLzce9hVt8n8b9jP+BCRS4AYOvZnTiSn4Z/Dn8VH4x8HadLz2Hr2Z0t1m6HDq1yJxnmPdEbyUdz8eDfNmHD7ky8+lgCnJUOPcBMREREdE0B7r74fNy7iPaJRL2xHhX1VXBxUsJJKsPBvGOYEjsGCpkc7b3D0D+kJ7af2wMASM5KwaiowdC4eMDTxQP3xQzHtvO7W6zdDp/ewgM88P6zA677PoPBAAA4derUzW6SzaUdC66emnCr8X35vo7wnnxfvq+jvCffl+97M1zKK3l5efDz87Ora2oapLPcGcW6Uszc9BoggMfjp6JIVwqZVAZfldZ2XYDaFym5R63Pri5EkPryswPcfZFbVQAhBCQSya3qmo3Dh9Ybdf78eQDA/PnzW7klRERERM0zZ86cRmUzZszAzJkzG5V7u2qwctJCnCo5g3d3f4pxHe+BQia3u0YhU8Bgsg7k6U0GKJ0UtjqlkwJCCBgtpkb33QoMrdcwZswYAEBkZCTk8lv/jSAiIiK6UUajESdOnEC/fv3g7m6/R/21RnhlUuu2oLG+HZEQ1BXnyi7AYDbaXWMwG+AsVwIAlDKFXb3eZIBMIm2RwAowtF6Tp6cnpk2b1trNICIiImqWPn36NOu6I/lp2JixDX8b+JytzGQxw1elhbngBEpryuDj5gUAyK8qQqDaHwAQqPZDflUROniHW+uqL9e1BIdeiEVERERE9iI0wThXlo3krBRYhAVH8tNwtCAN90QmokdgZ6w6vhZ6kwFnL2Zhd/ZBJIb2AgAkhvbCutNJuFhbjor6KqxN34oBYb1arN0SIYRosXcjIiIiolaXXnIGXx5dg4LqYvi7t8O0LhMQ6xsNnb4GSw6vQmrRKTg7KTG50xgMjugLALBYLPjmxHrsOL8PRosJiaG9WnSfVoZWIiIiImrzOD2AiIiIiNo8hlYiIiIiavMYWomIiIiozWNoJSIiIqI2j6GVbprdu3cjISEBHh4eiIyMxGeffQYAKC8vx/jx4+Hh4YGQkBAsXbrUdo8QAnPnzoVWq4VGo8Ff//pXmM3m1upCs33zzTeIiYmBSqVCp06dsHbtWgCO2dcrFRUVoV27dtiwYQMAx+3v+++/D4VCAZVKZfvatWuXQ/Y3NzcXY8aMgVqtRlBQEBYsWADAMb+3K1eutPueqlQqSKVSPPnkkw7ZXwDYu3cvevToAbVajejoaKxatQqAY35/W8v27dvRrVs3uLu7o0+fPkhJSQEAHDx4EDKZzO7X29///ncA/IxvmCC6CcrKyoRGoxH/+9//hNlsFocPHxYajUYkJSWJiRMnimnTpom6ujqRkpIivLy8xLFjx4QQQixcuFDExcWJ/Px8UVBQIOLj48WCBQtauTe/7fTp08LV1VXs2bNHCCFEUlKSUCgUoqSkxOH6erXRo0cLqVQq1q9fL4QQDtvfBx98ULz//vuNyh2tvxaLRcTHx4vZs2cLg8Eg0tLShEajEXv27HG4vjZl27Ztwt/fX+Tk5Dhkf00mk9BqteLbb78VQgiRnJwsnJycRGZmpkP2tzVkZmYKV1dXsWTJEmE0GsWGDRuEl5eXKCgoEEuWLBGjR49u8j5+xjeGoZVuiqNHj4pp06bZlU2YMEG88cYbQiaTiXPnztnKZ8yYIWbMmCGEEKJXr15i6dKltro1a9aI2NjYlmn0H1BdXW37cfny5cLb21tUVlY6ZF8v+eSTT8SUKVNEaGioWL9+vaiurnbY/sbExIikpCS7Mkfs7759+4S/v78wmUy2slOnTomsrCyH6+vVqqurRWBgoPjhhx8c8nsrhBClpaUCgPjqq6+ExWIRu3fvFq6uriI7O9sh+9saFi9eLBISEuzKRo4cKRYsWCCefvpp8corrzR5Hz/jG8PpAXRTdO3aFStWrLC9Li8vx65duyCEgFwuR0REhK0uOjoaJ06cAACcOnUKd911l13dyZMnIdr49sEqlQqZmZnw8PDAww8/jLfffhvnzp1zyL4CwJkzZ/Cvf/0Ln3zyiV2ZI/a3trYWGRkZ+Oijj+Dn54eYmBgsW7bMIft75MgRdOrUCXPmzIGfnx+ioqKwf/9+lJWVOVxfr/bee+8hLi4O9913n0N+bwHA29sbzzzzDKZOnQq5XI7ExEQsWrQIpaWlDtnf1mA2m+Hq6mpXJpVKcebMGRw9ehR79uxBeHg4QkJCMHv2bOj1egD8jG8UQyvddJWVlRg7dizi4+MxaNAguLi42NW7urqitrYWAFBTU2P3G97V1RUWi8X2G7stCw4ORn19PbZt24ZZs2Zh/fr1DtlXk8mEadOm4aOPPoKXl5etvKamxiH7W1RUhH79+uHpp59GdnY2lixZghdeeAEbNmxwuP6WlZVhx44d8PHxQXZ2Nr744gvMnDkTOp3O4fp6JZ1Oh4ULF+L1118H4Li/li0WC1xdXfHtt9+itrYW69evx3PPPYeqqiqH7G9rGD58OFJSUvDtt9/CaDRiy5Yt+Pnnn1FfXw+tVouxY8ciLS0Nv/zyC3bs2GH3a46f8fVjaKWbKjMzE3379oWXlxe+//57qFQq1NXV2V1TW1sLlUoFwPob9cr62tpaODk5wdnZuUXbfSOcnJwgl8sxePBgTJw4EYcOHXLIvs6fPx9du3bFqFGj7Mqv7g/gGP0NDw/Hzp07MWrUKCgUCiQmJuKhhx5CcnKyw/VXqVTCy8sLc+fOhUKhQN++fTFx4kS8/vrrDtfXK61duxahoaHo3bs3AMf9tfz9998jJSUFkyZNgkKhwOjRozF69GjMmzfPIfvbGjp06IBvvvkGb731Fvz9/bFq1Sr86U9/gqenJ9atW4cXXngBbm5uiIiIwMsvv4wffvgBAD/jG8XQSjfNkSNHkJCQgOHDh2Pt2rVwcXFBhw4dYDQakZ2dbbvu9OnTtv8WiYmJwenTp+3qYmJiWrzt12PTpk0YOnSoXZnBYEBkZKTD9RUAVq9eja+//hqenp7w9PREdnY2HnjgAWzcuNEh+3vkyBH84x//sCurr69HSEiIw/U3OjoaNTU1MJlMtjKz2Yxu3bo5XF+vtH79ekyZMsX22hH/nAKA7OzsRiN3crkc3bt3d8j+tobq6mqEhITg2LFjKC0txfLly5Gamopu3brhxRdfRHV1te3a+vp6WyjlZ3yDWm02LTmUwsJCodVqxT/+8Y9GdRMmTBAPPvigqKmpEQcOHBBeXl5i//79QgghFixYIGJjY0VOTo4oLCwU8fHx4r333mvp5l+XgoIC4eHhIZYvXy7MZrPYuHGjUKvVIj093eH62pRLC7GEcLzvrRDW3SGcnZ3Ft99+K8xms9i2bZtQqVTi8OHDDtff2tpaERgYKGbNmiWMRqPYs2ePUKlUYt++fQ7X1yuFhISIn3/+2a7MEft7/PhxoVAoxLJly4TFYhG//PKLcHd3FwcPHnTI/raGzMxM4ebmJg4fPiz0er34+OOPhb+/v9DpdCIiIkK88MILwmAwiDNnzoiYmBjxwQcfCCH4Gd8ohla6Kd5++20BQLi5udl9vfzyy+LixYti8uTJQqPRiODgYLsVkyaTSbzyyivC399feHl5iWeffdZuJXNblZycLOLj44W7u7uIj4+3/QXoiH292pWh1VH7u27dOhEXFydcXV1FVFSUbcsgR+zvmTNnxPDhw4VGoxGhoaFi2bJlQgjH7KsQ1rZLpVKRnp5uV+6o/V23bp3o0qWLcHd3F506dRLff/+9EMJx+9saVqxYIcLCwoS7u7u4++67RVpamhBCiBMnToghQ4YItVotfH19xWuvvSYsFosQgp/xjZIIwaVqRERERNS2cU4rEREREbV5DK1ERERE1OYxtBIRERFRm8fQSkRERERtHkMrEREREbV5DK1ERERE1OY5tXYDiIhupZdeesl2dGJTAgMDAQCTJk3CM88801LNwksvvYTCwkJ88cUXN/yM6OhovPfeexg3blyT9Y888gj8/PwanfBFRHQ7YmglIof2yiuvYNasWQCAgoICTJ48GYsXL0bnzp0BADKZDJMmTWrNJhIRUTMwtBKRQ3N3d4e7uzsA2M5h9/DwgFarbc1mERHRdeKcViIiAEVFRXjqqafQpUsX9O/fH59++qmtbuHChXjooYfw7LPPonv37vjggw8AANu2bcO9996LuLg4jBgxAkuXLoXFYrHdt2TJEgwZMgSxsbEYPnw4Vq5cafeeRqMRb7/9NhISEtC9e3fMmTMHtbW1tvqMjAw88cQT6NmzJ3r16oU5c+agrKysyfZbLBYsWLAA/fv3R7du3fDOO+/AbDbb6s1mM959910kJiYiNjYWY8eOxebNm2/KZ0dE1BIYWomIAHz33Xe4++67sWHDBkyfPh0ffPABDh48aKs/cOAAgoOD8cMPP2DSpEnYuXMnZs+ejenTp2Pjxo148cUXsXz5cixevBgA8PPPP2Pp0qV46623sHXrVjz++OOYP3++3TMPHToEk8mE1atX49///je2bNmCZcuWAQByc3MxdepUeHh4YOXKlVi8eDFOnTqFxx57zC6MXvLJJ59g+fLlePXVV7FmzRpUVlbiwIEDtvpVq1YhKSkJCxcuxJYtWzBixAjMmjULOTk5t+ojJSK6qTg9gIgIwPDhwzF16lQAwJNPPoklS5YgLS0NPXv2BABIJBLMnDkTzs7OAIA5c+Zg6tSptvmwISEhqKmpwd/+9jc888wzyM7OhlwuR0BAAAIDAzF58mQEBQUhIiLC9p5+fn547bXXIJFIEBYWhn79+iEtLQ2ANWSq1Wq88847kMvlAIAPPvgAo0aNwq5duzBw4EDbc4QQWLVqFR599FGMGDECAPDmm29i7969tmsuXLgAFxcXBAYGQqvV4plnnkHnzp3h6el5az5QIqKbjKGViAhAeHi43Wu1Wo36+nrba61WawusAJCeno7U1FR8/fXXtjKLxYL6+nrk5eVh7NixWLNmDYYNG4aoqCj0798f9957L7y9vW3Xh4SEQCKR2F57eHigqKgIAHDmzBnExcXZAisAREZGQqPRICMjwy60lpeXo7S0FLGxsbYyhUKBu+66y/b6wQcfRFJSEgYMGIDY2FgkJiZi3Lhxtvm+RERtHUMrEREAqbTxbCkhhO3nVwZWAJDL5Xj88ccxduzYRvf5+vpCoVBg3bp1OHz4MHbv3o2dO3fiyy+/xLvvvmu7RyaTXfM9lUplk+20WCx2QfZa7QWswfWSiIgIbNu2Dfv27cOePXuwceNGLFu2DJ999hkSEhKafB4RUVvCOa1ERDegffv2yMrKQmhoqO0rIyPDtkhr06ZN+Oqrr9CzZ088//zzWLt2Lfr164d169Y1+/mpqakwGo22srNnz6KyshKRkZF213p5ecHX1xdHjx61lVksFpw8edL2euXKlfjpp58wYMAAzJ07F5s3b0ZQUBAXYxHRbYOhlYjoBjz99NPYuHEjlixZgqysLPzyyy947bXX4OzsDIVCAYPBgHfffRfr1q1DXl4e9u3bh5MnT6JLly7Nev60adNQXV2NuXPn4syZMzh06BBmz56Njh07ok+fPo2uf+yxx7B8+XKsXbsW58+fx/z585Gfn2+rLy8vx/z587Fjxw7k5eVh+/btyM3NbXZ7iIhaG6cHEBHdgAEDBuC9997DkiVLsGDBAnh5eeG+++7D888/DwC47777cPHiRSxcuBAFBQXw9vbGhAkT8NRTTzXr+T4+Pli2bBnef/99TJw4ES4uLhg8eDBefPHFJqcHPPLIIxBC4MMPP0R5eTmGDx+OoUOH2uqfeuop1NfX44033kBpaSn8/f0xc+ZMjB8//uZ8IEREt5hEXD0JioiIiIiojeH0ACIiIiJq8xhaiYiIiKjNY2glIiIiojaPoZWIiIiI2jyGViIiIiJq8xhaiYiIiKjNY2glIiIiojaPoZWIiIiI2jyGViIiIiJq8/4/c2yobOlK0sQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 720x432 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "threshs = np.arange(0.1, 1.0, 0.05)\n",
    "rows = np.repeat(None, len(threshs))\n",
    "uniq_g1s, uniq_g2s = np.repeat(None, len(threshs)), np.repeat(None, len(threshs))\n",
    "for i, thresh in enumerate(threshs): \n",
    "    # print(f\"thresh: {thresh*1_000:3.0f}\\n{20*'='}\")\n",
    "    protein_temp = proteins2[proteins2['combined_score'] > (thresh * 1_000)]\n",
    "    rows[i] = protein_temp.shape[0]\n",
    "    uniq_g1s[i] = len(protein_temp.gene_symbol1.unique())\n",
    "    uniq_g2s[i] = len(protein_temp.gene_symbol2.unique())\n",
    "\n",
    "fig, ax1 = plt.subplots(figsize=(10, 6))\n",
    "\n",
    "ax1.plot(threshs*1_000, rows, 'bo-');\n",
    "ax1.set_xlabel(\"Thresholds\", fontsize=15);\n",
    "ax1.set_ylabel(\"Number of Rows\", color='b', fontsize=PlottingParameters.YLABEL_FONTSIZE.value);\n",
    "ax1.tick_params(axis='y', labelcolor = 'b', labelsize=PlottingParameters.YTICKS_LABELSIZE.value)\n",
    "\n",
    "ax2 = ax1.twinx()\n",
    "ax2.plot(threshs*1_000, uniq_g1s, 'go-');\n",
    "ax2.axhline(y=700, color='grey', linestyle='--');\n",
    "ax2.set_ylabel(\"Number of Unique Genes\", color='g', fontsize=PlottingParameters.YLABEL_FONTSIZE.value);\n",
    "ax2.tick_params(axis='y', labelcolor = 'g', labelsize=PlottingParameters.YTICKS_LABELSIZE.value);\n",
    "ax2.set_ylim(300, 900);\n",
    "\n",
    "for thresh in threshs:\n",
    "    ax1.axvline(x=thresh*1_000, color='grey', linestyle='-', alpha=0.2);\n",
    "\n",
    "plt.xticks([200, 300, 400, 500, 600, 700, 800, 950]);\n",
    "ax1.tick_params(axis='x', labelcolor='black', labelsize=PlottingParameters.YTICKS_LABELSIZE.value);\n",
    "plt.title(\"Number of Remaining Rows & Unique Columns for Score Thresholds\", \n",
    "          fontsize=PlottingParameters.TITLE_FONTSIZE.value);\n",
    "plt.show();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Choosing threshold 700 we end up with shape: (7794, 12)\n",
      "Number of unique gene_symbol1's: 696\n",
      "Number of unique gene_symbol2's: 696\n"
     ]
    }
   ],
   "source": [
    "SCORE_THRESH = 700\n",
    "\n",
    "# Select only the rows above the chose threshold.\n",
    "proteins3 = proteins2[proteins2['combined_score'] > SCORE_THRESH]\n",
    "print(f\"Choosing threshold {SCORE_THRESH} we end up with shape: {proteins3.shape}\")\n",
    "print(\"Number of unique gene_symbol1's:\", len(proteins3.gene_symbol1.unique()))\n",
    "print(\"Number of unique gene_symbol2's:\", len(proteins3.gene_symbol2.unique()))\n",
    "assert set(proteins3.gene_symbol1.unique()) == set(proteins3.gene_symbol2.unique()), \\\n",
    "    \"ERROR: The unique gene_symbol1's are not the same as the unique gene_symbol2's!\"\n",
    "UNIQ_GENE_SYMBOLS = list(set(proteins3.gene_symbol1.unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correct indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7794, 12)\n",
      "(696, 2)\n"
     ]
    }
   ],
   "source": [
    "def get_proteins_and_genes_above_thresh(proteins: pd.DataFrame,\n",
    "                                        genes: pd.DataFrame,\n",
    "                                        thresh: float):\n",
    "    \"\"\"Return new proteins dataframe and gene dataframe by only selecting \n",
    "    data above the chosen threshold.\"\"\"\n",
    "    p_sub = proteins[proteins['combined_score'] > thresh]\n",
    "    gene_symbols = list(set(p_sub.gene_symbol1.unique()))\n",
    "\n",
    "    # Select only the gene symbols with a combined_score > SCORE_THRESH.\n",
    "    gene_sub = genes[genes.GENE_SYMBOL.isin(gene_symbols)].reset_index(drop=True)\n",
    "    gene_sub.loc[:, 'GENE_INDEX'] = gene_sub.index\n",
    "\n",
    "    return p_sub, gene_sub\n",
    "\n",
    "p_sub, gene_sub = get_proteins_and_genes_above_thresh(proteins2, inter_genes, SCORE_THRESH)\n",
    "print(p_sub.shape)\n",
    "print(gene_sub.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape before: (858, 2)\n",
      "  GENE_SYMBOL  GENE_INDEX\n",
      "0       LPAR2           0\n",
      "1      PEX11A           1\n",
      "2    MAPKAPK2           2\n",
      "Shape after : (696, 2)\n",
      "  GENE_SYMBOL  GENE_INDEX\n",
      "0       LPAR2           0\n",
      "1      PEX11A           1\n",
      "2    MAPKAPK2           2\n"
     ]
    }
   ],
   "source": [
    "# Select only the gene symbols with a combined_score > SCORE_THRESH.\n",
    "temp = inter_genes[inter_genes.GENE_SYMBOL.isin(UNIQ_GENE_SYMBOLS)]\n",
    "temp.reset_index(inplace=True, drop=True)\n",
    "inter_genes_thresh_700 = temp.copy()\n",
    "inter_genes_thresh_700.loc[:, 'GENE_INDEX'] = inter_genes_thresh_700.index\n",
    "del temp\n",
    "print(\"Shape before:\", inter_genes.shape)\n",
    "print(inter_genes.head(3))\n",
    "print(\"Shape after :\", inter_genes_thresh_700.shape)\n",
    "print(inter_genes_thresh_700.head(3))\n",
    "\n",
    "inter_genes_thresh_700.to_csv(PROCESSED_DATA_PATH + 'thresh_700_inter_genes.csv', header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add gene index to proteins dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There will be 696 nodes in the graph.\n",
      "There will be 696 indices in the graph.\n",
      "(7794, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "      <th>index_gene_symbol1</th>\n",
       "      <th>index_gene_symbol2</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000387365</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>PWP1</td>\n",
       "      <td>289</td>\n",
       "      <td>519</td>\n",
       "      <td>726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000254322</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>DNAJB1</td>\n",
       "      <td>289</td>\n",
       "      <td>539</td>\n",
       "      <td>749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000302961</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>HSPA4</td>\n",
       "      <td>289</td>\n",
       "      <td>605</td>\n",
       "      <td>988</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2 gene_symbol1 gene_symbol2  \\\n",
       "0  ENSP00000001008  ENSP00000387365        FKBP4         PWP1   \n",
       "1  ENSP00000001008  ENSP00000254322        FKBP4       DNAJB1   \n",
       "2  ENSP00000001008  ENSP00000302961        FKBP4        HSPA4   \n",
       "\n",
       "   index_gene_symbol1  index_gene_symbol2  combined_score  \n",
       "0                 289                 519             726  \n",
       "1                 289                 539             749  \n",
       "2                 289                 605             988  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins4 = add_gene_index(proteins3, inter_genes_thresh_700)\n",
    "proteins4 = proteins4[['protein1', 'protein2', 'gene_symbol1', 'gene_symbol2', \n",
    "                       'index_gene_symbol1', 'index_gene_symbol2', 'combined_score']]\n",
    "\n",
    "NODES_AS_SYMBOLS = list(np.unique(proteins4[['gene_symbol1', 'gene_symbol2']].values))\n",
    "NODES_AS_INDECES = list(np.unique(proteins4[['index_gene_symbol1', 'index_gene_symbol2']].values))\n",
    "print(f\"There will be {len(NODES_AS_SYMBOLS)} nodes in the graph.\")\n",
    "print(f\"There will be {len(NODES_AS_INDECES)} indices in the graph.\")\n",
    "\n",
    "print(proteins4.shape)\n",
    "proteins4.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCNB1', 'GMNN'), ('MYC', 'PIK3CA'), ('POLR1C', 'RRP8'), ('DNMT3A', 'MAT2A'), ('PECR', 'ECH1')]\n",
      "[(529, 258), (292, 119), (564, 306), (149, 561), (141, 14)]\n",
      "Number of neighbor tuples: 7794\n"
     ]
    }
   ],
   "source": [
    "neighbor_gene_tuples = get_neighbor_tuples(proteins4, inter_genes_thresh_700)\n",
    "print(neighbor_gene_tuples[:5])\n",
    "print(\"Number of neighbor tuples:\", len(neighbor_gene_tuples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUCCESS: The given list of tuples is undirected!\n"
     ]
    }
   ],
   "source": [
    "neighbor_gene_tuples_undirected = neighbor_gene_tuples if is_undirected_list_of_tuples(neighbor_gene_tuples) else None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dictionary of neihbors per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_as_indices = create_graph_dict_with_indices(proteins4, NODES_AS_INDECES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAt0AAAGnCAYAAABrZuvIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAABGgElEQVR4nO3deVyVZf7/8fdhkQPjmmK2CG6EJrhvjWaGOaaJjmulknsQZu6aaS6huJCayxhM0aLiuI1pbjWWlVMuxWgS7gpuYy6MSi7snN8f/jxfj6AehJuD+Ho+Hj4enuu+7uv+3Bcne5+b676PyWKxWAQAAADAME6OLgAAAAAo7gjdAAAAgMEI3QAAAIDBCN0AAACAwQjdAAAAgMEI3QAAAIDBCN1AMRAQEKA2bdooJSUlx7agoCCNHz/esGOfPn1avr6+io2NNewY9tq3b5/at28vPz8/zZw5M8f2BQsWyNfXV9u2bcuxbc2aNXr66aftPlZe+7/99tvq27fvHbfv2rVLvr6+Onv2rN1jPuiuXr2qDz74QO3atVPdunXVsmVLjRw5UomJiXkaJyAgQIsWLZJ042fcpk2bu/b39fXVunXr7BrPSHv37tXw4cPVsmVL+fn56dlnn9WYMWN07Ngxw4+d1/cvgPwjdAPFxMmTJzVnzhxHl+FQUVFRcnFx0aZNm/T666/fsd/EiRN19erVfB2rffv2uYZ32CcpKUldunTR999/r5EjR2rDhg2aN2+erly5oldeeUVHjhxxWG2rV6++6wekgrBmzRr17NlTJUuW1Lx587RlyxZ98MEHSk1NVffu3R16/gCMQegGionKlStr6dKl2r17t6NLcZgrV66oVq1a8vLyUrly5XLtU7ZsWV29elWzZs3K17HMZrMqVKiQrzEeZpMnT5bFYtHSpUv1wgsvqHLlyqpfv77+9re/6dFHH831NxWF5ZFHHpGHh4dh4586dUpTpkzRgAEDFBYWpvr16+uxxx5Tw4YNNX/+fPn5+WnBggWGHR+AYxC6gWKic+fOql+/vsaPH6+0tLRc++S2FOT2tqCgIM2fP19jx45VvXr11KJFC61cuVKxsbHq2LGj6tatq1dffVUnT560GTs2Nlbt27eXv7+/evbsqYSEBOu27OxsRUZG6vnnn1e9evXUtWtX/fDDD9bta9asUdu2bTV58mQ1bNhQY8aMybX+w4cPa9CgQWrcuLGaNGmiMWPG6OLFi5JuLAnYvn271q5dK19fX50+fTrXMUqXLq0xY8Zo5cqV2rlz5x3n056ab/31fFJSkoYMGaIGDRqoRYsW+vjjj9WmTRutWbPG2icjI0PTpk1T06ZN1aBBA40ZM0bXr1+3Oe7XX3+tgIAA1alTRyEhITp//rx1W0pKit5//30FBATI399f3bt3144dO6zb3377bQ0bNkxBQUFq2LChli1bpoSEBPXv318NGjRQw4YNFRoaese5kW4svVi+fLm6dOmiOnXqqEuXLvrll19s+qxcuVJt27ZVnTp1FBgYqC+++MK6bdeuXfL399eiRYvUpEkTBQUF5TjGhQsXtHXrVvXp00clS5a02ebq6qrZs2drwoQJ1rbDhw9rwIAB1iUoEydO1B9//HHHc8iv25erDBgwQH/729/UokULNW7cWCEhITp37py1/++//6633npLDRo00J///GcNHz7cZvvtVq9eLWdnZw0ePDjX7e+//76mT59ufe3r66t58+apZcuWatmypS5cuKCDBw9q0KBBatSokfz8/NS2bVutXbvWuk9QUJBmzZqlIUOGqE6dOgoICNA//vGPHMdauXKl9f3Wu3dvHT9+PI+zBcBehG6gmDCZTAoPD9eZM2fyfZXso48+kq+vr9avX6/WrVvrvffe05QpUzRhwgQtXbpU586dy7GU5dNPP9WIESO0Zs0aVahQQUFBQdZAOXv2bK1Zs0bvvfee1q1bp86dO+vNN9/Url27rPsfP35cV69e1dq1axUcHJyjptOnT+vVV19VmTJlFBMTo0WLFungwYPq37+/srKytHr1ajVq1Ejt2rXTjz/+qMcee+yO59ejRw8988wzmjBhQq7r4O2t+abs7GwFBwfr3Llz+vzzz7VgwQKtX79ep06dsukXGxurzMxMrVixQnPmzNFXX32lTz75xKbP4sWLNW3aNK1YsUJXrlzRgAEDZLFYJEnDhw/X5s2bNWXKFK1du1Z169bVwIEDtXfvXuv+mzdvVps2bbRy5Uq1adNGo0aN0uOPP64vvvhCMTExunTpkt555507zo0kzZo1S6+88oq++OIL+fn5acCAAdZzWbZsmebOnavhw4drw4YNGjhwoKZNm2YTvNPT07Vr1y6tWrXKJjzfdODAAWVlZalu3bq5Ht/Hx0dVqlSRJJ07d05BQUF66qmn9MUXX2j+/Pk6evSo3nzzzbueQ0HatWuXDh06pE8//VRz587Vnj17NH/+fEnS9evXFRQUJDc3Ny1fvlzR0dHKyMhQnz59lJ6enut4sbGxql+/vtzc3HLdXrFiRf3pT3+yaVu1apWioqK0cOFC/elPf1L//v1VsWJFrVy5UuvWrVPjxo01YcIEJSUlWfdZvHixKlWqpC+++MJ6VX3Dhg3W7VlZWfryyy+1YMEC/eMf/9D//vc/TZw4Mb/TBeAOXBxdAICCU6VKFQ0ZMkRz5szRiy++KD8/v/sap3bt2urfv78kqXfv3lq+fLn69u2rJk2aSJLatWun77//3mafYcOG6YUXXpAkhYeHq2XLltq4caPat2+vxYsXa8GCBXr22WclSd7e3jp48KD+/ve/q2nTptYxQkNDVbly5VxrWrZsmUqXLq3p06fL1dVVkjR37ly1b99e//73v9WqVSu5urrKbDbL09PznucYFhamwMBAzZkzJ8eNpteuXbO7Zkn6+eefFR8fr2+++cZaf0REhAIDA236VapUSRMnTpTJZFKVKlXUvHlzxcfH2/SZNGmSnnnmGUnSzJkz1bp1a+3YsUMVK1bUd999p+joaLVo0UKSNGHCBMXFxSk6OtoaAj09PfXaa69Zxztx4oSaN2+uJ554Qi4uLoqIiLAJZrnp0aOHevToYa3np59+0sqVKzVy5EhFRkbqzTff1IsvvihJ8vLy0pkzZxQZGanOnTtbxxg4cKC8vb1zHf/mVerSpUvftQ7pxs/9ySef1NixY61tc+fOVcuWLbVnzx7Vr1//nmPkl8ViUXh4uEqWLCkfHx917NhR27dvlyRt3LhRKSkpmjFjhpydnSVJc+bMUdOmTfWvf/1LHTp0yDFeUlJSjv82P/rooxw3b+7Zs8f6986dO6tWrVqSpP/973/q27evgoKC5O7uLkkKDg7WqlWrdPz4ceuyp6eeesr63q5evbr27t2rJUuW2NQ0depU6wecl19+2fo+AlDwCN1AMdOvXz99/fXXGjdunM3Shry4NSzd/J+6l5eXtc1sNue4indr+ClZsqSqVaumw4cPy9fXV+np6Ro6dKicnP7vl2sZGRk2a6JNJpOefPLJO9Z05MgR+fv7WwO3dCNIlCtXTocPH1arVq3ydI5PPvmkRowYofDwcLVr185m27Fjx+yq+ab9+/erfPnyNh8YnnrqqRyh0svLSyaTyfq6TJkyOZYhNGjQwKbGRx55RIcPH9bly5dzbJekhg0b2nwAun0Ohw4dqpkzZ2rZsmVq1qyZWrVqlePDwO0aN25s/buzs7P8/Px0+PBhXbx4UefOndPMmTP1/vvvW/tkZmYqKyvL5j1xpw9Pkqzr7ZOTk+9ah3TjqviBAwdyDdfHjh27a+ieOHGi1q9fb309ZcoUdezY8Z7HvF2FChVslsGULl1aGRkZkm787C9evKhGjRrZ7JOSknLHp5CUK1cux7n36NFDf/nLXyRJ3377bY417bfOZ/ny5dWzZ0+tXbtWBw4c0PHjx3Xw4EFJN65e33Trz1GS6tatqy1btlhfm0wmm//WS5cufcelaQDyj9ANFDPOzs4KDw9X586dFRkZec/+t/5P+iYXl5z/NNwaFu903FtlZ2erRIkSKlGihKQba2Nvv/J5a6B1cnKy9s3NnX4Vn52dbRPE86J379766quvNH78eJurw/bWfJOzs7Oys7NztN9cFnJrv3v1uX38m/N48/xv75+dnW3z8zKbzTbbX3vtNbVv317fffedtm/frunTp2vZsmVasWLFHef79p9/dna2TCaTdZ7fffdd62897rTf7XXcys/PTy4uLvr1119Vp06dHNvXr19vDZ6urq5q3rx5rstUHnnkkTseQ7rxgWPAgAHW1+XLl79r/zvJbZ5u/hxcXV1Vo0YNLVy4MEefUqVK5Tpe/fr1tXr1amVkZFjntEyZMipTpswd67z1/X/+/Hm9/PLLevTRR/X888+rVatWqlixorp27Wqzz51+jjc5OTnl+O/69vcXgILDmm6gGPLx8dEbb7yhqKgomxseb/4P/tq1a9a2grpxav/+/da/X758WYmJifLx8ZG3t7dcXV117tw5eXt7W/+sX78+T1fia9Sood9++816hVGSjh49quTkZFWvXv2+ajaZTJo2bZrOnDmj6Ohoa3tea/b19dWlS5ds5johIUFXrlzJc023zmNCQoIuX74sHx8f+fj4SFKOp9Ps3r1bNWrUyHWsS5cuKSwsTJmZmerevbvmzp2rzz77TPv377deGc3Nvn37rH/PzMzUvn379PTTT6tUqVJ69NFHdfr0aZt52b59u6Kjo3P9QJKbMmXKqE2bNvr8889t3ouSlJaWpo8++kiXL1+Wm5ubatSooWPHjunxxx+3Hs/JyUnh4eH6/fff73qc8uXL29R5+02bBcHHx0enT59W2bJlrccpX768pk+frsOHD+e6z8svv6y0tDRFRUXluv1uN2FK0pYtW3Tt2jXFxMQoODhYAQEBunTpkiTb0Hzrz1GSfv31V57NDTgQoRsopoKDg1WjRg2bL1upWLGinnjiCX322WdKSEhQbGysPvjgg3texbZHRESEfvjhBx06dEijRo1ShQoV1L59e7m7u6tv376aPXu2Nm3apFOnTmnx4sX629/+dtclCLfr3bu3rly5onHjxunIkSOKjY3VqFGjVLNmTesa6Ptxcx38rTc95rXmZs2ayd/fX2PGjFF8fLzi4uKsT2DJ69y+++67io2NVVxcnEaPHq1GjRqpcePG8vLy0ksvvaTJkyfrxx9/1LFjxzR9+nTt27fP5ir9rcqUKaNt27Zp4sSJOnjwoE6cOKE1a9aodOnSqlq16h1r+OSTT7R582YdO3ZMEydOVHJysl5++WVJ0htvvKHPPvtMK1as0MmTJ7V+/XrNmDHDrnX0t3r77bdlsVjUq1cvbd26VadOndLOnTs1cOBAnTt3znpDX+/evfXHH3/o7bff1qFDh/Tbb79pxIgROn78uHUt8v04dOiQtm3bZvPnbh9E7iQwMFDlypXTsGHD9Ntvv+nw4cMaOXKk9u7da/2gdLsqVaooLCxMkZGRGjNmjH755Rf997//1S+//KK3337buib8TsqVK6erV6/q66+/1n//+199++23mjRpkiTZLPHZsWOHIiMjlZiYqMWLF2vz5s3WezUAFD6WlwDFlIuLi8LDw9W9e3drm8lk0qxZsxQeHq6OHTvK29tb48aNu+sXydgrNDRU06ZN0++//67GjRvr448/tv5aftiwYXJ1ddWsWbOUlJSkypUr67333lOXLl3sHr9ChQr65JNPFBERoa5du8rd3V0BAQEaPXr0fS8vuenmOvhbrwzmteYFCxZoypQp6tWrl0qVKqXXX39d8fHxea5t0KBBGjFihP744w+1bNnSGqakGzd/RkREaPTo0bp+/bpq1aql6OjoO65rdnJyUlRUlGbMmKGgoCClp6fL399f0dHRd1z6IN1YX3wzrNWpU0eLFy/Wo48+Kkl69dVXlZ6erujoaIWFhenRRx9VaGhont9DlSpV0ooVKxQVFaXw8HCdP39ejzzyiJo0aaJp06ZZ7yHw9PTUp59+qvfff189evSQ2WxW06ZNNW/evLsuR7qX6Ohom99uSDcC9K1r1e1hNpv16aefasaMGerTp49MJpPq1aunzz///K7LWTp16iRfX199/vnnGjNmjC5cuKBSpUqpbt26WrhwoVq3bn3Hfdu1a6fffvtNU6dO1fXr1+Xl5aXQ0FD9/e9/12+//aaWLVtKkv7yl78oLi5OixYt0hNPPKGIiAgFBATk6fwAFByThQVcAJAvFy9eVFxcnJ599lnruu0LFy6oRYsWiomJyXGTXVHm6+urWbNmqVOnTo4uBfkQFBQkLy8vTZs2zdGlAPj/uNINAPnk7OysoUOHqm/fvurWrZuuXbumefPmydvb+47PogYAPFxY0w0A+VSmTBlFRkZq586dCgwMVFBQkFxcXPTJJ5/ke+kLAKB4YHkJAAAAYDCudAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAZzcXQBRkpNTVV8fLw8PT3l7Ozs6HIAAABQTGVlZenChQvy8/OT2WzOsb1Yh+74+Hj16tXL0WUAAADgIRETE6NGjRrlaC/WodvT01PSjZOvVKmSg6sBAABAcXX27Fn16tXLmj9vV6xD980lJZUqVdKTTz7p4GoAAABQ3N1pSTM3UgIAAAAGK9ZXuouSjz76SAkJCXne79KlS5KkcuXKFXRJBaJatWoaNGiQo8sAAAAo0gjdhSQhIUHx+w/J2Vw2T/tlpV6WJJ29lF7wReXTzdoAAABwd4TuQuRsLisP79Z52uf6iW8lKc/7FYabtQEAAODuWNMNAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0G2QrVu3auvWrY4uA0UY7xEAAB4ePDLQIFu2bJEkBQQEOLgSFFW8RwAAeHhwpRsAAAAwGKEbAAAAMBihGwAAADAYoRsAAAAwGKEbAAAAMBihGwAAADAYoRsoIsaOHavAwECNGzdOkrRmzRoFBgZq7dq11j5TpkxRYGCgwsLCJEmJiYl67bXXdPz4cWufH3/8UR07dtT27dtzfZ3b2Lf3ye3Yt7NnXHvGuf0cchv3XnLb5/a23Obqftgzv7fL7di3txVUfQUht3O6V31Fqf7CZNR5P2jjAkVJUX2fmywWi8XRRRjl9OnTat26tb799ls9+eSThXrsN954QxcvXlS1atUkSQkJCUrJdFbJ6i/laZzrJ76VJHl4ty7wGvPr6rGNcnfJsp4j8iYhIUGPPPKIPvzwQ6Wlpalbt27WbStWrNDLL79sfb127VplZWWpa9eu1rbVq1drzJgxSkhIUPXq1TV37lxZLBZ1795d6enpKlGihHWcm69XrVoli8Wiv/71r9Zx1qxZo1deecXaZ/ny5erSpYvNsZ2dnW1qz87OtjlObuP+85//tKk3t3EsFouGDRtmPYfZs2erR48eNuM6Od392kButUiyaVu5cqVGjBhhM1cmk+mu49pzrNzm9/Z6bz/HuXPnSpJN25w5czR8+PB811cQcptPk8mU4xxurS+3c3RU/YXJqPN+0MYFihJHvs/vlTu50g0UAVOmTLF5PWzYMJvXmzZt0qxZs2zawsLCdOLECUnS8ePHdfLkSe3atUvp6emSpPT0dC1dutTm9c8//6wNGzbYjLNo0SKbPosWLcpx7Nvdfpzcxn3//ffvOc7JkydtzmHjxo05xr2X3Gq5vW3jxo055up+2DO/9zrHkydP5mj7+eefC6S+gpDbfOZ2Dre61/biyqjzftDGBYqSovw+5xspDVK2bFmVLVtW06dPlySNGzdOBxLOObiqguXkYla1ao9azxF5c3MZiSRNmjTJ5kr3Bx98YHOlu3379mrbtq3NleN3331XY8eO1bFjx1SlShV5eXmpcuXKKlGihPUqZe/evbV+/Xrr6yZNmshisejjjz+2jhMaGqpt27ZZ+4SGhuqbb76xOfbtmjZtanOc3MYdNWqUTb25jePl5aUqVapYz+Gll17S559/bjPuveRWiySbtpdeeklbt261mav7cfuxcpvfe53jzWPf2takSZNc+zhCbvNpMpnuWt+dzrG4M+q8H7RxgaKkKL/PudINFAFubm7y8/OTJPn5+cnDw0MDBgyQJA0YMEDOzs42oa5JkyZyc3PTsGHDVK5cOQ0bNkwmk0lOTk4aOXKkTCaTRo4cKRcXF5vXTk5OcnZ2thnb1dXVpo+rq2uOY9/u9uPkNm6JEiXuOc7NZQs3z8HZ2TnHuPeSWy23tzk7O+eYq/thz/ze6xxNJlOONicnpwKpryDkNp+5ncOt7rW9uDLqvB+0cYGipCi/z1nTbZCbVzFvv9Kd17XZRXlN9/UT36oWV7rv2+3vEQAA8OBiTTcAAADgYIRuAAAAwGCEbgAAAMBghG4AAADAYIRuAAAAwGA8p9sgbdq0cXQJKOJ4jwAA8PAgdBskICDA0SWgiOM9AgDAw4PlJQAAAIDBCN0AAACAwQjdAAAAgMEI3QAAAIDBCN0AAACAwQjdAAAAgMF4ZGAhykq9rOsnvs3zPpLyvF9huFHbo44uAwAAoMgjdBeSatWq3dd+ly6VkCSVK1euIMspII/e93kBAAA8TAjdhWTQoEGOLgEAAAAOwppuAAAAwGCEbgAAAMBgxXp5SVZWliTp7NmzDq4EAAAAxdnNvHkzf96uWIfuCxcuSJJ69erl4EoAAADwMLhw4YK8vb1ztJssFovFAfUUitTUVMXHx8vT01POzs6OLgcAAADFVFZWli5cuCA/Pz+ZzeYc24t16AYAAACKAm6kBAAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAzm4ugCjJSamqr4+Hh5enrK2dnZ0eUAAACgmMrKytKFCxfk5+cns9mcY3uxDt3x8fHq1auXo8sAAADAQyImJkaNGjXK0V6sQ7enp6ekGydfqVIlB1cDAACA4urs2bPq1auXNX/erliH7ptLSipVqqQnn3zSwdUAAACguLvTkmZupAQAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6DZQRma2of0BAADwYCjWXwPvaK4uTnpn0U929w8PbW5gNQAAAHAUrnQDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYr1NAdGxur7t27q2HDhnrhhRe0fPlySVJcXJxq1aql+vXrW/9ERkZKkiwWi2bPnq1mzZqpcePGmjp1qrKysgqzbAAAACBfXArrQMnJyQoNDdWECRPUoUMHHThwQP369ZOXl5dOnz6tli1bKioqKsd+MTEx+v777/Xll1/KZDIpODhYy5YtU1BQUGGVDgAAAORLoV3pPnPmjJ577jl17NhRTk5Oql27tpo2bardu3dr//79qlmzZq77rVu3Tn369FHFihXl6emp4OBgrVy5srDKBgAAAPKt0EJ3rVq1FBERYX2dnJys2NhY1axZUwcOHNDu3bsVEBCgVq1aaebMmUpPT5ckJSQkqEaNGtb9qlatqqNHj8pisRRW6QAAAEC+OORGyitXrigkJES1a9dWQECAypUrp4CAAG3YsEFLlizRrl27NH/+fElSSkqKzGazdV93d3dlZ2dbQzkAAABQ1BV66D516pReeeUVlSlTRgsXLpSTk5MiIyPVr18/eXh4qHLlygoODtaWLVskSWazWWlpadb9U1JS5OLiIjc3t8IuHQAAALgvhRq69+3bpx49eqhFixZatGiRzGazkpOTNXPmTF29etXaLy0tzRqqq1evrsTEROu2xMREVatWrTDLBgAAAPKl0J5ekpSUpIEDB6pfv356/fXXre2lSpXSli1bZLFYNHLkSJ05c0aRkZHq0aOHJKljx46Kjo5Ws2bN5OLioqioKHXq1KmwygYAAADyrdBC9+rVq3Xx4kV9+OGH+vDDD63tr732miIjIzV16lQ1a9ZMZrNZL7/8svr06SNJ6tmzp5KSktStWzdlZGQoMDBQ/fr1K6yyAQAAgHwrtNAdEhKikJCQO27/7LPPcm13dnbW8OHDNXz4cIMqAwAAAIzF18ADAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGK9TQHRsbq+7du6thw4Z64YUXtHz5cklScnKyBg8erIYNG6pVq1ZatWqVdR+LxaLZs2erWbNmaty4saZOnaqsrKzCLBsAAADIF5fCOlBycrJCQ0M1YcIEdejQQQcOHFC/fv3k5eWl5cuXy8PDQ9u3b9ehQ4c0aNAg+fv7q2bNmoqJidH333+vL7/8UiaTScHBwVq2bJmCgoIKq3QAAAAgXwrtSveZM2f03HPPqWPHjnJyclLt2rXVtGlT7d69W998843eeustubm5qU6dOurQoYP1ave6devUp08fVaxYUZ6engoODtbKlSsLq2wAAAAg3wotdNeqVUsRERHW18nJyYqNjZUkubi4qHLlytZtVatW1ZEjRyRJCQkJqlGjhs22o0ePymKxFFLlAAAAQP445EbKK1euKCQkxHq122w222w3m81KTU2VJKWkpNhsd3d3V3Z2ttLT0wu1ZgAAAOB+FXroPnXqlF555RWVKVNGCxculIeHhzVg35SamioPDw9JNwJ4WlqadVtKSopcXFzk5uZWqHUDAAAA96tQQ/e+ffvUo0cPtWjRQosWLZLZbJa3t7cyMzN15swZa7/ExETrkpLq1asrMTHRZlu1atUKs2wAAAAgXwotdCclJWngwIHq16+fxo0bJyenG4cuWbKkWrdurdmzZyslJUVxcXHasGGDAgMDJUkdO3ZUdHS0zp49q6SkJEVFRalTp06FVTYAAACQb4X2yMDVq1fr4sWL+vDDD/Xhhx9a21977TWFhYVp0qRJeu655+Th4aHRo0erbt26kqSePXsqKSlJ3bp1U0ZGhgIDA9WvX7/CKhsAAADIt0IL3SEhIQoJCbnj9nnz5uXa7uzsrOHDh2v48OFGlQYAAAAYiq+BBwAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxmd+geMmSItm7dqqysLCPrAQAAAIodF3s7ms1mjRw5Uu7u7urQoYO6dOmimjVrGlkbAAAAUCzYHbojIiKUkpKif/3rX/ryyy/VtWtX1ahRQ126dFFgYKAeeeQRI+sEAAAAHlh5WtPt7u6uTp06KTo6Wj/88INefPFFzZ07Vy1btlRoaKh27NhhVJ0AAADAA8vuK903JScna/PmzdqwYYN2794tHx8f/fWvf9X58+c1ePBgBQUFafjw4UbUCgAAADyQ7A7dX3/9tb788ktt27ZNJUuWVIcOHTR+/HjVqlXL2qdKlSqaMWMGoRsAAAC4hd2he+TIkWrZsqXmzp2rVq1aycUl567Vq1fXK6+8UqAFAgAAAA86u0P3tm3bVKZMGV2+fNkauHfv3i1/f3+5urpKkho1aqRGjRoZUykAAADwgLL7Rsrk5GT95S9/0ccff2xte/PNNxUYGKhTp04ZUhwAAABQHNgduqdOnaratWsrODjY2vavf/1LPj4+Cg8PN6Q4AAAAoDiwO3Tv2bNHI0aMUNmyZa1tJUuW1LBhwxQbG2tEbQAAAECxYHfodnd31/nz53O0X7p0SU5OeXrcNwAAAPBQsTst/+Uvf9HkyZMVGxurtLQ0paWlKTY2VlOmTNELL7xgZI0AAADAA83up5eMGjVKQ4cOVe/evWUymaztAQEBGjdunCHFAQAAAMWB3aH7T3/6kz7++GMlJCTo8OHDcnV1VfXq1VWlShUDywMAAAAefHn+Gvhq1aqpWrVqRtQCAAAAFEt2h+6EhARNnTpVe/bsUUZGRo7t8fHxBVoYAAAAUFzYHbqnTJmiM2fOaNiwYSpdurSRNQEAAADFit2h+7ffftOnn36qunXrGlkPAAAAUOzY/cjAsmXLqkSJEkbWAgAAABRLdofukJAQhYeHKzExUdnZ2UbWBAAAABQrdi8v+fTTT3Xq1Cm1b99eJpMpx7dQciMlAAAAkDu7Q/frr79uZB0AAABAsWV36O7cubORdQAAAADFlt1ruiXpl19+0cCBAxUQEKD//ve/WrBggdauXZvng8bFxalFixY2r2vVqqX69etb/0RGRkqSLBaLZs+erWbNmqlx48aaOnWqsrKy8nxMAAAAwFHsDt0//PCDBg4cqMcee0xJSUnKzs6WyWTS+PHj9c9//tOuMSwWi1avXq3+/fvbfMHOwYMH1bJlS+3Zs8f6JyQkRJIUExOj77//Xl9++aU2bdqk3bt3a9myZXk8TQAAAMBx7A7dCxcu1JgxYxQWFiZnZ2dJ0ptvvqmxY8fqk08+sWuMyMhILV682Bqob9q/f79q1qyZ6z7r1q1Tnz59VLFiRXl6eio4OFgrV660t2wAAADA4ewO3UePHlXLli1ztD///PM6deqUXWN07dpV69atk7+/v037gQMHtHv3bgUEBKhVq1aaOXOm0tPTJd34+vkaNWpY+1atWlVHjx6VxWKxt3QAAADAoewO3eXKlcs1XMfHx6tChQp2jVGxYkWZTKZcxw4ICNCGDRu0ZMkS7dq1S/Pnz5ckpaSkyGw2W/u6u7srOzvbGsoBAACAos7u0N2jRw9NmTJFP/zwgyTp5MmTWr16tcLCwvL9ZJPIyEj169dPHh4eqly5soKDg7VlyxZJktlsVlpamrVvSkqKXFxc5Obmlq9jAgAAAIXF7kcGBgcH68qVKxoyZIjS09M1YMAAubi4qF+/fgoNDb3vApKTkxUZGanBgwerZMmSkqS0tDRrqK5evboSExNVt25dSVJiYqKqVat238cDAAAACpvdodtkMmn06NEaPHiwjh07JldXV1WpUsVm6cf9KFWqlLZs2SKLxaKRI0fqzJkzioyMVI8ePSRJHTt2VHR0tJo1ayYXFxdFRUWpU6dO+TomAAAAUJjsDt27d++2eZ2RkaH9+/dbXzdo0OC+CnByclJkZKSmTp2qZs2ayWw26+WXX1afPn0kST179lRSUpK6deumjIwMBQYGql+/fvd1LAAAAMAR7A7dPXv2lMlksnlqiMlkkslkkpOTk+Lj4+0+aNOmTbVr1y7r6xo1auizzz7Lta+zs7OGDx+u4cOH2z0+AAAAUJTYHbq//fZbm9dZWVlKTEzUvHnzNGrUqAIvDAAAACgu7A7dTzzxRI42Ly8v/elPf9KUKVO0fv36Ai0MAAAAKC7sfmTgnZQvX14nTpwoiFoAAACAYum+b6SUpKtXr+rzzz+Xj49PgRYFAAAAFCf5upFSurHsJCIiosALAwAAAIqL+76RUpJcXV1VsWLFAi0IAAAAKG7ydSMlAAAAgHuzO3QHBQXJZDLZ1Xfx4sX3XRAAAABQ3Nj99JIGDRro119/VWpqqvz8/FS/fn25urrq559/VpkyZeTl5WX9AwAAAOD/2H2l+8SJE+rVq5fGjh1r0x4VFaVff/1VU6dOLfDiAAAAgOLA7ivdP/zwg3r06JGjvW3bttq+fXuBFgUAAAAUJ3aH7vLlyys2NjZH+7Zt21SpUqUCLQoAAAAoTuxeXtK/f39NmTJFv/76q/z8/GSxWLRnzx5t3rxZM2fONLJGAAAA4IGWpy/H8fDwUExMjDZt2iSz2aynnnpK0dHRatq0qZE1AgAAAA80u0O3JP31r3/VX//6V4NKAQAAAIonu9d0SzeeYDJp0iQFBQXp3LlziomJ0c6dO42qDQAAACgW7A7de/fuVadOnXTq1Cnt2bNH6enpOnr0qAYMGKDvvvvOyBoBAACAB5rdofv999/XoEGD9Mknn8jV1VWSNGnSJA0cOFALFiwwrEAAAADgQWd36N6/f786dOiQo71bt25KSEgo0KIAAACA4sTu0O3u7q7//e9/OdoTExNVsmTJAi0KAAAAKE7sDt0dOnTQ9OnTdezYMZlMJqWlpWnHjh0KCwvTiy++aGSNAAAAwAPN7tA9YsQIPfHEE3rppZd0/fp1BQYGqn///vL399fIkSONrBEAAAB4oNn9nO5jx45p7ty5GjFihPbv3y9XV1f5+PjIy8vLyPoAAACAB57doXvAgAGKjIxUnTp1CNoAAABAHti9vKR06dJKT083shYAAACgWLL7Svfzzz+vQYMGKSAgQJUrV5bZbLbZHhISUuDFAQAAAMWB3aH766+/Vrly5bRnzx7t2bPHZpvJZCJ0AwAAAHdw19AdExOjbt26yc3NTVu3bi2smgAAAIBi5a5ruqdOnaqrV6/atL377ru6ePGioUUBAAAAxcldQ7fFYsnRtnHjRl27ds2wggAAAIDixu6nl9yUWxAHAAAAcGd5Dt0wTkZmtqH9AQAA4Bj3fHqJyWQqjDogydXFSe8s+snu/uGhzQ2sBgAAAAXlnqF7+vTpNs/kzsjI0Jw5c1SyZEmbfmFhYXYfNC4uTqGhofrxxx8lScnJyXrnnXe0c+dOlSpVSoMHD1b37t0l3VjOMmfOHK1atUpZWVnq1KmTxo0bJ2dnZ7uPBwAAADjSXUN348aNdfbsWZu2+vXrKykpSUlJSdY2e6+GWywW/fOf/9SMGTNsQvO7774rDw8Pbd++XYcOHdKgQYPk7++vmjVrKiYmRt9//72+/PJLmUwmBQcHa9myZQoKCsrLeQIAAAAOc9fQvWTJkgI9WGRkpDZv3qyQkBB99NFHkqRr167pm2++0ddffy03NzfVqVNHHTp00KpVq/Tuu+9q3bp16tOnjypWrChJCg4O1sKFCwndAAAAeGAU6o2UXbt21bp16+Tv729tO3HihFxcXFS5cmVrW9WqVXXkyBFJUkJCgmrUqGGz7ejRozxFBQAAAA+MQg3dFStWzLEU5fr16zZrxiXJbDYrNTVVkpSSkmKz3d3dXdnZ2UpPTze+YAAAAKAAOPyRge7u7taAfVNqaqo8PDwk3QjgaWlp1m0pKSlycXGRm5tbodYJAAAA3C+Hh25vb29lZmbqzJkz1rbExETrkpLq1asrMTHRZlu1atUKvU4AAADgfjk8dJcsWVKtW7fW7NmzlZKSori4OG3YsEGBgYGSpI4dOyo6Olpnz55VUlKSoqKi1KlTJwdXDQAAANjvns/pLgxhYWGaNGmSnnvuOXl4eGj06NGqW7euJKlnz55KSkpSt27dlJGRocDAQPXr18/BFQMAAAD2c0jobtq0qXbt2mV9XbZsWc2bNy/Xvs7Ozho+fLiGDx9eWOUBAAAABcrhy0sAAACA4o7QDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjNANAAAAGIzQDQAAABiM0A0AAAAYjND9AMvIzDa0PwAAAAqGi6MLwP1zdXHSO4t+srt/eGhzA6sBAADAnXClGwAAADAYoRsAAAAwGKEbAAAAMBihGwAAADAYoRsAAAAwWJEJ3R9//LH8/PxUv35965/Y2FglJydr8ODBatiwoVq1aqVVq1Y5ulQAAAAgT4rMIwMPHDig4cOHa8CAATbtb731ljw8PLR9+3YdOnRIgwYNkr+/v2rWrOmgSgEAAIC8KTJXug8cOKBatWrZtF27dk3ffPON3nrrLbm5ualOnTrq0KEDV7sBAADwQCkSoTslJUXHjx/X4sWL1bx5c7Vr106rV6/WiRMn5OLiosqVK1v7Vq1aVUeOHHFgtQAAAEDeFInlJUlJSWrQoIFeffVVzZ8/X3FxcQoJCVG/fv1kNptt+prNZqWmpjqoUgAAACDvikTorly5spYuXWp93ahRI3Xq1EmxsbE5AnZqaqo8PDwKu8SHUkZmtlxd7P9lSF77AwAAPCyKROjet2+ffvrpJ73++uvWtrS0ND322GPKzMzUmTNn9Pjjj0uSEhMTVaNGDUeV+lBxdXHSO4t+srt/eGhzA6sBAAB4cBWJy5IeHh5auHChvvrqK2VnZ2vHjh3auHGjevXqpdatW2v27NlKSUlRXFycNmzYoMDAQEeXDAAAANitSFzprlq1qj744APNnTtXb7/9th599FFNnz5dtWvXVlhYmCZNmqTnnntOHh4eGj16tOrWrevokgEAAAC7FYnQLUkBAQEKCAjI0V62bFnNmzfPARUBAAAABaNILC8BAAAAijNCNwAAAGAwQjcAAABgMEI3AAAAYDBCNwAAAGAwQjcAAABgMEI3AAAAYDBCNwAAAGAwQvdDJCMz29ElAAAAPJSKzDdSwniuLk56Z9FPdvcPD21uYDUAAAAPD650AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdAMAAAAGI3QDAAAABiN0AwAAAAYjdKPA5PXLd/iyHgAA8LDgy3FQYIz88p2MzGy5utj/GTGv/QEAAIxE6MYDgW/TBAAADzIuBQIAAAAGI3QDAAAABiN0AwAAAAYjdMNhjHx6CU9SAQAARQk3UsJh8nJzZF5vjOTGSwAAUJRwpRsAAAAwGKEbEMtRAACAsVheAijvy1GmvP5MnsbPyMySq4tzHvrz5T4AABQnhG7gPtzPmnHWmAMA8PDiUhpQBLHcBQCA4oUr3UARZPTTV/K6fIXlLgAA5A+hG3gI8UhFAAAKF5euAAAAAIM9EKF7//796tatm+rVq6dOnTrp119/dXRJwEOFNeYAAORPkV9ekpaWppCQEIWEhKh79+5at26d3nzzTW3dulUlSpRwdHlAkWD0muuithyFNekAgAdNkQ/dO3fulJOTk3r27ClJ6tatmz7//HN99913atu2rYOrA4qGBz8U5+055g/y+Ra1DwB8gAGAwlHkQ3diYqKqV69u01a1alUdOXLknqE7KytLknT27FnD6ruXa38k2d339OnT9H8AaqH/vZ07e0YRS/9jd//RvRvmuX9e6jl+4pRcnE1298/MypaLc96Cpb31D3+1QR5rseSp//3Iy9wXxfqNlNf6H7b+QGF5EN7LN/Pmzfx5O5PFYrEUZkF5tWjRIu3fv18LFy60to0ZM0YVK1bUqFGj7rpvbGysevXqZXSJAAAAgCQpJiZGjRo1ytFe5K90u7u7KzU11aYtNTVVHh4e99zXz89PMTEx8vT0lLOz/b+6BgAAAPIiKytLFy5ckJ+fX67bi3zorlatmpYuXWrTlpiYqA4dOtxzX7PZnOsnDQAAAKCgeXt733Fbkb8b5plnnlF6erqWLFmijIwMrV69WklJSWrRooWjSwMAAADsUuTXdEvSwYMHNXnyZB06dEje3t6aPHmy6tWr5+iyAAAAALs8EKEbAAAAeJAV+eUlAAAAwIOO0A0AAAAYjNANAAAAGIzQDQAAABiM0J1P+/fvV7du3VSvXj116tRJv/76q6NLeuDFxsaqe/fuatiwoV544QUtX75ckpScnKzBgwerYcOGatWqlVatWuXgSh9sSUlJeuaZZ/Tdd99JYn4LytmzZxUcHKwGDRqoZcuWWrx4sSTmtyDt3r1bXbp0UYMGDdS2bVutX79eEnOcX3FxcTaP473bfFosFs2ePVvNmjVT48aNNXXq1Dt+9TVuuH1+z549q9DQUDVt2lTNmzdXWFiY0tPTJTG/9+v2Ob4pOztbQUFBmjlzprXNIXNswX1LTU21PPvss5aYmBhLenq6ZdWqVZbmzZtb0tLSHF3aA+vy5cuWxo0bW9atW2fJysqyxMfHWxo3bmz56aefLEOGDLGMGjXKkpqaatm7d6+lSZMmlgMHDji65AfW66+/bqlZs6Zl69atFovFwvwWgOzsbEvnzp0tM2bMsKSnp1sOHz5sady4seU///kP81tAMjMzLc2aNbNs3rzZYrFYLL/88ovl6aeftpw6dYo5vk/Z2dmWVatWWRo2bGhp0qSJtf1u87lkyRJLhw4dLOfOnbOcP3/e0rlzZ8vixYsddQpF2p3mt3fv3pYpU6ZYUlNTLefPn7d0797dMmfOHIvFwvzm1Z3m+KaPPvrIUrNmTcuMGTOsbY6YY65058POnTvl5OSknj17ytXVVd26dVO5cuWsVw6Rd2fOnNFzzz2njh07ysnJSbVr11bTpk21e/duffPNN3rrrbfk5uamOnXqqEOHDlzJuk//+Mc/5O7urscee0ySdO3aNea3AOzdu1fnz5/XqFGj5OrqKh8fHy1fvlyPPvoo81tA/vjjD128eFFZWVmyWCwymUxydXWVs7Mzc3yfIiMjtXjxYoWEhFjb7vVvwrp169SnTx9VrFhRnp6eCg4O1sqVKx11CkVabvObnp4ud3d3vfHGG3Jzc5Onp6cCAwO1Z88eScxvXuU2xzcdPHhQa9asUZs2bWzaHTHHhO58SExMVPXq1W3aqlatqiNHjjioogdfrVq1FBERYX2dnJys2NhYSZKLi4sqV65s3cZc35/jx4/r008/1eTJk61tJ06cYH4LwL59++Tj46OIiAg1b95cbdu21d69e5WcnMz8FpBy5cqpZ8+eGjFihGrXrq1evXrp3Xff1aVLl5jj+9S1a1etW7dO/v7+1rZ7/ZuQkJCgGjVq2Gw7evSoLHz1Rw65zW+JEiX097//XZ6enta27777TjVr1pTE/OZVbnMs3fhwM3bsWL333nvy8PCw2eaIOSZ058P169fl7u5u02Y2m5WamuqgioqXK1euKCQkxHq122w222xnrvMuMzNTo0eP1vjx41W2bFlr+/Xr15nfApCcnKxdu3ZZf+M1ffp0hYWFMb8FKDs7W2azWfPmzdOvv/6qyMhIhYeH6+rVq8zxfapYsaJMJpNN273esykpKTbb3d3dlZ2dbV2TjP+T2/zeymKxaOrUqUpISFBwcLAk5jev7jTHs2fPVosWLdSoUaMc2xwxx4TufHB3d8/xD3pqamqOT1PIu1OnTumVV15RmTJltHDhQnl4eDDXBWDRokWqVauWnnvuOZt23ssFo0SJEipTpoyCg4NVokQJ641+8+fPZ34LyL/+9S/FxcXpxRdfVIkSJdSqVSu1atVKCxYsYI4L0L3+TTCbzUpLS7NuS0lJkYuLi9zc3Aq1zgddamqqhg4dqn//+99asmSJypcvL4n5LQg7duzQzp07NXTo0Fy3O2KOCd35UK1aNSUmJtq0JSYm2vy6Anm3b98+9ejRQy1atNCiRYtkNpvl7e2tzMxMnTlzxtqPuc67TZs2aePGjWrUqJEaNWqkM2fOaMSIEfr++++Z3wJQtWpVpaSkKDMz09qWlZWlp59+mvktIL///nuOK1EuLi6qXbs2c1yA7vVvbvXq1W3+/5eYmKhq1aoVep0PssuXL6t37966fPmyVqxYYbOUh/nNv02bNunkyZP685//rEaNGmnDhg1aunSp9bcJjphjQnc+PPPMM0pPT9eSJUuUkZGh1atXKykpKdfH1cA+SUlJGjhwoPr166dx48bJyenGW7RkyZJq3bq1Zs+erZSUFMXFxWnDhg0KDAx0cMUPlq+++kr/+c9/FBsbq9jYWD3++OOaM2eOBg8ezPwWgObNm6t06dKaPXu2MjMztXv3bm3ZskUvvvgi81tA/vznP+vAgQP65z//KYvFop9//llbtmzRSy+9xBwXoHv9m9uxY0dFR0fr7NmzSkpKUlRUlDp16uTgqh8cFotFQ4YMUYUKFRQdHW2z3E9ifgtCWFiY9uzZY/3/XYcOHdS7d29FRUVJcswcuxg6ejFXokQJffTRR5o8ebLmzJkjb29vffjhh/w6Mx9Wr16tixcv6sMPP9SHH35obX/ttdcUFhamSZMm6bnnnpOHh4dGjx6tunXrOrDa4oX5zT+z2awlS5bovffe05///GeVLFlSEyZMUL169ZjfAuLr66v58+dr3rx5mjZtmh5//HHNnDlT/v7+zHEBu9t89uzZU0lJSerWrZsyMjIUGBiofv36ObjiB8eePXv0888/y83NTU2aNLG2P/3004qJiWF+C4Ej5thk4VZYAAAAwFAsLwEAAAAMRugGAAAADEboBgAAAAxG6AYAAAAMRugGAAAADEboBgAAAAxG6AbwUAsICFCbNm2UkpKSY1tQUJDGjx9v2LFPnz4tX19fxcbGGnYMe+3bt0/t27eXn5+fZs6cmWP7ggUL5Ovrq23btuXYtmbNGj399NN2Hyuv/d9++2317dv3jtt37dolX19fnT171u4xAaCwEboBPPROnjypOXPmOLoMh4qKipKLi4s2bdqk119//Y79Jk6cqKtXr+brWO3bt881vANAcUboBvDQq1y5spYuXardu3c7uhSHuXLlimrVqiUvLy+VK1cu1z5ly5bV1atXNWvWrHwdy2w2q0KFCvkaAwAeNIRuAA+9zp07q379+ho/frzS0tJy7ZPbUpDb24KCgjR//nyNHTtW9erVU4sWLbRy5UrFxsaqY8eOqlu3rl599VWdPHnSZuzY2Fi1b99e/v7+6tmzpxISEqzbsrOzFRkZqeeff1716tVT165d9cMPP1i3r1mzRm3bttXkyZPVsGFDjRkzJtf6Dx8+rEGDBqlx48Zq0qSJxowZo4sXL0q6scRm+/btWrt2rXx9fXX69OlcxyhdurTGjBmjlStXaufOnXecT3tqvnV5SVJSkoYMGaIGDRqoRYsW+vjjj9WmTRutWbPG2icjI0PTpk1T06ZN1aBBA40ZM0bXr1+3Oe7XX3+tgIAA1alTRyEhITp//rx1W0pKit5//30FBATI399f3bt3144dO6zb3377bQ0bNkxBQUFq2LChli1bpoSEBPXv318NGjRQw4YNFRoaese5AYB7IXQDeOiZTCaFh4frzJkzWrBgQb7G+uijj+Tr66v169erdevWeu+99zRlyhRNmDBBS5cu1blz53IsZfn00081YsQIrVmzRhUqVFBQUJA1UM6ePVtr1qzRe++9p3Xr1qlz58568803tWvXLuv+x48f19WrV7V27VoFBwfnqOn06dN69dVXVaZMGcXExGjRokU6ePCg+vfvr6ysLK1evVqNGjVSu3bt9OOPP+qxxx674/n16NFDzzzzjCZMmJDrOnh7a74pOztbwcHBOnfunD7//HMtWLBA69ev16lTp2z6xcbGKjMzUytWrNCcOXP01Vdf6ZNPPrHps3jxYk2bNk0rVqzQlStXNGDAAFksFknS8OHDtXnzZk2ZMkVr165V3bp1NXDgQO3du9e6/+bNm9WmTRutXLlSbdq00ahRo/T444/riy++UExMjC5duqR33nnnjnMDAHfj4ugCAKAoqFKlioYMGaI5c+boxRdflJ+f332NU7t2bfXv31+S1Lt3by1fvlx9+/ZVkyZNJEnt2rXT999/b7PPsGHD9MILL0iSwsPD1bJlS23cuFHt27fX4sWLtWDBAj377LOSJG9vbx08eFB///vf1bRpU+sYoaGhqly5cq41LVu2TKVLl9b06dPl6uoqSZo7d67at2+vf//732rVqpVcXV1lNpvl6el5z3MMCwtTYGCg5syZk+NG02vXrtldsyT9/PPPio+P1zfffGOtPyIiQoGBgTb9KlWqpIkTJ8pkMqlKlSpq3ry54uPjbfpMmjRJzzzzjCRp5syZat26tXbs2KGKFSvqu+++U3R0tFq0aCFJmjBhguLi4hQdHa358+dLkjw9PfXaa69Zxztx4oSaN2+uJ554Qi4uLoqIiFBSUtI95wcAckPoBoD/r1+/fvr66681btw4m6UNeeHt7W39u7u7uyTJy8vL2mY2m5Wenm6zT/369a1/L1mypKpVq6bDhw/L19dX6enpGjp0qJyc/u8XkxkZGTZrok0mk5588sk71nTkyBH5+/tbA7ckVa9eXeXKldPhw4fVqlWrPJ3jk08+qREjRig8PFzt2rWz2Xbs2DG7ar5p//79Kl++vM0HhqeeekqlS5e26efl5SWTyWR9XaZMGZ07d86mT4MGDWxqfOSRR3T48GFdvnw5x3ZJatiwoc0HoNvncOjQoZo5c6aWLVumZs2aqVWrVjk+DACAvQjdAPD/OTs7Kzw8XJ07d1ZkZOQ9+2dlZeVoc3HJ+c/qrWHxTse9VXZ2tkqUKKESJUpIuvG4vlvDvCSbQOvk5GTtmxs3N7dc27Ozs22CeF707t1bX331lcaPH29zddjemm9ydnZWdnZ2jvaby0Ju7XevPrePf3Meb57/7f2zs7Ntfl5ms9lm+2uvvab27dvru+++0/bt2zV9+nQtW7ZMK1asuOt8A0BuWNMNALfw8fHRG2+8oaioKJsbHm+G02vXrlnbjh8/XiDH3L9/v/Xvly9fVmJionx8fOTt7S1XV1edO3dO3t7e1j/r16/P05X4GjVq6LffflNGRoa17ejRo0pOTlb16tXvq2aTyaRp06bpzJkzio6OtrbntWZfX19dunTJZq4TEhJ05cqVPNd06zwmJCTo8uXL8vHxkY+PjyTleDrN7t27VaNGjVzHunTpksLCwpSZmanu3btr7ty5+uyzz7R//34dPHgwz7UBAKEbAG4THBysGjVq2HzZSsWKFfXEE0/os88+U0JCgmJjY/XBBx/c8yq2PSIiIvTDDz/o0KFDGjVqlCpUqKD27dvL3d1dffv21ezZs7Vp0yadOnVKixcv1t/+9rc7rt/OTe/evXXlyhWNGzdOR44cUWxsrEaNGqWaNWta10Dfj5vr4G+96TGvNTdr1kz+/v4aM2aM4uPjFRcXZ30CS17n9t1331VsbKzi4uI0evRoNWrUSI0bN5aXl5deeuklTZ48WT/++KOOHTum6dOna9++fTZX6W9VpkwZbdu2TRMnTtTBgwd14sQJrVmzRqVLl1bVqlXzVBcASCwvAYAcXFxcFB4eru7du1vbTCaTZs2apfDwcHXs2FHe3t4aN27cXb9Ixl6hoaGaNm2afv/9dzVu3Fgff/yxdfnCsGHD5OrqqlmzZikpKUmVK1fWe++9py5dutg9foUKFfTJJ58oIiJCXbt2lbu7uwICAjR69Oj7Xl5y08118Pv27bO25bXmBQsWaMqUKerVq5dKlSql119/XfHx8XmubdCgQRoxYoT++OMPtWzZUpMmTbJuCwsLU0REhEaPHq3r16+rVq1aio6OtllPfysnJydFRUVpxowZCgoKUnp6uvz9/RUdHa1SpUrlqS4AkCST5fZFbgAAFJKLFy8qLi5Ozz77rHXd9oULF9SiRQvFxMSoUaNGDq4QAAoGV7oBAA7j7OysoUOHqm/fvurWrZuuXbumefPmydvbW3Xr1nV0eQBQYLjSDQBwqB07duiDDz7QoUOH5OrqqmbNmmns2LF3fQwiADxoCN0AAACAwXh6CQAAAGAwQjcAAABgMEI3AAAAYDBCNwAAAGAwQjcAAABgMEI3AAAAYLD/B9nTXLVIEQwaAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 864x432 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nr_neighbors mean  : 11\n",
      "nr_neighbors median: 5\n",
      "Thus, on average there are  11 neighbors per node.\n",
      "Note that there are 696 keys/nodes/genes in total.\n",
      "Thus, on average 1.61 % of all genes are neighbors.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nr_neighbors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>696.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>11.198276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>14.727776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>139.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       nr_neighbors\n",
       "count    696.000000\n",
       "mean      11.198276\n",
       "std       14.727776\n",
       "min        1.000000\n",
       "25%        2.000000\n",
       "50%        5.000000\n",
       "75%       14.000000\n",
       "max      139.000000"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many neighbors are there on average?\n",
    "len_of_neighbors = [len(neighbors) for _, neighbors in dict_as_indices.items()]\n",
    "assert len(len_of_neighbors) == len(NODES_AS_INDECES)\n",
    "f, (ax_box, ax_hist) = plt.subplots(2, sharex=True, gridspec_kw={\"height_ratios\": (.15, .85)}, figsize=(12, 6));\n",
    "\n",
    "sns.histplot(x=len_of_neighbors, kde=False, ax=ax_hist); #, element=\"step\"\n",
    "sns.boxplot(x=len_of_neighbors, ax=ax_box, fliersize=3);\n",
    "plt.suptitle(\"Number of Neighbors per Cell-Line Graph\", fontsize=PlottingParameters.TITLE_FONTSIZE.value);\n",
    "plt.xlabel(\"Number of Neighbors\", fontsize=PlottingParameters.XLABEL_FONTSIZE.value);\n",
    "plt.ylabel(\"Frequency\", fontsize=PlottingParameters.YLABEL_FONTSIZE.value);\n",
    "ax_hist.tick_params(axis='both', which='major', labelsize=PlottingParameters.XTICKS_LABELSIZE.value);\n",
    "\n",
    "ax_box.set(xlabel='');\n",
    "plt.show();\n",
    "# f.savefig(f'{IMGS_PATH}/NumberOfNeighborsPerCellLineGraph.eps', format='eps');\n",
    "# f.savefig(f'{PATH_SUMMARY_DATASETS}/{NOTEBOOKS_SUMMARY_FOLDER}/NumberOfNeighborsPerCellLineGraph.eps', format='eps');\n",
    "\n",
    "print(\"nr_neighbors mean  :\", round(np.mean(len_of_neighbors)))\n",
    "print(\"nr_neighbors median:\", round(np.median(len_of_neighbors)))\n",
    "print(f\"Thus, on average there are {np.mean(len_of_neighbors):3.0f} neighbors per node.\")\n",
    "print(f\"Note that there are {len(dict_as_indices.keys())} keys/nodes/genes in total.\")\n",
    "print(f\"Thus, on average {100*np.mean(len_of_neighbors)/len(dict_as_indices.keys()):2.2f} %\" +\n",
    "    \" of all genes are neighbors.\")\n",
    "\n",
    "pd.DataFrame(len_of_neighbors, columns=['nr_neighbors']).describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load node features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [],
   "source": [
    "gexpr = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_gexpr.pkl')\n",
    "cnvg = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_cnvg.pkl')\n",
    "cnvp = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_cnvp.pkl')\n",
    "mut = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_mut.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniqs(df: pd.DataFrame, col: str):\n",
    "    return np.unique(df[col].values).tolist()\n",
    "\n",
    "def get_intersecting_cell_lines(gexpr, cnvg, cnvp, mut):\n",
    "    # Test that all feature datasets contains exactly the same cell-lines.\n",
    "    gexpr_cls = get_uniqs(gexpr, 'CELL_LINE_NAME')\n",
    "    cnvg_cls = get_uniqs(cnvg, 'CELL_LINE_NAME')\n",
    "    cnvp_cls = get_uniqs(cnvp, 'CELL_LINE_NAME')\n",
    "    mut_cls = get_uniqs(mut, 'CELL_LINE_NAME')\n",
    "    inter_cls = set(gexpr_cls) \\\n",
    "        .intersection(set(cnvp_cls)) \\\n",
    "        .intersection(set(cnvp_cls)) \\\n",
    "        .intersection(set(mut_cls))\n",
    "\n",
    "    assert len(inter_cls) == len(gexpr_cls) == len(cnvg_cls) == len(cnvp_cls) == len(mut_cls), \\\n",
    "        \"Not all feature datasets contain the exact same cell-lines as rows!\"\n",
    "    del gexpr_cls, cnvg_cls, cnvp_cls, mut_cls\n",
    "\n",
    "    return inter_cls\n",
    "\n",
    "def get_intersecting_genes(gexpr, cnvg, cnvp, mut):\n",
    "    # Test that all feature datasets contains exactly the same gene symbols.\n",
    "    inter_genes = set(np.unique(gexpr.columns.values).tolist()) \\\n",
    "        .intersection(set(np.unique(cnvg.columns.values).tolist())) \\\n",
    "        .intersection(set(np.unique(cnvp.columns.values).tolist())) \\\n",
    "        .intersection(set(np.unique(mut.columns.values).tolist()))\n",
    "\n",
    "    assert len(inter_genes) == \\\n",
    "        len(np.unique(gexpr.columns.values).tolist()) == \\\n",
    "        len(np.unique(cnvg.columns.values).tolist()) == \\\n",
    "        len(np.unique(cnvp.columns.values).tolist()) == \\\n",
    "        len(np.unique(mut.columns.values).tolist()), \\\n",
    "            \"Not all feature datasets contain the exact same gene symbols as columns!\"\n",
    "\n",
    "    return inter_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "870\n",
      "859\n"
     ]
    }
   ],
   "source": [
    "inter_cls = get_intersecting_cell_lines(gexpr, cnvg, cnvp, mut)\n",
    "inter_genes = get_intersecting_genes(gexpr, cnvg, cnvp, mut)\n",
    "print(len(inter_cls))\n",
    "print(len(inter_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "gexpr.set_index('CELL_LINE_NAME', inplace=True)\n",
    "cnvg.set_index('CELL_LINE_NAME', inplace=True)\n",
    "cnvp.set_index('CELL_LINE_NAME', inplace=True)\n",
    "mut.set_index('CELL_LINE_NAME', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build gene-gene interaction graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 870/870 [00:06<00:00, 132.94it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell-line: 22RV1      Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 23132-87   Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 42-MG-BA   Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 451Lu      Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 5637       Graph: Data(x=[696, 4], edge_index=[2, 7794])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from typing import Tuple, Set, FrozenSet\n",
    "from tqdm import tqdm\n",
    "\n",
    "def create_cell_line_gene_graphs(gene_symbols: List[str], \n",
    "                                 gene_features: List[pd.DataFrame], \n",
    "                                 cls: Set[str],\n",
    "                                 neighbor_gene_tuples: List[Tuple[int]]):\n",
    "    \"\"\" \n",
    "    Creates pytorch geometric gene-gene interaction graphs for each of the\n",
    "    given cell-lines. The each graph has the exact same topology, meaning\n",
    "    the same gene symbols as nodes and the same edges. However, each graph\n",
    "    has different gene feature values per node.\n",
    "\n",
    "    Args: \n",
    "        gene_symbols (`List[str]`):\n",
    "            List of gene symbols which need to be available as columns in the \n",
    "            gene features (`gene_features`).\n",
    "        gene_features (`List[pd.DataFrame]`):\n",
    "            List of feature pd.DataFrame's. All dataframe need to have all the \n",
    "            given cell-lines (`cls`) as indices and gene symbols (`gene_symbols`) as \n",
    "            columns.\n",
    "        cls (`Set[str]`):\n",
    "            List of cell-lines for which to create graphs.\n",
    "        neighbor_gene_tuples (`List[Tuple[int]]`):\n",
    "            List of tuples of neighbors, where the neighbors are given as gene indices, \n",
    "            not symbols.\n",
    "    Returns:\n",
    "        `Dict[torch_geometric.data.Data]`:\n",
    "            Dictionary with the cell-line names as keys and pytorch geometric\n",
    "            gene-gene interaction graphs as values.\n",
    "    \"\"\"\n",
    "    Gs = {}\n",
    "    for cl in tqdm(cls):\n",
    "        # Convert the feature values to tensors and stack them up.\n",
    "        cl_features = []\n",
    "        for feature in gene_features:\n",
    "            cl_features.append(torch.tensor(feature.loc[cl][gene_symbols].values, dtype=torch.float64))\n",
    "        features = torch.stack(cl_features).t()\n",
    "\n",
    "        # Generate the graph.\n",
    "        edge_index = torch.tensor(neighbor_gene_tuples, dtype=torch.long).t().contiguous()\n",
    "        G_cl = Data(x=features, edge_index=edge_index)\n",
    "\n",
    "        Gs[cl] = G_cl\n",
    "\n",
    "    return Gs\n",
    "\n",
    "cl_graphs = create_cell_line_gene_graphs(NODES_AS_SYMBOLS, \n",
    "                                         [gexpr, cnvg, cnvp, mut],\n",
    "                                         inter_cls,\n",
    "                                         neighbor_gene_tuples_undirected)\n",
    "\n",
    "# Showcase topology for some cell-line examples.\n",
    "for cl in gexpr.index[:5].tolist():\n",
    "    print(f\"Cell-line: {cl:8s}   Graph: {cl_graphs[cl]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle \n",
    "with open(PROCESSED_DATA_PATH + 'thresh_700_gene_graphs.pkl', 'wb') as f:\n",
    "    pickle.dump(cl_graphs, f, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "987\n",
      "809\n"
     ]
    }
   ],
   "source": [
    "print(len(drm[drm.DATASET=='GDSC1'].CELL_LINE_NAME.unique()))\n",
    "print(len(drm[drm.DATASET=='GDSC2'].CELL_LINE_NAME.unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature subset with only the genes over the chosen threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(870, 859)\n",
      "(696, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_INDEX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPAR2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PEX11A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAPKAPK2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GENE_SYMBOL  GENE_INDEX\n",
       "0       LPAR2           0\n",
       "1      PEX11A           1\n",
       "2    MAPKAPK2           2"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gexpr = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_gexpr.pkl')\n",
    "cnvg = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_cnvg.pkl')\n",
    "cnvp = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_cnvp.pkl')\n",
    "mut = pd.read_pickle(PROCESSED_DATA_PATH + 'sparse_mut.pkl')\n",
    "print(gexpr.shape)\n",
    "\n",
    "inter_genes_above_thresh = pd.read_csv(PROCESSED_DATA_PATH + 'thresh_700_inter_genes.csv')\n",
    "print(inter_genes_above_thresh.shape)\n",
    "inter_genes_above_thresh.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(870, 697)\n"
     ]
    }
   ],
   "source": [
    "gexpr2 = gexpr.loc[:, gexpr.columns.isin(['CELL_LINE_NAME'] + inter_genes_above_thresh.GENE_SYMBOL.values.tolist())]\n",
    "cnvg2 = cnvg.loc[:, cnvg.columns.isin(['CELL_LINE_NAME'] + inter_genes_above_thresh.GENE_SYMBOL.values.tolist())]\n",
    "cnvp2 = cnvp.loc[:, cnvp.columns.isin(['CELL_LINE_NAME'] + inter_genes_above_thresh.GENE_SYMBOL.values.tolist())]\n",
    "mut2 = mut.loc[:, mut.columns.isin(['CELL_LINE_NAME'] + inter_genes_above_thresh.GENE_SYMBOL.values.tolist())]\n",
    "assert gexpr2.shape == cnvg2.shape == cnvp2.shape == mut2.shape, \\\n",
    "    \"ERROR: The shapes of all feature dataframes are not equal.\"\n",
    "print(gexpr2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These dataframe have been sparsed by intersecting genes after chosing \n",
    "# only the genes with combined_score > 700.\n",
    "gexpr2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gexpr.pkl')\n",
    "cnvg2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_cnvg.pkl')\n",
    "cnvp2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_cnvp.pkl')\n",
    "mut2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_mut.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split by GDSC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "metadata": {},
   "outputs": [],
   "source": [
    "gexpr2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_gexpr.pkl')\n",
    "cnvg2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_cnvg.pkl')\n",
    "cnvp2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_cnvp.pkl')\n",
    "mut2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_mut.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "When choosing both GDSC databases the number of intersecting cell-lines is 870\n",
      "When choosing only GDSC1 database the number of intersecting cell-lines is 869\n",
      "When choosing both GDSC2 databases the number of intersecting cell-lines is 729\n"
     ]
    }
   ],
   "source": [
    "inter_cls_gdsc = set(drm.CELL_LINE_NAME.unique())\\\n",
    "    .intersection(set(gexpr2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvg2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvp2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(mut2.CELL_LINE_NAME.unique()))\n",
    "\n",
    "inter_cls_gdsc1 = set(drm[drm.DATASET=='GDSC1'].CELL_LINE_NAME.unique())\\\n",
    "    .intersection(set(gexpr2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvg2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvp2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(mut2.CELL_LINE_NAME.unique()))\n",
    "\n",
    "inter_cls_gdsc2 = set(drm[drm.DATASET=='GDSC2'].CELL_LINE_NAME.unique())\\\n",
    "    .intersection(set(gexpr2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvg2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(cnvp2.CELL_LINE_NAME.unique()))\\\n",
    "    .intersection(set(mut2.CELL_LINE_NAME.unique()))\n",
    "\n",
    "print(f\"When choosing both GDSC databases the number of intersecting cell-lines is {len(list(inter_cls_gdsc))}\\n\" + \\\n",
    "      f\"When choosing only GDSC1 database the number of intersecting cell-lines is {len(list(inter_cls_gdsc1))}\\n\" + \\\n",
    "      f\"When choosing both GDSC2 databases the number of intersecting cell-lines is {len(list(inter_cls_gdsc2))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame({'CELL_LINE_NAME': list(inter_cls_gdsc2)})\\\n",
    "    .to_csv(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_inter_cls.csv', header=True, index=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose only GDSC1.\n",
    "drmGDSC1 = drm[drm.DATASET=='GDSC1']\n",
    "drmGDSC1 = drmGDSC1[drmGDSC1.CELL_LINE_NAME.isin(inter_cls_gdsc1)]\n",
    "gexprGDSC1 = gexpr2[gexpr2.CELL_LINE_NAME.isin(inter_cls_gdsc1)]\n",
    "cnvgGDSC1 = cnvg2[cnvg2.CELL_LINE_NAME.isin(inter_cls_gdsc1)]\n",
    "cnvpGDSC1 = cnvp2[cnvp2.CELL_LINE_NAME.isin(inter_cls_gdsc1)]\n",
    "mutGDSC1 = mut2[mut2.CELL_LINE_NAME.isin(inter_cls_gdsc1)]\n",
    "\n",
    "# Choose only GDSC2.\n",
    "drmGDSC2 = drm[drm.DATASET=='GDSC2']\n",
    "drmGDSC2 = drmGDSC2[drmGDSC2.CELL_LINE_NAME.isin(inter_cls_gdsc2)]\n",
    "gexprGDSC2 = gexpr2[gexpr2.CELL_LINE_NAME.isin(inter_cls_gdsc2)]\n",
    "cnvgGDSC2 = cnvg2[cnvg2.CELL_LINE_NAME.isin(inter_cls_gdsc2)]\n",
    "cnvpGDSC2 = cnvp2[cnvp2.CELL_LINE_NAME.isin(inter_cls_gdsc2)]\n",
    "mutGDSC2 = mut2[mut2.CELL_LINE_NAME.isin(inter_cls_gdsc2)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GDSC2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### As graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparsed by combined score > 700 and intersecting cell-lines for only GDSC2\n",
    "gexprGDSC2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_gexpr.pkl')\n",
    "cnvgGDSC2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_cnvg.pkl')\n",
    "cnvpGDSC2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_cnvp.pkl')\n",
    "mutGDSC2.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_mut.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gexprGDSC2.set_index('CELL_LINE_NAME', inplace=True)\n",
    "cnvgGDSC2.set_index('CELL_LINE_NAME', inplace=True)\n",
    "cnvpGDSC2.set_index('CELL_LINE_NAME', inplace=True)\n",
    "mutGDSC2.set_index('CELL_LINE_NAME', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(696, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 729/729 [00:05<00:00, 129.02it/s]\n"
     ]
    }
   ],
   "source": [
    "inter_genes_above_thresh = pd.read_csv(PROCESSED_DATA_PATH + 'thresh_700_inter_genes.csv')\n",
    "print(inter_genes_above_thresh.shape)\n",
    "\n",
    "cl_graphs_GDSC2 = create_cell_line_gene_graphs(inter_genes_above_thresh.GENE_SYMBOL.values.tolist(), \n",
    "                                               [gexprGDSC2, cnvgGDSC2, cnvpGDSC2, mutGDSC2],\n",
    "                                               inter_cls_gdsc2,\n",
    "                                               neighbor_gene_tuples_undirected)                                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell-line: 22RV1      Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 23132-87   Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 42-MG-BA   Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 5637       Graph: Data(x=[696, 4], edge_index=[2, 7794])\n",
      "Cell-line: 639-V      Graph: Data(x=[696, 4], edge_index=[2, 7794])\n"
     ]
    }
   ],
   "source": [
    "# Showcase topology for some cell-line examples.\n",
    "for cl in gexprGDSC2.index[:5].tolist():\n",
    "    print(f\"Cell-line: {cl:8s}   Graph: {cl_graphs_GDSC2[cl]}\")     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle \n",
    "with open(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_gene_graphs.pkl', 'wb') as f:\n",
    "    pickle.dump(cl_graphs_GDSC2, f, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### As table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique gene symbols: (696, 2)\n",
      "Drug-response matrix shape: (91639, 9)\n",
      "Feature matrix shapes: (729, 697)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>LPAR2</th>\n",
       "      <th>PEX11A</th>\n",
       "      <th>MAPKAPK2</th>\n",
       "      <th>RSU1</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>PRKACA</th>\n",
       "      <th>BZW2</th>\n",
       "      <th>CD40</th>\n",
       "      <th>TLR4</th>\n",
       "      <th>...</th>\n",
       "      <th>SPP1</th>\n",
       "      <th>PTPRK</th>\n",
       "      <th>HTRA1</th>\n",
       "      <th>CXCR4</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>TBX2</th>\n",
       "      <th>RRP1B</th>\n",
       "      <th>VAPB</th>\n",
       "      <th>CDK5R1</th>\n",
       "      <th>STAT5B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>4.865896</td>\n",
       "      <td>4.895203</td>\n",
       "      <td>4.905969</td>\n",
       "      <td>8.326618</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>3.398944</td>\n",
       "      <td>8.663252</td>\n",
       "      <td>3.427235</td>\n",
       "      <td>3.046879</td>\n",
       "      <td>...</td>\n",
       "      <td>2.991190</td>\n",
       "      <td>5.850645</td>\n",
       "      <td>5.483757</td>\n",
       "      <td>3.233630</td>\n",
       "      <td>3.468800</td>\n",
       "      <td>3.235685</td>\n",
       "      <td>7.408098</td>\n",
       "      <td>5.510225</td>\n",
       "      <td>3.419874</td>\n",
       "      <td>4.033523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>6.936096</td>\n",
       "      <td>5.255831</td>\n",
       "      <td>4.253763</td>\n",
       "      <td>8.714088</td>\n",
       "      <td>9.923119</td>\n",
       "      <td>3.274924</td>\n",
       "      <td>8.563217</td>\n",
       "      <td>3.581605</td>\n",
       "      <td>3.082267</td>\n",
       "      <td>...</td>\n",
       "      <td>2.916594</td>\n",
       "      <td>6.893173</td>\n",
       "      <td>3.471884</td>\n",
       "      <td>2.983593</td>\n",
       "      <td>3.438102</td>\n",
       "      <td>3.156897</td>\n",
       "      <td>7.211438</td>\n",
       "      <td>5.939798</td>\n",
       "      <td>3.186034</td>\n",
       "      <td>4.070979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>5.878033</td>\n",
       "      <td>4.454860</td>\n",
       "      <td>5.047313</td>\n",
       "      <td>8.690356</td>\n",
       "      <td>7.778161</td>\n",
       "      <td>3.333005</td>\n",
       "      <td>9.657946</td>\n",
       "      <td>3.146350</td>\n",
       "      <td>3.248642</td>\n",
       "      <td>...</td>\n",
       "      <td>2.891011</td>\n",
       "      <td>6.799222</td>\n",
       "      <td>11.422810</td>\n",
       "      <td>2.930109</td>\n",
       "      <td>3.218940</td>\n",
       "      <td>3.079025</td>\n",
       "      <td>7.807404</td>\n",
       "      <td>5.546324</td>\n",
       "      <td>3.439521</td>\n",
       "      <td>4.514098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 697 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME     LPAR2    PEX11A  MAPKAPK2      RSU1       FIS1    PRKACA  \\\n",
       "0          22RV1  4.865896  4.895203  4.905969  8.326618  10.102478  3.398944   \n",
       "1       23132-87  6.936096  5.255831  4.253763  8.714088   9.923119  3.274924   \n",
       "2       42-MG-BA  5.878033  4.454860  5.047313  8.690356   7.778161  3.333005   \n",
       "\n",
       "       BZW2      CD40      TLR4  ...      SPP1     PTPRK      HTRA1     CXCR4  \\\n",
       "0  8.663252  3.427235  3.046879  ...  2.991190  5.850645   5.483757  3.233630   \n",
       "1  8.563217  3.581605  3.082267  ...  2.916594  6.893173   3.471884  2.983593   \n",
       "2  9.657946  3.146350  3.248642  ...  2.891011  6.799222  11.422810  2.930109   \n",
       "\n",
       "     NOTCH1      TBX2     RRP1B      VAPB    CDK5R1    STAT5B  \n",
       "0  3.468800  3.235685  7.408098  5.510225  3.419874  4.033523  \n",
       "1  3.438102  3.156897  7.211438  5.939798  3.186034  4.070979  \n",
       "2  3.218940  3.079025  7.807404  5.546324  3.439521  4.514098  \n",
       "\n",
       "[3 rows x 697 columns]"
      ]
     },
     "execution_count": 458,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drmGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'gdsc2_drm.pkl')\n",
    "gexprGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_gexpr.pkl')\n",
    "cnvgGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_cnvg.pkl')\n",
    "cnvpGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_cnvp.pkl')\n",
    "mutGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_mut.pkl')\n",
    "inter_genes = pd.read_csv(PROCESSED_DATA_PATH + 'thresh_700_inter_genes.csv')\n",
    "print(\"Number of unique gene symbols:\", inter_genes.shape)\n",
    "print(\"Drug-response matrix shape:\", drmGDSC2.shape)\n",
    "print(\"Feature matrix shapes:\", gexprGDSC2.shape)\n",
    "gexprGDSC2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(729, 2785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>LPAR2_gexpr</th>\n",
       "      <th>PEX11A_gexpr</th>\n",
       "      <th>MAPKAPK2_gexpr</th>\n",
       "      <th>RSU1_gexpr</th>\n",
       "      <th>FIS1_gexpr</th>\n",
       "      <th>PRKACA_gexpr</th>\n",
       "      <th>BZW2_gexpr</th>\n",
       "      <th>CD40_gexpr</th>\n",
       "      <th>TLR4_gexpr</th>\n",
       "      <th>...</th>\n",
       "      <th>SPP1_mut</th>\n",
       "      <th>PTPRK_mut</th>\n",
       "      <th>HTRA1_mut</th>\n",
       "      <th>CXCR4_mut</th>\n",
       "      <th>NOTCH1_mut</th>\n",
       "      <th>TBX2_mut</th>\n",
       "      <th>RRP1B_mut</th>\n",
       "      <th>VAPB_mut</th>\n",
       "      <th>CDK5R1_mut</th>\n",
       "      <th>STAT5B_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>639-V</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME  LPAR2_gexpr  PEX11A_gexpr  MAPKAPK2_gexpr  RSU1_gexpr  \\\n",
       "0          22RV1          0.0           0.0             1.0         0.0   \n",
       "1       23132-87         -1.0           0.0             0.0         0.0   \n",
       "2       42-MG-BA          0.0           1.0            -1.0         0.0   \n",
       "3           5637          0.0           0.0             0.0        -1.0   \n",
       "4          639-V          1.0          -1.0            -1.0         1.0   \n",
       "\n",
       "   FIS1_gexpr  PRKACA_gexpr  BZW2_gexpr  CD40_gexpr  TLR4_gexpr  ...  \\\n",
       "0         1.0           0.0         1.0         0.0         0.0  ...   \n",
       "1        -1.0          -1.0         0.0         1.0         0.0  ...   \n",
       "2        -1.0           0.0         1.0         1.0        -1.0  ...   \n",
       "3         1.0           0.0         1.0         1.0         1.0  ...   \n",
       "4         0.0           1.0         1.0         1.0        -1.0  ...   \n",
       "\n",
       "   SPP1_mut  PTPRK_mut  HTRA1_mut  CXCR4_mut  NOTCH1_mut  TBX2_mut  RRP1B_mut  \\\n",
       "0       0.0        1.0        0.0        0.0         1.0       0.0        1.0   \n",
       "1       0.0        1.0        0.0        0.0         1.0       1.0        0.0   \n",
       "2       0.0        0.0        0.0        0.0         0.0       0.0        0.0   \n",
       "3       0.0        1.0        0.0        0.0         0.0       0.0        0.0   \n",
       "4       0.0        1.0        0.0        0.0         1.0       1.0        1.0   \n",
       "\n",
       "   VAPB_mut  CDK5R1_mut  STAT5B_mut  \n",
       "0       0.0         0.0         1.0  \n",
       "1       0.0         0.0         0.0  \n",
       "2       0.0         0.0         0.0  \n",
       "3       0.0         0.0         1.0  \n",
       "4       1.0         0.0         1.0  \n",
       "\n",
       "[5 rows x 2785 columns]"
      ]
     },
     "execution_count": 459,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_gene_feature_matrix(features: List[pd.DataFrame],\n",
    "                               suffixes: List[str]):\n",
    "    merged = features[1]\n",
    "    for i, feature in enumerate(features[1:]):\n",
    "        left_suffix = suffixes[i] if i==0 else ''\n",
    "        merged = pd.merge(left=merged, right=feature,\n",
    "                          on=['CELL_LINE_NAME'],\n",
    "                          suffixes=[left_suffix, suffixes[i+1]])\n",
    "    assert merged.shape == (features[0].shape[0], 4*(features[0].shape[1]-1) + 1)\n",
    "    return merged\n",
    "\n",
    "merged = create_gene_feature_matrix([gexprGDSC2, cnvgGDSC2, cnvpGDSC2, mutGDSC2],\n",
    "                                    ['_gexpr', '_cnvg', '_cnvp', '_mut'])\n",
    "print(merged.shape)\n",
    "merged.head(5)                                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes: 696\n",
      "Number of cell-lines: 729\n",
      "(729, 2785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>LPAR2_gexpr</th>\n",
       "      <th>PEX11A_gexpr</th>\n",
       "      <th>MAPKAPK2_gexpr</th>\n",
       "      <th>RSU1_gexpr</th>\n",
       "      <th>FIS1_gexpr</th>\n",
       "      <th>PRKACA_gexpr</th>\n",
       "      <th>BZW2_gexpr</th>\n",
       "      <th>CD40_gexpr</th>\n",
       "      <th>TLR4_gexpr</th>\n",
       "      <th>...</th>\n",
       "      <th>SPP1_mut</th>\n",
       "      <th>PTPRK_mut</th>\n",
       "      <th>HTRA1_mut</th>\n",
       "      <th>CXCR4_mut</th>\n",
       "      <th>NOTCH1_mut</th>\n",
       "      <th>TBX2_mut</th>\n",
       "      <th>RRP1B_mut</th>\n",
       "      <th>VAPB_mut</th>\n",
       "      <th>CDK5R1_mut</th>\n",
       "      <th>STAT5B_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>4.865896</td>\n",
       "      <td>4.895203</td>\n",
       "      <td>4.905969</td>\n",
       "      <td>8.326618</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>3.398944</td>\n",
       "      <td>8.663252</td>\n",
       "      <td>3.427235</td>\n",
       "      <td>3.046879</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>6.936096</td>\n",
       "      <td>5.255831</td>\n",
       "      <td>4.253763</td>\n",
       "      <td>8.714088</td>\n",
       "      <td>9.923119</td>\n",
       "      <td>3.274924</td>\n",
       "      <td>8.563217</td>\n",
       "      <td>3.581605</td>\n",
       "      <td>3.082267</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>5.878033</td>\n",
       "      <td>4.454860</td>\n",
       "      <td>5.047313</td>\n",
       "      <td>8.690356</td>\n",
       "      <td>7.778161</td>\n",
       "      <td>3.333005</td>\n",
       "      <td>9.657946</td>\n",
       "      <td>3.146350</td>\n",
       "      <td>3.248642</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 2785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME  LPAR2_gexpr  PEX11A_gexpr  MAPKAPK2_gexpr  RSU1_gexpr  \\\n",
       "0          22RV1     4.865896      4.895203        4.905969    8.326618   \n",
       "1       23132-87     6.936096      5.255831        4.253763    8.714088   \n",
       "2       42-MG-BA     5.878033      4.454860        5.047313    8.690356   \n",
       "\n",
       "   FIS1_gexpr  PRKACA_gexpr  BZW2_gexpr  CD40_gexpr  TLR4_gexpr  ...  \\\n",
       "0   10.102478      3.398944    8.663252    3.427235    3.046879  ...   \n",
       "1    9.923119      3.274924    8.563217    3.581605    3.082267  ...   \n",
       "2    7.778161      3.333005    9.657946    3.146350    3.248642  ...   \n",
       "\n",
       "   SPP1_mut  PTPRK_mut  HTRA1_mut  CXCR4_mut  NOTCH1_mut  TBX2_mut  RRP1B_mut  \\\n",
       "0       0.0        1.0        0.0        0.0         1.0       0.0        1.0   \n",
       "1       0.0        1.0        0.0        0.0         1.0       1.0        0.0   \n",
       "2       0.0        0.0        0.0        0.0         0.0       0.0        0.0   \n",
       "\n",
       "   VAPB_mut  CDK5R1_mut  STAT5B_mut  \n",
       "0       0.0         0.0         1.0  \n",
       "1       0.0         0.0         0.0  \n",
       "2       0.0         0.0         0.0  \n",
       "\n",
       "[3 rows x 2785 columns]"
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.concat([gexprGDSC2.set_index('CELL_LINE_NAME'), \n",
    "                    cnvgGDSC2.set_index('CELL_LINE_NAME'),\n",
    "                    cnvpGDSC2.set_index('CELL_LINE_NAME'),\n",
    "                    mutGDSC2.set_index('CELL_LINE_NAME')], \n",
    "                   keys=('gexpr', 'cnvg', 'cnvp', 'mut'), \n",
    "                   axis=1)\n",
    "merged.columns = merged.columns.map(lambda x: f\"{x[1]}_{x[0]}\")\n",
    "assert len([col for col in merged.columns if '_gexpr' in col]) == \\\n",
    "       len([col for col in merged.columns if '_cnvg' in col]) == \\\n",
    "       len([col for col in merged.columns if '_cnvp' in col]) == \\\n",
    "       len([col for col in merged.columns if '_mut' in col]), \\\n",
    "    \"ERROR: Number of gene columns per feature is not equal.\"\n",
    "print(f\"Number of genes:\", len([col for col in merged.columns if '_gexpr' in col]))\n",
    "merged.reset_index(inplace=True)                   \n",
    "print(f\"Number of cell-lines:\", len(merged.CELL_LINE_NAME.unique()))\n",
    "print(merged.shape)\n",
    "merged.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.to_pickle(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_gene_mat.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: make the below working to later perform ablation studies by adding different features.\n",
    "# def create_gene_feature_matrix(features: List[pd.DataFrame],\n",
    "#                                suffixes: List[str]):\n",
    "#     merged = features[0]\n",
    "#     merged.set_index('CELL_LINE_NAME')\n",
    "#     if len(features) > 1:\n",
    "#         for i, feature in enumerate(features[1:]):\n",
    "#             merged = pd.concat([merged, \n",
    "#                                 feature.set_index('CELL_LINE_NAME')],\n",
    "#                                 keys=(suffixes[i], suffixes[i+1]), axis=1)\n",
    "#         merged.columns = merged.columns.map(lambda x: f\"{x[1]}_{x[0]}\")\n",
    "#     return merged        \n",
    "\n",
    "# merged = create_gene_feature_matrix([gexprGDSC2, cnvgGDSC2, cnvpGDSC2, mutGDSC2],\n",
    "#                                     ['_gexpr', '_cnvg', '_cnvp', '_mut'])\n",
    "# print(merged.shape)                            \n",
    "# merged.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_demorgen_fingerprints(\n",
    "    drugs: pd.Series,\n",
    "    radius: int = 2,\n",
    "    n_bits: int = 256,\n",
    "    path_drug_smiles: str = \"data/GDSC/GDSC_compounds_inchi_key_with_smiles.csv\",\n",
    "):\n",
    "    \"\"\"\n",
    "\n",
    "    :param drugs: pandas Series of GDSC drug names for which to query the fingerprints\n",
    "    :param radius: see rdkit documentation GetMorganFingerprintAsBitVect\n",
    "    :param n_bits: length of the fingerprint\n",
    "    :param path_drug_smiles: path to .csv file containing \"smiles\" and \"drug_name\" column\n",
    "    :return: List of fingerprints or nans, if fingerprint inaccessible\n",
    "    \"\"\"\n",
    "    from rdkit import Chem\n",
    "    from rdkit.Chem import AllChem\n",
    "\n",
    "    # load smiles data for the GDSC compounds\n",
    "    GDSC_compound_identification = pd.read_csv(path_drug_smiles, index_col=0)\n",
    "    GDSC_compound_identification = GDSC_compound_identification.drop_duplicates(\n",
    "        \"drug_name\"\n",
    "    )\n",
    "    GDSC_compound_identification = GDSC_compound_identification.loc[\n",
    "        GDSC_compound_identification.smiles != \"not_found\"\n",
    "    ]\n",
    "\n",
    "    # convert smiles to RDkit mols\n",
    "    compound_name_to_rdkit_molecule_map = dict()\n",
    "    for compound_name, compound_smiles in zip(\n",
    "        GDSC_compound_identification.drug_name, GDSC_compound_identification.smiles\n",
    "    ):\n",
    "        compound_name_to_rdkit_molecule_map[compound_name] = Chem.MolFromSmiles(\n",
    "            compound_smiles\n",
    "        )\n",
    "\n",
    "    # convert RDkit mols to demorgan fingerprints\n",
    "    compound_name_to_demorgan_fingerprint = dict()\n",
    "    for compound_name in GDSC_compound_identification.drug_name:\n",
    "        mol = compound_name_to_rdkit_molecule_map[compound_name]\n",
    "        if mol is not None:\n",
    "            fingerprint = AllChem.GetMorganFingerprintAsBitVect(\n",
    "                mol, radius=radius, nBits=n_bits\n",
    "            )\n",
    "            compound_name_to_demorgan_fingerprint[compound_name] = fingerprint\n",
    "        else:\n",
    "            compound_name_to_demorgan_fingerprint[compound_name] = None\n",
    "\n",
    "    # apply conversion to input drugs\n",
    "    fingerprints = []\n",
    "    for drug in drugs:\n",
    "        if drug in compound_name_to_demorgan_fingerprint:\n",
    "            fingerprints.append(np.array(compound_name_to_demorgan_fingerprint[drug]))\n",
    "        else:\n",
    "            fingerprints.append(None)\n",
    "\n",
    "    return fingerprints"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GDSC2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### As table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique cell-lines: 729\n",
      "Number of unique DRUG_NAME's: 192\n",
      "(122269, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>310905</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1143</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00866</td>\n",
       "      <td>0.978464</td>\n",
       "      <td>0.067577</td>\n",
       "      <td>1.683567</td>\n",
       "      <td>1.343315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310906</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1187</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00885</td>\n",
       "      <td>0.995100</td>\n",
       "      <td>0.045205</td>\n",
       "      <td>1.875399</td>\n",
       "      <td>1.736985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310907</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1395</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00884</td>\n",
       "      <td>0.867832</td>\n",
       "      <td>0.107282</td>\n",
       "      <td>-0.096212</td>\n",
       "      <td>-2.309078</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATASET CELL_LINE_NAME     DRUG_NAME  DRUG_ID SANGER_MODEL_ID  \\\n",
       "310905   GDSC2        HCC1143  Camptothecin     1003       SIDM00866   \n",
       "310906   GDSC2        HCC1187  Camptothecin     1003       SIDM00885   \n",
       "310907   GDSC2        HCC1395  Camptothecin     1003       SIDM00884   \n",
       "\n",
       "             AUC      RMSE   Z_SCORE   LN_IC50  \n",
       "310905  0.978464  0.067577  1.683567  1.343315  \n",
       "310906  0.995100  0.045205  1.875399  1.736985  \n",
       "310907  0.867832  0.107282 -0.096212 -2.309078  "
      ]
     },
     "execution_count": 349,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drm = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "cls = pd.read_csv(PROCESSED_DATA_PATH + 'thresh_700_gdsc2_inter_cls.csv')\n",
    "drmGDSC2 = drm[drm.DATASET=='GDSC2']\n",
    "drmGDSC2 = drmGDSC2[drmGDSC2.CELL_LINE_NAME.isin(cls.CELL_LINE_NAME.values.tolist())]\n",
    "print(\"Number of unique cell-lines:\", len(drmGDSC2.CELL_LINE_NAME.unique()))\n",
    "uniq_drug_names = list(drmGDSC2.DRUG_NAME.unique())\n",
    "print(\"Number of unique DRUG_NAME's:\", len(uniq_drug_names))\n",
    "print(drmGDSC2.shape)\n",
    "drmGDSC2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(425, 3)\n",
      "|    | drug_name         | inchi_key                   | smiles                                                     |\n",
      "|---:|:------------------|:----------------------------|:-----------------------------------------------------------|\n",
      "|  0 | (5Z)-7-Oxozeaenol | NEQZWEXWOFPKOT-BYRRXHGESA-N | C[C@H]1CC=CC(=O)[C@H]([C@H](CC=Cc2cc(cc(c2C(=O)O1)O)OC)O)O |\n",
      "|  1 | 5-Fluorouracil    | GHASVSINZRGABV-UHFFFAOYSA-N | c1c(c(nc(n1)O)O)F                                          |\n",
      "|  2 | A-443654          | YWTBGJGMTBHQTM-IBGZPJMESA-N | Cc1c2cc(ccc2n[nH]1)c1cc(cnc1)OC[C@H](Cc1c[nH]c2ccccc12)N   |\n"
     ]
    }
   ],
   "source": [
    "smiles = pd.read_csv(RAW_DATA_PATH + 'GDSC_compounds_inchi_key_with_smiles.csv', index_col=0)\n",
    "print(smiles.shape)\n",
    "print(smiles.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of drugs: 192\n",
      "Number of drugs with not None fingerprint: 155\n",
      "Number of drugs which have a None fingerprint: 37\n"
     ]
    }
   ],
   "source": [
    "N_BITS = 256\n",
    "\n",
    "# Returns a list of fingerprints or nans.\n",
    "fps = get_demorgen_fingerprints(drugs=uniq_drug_names,\n",
    "                                n_bits=N_BITS,\n",
    "                                path_drug_smiles=RAW_DATA_PATH + 'GDSC_compounds_inchi_key_with_smiles.csv')\n",
    "\n",
    "# Append the fingerprints to the corresponding drug names. \n",
    "drug_name_fps_full = {uniq_drug_name: fps[i] for i, uniq_drug_name in enumerate(uniq_drug_names)}\n",
    "print(\"Number of drugs:\", len(drug_name_fps_full.keys()))\n",
    "\n",
    "drug_name_fps = {drug_name: fp for drug_name, fp in drug_name_fps_full.items() if fp is not None}\n",
    "print(\"Number of drugs with not None fingerprint:\", len(drug_name_fps.keys()))\n",
    "\n",
    "non_drugs = list(set(drug_name_fps_full.keys()).difference(set(drug_name_fps.keys())))\n",
    "print(\"Number of drugs which have a None fingerprint:\", len(non_drugs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `DRUG_NAME`'s which have a `None` fingerprint need to be removed from the drug-response matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(122269, 9)\n",
      "(95626, 9)\n",
      "729\n",
      "729\n",
      "Number of DRUG_NAME's: 155\n",
      "Number of DRUG_ID's: 158\n"
     ]
    }
   ],
   "source": [
    "drmGDSC2_v2 = drmGDSC2[~drmGDSC2.DRUG_NAME.isin(non_drugs)]\n",
    "print(drmGDSC2.shape)\n",
    "print(drmGDSC2_v2.shape)\n",
    "assert len(drmGDSC2.CELL_LINE_NAME.unique()) == len(drmGDSC2_v2.CELL_LINE_NAME.unique()), \\\n",
    "    \"Some cell-line have been removed because they have only drugs which have a None fingerprint.\"\n",
    "print(len(drmGDSC2.CELL_LINE_NAME.unique()))\n",
    "print(len(drmGDSC2_v2.CELL_LINE_NAME.unique()))\n",
    "print(\"Number of DRUG_NAME's:\", len(drmGDSC2_v2.DRUG_NAME.unique()))\n",
    "print(\"Number of DRUG_ID's:\", len(drmGDSC2_v2.DRUG_ID.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DRUG_NAME's which have more then 1 DRUG_ID: 3\n",
      "['Docetaxel', 'Ulixertinib', 'Uprosertib']\n",
      "(95626, 9)\n",
      "(91639, 9)\n",
      "Number of unique cell-lines before: 729\n",
      "Number of unique cell-lines after: 729\n",
      "Number of unique drug names: 152\n",
      "Number of unique drug id: 152\n",
      "Number of drug name keys: 152\n"
     ]
    }
   ],
   "source": [
    "# Remove the DRUG_NAME's which have more then one corresponding DRUG_ID.\n",
    "non_uniqs = drmGDSC2_v2[['DRUG_ID', 'DRUG_NAME']]\\\n",
    "    .groupby(['DRUG_NAME']).nunique()\\\n",
    "    .sort_values(['DRUG_ID'], ascending=False)\\\n",
    "    .reset_index().rename(columns={'DRUG_ID': 'count'})\n",
    "non_uniqs = non_uniqs[non_uniqs['count'] > 1]\n",
    "print(\"Number of DRUG_NAME's which have more then 1 DRUG_ID:\", non_uniqs.shape[0])\n",
    "non_uniq_drug_names = non_uniqs.DRUG_NAME.tolist()\n",
    "print(non_uniq_drug_names)\n",
    "\n",
    "# Remove these drug names from the drug response matrix and the smiles matrix.\n",
    "drmGDSC2_v3 = drmGDSC2_v2[~drmGDSC2_v2.DRUG_NAME.isin(non_uniq_drug_names)]\n",
    "print(drmGDSC2_v2.shape)\n",
    "print(drmGDSC2_v3.shape)\n",
    "print(\"Number of unique cell-lines before:\", len(drmGDSC2_v2.CELL_LINE_NAME.unique()))\n",
    "print(\"Number of unique cell-lines after:\", len(drmGDSC2_v3.CELL_LINE_NAME.unique()))\n",
    "print(\"Number of unique drug names:\", len(drmGDSC2_v3.DRUG_NAME.unique()))\n",
    "print(\"Number of unique drug id:\", len(drmGDSC2_v3.DRUG_ID.unique()))\n",
    "\n",
    "drug_name_fps_v2 = drug_name_fps\n",
    "for drug_name in non_uniq_drug_names:\n",
    "    drug_name_fps_v2.pop(drug_name, None)\n",
    "print(\"Number of drug name keys:\", len(drug_name_fps_v2.keys()))\n",
    "\n",
    "assert len(drmGDSC2_v3.DRUG_NAME.unique()) == len(drmGDSC2_v3.DRUG_ID.unique()) == len(drug_name_fps_v2.keys()), \\\n",
    "    \"ERROR: There is some mismatch in the DRUG_NAME's and DRUG_ID's between the drug response matrix and drug smiles dictionary.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The reason why the number of unique `CELL_LINE_NAME`'s is still the same after we removed all rows with a `DRUG_NAME` for which the fingprint is `None` is that the cell-lines have multiple drug's and it is very unlike that the 37 drugs we removed are the only drugs for a specific cell-line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(152, 257)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>246</th>\n",
       "      <th>247</th>\n",
       "      <th>248</th>\n",
       "      <th>249</th>\n",
       "      <th>250</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vinblastine</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cytarabine</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 257 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      DRUG_NAME  0  1  2  3  4  5  6  7  8  ...  246  247  248  249  250  251  \\\n",
       "0  Camptothecin  0  0  0  0  0  0  0  0  0  ...    0    0    1    0    1    0   \n",
       "1   Vinblastine  1  0  0  0  0  0  0  0  1  ...    0    1    0    1    0    1   \n",
       "2    Cytarabine  0  0  0  0  0  0  0  0  0  ...    0    0    0    0    0    1   \n",
       "\n",
       "   252  253  254  255  \n",
       "0    0    0    0    0  \n",
       "1    0    0    0    0  \n",
       "2    0    0    0    0  \n",
       "\n",
       "[3 rows x 257 columns]"
      ]
     },
     "execution_count": 479,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_name_fps_df = pd.DataFrame\\\n",
    "    .from_dict(drug_name_fps_v2, orient='index')\\\n",
    "    .rename_axis('DRUG_NAME').reset_index()\n",
    "print(drug_name_fps_df.shape)\n",
    "drug_name_fps_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(152, 258)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>...</th>\n",
       "      <th>246</th>\n",
       "      <th>247</th>\n",
       "      <th>248</th>\n",
       "      <th>249</th>\n",
       "      <th>250</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vinblastine</td>\n",
       "      <td>1004</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cytarabine</td>\n",
       "      <td>1006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 258 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      DRUG_NAME  DRUG_ID  0  1  2  3  4  5  6  7  ...  246  247  248  249  \\\n",
       "0  Camptothecin     1003  0  0  0  0  0  0  0  0  ...    0    0    1    0   \n",
       "1   Vinblastine     1004  1  0  0  0  0  0  0  0  ...    0    1    0    1   \n",
       "2    Cytarabine     1006  0  0  0  0  0  0  0  0  ...    0    0    0    0   \n",
       "\n",
       "   250  251  252  253  254  255  \n",
       "0    1    0    0    0    0    0  \n",
       "1    0    1    0    0    0    0  \n",
       "2    0    1    0    0    0    0  \n",
       "\n",
       "[3 rows x 258 columns]"
      ]
     },
     "execution_count": 480,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_id_fps_df = pd.merge(left=drug_name_fps_df, \n",
    "                     right=drmGDSC2[['DRUG_NAME', 'DRUG_ID']], \n",
    "                     how='left', on=['DRUG_NAME']).drop_duplicates()\n",
    "drug_id_fps_df.reset_index(inplace=True, drop=True)\n",
    "drug_id_fps_df.insert(1, 'DRUG_ID', drug_id_fps_df.pop('DRUG_ID'))\n",
    "print(drug_id_fps_df.shape)\n",
    "drug_id_fps_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>246</th>\n",
       "      <th>247</th>\n",
       "      <th>248</th>\n",
       "      <th>249</th>\n",
       "      <th>250</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1004</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 257 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   DRUG_ID  0  1  2  3  4  5  6  7  8  ...  246  247  248  249  250  251  252  \\\n",
       "0     1003  0  0  0  0  0  0  0  0  0  ...    0    0    1    0    1    0    0   \n",
       "1     1004  1  0  0  0  0  0  0  0  1  ...    0    1    0    1    0    1    0   \n",
       "2     1006  0  0  0  0  0  0  0  0  0  ...    0    0    0    0    0    1    0   \n",
       "\n",
       "   253  254  255  \n",
       "0    0    0    0  \n",
       "1    0    0    0  \n",
       "2    0    0    0  \n",
       "\n",
       "[3 rows x 257 columns]"
      ]
     },
     "execution_count": 486,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_id_fps_df = drug_id_fps_df.loc[:, ~drug_id_fps_df.columns.isin(['DRUG_NAME'])]\n",
    "drug_id_fps_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_id_fps_dict = drug_id_fps_df.loc[:, ~drug_id_fps_df.columns.isin(['DRUG_NAME'])]\\\n",
    "    .set_index('DRUG_ID').T.to_dict('list')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DRUG_NAME - fingerprint dictionary to a file.\n",
    "with open(PROCESSED_DATA_PATH + 'gdsc2_smiles_dict.pkl', 'wb') as f:\n",
    "    pickle.dump(drug_id_fps_dict, f) # As dictionary.\n",
    "drug_id_fps_df.to_pickle(PROCESSED_DATA_PATH + 'gdsc2_smiles_mat.pkl') # As matrix.\n",
    "\n",
    "# Save the new drug response matrix with only DRUG_ID's which have a fingerprint.\n",
    "drmGDSC2_v3.to_pickle(PROCESSED_DATA_PATH + 'gdsc2_drm.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### As graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(91639, 9)\n",
      "(152, 257)\n"
     ]
    }
   ],
   "source": [
    "drmGDSC2 = pd.read_pickle(PROCESSED_DATA_PATH + 'gdsc2_drm.pkl')\n",
    "drug_name_fps_df = pd.read_pickle(PROCESSED_DATA_PATH + 'gdsc2_smiles_mat.pkl')\n",
    "print(drmGDSC2.shape)\n",
    "print(drug_name_fps_df.shape)\n",
    "\n",
    "assert drug_name_fps_df.shape[0] == len(drmGDSC2.DRUG_NAME.unique()) == len(drmGDSC2.DRUG_ID.unique()),\\\n",
    "    \"ERROR: mismatch in the number of unique DRUG_NAME's.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(152, 2)\n",
      "(152, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SMILES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>CC[C@@]1(c2cc3c4c(cc5ccccc5n4)Cn3c(=O)c2COC1=O)O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vinblastine</td>\n",
       "      <td>1004</td>\n",
       "      <td>CC[C@@]1(C[C@@H]2C[C@](c3cc4c(cc3OC)N(C)[C@@H]...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cytarabine</td>\n",
       "      <td>1006</td>\n",
       "      <td>c1cn([C@H]2[C@H]([C@@H]([C@@H](CO)O2)O)O)c(nc1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      DRUG_NAME  DRUG_ID                                             SMILES\n",
       "0  Camptothecin     1003   CC[C@@]1(c2cc3c4c(cc5ccccc5n4)Cn3c(=O)c2COC1=O)O\n",
       "1   Vinblastine     1004  CC[C@@]1(C[C@@H]2C[C@](c3cc4c(cc3OC)N(C)[C@@H]...\n",
       "2    Cytarabine     1006  c1cn([C@H]2[C@H]([C@@H]([C@@H](CO)O2)O)O)c(nc1..."
      ]
     },
     "execution_count": 472,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the fingerprints to get also the DRUG_ID's.\n",
    "smiles = pd.merge(left=drug_name_fps_df['DRUG_NAME'], \n",
    "                     right=drmGDSC2[['DRUG_NAME', 'DRUG_ID']], \n",
    "                     how='left', on=['DRUG_NAME']).drop_duplicates()\n",
    "smiles.reset_index(inplace=True, drop=True)\n",
    "print(smiles.shape)\n",
    "# Root smiles dataframe with the SMILES string.\n",
    "smiles_root = pd.read_csv(RAW_DATA_PATH + 'GDSC_compounds_inchi_key_with_smiles.csv', index_col=0)\n",
    "\n",
    "# Join the SMILES string onto the subset.\n",
    "smiles2 = pd.merge(left=smiles,\n",
    "                   right=smiles_root[['drug_name', 'smiles']],\n",
    "                   left_on=['DRUG_NAME'],\n",
    "                   right_on=['drug_name'],\n",
    "                   how='left')[['DRUG_NAME', 'DRUG_ID', 'smiles']]\\\n",
    "            .rename(columns={'smiles': 'SMILES'})\\\n",
    "            .drop_duplicates()\n",
    "smiles2 = smiles2[smiles2.SMILES != 'not_found']\n",
    "print(smiles2.shape)\n",
    "smiles2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [],
   "source": [
    "smiles2[['DRUG_NAME', 'DRUG_ID']].to_csv(PROCESSED_DATA_PATH + 'gdsc2_drug_name_id_map.csv', \n",
    "                                         header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of keys/drugs : 152\n",
      "drug_id:  1003 | drug_name: Camptothecin    | graph: Data(x=[26, 9], edge_index=[2, 60], edge_attr=[60, 3], smiles='CC[C@@]1(c2cc3c4c(cc5ccccc5n4)Cn3c(=O)c2COC1=O)O')\n",
      "drug_id:  1004 | drug_name: Vinblastine     | graph: Data(x=[59, 9], edge_index=[2, 134], edge_attr=[134, 3], smiles='CC[C@@]1(C[C@@H]2C[C@](c3cc4c(cc3OC)N(C)[C@@H]3[C@@]54CCN4CC=C[C@](CC)([C@@H]54)[C@H]([C@@]3(C(=O)OC)O)OC(=O)C)(c3c(CCN(C2)C1)c1ccccc1[nH]3)C(=O)OC)O')\n",
      "drug_id:  1006 | drug_name: Cytarabine      | graph: Data(x=[17, 9], edge_index=[2, 36], edge_attr=[36, 3], smiles='c1cn([C@H]2[C@H]([C@@H]([C@@H](CO)O2)O)O)c(nc1=N)O')\n",
      "drug_id:  1010 | drug_name: Gefitinib       | graph: Data(x=[31, 9], edge_index=[2, 68], edge_attr=[68, 3], smiles='COc1cc2c(cc1OCCCN1CCOCC1)c(ncn2)Nc1ccc(c(c1)Cl)F')\n",
      "drug_id:  1011 | drug_name: Navitoclax      | graph: Data(x=[65, 9], edge_index=[2, 142], edge_attr=[142, 3], smiles='CC1(C)CCC(=C(C1)CN1CCN(CC1)c1ccc(cc1)C(=O)NS(=O)(=O)c1ccc(c(c1)S(=O)(=O)C(F)(F)F)N[C@H](CCN1CCOCC1)CSc1ccccc1)c1ccc(cc1)Cl')\n"
     ]
    }
   ],
   "source": [
    "# ref: https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/smiles.html \n",
    "from torch_geometric.utils import from_smiles \n",
    "\n",
    "# Create dictionary with DRUG_ID as key and smiles molecular graph as value.\n",
    "smiles_graphs = {}\n",
    "for i in range(smiles2.shape[0]):\n",
    "    drug_name, drug_id, smiles = smiles2.iloc[i]\n",
    "    smiles_graphs[drug_id] = from_smiles(smiles)\n",
    "\n",
    "print(f'Number of keys/drugs : {len(smiles_graphs.keys())}')\n",
    "# Print some examples.\n",
    "for i in range(5):\n",
    "    print(f'drug_id: {smiles2.iloc[i].DRUG_ID:5.0f} | drug_name: {smiles2.iloc[i].DRUG_NAME:15s} | graph: {smiles_graphs[smiles2.iloc[i].DRUG_ID]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(PROCESSED_DATA_PATH + 'gdsc2_smiles_graphs.pkl', 'wb') as f:\n",
    "    pickle.dump(smiles_graphs, f, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO \n",
    "\n",
    "- [x] create full feature and drug-response datasets, with only landmark genes\n",
    "  - [x] drug-response matrix\n",
    "  - [x] gene expression\n",
    "  - [x] copy number variations\n",
    "  - [x] mutations\n",
    "- [x] create sparsed feature and drug-response datasets, with only intersecting genes\n",
    "  - [x] drug-response matrix\n",
    "  - [x] gene expression\n",
    "  - [x] copy number variations\n",
    "  - [x] mutations\n",
    "- [ ] create smiles drug dataset\n",
    "  - [x] for GDSC 2\n",
    "  - [ ] for GDSC 1 \n",
    "  - [ ] for both \n",
    "- [ ] create the graphs for genes\n",
    "  - [x] for GDSC 2\n",
    "  - [ ] for GDSC 1 \n",
    "  - [ ] for both  \n",
    "- [ ] create the graphs for drugs\n",
    "  - [x] for GDSC 2\n",
    "  - [ ] for GDSC 1 \n",
    "  - [ ] for both \n",
    "- [ ] create tabular matrix for the drugs\n",
    "  - [x] for GDSC 2\n",
    "  - [ ] for GDSC 1\n",
    "  - [ ] for both \n",
    "- [x] choose threshold and create final datasets for v1 (this will contain only a subset of the genes)\n",
    "  - [x] create the newly sparsed gene feature datasets with only these genes\n",
    "- [x] Write generalized methods for all datasets and integrate in separate `.py` scripts\n",
    "- [ ] Later on: separate table/dataset with drug concentration for each row in the final drug-response matrix. (use raw GDSC dataset for that)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Model Input Datasets\n",
    "\n",
    "In the previous section we created the processed full datasets. In this section we are going to create the data used for the model inputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>310905</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1143</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00866</td>\n",
       "      <td>0.978464</td>\n",
       "      <td>0.067577</td>\n",
       "      <td>1.683567</td>\n",
       "      <td>1.343315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310906</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1187</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00885</td>\n",
       "      <td>0.995100</td>\n",
       "      <td>0.045205</td>\n",
       "      <td>1.875399</td>\n",
       "      <td>1.736985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310907</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>HCC1395</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>1003</td>\n",
       "      <td>SIDM00884</td>\n",
       "      <td>0.867832</td>\n",
       "      <td>0.107282</td>\n",
       "      <td>-0.096212</td>\n",
       "      <td>-2.309078</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATASET CELL_LINE_NAME     DRUG_NAME  DRUG_ID SANGER_MODEL_ID  \\\n",
       "310905   GDSC2        HCC1143  Camptothecin     1003       SIDM00866   \n",
       "310906   GDSC2        HCC1187  Camptothecin     1003       SIDM00885   \n",
       "310907   GDSC2        HCC1395  Camptothecin     1003       SIDM00884   \n",
       "\n",
       "             AUC      RMSE   Z_SCORE   LN_IC50  \n",
       "310905  0.978464  0.067577  1.683567  1.343315  \n",
       "310906  0.995100  0.045205  1.875399  1.736985  \n",
       "310907  0.867832  0.107282 -0.096212 -2.309078  "
      ]
     },
     "execution_count": 471,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(PROCESSED_DATA_PATH + 'gdsc2_drm.pkl', 'rb') as f: \n",
    "    drm = pickle.load(f)\n",
    "drm.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c78b81650a0bd32063743affb6953ff71b1a0dba806fbca9e2db842718495748"
  },
  "kernelspec": {
   "display_name": "conda_amazonei_mxnet_p27",
   "language": "python",
   "name": "conda_amazonei_mxnet_p27"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
