{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from glob import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Load blast hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Load blast hits\n",
    "blastp_hits = pd.read_csv(\"2_blastp_hits.csv\")\n",
    "blastp_hits.head()\n",
    "#Filter out Metahit 2010 hits, keep only Metahit 2014\n",
    "blastp_hits = blastp_hits[blastp_hits.db != \"metahit_pep\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Process blastp results\n",
    "\n",
    "## 2.1 Extract ORF stats from fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(52, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>q_id</th>\n",
       "      <th>orf_len</th>\n",
       "      <th>q_cds_start</th>\n",
       "      <th>q_cds_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d9539_asm_v1.2_1</td>\n",
       "      <td>27.0</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d9539_asm_v1.2_2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>98</td>\n",
       "      <td>127</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d9539_asm_v1.2_3</td>\n",
       "      <td>36.0</td>\n",
       "      <td>130</td>\n",
       "      <td>237</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d9539_asm_v1.2_4</td>\n",
       "      <td>52.0</td>\n",
       "      <td>105</td>\n",
       "      <td>260</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               q_id  orf_len  q_cds_start  q_cds_end strand\n",
       "0  d9539_asm_v1.2_1     27.0           25        105      +\n",
       "1  d9539_asm_v1.2_2     10.0           98        127      +\n",
       "2  d9539_asm_v1.2_3     36.0          130        237      +\n",
       "3  d9539_asm_v1.2_4     52.0          105        260      +\n",
       "4  d9539_asm_v1.2_5     58.0          215        388      +"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Assumes the Fasta file comes with the header format of EMBOSS getorf\n",
    "fh = open(\"1_orf/d9539_asm_v1.2_orf.fa\")\n",
    "header_regex = re.compile(r\">([^ ]+?) \\[([0-9]+) - ([0-9]+)\\]\")\n",
    "orf_stats = []\n",
    "for line in fh:\n",
    "    header_match = header_regex.match(line)\n",
    "    if header_match:\n",
    "        is_reverse = line.rstrip(\" \\n\").endswith(\"(REVERSE SENSE)\")\n",
    "        q_id = header_match.group(1)\n",
    "        #Position in contig\n",
    "        q_cds_start = int(header_match.group(2) if not is_reverse else header_match.group(3))\n",
    "        q_cds_end = int(header_match.group(3) if not is_reverse else header_match.group(2))\n",
    "        #Length of orf in aminoacids\n",
    "        q_len = (q_cds_end - q_cds_start + 1) / 3\n",
    "        orf_stats.append( pd.Series(data=[q_id,q_len,q_cds_start,q_cds_end,(\"-\" if is_reverse else \"+\")],\n",
    "                                     index=[\"q_id\",\"orf_len\",\"q_cds_start\",\"q_cds_end\",\"strand\"]))\n",
    "        \n",
    "orf_stats_df = pd.DataFrame(orf_stats)\n",
    "print(orf_stats_df.shape)\n",
    "orf_stats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Write orf stats to fasta\n",
    "orf_stats_df.to_csv(\"1_orf/orf_stats.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Annotate blast hits with orf stats "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>pct_id</th>\n",
       "      <th>ali_len</th>\n",
       "      <th>mism</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>...</th>\n",
       "      <th>q_cov</th>\n",
       "      <th>s_description</th>\n",
       "      <th>tool</th>\n",
       "      <th>db</th>\n",
       "      <th>q_id</th>\n",
       "      <th>orf_len</th>\n",
       "      <th>q_cds_start</th>\n",
       "      <th>q_cds_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>q_cov_calc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>MH0413_GL0135536</td>\n",
       "      <td>93.354</td>\n",
       "      <td>647</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0413_GL0135536  [gene]  locus=scaffold143833...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>647.0</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>MH0039_GL0018307</td>\n",
       "      <td>93.045</td>\n",
       "      <td>647</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0039_GL0018307  [gene]  locus=scaffold66015_...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>647.0</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>MH0409_GL0129297</td>\n",
       "      <td>91.963</td>\n",
       "      <td>647</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0409_GL0129297  [gene]  locus=scaffold117357...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>647.0</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>MH0033_GL0005946</td>\n",
       "      <td>89.352</td>\n",
       "      <td>648</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>648</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0033_GL0005946  [gene]  locus=scaffold47970_...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>647.0</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>MH0136_GL0000112</td>\n",
       "      <td>89.352</td>\n",
       "      <td>648</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>648</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0136_GL0000112  [gene]  locus=C1259564_1:206...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>647.0</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              query_id        subject_id  pct_id  ali_len  mism  gap_open  \\\n",
       "157  d9539_asm_v1.2_29  MH0413_GL0135536  93.354      647    43         0   \n",
       "158  d9539_asm_v1.2_29  MH0039_GL0018307  93.045      647    45         0   \n",
       "159  d9539_asm_v1.2_29  MH0409_GL0129297  91.963      647    52         0   \n",
       "160  d9539_asm_v1.2_29  MH0033_GL0005946  89.352      648    68         1   \n",
       "161  d9539_asm_v1.2_29  MH0136_GL0000112  89.352      648    68         1   \n",
       "\n",
       "     q_start  q_end  s_start  s_end    ...      q_cov  \\\n",
       "157        1    647        1    647    ...        100   \n",
       "158        1    647        1    647    ...        100   \n",
       "159        1    647        1    647    ...        100   \n",
       "160        1    647        1    648    ...        100   \n",
       "161        1    647        1    648    ...        100   \n",
       "\n",
       "                                         s_description    tool  \\\n",
       "157  MH0413_GL0135536  [gene]  locus=scaffold143833...  blastp   \n",
       "158  MH0039_GL0018307  [gene]  locus=scaffold66015_...  blastp   \n",
       "159  MH0409_GL0129297  [gene]  locus=scaffold117357...  blastp   \n",
       "160  MH0033_GL0005946  [gene]  locus=scaffold47970_...  blastp   \n",
       "161  MH0136_GL0000112  [gene]  locus=C1259564_1:206...  blastp   \n",
       "\n",
       "                   db               q_id orf_len q_cds_start q_cds_end  \\\n",
       "157  metahit_2014_pep  d9539_asm_v1.2_29   647.0        3429      5369   \n",
       "158  metahit_2014_pep  d9539_asm_v1.2_29   647.0        3429      5369   \n",
       "159  metahit_2014_pep  d9539_asm_v1.2_29   647.0        3429      5369   \n",
       "160  metahit_2014_pep  d9539_asm_v1.2_29   647.0        3429      5369   \n",
       "161  metahit_2014_pep  d9539_asm_v1.2_29   647.0        3429      5369   \n",
       "\n",
       "     strand q_cov_calc  \n",
       "157       +      100.0  \n",
       "158       +      100.0  \n",
       "159       +      100.0  \n",
       "160       +      100.0  \n",
       "161       +      100.0  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blastp_hits_annot = blastp_hits.merge(orf_stats_df,left_on=\"query_id\",right_on=\"q_id\")\n",
    "#Add query coverage calculation\n",
    "blastp_hits_annot[\"q_cov_calc\"] = (blastp_hits_annot[\"q_end\"] - blastp_hits_annot[\"q_start\"] + 1 ) * 100 / blastp_hits_annot[\"q_len\"]\n",
    "blastp_hits_annot.sort_values(by=\"bitscore\",ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "assert blastp_hits_annot.shape[0] == blastp_hits.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Extract best hit for each ORF ( q_cov > 0.8 and pct_id > 40% and e-value < 1)\n",
    "\n",
    "Define these resulting 7 ORFs as the core ORFs for the d9539 assembly. \n",
    "\n",
    "The homology between the Metahit gene catalogue is very good, and considering the catalogue was curated \n",
    "on a big set of gut metagenomes, it is reasonable to assume that these putative proteins would come \n",
    "from our detected circular putative virus/phage genome\n",
    "\n",
    "Two extra notes:\n",
    "  * Additionally, considering only these 7 ORFs ,  almost the entire genomic region is covered, with very few non-coding regions, still consistent with the hypothesis of a small viral genome which should be mainly coding\n",
    "\n",
    "  * Also, even though the naive ORF finder detected putative ORFs in both positive and negative strands, the supported ORFs only occur in the positive strand. This could be an indication of a ssDNA or ssRNA virus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! mkdir -p 4_msa_prots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>db</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>pct_id</th>\n",
       "      <th>q_cov</th>\n",
       "      <th>q_len</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>e_value</th>\n",
       "      <th>strand</th>\n",
       "      <th>q_cds_start</th>\n",
       "      <th>q_cds_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>d9539_asm_v1.2_1</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0148_GL0013503</td>\n",
       "      <td>92.593</td>\n",
       "      <td>100</td>\n",
       "      <td>27</td>\n",
       "      <td>57.4</td>\n",
       "      <td>2.310000e-10</td>\n",
       "      <td>+</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>d9539_asm_v1.2_3</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0203_GL0004312</td>\n",
       "      <td>90.323</td>\n",
       "      <td>86</td>\n",
       "      <td>36</td>\n",
       "      <td>57.8</td>\n",
       "      <td>1.450000e-10</td>\n",
       "      <td>+</td>\n",
       "      <td>130</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0409_GL0061975</td>\n",
       "      <td>91.379</td>\n",
       "      <td>100</td>\n",
       "      <td>58</td>\n",
       "      <td>113.0</td>\n",
       "      <td>1.220000e-31</td>\n",
       "      <td>+</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>d9539_asm_v1.2_11</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>V1.FI30_GL0086214</td>\n",
       "      <td>91.449</td>\n",
       "      <td>100</td>\n",
       "      <td>421</td>\n",
       "      <td>801.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>393</td>\n",
       "      <td>1655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>d9539_asm_v1.2_14</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>O2.UC35-0_GL0103272</td>\n",
       "      <td>73.333</td>\n",
       "      <td>86</td>\n",
       "      <td>174</td>\n",
       "      <td>221.0</td>\n",
       "      <td>2.390000e-71</td>\n",
       "      <td>+</td>\n",
       "      <td>1603</td>\n",
       "      <td>2124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>d9539_asm_v1.2_16</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>DOM013_GL0003401</td>\n",
       "      <td>82.857</td>\n",
       "      <td>100</td>\n",
       "      <td>70</td>\n",
       "      <td>115.0</td>\n",
       "      <td>2.460000e-32</td>\n",
       "      <td>+</td>\n",
       "      <td>2179</td>\n",
       "      <td>2388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>d9539_asm_v1.2_22</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0398_GL0192275</td>\n",
       "      <td>90.831</td>\n",
       "      <td>97</td>\n",
       "      <td>361</td>\n",
       "      <td>658.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>2351</td>\n",
       "      <td>3433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>d9539_asm_v1.2_29</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0413_GL0135536</td>\n",
       "      <td>93.354</td>\n",
       "      <td>100</td>\n",
       "      <td>647</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>3429</td>\n",
       "      <td>5369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>d9539_asm_v1.2_30</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>MH0039_GL0018308</td>\n",
       "      <td>98.969</td>\n",
       "      <td>100</td>\n",
       "      <td>97</td>\n",
       "      <td>190.0</td>\n",
       "      <td>5.350000e-61</td>\n",
       "      <td>+</td>\n",
       "      <td>5374</td>\n",
       "      <td>5664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              query_id                db           subject_id  pct_id  q_cov  \\\n",
       "317   d9539_asm_v1.2_1  metahit_2014_pep     MH0148_GL0013503  92.593    100   \n",
       "147   d9539_asm_v1.2_3  metahit_2014_pep     MH0203_GL0004312  90.323     86   \n",
       "10    d9539_asm_v1.2_5  metahit_2014_pep     MH0409_GL0061975  91.379    100   \n",
       "282  d9539_asm_v1.2_11  metahit_2014_pep    V1.FI30_GL0086214  91.449    100   \n",
       "261  d9539_asm_v1.2_14  metahit_2014_pep  O2.UC35-0_GL0103272  73.333     86   \n",
       "247  d9539_asm_v1.2_16  metahit_2014_pep     DOM013_GL0003401  82.857    100   \n",
       "205  d9539_asm_v1.2_22  metahit_2014_pep     MH0398_GL0192275  90.831     97   \n",
       "157  d9539_asm_v1.2_29  metahit_2014_pep     MH0413_GL0135536  93.354    100   \n",
       "124  d9539_asm_v1.2_30  metahit_2014_pep     MH0039_GL0018308  98.969    100   \n",
       "\n",
       "     q_len  bitscore       e_value strand  q_cds_start  q_cds_end  \n",
       "317     27      57.4  2.310000e-10      +           25        105  \n",
       "147     36      57.8  1.450000e-10      +          130        237  \n",
       "10      58     113.0  1.220000e-31      +          215        388  \n",
       "282    421     801.0  0.000000e+00      +          393       1655  \n",
       "261    174     221.0  2.390000e-71      +         1603       2124  \n",
       "247     70     115.0  2.460000e-32      +         2179       2388  \n",
       "205    361     658.0  0.000000e+00      +         2351       3433  \n",
       "157    647    1255.0  0.000000e+00      +         3429       5369  \n",
       "124     97     190.0  5.350000e-61      +         5374       5664  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Get best hit (highest bitscore) for each ORF\n",
    "gb = blastp_hits_annot[ (blastp_hits_annot.q_cov > 80) & (blastp_hits_annot.pct_id > 40) & (blastp_hits_annot.e_value < 1) ].groupby(\"query_id\")\n",
    "reliable_orfs = pd.DataFrame( hits.ix[hits.bitscore.idxmax()] for q_id,hits in gb )[[\"query_id\",\"db\",\"subject_id\",\"pct_id\",\"q_cov\",\"q_len\",\n",
    "                                                                                      \"bitscore\",\"e_value\",\"strand\",\"q_cds_start\",\"q_cds_end\"]]\n",
    "reliable_orfs = reliable_orfs.sort_values(by=\"q_cds_start\",ascending=True)\n",
    "reliable_orfs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4 Extract selected orfs for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "reliable_orfs[\"orf_id\"] = [\"orf{}\".format(x) for x in range(1,reliable_orfs.shape[0]+1) ]\n",
    "reliable_orfs[\"cds_len\"] = reliable_orfs[\"q_cds_end\"] - reliable_orfs[\"q_cds_start\"] +1\n",
    "reliable_orfs.sort_values(by=\"q_cds_start\",ascending=True).to_csv(\"3_filtered_orfs/filt_orf_stats.csv\",index=False,header=True)\n",
    "reliable_orfs.sort_values(by=\"q_cds_start\",ascending=True).to_csv(\"3_filtered_orfs/filt_orf_list.txt\",index=False,header=False,columns=[\"query_id\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4.2 Extract fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "! ~/utils/bin/seqtk subseq 1_orf/d9539_asm_v1.2_orf.fa 3_filtered_orfs/filt_orf_list.txt > 3_filtered_orfs/d9539_asm_v1.2_orf_filt.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4.3 Write out filtered blast hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>pct_id</th>\n",
       "      <th>ali_len</th>\n",
       "      <th>mism</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>...</th>\n",
       "      <th>q_cov</th>\n",
       "      <th>s_description</th>\n",
       "      <th>tool</th>\n",
       "      <th>db</th>\n",
       "      <th>q_id</th>\n",
       "      <th>orf_len</th>\n",
       "      <th>q_cds_start</th>\n",
       "      <th>q_cds_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>q_cov_calc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>MH0409_GL0061975</td>\n",
       "      <td>91.379</td>\n",
       "      <td>58</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>29</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0409_GL0061975  [gene]  locus=scaffold117357...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>675950834-stool1_revised_C814893_1_gene71376</td>\n",
       "      <td>89.655</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>675950834-stool1_revised_C814893_1_gene71376 s...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>MH0105_GL0047475</td>\n",
       "      <td>89.655</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>29</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0105_GL0047475  [gene]  locus=scaffold1014_1...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>MH0203_GL0004312</td>\n",
       "      <td>89.655</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>29</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0203_GL0004312  [gene]  locus=C2075218_1:153...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>MH0033_GL0005948</td>\n",
       "      <td>89.655</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>29</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>MH0033_GL0005948  [gene]  locus=scaffold47970_...</td>\n",
       "      <td>blastp</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>d9539_asm_v1.2_5</td>\n",
       "      <td>58.0</td>\n",
       "      <td>215</td>\n",
       "      <td>388</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            query_id                                    subject_id  pct_id  \\\n",
       "10  d9539_asm_v1.2_5                              MH0409_GL0061975  91.379   \n",
       "11  d9539_asm_v1.2_5  675950834-stool1_revised_C814893_1_gene71376  89.655   \n",
       "12  d9539_asm_v1.2_5                              MH0105_GL0047475  89.655   \n",
       "13  d9539_asm_v1.2_5                              MH0203_GL0004312  89.655   \n",
       "14  d9539_asm_v1.2_5                              MH0033_GL0005948  89.655   \n",
       "\n",
       "    ali_len  mism  gap_open  q_start  q_end  s_start  s_end    ...      q_cov  \\\n",
       "10       58     5         0        1     58       29     86    ...        100   \n",
       "11       58     6         0        1     58        1     58    ...        100   \n",
       "12       58     6         0        1     58       29     86    ...        100   \n",
       "13       58     6         0        1     58       29     86    ...        100   \n",
       "14       58     6         0        1     58       29     86    ...        100   \n",
       "\n",
       "                                        s_description    tool  \\\n",
       "10  MH0409_GL0061975  [gene]  locus=scaffold117357...  blastp   \n",
       "11  675950834-stool1_revised_C814893_1_gene71376 s...  blastp   \n",
       "12  MH0105_GL0047475  [gene]  locus=scaffold1014_1...  blastp   \n",
       "13  MH0203_GL0004312  [gene]  locus=C2075218_1:153...  blastp   \n",
       "14  MH0033_GL0005948  [gene]  locus=scaffold47970_...  blastp   \n",
       "\n",
       "                  db              q_id orf_len q_cds_start q_cds_end  strand  \\\n",
       "10  metahit_2014_pep  d9539_asm_v1.2_5    58.0         215       388       +   \n",
       "11  metahit_2014_pep  d9539_asm_v1.2_5    58.0         215       388       +   \n",
       "12  metahit_2014_pep  d9539_asm_v1.2_5    58.0         215       388       +   \n",
       "13  metahit_2014_pep  d9539_asm_v1.2_5    58.0         215       388       +   \n",
       "14  metahit_2014_pep  d9539_asm_v1.2_5    58.0         215       388       +   \n",
       "\n",
       "   q_cov_calc  \n",
       "10      100.0  \n",
       "11      100.0  \n",
       "12      100.0  \n",
       "13      100.0  \n",
       "14      100.0  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_blastp_hits = blastp_hits_annot[ blastp_hits_annot.query_id.apply(lambda x: x in reliable_orfs.query_id.tolist())]\n",
    "filt_blastp_hits.to_csv(\"3_filtered_orfs/d9539_asm_v1.2_orf_filt_blastp.csv\")\n",
    "filt_blastp_hits.head()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gb = filt_blastp_hits.groupby(\"query_id\")\n",
    "for q_id,hits in gb:\n",
    "    for db in [\"metahit_pep\",\"hmp_pep\",\"env_nr\",\"nr\"]:\n",
    "        try:\n",
    "            os.mkdir(\"3_filtered_orfs/msa/{}\".format(q_id))\n",
    "        except:\n",
    "            pass\n",
    "        hits_to_write = hits[hits.db == db][\"subject_id\"].unique()\n",
    "        if hits_to_write.shape != (0,):\n",
    "            hits_to_write.tofile(\"3_filtered_orfs/msa/{0}/{0}_{1}.txt\".format(q_id,db),sep=\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
